
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>18. Compositional analysis &#8212; Single-cell best practices</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="../_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" href="../_static/styles/sphinx-book-theme.css?digest=5115cc725059bd94278eecd172e13a965bf8f5a9" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/book.css" />
    <link rel="stylesheet" type="text/css" href="../_static/design-style.b7bb847fb20b106c3d81b95245e65545.min.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?digest=9c920249402e914e316237a7dbc6769907cce411"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js"></script>
    <script async="async" kind="hypothesis" src="https://hypothes.is/embed.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../_static/book.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="19. Gene set enrichment and pathway analysis" href="gsea_pathway.html" />
    <link rel="prev" title="17. Differential gene expression analysis" href="differential_gene_expression.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="en">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../_static/logo.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">Single-cell best practices</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Introduction
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../introduction/prior_art.html">
   1. Prior art
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../introduction/scrna_seq.html">
   2. Single-cell RNA sequencing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../introduction/raw_data_processing.html">
   3. Raw data processing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../introduction/analysis_tools.html">
   4. Analysis frameworks and tools
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../introduction/data_infrastructure.html">
   5. Data infrastructure
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../introduction/interoperability.html">
   6. Interoperability
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Preprocessing and visualization
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../preprocessing_visualization/quality_control.html">
   7. Quality Control
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../preprocessing_visualization/normalization.html">
   8. Normalization
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../preprocessing_visualization/feature_selection.html">
   9. Feature selection
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../preprocessing_visualization/dimensionality_reduction.html">
   10. Dimensionality Reduction
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Identifying cellular structure
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../cellular_structure/clustering.html">
   11. Clustering
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cellular_structure/annotation.html">
   12. Annotation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cellular_structure/integration.html">
   13. Data integration
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Inferring trajectories
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../trajectories/pseudotemporal.html">
   14. Pseudotemporal ordering
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../trajectories/rna_velocity.html">
   15. RNA velocity
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../trajectories/lineage_tracing.html">
   16. Lineage tracing
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Dealing with conditions
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="differential_gene_expression.html">
   17. Differential gene expression analysis
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   18. Compositional analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="gsea_pathway.html">
   19. Gene set enrichment and pathway analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="perturbation_modeling.html">
   20. Perturbation modeling
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Modeling mechanisms
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../mechanisms/grns.html">
   21. Gene regulatory networks
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../mechanisms/cell_cell_communication.html">
   22. Cell-cell communication
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Deconvolution
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../deconvolution/bulk_deconvolution.html">
   23. Bulk deconvolution
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Surface protein
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/quality_control.html">
   24. Quality control
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/normalization.html">
   25. Normalization
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/doublet_detection.html">
   26. Doublet detection
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/dimensionality_reduction.html">
   27. Dimensionality Reduction
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/batch_correction.html">
   28. Batch correction
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/annotation.html">
   29. Annotation
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  AIR repertoire
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../air_repertoire/ir_profiling.html">
   30. Immune Receptor Profiling
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../air_repertoire/clonotype.html">
   31. Clonotype analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../air_repertoire/specificity.html">
   35. Specificity analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../air_repertoire/multimodal_integration.html">
   36. Integrating AIR and transcriptomics
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multimodal integration
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../multimodal_integration/paired_integration.html">
   37. Paired integration
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../multimodal_integration/advanced_integration.html">
   38. Advanced integration
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Reproducibility
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../reproducibility/introduction.html">
   39. Reproducibility
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Outlook
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../outlook.html">
   40. Outlook
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Acknowledgements
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../acknowledgements.html">
   41. Acknowledgements
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Glossary
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../glossary.html">
   42. Glossary
  </a>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
            <div class="navbar_extra_footer">
            Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
            </div>
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right">
<div class="menu-dropdown menu-dropdown-launch-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Launch interactive content">
      <i class="fas fa-rocket"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://mybinder.org/v2/gh/theislab/single-cell-best-practices/master?urlpath=tree/conditions/compositional.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Launch on Binder"
>
  

<span class="headerbtn__icon-container">
  
    <img src="../_static/images/logo_binder.svg">
  </span>
<span class="headerbtn__text-container">Binder</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

<button onclick="toggleFullScreen()"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="bottom"
title="Fullscreen mode"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>

<div class="menu-dropdown menu-dropdown-repository-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Source repositories">
      <i class="fab fa-github"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://github.com/theislab/single-cell-best-practices"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Source repository"
>
  

<span class="headerbtn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="headerbtn__text-container">repository</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/theislab/single-cell-best-practices/issues/new?title=Issue%20on%20page%20%2Fconditions/compositional.html&body=Your%20issue%20content%20here."
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Open an issue"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="headerbtn__text-container">open issue</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/theislab/single-cell-best-practices/edit/master/conditions/compositional.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Edit this page"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="headerbtn__text-container">suggest edit</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

<div class="menu-dropdown menu-dropdown-download-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Download this page">
      <i class="fas fa-download"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="../_sources/conditions/compositional.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Download source file"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="headerbtn__text-container">.ipynb</span>
</a>

      </li>
      
      <li>
        
<button onclick="printPdf(this)"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="left"
title="Print to PDF"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="headerbtn__text-container">.pdf</span>
</button>

      </li>
      
    </ul>
  </div>
</div>
<label for="__page-toc"
  class="headerbtn headerbtn-page-toc"
  
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-list"></i>
  </span>

</label>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
    <div class="tocsection onthispage pt-5 pb-3">
        <i class="fas fa-list"></i> Contents
    </div>
    <nav id="bd-toc-nav" aria-label="Page">
        <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#motivation">
   18.1. Motivation
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#data-loading">
   18.2. Data loading
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#why-cell-type-count-data-is-compositional">
   18.3. Why cell-type count data is compositional
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#with-labeled-clusters">
   18.4. With labeled clusters
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#with-labeled-clusters-and-hierarchical-structure">
   18.5. With labeled clusters and hierarchical structure
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#without-labeled-clusters">
   18.6. Without labeled clusters
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#define-neighbourhoods">
     18.6.1. Define neighbourhoods
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#count-cells-in-neighbourhoods">
     18.6.2. Count cells in neighbourhoods
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#key-takeaways">
   18.7. Key Takeaways
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#quiz">
   18.8. Quiz
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#references">
   18.9. References
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#contributors">
   18.10. Contributors
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#authors">
     18.10.1. Authors
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#reviewers">
     18.10.2. Reviewers
    </a>
   </li>
  </ul>
 </li>
</ul>

    </nav>
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Compositional analysis</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#motivation">
   18.1. Motivation
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#data-loading">
   18.2. Data loading
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#why-cell-type-count-data-is-compositional">
   18.3. Why cell-type count data is compositional
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#with-labeled-clusters">
   18.4. With labeled clusters
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#with-labeled-clusters-and-hierarchical-structure">
   18.5. With labeled clusters and hierarchical structure
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#without-labeled-clusters">
   18.6. Without labeled clusters
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#define-neighbourhoods">
     18.6.1. Define neighbourhoods
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#count-cells-in-neighbourhoods">
     18.6.2. Count cells in neighbourhoods
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#key-takeaways">
   18.7. Key Takeaways
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#quiz">
   18.8. Quiz
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#references">
   18.9. References
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#contributors">
   18.10. Contributors
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#authors">
     18.10.1. Authors
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#reviewers">
     18.10.2. Reviewers
    </a>
   </li>
  </ul>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                
  <section class="tex2jax_ignore mathjax_ignore" id="compositional-analysis">
<h1><span class="section-number">18. </span>Compositional analysis<a class="headerlink" href="#compositional-analysis" title="Permalink to this headline">#</a></h1>
<section id="motivation">
<h2><span class="section-number">18.1. </span>Motivation<a class="headerlink" href="#motivation" title="Permalink to this headline">#</a></h2>
<p>Beyond changes in gene expression patterns, cell compositions, such as the proportions of cell-types, can change between conditions. A specific drug may, for example, induce a transdifferentiation of a cell type which will be reflected in the cell identity composition. Sufficient cell and sample numbers are required to accurately determine cell-identity cluster proportions and background variation. Compositional analysis can be done on the level of cell identity clusters in the form of known cell types or cell states corresponding to, for example, cells recently affected by perturbations.</p>
<figure class="align-default" id="compositional-analysis-overview">
<a class="bg-primary mb-1 reference internal image-reference" href="../_images/compositional.jpg"><img alt="Compositional analysis overview" class="bg-primary mb-1" src="../_images/compositional.jpg" style="width: 800px;" /></a>
<figcaption>
<p><span class="caption-number">Fig. 18.1 </span><span class="caption-text">Differential abundance analysis compares the composition of cell types between two conditions. The samples from both modalities contain different proportions of cell types, which can be tested for significant shifts in abundance.</span><a class="headerlink" href="#compositional-analysis-overview" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p>This chapter will introduce both approaches and apply them to the Haber dataset<span id="id1">[<a class="reference internal" href="#id313" title="Adam L. Haber, Moshe Biton, Noga Rogel, Rebecca H. Herbst, Karthik Shekhar, Christopher Smillie, Grace Burgin, Toni M. Delorey, Michael R. Howitt, Yarden Katz, Itay Tirosh, Semir Beyaz, Danielle Dionne, Mei Zhang, Raktima Raychowdhury, Wendy S. Garrett, Orit Rozenblatt-Rosen, Hai Ning Shi, Omer Yilmaz, Ramnik J. Xavier, and Aviv Regev. A single-cell survey of the small intestinal epithelium. Nature, 551(7680):333-339, Nov 2017. URL: https://doi.org/10.1038/nature24489, doi:10.1038/nature24489.">Haber <em>et al.</em>, 2017</a>]</span>. This dataset contains 53,193 individual epithelial cells from the small intestine and organoids of mice. Some of the cells were also subject to bacterial or helminth infection such as through Salmonella and Heligmosomoides polygyrus respectively.
Throughout this tutorial we are using a subset of the complete Haber dataset which only includes control and infected cells that were collected specifically for this purpose. Notably, we are excluding an additional dataset which collected only large cells for faster computation and reduced complexity.</p>
<p>As a first step, we load the dataset.</p>
</section>
<section id="data-loading">
<h2><span class="section-number">18.2. </span>Data loading<a class="headerlink" href="#data-loading" title="Permalink to this headline">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">warnings</span>

<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span><span class="s2">&quot;ignore&quot;</span><span class="p">)</span>
<span class="n">warnings</span><span class="o">.</span><span class="n">simplefilter</span><span class="p">(</span><span class="s2">&quot;ignore&quot;</span><span class="p">)</span>

<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">tensorflow</span> <span class="k">as</span> <span class="nn">tf</span>

<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">matplotlib</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
<span class="kn">import</span> <span class="nn">altair</span> <span class="k">as</span> <span class="nn">alt</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&quot;haber_count.h5ad&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 9842 × 15215
    obs: &#39;batch&#39;, &#39;barcode&#39;, &#39;condition&#39;, &#39;cell_label&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>batch</th>
      <th>barcode</th>
      <th>condition</th>
      <th>cell_label</th>
    </tr>
    <tr>
      <th>index</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>B1_AAACATACCACAAC_Control_Enterocyte.Progenitor</th>
      <td>B1</td>
      <td>AAACATACCACAAC</td>
      <td>Control</td>
      <td>Enterocyte.Progenitor</td>
    </tr>
    <tr>
      <th>B1_AAACGCACGAGGAC_Control_Stem</th>
      <td>B1</td>
      <td>AAACGCACGAGGAC</td>
      <td>Control</td>
      <td>Stem</td>
    </tr>
    <tr>
      <th>B1_AAACGCACTAGCCA_Control_Stem</th>
      <td>B1</td>
      <td>AAACGCACTAGCCA</td>
      <td>Control</td>
      <td>Stem</td>
    </tr>
    <tr>
      <th>B1_AAACGCACTGTCCC_Control_Stem</th>
      <td>B1</td>
      <td>AAACGCACTGTCCC</td>
      <td>Control</td>
      <td>Stem</td>
    </tr>
    <tr>
      <th>B1_AAACTTGACCACCT_Control_Enterocyte.Progenitor</th>
      <td>B1</td>
      <td>AAACTTGACCACCT</td>
      <td>Control</td>
      <td>Enterocyte.Progenitor</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>B10_TTTCACGACAAGCT_Salmonella_TA</th>
      <td>B10</td>
      <td>TTTCACGACAAGCT</td>
      <td>Salmonella</td>
      <td>TA</td>
    </tr>
    <tr>
      <th>B10_TTTCAGTGAGGCGA_Salmonella_Enterocyte</th>
      <td>B10</td>
      <td>TTTCAGTGAGGCGA</td>
      <td>Salmonella</td>
      <td>Enterocyte</td>
    </tr>
    <tr>
      <th>B10_TTTCAGTGCGACAT_Salmonella_Stem</th>
      <td>B10</td>
      <td>TTTCAGTGCGACAT</td>
      <td>Salmonella</td>
      <td>Stem</td>
    </tr>
    <tr>
      <th>B10_TTTCAGTGTGACCA_Salmonella_Endocrine</th>
      <td>B10</td>
      <td>TTTCAGTGTGACCA</td>
      <td>Salmonella</td>
      <td>Endocrine</td>
    </tr>
    <tr>
      <th>B10_TTTCAGTGTTCTCA_Salmonella_Enterocyte.Progenitor</th>
      <td>B10</td>
      <td>TTTCAGTGTTCTCA</td>
      <td>Salmonella</td>
      <td>Enterocyte.Progenitor</td>
    </tr>
  </tbody>
</table>
<p>9842 rows × 4 columns</p>
</div></div></div>
</div>
<p>The data was collected in 10 batches. Unique conditions are Control, Salmonella, Hpoly.Day3 and Hpoly.Day10 which correspond to the healthy control state, Salmonella infection, Heligmosomoides polygyrus infected cells after 3 days and Heligmosomoides polygyrus infected cells after 10 days. The <code class="docutils literal notranslate"><span class="pre">cell_label</span></code> corresponds to the cell types.</p>
</section>
<section id="why-cell-type-count-data-is-compositional">
<h2><span class="section-number">18.3. </span>Why cell-type count data is compositional<a class="headerlink" href="#why-cell-type-count-data-is-compositional" title="Permalink to this headline">#</a></h2>
<p>When analyzing the compositional shifts in cell count data, multiple technical and methodological limitations need to be accounted for. One challenge is the characteristically low number of experimental replicates, which leads to large confidence intervals when conducting differential abundance analysis with frequentist statistical tests.
Even more important, single-cell sequencing is naturally limited in the number of cells per sample - we can’t sequence every cell in a tissue or organ, but use a small, representative snapshot instead. This, however, forces us to view the cell type counts as purely proportional, i.e. the total number of cells in a sample is only a scaling factor. In the statistical literature, such data is known as compositional data<span id="id2">[<a class="reference internal" href="#id314" title="J. Aitchison. The statistical analysis of compositional data. Journal of the Royal Statistical Society: Series B (Methodological), 44(2):139-160, 1982. URL: https://rss.onlinelibrary.wiley.com/doi/abs/10.1111/j.2517-6161.1982.tb01195.x, arXiv:https://rss.onlinelibrary.wiley.com/doi/pdf/10.1111/j.2517-6161.1982.tb01195.x, doi:https://doi.org/10.1111/j.2517-6161.1982.tb01195.x.">Aitchison, 1982</a>]</span>, and characterized by the relative abundances of all features (cell types in our case) in one sample always adding up to one.</p>
<p>Because of this sum-to-one constraint, a negative correlation between the cell type abundances is induced. To illustrate this, let’s consider the following example:</p>
<p>In a case-control study, we want to compare the cell type composition of a healthy and a diseased organ. In both cases, we have three cell types (A, B and C), but their abundances differ:</p>
<ul class="simple">
<li><p>The healthy organ consists of 2,000 cells of each type (6,000 cells total).</p></li>
<li><p>The disease leads to a doubling of cell type A, while cell types B and C are not affected, so that the diseased organ has 8,000 cells.</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">healthy_tissue</span> <span class="o">=</span> <span class="p">[</span><span class="mi">2000</span><span class="p">,</span> <span class="mi">2000</span><span class="p">,</span> <span class="mi">2000</span><span class="p">]</span>
<span class="n">diseased_tissue</span> <span class="o">=</span> <span class="p">[</span><span class="mi">4000</span><span class="p">,</span> <span class="mi">2000</span><span class="p">,</span> <span class="mi">2000</span><span class="p">]</span>
<span class="n">example_data_global</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
    <span class="n">data</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">healthy_tissue</span><span class="p">,</span> <span class="n">diseased_tissue</span><span class="p">]),</span>
    <span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span>
    <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">],</span>
<span class="p">)</span>
<span class="n">example_data_global</span><span class="p">[</span><span class="s2">&quot;Disease status&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;Healthy&quot;</span><span class="p">,</span> <span class="s2">&quot;Diseased&quot;</span><span class="p">]</span>
<span class="n">example_data_global</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>A</th>
      <th>B</th>
      <th>C</th>
      <th>Disease status</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>1</th>
      <td>2000</td>
      <td>2000</td>
      <td>2000</td>
      <td>Healthy</td>
    </tr>
    <tr>
      <th>2</th>
      <td>4000</td>
      <td>2000</td>
      <td>2000</td>
      <td>Diseased</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">plot_data_global</span> <span class="o">=</span> <span class="n">example_data_global</span><span class="o">.</span><span class="n">melt</span><span class="p">(</span>
    <span class="s2">&quot;Disease status&quot;</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">],</span> <span class="s2">&quot;Cell type&quot;</span><span class="p">,</span> <span class="s2">&quot;count&quot;</span>
<span class="p">)</span>

<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">sns</span><span class="o">.</span><span class="n">barplot</span><span class="p">(</span>
    <span class="n">data</span><span class="o">=</span><span class="n">plot_data_global</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="s2">&quot;Disease status&quot;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s2">&quot;count&quot;</span><span class="p">,</span> <span class="n">hue</span><span class="o">=</span><span class="s2">&quot;Cell type&quot;</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="p">)</span>
<span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Global abundances, by status&quot;</span><span class="p">)</span>

<span class="n">sns</span><span class="o">.</span><span class="n">barplot</span><span class="p">(</span>
    <span class="n">data</span><span class="o">=</span><span class="n">plot_data_global</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="s2">&quot;Cell type&quot;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s2">&quot;count&quot;</span><span class="p">,</span> <span class="n">hue</span><span class="o">=</span><span class="s2">&quot;Disease status&quot;</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="p">)</span>
<span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Global abundances, by cell type&quot;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/compositional_14_0.png" src="../_images/compositional_14_0.png" />
</div>
</div>
<p>We want to find out which cell types increase or decrease in abundance in the diseased organ. If we are able to determine the type of every cell in both organs, the case would be clear, as we can see in the right plot above. Unfortunately, this is not possible. Since our sequencing process has a limited capacity, we can only take a representative sample of 600 cells from both populations. To simulate this step, we can use numpy’s <code class="docutils literal notranslate"><span class="pre">random.multinomial</span></code> function to sample 600 cells from the populations without replacement:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">1234</span><span class="p">)</span>
<span class="n">healthy_sample</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">multinomial</span><span class="p">(</span>
    <span class="n">pvals</span><span class="o">=</span><span class="n">healthy_tissue</span> <span class="o">/</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">healthy_tissue</span><span class="p">),</span> <span class="n">n</span><span class="o">=</span><span class="mi">600</span>
<span class="p">)</span>
<span class="n">diseased_sample</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">multinomial</span><span class="p">(</span>
    <span class="n">pvals</span><span class="o">=</span><span class="n">diseased_tissue</span> <span class="o">/</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">diseased_tissue</span><span class="p">),</span> <span class="n">n</span><span class="o">=</span><span class="mi">600</span>
<span class="p">)</span>
<span class="n">example_data_sample</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
    <span class="n">data</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">healthy_sample</span><span class="p">,</span> <span class="n">diseased_sample</span><span class="p">]),</span>
    <span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span>
    <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">],</span>
<span class="p">)</span>
<span class="n">example_data_sample</span><span class="p">[</span><span class="s2">&quot;Disease status&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;Healthy&quot;</span><span class="p">,</span> <span class="s2">&quot;Diseased&quot;</span><span class="p">]</span>
<span class="n">example_data_sample</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>A</th>
      <th>B</th>
      <th>C</th>
      <th>Disease status</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>1</th>
      <td>193</td>
      <td>201</td>
      <td>206</td>
      <td>Healthy</td>
    </tr>
    <tr>
      <th>2</th>
      <td>296</td>
      <td>146</td>
      <td>158</td>
      <td>Diseased</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">plot_data_sample</span> <span class="o">=</span> <span class="n">example_data_sample</span><span class="o">.</span><span class="n">melt</span><span class="p">(</span>
    <span class="s2">&quot;Disease status&quot;</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">],</span> <span class="s2">&quot;Cell type&quot;</span><span class="p">,</span> <span class="s2">&quot;count&quot;</span>
<span class="p">)</span>

<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">sns</span><span class="o">.</span><span class="n">barplot</span><span class="p">(</span>
    <span class="n">data</span><span class="o">=</span><span class="n">plot_data_sample</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="s2">&quot;Disease status&quot;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s2">&quot;count&quot;</span><span class="p">,</span> <span class="n">hue</span><span class="o">=</span><span class="s2">&quot;Cell type&quot;</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="p">)</span>
<span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Sampled abundances, by status&quot;</span><span class="p">)</span>

<span class="n">sns</span><span class="o">.</span><span class="n">barplot</span><span class="p">(</span>
    <span class="n">data</span><span class="o">=</span><span class="n">plot_data_sample</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="s2">&quot;Cell type&quot;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s2">&quot;count&quot;</span><span class="p">,</span> <span class="n">hue</span><span class="o">=</span><span class="s2">&quot;Disease status&quot;</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="p">)</span>
<span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Sampled abundances, by cell type&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/compositional_17_0.png" src="../_images/compositional_17_0.png" />
</div>
</div>
<p>Now the picture is not clear anymore. While the counts of cell type A still increase (approx. from 200 to 300), the other two cell types seem to decrease from about 200 to 150. This apparent decrease is caused by our constraint to 600 cells - If a larger fraction of the sample is taken up by cell type A, the share of cell types B and C must be lower. Therefore, determining the change in abundance of one cell type is impossible without taking the other cell types into account.</p>
<p>If we ignore the compositionality of the data, and use univariate methods like Wilcoxon rank-sum tests or scDC, a method which performs differential cell-type composition analysis by bootstrap resampling<span id="id3">[<a class="reference internal" href="#id315" title="Yue Cao, Yingxin Lin, John T. Ormerod, Pengyi Yang, Jean Y.H. Yang, and Kitty K. Lo. Scdc: single cell differential composition analysis. BMC Bioinformatics, 20(19):721, Dec 2019. URL: https://doi.org/10.1186/s12859-019-3211-9, doi:10.1186/s12859-019-3211-9.">Cao <em>et al.</em>, 2019</a>]</span>, we may falsely perceive cell-type population shifts as statistically sound effects, although they were induced by inherent negative correlations of the cell-type proportions.</p>
<p>Furthermore, the subsampled data does not only give us one valid solution to our question. If both cell types B and C decreased by 1,000 cells in the diseased case, we would obtain the same representative samples of 600 cells as above. To get a unique result, we can fix a reference point for the data, which is assumed to be unchanged throughout all samples<span id="id4">[<a class="reference internal" href="#id323" title="Barak Brill, Amnon Amir, and Ruth Heller. Testing for differential abundance in compositional counts data, with application to microbiome studies. ArXiv, April 2019. URL: http://arxiv.org/abs/1904.08937, arXiv:1904.08937.">Brill <em>et al.</em>, 2019</a>]</span>. This can be a single cell type, an aggregation over multiple cell types such as the geometric mean, or a set of orthogonal bases{cite}Egozcue2003.</p>
<p>While single-cell datasets of sufficient size and replicate number have only been around for a few years, the same statistical property has also been discussed in the context of microbial analysis<span id="id5">[<a class="reference internal" href="#id325" title="Gregory B Gloor, Jean M Macklaim, Vera Pawlowsky-Glahn, and Juan J Egozcue. Microbiome datasets are compositional: and this is not optional. Front. Microbiol., 8:2224, November 2017. URL: http://dx.doi.org/10.3389/fmicb.2017.02224, doi:10.3389/fmicb.2017.02224.">Gloor <em>et al.</em>, 2017</a>]</span>. There, some popular approaches include ANCOM-BC <span id="id6">[<a class="reference internal" href="#id326" title="Huang Lin and Shyamal Das Peddada. Analysis of compositions of microbiomes with bias correction. Nat. Commun., 11(1):3514, July 2020. URL: http://dx.doi.org/10.1038/s41467-020-17041-7, doi:10.1038/s41467-020-17041-7.">Lin and Peddada, 2020</a>]</span> and ALDEx2 <span id="id7">[<a class="reference internal" href="#id327" title="Andrew D Fernandes, Jennifer Ns Reid, Jean M Macklaim, Thomas A McMurrough, David R Edgell, and Gregory B Gloor. Unifying the analysis of high-throughput sequencing datasets: characterizing RNA-seq, 16S rRNA gene sequencing and selective growth experiments by compositional data analysis. Microbiome, 2:15, May 2014. URL: http://dx.doi.org/10.1186/2049-2618-2-15, doi:10.1186/2049-2618-2-15.">Fernandes <em>et al.</em>, 2014</a>]</span>. However, these approaches often struggle with single-cell datasets due to the small number of experimental replicates.</p>
<p>This issue has been tackled by scCODA<span id="id8">[<a class="reference internal" href="#id316" title="M. Büttner, J. Ostner, C. L. Müller, F. J. Theis, and B. Schubert. Sccoda is a bayesian model for compositional single-cell data analysis. Nature Communications, 12(1):6876, Nov 2021. URL: https://doi.org/10.1038/s41467-021-27150-6, doi:10.1038/s41467-021-27150-6.">Büttner <em>et al.</em>, 2021</a>]</span>, which we are going to introduce and apply to our dataset in the following section.</p>
</section>
<section id="with-labeled-clusters">
<h2><span class="section-number">18.4. </span>With labeled clusters<a class="headerlink" href="#with-labeled-clusters" title="Permalink to this headline">#</a></h2>
<p><a class="reference external" href="https://sccoda.readthedocs.io/en/latest">scCODA</a> belongs to the family of tools that require pre-defined clusters, most commony cell types, to statistically derive changes in composition. Inspired by methods for compositional analysis of microbiome data, scCODA proposes a Bayesian approach to address the low replicate issue as commonly encountered in single-cell analysis<span id="id9">[<a class="reference internal" href="#id316" title="M. Büttner, J. Ostner, C. L. Müller, F. J. Theis, and B. Schubert. Sccoda is a bayesian model for compositional single-cell data analysis. Nature Communications, 12(1):6876, Nov 2021. URL: https://doi.org/10.1038/s41467-021-27150-6, doi:10.1038/s41467-021-27150-6.">Büttner <em>et al.</em>, 2021</a>]</span>. It models cell-type counts using a hierarchical Dirichlet-Multinomial model, which accounts for uncertainty in cell-type proportions and the negative correlative bias via joint modeling of all measured cell-type proportions. To ensure a uniquely identifiable solution and easy interpretability, the reference in scCODA is chosen to be a specific cell type. Hence, any detected compositional changes by scCODA always have to be viewed in relation to the selected reference.</p>
<p>However, scCODA assumes a log-linear relationship between covariates and cell abundance, which may not always reflect the underlying biological processes when using continuoaus covariates. A further limitation of scCODA is the inability to infer correlation structures among cell compositions beyond compositional effects. Furthermore, scCODA only models shifts in mean abundance, but does not detect changes in response variability<span id="id10">[<a class="reference internal" href="#id316" title="M. Büttner, J. Ostner, C. L. Müller, F. J. Theis, and B. Schubert. Sccoda is a bayesian model for compositional single-cell data analysis. Nature Communications, 12(1):6876, Nov 2021. URL: https://doi.org/10.1038/s41467-021-27150-6, doi:10.1038/s41467-021-27150-6.">Büttner <em>et al.</em>, 2021</a>]</span>.</p>
<p>As a first step, we import the required scCODA functions.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">sccoda.util.cell_composition_data</span> <span class="k">as</span> <span class="nn">scc_dat</span>
<span class="kn">import</span> <span class="nn">sccoda.util.comp_ana</span> <span class="k">as</span> <span class="nn">scc_ana</span>
<span class="kn">import</span> <span class="nn">sccoda.util.data_visualization</span> <span class="k">as</span> <span class="nn">scc_viz</span>
</pre></div>
</div>
</div>
</div>
<p>scCODA cannot be applied to the main AnnData object directly. The <a class="reference external" href="https://sccoda.readthedocs.io/en/latest/Data_import_and_visualization.html">scCODA data import and visualization tutorial</a> describes the various options of generating a specific cell counts AnnData object from either a Pandas DataFrame or an AnnData object. Instead of using scCODA’s API to generate such an AnnData object, we create an intermediate Pandas DataFrame to better illustrate the expected types of input objects for scCODA.
As a first step, we select all conditions that we want to test, group them by the conditions and batches and use this intermediate representation to count the cell types.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">frac_by_condition</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
        <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;condition&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span>
            <span class="p">[</span><span class="s2">&quot;Control&quot;</span><span class="p">,</span> <span class="s2">&quot;Salmonella&quot;</span><span class="p">,</span> <span class="s2">&quot;Hpoly.Day3&quot;</span><span class="p">,</span> <span class="s2">&quot;Hpoly.Day10&quot;</span><span class="p">]</span>
        <span class="p">)</span>
    <span class="p">]</span>
    <span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s2">&quot;condition&quot;</span><span class="p">,</span> <span class="s2">&quot;batch&quot;</span><span class="p">])</span>
    <span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">value_counts</span><span class="p">(</span><span class="s2">&quot;cell_label&quot;</span><span class="p">,</span> <span class="n">normalize</span><span class="o">=</span><span class="kc">False</span><span class="p">))</span>
    <span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;n_cells&quot;</span><span class="p">)</span>
    <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">condition</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;condition&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">))</span>
<span class="p">)</span>
<span class="n">frac_by_condition</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>condition</th>
      <th>batch</th>
      <th>cell_label</th>
      <th>n_cells</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>Control</td>
      <td>B1</td>
      <td>Stem</td>
      <td>239</td>
    </tr>
    <tr>
      <th>1</th>
      <td>Control</td>
      <td>B1</td>
      <td>TA.Early</td>
      <td>191</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Control</td>
      <td>B1</td>
      <td>Enterocyte.Progenitor</td>
      <td>136</td>
    </tr>
    <tr>
      <th>3</th>
      <td>Control</td>
      <td>B1</td>
      <td>TA</td>
      <td>125</td>
    </tr>
    <tr>
      <th>4</th>
      <td>Control</td>
      <td>B1</td>
      <td>Enterocyte</td>
      <td>59</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>75</th>
      <td>Salmonella</td>
      <td>B10</td>
      <td>Enterocyte.Progenitor</td>
      <td>116</td>
    </tr>
    <tr>
      <th>76</th>
      <td>Salmonella</td>
      <td>B10</td>
      <td>Goblet</td>
      <td>67</td>
    </tr>
    <tr>
      <th>77</th>
      <td>Salmonella</td>
      <td>B10</td>
      <td>TA</td>
      <td>65</td>
    </tr>
    <tr>
      <th>78</th>
      <td>Salmonella</td>
      <td>B10</td>
      <td>Endocrine</td>
      <td>32</td>
    </tr>
    <tr>
      <th>79</th>
      <td>Salmonella</td>
      <td>B10</td>
      <td>Tuft</td>
      <td>12</td>
    </tr>
  </tbody>
</table>
<p>80 rows × 4 columns</p>
</div></div></div>
</div>
<p>To group the representations by conditions which are ordered by batches, we <a class="reference external" href="https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.pivot.html">pivot</a> the table.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">frac_pivot</span> <span class="o">=</span> <span class="n">frac_by_condition</span><span class="o">.</span><span class="n">pivot</span><span class="p">(</span>
    <span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;batch&quot;</span><span class="p">,</span> <span class="s2">&quot;condition&quot;</span><span class="p">],</span>
    <span class="n">columns</span><span class="o">=</span><span class="s2">&quot;cell_label&quot;</span><span class="p">,</span>
    <span class="n">values</span><span class="o">=</span><span class="s2">&quot;n_cells&quot;</span><span class="p">,</span>
<span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
<span class="n">frac_pivot</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>cell_label</th>
      <th>batch</th>
      <th>condition</th>
      <th>Endocrine</th>
      <th>Enterocyte</th>
      <th>Enterocyte.Progenitor</th>
      <th>Goblet</th>
      <th>Stem</th>
      <th>TA</th>
      <th>TA.Early</th>
      <th>Tuft</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>B1</td>
      <td>Control</td>
      <td>36</td>
      <td>59</td>
      <td>136</td>
      <td>36</td>
      <td>239</td>
      <td>125</td>
      <td>191</td>
      <td>18</td>
    </tr>
    <tr>
      <th>1</th>
      <td>B2</td>
      <td>Control</td>
      <td>5</td>
      <td>46</td>
      <td>23</td>
      <td>20</td>
      <td>50</td>
      <td>11</td>
      <td>40</td>
      <td>5</td>
    </tr>
    <tr>
      <th>2</th>
      <td>B3</td>
      <td>Control</td>
      <td>45</td>
      <td>98</td>
      <td>188</td>
      <td>124</td>
      <td>250</td>
      <td>155</td>
      <td>365</td>
      <td>33</td>
    </tr>
    <tr>
      <th>3</th>
      <td>B4</td>
      <td>Control</td>
      <td>26</td>
      <td>221</td>
      <td>198</td>
      <td>36</td>
      <td>131</td>
      <td>130</td>
      <td>196</td>
      <td>4</td>
    </tr>
    <tr>
      <th>4</th>
      <td>B5</td>
      <td>Hpoly.Day3</td>
      <td>52</td>
      <td>75</td>
      <td>347</td>
      <td>66</td>
      <td>323</td>
      <td>263</td>
      <td>313</td>
      <td>51</td>
    </tr>
    <tr>
      <th>5</th>
      <td>B6</td>
      <td>Hpoly.Day3</td>
      <td>65</td>
      <td>126</td>
      <td>115</td>
      <td>33</td>
      <td>65</td>
      <td>39</td>
      <td>129</td>
      <td>59</td>
    </tr>
    <tr>
      <th>6</th>
      <td>B7</td>
      <td>Hpoly.Day10</td>
      <td>42</td>
      <td>71</td>
      <td>203</td>
      <td>147</td>
      <td>271</td>
      <td>109</td>
      <td>180</td>
      <td>146</td>
    </tr>
    <tr>
      <th>7</th>
      <td>B8</td>
      <td>Hpoly.Day10</td>
      <td>40</td>
      <td>57</td>
      <td>383</td>
      <td>170</td>
      <td>321</td>
      <td>244</td>
      <td>256</td>
      <td>71</td>
    </tr>
    <tr>
      <th>8</th>
      <td>B9</td>
      <td>Salmonella</td>
      <td>37</td>
      <td>332</td>
      <td>113</td>
      <td>59</td>
      <td>90</td>
      <td>47</td>
      <td>132</td>
      <td>10</td>
    </tr>
    <tr>
      <th>9</th>
      <td>B10</td>
      <td>Salmonella</td>
      <td>32</td>
      <td>373</td>
      <td>116</td>
      <td>67</td>
      <td>117</td>
      <td>65</td>
      <td>168</td>
      <td>12</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>This allows us to finally use scCODA’s <code class="docutils literal notranslate"><span class="pre">from_pandas</span></code> function to create a scCODA specific AnnData object by specifying the covariate columns which are <code class="docutils literal notranslate"><span class="pre">batch</span></code> and <code class="docutils literal notranslate"><span class="pre">condition</span></code> in our case.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">scc_df</span> <span class="o">=</span> <span class="n">scc_dat</span><span class="o">.</span><span class="n">from_pandas</span><span class="p">(</span><span class="n">frac_pivot</span><span class="p">,</span> <span class="n">covariate_columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;batch&quot;</span><span class="p">,</span> <span class="s2">&quot;condition&quot;</span><span class="p">])</span>
<span class="n">scc_df</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 10 × 8
    obs: &#39;batch&#39;, &#39;condition&#39;
</pre></div>
</div>
</div>
</div>
<p>To get an overview of the cell type distributions across conditions we can use scCODA’s <code class="docutils literal notranslate"><span class="pre">boxplots</span></code>. To get an even better understanding of how the data is distributed, the red dots show the actual data points.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">scc_viz</span><span class="o">.</span><span class="n">boxplots</span><span class="p">(</span>
    <span class="n">scc_df</span><span class="p">,</span>
    <span class="n">feature_name</span><span class="o">=</span><span class="s2">&quot;condition&quot;</span><span class="p">,</span>
    <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">5</span><span class="p">),</span>
    <span class="n">add_dots</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">args_swarmplot</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;palette&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;red&quot;</span><span class="p">]},</span>
<span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/compositional_32_0.png" src="../_images/compositional_32_0.png" />
</div>
</div>
<p>The boxplots highlight some differences in the distributions of the cell types. Clearly noticable is the high proportion of enterocytes for the Salmonella condition. But other cell types such as transit-amplifying (TA) cells also show stark differences in abundance for the Salmonella condition compared to control. Whether any of these differences are statistically significant has to be properly evaluated.</p>
<p>An alternative visualization is a stacked barplot as provided by scCODA. This visualization nicely displays the characteristics of compositional data: If we compare the Control and Salmonella groups, we can see that the proportion of Enterocytes greatly increases in the infected mice. Since the data is proportional, this leads to a decreased share of all other cell types to fulfil the sum-to-one constraint.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">scc_viz</span><span class="o">.</span><span class="n">stacked_barplot</span><span class="p">(</span><span class="n">scc_df</span><span class="p">,</span> <span class="n">feature_name</span><span class="o">=</span><span class="s2">&quot;condition&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/compositional_34_0.png" src="../_images/compositional_34_0.png" />
</div>
</div>
<p>scCODA requires two major parameters beyond the cell count AnnData object: A formula and a reference cell type. The formula describes the covariates, which are specified using the <a class="reference external" href="https://www.statsmodels.org/stable/example_formulas.html">R-style</a>. In our case we specify the condition as the only covariate. Since it is a discrete covariate with four levels (control and three disease states), this models a comparison of each state with the other samples.
If we wanted to model multiple covariates at once, simply adding them in the formula (i.e. <code class="docutils literal notranslate"><span class="pre">formula</span> <span class="pre">=</span> <span class="pre">&quot;covariate_1</span> <span class="pre">+</span> <span class="pre">covariate_2&quot;</span></code>) is enough.
As mentioned above, scCODA requires a reference cell type to compare against, which is believed to be unchanged by the covariates. scCODA can either automatically select an appropriate cell type as reference, which is a cell type that has nearly constant relative abundance over all samples, or be run with a user specified reference cell type. Here we set Endocrine cells as the reference since visually their abundance seems to be rather constant. An alternative to setting a reference cell type manually is to set the <code class="docutils literal notranslate"><span class="pre">reference_cell_type</span></code> to <code class="docutils literal notranslate"><span class="pre">&quot;automatic&quot;</span></code> which will force scCODA to select a suitable reference cell type itself. If the choice of reference cell type is unclear, we recommend to use this option to get an indicator or even a final selection.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">1234</span><span class="p">)</span>
<span class="n">tf</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">set_seed</span><span class="p">(</span><span class="mi">1234</span><span class="p">)</span>

<span class="n">sccoda_mod</span> <span class="o">=</span> <span class="n">scc_ana</span><span class="o">.</span><span class="n">CompositionalAnalysis</span><span class="p">(</span>
    <span class="n">scc_df</span><span class="p">,</span>
    <span class="n">formula</span><span class="o">=</span><span class="s2">&quot;condition&quot;</span><span class="p">,</span>
    <span class="n">reference_cell_type</span><span class="o">=</span><span class="s2">&quot;Endocrine&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">sccoda_res</span> <span class="o">=</span> <span class="n">sccoda_mod</span><span class="o">.</span><span class="n">sample_hmc</span><span class="p">(</span><span class="n">num_results</span><span class="o">=</span><span class="mi">20000</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>100%|██████████| 20000/20000 [00:53&lt;00:00, 374.73it/s]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>MCMC sampling finished. (69.540 sec)
Acceptance rate: 55.2%
</pre></div>
</div>
</div>
</div>
<p>The acceptance rate describes the fraction of proposed samples that are accepted after the initial burn-in phase, and can be an ad-hoc indicator for a bad optimization run. In the case of scCODA, the desired acceptance rate is between 0.4 and 0.9. Acceptance rates that are way higher or too low indicate issues with the sampling process.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sccoda_res</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
            <div>
              <div class='xr-header'>
                <div class="xr-obj-type">arviz.InferenceData</div>
              </div>
              <ul class="xr-sections group-sections">

            <li class = "xr-section-item">
                  <input id="idata_posterior34524f5a-7458-4b42-a439-8610e154856f" class="xr-section-summary-in" type="checkbox">
                  <label for="idata_posterior34524f5a-7458-4b42-a439-8610e154856f" class = "xr-section-summary">posterior</label>
                  <div class="xr-section-inline-details"></div>
                  <div class="xr-section-details">
                      <ul id="xr-dataset-coord-list" class="xr-var-list">
                          <div style="padding-left:2rem;"><div><svg style="position: absolute; width: 0; height: 0; overflow: hidden">
<defs>
<symbol id="icon-database" viewBox="0 0 32 32">
<path d="M16 0c-8.837 0-16 2.239-16 5v4c0 2.761 7.163 5 16 5s16-2.239 16-5v-4c0-2.761-7.163-5-16-5z"></path>
<path d="M16 17c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z"></path>
<path d="M16 26c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z"></path>
</symbol>
<symbol id="icon-file-text2" viewBox="0 0 32 32">
<path d="M28.681 7.159c-0.694-0.947-1.662-2.053-2.724-3.116s-2.169-2.030-3.116-2.724c-1.612-1.182-2.393-1.319-2.841-1.319h-15.5c-1.378 0-2.5 1.121-2.5 2.5v27c0 1.378 1.122 2.5 2.5 2.5h23c1.378 0 2.5-1.122 2.5-2.5v-19.5c0-0.448-0.137-1.23-1.319-2.841zM24.543 5.457c0.959 0.959 1.712 1.825 2.268 2.543h-4.811v-4.811c0.718 0.556 1.584 1.309 2.543 2.268zM28 29.5c0 0.271-0.229 0.5-0.5 0.5h-23c-0.271 0-0.5-0.229-0.5-0.5v-27c0-0.271 0.229-0.5 0.5-0.5 0 0 15.499-0 15.5 0v7c0 0.552 0.448 1 1 1h7v19.5z"></path>
<path d="M23 26h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z"></path>
<path d="M23 22h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z"></path>
<path d="M23 18h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z"></path>
</symbol>
</defs>
</svg>
<style>/* CSS stylesheet for displaying xarray objects in jupyterlab.
 *
 */

:root {
  --xr-font-color0: var(--jp-content-font-color0, rgba(0, 0, 0, 1));
  --xr-font-color2: var(--jp-content-font-color2, rgba(0, 0, 0, 0.54));
  --xr-font-color3: var(--jp-content-font-color3, rgba(0, 0, 0, 0.38));
  --xr-border-color: var(--jp-border-color2, #e0e0e0);
  --xr-disabled-color: var(--jp-layout-color3, #bdbdbd);
  --xr-background-color: var(--jp-layout-color0, white);
  --xr-background-color-row-even: var(--jp-layout-color1, white);
  --xr-background-color-row-odd: var(--jp-layout-color2, #eeeeee);
}

html[theme=dark],
body[data-theme=dark],
body.vscode-dark {
  --xr-font-color0: rgba(255, 255, 255, 1);
  --xr-font-color2: rgba(255, 255, 255, 0.54);
  --xr-font-color3: rgba(255, 255, 255, 0.38);
  --xr-border-color: #1F1F1F;
  --xr-disabled-color: #515151;
  --xr-background-color: #111111;
  --xr-background-color-row-even: #111111;
  --xr-background-color-row-odd: #313131;
}

.xr-wrap {
  display: block !important;
  min-width: 300px;
  max-width: 700px;
}

.xr-text-repr-fallback {
  /* fallback to plain text repr when CSS is not injected (untrusted notebook) */
  display: none;
}

.xr-header {
  padding-top: 6px;
  padding-bottom: 6px;
  margin-bottom: 4px;
  border-bottom: solid 1px var(--xr-border-color);
}

.xr-header > div,
.xr-header > ul {
  display: inline;
  margin-top: 0;
  margin-bottom: 0;
}

.xr-obj-type,
.xr-array-name {
  margin-left: 2px;
  margin-right: 10px;
}

.xr-obj-type {
  color: var(--xr-font-color2);
}

.xr-sections {
  padding-left: 0 !important;
  display: grid;
  grid-template-columns: 150px auto auto 1fr 20px 20px;
}

.xr-section-item {
  display: contents;
}

.xr-section-item input {
  display: none;
}

.xr-section-item input + label {
  color: var(--xr-disabled-color);
}

.xr-section-item input:enabled + label {
  cursor: pointer;
  color: var(--xr-font-color2);
}

.xr-section-item input:enabled + label:hover {
  color: var(--xr-font-color0);
}

.xr-section-summary {
  grid-column: 1;
  color: var(--xr-font-color2);
  font-weight: 500;
}

.xr-section-summary > span {
  display: inline-block;
  padding-left: 0.5em;
}

.xr-section-summary-in:disabled + label {
  color: var(--xr-font-color2);
}

.xr-section-summary-in + label:before {
  display: inline-block;
  content: '►';
  font-size: 11px;
  width: 15px;
  text-align: center;
}

.xr-section-summary-in:disabled + label:before {
  color: var(--xr-disabled-color);
}

.xr-section-summary-in:checked + label:before {
  content: '▼';
}

.xr-section-summary-in:checked + label > span {
  display: none;
}

.xr-section-summary,
.xr-section-inline-details {
  padding-top: 4px;
  padding-bottom: 4px;
}

.xr-section-inline-details {
  grid-column: 2 / -1;
}

.xr-section-details {
  display: none;
  grid-column: 1 / -1;
  margin-bottom: 5px;
}

.xr-section-summary-in:checked ~ .xr-section-details {
  display: contents;
}

.xr-array-wrap {
  grid-column: 1 / -1;
  display: grid;
  grid-template-columns: 20px auto;
}

.xr-array-wrap > label {
  grid-column: 1;
  vertical-align: top;
}

.xr-preview {
  color: var(--xr-font-color3);
}

.xr-array-preview,
.xr-array-data {
  padding: 0 5px !important;
  grid-column: 2;
}

.xr-array-data,
.xr-array-in:checked ~ .xr-array-preview {
  display: none;
}

.xr-array-in:checked ~ .xr-array-data,
.xr-array-preview {
  display: inline-block;
}

.xr-dim-list {
  display: inline-block !important;
  list-style: none;
  padding: 0 !important;
  margin: 0;
}

.xr-dim-list li {
  display: inline-block;
  padding: 0;
  margin: 0;
}

.xr-dim-list:before {
  content: '(';
}

.xr-dim-list:after {
  content: ')';
}

.xr-dim-list li:not(:last-child):after {
  content: ',';
  padding-right: 5px;
}

.xr-has-index {
  font-weight: bold;
}

.xr-var-list,
.xr-var-item {
  display: contents;
}

.xr-var-item > div,
.xr-var-item label,
.xr-var-item > .xr-var-name span {
  background-color: var(--xr-background-color-row-even);
  margin-bottom: 0;
}

.xr-var-item > .xr-var-name:hover span {
  padding-right: 5px;
}

.xr-var-list > li:nth-child(odd) > div,
.xr-var-list > li:nth-child(odd) > label,
.xr-var-list > li:nth-child(odd) > .xr-var-name span {
  background-color: var(--xr-background-color-row-odd);
}

.xr-var-name {
  grid-column: 1;
}

.xr-var-dims {
  grid-column: 2;
}

.xr-var-dtype {
  grid-column: 3;
  text-align: right;
  color: var(--xr-font-color2);
}

.xr-var-preview {
  grid-column: 4;
}

.xr-var-name,
.xr-var-dims,
.xr-var-dtype,
.xr-preview,
.xr-attrs dt {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  padding-right: 10px;
}

.xr-var-name:hover,
.xr-var-dims:hover,
.xr-var-dtype:hover,
.xr-attrs dt:hover {
  overflow: visible;
  width: auto;
  z-index: 1;
}

.xr-var-attrs,
.xr-var-data {
  display: none;
  background-color: var(--xr-background-color) !important;
  padding-bottom: 5px !important;
}

.xr-var-attrs-in:checked ~ .xr-var-attrs,
.xr-var-data-in:checked ~ .xr-var-data {
  display: block;
}

.xr-var-data > table {
  float: right;
}

.xr-var-name span,
.xr-var-data,
.xr-attrs {
  padding-left: 25px !important;
}

.xr-attrs,
.xr-var-attrs,
.xr-var-data {
  grid-column: 1 / -1;
}

dl.xr-attrs {
  padding: 0;
  margin: 0;
  display: grid;
  grid-template-columns: 125px auto;
}

.xr-attrs dt,
.xr-attrs dd {
  padding: 0;
  margin: 0;
  float: left;
  padding-right: 10px;
  width: auto;
}

.xr-attrs dt {
  font-weight: normal;
  grid-column: 1;
}

.xr-attrs dt:hover span {
  display: inline-block;
  background: var(--xr-background-color);
  padding-right: 10px;
}

.xr-attrs dd {
  grid-column: 2;
  white-space: pre-wrap;
  word-break: break-all;
}

.xr-icon-database,
.xr-icon-file-text2 {
  display: inline-block;
  vertical-align: middle;
  width: 1em;
  height: 1.5em !important;
  stroke-width: 0;
  stroke: currentColor;
  fill: currentColor;
}
</style><pre class='xr-text-repr-fallback'>&lt;xarray.Dataset&gt;
Dimensions:        (chain: 1, draw: 15000, covariate: 3, sigma_d_dim_1: 1,
                    cell_type_nb: 7, cell_type: 8, sample: 10)
Coordinates:
  * chain          (chain) int64 0
  * draw           (draw) int64 0 1 2 3 4 5 ... 14995 14996 14997 14998 14999
  * covariate      (covariate) &lt;U24 &#x27;condition[T.Hpoly.Day10]&#x27; ... &#x27;condition...
  * sigma_d_dim_1  (sigma_d_dim_1) int64 0
  * cell_type_nb   (cell_type_nb) &lt;U21 &#x27;Enterocyte&#x27; ... &#x27;Tuft&#x27;
  * cell_type      (cell_type) &lt;U21 &#x27;Endocrine&#x27; &#x27;Enterocyte&#x27; ... &#x27;Tuft&#x27;
  * sample         (sample) int64 0 1 2 3 4 5 6 7 8 9
Data variables:
    sigma_d        (chain, draw, covariate, sigma_d_dim_1) float64 0.7184 ......
    b_offset       (chain, draw, covariate, cell_type_nb) float64 -0.7204 ......
    ind_raw        (chain, draw, covariate, cell_type_nb) float64 0.3983 ... ...
    alpha          (chain, draw, cell_type) float64 1.033 2.106 ... 2.72 0.7746
    ind            (chain, draw, covariate, cell_type_nb) float64 1.0 ... 1.2...
    b_raw          (chain, draw, covariate, cell_type_nb) float64 -0.5175 ......
    beta           (chain, draw, covariate, cell_type) float64 0.0 ... 1.457e-29
    concentration  (chain, draw, sample, cell_type) float64 2.81 8.217 ... 2.17
Attributes:
    created_at:     2022-08-25T13:07:26.604429
    arviz_version:  0.12.1</pre><div class='xr-wrap' style='display:none'><div class='xr-header'><div class='xr-obj-type'>xarray.Dataset</div></div><ul class='xr-sections'><li class='xr-section-item'><input id='section-ec736639-fa63-4871-8e2f-3dabc52f19aa' class='xr-section-summary-in' type='checkbox' disabled ><label for='section-ec736639-fa63-4871-8e2f-3dabc52f19aa' class='xr-section-summary'  title='Expand/collapse section'>Dimensions:</label><div class='xr-section-inline-details'><ul class='xr-dim-list'><li><span class='xr-has-index'>chain</span>: 1</li><li><span class='xr-has-index'>draw</span>: 15000</li><li><span class='xr-has-index'>covariate</span>: 3</li><li><span class='xr-has-index'>sigma_d_dim_1</span>: 1</li><li><span class='xr-has-index'>cell_type_nb</span>: 7</li><li><span class='xr-has-index'>cell_type</span>: 8</li><li><span class='xr-has-index'>sample</span>: 10</li></ul></div><div class='xr-section-details'></div></li><li class='xr-section-item'><input id='section-b1b57f68-ee05-4785-b2b1-f938d1d975d5' class='xr-section-summary-in' type='checkbox'  checked><label for='section-b1b57f68-ee05-4785-b2b1-f938d1d975d5' class='xr-section-summary' >Coordinates: <span>(7)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>chain</span></div><div class='xr-var-dims'>(chain)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>0</div><input id='attrs-f5b7cf68-1c56-4178-a1ee-41c4ef000436' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-f5b7cf68-1c56-4178-a1ee-41c4ef000436' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-3be9b2f7-2b94-4fc0-9b5e-c16d58013d49' class='xr-var-data-in' type='checkbox'><label for='data-3be9b2f7-2b94-4fc0-9b5e-c16d58013d49' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([0])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>draw</span></div><div class='xr-var-dims'>(draw)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>0 1 2 3 ... 14996 14997 14998 14999</div><input id='attrs-6f0186e8-2e0c-4491-a5e0-3269973a10b1' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-6f0186e8-2e0c-4491-a5e0-3269973a10b1' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-35a7983d-6de6-473c-b372-14b3e6ed87b0' class='xr-var-data-in' type='checkbox'><label for='data-35a7983d-6de6-473c-b372-14b3e6ed87b0' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([    0,     1,     2, ..., 14997, 14998, 14999])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>covariate</span></div><div class='xr-var-dims'>(covariate)</div><div class='xr-var-dtype'>&lt;U24</div><div class='xr-var-preview xr-preview'>&#x27;condition[T.Hpoly.Day10]&#x27; ... &#x27;...</div><input id='attrs-b283ec15-0168-4784-9697-27b62984d5cf' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-b283ec15-0168-4784-9697-27b62984d5cf' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-83f6a9b2-528e-4815-b37b-c16d5dec8b5d' class='xr-var-data-in' type='checkbox'><label for='data-83f6a9b2-528e-4815-b37b-c16d5dec8b5d' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;condition[T.Hpoly.Day10]&#x27;, &#x27;condition[T.Hpoly.Day3]&#x27;,
       &#x27;condition[T.Salmonella]&#x27;], dtype=&#x27;&lt;U24&#x27;)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>sigma_d_dim_1</span></div><div class='xr-var-dims'>(sigma_d_dim_1)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>0</div><input id='attrs-5bb37b68-77c5-4cd8-9078-b49c176ad6b0' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-5bb37b68-77c5-4cd8-9078-b49c176ad6b0' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-60877c36-966a-4c47-aa7e-23f5977c8bed' class='xr-var-data-in' type='checkbox'><label for='data-60877c36-966a-4c47-aa7e-23f5977c8bed' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([0])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>cell_type_nb</span></div><div class='xr-var-dims'>(cell_type_nb)</div><div class='xr-var-dtype'>&lt;U21</div><div class='xr-var-preview xr-preview'>&#x27;Enterocyte&#x27; ... &#x27;Tuft&#x27;</div><input id='attrs-0ac69b89-a826-449c-834a-61f651f4084e' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-0ac69b89-a826-449c-834a-61f651f4084e' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-f635dfea-c024-42b3-b71a-2ddfdc692c88' class='xr-var-data-in' type='checkbox'><label for='data-f635dfea-c024-42b3-b71a-2ddfdc692c88' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;Enterocyte&#x27;, &#x27;Enterocyte.Progenitor&#x27;, &#x27;Goblet&#x27;, &#x27;Stem&#x27;, &#x27;TA&#x27;,
       &#x27;TA.Early&#x27;, &#x27;Tuft&#x27;], dtype=&#x27;&lt;U21&#x27;)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>cell_type</span></div><div class='xr-var-dims'>(cell_type)</div><div class='xr-var-dtype'>&lt;U21</div><div class='xr-var-preview xr-preview'>&#x27;Endocrine&#x27; &#x27;Enterocyte&#x27; ... &#x27;Tuft&#x27;</div><input id='attrs-c86c6ffc-ab3a-45a8-9014-7f48233819b3' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-c86c6ffc-ab3a-45a8-9014-7f48233819b3' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-32b502ff-5a4d-4f44-8405-52d32f5289d2' class='xr-var-data-in' type='checkbox'><label for='data-32b502ff-5a4d-4f44-8405-52d32f5289d2' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;Endocrine&#x27;, &#x27;Enterocyte&#x27;, &#x27;Enterocyte.Progenitor&#x27;, &#x27;Goblet&#x27;, &#x27;Stem&#x27;,
       &#x27;TA&#x27;, &#x27;TA.Early&#x27;, &#x27;Tuft&#x27;], dtype=&#x27;&lt;U21&#x27;)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>sample</span></div><div class='xr-var-dims'>(sample)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>0 1 2 3 4 5 6 7 8 9</div><input id='attrs-06a4a1c9-4a2a-407c-a909-57f0f21dfa84' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-06a4a1c9-4a2a-407c-a909-57f0f21dfa84' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-2029fa3f-8f8f-4aa6-b668-3b2e56a7a3b3' class='xr-var-data-in' type='checkbox'><label for='data-2029fa3f-8f8f-4aa6-b668-3b2e56a7a3b3' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-5428a3fb-a95e-44bd-8fb0-783356c5a5be' class='xr-section-summary-in' type='checkbox'  checked><label for='section-5428a3fb-a95e-44bd-8fb0-783356c5a5be' class='xr-section-summary' >Data variables: <span>(8)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span>sigma_d</span></div><div class='xr-var-dims'>(chain, draw, covariate, sigma_d_dim_1)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.7184 0.2872 1.23 ... 0.4437 2.847</div><input id='attrs-b236e4cd-998c-4c84-bb96-f7787a7923fc' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-b236e4cd-998c-4c84-bb96-f7787a7923fc' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-ad23fd6a-2cd4-4e25-91f0-7743b40b7c7c' class='xr-var-data-in' type='checkbox'><label for='data-ad23fd6a-2cd4-4e25-91f0-7743b40b7c7c' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[[[0.7183928 ],
         [0.28721429],
         [1.23009369]],

        [[0.7183928 ],
         [0.28721429],
         [1.23009369]],

        [[0.7183928 ],
         [0.28721429],
         [1.23009369]],

        ...,

        [[0.41285775],
         [0.10208883],
         [3.60454351]],

        [[0.41285775],
         [0.10208883],
         [3.60454351]],

        [[0.24838339],
         [0.44368568],
         [2.84706002]]]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>b_offset</span></div><div class='xr-var-dims'>(chain, draw, covariate, cell_type_nb)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>-0.7204 0.4881 ... -0.3666 0.4097</div><input id='attrs-1291c1cf-8928-4031-8c00-50f1df50ee52' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-1291c1cf-8928-4031-8c00-50f1df50ee52' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-8553fa70-5603-4b3f-9640-f492a0011131' class='xr-var-data-in' type='checkbox'><label for='data-8553fa70-5603-4b3f-9640-f492a0011131' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[[[-0.72036132,  0.48809068,  0.40063487, ...,  0.57989878,
           0.35498095, -1.80873918],
         [ 1.45264592, -0.91102803,  1.65296354, ...,  0.5928626 ,
           0.2097078 ,  1.21335494],
         [ 1.15147592,  0.73651232,  1.23028075, ...,  0.41767022,
           0.11649621, -0.30908063]],

        [[-0.72036132,  0.48809068,  0.40063487, ...,  0.57989878,
           0.35498095, -1.80873918],
         [ 1.45264592, -0.91102803,  1.65296354, ...,  0.5928626 ,
           0.2097078 ,  1.21335494],
         [ 1.15147592,  0.73651232,  1.23028075, ...,  0.41767022,
           0.11649621, -0.30908063]],

        [[-0.72036132,  0.48809068,  0.40063487, ...,  0.57989878,
           0.35498095, -1.80873918],
         [ 1.45264592, -0.91102803,  1.65296354, ...,  0.5928626 ,
           0.2097078 ,  1.21335494],
         [ 1.15147592,  0.73651232,  1.23028075, ...,  0.41767022,
           0.11649621, -0.30908063]],
...
        [[-1.02515559,  1.27905136,  0.47152558, ...,  0.27618756,
          -0.80909793,  2.10454261],
         [-0.50493417,  1.16372832,  0.27319134, ..., -0.88399111,
           1.60165557, -0.89061015],
         [ 0.44575511,  0.10482996,  0.02745633, ...,  0.07018678,
          -0.17119197,  0.39788988]],

        [[-1.02515559,  1.27905136,  0.47152558, ...,  0.27618756,
          -0.80909793,  2.10454261],
         [-0.50493417,  1.16372832,  0.27319134, ..., -0.88399111,
           1.60165557, -0.89061015],
         [ 0.44575511,  0.10482996,  0.02745633, ...,  0.07018678,
          -0.17119197,  0.39788988]],

        [[-0.43502474,  1.29071468,  0.42777639, ...,  0.55848301,
          -0.81595819,  2.03068491],
         [-0.29468284,  1.05975253,  0.4907459 , ..., -0.44289866,
           1.7972246 , -0.73725547],
         [ 0.56555757,  0.2610112 ,  0.01562117, ..., -0.27297312,
          -0.36657669,  0.40974094]]]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>ind_raw</span></div><div class='xr-var-dims'>(chain, draw, covariate, cell_type_nb)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.3983 -0.6396 ... -1.337 -1.331</div><input id='attrs-62b238c6-5ad1-444b-8f15-cb44fb211ef2' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-62b238c6-5ad1-444b-8f15-cb44fb211ef2' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-0d9c0a19-5689-4aeb-bf8d-d38798345ff2' class='xr-var-data-in' type='checkbox'><label for='data-0d9c0a19-5689-4aeb-bf8d-d38798345ff2' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[[[ 3.98285879e-01, -6.39616102e-01, -1.43969698e+00, ...,
          -6.07428437e-01, -2.03155060e+00, -1.93736323e-01],
         [-1.09760877e+00, -1.80036947e+00,  1.13296134e+00, ...,
           1.66023253e+00, -2.95659674e-01,  1.10366977e+00],
         [ 5.51144119e-01, -4.64405280e-02, -8.23004469e-01, ...,
          -2.89004463e-01, -1.76638546e+00,  3.78405759e-01]],

        [[ 3.98285879e-01, -6.39616102e-01, -1.43969698e+00, ...,
          -6.07428437e-01, -2.03155060e+00, -1.93736323e-01],
         [-1.09760877e+00, -1.80036947e+00,  1.13296134e+00, ...,
           1.66023253e+00, -2.95659674e-01,  1.10366977e+00],
         [ 5.51144119e-01, -4.64405280e-02, -8.23004469e-01, ...,
          -2.89004463e-01, -1.76638546e+00,  3.78405759e-01]],

        [[ 3.98285879e-01, -6.39616102e-01, -1.43969698e+00, ...,
          -6.07428437e-01, -2.03155060e+00, -1.93736323e-01],
         [-1.09760877e+00, -1.80036947e+00,  1.13296134e+00, ...,
           1.66023253e+00, -2.95659674e-01,  1.10366977e+00],
         [ 5.51144119e-01, -4.64405280e-02, -8.23004469e-01, ...,
          -2.89004463e-01, -1.76638546e+00,  3.78405759e-01]],
...
        [[ 6.76079885e-02,  1.19397158e-01, -6.39339160e-01, ...,
           1.65298392e+00, -1.10210939e+00, -8.25283181e-01],
         [ 1.89846055e+00,  3.30264461e-01, -7.91264418e-01, ...,
          -1.77878435e+00, -1.03228540e+00,  1.71992797e-01],
         [ 8.27668324e-01,  2.75101624e-02,  4.48084062e-01, ...,
           5.33658654e-01, -1.41290010e+00, -1.20985252e+00]],

        [[ 6.76079885e-02,  1.19397158e-01, -6.39339160e-01, ...,
           1.65298392e+00, -1.10210939e+00, -8.25283181e-01],
         [ 1.89846055e+00,  3.30264461e-01, -7.91264418e-01, ...,
          -1.77878435e+00, -1.03228540e+00,  1.71992797e-01],
         [ 8.27668324e-01,  2.75101624e-02,  4.48084062e-01, ...,
           5.33658654e-01, -1.41290010e+00, -1.20985252e+00]],

        [[ 1.68704064e-01,  6.23359873e-01, -8.00485905e-01, ...,
           1.56037013e+00, -1.22799582e+00, -6.20215735e-01],
         [ 1.28989294e+00,  1.09824372e+00, -9.96934806e-01, ...,
          -2.05813286e+00, -1.02698242e+00,  1.80377377e-01],
         [ 5.17487928e-01, -9.11387063e-02, -9.01734744e-04, ...,
           2.66378957e-01, -1.33733121e+00, -1.33105857e+00]]]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>alpha</span></div><div class='xr-var-dims'>(chain, draw, cell_type)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>1.033 2.106 2.517 ... 2.72 0.7746</div><input id='attrs-08b68c03-b97c-4fc6-9e6e-5a3eaad0487a' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-08b68c03-b97c-4fc6-9e6e-5a3eaad0487a' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-10bc6f35-f3c2-40c1-b8da-086f0d7fd821' class='xr-var-data-in' type='checkbox'><label for='data-10bc6f35-f3c2-40c1-b8da-086f0d7fd821' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[[1.03324679, 2.10618987, 2.51695856, ..., 1.97746285,
         2.82171086, 0.67258824],
        [1.03324679, 2.10618987, 2.51695856, ..., 1.97746285,
         2.82171086, 0.67258824],
        [1.03324679, 2.10618987, 2.51695856, ..., 1.97746285,
         2.82171086, 0.67258824],
        ...,
        [1.05860248, 2.12629371, 2.37622723, ..., 2.10239297,
         2.83199593, 1.19757136],
        [1.05860248, 2.12629371, 2.37622723, ..., 2.10239297,
         2.83199593, 1.19757136],
        [1.3070755 , 1.98665657, 2.37548325, ..., 2.21305882,
         2.72040604, 0.77459976]]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>ind</span></div><div class='xr-var-dims'>(chain, draw, covariate, cell_type_nb)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>1.0 1.291e-14 ... 1.249e-29</div><input id='attrs-c7eaf079-ac2c-411f-9b2e-731827cd5f6f' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-c7eaf079-ac2c-411f-9b2e-731827cd5f6f' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-ce22f982-2940-4b8e-8a93-f95a7ba0a9eb' class='xr-var-data-in' type='checkbox'><label for='data-ce22f982-2940-4b8e-8a93-f95a7ba0a9eb' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[[[9.99999998e-01, 1.29096010e-14, 5.46232104e-32, ...,
          6.45444770e-14, 7.68138108e-45, 6.20929408e-05],
         [1.46463152e-24, 8.04402933e-40, 1.00000000e+00, ...,
          1.00000000e+00, 3.80041991e-07, 1.00000000e+00],
         [1.00000000e+00, 8.93150990e-02, 1.34488462e-18, ...,
          5.30087526e-07, 4.39973975e-39, 9.99999994e-01]],

        [[9.99999998e-01, 1.29096010e-14, 5.46232104e-32, ...,
          6.45444770e-14, 7.68138108e-45, 6.20929408e-05],
         [1.46463152e-24, 8.04402933e-40, 1.00000000e+00, ...,
          1.00000000e+00, 3.80041991e-07, 1.00000000e+00],
         [1.00000000e+00, 8.93150990e-02, 1.34488462e-18, ...,
          5.30087526e-07, 4.39973975e-39, 9.99999994e-01]],

        [[9.99999998e-01, 1.29096010e-14, 5.46232104e-32, ...,
          6.45444770e-14, 7.68138108e-45, 6.20929408e-05],
         [1.46463152e-24, 8.04402933e-40, 1.00000000e+00, ...,
          1.00000000e+00, 3.80041991e-07, 1.00000000e+00],
         [1.00000000e+00, 8.93150990e-02, 1.34488462e-18, ...,
          5.30087526e-07, 4.39973975e-39, 9.99999994e-01]],
...
        [[9.67086321e-01, 9.97451905e-01, 1.30896051e-14, ...,
          1.00000000e+00, 1.16949574e-24, 1.20006128e-18],
         [1.00000000e+00, 9.99999933e-01, 6.57523791e-18, ...,
          2.36694698e-39, 3.83890399e-23, 9.99815862e-01],
         [1.00000000e+00, 7.98268616e-01, 1.00000000e+00, ...,
          1.00000000e+00, 2.08575919e-31, 5.35040146e-27]],

        [[9.67086321e-01, 9.97451905e-01, 1.30896051e-14, ...,
          1.00000000e+00, 1.16949574e-24, 1.20006128e-18],
         [1.00000000e+00, 9.99999933e-01, 6.57523791e-18, ...,
          2.36694698e-39, 3.83890399e-23, 9.99815862e-01],
         [1.00000000e+00, 7.98268616e-01, 1.00000000e+00, ...,
          1.00000000e+00, 2.08575919e-31, 5.35040146e-27]],

        [[9.99782958e-01, 1.00000000e+00, 4.14638318e-18, ...,
          1.00000000e+00, 2.15978015e-27, 3.40554352e-14],
         [1.00000000e+00, 1.00000000e+00, 2.24820411e-22, ...,
          2.03335345e-45, 5.00449344e-23, 9.99878912e-01],
         [1.00000000e+00, 1.03851871e-02, 4.88730225e-01, ...,
          9.99998357e-01, 9.12476653e-30, 1.24862054e-29]]]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>b_raw</span></div><div class='xr-var-dims'>(chain, draw, covariate, cell_type_nb)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>-0.5175 0.3506 ... -1.044 1.167</div><input id='attrs-e534018e-cd36-4fb9-92e3-cdf111994369' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-e534018e-cd36-4fb9-92e3-cdf111994369' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-d4940cd1-fcdc-4db9-81c1-a152633354be' class='xr-var-data-in' type='checkbox'><label for='data-d4940cd1-fcdc-4db9-81c1-a152633354be' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[[[-0.51750238,  0.35064083,  0.28781321, ...,  0.41659511,
           0.25501576, -1.2993852 ],
         [ 0.41722067, -0.26166027,  0.47475475, ...,  0.17027861,
           0.06023108,  0.34849288],
         [ 1.41642327,  0.90597916,  1.51336059, ...,  0.5137735 ,
           0.14330126, -0.38019813]],

        [[-0.51750238,  0.35064083,  0.28781321, ...,  0.41659511,
           0.25501576, -1.2993852 ],
         [ 0.41722067, -0.26166027,  0.47475475, ...,  0.17027861,
           0.06023108,  0.34849288],
         [ 1.41642327,  0.90597916,  1.51336059, ...,  0.5137735 ,
           0.14330126, -0.38019813]],

        [[-0.51750238,  0.35064083,  0.28781321, ...,  0.41659511,
           0.25501576, -1.2993852 ],
         [ 0.41722067, -0.26166027,  0.47475475, ...,  0.17027861,
           0.06023108,  0.34849288],
         [ 1.41642327,  0.90597916,  1.51336059, ...,  0.5137735 ,
           0.14330126, -0.38019813]],
...
        [[-0.42324344,  0.52806627,  0.19467299, ...,  0.11402617,
          -0.33404235,  0.86887674],
         [-0.05154814,  0.11880366,  0.02788978, ..., -0.09024562,
           0.16351114, -0.09092135],
         [ 1.60674369,  0.37786417,  0.09896754, ...,  0.25299132,
          -0.6170689 ,  1.43421138]],

        [[-0.42324344,  0.52806627,  0.19467299, ...,  0.11402617,
          -0.33404235,  0.86887674],
         [-0.05154814,  0.11880366,  0.02788978, ..., -0.09024562,
           0.16351114, -0.09092135],
         [ 1.60674369,  0.37786417,  0.09896754, ...,  0.25299132,
          -0.6170689 ,  1.43421138]],

        [[-0.10805292,  0.32059208,  0.10625255, ...,  0.1387179 ,
          -0.20267046,  0.50438839],
         [-0.13074656,  0.47019702,  0.21773693, ..., -0.19650779,
           0.79740281, -0.32710969],
         [ 1.61017633,  0.74311455,  0.0444744 , ..., -0.77717085,
          -1.04366585,  1.16655705]]]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>beta</span></div><div class='xr-var-dims'>(chain, draw, covariate, cell_type)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.0 -0.5175 ... 1.457e-29</div><input id='attrs-52dffbf2-f61d-4b66-91f1-b83a61b9430e' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-52dffbf2-f61d-4b66-91f1-b83a61b9430e' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-585859ce-9d50-466d-8835-ee8684f485b8' class='xr-var-data-in' type='checkbox'><label for='data-585859ce-9d50-466d-8835-ee8684f485b8' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[[[ 0.00000000e+00, -5.17502381e-01,  4.52663315e-15, ...,
           2.68889134e-14,  1.95887321e-45, -8.06826482e-05],
         [ 0.00000000e+00,  6.11074542e-25, -2.10480289e-40, ...,
           1.70278612e-01,  2.28903384e-08,  3.48492877e-01],
         [ 0.00000000e+00,  1.41642327e+00,  8.09176182e-02, ...,
           2.72344925e-07,  6.30488238e-40, -3.80198130e-01]],

        [[ 0.00000000e+00, -5.17502381e-01,  4.52663315e-15, ...,
           2.68889134e-14,  1.95887321e-45, -8.06826482e-05],
         [ 0.00000000e+00,  6.11074542e-25, -2.10480289e-40, ...,
           1.70278612e-01,  2.28903384e-08,  3.48492877e-01],
         [ 0.00000000e+00,  1.41642327e+00,  8.09176182e-02, ...,
           2.72344925e-07,  6.30488238e-40, -3.80198130e-01]],

        [[ 0.00000000e+00, -5.17502381e-01,  4.52663315e-15, ...,
           2.68889134e-14,  1.95887321e-45, -8.06826482e-05],
         [ 0.00000000e+00,  6.11074542e-25, -2.10480289e-40, ...,
           1.70278612e-01,  2.28903384e-08,  3.48492877e-01],
         [ 0.00000000e+00,  1.41642327e+00,  8.09176182e-02, ...,
           2.72344925e-07,  6.30488238e-40, -3.80198130e-01]],
...
        [[ 0.00000000e+00, -4.09312937e-01,  5.26720710e-01, ...,
           1.14026174e-01, -3.90661110e-25,  1.04270533e-18],
         [ 0.00000000e+00, -5.15481370e-02,  1.18803652e-01, ...,
          -2.13606588e-40,  6.27703567e-24, -9.09046034e-02],
         [ 0.00000000e+00,  1.60674369e+00,  3.01637106e-01, ...,
           2.52991315e-01, -1.28705713e-31,  7.67360664e-27]],

        [[ 0.00000000e+00, -4.09312937e-01,  5.26720710e-01, ...,
           1.14026174e-01, -3.90661110e-25,  1.04270533e-18],
         [ 0.00000000e+00, -5.15481370e-02,  1.18803652e-01, ...,
          -2.13606588e-40,  6.27703567e-24, -9.09046034e-02],
         [ 0.00000000e+00,  1.60674369e+00,  3.01637106e-01, ...,
           2.52991315e-01, -1.28705713e-31,  7.67360664e-27]],

        [[ 0.00000000e+00, -1.08029465e-01,  3.20592083e-01, ...,
           1.38717901e-01, -4.37723632e-28,  1.71771663e-14],
         [ 0.00000000e+00, -1.30746556e-01,  4.70197020e-01, ...,
          -3.99569793e-46,  3.99059715e-23, -3.27070083e-01],
         [ 0.00000000e+00,  1.61017633e+00,  7.71738367e-03, ...,
          -7.77169574e-01, -9.52320718e-30,  1.45658710e-29]]]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>concentration</span></div><div class='xr-var-dims'>(chain, draw, sample, cell_type)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>2.81 8.217 12.39 ... 15.19 2.17</div><input id='attrs-75e540c0-f3d5-489f-8027-16b2a3f56af5' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-75e540c0-f3d5-489f-8027-16b2a3f56af5' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-c5e54b27-5ecf-4f9d-a870-3ce656c49a29' class='xr-var-data-in' type='checkbox'><label for='data-c5e54b27-5ecf-4f9d-a870-3ce656c49a29' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[[[ 2.81017508,  8.21687422, 12.39085323, ...,  7.22439033,
          16.80557803,  1.9593019 ],
         [ 2.81017508,  8.21687422, 12.39085323, ...,  7.22439033,
          16.80557803,  1.9593019 ],
         [ 2.81017508,  8.21687422, 12.39085323, ...,  7.22439033,
          16.80557803,  1.9593019 ],
         ...,
         [ 2.81017508,  4.89731693, 12.39085323, ...,  7.22439033,
          16.80557803,  1.95914382],
         [ 2.81017508, 33.87282726, 13.43517377, ...,  7.22439229,
          16.80557803,  1.33962551],
         [ 2.81017508, 33.87282726, 13.43517377, ...,  7.22439229,
          16.80557803,  1.33962551]],

        [[ 2.81017508,  8.21687422, 12.39085323, ...,  7.22439033,
          16.80557803,  1.9593019 ],
         [ 2.81017508,  8.21687422, 12.39085323, ...,  7.22439033,
          16.80557803,  1.9593019 ],
         [ 2.81017508,  8.21687422, 12.39085323, ...,  7.22439033,
          16.80557803,  1.9593019 ],
...
         [ 2.88234004,  5.56769293, 18.22780072, ...,  9.17441967,
          16.97931656,  3.31206334],
         [ 2.88234004, 41.80589653, 14.55397768, ..., 10.54217925,
          16.97931656,  3.31206334],
         [ 2.88234004, 41.80589653, 14.55397768, ..., 10.54217925,
          16.97931656,  3.31206334]],

        [[ 3.69535085,  7.29111562, 10.75620988, ...,  9.14364239,
          15.18648732,  2.16972354],
         [ 3.69535085,  7.29111562, 10.75620988, ...,  9.14364239,
          15.18648732,  2.16972354],
         [ 3.69535085,  7.29111562, 10.75620988, ...,  9.14364239,
          15.18648732,  2.16972354],
         ...,
         [ 3.69535085,  6.54451375, 14.82144819, ..., 10.50421619,
          15.18648732,  2.16972354],
         [ 3.69535085, 36.48250758, 10.83954081, ...,  4.20338118,
          15.18648732,  2.16972354],
         [ 3.69535085, 36.48250758, 10.83954081, ...,  4.20338118,
          15.18648732,  2.16972354]]]])</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-0fd00d05-424e-4a58-a94c-77cc660fca0e' class='xr-section-summary-in' type='checkbox'  checked><label for='section-0fd00d05-424e-4a58-a94c-77cc660fca0e' class='xr-section-summary' >Attributes: <span>(2)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><dl class='xr-attrs'><dt><span>created_at :</span></dt><dd>2022-08-25T13:07:26.604429</dd><dt><span>arviz_version :</span></dt><dd>0.12.1</dd></dl></div></li></ul></div></div><br></div>
                      </ul>
                  </div>
            </li>

            <li class = "xr-section-item">
                  <input id="idata_posterior_predictiveed12612c-7069-45f5-93f9-86c721030cf4" class="xr-section-summary-in" type="checkbox">
                  <label for="idata_posterior_predictiveed12612c-7069-45f5-93f9-86c721030cf4" class = "xr-section-summary">posterior_predictive</label>
                  <div class="xr-section-inline-details"></div>
                  <div class="xr-section-details">
                      <ul id="xr-dataset-coord-list" class="xr-var-list">
                          <div style="padding-left:2rem;"><div><svg style="position: absolute; width: 0; height: 0; overflow: hidden">
<defs>
<symbol id="icon-database" viewBox="0 0 32 32">
<path d="M16 0c-8.837 0-16 2.239-16 5v4c0 2.761 7.163 5 16 5s16-2.239 16-5v-4c0-2.761-7.163-5-16-5z"></path>
<path d="M16 17c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z"></path>
<path d="M16 26c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z"></path>
</symbol>
<symbol id="icon-file-text2" viewBox="0 0 32 32">
<path d="M28.681 7.159c-0.694-0.947-1.662-2.053-2.724-3.116s-2.169-2.030-3.116-2.724c-1.612-1.182-2.393-1.319-2.841-1.319h-15.5c-1.378 0-2.5 1.121-2.5 2.5v27c0 1.378 1.122 2.5 2.5 2.5h23c1.378 0 2.5-1.122 2.5-2.5v-19.5c0-0.448-0.137-1.23-1.319-2.841zM24.543 5.457c0.959 0.959 1.712 1.825 2.268 2.543h-4.811v-4.811c0.718 0.556 1.584 1.309 2.543 2.268zM28 29.5c0 0.271-0.229 0.5-0.5 0.5h-23c-0.271 0-0.5-0.229-0.5-0.5v-27c0-0.271 0.229-0.5 0.5-0.5 0 0 15.499-0 15.5 0v7c0 0.552 0.448 1 1 1h7v19.5z"></path>
<path d="M23 26h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z"></path>
<path d="M23 22h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z"></path>
<path d="M23 18h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z"></path>
</symbol>
</defs>
</svg>
<style>/* CSS stylesheet for displaying xarray objects in jupyterlab.
 *
 */

:root {
  --xr-font-color0: var(--jp-content-font-color0, rgba(0, 0, 0, 1));
  --xr-font-color2: var(--jp-content-font-color2, rgba(0, 0, 0, 0.54));
  --xr-font-color3: var(--jp-content-font-color3, rgba(0, 0, 0, 0.38));
  --xr-border-color: var(--jp-border-color2, #e0e0e0);
  --xr-disabled-color: var(--jp-layout-color3, #bdbdbd);
  --xr-background-color: var(--jp-layout-color0, white);
  --xr-background-color-row-even: var(--jp-layout-color1, white);
  --xr-background-color-row-odd: var(--jp-layout-color2, #eeeeee);
}

html[theme=dark],
body[data-theme=dark],
body.vscode-dark {
  --xr-font-color0: rgba(255, 255, 255, 1);
  --xr-font-color2: rgba(255, 255, 255, 0.54);
  --xr-font-color3: rgba(255, 255, 255, 0.38);
  --xr-border-color: #1F1F1F;
  --xr-disabled-color: #515151;
  --xr-background-color: #111111;
  --xr-background-color-row-even: #111111;
  --xr-background-color-row-odd: #313131;
}

.xr-wrap {
  display: block !important;
  min-width: 300px;
  max-width: 700px;
}

.xr-text-repr-fallback {
  /* fallback to plain text repr when CSS is not injected (untrusted notebook) */
  display: none;
}

.xr-header {
  padding-top: 6px;
  padding-bottom: 6px;
  margin-bottom: 4px;
  border-bottom: solid 1px var(--xr-border-color);
}

.xr-header > div,
.xr-header > ul {
  display: inline;
  margin-top: 0;
  margin-bottom: 0;
}

.xr-obj-type,
.xr-array-name {
  margin-left: 2px;
  margin-right: 10px;
}

.xr-obj-type {
  color: var(--xr-font-color2);
}

.xr-sections {
  padding-left: 0 !important;
  display: grid;
  grid-template-columns: 150px auto auto 1fr 20px 20px;
}

.xr-section-item {
  display: contents;
}

.xr-section-item input {
  display: none;
}

.xr-section-item input + label {
  color: var(--xr-disabled-color);
}

.xr-section-item input:enabled + label {
  cursor: pointer;
  color: var(--xr-font-color2);
}

.xr-section-item input:enabled + label:hover {
  color: var(--xr-font-color0);
}

.xr-section-summary {
  grid-column: 1;
  color: var(--xr-font-color2);
  font-weight: 500;
}

.xr-section-summary > span {
  display: inline-block;
  padding-left: 0.5em;
}

.xr-section-summary-in:disabled + label {
  color: var(--xr-font-color2);
}

.xr-section-summary-in + label:before {
  display: inline-block;
  content: '►';
  font-size: 11px;
  width: 15px;
  text-align: center;
}

.xr-section-summary-in:disabled + label:before {
  color: var(--xr-disabled-color);
}

.xr-section-summary-in:checked + label:before {
  content: '▼';
}

.xr-section-summary-in:checked + label > span {
  display: none;
}

.xr-section-summary,
.xr-section-inline-details {
  padding-top: 4px;
  padding-bottom: 4px;
}

.xr-section-inline-details {
  grid-column: 2 / -1;
}

.xr-section-details {
  display: none;
  grid-column: 1 / -1;
  margin-bottom: 5px;
}

.xr-section-summary-in:checked ~ .xr-section-details {
  display: contents;
}

.xr-array-wrap {
  grid-column: 1 / -1;
  display: grid;
  grid-template-columns: 20px auto;
}

.xr-array-wrap > label {
  grid-column: 1;
  vertical-align: top;
}

.xr-preview {
  color: var(--xr-font-color3);
}

.xr-array-preview,
.xr-array-data {
  padding: 0 5px !important;
  grid-column: 2;
}

.xr-array-data,
.xr-array-in:checked ~ .xr-array-preview {
  display: none;
}

.xr-array-in:checked ~ .xr-array-data,
.xr-array-preview {
  display: inline-block;
}

.xr-dim-list {
  display: inline-block !important;
  list-style: none;
  padding: 0 !important;
  margin: 0;
}

.xr-dim-list li {
  display: inline-block;
  padding: 0;
  margin: 0;
}

.xr-dim-list:before {
  content: '(';
}

.xr-dim-list:after {
  content: ')';
}

.xr-dim-list li:not(:last-child):after {
  content: ',';
  padding-right: 5px;
}

.xr-has-index {
  font-weight: bold;
}

.xr-var-list,
.xr-var-item {
  display: contents;
}

.xr-var-item > div,
.xr-var-item label,
.xr-var-item > .xr-var-name span {
  background-color: var(--xr-background-color-row-even);
  margin-bottom: 0;
}

.xr-var-item > .xr-var-name:hover span {
  padding-right: 5px;
}

.xr-var-list > li:nth-child(odd) > div,
.xr-var-list > li:nth-child(odd) > label,
.xr-var-list > li:nth-child(odd) > .xr-var-name span {
  background-color: var(--xr-background-color-row-odd);
}

.xr-var-name {
  grid-column: 1;
}

.xr-var-dims {
  grid-column: 2;
}

.xr-var-dtype {
  grid-column: 3;
  text-align: right;
  color: var(--xr-font-color2);
}

.xr-var-preview {
  grid-column: 4;
}

.xr-var-name,
.xr-var-dims,
.xr-var-dtype,
.xr-preview,
.xr-attrs dt {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  padding-right: 10px;
}

.xr-var-name:hover,
.xr-var-dims:hover,
.xr-var-dtype:hover,
.xr-attrs dt:hover {
  overflow: visible;
  width: auto;
  z-index: 1;
}

.xr-var-attrs,
.xr-var-data {
  display: none;
  background-color: var(--xr-background-color) !important;
  padding-bottom: 5px !important;
}

.xr-var-attrs-in:checked ~ .xr-var-attrs,
.xr-var-data-in:checked ~ .xr-var-data {
  display: block;
}

.xr-var-data > table {
  float: right;
}

.xr-var-name span,
.xr-var-data,
.xr-attrs {
  padding-left: 25px !important;
}

.xr-attrs,
.xr-var-attrs,
.xr-var-data {
  grid-column: 1 / -1;
}

dl.xr-attrs {
  padding: 0;
  margin: 0;
  display: grid;
  grid-template-columns: 125px auto;
}

.xr-attrs dt,
.xr-attrs dd {
  padding: 0;
  margin: 0;
  float: left;
  padding-right: 10px;
  width: auto;
}

.xr-attrs dt {
  font-weight: normal;
  grid-column: 1;
}

.xr-attrs dt:hover span {
  display: inline-block;
  background: var(--xr-background-color);
  padding-right: 10px;
}

.xr-attrs dd {
  grid-column: 2;
  white-space: pre-wrap;
  word-break: break-all;
}

.xr-icon-database,
.xr-icon-file-text2 {
  display: inline-block;
  vertical-align: middle;
  width: 1em;
  height: 1.5em !important;
  stroke-width: 0;
  stroke: currentColor;
  fill: currentColor;
}
</style><pre class='xr-text-repr-fallback'>&lt;xarray.Dataset&gt;
Dimensions:     (chain: 1, draw: 15000, sample: 10, cell_type: 8)
Coordinates:
  * chain       (chain) int64 0
  * draw        (draw) int64 0 1 2 3 4 5 ... 14994 14995 14996 14997 14998 14999
  * sample      (sample) int64 0 1 2 3 4 5 6 7 8 9
  * cell_type   (cell_type) &lt;U21 &#x27;Endocrine&#x27; &#x27;Enterocyte&#x27; ... &#x27;TA.Early&#x27; &#x27;Tuft&#x27;
Data variables:
    prediction  (chain, draw, sample, cell_type) float64 35.62 104.2 ... 22.26
Attributes:
    created_at:     2022-08-25T13:07:26.620810
    arviz_version:  0.12.1</pre><div class='xr-wrap' style='display:none'><div class='xr-header'><div class='xr-obj-type'>xarray.Dataset</div></div><ul class='xr-sections'><li class='xr-section-item'><input id='section-be3bd124-3c6a-47ee-897b-a142c18bb571' class='xr-section-summary-in' type='checkbox' disabled ><label for='section-be3bd124-3c6a-47ee-897b-a142c18bb571' class='xr-section-summary'  title='Expand/collapse section'>Dimensions:</label><div class='xr-section-inline-details'><ul class='xr-dim-list'><li><span class='xr-has-index'>chain</span>: 1</li><li><span class='xr-has-index'>draw</span>: 15000</li><li><span class='xr-has-index'>sample</span>: 10</li><li><span class='xr-has-index'>cell_type</span>: 8</li></ul></div><div class='xr-section-details'></div></li><li class='xr-section-item'><input id='section-09a3baa4-6e93-4cb1-baef-d4cc45d3697c' class='xr-section-summary-in' type='checkbox'  checked><label for='section-09a3baa4-6e93-4cb1-baef-d4cc45d3697c' class='xr-section-summary' >Coordinates: <span>(4)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>chain</span></div><div class='xr-var-dims'>(chain)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>0</div><input id='attrs-85d42a46-fff3-4f06-bfca-7eca62cc5b4f' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-85d42a46-fff3-4f06-bfca-7eca62cc5b4f' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-78a81393-cfaa-4deb-a89d-f77ef885f93b' class='xr-var-data-in' type='checkbox'><label for='data-78a81393-cfaa-4deb-a89d-f77ef885f93b' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([0])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>draw</span></div><div class='xr-var-dims'>(draw)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>0 1 2 3 ... 14996 14997 14998 14999</div><input id='attrs-e23391fe-b36a-4d23-9907-0be96fbea68e' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-e23391fe-b36a-4d23-9907-0be96fbea68e' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-fffc259e-8a53-40c2-a3e6-9784d8f564ed' class='xr-var-data-in' type='checkbox'><label for='data-fffc259e-8a53-40c2-a3e6-9784d8f564ed' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([    0,     1,     2, ..., 14997, 14998, 14999])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>sample</span></div><div class='xr-var-dims'>(sample)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>0 1 2 3 4 5 6 7 8 9</div><input id='attrs-77414dc5-26cb-48e5-a8a0-63e228f91a9d' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-77414dc5-26cb-48e5-a8a0-63e228f91a9d' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-449abb9a-b143-43a8-9ba3-d346a4c4c1bf' class='xr-var-data-in' type='checkbox'><label for='data-449abb9a-b143-43a8-9ba3-d346a4c4c1bf' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>cell_type</span></div><div class='xr-var-dims'>(cell_type)</div><div class='xr-var-dtype'>&lt;U21</div><div class='xr-var-preview xr-preview'>&#x27;Endocrine&#x27; &#x27;Enterocyte&#x27; ... &#x27;Tuft&#x27;</div><input id='attrs-62fa73bd-2a19-4150-bc78-27fe24b33be3' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-62fa73bd-2a19-4150-bc78-27fe24b33be3' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-0213a5b7-cd76-4045-8a68-200f781e6d76' class='xr-var-data-in' type='checkbox'><label for='data-0213a5b7-cd76-4045-8a68-200f781e6d76' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;Endocrine&#x27;, &#x27;Enterocyte&#x27;, &#x27;Enterocyte.Progenitor&#x27;, &#x27;Goblet&#x27;, &#x27;Stem&#x27;,
       &#x27;TA&#x27;, &#x27;TA.Early&#x27;, &#x27;Tuft&#x27;], dtype=&#x27;&lt;U21&#x27;)</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-f2d23cdc-8492-4e10-bd26-b1b5a7729ed7' class='xr-section-summary-in' type='checkbox'  checked><label for='section-f2d23cdc-8492-4e10-bd26-b1b5a7729ed7' class='xr-section-summary' >Data variables: <span>(1)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span>prediction</span></div><div class='xr-var-dims'>(chain, draw, sample, cell_type)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>35.62 104.2 157.1 ... 155.8 22.26</div><input id='attrs-3967b36a-815f-45a9-9443-d2762039793e' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-3967b36a-815f-45a9-9443-d2762039793e' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-8af072bb-a2d8-4b60-8b14-af1637838005' class='xr-var-data-in' type='checkbox'><label for='data-8af072bb-a2d8-4b60-8b14-af1637838005' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[[[ 35.6248478 , 104.16606965, 157.07998517, ...,  91.58426013,
          213.04585724,  24.83825022],
         [  8.48210662,  24.80144515,  37.39999647, ...,  21.80577622,
           50.7252041 ,   5.9138691 ],
         [ 53.35245064, 156.00109002, 235.2459778 , ..., 137.15833244,
          319.06153382,  37.19823664],
         ...,
         [ 68.84626764, 119.97899858, 303.56257972, ..., 176.98979432,
          411.71859005,  47.99691689],
         [ 24.9569678 , 300.82220301, 119.31683585, ...,  64.15932124,
          149.24915973,  11.89712019],
         [ 28.91356026, 348.51352788, 138.23291958, ...,  74.33092094,
          172.91061189,  13.78324901]],

        [[ 35.6248478 , 104.16606965, 157.07998517, ...,  91.58426013,
          213.04585724,  24.83825022],
         [  8.48210662,  24.80144515,  37.39999647, ...,  21.80577622,
           50.7252041 ,   5.9138691 ],
         [ 53.35245064, 156.00109002, 235.2459778 , ..., 137.15833244,
          319.06153382,  37.19823664],
...
         [ 55.54719421, 107.29813852, 351.27818821, ..., 176.80539572,
          327.21794856,  63.82863339],
         [ 21.60203659, 313.31921101, 109.07649836, ...,  79.00960289,
          127.25348595,  24.82264845],
         [ 25.02674971, 362.99176885, 126.36911396, ...,  91.53551555,
          147.42781909,  28.75794637]],

        [[ 46.70848814,  92.15822838, 135.95631972, ..., 115.57379252,
          191.95413158,  27.42486721],
         [ 11.12106861,  21.94243533,  32.37055231, ...,  27.51756965,
           45.70336466,   6.52973029],
         [ 69.95152153, 138.01791822, 203.61077406, ..., 173.08551308,
          287.47416373,  41.07200351],
         ...,
         [ 80.10337268, 141.8641004 , 321.28153369, ..., 227.69776897,
          329.19441306,  47.03265822],
         [ 32.7183248 , 323.01304561,  95.97238026, ...,  37.21638249,
          134.45987813,  19.21054927],
         [ 37.90537629, 374.22243089, 111.18751372, ...,  43.11654069,
          155.77668808,  22.25612416]]]])</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-cc6582cb-c32d-4457-8de0-7246a4c59fa1' class='xr-section-summary-in' type='checkbox'  checked><label for='section-cc6582cb-c32d-4457-8de0-7246a4c59fa1' class='xr-section-summary' >Attributes: <span>(2)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><dl class='xr-attrs'><dt><span>created_at :</span></dt><dd>2022-08-25T13:07:26.620810</dd><dt><span>arviz_version :</span></dt><dd>0.12.1</dd></dl></div></li></ul></div></div><br></div>
                      </ul>
                  </div>
            </li>

            <li class = "xr-section-item">
                  <input id="idata_sample_statsc7802028-45e5-4cd4-9184-1cb7fa7e7dc9" class="xr-section-summary-in" type="checkbox">
                  <label for="idata_sample_statsc7802028-45e5-4cd4-9184-1cb7fa7e7dc9" class = "xr-section-summary">sample_stats</label>
                  <div class="xr-section-inline-details"></div>
                  <div class="xr-section-details">
                      <ul id="xr-dataset-coord-list" class="xr-var-list">
                          <div style="padding-left:2rem;"><div><svg style="position: absolute; width: 0; height: 0; overflow: hidden">
<defs>
<symbol id="icon-database" viewBox="0 0 32 32">
<path d="M16 0c-8.837 0-16 2.239-16 5v4c0 2.761 7.163 5 16 5s16-2.239 16-5v-4c0-2.761-7.163-5-16-5z"></path>
<path d="M16 17c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z"></path>
<path d="M16 26c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z"></path>
</symbol>
<symbol id="icon-file-text2" viewBox="0 0 32 32">
<path d="M28.681 7.159c-0.694-0.947-1.662-2.053-2.724-3.116s-2.169-2.030-3.116-2.724c-1.612-1.182-2.393-1.319-2.841-1.319h-15.5c-1.378 0-2.5 1.121-2.5 2.5v27c0 1.378 1.122 2.5 2.5 2.5h23c1.378 0 2.5-1.122 2.5-2.5v-19.5c0-0.448-0.137-1.23-1.319-2.841zM24.543 5.457c0.959 0.959 1.712 1.825 2.268 2.543h-4.811v-4.811c0.718 0.556 1.584 1.309 2.543 2.268zM28 29.5c0 0.271-0.229 0.5-0.5 0.5h-23c-0.271 0-0.5-0.229-0.5-0.5v-27c0-0.271 0.229-0.5 0.5-0.5 0 0 15.499-0 15.5 0v7c0 0.552 0.448 1 1 1h7v19.5z"></path>
<path d="M23 26h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z"></path>
<path d="M23 22h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z"></path>
<path d="M23 18h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z"></path>
</symbol>
</defs>
</svg>
<style>/* CSS stylesheet for displaying xarray objects in jupyterlab.
 *
 */

:root {
  --xr-font-color0: var(--jp-content-font-color0, rgba(0, 0, 0, 1));
  --xr-font-color2: var(--jp-content-font-color2, rgba(0, 0, 0, 0.54));
  --xr-font-color3: var(--jp-content-font-color3, rgba(0, 0, 0, 0.38));
  --xr-border-color: var(--jp-border-color2, #e0e0e0);
  --xr-disabled-color: var(--jp-layout-color3, #bdbdbd);
  --xr-background-color: var(--jp-layout-color0, white);
  --xr-background-color-row-even: var(--jp-layout-color1, white);
  --xr-background-color-row-odd: var(--jp-layout-color2, #eeeeee);
}

html[theme=dark],
body[data-theme=dark],
body.vscode-dark {
  --xr-font-color0: rgba(255, 255, 255, 1);
  --xr-font-color2: rgba(255, 255, 255, 0.54);
  --xr-font-color3: rgba(255, 255, 255, 0.38);
  --xr-border-color: #1F1F1F;
  --xr-disabled-color: #515151;
  --xr-background-color: #111111;
  --xr-background-color-row-even: #111111;
  --xr-background-color-row-odd: #313131;
}

.xr-wrap {
  display: block !important;
  min-width: 300px;
  max-width: 700px;
}

.xr-text-repr-fallback {
  /* fallback to plain text repr when CSS is not injected (untrusted notebook) */
  display: none;
}

.xr-header {
  padding-top: 6px;
  padding-bottom: 6px;
  margin-bottom: 4px;
  border-bottom: solid 1px var(--xr-border-color);
}

.xr-header > div,
.xr-header > ul {
  display: inline;
  margin-top: 0;
  margin-bottom: 0;
}

.xr-obj-type,
.xr-array-name {
  margin-left: 2px;
  margin-right: 10px;
}

.xr-obj-type {
  color: var(--xr-font-color2);
}

.xr-sections {
  padding-left: 0 !important;
  display: grid;
  grid-template-columns: 150px auto auto 1fr 20px 20px;
}

.xr-section-item {
  display: contents;
}

.xr-section-item input {
  display: none;
}

.xr-section-item input + label {
  color: var(--xr-disabled-color);
}

.xr-section-item input:enabled + label {
  cursor: pointer;
  color: var(--xr-font-color2);
}

.xr-section-item input:enabled + label:hover {
  color: var(--xr-font-color0);
}

.xr-section-summary {
  grid-column: 1;
  color: var(--xr-font-color2);
  font-weight: 500;
}

.xr-section-summary > span {
  display: inline-block;
  padding-left: 0.5em;
}

.xr-section-summary-in:disabled + label {
  color: var(--xr-font-color2);
}

.xr-section-summary-in + label:before {
  display: inline-block;
  content: '►';
  font-size: 11px;
  width: 15px;
  text-align: center;
}

.xr-section-summary-in:disabled + label:before {
  color: var(--xr-disabled-color);
}

.xr-section-summary-in:checked + label:before {
  content: '▼';
}

.xr-section-summary-in:checked + label > span {
  display: none;
}

.xr-section-summary,
.xr-section-inline-details {
  padding-top: 4px;
  padding-bottom: 4px;
}

.xr-section-inline-details {
  grid-column: 2 / -1;
}

.xr-section-details {
  display: none;
  grid-column: 1 / -1;
  margin-bottom: 5px;
}

.xr-section-summary-in:checked ~ .xr-section-details {
  display: contents;
}

.xr-array-wrap {
  grid-column: 1 / -1;
  display: grid;
  grid-template-columns: 20px auto;
}

.xr-array-wrap > label {
  grid-column: 1;
  vertical-align: top;
}

.xr-preview {
  color: var(--xr-font-color3);
}

.xr-array-preview,
.xr-array-data {
  padding: 0 5px !important;
  grid-column: 2;
}

.xr-array-data,
.xr-array-in:checked ~ .xr-array-preview {
  display: none;
}

.xr-array-in:checked ~ .xr-array-data,
.xr-array-preview {
  display: inline-block;
}

.xr-dim-list {
  display: inline-block !important;
  list-style: none;
  padding: 0 !important;
  margin: 0;
}

.xr-dim-list li {
  display: inline-block;
  padding: 0;
  margin: 0;
}

.xr-dim-list:before {
  content: '(';
}

.xr-dim-list:after {
  content: ')';
}

.xr-dim-list li:not(:last-child):after {
  content: ',';
  padding-right: 5px;
}

.xr-has-index {
  font-weight: bold;
}

.xr-var-list,
.xr-var-item {
  display: contents;
}

.xr-var-item > div,
.xr-var-item label,
.xr-var-item > .xr-var-name span {
  background-color: var(--xr-background-color-row-even);
  margin-bottom: 0;
}

.xr-var-item > .xr-var-name:hover span {
  padding-right: 5px;
}

.xr-var-list > li:nth-child(odd) > div,
.xr-var-list > li:nth-child(odd) > label,
.xr-var-list > li:nth-child(odd) > .xr-var-name span {
  background-color: var(--xr-background-color-row-odd);
}

.xr-var-name {
  grid-column: 1;
}

.xr-var-dims {
  grid-column: 2;
}

.xr-var-dtype {
  grid-column: 3;
  text-align: right;
  color: var(--xr-font-color2);
}

.xr-var-preview {
  grid-column: 4;
}

.xr-var-name,
.xr-var-dims,
.xr-var-dtype,
.xr-preview,
.xr-attrs dt {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  padding-right: 10px;
}

.xr-var-name:hover,
.xr-var-dims:hover,
.xr-var-dtype:hover,
.xr-attrs dt:hover {
  overflow: visible;
  width: auto;
  z-index: 1;
}

.xr-var-attrs,
.xr-var-data {
  display: none;
  background-color: var(--xr-background-color) !important;
  padding-bottom: 5px !important;
}

.xr-var-attrs-in:checked ~ .xr-var-attrs,
.xr-var-data-in:checked ~ .xr-var-data {
  display: block;
}

.xr-var-data > table {
  float: right;
}

.xr-var-name span,
.xr-var-data,
.xr-attrs {
  padding-left: 25px !important;
}

.xr-attrs,
.xr-var-attrs,
.xr-var-data {
  grid-column: 1 / -1;
}

dl.xr-attrs {
  padding: 0;
  margin: 0;
  display: grid;
  grid-template-columns: 125px auto;
}

.xr-attrs dt,
.xr-attrs dd {
  padding: 0;
  margin: 0;
  float: left;
  padding-right: 10px;
  width: auto;
}

.xr-attrs dt {
  font-weight: normal;
  grid-column: 1;
}

.xr-attrs dt:hover span {
  display: inline-block;
  background: var(--xr-background-color);
  padding-right: 10px;
}

.xr-attrs dd {
  grid-column: 2;
  white-space: pre-wrap;
  word-break: break-all;
}

.xr-icon-database,
.xr-icon-file-text2 {
  display: inline-block;
  vertical-align: middle;
  width: 1em;
  height: 1.5em !important;
  stroke-width: 0;
  stroke: currentColor;
  fill: currentColor;
}
</style><pre class='xr-text-repr-fallback'>&lt;xarray.Dataset&gt;
Dimensions:          (chain: 1, draw: 15000)
Coordinates:
  * chain            (chain) int64 0
  * draw             (draw) int64 0 1 2 3 4 5 ... 14995 14996 14997 14998 14999
Data variables:
    target_log_prob  (chain, draw) float64 -420.8 -420.8 ... -420.1 -420.7
    diverging        (chain, draw) bool True True True True ... False True False
    is_accepted      (chain, draw) bool False False False ... True False True
    step_size        (chain, draw) float64 0.03172 0.03172 ... 0.03172 0.03172
Attributes:
    created_at:     2022-08-25T13:07:26.612449
    arviz_version:  0.12.1</pre><div class='xr-wrap' style='display:none'><div class='xr-header'><div class='xr-obj-type'>xarray.Dataset</div></div><ul class='xr-sections'><li class='xr-section-item'><input id='section-1e9d2310-a1c7-4835-8348-b9287dc2cdee' class='xr-section-summary-in' type='checkbox' disabled ><label for='section-1e9d2310-a1c7-4835-8348-b9287dc2cdee' class='xr-section-summary'  title='Expand/collapse section'>Dimensions:</label><div class='xr-section-inline-details'><ul class='xr-dim-list'><li><span class='xr-has-index'>chain</span>: 1</li><li><span class='xr-has-index'>draw</span>: 15000</li></ul></div><div class='xr-section-details'></div></li><li class='xr-section-item'><input id='section-356f3e30-7117-4c9e-8485-2b5d4589b0e0' class='xr-section-summary-in' type='checkbox'  checked><label for='section-356f3e30-7117-4c9e-8485-2b5d4589b0e0' class='xr-section-summary' >Coordinates: <span>(2)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>chain</span></div><div class='xr-var-dims'>(chain)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>0</div><input id='attrs-5c6c12ae-aac1-41e8-86b3-efe0d601a70a' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-5c6c12ae-aac1-41e8-86b3-efe0d601a70a' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-9c7cef20-d500-4ed6-b635-55fe5b1dd035' class='xr-var-data-in' type='checkbox'><label for='data-9c7cef20-d500-4ed6-b635-55fe5b1dd035' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([0])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>draw</span></div><div class='xr-var-dims'>(draw)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>0 1 2 3 ... 14996 14997 14998 14999</div><input id='attrs-0e81fe28-5bdd-43b2-8f17-96a9fd2f5974' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-0e81fe28-5bdd-43b2-8f17-96a9fd2f5974' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-e2036926-961b-4cd4-8c97-51b8ef0a3e69' class='xr-var-data-in' type='checkbox'><label for='data-e2036926-961b-4cd4-8c97-51b8ef0a3e69' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([    0,     1,     2, ..., 14997, 14998, 14999])</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-80797644-4f0a-4fe7-84ec-78eda55978a6' class='xr-section-summary-in' type='checkbox'  checked><label for='section-80797644-4f0a-4fe7-84ec-78eda55978a6' class='xr-section-summary' >Data variables: <span>(4)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span>target_log_prob</span></div><div class='xr-var-dims'>(chain, draw)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>-420.8 -420.8 ... -420.1 -420.7</div><input id='attrs-5abedbcf-0790-4315-8cc9-c99ee21848b6' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-5abedbcf-0790-4315-8cc9-c99ee21848b6' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-ae0b04b3-e9c4-4e0e-8416-3175c82b5c30' class='xr-var-data-in' type='checkbox'><label for='data-ae0b04b3-e9c4-4e0e-8416-3175c82b5c30' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[-420.77255402, -420.77255402, -420.77255402, ..., -420.13972707,
        -420.13972707, -420.73493243]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>diverging</span></div><div class='xr-var-dims'>(chain, draw)</div><div class='xr-var-dtype'>bool</div><div class='xr-var-preview xr-preview'>True True True ... False True False</div><input id='attrs-a8502786-7ac6-40e3-b3c8-d6d61a6eae9d' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-a8502786-7ac6-40e3-b3c8-d6d61a6eae9d' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-3212bc09-98fb-4a12-a7fc-fb4eef3bb474' class='xr-var-data-in' type='checkbox'><label for='data-3212bc09-98fb-4a12-a7fc-fb4eef3bb474' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[ True,  True,  True, ..., False,  True, False]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>is_accepted</span></div><div class='xr-var-dims'>(chain, draw)</div><div class='xr-var-dtype'>bool</div><div class='xr-var-preview xr-preview'>False False False ... False True</div><input id='attrs-b739adfc-57ac-4d9b-8915-1cff9eb571b0' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-b739adfc-57ac-4d9b-8915-1cff9eb571b0' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-5e6d169d-2a02-4588-92fd-a43c46b7c4e0' class='xr-var-data-in' type='checkbox'><label for='data-5e6d169d-2a02-4588-92fd-a43c46b7c4e0' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[False, False, False, ...,  True, False,  True]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>step_size</span></div><div class='xr-var-dims'>(chain, draw)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.03172 0.03172 ... 0.03172 0.03172</div><input id='attrs-45abe4e1-fc86-424d-93bf-c60fad888940' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-45abe4e1-fc86-424d-93bf-c60fad888940' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-8ec49bd2-6681-48fd-a170-0156a2ad749b' class='xr-var-data-in' type='checkbox'><label for='data-8ec49bd2-6681-48fd-a170-0156a2ad749b' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[0.03171607, 0.03171607, 0.03171607, ..., 0.03171607, 0.03171607,
        0.03171607]])</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-a91d520c-6ab8-4f4f-87c8-0126ebb32a00' class='xr-section-summary-in' type='checkbox'  checked><label for='section-a91d520c-6ab8-4f4f-87c8-0126ebb32a00' class='xr-section-summary' >Attributes: <span>(2)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><dl class='xr-attrs'><dt><span>created_at :</span></dt><dd>2022-08-25T13:07:26.612449</dd><dt><span>arviz_version :</span></dt><dd>0.12.1</dd></dl></div></li></ul></div></div><br></div>
                      </ul>
                  </div>
            </li>

            <li class = "xr-section-item">
                  <input id="idata_observed_data8b66119b-5938-4085-a9c6-481f82a49503" class="xr-section-summary-in" type="checkbox">
                  <label for="idata_observed_data8b66119b-5938-4085-a9c6-481f82a49503" class = "xr-section-summary">observed_data</label>
                  <div class="xr-section-inline-details"></div>
                  <div class="xr-section-details">
                      <ul id="xr-dataset-coord-list" class="xr-var-list">
                          <div style="padding-left:2rem;"><div><svg style="position: absolute; width: 0; height: 0; overflow: hidden">
<defs>
<symbol id="icon-database" viewBox="0 0 32 32">
<path d="M16 0c-8.837 0-16 2.239-16 5v4c0 2.761 7.163 5 16 5s16-2.239 16-5v-4c0-2.761-7.163-5-16-5z"></path>
<path d="M16 17c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z"></path>
<path d="M16 26c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z"></path>
</symbol>
<symbol id="icon-file-text2" viewBox="0 0 32 32">
<path d="M28.681 7.159c-0.694-0.947-1.662-2.053-2.724-3.116s-2.169-2.030-3.116-2.724c-1.612-1.182-2.393-1.319-2.841-1.319h-15.5c-1.378 0-2.5 1.121-2.5 2.5v27c0 1.378 1.122 2.5 2.5 2.5h23c1.378 0 2.5-1.122 2.5-2.5v-19.5c0-0.448-0.137-1.23-1.319-2.841zM24.543 5.457c0.959 0.959 1.712 1.825 2.268 2.543h-4.811v-4.811c0.718 0.556 1.584 1.309 2.543 2.268zM28 29.5c0 0.271-0.229 0.5-0.5 0.5h-23c-0.271 0-0.5-0.229-0.5-0.5v-27c0-0.271 0.229-0.5 0.5-0.5 0 0 15.499-0 15.5 0v7c0 0.552 0.448 1 1 1h7v19.5z"></path>
<path d="M23 26h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z"></path>
<path d="M23 22h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z"></path>
<path d="M23 18h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z"></path>
</symbol>
</defs>
</svg>
<style>/* CSS stylesheet for displaying xarray objects in jupyterlab.
 *
 */

:root {
  --xr-font-color0: var(--jp-content-font-color0, rgba(0, 0, 0, 1));
  --xr-font-color2: var(--jp-content-font-color2, rgba(0, 0, 0, 0.54));
  --xr-font-color3: var(--jp-content-font-color3, rgba(0, 0, 0, 0.38));
  --xr-border-color: var(--jp-border-color2, #e0e0e0);
  --xr-disabled-color: var(--jp-layout-color3, #bdbdbd);
  --xr-background-color: var(--jp-layout-color0, white);
  --xr-background-color-row-even: var(--jp-layout-color1, white);
  --xr-background-color-row-odd: var(--jp-layout-color2, #eeeeee);
}

html[theme=dark],
body[data-theme=dark],
body.vscode-dark {
  --xr-font-color0: rgba(255, 255, 255, 1);
  --xr-font-color2: rgba(255, 255, 255, 0.54);
  --xr-font-color3: rgba(255, 255, 255, 0.38);
  --xr-border-color: #1F1F1F;
  --xr-disabled-color: #515151;
  --xr-background-color: #111111;
  --xr-background-color-row-even: #111111;
  --xr-background-color-row-odd: #313131;
}

.xr-wrap {
  display: block !important;
  min-width: 300px;
  max-width: 700px;
}

.xr-text-repr-fallback {
  /* fallback to plain text repr when CSS is not injected (untrusted notebook) */
  display: none;
}

.xr-header {
  padding-top: 6px;
  padding-bottom: 6px;
  margin-bottom: 4px;
  border-bottom: solid 1px var(--xr-border-color);
}

.xr-header > div,
.xr-header > ul {
  display: inline;
  margin-top: 0;
  margin-bottom: 0;
}

.xr-obj-type,
.xr-array-name {
  margin-left: 2px;
  margin-right: 10px;
}

.xr-obj-type {
  color: var(--xr-font-color2);
}

.xr-sections {
  padding-left: 0 !important;
  display: grid;
  grid-template-columns: 150px auto auto 1fr 20px 20px;
}

.xr-section-item {
  display: contents;
}

.xr-section-item input {
  display: none;
}

.xr-section-item input + label {
  color: var(--xr-disabled-color);
}

.xr-section-item input:enabled + label {
  cursor: pointer;
  color: var(--xr-font-color2);
}

.xr-section-item input:enabled + label:hover {
  color: var(--xr-font-color0);
}

.xr-section-summary {
  grid-column: 1;
  color: var(--xr-font-color2);
  font-weight: 500;
}

.xr-section-summary > span {
  display: inline-block;
  padding-left: 0.5em;
}

.xr-section-summary-in:disabled + label {
  color: var(--xr-font-color2);
}

.xr-section-summary-in + label:before {
  display: inline-block;
  content: '►';
  font-size: 11px;
  width: 15px;
  text-align: center;
}

.xr-section-summary-in:disabled + label:before {
  color: var(--xr-disabled-color);
}

.xr-section-summary-in:checked + label:before {
  content: '▼';
}

.xr-section-summary-in:checked + label > span {
  display: none;
}

.xr-section-summary,
.xr-section-inline-details {
  padding-top: 4px;
  padding-bottom: 4px;
}

.xr-section-inline-details {
  grid-column: 2 / -1;
}

.xr-section-details {
  display: none;
  grid-column: 1 / -1;
  margin-bottom: 5px;
}

.xr-section-summary-in:checked ~ .xr-section-details {
  display: contents;
}

.xr-array-wrap {
  grid-column: 1 / -1;
  display: grid;
  grid-template-columns: 20px auto;
}

.xr-array-wrap > label {
  grid-column: 1;
  vertical-align: top;
}

.xr-preview {
  color: var(--xr-font-color3);
}

.xr-array-preview,
.xr-array-data {
  padding: 0 5px !important;
  grid-column: 2;
}

.xr-array-data,
.xr-array-in:checked ~ .xr-array-preview {
  display: none;
}

.xr-array-in:checked ~ .xr-array-data,
.xr-array-preview {
  display: inline-block;
}

.xr-dim-list {
  display: inline-block !important;
  list-style: none;
  padding: 0 !important;
  margin: 0;
}

.xr-dim-list li {
  display: inline-block;
  padding: 0;
  margin: 0;
}

.xr-dim-list:before {
  content: '(';
}

.xr-dim-list:after {
  content: ')';
}

.xr-dim-list li:not(:last-child):after {
  content: ',';
  padding-right: 5px;
}

.xr-has-index {
  font-weight: bold;
}

.xr-var-list,
.xr-var-item {
  display: contents;
}

.xr-var-item > div,
.xr-var-item label,
.xr-var-item > .xr-var-name span {
  background-color: var(--xr-background-color-row-even);
  margin-bottom: 0;
}

.xr-var-item > .xr-var-name:hover span {
  padding-right: 5px;
}

.xr-var-list > li:nth-child(odd) > div,
.xr-var-list > li:nth-child(odd) > label,
.xr-var-list > li:nth-child(odd) > .xr-var-name span {
  background-color: var(--xr-background-color-row-odd);
}

.xr-var-name {
  grid-column: 1;
}

.xr-var-dims {
  grid-column: 2;
}

.xr-var-dtype {
  grid-column: 3;
  text-align: right;
  color: var(--xr-font-color2);
}

.xr-var-preview {
  grid-column: 4;
}

.xr-var-name,
.xr-var-dims,
.xr-var-dtype,
.xr-preview,
.xr-attrs dt {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  padding-right: 10px;
}

.xr-var-name:hover,
.xr-var-dims:hover,
.xr-var-dtype:hover,
.xr-attrs dt:hover {
  overflow: visible;
  width: auto;
  z-index: 1;
}

.xr-var-attrs,
.xr-var-data {
  display: none;
  background-color: var(--xr-background-color) !important;
  padding-bottom: 5px !important;
}

.xr-var-attrs-in:checked ~ .xr-var-attrs,
.xr-var-data-in:checked ~ .xr-var-data {
  display: block;
}

.xr-var-data > table {
  float: right;
}

.xr-var-name span,
.xr-var-data,
.xr-attrs {
  padding-left: 25px !important;
}

.xr-attrs,
.xr-var-attrs,
.xr-var-data {
  grid-column: 1 / -1;
}

dl.xr-attrs {
  padding: 0;
  margin: 0;
  display: grid;
  grid-template-columns: 125px auto;
}

.xr-attrs dt,
.xr-attrs dd {
  padding: 0;
  margin: 0;
  float: left;
  padding-right: 10px;
  width: auto;
}

.xr-attrs dt {
  font-weight: normal;
  grid-column: 1;
}

.xr-attrs dt:hover span {
  display: inline-block;
  background: var(--xr-background-color);
  padding-right: 10px;
}

.xr-attrs dd {
  grid-column: 2;
  white-space: pre-wrap;
  word-break: break-all;
}

.xr-icon-database,
.xr-icon-file-text2 {
  display: inline-block;
  vertical-align: middle;
  width: 1em;
  height: 1.5em !important;
  stroke-width: 0;
  stroke: currentColor;
  fill: currentColor;
}
</style><pre class='xr-text-repr-fallback'>&lt;xarray.Dataset&gt;
Dimensions:  (y_dim_0: 10, y_dim_1: 8)
Coordinates:
  * y_dim_0  (y_dim_0) int64 0 1 2 3 4 5 6 7 8 9
  * y_dim_1  (y_dim_1) int64 0 1 2 3 4 5 6 7
Data variables:
    y        (y_dim_0, y_dim_1) float64 36.0 59.0 136.0 36.0 ... 65.0 168.0 12.0
Attributes:
    created_at:     2022-08-25T13:07:26.622125
    arviz_version:  0.12.1</pre><div class='xr-wrap' style='display:none'><div class='xr-header'><div class='xr-obj-type'>xarray.Dataset</div></div><ul class='xr-sections'><li class='xr-section-item'><input id='section-ba344ba1-c0a4-4f70-9784-f5e108304ce9' class='xr-section-summary-in' type='checkbox' disabled ><label for='section-ba344ba1-c0a4-4f70-9784-f5e108304ce9' class='xr-section-summary'  title='Expand/collapse section'>Dimensions:</label><div class='xr-section-inline-details'><ul class='xr-dim-list'><li><span class='xr-has-index'>y_dim_0</span>: 10</li><li><span class='xr-has-index'>y_dim_1</span>: 8</li></ul></div><div class='xr-section-details'></div></li><li class='xr-section-item'><input id='section-d7c0b919-cefd-461c-9128-d074a937b8b5' class='xr-section-summary-in' type='checkbox'  checked><label for='section-d7c0b919-cefd-461c-9128-d074a937b8b5' class='xr-section-summary' >Coordinates: <span>(2)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>y_dim_0</span></div><div class='xr-var-dims'>(y_dim_0)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>0 1 2 3 4 5 6 7 8 9</div><input id='attrs-65b7e30f-d8bf-48ac-9434-1afa1c20def7' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-65b7e30f-d8bf-48ac-9434-1afa1c20def7' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-a5fdac5a-573d-44da-9c03-cb910445b07f' class='xr-var-data-in' type='checkbox'><label for='data-a5fdac5a-573d-44da-9c03-cb910445b07f' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>y_dim_1</span></div><div class='xr-var-dims'>(y_dim_1)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>0 1 2 3 4 5 6 7</div><input id='attrs-3cae9796-b212-4a7b-b411-c0d67070423b' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-3cae9796-b212-4a7b-b411-c0d67070423b' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-30d810cc-864a-45d6-be12-a3d534b8ae74' class='xr-var-data-in' type='checkbox'><label for='data-30d810cc-864a-45d6-be12-a3d534b8ae74' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([0, 1, 2, 3, 4, 5, 6, 7])</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-4e2357ec-a345-4ae8-882c-aea0aba0d82f' class='xr-section-summary-in' type='checkbox'  checked><label for='section-4e2357ec-a345-4ae8-882c-aea0aba0d82f' class='xr-section-summary' >Data variables: <span>(1)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span>y</span></div><div class='xr-var-dims'>(y_dim_0, y_dim_1)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>36.0 59.0 136.0 ... 65.0 168.0 12.0</div><input id='attrs-fc2f81af-c8e4-4a2e-bf4c-d83c2b81b582' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-fc2f81af-c8e4-4a2e-bf4c-d83c2b81b582' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-b57038ec-7371-408b-bb5a-2583554112d1' class='xr-var-data-in' type='checkbox'><label for='data-b57038ec-7371-408b-bb5a-2583554112d1' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[ 36.,  59., 136.,  36., 239., 125., 191.,  18.],
       [  5.,  46.,  23.,  20.,  50.,  11.,  40.,   5.],
       [ 45.,  98., 188., 124., 250., 155., 365.,  33.],
       [ 26., 221., 198.,  36., 131., 130., 196.,   4.],
       [ 52.,  75., 347.,  66., 323., 263., 313.,  51.],
       [ 65., 126., 115.,  33.,  65.,  39., 129.,  59.],
       [ 42.,  71., 203., 147., 271., 109., 180., 146.],
       [ 40.,  57., 383., 170., 321., 244., 256.,  71.],
       [ 37., 332., 113.,  59.,  90.,  47., 132.,  10.],
       [ 32., 373., 116.,  67., 117.,  65., 168.,  12.]])</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-55851cdd-a426-4b0b-b5de-c93b92e22456' class='xr-section-summary-in' type='checkbox'  checked><label for='section-55851cdd-a426-4b0b-b5de-c93b92e22456' class='xr-section-summary' >Attributes: <span>(2)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><dl class='xr-attrs'><dt><span>created_at :</span></dt><dd>2022-08-25T13:07:26.622125</dd><dt><span>arviz_version :</span></dt><dd>0.12.1</dd></dl></div></li></ul></div></div><br></div>
                      </ul>
                  </div>
            </li>

              </ul>
            </div>
            <style> /* CSS stylesheet for displaying InferenceData objects in jupyterlab.
 *
 */

:root {
  --xr-font-color0: var(--jp-content-font-color0, rgba(0, 0, 0, 1));
  --xr-font-color2: var(--jp-content-font-color2, rgba(0, 0, 0, 0.54));
  --xr-font-color3: var(--jp-content-font-color3, rgba(0, 0, 0, 0.38));
  --xr-border-color: var(--jp-border-color2, #e0e0e0);
  --xr-disabled-color: var(--jp-layout-color3, #bdbdbd);
  --xr-background-color: var(--jp-layout-color0, white);
  --xr-background-color-row-even: var(--jp-layout-color1, white);
  --xr-background-color-row-odd: var(--jp-layout-color2, #eeeeee);
}

html[theme=dark],
body.vscode-dark {
  --xr-font-color0: rgba(255, 255, 255, 1);
  --xr-font-color2: rgba(255, 255, 255, 0.54);
  --xr-font-color3: rgba(255, 255, 255, 0.38);
  --xr-border-color: #1F1F1F;
  --xr-disabled-color: #515151;
  --xr-background-color: #111111;
  --xr-background-color-row-even: #111111;
  --xr-background-color-row-odd: #313131;
}

.xr-wrap {
  display: block;
  min-width: 300px;
  max-width: 700px;
}

.xr-text-repr-fallback {
  /* fallback to plain text repr when CSS is not injected (untrusted notebook) */
  display: none;
}

.xr-header {
  padding-top: 6px;
  padding-bottom: 6px;
  margin-bottom: 4px;
  border-bottom: solid 1px var(--xr-border-color);
}

.xr-header > div,
.xr-header > ul {
  display: inline;
  margin-top: 0;
  margin-bottom: 0;
}

.xr-obj-type,
.xr-array-name {
  margin-left: 2px;
  margin-right: 10px;
}

.xr-obj-type {
  color: var(--xr-font-color2);
}

.xr-sections {
  padding-left: 0 !important;
  display: grid;
  grid-template-columns: 150px auto auto 1fr 20px 20px;
}

.xr-sections.group-sections {
  grid-template-columns: auto;
}

.xr-section-item {
  display: contents;
}

.xr-section-item input {
  display: none;
}

.xr-section-item input + label {
  color: var(--xr-disabled-color);
}

.xr-section-item input:enabled + label {
  cursor: pointer;
  color: var(--xr-font-color2);
}

.xr-section-item input:enabled + label:hover {
  color: var(--xr-font-color0);
}

.xr-section-summary {
  grid-column: 1;
  color: var(--xr-font-color2);
  font-weight: 500;
}

.xr-section-summary > span {
  display: inline-block;
  padding-left: 0.5em;
}

.xr-section-summary-in:disabled + label {
  color: var(--xr-font-color2);
}

.xr-section-summary-in + label:before {
  display: inline-block;
  content: '►';
  font-size: 11px;
  width: 15px;
  text-align: center;
}

.xr-section-summary-in:disabled + label:before {
  color: var(--xr-disabled-color);
}

.xr-section-summary-in:checked + label:before {
  content: '▼';
}

.xr-section-summary-in:checked + label > span {
  display: none;
}

.xr-section-summary,
.xr-section-inline-details {
  padding-top: 4px;
  padding-bottom: 4px;
}

.xr-section-inline-details {
  grid-column: 2 / -1;
}

.xr-section-details {
  display: none;
  grid-column: 1 / -1;
  margin-bottom: 5px;
}

.xr-section-summary-in:checked ~ .xr-section-details {
  display: contents;
}

.xr-array-wrap {
  grid-column: 1 / -1;
  display: grid;
  grid-template-columns: 20px auto;
}

.xr-array-wrap > label {
  grid-column: 1;
  vertical-align: top;
}

.xr-preview {
  color: var(--xr-font-color3);
}

.xr-array-preview,
.xr-array-data {
  padding: 0 5px !important;
  grid-column: 2;
}

.xr-array-data,
.xr-array-in:checked ~ .xr-array-preview {
  display: none;
}

.xr-array-in:checked ~ .xr-array-data,
.xr-array-preview {
  display: inline-block;
}

.xr-dim-list {
  display: inline-block !important;
  list-style: none;
  padding: 0 !important;
  margin: 0;
}

.xr-dim-list li {
  display: inline-block;
  padding: 0;
  margin: 0;
}

.xr-dim-list:before {
  content: '(';
}

.xr-dim-list:after {
  content: ')';
}

.xr-dim-list li:not(:last-child):after {
  content: ',';
  padding-right: 5px;
}

.xr-has-index {
  font-weight: bold;
}

.xr-var-list,
.xr-var-item {
  display: contents;
}

.xr-var-item > div,
.xr-var-item label,
.xr-var-item > .xr-var-name span {
  background-color: var(--xr-background-color-row-even);
  margin-bottom: 0;
}

.xr-var-item > .xr-var-name:hover span {
  padding-right: 5px;
}

.xr-var-list > li:nth-child(odd) > div,
.xr-var-list > li:nth-child(odd) > label,
.xr-var-list > li:nth-child(odd) > .xr-var-name span {
  background-color: var(--xr-background-color-row-odd);
}

.xr-var-name {
  grid-column: 1;
}

.xr-var-dims {
  grid-column: 2;
}

.xr-var-dtype {
  grid-column: 3;
  text-align: right;
  color: var(--xr-font-color2);
}

.xr-var-preview {
  grid-column: 4;
}

.xr-var-name,
.xr-var-dims,
.xr-var-dtype,
.xr-preview,
.xr-attrs dt {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  padding-right: 10px;
}

.xr-var-name:hover,
.xr-var-dims:hover,
.xr-var-dtype:hover,
.xr-attrs dt:hover {
  overflow: visible;
  width: auto;
  z-index: 1;
}

.xr-var-attrs,
.xr-var-data {
  display: none;
  background-color: var(--xr-background-color) !important;
  padding-bottom: 5px !important;
}

.xr-var-attrs-in:checked ~ .xr-var-attrs,
.xr-var-data-in:checked ~ .xr-var-data {
  display: block;
}

.xr-var-data > table {
  float: right;
}

.xr-var-name span,
.xr-var-data,
.xr-attrs {
  padding-left: 25px !important;
}

.xr-attrs,
.xr-var-attrs,
.xr-var-data {
  grid-column: 1 / -1;
}

dl.xr-attrs {
  padding: 0;
  margin: 0;
  display: grid;
  grid-template-columns: 125px auto;
}

.xr-attrs dt, dd {
  padding: 0;
  margin: 0;
  float: left;
  padding-right: 10px;
  width: auto;
}

.xr-attrs dt {
  font-weight: normal;
  grid-column: 1;
}

.xr-attrs dt:hover span {
  display: inline-block;
  background: var(--xr-background-color);
  padding-right: 10px;
}

.xr-attrs dd {
  grid-column: 2;
  white-space: pre-wrap;
  word-break: break-all;
}

.xr-icon-database,
.xr-icon-file-text2 {
  display: inline-block;
  vertical-align: middle;
  width: 1em;
  height: 1.5em !important;
  stroke-width: 0;
  stroke: currentColor;
  fill: currentColor;
}
.xr-wrap{width:700px!important;} </style></div></div>
</div>
<p>scCODA selects credible effects based on their inclusion probability. The cutoff between credible and non-credible effects depends on the desired false discovery rate (FDR). A smaller FDR value will produce more conservative results, but might miss some effects, while a larger FDR value selects more effects at the cost of a larger number of false discoveries.
The desired FDR level can be easily set after inference via sim_results.set_fdr(). Per default, the value is 0.05. Since, depending on the dataset, the FDR can have a major influence on the result, we recommend to try out different FDRs up to 0.2 to get the most prominent effects.</p>
<p>In our case, we use less strict FDR of 0.2.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sccoda_res</span><span class="o">.</span><span class="n">set_fdr</span><span class="p">(</span><span class="mf">0.2</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>To get the binary classification of compositional changes per cell type we use the <code class="docutils literal notranslate"><span class="pre">credible_effects</span></code> function of scCODA on the result object. Every cell type labeled as “True” is significantly more or less present. The fold-changes describe whether the cell type is more or less present. Hence, we will plot them alongside the binary classification below.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sccoda_res</span><span class="o">.</span><span class="n">credible_effects</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Covariate                 Cell Type            
condition[T.Hpoly.Day10]  Endocrine                False
                          Enterocyte                True
                          Enterocyte.Progenitor    False
                          Goblet                   False
                          Stem                     False
                          TA                       False
                          TA.Early                 False
                          Tuft                      True
condition[T.Hpoly.Day3]   Endocrine                False
                          Enterocyte               False
                          Enterocyte.Progenitor    False
                          Goblet                   False
                          Stem                     False
                          TA                       False
                          TA.Early                 False
                          Tuft                     False
condition[T.Salmonella]   Endocrine                False
                          Enterocyte                True
                          Enterocyte.Progenitor    False
                          Goblet                   False
                          Stem                     False
                          TA                       False
                          TA.Early                 False
                          Tuft                     False
Name: Final Parameter, dtype: bool
</pre></div>
</div>
</div>
</div>
<p>To be able to easily plot the fold changes together with the binary classification, we create intermediate objects for all conditions.</p>
<ol class="simple">
<li><p>Pandas DataFrames which only store the cell types and whether they had significant compositional changes</p></li>
<li><p>Pandas DataFrames which have the cell types as indices and the scCODA learned metrics as values.</p></li>
</ol>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">credible_effects_salmonella</span> <span class="o">=</span> <span class="n">sccoda_res</span><span class="o">.</span><span class="n">credible_effects</span><span class="p">()[</span><span class="s2">&quot;condition[T.Salmonella]&quot;</span><span class="p">]</span>
<span class="n">credible_effects_hpoly_day3</span> <span class="o">=</span> <span class="n">sccoda_res</span><span class="o">.</span><span class="n">credible_effects</span><span class="p">()[</span><span class="s2">&quot;condition[T.Hpoly.Day3]&quot;</span><span class="p">]</span>
<span class="n">credible_effects_hpoly_day10</span> <span class="o">=</span> <span class="n">sccoda_res</span><span class="o">.</span><span class="n">credible_effects</span><span class="p">()[</span><span class="s2">&quot;condition[T.Hpoly.Day10]&quot;</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">credible_effects_salmonella</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Cell Type
Endocrine                False
Enterocyte                True
Enterocyte.Progenitor    False
Goblet                   False
Stem                     False
TA                       False
TA.Early                 False
Tuft                     False
Name: Final Parameter, dtype: bool
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">salmonella_effect</span> <span class="o">=</span> <span class="n">sccoda_res</span><span class="o">.</span><span class="n">effect_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s2">&quot;condition[T.Salmonella]&quot;</span><span class="p">]</span>
<span class="n">hpoly_effect_3</span> <span class="o">=</span> <span class="n">sccoda_res</span><span class="o">.</span><span class="n">effect_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s2">&quot;condition[T.Hpoly.Day3]&quot;</span><span class="p">]</span>
<span class="n">hpoly_effected_10</span> <span class="o">=</span> <span class="n">sccoda_res</span><span class="o">.</span><span class="n">effect_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s2">&quot;condition[T.Hpoly.Day10]&quot;</span><span class="p">]</span>

<span class="n">salmonella_effect</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Final Parameter</th>
      <th>HDI 3%</th>
      <th>HDI 97%</th>
      <th>SD</th>
      <th>Inclusion probability</th>
      <th>Expected Sample</th>
      <th>log2-fold change</th>
    </tr>
    <tr>
      <th>Cell Type</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Endocrine</th>
      <td>0.000000</td>
      <td>0.000</td>
      <td>0.000</td>
      <td>0.000</td>
      <td>0.000000</td>
      <td>32.728788</td>
      <td>-0.529448</td>
    </tr>
    <tr>
      <th>Enterocyte</th>
      <td>1.560117</td>
      <td>0.998</td>
      <td>2.074</td>
      <td>0.285</td>
      <td>1.000000</td>
      <td>382.746152</td>
      <td>1.721325</td>
    </tr>
    <tr>
      <th>Enterocyte.Progenitor</th>
      <td>0.000000</td>
      <td>-0.413</td>
      <td>0.609</td>
      <td>0.148</td>
      <td>0.313267</td>
      <td>125.870682</td>
      <td>-0.529448</td>
    </tr>
    <tr>
      <th>Goblet</th>
      <td>0.000000</td>
      <td>-0.314</td>
      <td>1.025</td>
      <td>0.296</td>
      <td>0.455000</td>
      <td>52.470707</td>
      <td>-0.529448</td>
    </tr>
    <tr>
      <th>Stem</th>
      <td>0.000000</td>
      <td>-0.715</td>
      <td>0.345</td>
      <td>0.180</td>
      <td>0.332733</td>
      <td>135.809755</td>
      <td>-0.529448</td>
    </tr>
    <tr>
      <th>TA</th>
      <td>0.000000</td>
      <td>-0.857</td>
      <td>0.421</td>
      <td>0.218</td>
      <td>0.364200</td>
      <td>78.905825</td>
      <td>-0.529448</td>
    </tr>
    <tr>
      <th>TA.Early</th>
      <td>0.000000</td>
      <td>-0.344</td>
      <td>0.637</td>
      <td>0.160</td>
      <td>0.322400</td>
      <td>152.209072</td>
      <td>-0.529448</td>
    </tr>
    <tr>
      <th>Tuft</th>
      <td>0.000000</td>
      <td>-1.371</td>
      <td>0.467</td>
      <td>0.347</td>
      <td>0.369867</td>
      <td>23.459019</td>
      <td>-0.529448</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>Inspired by plots from “High-resolution single-cell atlas reveals diversity and plasticity of tissue-resident neutrophils in non-small cell lung cancer” <span id="id11">[<a class="reference internal" href="#id318" title="Stefan Salcher, Gregor Sturm, Lena Horvath, Gerold Untergasser, Georgios Fotakis, Elisa Panizzolo, Agnieszka Martowicz, Georg Pall, Gabriele Gamerith, Martina Sykora, Florian Augustin, Katja Schmitz, Francesca Finotello, Dietmar Rieder, Sieghart Sopper, Dominik Wolf, Andreas Pircher, and Zlatko Trajanoski. High-resolution single-cell atlas reveals diversity and plasticity of tissue-resident neutrophils in non-small cell lung cancer. bioRxiv, 2022. URL: https://www.biorxiv.org/content/early/2022/05/10/2022.05.09.491204, arXiv:https://www.biorxiv.org/content/early/2022/05/10/2022.05.09.491204.full.pdf, doi:10.1101/2022.05.09.491204.">Salcher <em>et al.</em>, 2022</a>]</span>, we can now easily filter by the cell types that had credible changes and show their log fold changes. This is easily done with the plotting library <a class="reference external" href="https://altair-viz.github.io/">Altair</a>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="p">(</span>
    <span class="n">alt</span><span class="o">.</span><span class="n">Chart</span><span class="p">(</span>
        <span class="n">salmonella_effect</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">credible_effects_salmonella</span><span class="p">]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(),</span>
        <span class="n">title</span><span class="o">=</span><span class="s2">&quot;Salmonella&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="o">.</span><span class="n">mark_bar</span><span class="p">()</span>
    <span class="o">.</span><span class="n">encode</span><span class="p">(</span>
        <span class="n">x</span><span class="o">=</span><span class="n">alt</span><span class="o">.</span><span class="n">X</span><span class="p">(</span><span class="s2">&quot;Cell Type&quot;</span><span class="p">,</span> <span class="n">sort</span><span class="o">=</span><span class="s2">&quot;y&quot;</span><span class="p">),</span>
        <span class="n">y</span><span class="o">=</span><span class="s2">&quot;log2-fold change&quot;</span><span class="p">,</span>
        <span class="n">color</span><span class="o">=</span><span class="n">alt</span><span class="o">.</span><span class="n">Color</span><span class="p">(</span><span class="s2">&quot;Cell Type&quot;</span><span class="p">),</span>
    <span class="p">)</span>
    <span class="o">|</span> <span class="n">alt</span><span class="o">.</span><span class="n">Chart</span><span class="p">(</span>
        <span class="n">hpoly_effect_3</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">credible_effects_hpoly_day3</span><span class="p">]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(),</span>
        <span class="n">title</span><span class="o">=</span><span class="s2">&quot;Heligmosomoides polygyrus day 3&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="o">.</span><span class="n">mark_bar</span><span class="p">()</span>
    <span class="o">.</span><span class="n">encode</span><span class="p">(</span>
        <span class="n">x</span><span class="o">=</span><span class="n">alt</span><span class="o">.</span><span class="n">X</span><span class="p">(</span><span class="s2">&quot;Cell Type&quot;</span><span class="p">,</span> <span class="n">sort</span><span class="o">=</span><span class="s2">&quot;y&quot;</span><span class="p">),</span>
        <span class="n">y</span><span class="o">=</span><span class="s2">&quot;log2-fold change&quot;</span><span class="p">,</span>
        <span class="n">color</span><span class="o">=</span><span class="n">alt</span><span class="o">.</span><span class="n">Color</span><span class="p">(</span><span class="s2">&quot;Cell Type&quot;</span><span class="p">),</span>
    <span class="p">)</span>
    <span class="o">|</span> <span class="n">alt</span><span class="o">.</span><span class="n">Chart</span><span class="p">(</span>
        <span class="n">hpoly_effected_10</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">credible_effects_hpoly_day10</span><span class="p">]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(),</span>
        <span class="n">title</span><span class="o">=</span><span class="s2">&quot;Heligmosomoides polygyrus day 10&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="o">.</span><span class="n">mark_bar</span><span class="p">()</span>
    <span class="o">.</span><span class="n">encode</span><span class="p">(</span>
        <span class="n">x</span><span class="o">=</span><span class="n">alt</span><span class="o">.</span><span class="n">X</span><span class="p">(</span><span class="s2">&quot;Cell Type&quot;</span><span class="p">,</span> <span class="n">sort</span><span class="o">=</span><span class="s2">&quot;y&quot;</span><span class="p">),</span>
        <span class="n">y</span><span class="o">=</span><span class="s2">&quot;log2-fold change&quot;</span><span class="p">,</span>
        <span class="n">color</span><span class="o">=</span><span class="n">alt</span><span class="o">.</span><span class="n">Color</span><span class="p">(</span><span class="s2">&quot;Cell Type&quot;</span><span class="p">),</span>
    <span class="p">)</span>
<span class="p">)</span><span class="o">.</span><span class="n">resolve_scale</span><span class="p">(</span><span class="n">y</span><span class="o">=</span><span class="s2">&quot;shared&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;shared&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
<div id="altair-viz-d2daf180bc6442dca8430c0a0301f1c8"></div>
<script type="text/javascript">
  var VEGA_DEBUG = (typeof VEGA_DEBUG == "undefined") ? {} : VEGA_DEBUG;
  (function(spec, embedOpt){
    let outputDiv = document.currentScript.previousElementSibling;
    if (outputDiv.id !== "altair-viz-d2daf180bc6442dca8430c0a0301f1c8") {
      outputDiv = document.getElementById("altair-viz-d2daf180bc6442dca8430c0a0301f1c8");
    }
    const paths = {
      "vega": "https://cdn.jsdelivr.net/npm//vega@5?noext",
      "vega-lib": "https://cdn.jsdelivr.net/npm//vega-lib?noext",
      "vega-lite": "https://cdn.jsdelivr.net/npm//vega-lite@4.17.0?noext",
      "vega-embed": "https://cdn.jsdelivr.net/npm//vega-embed@6?noext",
    };

    function maybeLoadScript(lib, version) {
      var key = `${lib.replace("-", "")}_version`;
      return (VEGA_DEBUG[key] == version) ?
        Promise.resolve(paths[lib]) :
        new Promise(function(resolve, reject) {
          var s = document.createElement('script');
          document.getElementsByTagName("head")[0].appendChild(s);
          s.async = true;
          s.onload = () => {
            VEGA_DEBUG[key] = version;
            return resolve(paths[lib]);
          };
          s.onerror = () => reject(`Error loading script: ${paths[lib]}`);
          s.src = paths[lib];
        });
    }

    function showError(err) {
      outputDiv.innerHTML = `<div class="error" style="color:red;">${err}</div>`;
      throw err;
    }

    function displayChart(vegaEmbed) {
      vegaEmbed(outputDiv, spec, embedOpt)
        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));
    }

    if(typeof define === "function" && define.amd) {
      requirejs.config({paths});
      require(["vega-embed"], displayChart, err => showError(`Error loading script: ${err.message}`));
    } else {
      maybeLoadScript("vega", "5")
        .then(() => maybeLoadScript("vega-lite", "4.17.0"))
        .then(() => maybeLoadScript("vega-embed", "6"))
        .catch(showError)
        .then(() => displayChart(vegaEmbed));
    }
  })({"config": {"view": {"continuousWidth": 400, "continuousHeight": 300}}, "hconcat": [{"data": {"name": "data-97711a13ba5763d8d65bb3c468af5f74"}, "mark": "bar", "encoding": {"color": {"field": "Cell Type", "type": "nominal"}, "x": {"field": "Cell Type", "sort": "y", "type": "nominal"}, "y": {"field": "log2-fold change", "type": "quantitative"}}, "title": "Salmonella"}, {"data": {"name": "data-d751713988987e9331980363e24189ce"}, "mark": "bar", "encoding": {"color": {"field": "Cell Type", "type": "nominal"}, "x": {"field": "Cell Type", "sort": "y", "type": "nominal"}, "y": {"field": "log2-fold change", "type": "quantitative"}}, "title": "Heligmosomoides polygyrus day 3"}, {"data": {"name": "data-448bcf85bb2d43409fc2182c8325557c"}, "mark": "bar", "encoding": {"color": {"field": "Cell Type", "type": "nominal"}, "x": {"field": "Cell Type", "sort": "y", "type": "nominal"}, "y": {"field": "log2-fold change", "type": "quantitative"}}, "title": "Heligmosomoides polygyrus day 10"}], "resolve": {"scale": {"color": "shared", "y": "shared"}}, "$schema": "https://vega.github.io/schema/vega-lite/v4.17.0.json", "datasets": {"data-97711a13ba5763d8d65bb3c468af5f74": [{"Cell Type": "Enterocyte", "Final Parameter": 1.5601169343226744, "HDI 3%": 0.998, "HDI 97%": 2.074, "SD": 0.285, "Inclusion probability": 1.0, "Expected Sample": 382.7461523952098, "log2-fold change": 1.7213249647768392}], "data-d751713988987e9331980363e24189ce": [], "data-448bcf85bb2d43409fc2182c8325557c": [{"Cell Type": "Enterocyte", "Final Parameter": -0.5983571279468299, "HDI 3%": -1.399, "HDI 97%": 0.07, "SD": 0.447, "Inclusion probability": 0.7592, "Expected Sample": 65.58725146478338, "log2-fold change": -0.8235755870771747}, {"Cell Type": "Tuft", "Final Parameter": 0.562600832496352, "HDI 3%": -0.252, "HDI 97%": 1.445, "SD": 0.483, "Inclusion probability": 0.7031333333333334, "Expected Sample": 61.08945712114848, "log2-fold change": 0.8513327051347692}]}}, {"mode": "vega-lite"});
</script></div></div>
</div>
<p>The plots nicely show the significant and credible effects of conditions on the cell types. These effects largely agree with the findings in the Haber paper, who used a non-compositional Poisson regression model their findings:</p>
<ol class="simple">
<li><p>“After Salmonella infection, the frequency of mature enterocytes increased substantially.”<span id="id12">[<a class="reference internal" href="#id313" title="Adam L. Haber, Moshe Biton, Noga Rogel, Rebecca H. Herbst, Karthik Shekhar, Christopher Smillie, Grace Burgin, Toni M. Delorey, Michael R. Howitt, Yarden Katz, Itay Tirosh, Semir Beyaz, Danielle Dionne, Mei Zhang, Raktima Raychowdhury, Wendy S. Garrett, Orit Rozenblatt-Rosen, Hai Ning Shi, Omer Yilmaz, Ramnik J. Xavier, and Aviv Regev. A single-cell survey of the small intestinal epithelium. Nature, 551(7680):333-339, Nov 2017. URL: https://doi.org/10.1038/nature24489, doi:10.1038/nature24489.">Haber <em>et al.</em>, 2017</a>]</span></p></li>
<li><p>“Heligmosomoides polygyrus caused an increase in the abundance of goblet and tuft cells.”<span id="id13">[<a class="reference internal" href="#id313" title="Adam L. Haber, Moshe Biton, Noga Rogel, Rebecca H. Herbst, Karthik Shekhar, Christopher Smillie, Grace Burgin, Toni M. Delorey, Michael R. Howitt, Yarden Katz, Itay Tirosh, Semir Beyaz, Danielle Dionne, Mei Zhang, Raktima Raychowdhury, Wendy S. Garrett, Orit Rozenblatt-Rosen, Hai Ning Shi, Omer Yilmaz, Ramnik J. Xavier, and Aviv Regev. A single-cell survey of the small intestinal epithelium. Nature, 551(7680):333-339, Nov 2017. URL: https://doi.org/10.1038/nature24489, doi:10.1038/nature24489.">Haber <em>et al.</em>, 2017</a>]</span></p></li>
</ol>
<p>Readers familiar with the original publication may wonder why the model used by Haber et al. found more significant effects than scCODA, for example a decrease in Stem and Transit-Amplifying cells in the case of Salmonella infection<span id="id14">[<a class="reference internal" href="#id313" title="Adam L. Haber, Moshe Biton, Noga Rogel, Rebecca H. Herbst, Karthik Shekhar, Christopher Smillie, Grace Burgin, Toni M. Delorey, Michael R. Howitt, Yarden Katz, Itay Tirosh, Semir Beyaz, Danielle Dionne, Mei Zhang, Raktima Raychowdhury, Wendy S. Garrett, Orit Rozenblatt-Rosen, Hai Ning Shi, Omer Yilmaz, Ramnik J. Xavier, and Aviv Regev. A single-cell survey of the small intestinal epithelium. Nature, 551(7680):333-339, Nov 2017. URL: https://doi.org/10.1038/nature24489, doi:10.1038/nature24489.">Haber <em>et al.</em>, 2017</a>]</span>. To explain this discrepancy, remember that cell count data is compositional and therefore an increase in the relative abundance of one cell type will lead to a decrease in the relative abundance of all other cell types.
Due to the stark increase of Enterocytes in the small intestinal epithelium of Salmonella-infected mice, all other cell types appear to decrease, even though this shift is only caused by the compositional properties of the data. While the original (univariate) Poisson regression model will pick up these likely false positive effects, scCODA is able to account for the compositionality of the data and therefore does not fall into this trap.</p>
</section>
<section id="with-labeled-clusters-and-hierarchical-structure">
<h2><span class="section-number">18.5. </span>With labeled clusters and hierarchical structure<a class="headerlink" href="#with-labeled-clusters-and-hierarchical-structure" title="Permalink to this headline">#</a></h2>
<p>In addition to the abundance of each cell type, a typical single-cell dataset also contains information about the similarity of the different cells in the form of a tree-based hierarchical ordering. These hierarchies can either be determined automatically via clustering of the gene expression (which is usually done to discover the clusters of cells that belong to the same cell type), or through biologically informed hierarchies like cell lineages.
<a class="reference external" href="https://tasccoda.readthedocs.io/en/latest/">tascCODA</a> is an extension of scCODA that integrates hierarchical information and experimental covariate data into the generative modeling of compositional count data<span id="id15">[<a class="reference internal" href="#id317" title="Johannes Ostner, Salomé Carcy, and Christian L. Müller. Tasccoda: bayesian tree-aggregated analysis of compositional amplicon and single-cell data. Frontiers in Genetics, 2021. URL: https://www.frontiersin.org/article/10.3389/fgene.2021.766405, doi:10.3389/fgene.2021.766405.">Ostner <em>et al.</em>, 2021</a>]</span>. This is especially beneficial for cell atlassing efforts with increased resolution.</p>
<p>At its core, it uses almost the same Dirichlet-Multinomial setup as scCODA, but extends the model, such that effects on sets of cell types, which are defined as internal nodes in the tree structure.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">toytree</span> <span class="k">as</span> <span class="nn">tt</span>
<span class="kn">import</span> <span class="nn">tasccoda.tree_ana</span> <span class="k">as</span> <span class="nn">ta</span>
<span class="kn">import</span> <span class="nn">tasccoda.tree_utils</span> <span class="k">as</span> <span class="nn">tu</span>

<span class="kn">import</span> <span class="nn">schist</span>

<span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span><span class="s2">&quot;ignore&quot;</span><span class="p">)</span>
<span class="n">warnings</span><span class="o">.</span><span class="n">simplefilter</span><span class="p">(</span><span class="s2">&quot;ignore&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>To use tascCODA, we first have to define a hierarchical ordering of the cell types. One possible hierarchical clustering uses the eight cell types and orders them by their similarity (pearson correlation) in the PCA representation with <code class="docutils literal notranslate"><span class="pre">sc.tl.dendrogram</span></code>.
Since this structure is very simple in our data and will therefore not give us many new insights, we want to have a more complex clustering. One recent method to get such clusters, is the <a class="reference external" href="https://github.com/dawe/schist">schist</a> package <span id="id16">[<a class="reference internal" href="#id322" title="Leonardo Morelli, Valentina Giansanti, and Davide Cittaro. Nested stochastic block models applied to the analysis of single cell data. BMC Bioinformatics, 22(1):576, November 2021. URL: http://dx.doi.org/10.1186/s12859-021-04489-7, doi:10.1186/s12859-021-04489-7.">Morelli <em>et al.</em>, 2021</a>]</span>, which uses a nested stochastic block model that clusters the cell population at different resolution levels. Running the method with standard settings takes some time (~15 minutes on our data), and gives us an assignment of each cell to a hierarchical clustering in <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code>.
First, we need to define a distance measure between the cells through a PCA embedding:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># use logcounts to calculate PCA and neighbors</span>
<span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;counts&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;logcounts&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;counts&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">adata</span><span class="o">.</span><span class="n">X</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;logcounts&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">n_pcs</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">1234</span><span class="p">)</span>

<span class="c1"># Calculate UMAP for visualization purposes</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>WARNING: You’re trying to run this on 15215 dimensions of `.X`, if you really want this, set `use_rep=&#39;X&#39;`.
         Falling back to preprocessing with `sc.pp.pca` and default params.
</pre></div>
</div>
</div>
</div>
<p>Then, we can run schist on the AnnData object, which results in a clustering that is defined through a set of columns “nsbm_level_{i}” in <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">schist</span><span class="o">.</span><span class="n">inference</span><span class="o">.</span><span class="n">nested_model</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">5678</span><span class="p">)</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>batch</th>
      <th>barcode</th>
      <th>condition</th>
      <th>cell_label</th>
      <th>nsbm_level_0</th>
      <th>nsbm_level_1</th>
      <th>nsbm_level_2</th>
      <th>nsbm_level_3</th>
      <th>nsbm_level_4</th>
      <th>nsbm_level_5</th>
      <th>nsbm_level_6</th>
    </tr>
    <tr>
      <th>index</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>B1_AAACATACCACAAC_Control_Enterocyte.Progenitor</th>
      <td>B1</td>
      <td>AAACATACCACAAC</td>
      <td>Control</td>
      <td>Enterocyte.Progenitor</td>
      <td>0</td>
      <td>0</td>
      <td>3</td>
      <td>3</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>B1_AAACGCACGAGGAC_Control_Stem</th>
      <td>B1</td>
      <td>AAACGCACGAGGAC</td>
      <td>Control</td>
      <td>Stem</td>
      <td>5</td>
      <td>3</td>
      <td>1</td>
      <td>1</td>
      <td>2</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>B1_AAACGCACTAGCCA_Control_Stem</th>
      <td>B1</td>
      <td>AAACGCACTAGCCA</td>
      <td>Control</td>
      <td>Stem</td>
      <td>9</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>2</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>B1_AAACGCACTGTCCC_Control_Stem</th>
      <td>B1</td>
      <td>AAACGCACTGTCCC</td>
      <td>Control</td>
      <td>Stem</td>
      <td>2</td>
      <td>6</td>
      <td>2</td>
      <td>1</td>
      <td>2</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>B1_AAACTTGACCACCT_Control_Enterocyte.Progenitor</th>
      <td>B1</td>
      <td>AAACTTGACCACCT</td>
      <td>Control</td>
      <td>Enterocyte.Progenitor</td>
      <td>3</td>
      <td>2</td>
      <td>3</td>
      <td>3</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>B10_TTTCACGACAAGCT_Salmonella_TA</th>
      <td>B10</td>
      <td>TTTCACGACAAGCT</td>
      <td>Salmonella</td>
      <td>TA</td>
      <td>16</td>
      <td>3</td>
      <td>1</td>
      <td>1</td>
      <td>2</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>B10_TTTCAGTGAGGCGA_Salmonella_Enterocyte</th>
      <td>B10</td>
      <td>TTTCAGTGAGGCGA</td>
      <td>Salmonella</td>
      <td>Enterocyte</td>
      <td>165</td>
      <td>22</td>
      <td>8</td>
      <td>1</td>
      <td>2</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>B10_TTTCAGTGCGACAT_Salmonella_Stem</th>
      <td>B10</td>
      <td>TTTCAGTGCGACAT</td>
      <td>Salmonella</td>
      <td>Stem</td>
      <td>118</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>2</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>B10_TTTCAGTGTGACCA_Salmonella_Endocrine</th>
      <td>B10</td>
      <td>TTTCAGTGTGACCA</td>
      <td>Salmonella</td>
      <td>Endocrine</td>
      <td>160</td>
      <td>17</td>
      <td>8</td>
      <td>1</td>
      <td>2</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>B10_TTTCAGTGTTCTCA_Salmonella_Enterocyte.Progenitor</th>
      <td>B10</td>
      <td>TTTCAGTGTTCTCA</td>
      <td>Salmonella</td>
      <td>Enterocyte.Progenitor</td>
      <td>32</td>
      <td>8</td>
      <td>5</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
    </tr>
  </tbody>
</table>
<p>9842 rows × 11 columns</p>
</div></div></div>
</div>
<p>A UMAP plot nicely shows how the clustering from schist (here on levels 1 and 2) is connected to the cell type assignments. The representation on level 1 of the hierarchy is hereby a strict refinement of the level above, i.e. each cluster from level 2 is split into multiple smaller clusters:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;nsbm_level_1&quot;</span><span class="p">,</span> <span class="s2">&quot;nsbm_level_2&quot;</span><span class="p">,</span> <span class="s2">&quot;cell_label&quot;</span><span class="p">],</span> <span class="n">ncols</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">wspace</span><span class="o">=</span><span class="mf">0.5</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/compositional_57_0.png" src="../_images/compositional_57_0.png" />
</div>
</div>
<p>To get a data structure that can be used by tascCODA, we must create a new AnnData object in the same way as for scCODA, but with the clustering defined by schist instead of the cell types from before.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Group data</span>
<span class="n">tasccoda_frac</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
        <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;condition&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span>
            <span class="p">[</span><span class="s2">&quot;Control&quot;</span><span class="p">,</span> <span class="s2">&quot;Salmonella&quot;</span><span class="p">,</span> <span class="s2">&quot;Hpoly.Day3&quot;</span><span class="p">,</span> <span class="s2">&quot;Hpoly.Day10&quot;</span><span class="p">]</span>
        <span class="p">)</span>
    <span class="p">]</span>
    <span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s2">&quot;condition&quot;</span><span class="p">,</span> <span class="s2">&quot;batch&quot;</span><span class="p">])</span>
    <span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">value_counts</span><span class="p">(</span><span class="s2">&quot;nsbm_level_1&quot;</span><span class="p">,</span> <span class="n">normalize</span><span class="o">=</span><span class="kc">False</span><span class="p">))</span>
    <span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;n_cells&quot;</span><span class="p">)</span>
    <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">condition</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;condition&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">))</span>
<span class="p">)</span>

<span class="c1"># Pivot data</span>
<span class="n">tasccoda_pivot</span> <span class="o">=</span> <span class="n">tasccoda_frac</span><span class="o">.</span><span class="n">pivot</span><span class="p">(</span>
    <span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;batch&quot;</span><span class="p">,</span> <span class="s2">&quot;condition&quot;</span><span class="p">],</span>
    <span class="n">columns</span><span class="o">=</span><span class="s2">&quot;nsbm_level_1&quot;</span><span class="p">,</span>
    <span class="n">values</span><span class="o">=</span><span class="s2">&quot;n_cells&quot;</span><span class="p">,</span>
<span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>

<span class="n">tasccoda_data</span> <span class="o">=</span> <span class="n">scc_dat</span><span class="o">.</span><span class="n">from_pandas</span><span class="p">(</span>
    <span class="n">tasccoda_pivot</span><span class="p">,</span> <span class="n">covariate_columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;batch&quot;</span><span class="p">,</span> <span class="s2">&quot;condition&quot;</span><span class="p">]</span>
<span class="p">)</span>
<span class="n">tasccoda_data</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 10 × 38
    obs: &#39;batch&#39;, &#39;condition&#39;
</pre></div>
</div>
</div>
</div>
<p>Additionally, we need to convert our tree representation into a <a class="reference external" href="https://toytree.readthedocs.io/">toytree</a> structure and save it as <code class="docutils literal notranslate"><span class="pre">tasccoda_data.uns[&quot;phylo_tree&quot;]</span></code>. The <a class="reference external" href="https://tasccoda.readthedocs.io/en/latest/tascCODA_tutorial.html">tascCODA tutorial</a> shows a way to create a toytree object from a DataFrame via a <a class="reference external" href="https://en.wikipedia.org/wiki/Newick_format">Newick string</a>, which we replicate here. To get some clusters that are not too small, we cut the tree before the last level by leaving out <code class="docutils literal notranslate"><span class="pre">&quot;nsbm_level_0&quot;</span></code>. Note that not only the leaves of the trees have names (0-39), but also the internal nodes are named now with labels 40-61:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Convert cluster names to strings</span>
<span class="n">n_levels</span> <span class="o">=</span> <span class="mi">6</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_levels</span><span class="p">):</span>
    <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;nsbm_level_</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;nsbm_level_</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>

<span class="c1"># Generate tree</span>
<span class="n">newick</span> <span class="o">=</span> <span class="n">tu</span><span class="o">.</span><span class="n">df2newick</span><span class="p">(</span>
    <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(),</span>
    <span class="n">levels</span><span class="o">=</span><span class="p">[</span>
        <span class="s2">&quot;nsbm_level_5&quot;</span><span class="p">,</span>
        <span class="s2">&quot;nsbm_level_4&quot;</span><span class="p">,</span>
        <span class="s2">&quot;nsbm_level_3&quot;</span><span class="p">,</span>
        <span class="s2">&quot;nsbm_level_2&quot;</span><span class="p">,</span>
        <span class="s2">&quot;nsbm_level_1&quot;</span><span class="p">,</span>
    <span class="p">],</span>
<span class="p">)</span>
<span class="n">tree</span> <span class="o">=</span> <span class="n">tt</span><span class="o">.</span><span class="n">tree</span><span class="p">(</span><span class="n">newick</span><span class="p">)</span>

<span class="n">tasccoda_data</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s2">&quot;phylo_tree&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">tree</span>
<span class="n">tasccoda_data</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">tasccoda_data</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
<span class="c1"># draw tree</span>
<span class="n">tree</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span>
    <span class="n">tip_labels_align</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">node_sizes</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
    <span class="n">node_labels</span><span class="o">=</span><span class="s2">&quot;name&quot;</span><span class="p">,</span>
    <span class="n">width</span><span class="o">=</span><span class="mi">600</span><span class="p">,</span>
    <span class="n">node_markers</span><span class="o">=</span><span class="s2">&quot;r2x1.25&quot;</span><span class="p">,</span>
    <span class="n">node_colors</span><span class="o">=</span><span class="s2">&quot;lightblue&quot;</span><span class="p">,</span>
    <span class="n">layout</span><span class="o">=</span><span class="s2">&quot;d&quot;</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># Replace NaN values with 0 counts</span>
<span class="n">tasccoda_data</span><span class="o">.</span><span class="n">X</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan_to_num</span><span class="p">(</span><span class="n">tasccoda_data</span><span class="o">.</span><span class="n">X</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div class="toyplot" id="tb6ccbb640d2b4c69a8983405e956a388" style="text-align:center"><svg class="toyplot-canvas-Canvas" xmlns:toyplot="http://www.sandia.gov/toyplot" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns="http://www.w3.org/2000/svg" width="600.0px" height="260.0px" viewBox="0 0 600.0 260.0" preserveAspectRatio="xMidYMid meet" style="background-color:transparent;border-color:#292724;border-style:none;border-width:1.0;fill:rgb(16.1%,15.3%,14.1%);fill-opacity:1.0;font-family:Helvetica;font-size:12px;opacity:1.0;stroke:rgb(16.1%,15.3%,14.1%);stroke-opacity:1.0;stroke-width:1.0" id="t00eb7b69b5754338a5d7beda9ed4e118"><g class="toyplot-coordinates-Cartesian" id="t3abe098f2f5345e5b066d3157c62f435"><clipPath id="t958a44aff2e6450bbfac66a5b1503990"><rect x="30.0" y="30.0" width="540.0" height="200.0"></rect></clipPath><g clip-path="url(#t958a44aff2e6450bbfac66a5b1503990)"><g class="toytree-mark-Toytree" id="t99122de1ae3c4661ba8cb7d26acd8747"><g class="toytree-Edges" style="fill:none;stroke:rgb(14.9%,14.9%,14.9%);stroke-linecap:round;stroke-opacity:1;stroke-width:2"><path d="M 360.8 50.0 L 360.8 50.0 L 360.8 79.3" id="61,60"></path><path d="M 360.8 79.3 L 513.8 79.3 L 513.8 108.5" id="60,59"></path><path d="M 360.8 79.3 L 387.4 79.3 L 387.4 108.5" id="60,58"></path><path d="M 360.8 79.3 L 181.3 79.3 L 181.3 108.5" id="60,57"></path><path d="M 513.8 108.5 L 513.8 108.5 L 513.8 137.8" id="59,56"></path><path d="M 387.4 108.5 L 434.7 108.5 L 434.7 137.8" id="58,55"></path><path d="M 387.4 108.5 L 340.1 108.5 L 340.1 137.8" id="58,54"></path><path d="M 181.3 108.5 L 181.3 108.5 L 181.3 137.8" id="57,53"></path><path d="M 513.8 137.8 L 540.2 137.8 L 540.2 167.0" id="56,52"></path><path d="M 513.8 137.8 L 527.0 137.8 L 527.0 167.0" id="56,51"></path><path d="M 513.8 137.8 L 507.2 137.8 L 507.2 167.0" id="56,50"></path><path d="M 513.8 137.8 L 480.8 137.8 L 480.8 167.0" id="56,49"></path><path d="M 434.7 137.8 L 461.1 137.8 L 461.1 167.0" id="55,48"></path><path d="M 434.7 137.8 L 441.3 137.8 L 441.3 167.0" id="55,47"></path><path d="M 434.7 137.8 L 401.7 137.8 L 401.7 167.0" id="55,46"></path><path d="M 340.1 137.8 L 368.7 137.8 L 368.7 167.0" id="54,45"></path><path d="M 340.1 137.8 L 348.9 137.8 L 348.9 167.0" id="54,44"></path><path d="M 340.1 137.8 L 302.7 137.8 L 302.7 167.0" id="54,43"></path><path d="M 181.3 137.8 L 263.1 137.8 L 263.1 167.0" id="53,42"></path><path d="M 181.3 137.8 L 236.7 137.8 L 236.7 167.0" id="53,41"></path><path d="M 181.3 137.8 L 190.5 137.8 L 190.5 167.0" id="53,40"></path><path d="M 181.3 137.8 L 137.7 137.8 L 137.7 167.0" id="53,39"></path><path d="M 181.3 137.8 L 78.3 137.8 L 78.3 167.0" id="53,38"></path><path d="M 540.2 167.0 L 540.2 167.0 L 540.2 196.3" id="52,37"></path><path d="M 527.0 167.0 L 527.0 167.0 L 527.0 196.3" id="51,36"></path><path d="M 507.2 167.0 L 513.8 167.0 L 513.8 196.3" id="50,35"></path><path d="M 507.2 167.0 L 500.6 167.0 L 500.6 196.3" id="50,34"></path><path d="M 480.8 167.0 L 487.4 167.0 L 487.4 196.3" id="49,33"></path><path d="M 480.8 167.0 L 474.3 167.0 L 474.3 196.3" id="49,32"></path><path d="M 461.1 167.0 L 461.1 167.0 L 461.1 196.3" id="48,31"></path><path d="M 441.3 167.0 L 447.9 167.0 L 447.9 196.3" id="47,30"></path><path d="M 441.3 167.0 L 434.7 167.0 L 434.7 196.3" id="47,29"></path><path d="M 401.7 167.0 L 421.5 167.0 L 421.5 196.3" id="46,28"></path><path d="M 401.7 167.0 L 408.3 167.0 L 408.3 196.3" id="46,27"></path><path d="M 401.7 167.0 L 395.1 167.0 L 395.1 196.3" id="46,26"></path><path d="M 401.7 167.0 L 381.9 167.0 L 381.9 196.3" id="46,25"></path><path d="M 368.7 167.0 L 368.7 167.0 L 368.7 196.3" id="45,24"></path><path d="M 348.9 167.0 L 355.5 167.0 L 355.5 196.3" id="44,23"></path><path d="M 348.9 167.0 L 342.3 167.0 L 342.3 196.3" id="44,22"></path><path d="M 302.7 167.0 L 329.1 167.0 L 329.1 196.3" id="43,21"></path><path d="M 302.7 167.0 L 315.9 167.0 L 315.9 196.3" id="43,20"></path><path d="M 302.7 167.0 L 302.7 167.0 L 302.7 196.3" id="43,19"></path><path d="M 302.7 167.0 L 289.5 167.0 L 289.5 196.3" id="43,18"></path><path d="M 302.7 167.0 L 276.3 167.0 L 276.3 196.3" id="43,17"></path><path d="M 263.1 167.0 L 263.1 167.0 L 263.1 196.3" id="42,16"></path><path d="M 236.7 167.0 L 249.9 167.0 L 249.9 196.3" id="41,15"></path><path d="M 236.7 167.0 L 236.7 167.0 L 236.7 196.3" id="41,14"></path><path d="M 236.7 167.0 L 223.5 167.0 L 223.5 196.3" id="41,13"></path><path d="M 190.5 167.0 L 210.3 167.0 L 210.3 196.3" id="40,12"></path><path d="M 190.5 167.0 L 197.1 167.0 L 197.1 196.3" id="40,11"></path><path d="M 190.5 167.0 L 183.9 167.0 L 183.9 196.3" id="40,10"></path><path d="M 190.5 167.0 L 170.7 167.0 L 170.7 196.3" id="40,9"></path><path d="M 137.7 167.0 L 157.5 167.0 L 157.5 196.3" id="39,8"></path><path d="M 137.7 167.0 L 144.3 167.0 L 144.3 196.3" id="39,7"></path><path d="M 137.7 167.0 L 131.1 167.0 L 131.1 196.3" id="39,6"></path><path d="M 137.7 167.0 L 117.9 167.0 L 117.9 196.3" id="39,5"></path><path d="M 78.3 167.0 L 104.7 167.0 L 104.7 196.3" id="38,4"></path><path d="M 78.3 167.0 L 91.5 167.0 L 91.5 196.3" id="38,3"></path><path d="M 78.3 167.0 L 78.3 167.0 L 78.3 196.3" id="38,2"></path><path d="M 78.3 167.0 L 65.1 167.0 L 65.1 196.3" id="38,1"></path><path d="M 78.3 167.0 L 52.0 167.0 L 52.0 196.3" id="38,0"></path></g><g class="toytree-AlignEdges" style="stroke:rgb(66.3%,66.3%,66.3%);stroke-dasharray:2, 4;stroke-linecap:round;stroke-opacity:1.0;stroke-width:2"><path d="M 52.0 196.3 L 52.0 196.3"></path><path d="M 65.1 196.3 L 65.1 196.3"></path><path d="M 78.3 196.3 L 78.3 196.3"></path><path d="M 91.5 196.3 L 91.5 196.3"></path><path d="M 104.7 196.3 L 104.7 196.3"></path><path d="M 117.9 196.3 L 117.9 196.3"></path><path d="M 131.1 196.3 L 131.1 196.3"></path><path d="M 144.3 196.3 L 144.3 196.3"></path><path d="M 157.5 196.3 L 157.5 196.3"></path><path d="M 170.7 196.3 L 170.7 196.3"></path><path d="M 183.9 196.3 L 183.9 196.3"></path><path d="M 197.1 196.3 L 197.1 196.3"></path><path d="M 210.3 196.3 L 210.3 196.3"></path><path d="M 223.5 196.3 L 223.5 196.3"></path><path d="M 236.7 196.3 L 236.7 196.3"></path><path d="M 249.9 196.3 L 249.9 196.3"></path><path d="M 263.1 196.3 L 263.1 196.3"></path><path d="M 276.3 196.3 L 276.3 196.3"></path><path d="M 289.5 196.3 L 289.5 196.3"></path><path d="M 302.7 196.3 L 302.7 196.3"></path><path d="M 315.9 196.3 L 315.9 196.3"></path><path d="M 329.1 196.3 L 329.1 196.3"></path><path d="M 342.3 196.3 L 342.3 196.3"></path><path d="M 355.5 196.3 L 355.5 196.3"></path><path d="M 368.7 196.3 L 368.7 196.3"></path><path d="M 381.9 196.3 L 381.9 196.3"></path><path d="M 395.1 196.3 L 395.1 196.3"></path><path d="M 408.3 196.3 L 408.3 196.3"></path><path d="M 421.5 196.3 L 421.5 196.3"></path><path d="M 434.7 196.3 L 434.7 196.3"></path><path d="M 447.9 196.3 L 447.9 196.3"></path><path d="M 461.1 196.3 L 461.1 196.3"></path><path d="M 474.3 196.3 L 474.3 196.3"></path><path d="M 487.4 196.3 L 487.4 196.3"></path><path d="M 500.6 196.3 L 500.6 196.3"></path><path d="M 513.8 196.3 L 513.8 196.3"></path><path d="M 527.0 196.3 L 527.0 196.3"></path><path d="M 540.2 196.3 L 540.2 196.3"></path></g><g class="toytree-Nodes" style="fill:rgb(67.8%,84.7%,90.2%);fill-opacity:1.0;stroke:None;stroke-width:1"><g id="node-0" transform="translate(51.953,196.286)"><rect x="0.0" y="0.0" width="0.0" height="0.0"></rect></g><g id="node-1" transform="translate(65.150,196.286)"><rect x="0.0" y="0.0" width="0.0" height="0.0"></rect></g><g id="node-2" transform="translate(78.347,196.286)"><rect x="0.0" y="0.0" width="0.0" height="0.0"></rect></g><g id="node-3" transform="translate(91.543,196.286)"><rect x="0.0" y="0.0" width="0.0" height="0.0"></rect></g><g id="node-4" transform="translate(104.740,196.286)"><rect x="0.0" y="0.0" width="0.0" height="0.0"></rect></g><g id="node-5" transform="translate(117.937,196.286)"><rect x="0.0" y="0.0" width="0.0" height="0.0"></rect></g><g id="node-6" transform="translate(131.134,196.286)"><rect x="0.0" y="0.0" width="0.0" height="0.0"></rect></g><g id="node-7" transform="translate(144.331,196.286)"><rect x="0.0" y="0.0" width="0.0" height="0.0"></rect></g><g id="node-8" transform="translate(157.527,196.286)"><rect x="0.0" y="0.0" width="0.0" height="0.0"></rect></g><g id="node-9" transform="translate(170.724,196.286)"><rect x="0.0" y="0.0" width="0.0" height="0.0"></rect></g><g id="node-10" transform="translate(183.921,196.286)"><rect x="0.0" y="0.0" width="0.0" height="0.0"></rect></g><g id="node-11" transform="translate(197.118,196.286)"><rect x="0.0" y="0.0" width="0.0" height="0.0"></rect></g><g id="node-12" transform="translate(210.315,196.286)"><rect x="0.0" y="0.0" width="0.0" height="0.0"></rect></g><g id="node-13" transform="translate(223.511,196.286)"><rect x="0.0" y="0.0" width="0.0" height="0.0"></rect></g><g id="node-14" transform="translate(236.708,196.286)"><rect x="0.0" y="0.0" width="0.0" height="0.0"></rect></g><g id="node-15" transform="translate(249.905,196.286)"><rect x="0.0" y="0.0" width="0.0" height="0.0"></rect></g><g id="node-16" transform="translate(263.102,196.286)"><rect x="0.0" y="0.0" width="0.0" height="0.0"></rect></g><g id="node-17" transform="translate(276.299,196.286)"><rect x="0.0" y="0.0" width="0.0" height="0.0"></rect></g><g id="node-18" transform="translate(289.495,196.286)"><rect x="0.0" y="0.0" width="0.0" height="0.0"></rect></g><g id="node-19" transform="translate(302.692,196.286)"><rect x="0.0" y="0.0" width="0.0" height="0.0"></rect></g><g id="node-20" transform="translate(315.889,196.286)"><rect x="0.0" y="0.0" width="0.0" height="0.0"></rect></g><g id="node-21" transform="translate(329.086,196.286)"><rect x="0.0" y="0.0" width="0.0" height="0.0"></rect></g><g id="node-22" transform="translate(342.283,196.286)"><rect x="0.0" y="0.0" width="0.0" height="0.0"></rect></g><g id="node-23" transform="translate(355.479,196.286)"><rect x="0.0" y="0.0" width="0.0" height="0.0"></rect></g><g id="node-24" transform="translate(368.676,196.286)"><rect x="0.0" y="0.0" width="0.0" height="0.0"></rect></g><g id="node-25" transform="translate(381.873,196.286)"><rect x="0.0" y="0.0" width="0.0" height="0.0"></rect></g><g id="node-26" transform="translate(395.070,196.286)"><rect x="0.0" y="0.0" width="0.0" height="0.0"></rect></g><g id="node-27" transform="translate(408.266,196.286)"><rect x="0.0" y="0.0" width="0.0" height="0.0"></rect></g><g id="node-28" transform="translate(421.463,196.286)"><rect x="0.0" y="0.0" width="0.0" height="0.0"></rect></g><g id="node-29" transform="translate(434.660,196.286)"><rect x="0.0" y="0.0" width="0.0" height="0.0"></rect></g><g id="node-30" transform="translate(447.857,196.286)"><rect x="0.0" y="0.0" width="0.0" height="0.0"></rect></g><g id="node-31" transform="translate(461.054,196.286)"><rect x="0.0" y="0.0" width="0.0" height="0.0"></rect></g><g id="node-32" transform="translate(474.250,196.286)"><rect x="0.0" y="0.0" width="0.0" height="0.0"></rect></g><g id="node-33" transform="translate(487.447,196.286)"><rect x="0.0" y="0.0" width="0.0" height="0.0"></rect></g><g id="node-34" transform="translate(500.644,196.286)"><rect x="0.0" y="0.0" width="0.0" height="0.0"></rect></g><g id="node-35" transform="translate(513.841,196.286)"><rect x="0.0" y="0.0" width="0.0" height="0.0"></rect></g><g id="node-36" transform="translate(527.038,196.286)"><rect x="0.0" y="0.0" width="0.0" height="0.0"></rect></g><g id="node-37" transform="translate(540.234,196.286)"><rect x="0.0" y="0.0" width="0.0" height="0.0"></rect></g><g id="node-38" transform="translate(78.347,167.029)"><rect x="-10.0" y="-6.25" width="20.0" height="12.5"></rect></g><g id="node-39" transform="translate(137.732,167.029)"><rect x="-10.0" y="-6.25" width="20.0" height="12.5"></rect></g><g id="node-40" transform="translate(190.519,167.029)"><rect x="-10.0" y="-6.25" width="20.0" height="12.5"></rect></g><g id="node-41" transform="translate(236.708,167.029)"><rect x="-10.0" y="-6.25" width="20.0" height="12.5"></rect></g><g id="node-42" transform="translate(263.102,167.029)"><rect x="-10.0" y="-6.25" width="20.0" height="12.5"></rect></g><g id="node-43" transform="translate(302.692,167.029)"><rect x="-10.0" y="-6.25" width="20.0" height="12.5"></rect></g><g id="node-44" transform="translate(348.881,167.029)"><rect x="-10.0" y="-6.25" width="20.0" height="12.5"></rect></g><g id="node-45" transform="translate(368.676,167.029)"><rect x="-10.0" y="-6.25" width="20.0" height="12.5"></rect></g><g id="node-46" transform="translate(401.668,167.029)"><rect x="-10.0" y="-6.25" width="20.0" height="12.5"></rect></g><g id="node-47" transform="translate(441.258,167.029)"><rect x="-10.0" y="-6.25" width="20.0" height="12.5"></rect></g><g id="node-48" transform="translate(461.054,167.029)"><rect x="-10.0" y="-6.25" width="20.0" height="12.5"></rect></g><g id="node-49" transform="translate(480.849,167.029)"><rect x="-10.0" y="-6.25" width="20.0" height="12.5"></rect></g><g id="node-50" transform="translate(507.242,167.029)"><rect x="-10.0" y="-6.25" width="20.0" height="12.5"></rect></g><g id="node-51" transform="translate(527.038,167.029)"><rect x="-10.0" y="-6.25" width="20.0" height="12.5"></rect></g><g id="node-52" transform="translate(540.234,167.029)"><rect x="-10.0" y="-6.25" width="20.0" height="12.5"></rect></g><g id="node-53" transform="translate(181.282,137.771)"><rect x="-10.0" y="-6.25" width="20.0" height="12.5"></rect></g><g id="node-54" transform="translate(340.083,137.771)"><rect x="-10.0" y="-6.25" width="20.0" height="12.5"></rect></g><g id="node-55" transform="translate(434.660,137.771)"><rect x="-10.0" y="-6.25" width="20.0" height="12.5"></rect></g><g id="node-56" transform="translate(513.841,137.771)"><rect x="-10.0" y="-6.25" width="20.0" height="12.5"></rect></g><g id="node-57" transform="translate(181.282,108.514)"><rect x="-10.0" y="-6.25" width="20.0" height="12.5"></rect></g><g id="node-58" transform="translate(387.372,108.514)"><rect x="-10.0" y="-6.25" width="20.0" height="12.5"></rect></g><g id="node-59" transform="translate(513.841,108.514)"><rect x="-10.0" y="-6.25" width="20.0" height="12.5"></rect></g><g id="node-60" transform="translate(360.831,79.257)"><rect x="-10.0" y="-6.25" width="20.0" height="12.5"></rect></g><g id="node-61" transform="translate(360.831,50.000)"><rect x="0.0" y="0.0" width="0.0" height="0.0"></rect></g></g><g class="toytree-NodeLabels" style="fill:rgb(14.9%,14.9%,14.9%);fill-opacity:1.0;font-size:9px;stroke:none"><g transform="translate(73.34,169.33)"><text>38</text></g><g transform="translate(132.73,169.33)"><text>39</text></g><g transform="translate(185.52,169.33)"><text>40</text></g><g transform="translate(231.70,169.33)"><text>41</text></g><g transform="translate(258.10,169.33)"><text>42</text></g><g transform="translate(297.69,169.33)"><text>43</text></g><g transform="translate(343.88,169.33)"><text>44</text></g><g transform="translate(363.67,169.33)"><text>45</text></g><g transform="translate(396.66,169.33)"><text>46</text></g><g transform="translate(436.25,169.33)"><text>47</text></g><g transform="translate(456.05,169.33)"><text>48</text></g><g transform="translate(475.84,169.33)"><text>49</text></g><g transform="translate(502.24,169.33)"><text>50</text></g><g transform="translate(522.03,169.33)"><text>51</text></g><g transform="translate(535.23,169.33)"><text>52</text></g><g transform="translate(176.28,140.07)"><text>53</text></g><g transform="translate(335.08,140.07)"><text>54</text></g><g transform="translate(429.66,140.07)"><text>55</text></g><g transform="translate(508.84,140.07)"><text>56</text></g><g transform="translate(176.28,110.81)"><text>57</text></g><g transform="translate(382.37,110.81)"><text>58</text></g><g transform="translate(508.84,110.81)"><text>59</text></g><g transform="translate(355.83,81.56)"><text>60</text></g></g><g class="toytree-TipLabels" style="fill:rgb(14.9%,14.9%,14.9%);fill-opacity:1.0;font-family:helvetica;font-size:11px;font-weight:normal;stroke:none;white-space:pre"><g transform="translate(51.95,196.29)rotate(-90)"><text x="-27.23" y="2.81" style="">36</text></g><g transform="translate(65.15,196.29)rotate(-90)"><text x="-27.23" y="2.81" style="">14</text></g><g transform="translate(78.35,196.29)rotate(-90)"><text x="-27.23" y="2.81" style="">17</text></g><g transform="translate(91.54,196.29)rotate(-90)"><text x="-27.23" y="2.81" style="">11</text></g><g transform="translate(104.74,196.29)rotate(-90)"><text x="-27.23" y="2.81" style="">22</text></g><g transform="translate(117.94,196.29)rotate(-90)"><text x="-27.23" y="2.81" style="">31</text></g><g transform="translate(131.13,196.29)rotate(-90)"><text x="-27.23" y="2.81" style="">21</text></g><g transform="translate(144.33,196.29)rotate(-90)"><text x="-21.12" y="2.81" style="">5</text></g><g transform="translate(157.53,196.29)rotate(-90)"><text x="-27.23" y="2.81" style="">25</text></g><g transform="translate(170.72,196.29)rotate(-90)"><text x="-21.12" y="2.81" style="">4</text></g><g transform="translate(183.92,196.29)rotate(-90)"><text x="-27.23" y="2.81" style="">19</text></g><g transform="translate(197.12,196.29)rotate(-90)"><text x="-21.12" y="2.81" style="">1</text></g><g transform="translate(210.31,196.29)rotate(-90)"><text x="-21.12" y="2.81" style="">3</text></g><g transform="translate(223.51,196.29)rotate(-90)"><text x="-21.12" y="2.81" style="">9</text></g><g transform="translate(236.71,196.29)rotate(-90)"><text x="-27.23" y="2.81" style="">10</text></g><g transform="translate(249.90,196.29)rotate(-90)"><text x="-21.12" y="2.81" style="">6</text></g><g transform="translate(263.10,196.29)rotate(-90)"><text x="-27.23" y="2.81" style="">34</text></g><g transform="translate(276.30,196.29)rotate(-90)"><text x="-27.23" y="2.81" style="">37</text></g><g transform="translate(289.50,196.29)rotate(-90)"><text x="-27.23" y="2.81" style="">16</text></g><g transform="translate(302.69,196.29)rotate(-90)"><text x="-27.23" y="2.81" style="">12</text></g><g transform="translate(315.89,196.29)rotate(-90)"><text x="-21.12" y="2.81" style="">8</text></g><g transform="translate(329.09,196.29)rotate(-90)"><text x="-21.12" y="2.81" style="">7</text></g><g transform="translate(342.28,196.29)rotate(-90)"><text x="-27.23" y="2.81" style="">20</text></g><g transform="translate(355.48,196.29)rotate(-90)"><text x="-27.23" y="2.81" style="">26</text></g><g transform="translate(368.68,196.29)rotate(-90)"><text x="-27.23" y="2.81" style="">15</text></g><g transform="translate(381.87,196.29)rotate(-90)"><text x="-27.23" y="2.81" style="">32</text></g><g transform="translate(395.07,196.29)rotate(-90)"><text x="-27.23" y="2.81" style="">35</text></g><g transform="translate(408.27,196.29)rotate(-90)"><text x="-21.12" y="2.81" style="">2</text></g><g transform="translate(421.46,196.29)rotate(-90)"><text x="-21.12" y="2.81" style="">0</text></g><g transform="translate(434.66,196.29)rotate(-90)"><text x="-27.23" y="2.81" style="">29</text></g><g transform="translate(447.86,196.29)rotate(-90)"><text x="-27.23" y="2.81" style="">27</text></g><g transform="translate(461.05,196.29)rotate(-90)"><text x="-27.23" y="2.81" style="">33</text></g><g transform="translate(474.25,196.29)rotate(-90)"><text x="-27.23" y="2.81" style="">30</text></g><g transform="translate(487.45,196.29)rotate(-90)"><text x="-27.23" y="2.81" style="">28</text></g><g transform="translate(500.64,196.29)rotate(-90)"><text x="-27.23" y="2.81" style="">18</text></g><g transform="translate(513.84,196.29)rotate(-90)"><text x="-27.23" y="2.81" style="">23</text></g><g transform="translate(527.04,196.29)rotate(-90)"><text x="-27.23" y="2.81" style="">24</text></g><g transform="translate(540.23,196.29)rotate(-90)"><text x="-27.23" y="2.81" style="">13</text></g></g></g></g></g></svg><div class="toyplot-behavior"><script>(function()
{
var modules={};
})();</script></div></div></div></div>
</div>
<p>The model setup and execution in tascCODA works analogous to scCODA, and also the free parameters for the reference and the formula are the same. Additionally, we can adjust the tree aggregation and model selection via the parameters <code class="docutils literal notranslate"><span class="pre">phi</span></code> and <code class="docutils literal notranslate"><span class="pre">lambda_1</span></code> in the <code class="docutils literal notranslate"><span class="pre">pen_args</span></code> argument (see {cite}Ostner2021 for more information). Here, we use an unbiased setting <code class="docutils literal notranslate"><span class="pre">phi=0</span></code> and a model selection that is slightly less strict than the default with <code class="docutils literal notranslate"><span class="pre">lambda_1=1.7</span></code>. We use cluster 18 as our reference, since it is almost identical to the set of Endocrine cells.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">42</span><span class="p">)</span>
<span class="n">tf</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">set_seed</span><span class="p">(</span><span class="mi">42</span><span class="p">)</span>

<span class="n">tasccoda_mod</span> <span class="o">=</span> <span class="n">ta</span><span class="o">.</span><span class="n">CompositionalAnalysisTree</span><span class="p">(</span>
    <span class="n">tasccoda_data</span><span class="p">,</span>
    <span class="n">reference_cell_type</span><span class="o">=</span><span class="s2">&quot;18&quot;</span><span class="p">,</span>
    <span class="n">formula</span><span class="o">=</span><span class="s2">&quot;condition&quot;</span><span class="p">,</span>
    <span class="n">pen_args</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;phi&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;lambda_1&quot;</span><span class="p">:</span> <span class="mf">1.7</span><span class="p">},</span>
<span class="p">)</span>

<span class="n">tasccoda_res</span> <span class="o">=</span> <span class="n">tasccoda_mod</span><span class="o">.</span><span class="n">sample_hmc_da</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Zero counts encountered in data! Added a pseudocount of 0.5.
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>  0%|          | 0/20000 [00:00&lt;?, ?it/s]2022-08-25 17:27:46.895341: W tensorflow/core/framework/op_kernel.cc:1745] OP_REQUIRES failed at functional_ops.cc:374 : INTERNAL: No function library
2022-08-25 17:27:46.896194: W tensorflow/core/framework/op_kernel.cc:1745] OP_REQUIRES failed at functional_ops.cc:374 : INTERNAL: No function library
2022-08-25 17:27:46.915013: W tensorflow/core/framework/op_kernel.cc:1745] OP_REQUIRES failed at functional_ops.cc:374 : INTERNAL: No function library
2022-08-25 17:27:46.915082: W tensorflow/core/framework/op_kernel.cc:1745] OP_REQUIRES failed at functional_ops.cc:374 : INTERNAL: No function library
2022-08-25 17:27:46.928975: W tensorflow/core/framework/op_kernel.cc:1745] OP_REQUIRES failed at functional_ops.cc:374 : INTERNAL: No function library
2022-08-25 17:27:46.929045: W tensorflow/core/framework/op_kernel.cc:1745] OP_REQUIRES failed at functional_ops.cc:374 : INTERNAL: No function library
2022-08-25 17:27:46.950307: W tensorflow/core/framework/op_kernel.cc:1745] OP_REQUIRES failed at functional_ops.cc:374 : INTERNAL: No function library
2022-08-25 17:27:46.950348: W tensorflow/core/framework/op_kernel.cc:1745] OP_REQUIRES failed at functional_ops.cc:374 : INTERNAL: No function library
100%|██████████| 20000/20000 [01:44&lt;00:00, 191.22it/s]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>MCMC sampling finished. (137.542 sec)
Acceptance rate: 87.2%
</pre></div>
</div>
</div>
</div>
<p>Again, the acceptance probability is right around the desired value of 0.85 for tascCODA, indicating no apparent problems with the optimization.</p>
<p>The result from tascCODA should first and foremost be interpreted as effects on the nodes of the tree. A nonzero parameter on a node means that the aggregated count of all cell types under that node changes significantly.
We can easily visualize this in a tree plot for each of the three disease states. Blue circles indicate an increase, red circles a decrease:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">covariate</span> <span class="o">=</span> <span class="s2">&quot;condition[T.Salmonella]_node&quot;</span>
<span class="n">tasccoda_res</span><span class="o">.</span><span class="n">draw_tree_effects</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span> <span class="n">covariate</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div class="toyplot" id="t3b65fe0a02ef4287a53b3a45b9b4d6fa" style="text-align:center"><svg class="toyplot-canvas-Canvas" xmlns:toyplot="http://www.sandia.gov/toyplot" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns="http://www.w3.org/2000/svg" width="260.0px" height="684.0px" viewBox="0 0 260.0 684.0" preserveAspectRatio="xMidYMid meet" style="background-color:transparent;border-color:#292724;border-style:none;border-width:1.0;fill:rgb(16.1%,15.3%,14.1%);fill-opacity:1.0;font-family:Helvetica;font-size:12px;opacity:1.0;stroke:rgb(16.1%,15.3%,14.1%);stroke-opacity:1.0;stroke-width:1.0" id="td4f47cf52a964bcf86d1bdd4974855fa"><g class="toyplot-coordinates-Cartesian" id="te5feedfc8f0e48f5ac570a6ebedc516b"><clipPath id="t12460c809eeb4edfb8ed9b9bb7658bd1"><rect x="30.0" y="30.0" width="200.0" height="624.0"></rect></clipPath><g clip-path="url(#t12460c809eeb4edfb8ed9b9bb7658bd1)"><g class="toytree-mark-Toytree" id="t951211b30a86410aa19701d335dabeda"><g class="toytree-Edges" style="fill:none;stroke:rgb(14.9%,14.9%,14.9%);stroke-linecap:round;stroke-opacity:1;stroke-width:2"><path d="M 50.0 337.9 L 50.0 337.9 L 77.9 337.9" id="54,53"></path><path d="M 77.9 337.9 L 77.9 169.3 L 105.8 169.3" id="53,52"></path><path d="M 77.9 337.9 L 77.9 334.4 L 105.8 334.4" id="53,51"></path><path d="M 77.9 337.9 L 77.9 510.0 L 105.8 510.0" id="53,50"></path><path d="M 105.8 169.3 L 105.8 112.5 L 133.8 112.5" id="52,49"></path><path d="M 105.8 169.3 L 105.8 226.0 L 133.8 226.0" id="52,48"></path><path d="M 105.8 334.4 L 105.8 296.6 L 133.8 296.6" id="51,47"></path><path d="M 105.8 334.4 L 105.8 326.9 L 133.8 326.9" id="51,46"></path><path d="M 105.8 510.0 L 105.8 395.0 L 133.8 395.0" id="50,45"></path><path d="M 105.8 510.0 L 105.8 447.9 L 133.8 447.9" id="50,44"></path><path d="M 105.8 510.0 L 105.8 508.4 L 133.8 508.4" id="50,43"></path><path d="M 105.8 510.0 L 105.8 576.5 L 133.8 576.5" id="50,42"></path><path d="M 133.8 112.5 L 133.8 69.6 L 161.7 69.6" id="49,41"></path><path d="M 133.8 112.5 L 133.8 115.0 L 161.7 115.0" id="49,40"></path><path d="M 133.8 226.0 L 133.8 175.6 L 161.7 175.6" id="48,39"></path><path d="M 133.8 226.0 L 133.8 228.5 L 161.7 228.5" id="48,38"></path><path d="M 161.7 69.6 L 161.7 62.1 L 189.6 62.1" id="41,37"></path><path d="M 161.7 69.6 L 161.7 77.2 L 189.6 77.2" id="41,36"></path><path d="M 161.7 115.0 L 161.7 92.3 L 189.6 92.3" id="40,35"></path><path d="M 161.7 115.0 L 161.7 107.5 L 189.6 107.5" id="40,34"></path><path d="M 161.7 115.0 L 161.7 122.6 L 189.6 122.6" id="40,33"></path><path d="M 161.7 115.0 L 161.7 137.7 L 189.6 137.7" id="40,32"></path><path d="M 133.8 112.5 L 133.8 152.9 L 161.7 152.9" id="49,31"></path><path d="M 161.7 175.6 L 161.7 168.0 L 189.6 168.0" id="39,30"></path><path d="M 161.7 175.6 L 161.7 183.1 L 189.6 183.1" id="39,29"></path><path d="M 161.7 228.5 L 161.7 198.3 L 189.6 198.3" id="38,28"></path><path d="M 161.7 228.5 L 161.7 213.4 L 189.6 213.4" id="38,27"></path><path d="M 161.7 228.5 L 161.7 228.5 L 189.6 228.5" id="38,26"></path><path d="M 161.7 228.5 L 161.7 243.6 L 189.6 243.6" id="38,25"></path><path d="M 161.7 228.5 L 161.7 258.8 L 189.6 258.8" id="38,24"></path><path d="M 133.8 226.0 L 133.8 273.9 L 161.7 273.9" id="48,23"></path><path d="M 133.8 296.6 L 133.8 289.0 L 161.7 289.0" id="47,22"></path><path d="M 133.8 296.6 L 133.8 304.2 L 161.7 304.2" id="47,21"></path><path d="M 133.8 326.9 L 133.8 319.3 L 161.7 319.3" id="46,20"></path><path d="M 133.8 326.9 L 133.8 334.4 L 161.7 334.4" id="46,19"></path><path d="M 105.8 334.4 L 105.8 349.6 L 133.8 349.6" id="51,18"></path><path d="M 105.8 334.4 L 105.8 364.7 L 133.8 364.7" id="51,17"></path><path d="M 133.8 395.0 L 133.8 379.8 L 161.7 379.8" id="45,16"></path><path d="M 133.8 395.0 L 133.8 395.0 L 161.7 395.0" id="45,15"></path><path d="M 133.8 395.0 L 133.8 410.1 L 161.7 410.1" id="45,14"></path><path d="M 133.8 447.9 L 133.8 425.2 L 161.7 425.2" id="44,13"></path><path d="M 133.8 447.9 L 133.8 440.4 L 161.7 440.4" id="44,12"></path><path d="M 133.8 447.9 L 133.8 455.5 L 161.7 455.5" id="44,11"></path><path d="M 133.8 447.9 L 133.8 470.6 L 161.7 470.6" id="44,10"></path><path d="M 133.8 508.4 L 133.8 485.7 L 161.7 485.7" id="43,9"></path><path d="M 133.8 508.4 L 133.8 500.9 L 161.7 500.9" id="43,8"></path><path d="M 133.8 508.4 L 133.8 516.0 L 161.7 516.0" id="43,7"></path><path d="M 133.8 508.4 L 133.8 531.1 L 161.7 531.1" id="43,6"></path><path d="M 133.8 576.5 L 133.8 546.3 L 161.7 546.3" id="42,5"></path><path d="M 133.8 576.5 L 133.8 561.4 L 161.7 561.4" id="42,4"></path><path d="M 133.8 576.5 L 133.8 576.5 L 161.7 576.5" id="42,3"></path><path d="M 133.8 576.5 L 133.8 591.7 L 161.7 591.7" id="42,2"></path><path d="M 133.8 576.5 L 133.8 606.8 L 161.7 606.8" id="42,1"></path><path d="M 105.8 510.0 L 105.8 621.9 L 133.8 621.9" id="50,0"></path></g><g class="toytree-Nodes" style="fill:rgb(40%,76.1%,64.7%);fill-opacity:1.0;stroke:rgb(0%,0%,0%);stroke-opacity:1.0;stroke-width:1"><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-0" transform="translate(133.773,621.921)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-1" transform="translate(161.698,606.790)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-2" transform="translate(161.698,591.659)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-3" transform="translate(161.698,576.529)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-4" transform="translate(161.698,561.398)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-5" transform="translate(161.698,546.267)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-6" transform="translate(161.698,531.136)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-7" transform="translate(161.698,516.005)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-8" transform="translate(161.698,500.874)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-9" transform="translate(161.698,485.743)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-10" transform="translate(161.698,470.612)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-11" transform="translate(161.698,455.482)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-12" transform="translate(161.698,440.351)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-13" transform="translate(161.698,425.220)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-14" transform="translate(161.698,410.089)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-15" transform="translate(161.698,394.958)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-16" transform="translate(161.698,379.827)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-17" transform="translate(133.773,364.696)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-18" transform="translate(133.773,349.565)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-19" transform="translate(161.698,334.435)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-20" transform="translate(161.698,319.304)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-21" transform="translate(161.698,304.173)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-22" transform="translate(161.698,289.042)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-23" transform="translate(161.698,273.911)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-24" transform="translate(189.622,258.780)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-25" transform="translate(189.622,243.649)"><circle r="0.0"></circle></g><g fill="rgb(0%,0%,100%)" fill-opacity="1.0" id="node-26" transform="translate(189.622,228.518)"><circle r="4.850486787204451"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-27" transform="translate(189.622,213.388)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-28" transform="translate(189.622,198.257)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-29" transform="translate(189.622,183.126)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-30" transform="translate(189.622,167.995)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-31" transform="translate(161.698,152.864)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-32" transform="translate(189.622,137.733)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-33" transform="translate(189.622,122.602)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-34" transform="translate(189.622,107.471)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-35" transform="translate(189.622,92.341)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-36" transform="translate(189.622,77.210)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-37" transform="translate(189.622,62.079)"><circle r="0.0"></circle></g><g fill="rgb(0%,0%,100%)" fill-opacity="1.0" id="node-38" transform="translate(161.698,228.518)"><circle r="12.5"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-39" transform="translate(161.698,175.560)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-40" transform="translate(161.698,115.037)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-41" transform="translate(161.698,69.644)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-42" transform="translate(133.773,576.529)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-43" transform="translate(133.773,508.440)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-44" transform="translate(133.773,447.916)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-45" transform="translate(133.773,394.958)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-46" transform="translate(133.773,326.869)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-47" transform="translate(133.773,296.607)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-48" transform="translate(133.773,225.997)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-49" transform="translate(133.773,112.515)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-50" transform="translate(105.849,509.953)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-51" transform="translate(105.849,334.435)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-52" transform="translate(105.849,169.256)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-53" transform="translate(77.924,337.881)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-54" transform="translate(50.000,337.881)"><circle r="0.0"></circle></g></g><g class="toytree-TipLabels" style="fill:rgb(14.9%,14.9%,14.9%);fill-opacity:1.0;font-family:helvetica;font-size:11px;font-weight:normal;stroke:none;white-space:pre"><g transform="translate(133.77,621.92)rotate(0)"><text x="15.00" y="2.81" style="">34</text></g><g transform="translate(161.70,606.79)rotate(0)"><text x="15.00" y="2.81" style="">36</text></g><g transform="translate(161.70,591.66)rotate(0)"><text x="15.00" y="2.81" style="">14</text></g><g transform="translate(161.70,576.53)rotate(0)"><text x="15.00" y="2.81" style="">17</text></g><g transform="translate(161.70,561.40)rotate(0)"><text x="15.00" y="2.81" style="">11</text></g><g transform="translate(161.70,546.27)rotate(0)"><text x="15.00" y="2.81" style="">22</text></g><g transform="translate(161.70,531.14)rotate(0)"><text x="15.00" y="2.81" style="">31</text></g><g transform="translate(161.70,516.01)rotate(0)"><text x="15.00" y="2.81" style="">21</text></g><g transform="translate(161.70,500.87)rotate(0)"><text x="15.00" y="2.81" style="">5</text></g><g transform="translate(161.70,485.74)rotate(0)"><text x="15.00" y="2.81" style="">25</text></g><g transform="translate(161.70,470.61)rotate(0)"><text x="15.00" y="2.81" style="">4</text></g><g transform="translate(161.70,455.48)rotate(0)"><text x="15.00" y="2.81" style="">19</text></g><g transform="translate(161.70,440.35)rotate(0)"><text x="15.00" y="2.81" style="">1</text></g><g transform="translate(161.70,425.22)rotate(0)"><text x="15.00" y="2.81" style="">3</text></g><g transform="translate(161.70,410.09)rotate(0)"><text x="15.00" y="2.81" style="">9</text></g><g transform="translate(161.70,394.96)rotate(0)"><text x="15.00" y="2.81" style="">10</text></g><g transform="translate(161.70,379.83)rotate(0)"><text x="15.00" y="2.81" style="">6</text></g><g transform="translate(133.77,364.70)rotate(0)"><text x="15.00" y="2.81" style="">24</text></g><g transform="translate(133.77,349.57)rotate(0)"><text x="15.00" y="2.81" style="">13</text></g><g transform="translate(161.70,334.43)rotate(0)"><text x="15.00" y="2.81" style="">30</text></g><g transform="translate(161.70,319.30)rotate(0)"><text x="15.00" y="2.81" style="">28</text></g><g transform="translate(161.70,304.17)rotate(0)"><text x="15.00" y="2.81" style="">18</text></g><g transform="translate(161.70,289.04)rotate(0)"><text x="15.00" y="2.81" style="">23</text></g><g transform="translate(161.70,273.91)rotate(0)"><text x="15.00" y="2.81" style="">15</text></g><g transform="translate(189.62,258.78)rotate(0)"><text x="15.00" y="2.81" style="">37</text></g><g transform="translate(189.62,243.65)rotate(0)"><text x="15.00" y="2.81" style="">16</text></g><g transform="translate(189.62,228.52)rotate(0)"><text x="15.00" y="2.81" style="">12</text></g><g transform="translate(189.62,213.39)rotate(0)"><text x="15.00" y="2.81" style="">8</text></g><g transform="translate(189.62,198.26)rotate(0)"><text x="15.00" y="2.81" style="">7</text></g><g transform="translate(189.62,183.13)rotate(0)"><text x="15.00" y="2.81" style="">20</text></g><g transform="translate(189.62,167.99)rotate(0)"><text x="15.00" y="2.81" style="">26</text></g><g transform="translate(161.70,152.86)rotate(0)"><text x="15.00" y="2.81" style="">33</text></g><g transform="translate(189.62,137.73)rotate(0)"><text x="15.00" y="2.81" style="">32</text></g><g transform="translate(189.62,122.60)rotate(0)"><text x="15.00" y="2.81" style="">35</text></g><g transform="translate(189.62,107.47)rotate(0)"><text x="15.00" y="2.81" style="">2</text></g><g transform="translate(189.62,92.34)rotate(0)"><text x="15.00" y="2.81" style="">0</text></g><g transform="translate(189.62,77.21)rotate(0)"><text x="15.00" y="2.81" style="">29</text></g><g transform="translate(189.62,62.08)rotate(0)"><text x="15.00" y="2.81" style="">27</text></g></g></g></g></g></svg><div class="toyplot-behavior"><script>(function()
{
var modules={};
})();</script></div></div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">covariate</span> <span class="o">=</span> <span class="s2">&quot;condition[T.Hpoly.Day3]_node&quot;</span>
<span class="n">tasccoda_res</span><span class="o">.</span><span class="n">draw_tree_effects</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span> <span class="n">covariate</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div class="toyplot" id="t17afcbea51a14d47a61e4e89b3738046" style="text-align:center"><svg class="toyplot-canvas-Canvas" xmlns:toyplot="http://www.sandia.gov/toyplot" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns="http://www.w3.org/2000/svg" width="260.0px" height="684.0px" viewBox="0 0 260.0 684.0" preserveAspectRatio="xMidYMid meet" style="background-color:transparent;border-color:#292724;border-style:none;border-width:1.0;fill:rgb(16.1%,15.3%,14.1%);fill-opacity:1.0;font-family:Helvetica;font-size:12px;opacity:1.0;stroke:rgb(16.1%,15.3%,14.1%);stroke-opacity:1.0;stroke-width:1.0" id="t9f927dc0f8584e93a942642f31cb9c8d"><g class="toyplot-coordinates-Cartesian" id="tddd1b4ede5254cf2923ed94ac8e89764"><clipPath id="te24734814e1b420489ef4c093d566919"><rect x="30.0" y="30.0" width="200.0" height="624.0"></rect></clipPath><g clip-path="url(#te24734814e1b420489ef4c093d566919)"><g class="toytree-mark-Toytree" id="t83d13a7f5c9f451eb530faba909d7faf"><g class="toytree-Edges" style="fill:none;stroke:rgb(14.9%,14.9%,14.9%);stroke-linecap:round;stroke-opacity:1;stroke-width:2"><path d="M 50.0 337.9 L 50.0 337.9 L 77.9 337.9" id="54,53"></path><path d="M 77.9 337.9 L 77.9 169.3 L 105.8 169.3" id="53,52"></path><path d="M 77.9 337.9 L 77.9 334.4 L 105.8 334.4" id="53,51"></path><path d="M 77.9 337.9 L 77.9 510.0 L 105.8 510.0" id="53,50"></path><path d="M 105.8 169.3 L 105.8 112.5 L 133.8 112.5" id="52,49"></path><path d="M 105.8 169.3 L 105.8 226.0 L 133.8 226.0" id="52,48"></path><path d="M 105.8 334.4 L 105.8 296.6 L 133.8 296.6" id="51,47"></path><path d="M 105.8 334.4 L 105.8 326.9 L 133.8 326.9" id="51,46"></path><path d="M 105.8 510.0 L 105.8 395.0 L 133.8 395.0" id="50,45"></path><path d="M 105.8 510.0 L 105.8 447.9 L 133.8 447.9" id="50,44"></path><path d="M 105.8 510.0 L 105.8 508.4 L 133.8 508.4" id="50,43"></path><path d="M 105.8 510.0 L 105.8 576.5 L 133.8 576.5" id="50,42"></path><path d="M 133.8 112.5 L 133.8 69.6 L 161.7 69.6" id="49,41"></path><path d="M 133.8 112.5 L 133.8 115.0 L 161.7 115.0" id="49,40"></path><path d="M 133.8 226.0 L 133.8 175.6 L 161.7 175.6" id="48,39"></path><path d="M 133.8 226.0 L 133.8 228.5 L 161.7 228.5" id="48,38"></path><path d="M 161.7 69.6 L 161.7 62.1 L 189.6 62.1" id="41,37"></path><path d="M 161.7 69.6 L 161.7 77.2 L 189.6 77.2" id="41,36"></path><path d="M 161.7 115.0 L 161.7 92.3 L 189.6 92.3" id="40,35"></path><path d="M 161.7 115.0 L 161.7 107.5 L 189.6 107.5" id="40,34"></path><path d="M 161.7 115.0 L 161.7 122.6 L 189.6 122.6" id="40,33"></path><path d="M 161.7 115.0 L 161.7 137.7 L 189.6 137.7" id="40,32"></path><path d="M 133.8 112.5 L 133.8 152.9 L 161.7 152.9" id="49,31"></path><path d="M 161.7 175.6 L 161.7 168.0 L 189.6 168.0" id="39,30"></path><path d="M 161.7 175.6 L 161.7 183.1 L 189.6 183.1" id="39,29"></path><path d="M 161.7 228.5 L 161.7 198.3 L 189.6 198.3" id="38,28"></path><path d="M 161.7 228.5 L 161.7 213.4 L 189.6 213.4" id="38,27"></path><path d="M 161.7 228.5 L 161.7 228.5 L 189.6 228.5" id="38,26"></path><path d="M 161.7 228.5 L 161.7 243.6 L 189.6 243.6" id="38,25"></path><path d="M 161.7 228.5 L 161.7 258.8 L 189.6 258.8" id="38,24"></path><path d="M 133.8 226.0 L 133.8 273.9 L 161.7 273.9" id="48,23"></path><path d="M 133.8 296.6 L 133.8 289.0 L 161.7 289.0" id="47,22"></path><path d="M 133.8 296.6 L 133.8 304.2 L 161.7 304.2" id="47,21"></path><path d="M 133.8 326.9 L 133.8 319.3 L 161.7 319.3" id="46,20"></path><path d="M 133.8 326.9 L 133.8 334.4 L 161.7 334.4" id="46,19"></path><path d="M 105.8 334.4 L 105.8 349.6 L 133.8 349.6" id="51,18"></path><path d="M 105.8 334.4 L 105.8 364.7 L 133.8 364.7" id="51,17"></path><path d="M 133.8 395.0 L 133.8 379.8 L 161.7 379.8" id="45,16"></path><path d="M 133.8 395.0 L 133.8 395.0 L 161.7 395.0" id="45,15"></path><path d="M 133.8 395.0 L 133.8 410.1 L 161.7 410.1" id="45,14"></path><path d="M 133.8 447.9 L 133.8 425.2 L 161.7 425.2" id="44,13"></path><path d="M 133.8 447.9 L 133.8 440.4 L 161.7 440.4" id="44,12"></path><path d="M 133.8 447.9 L 133.8 455.5 L 161.7 455.5" id="44,11"></path><path d="M 133.8 447.9 L 133.8 470.6 L 161.7 470.6" id="44,10"></path><path d="M 133.8 508.4 L 133.8 485.7 L 161.7 485.7" id="43,9"></path><path d="M 133.8 508.4 L 133.8 500.9 L 161.7 500.9" id="43,8"></path><path d="M 133.8 508.4 L 133.8 516.0 L 161.7 516.0" id="43,7"></path><path d="M 133.8 508.4 L 133.8 531.1 L 161.7 531.1" id="43,6"></path><path d="M 133.8 576.5 L 133.8 546.3 L 161.7 546.3" id="42,5"></path><path d="M 133.8 576.5 L 133.8 561.4 L 161.7 561.4" id="42,4"></path><path d="M 133.8 576.5 L 133.8 576.5 L 161.7 576.5" id="42,3"></path><path d="M 133.8 576.5 L 133.8 591.7 L 161.7 591.7" id="42,2"></path><path d="M 133.8 576.5 L 133.8 606.8 L 161.7 606.8" id="42,1"></path><path d="M 105.8 510.0 L 105.8 621.9 L 133.8 621.9" id="50,0"></path></g><g class="toytree-TipLabels" style="fill:rgb(14.9%,14.9%,14.9%);fill-opacity:1.0;font-family:helvetica;font-size:11px;font-weight:normal;stroke:none;white-space:pre"><g transform="translate(133.77,621.92)rotate(0)"><text x="15.00" y="2.81" style="">34</text></g><g transform="translate(161.70,606.79)rotate(0)"><text x="15.00" y="2.81" style="">36</text></g><g transform="translate(161.70,591.66)rotate(0)"><text x="15.00" y="2.81" style="">14</text></g><g transform="translate(161.70,576.53)rotate(0)"><text x="15.00" y="2.81" style="">17</text></g><g transform="translate(161.70,561.40)rotate(0)"><text x="15.00" y="2.81" style="">11</text></g><g transform="translate(161.70,546.27)rotate(0)"><text x="15.00" y="2.81" style="">22</text></g><g transform="translate(161.70,531.14)rotate(0)"><text x="15.00" y="2.81" style="">31</text></g><g transform="translate(161.70,516.01)rotate(0)"><text x="15.00" y="2.81" style="">21</text></g><g transform="translate(161.70,500.87)rotate(0)"><text x="15.00" y="2.81" style="">5</text></g><g transform="translate(161.70,485.74)rotate(0)"><text x="15.00" y="2.81" style="">25</text></g><g transform="translate(161.70,470.61)rotate(0)"><text x="15.00" y="2.81" style="">4</text></g><g transform="translate(161.70,455.48)rotate(0)"><text x="15.00" y="2.81" style="">19</text></g><g transform="translate(161.70,440.35)rotate(0)"><text x="15.00" y="2.81" style="">1</text></g><g transform="translate(161.70,425.22)rotate(0)"><text x="15.00" y="2.81" style="">3</text></g><g transform="translate(161.70,410.09)rotate(0)"><text x="15.00" y="2.81" style="">9</text></g><g transform="translate(161.70,394.96)rotate(0)"><text x="15.00" y="2.81" style="">10</text></g><g transform="translate(161.70,379.83)rotate(0)"><text x="15.00" y="2.81" style="">6</text></g><g transform="translate(133.77,364.70)rotate(0)"><text x="15.00" y="2.81" style="">24</text></g><g transform="translate(133.77,349.57)rotate(0)"><text x="15.00" y="2.81" style="">13</text></g><g transform="translate(161.70,334.43)rotate(0)"><text x="15.00" y="2.81" style="">30</text></g><g transform="translate(161.70,319.30)rotate(0)"><text x="15.00" y="2.81" style="">28</text></g><g transform="translate(161.70,304.17)rotate(0)"><text x="15.00" y="2.81" style="">18</text></g><g transform="translate(161.70,289.04)rotate(0)"><text x="15.00" y="2.81" style="">23</text></g><g transform="translate(161.70,273.91)rotate(0)"><text x="15.00" y="2.81" style="">15</text></g><g transform="translate(189.62,258.78)rotate(0)"><text x="15.00" y="2.81" style="">37</text></g><g transform="translate(189.62,243.65)rotate(0)"><text x="15.00" y="2.81" style="">16</text></g><g transform="translate(189.62,228.52)rotate(0)"><text x="15.00" y="2.81" style="">12</text></g><g transform="translate(189.62,213.39)rotate(0)"><text x="15.00" y="2.81" style="">8</text></g><g transform="translate(189.62,198.26)rotate(0)"><text x="15.00" y="2.81" style="">7</text></g><g transform="translate(189.62,183.13)rotate(0)"><text x="15.00" y="2.81" style="">20</text></g><g transform="translate(189.62,167.99)rotate(0)"><text x="15.00" y="2.81" style="">26</text></g><g transform="translate(161.70,152.86)rotate(0)"><text x="15.00" y="2.81" style="">33</text></g><g transform="translate(189.62,137.73)rotate(0)"><text x="15.00" y="2.81" style="">32</text></g><g transform="translate(189.62,122.60)rotate(0)"><text x="15.00" y="2.81" style="">35</text></g><g transform="translate(189.62,107.47)rotate(0)"><text x="15.00" y="2.81" style="">2</text></g><g transform="translate(189.62,92.34)rotate(0)"><text x="15.00" y="2.81" style="">0</text></g><g transform="translate(189.62,77.21)rotate(0)"><text x="15.00" y="2.81" style="">29</text></g><g transform="translate(189.62,62.08)rotate(0)"><text x="15.00" y="2.81" style="">27</text></g></g></g></g></g></svg><div class="toyplot-behavior"><script>(function()
{
var modules={};
})();</script></div></div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">covariate</span> <span class="o">=</span> <span class="s2">&quot;condition[T.Hpoly.Day10]_node&quot;</span>
<span class="n">tasccoda_res</span><span class="o">.</span><span class="n">draw_tree_effects</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span> <span class="n">covariate</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div class="toyplot" id="t4b9205ca9088424b9571403b072a0cc7" style="text-align:center"><svg class="toyplot-canvas-Canvas" xmlns:toyplot="http://www.sandia.gov/toyplot" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns="http://www.w3.org/2000/svg" width="260.0px" height="684.0px" viewBox="0 0 260.0 684.0" preserveAspectRatio="xMidYMid meet" style="background-color:transparent;border-color:#292724;border-style:none;border-width:1.0;fill:rgb(16.1%,15.3%,14.1%);fill-opacity:1.0;font-family:Helvetica;font-size:12px;opacity:1.0;stroke:rgb(16.1%,15.3%,14.1%);stroke-opacity:1.0;stroke-width:1.0" id="tfd8f7e584c7f42eeb90e3a3d4300f171"><g class="toyplot-coordinates-Cartesian" id="t4365336d80014d05a9583cbd6db8e1e8"><clipPath id="t25736528b2f6465b95a666c040f26224"><rect x="30.0" y="30.0" width="200.0" height="624.0"></rect></clipPath><g clip-path="url(#t25736528b2f6465b95a666c040f26224)"><g class="toytree-mark-Toytree" id="tc169eded5b54466482cac7ed9032174f"><g class="toytree-Edges" style="fill:none;stroke:rgb(14.9%,14.9%,14.9%);stroke-linecap:round;stroke-opacity:1;stroke-width:2"><path d="M 50.0 337.9 L 50.0 337.9 L 77.9 337.9" id="54,53"></path><path d="M 77.9 337.9 L 77.9 169.3 L 105.8 169.3" id="53,52"></path><path d="M 77.9 337.9 L 77.9 334.4 L 105.8 334.4" id="53,51"></path><path d="M 77.9 337.9 L 77.9 510.0 L 105.8 510.0" id="53,50"></path><path d="M 105.8 169.3 L 105.8 112.5 L 133.8 112.5" id="52,49"></path><path d="M 105.8 169.3 L 105.8 226.0 L 133.8 226.0" id="52,48"></path><path d="M 105.8 334.4 L 105.8 296.6 L 133.8 296.6" id="51,47"></path><path d="M 105.8 334.4 L 105.8 326.9 L 133.8 326.9" id="51,46"></path><path d="M 105.8 510.0 L 105.8 395.0 L 133.8 395.0" id="50,45"></path><path d="M 105.8 510.0 L 105.8 447.9 L 133.8 447.9" id="50,44"></path><path d="M 105.8 510.0 L 105.8 508.4 L 133.8 508.4" id="50,43"></path><path d="M 105.8 510.0 L 105.8 576.5 L 133.8 576.5" id="50,42"></path><path d="M 133.8 112.5 L 133.8 69.6 L 161.7 69.6" id="49,41"></path><path d="M 133.8 112.5 L 133.8 115.0 L 161.7 115.0" id="49,40"></path><path d="M 133.8 226.0 L 133.8 175.6 L 161.7 175.6" id="48,39"></path><path d="M 133.8 226.0 L 133.8 228.5 L 161.7 228.5" id="48,38"></path><path d="M 161.7 69.6 L 161.7 62.1 L 189.6 62.1" id="41,37"></path><path d="M 161.7 69.6 L 161.7 77.2 L 189.6 77.2" id="41,36"></path><path d="M 161.7 115.0 L 161.7 92.3 L 189.6 92.3" id="40,35"></path><path d="M 161.7 115.0 L 161.7 107.5 L 189.6 107.5" id="40,34"></path><path d="M 161.7 115.0 L 161.7 122.6 L 189.6 122.6" id="40,33"></path><path d="M 161.7 115.0 L 161.7 137.7 L 189.6 137.7" id="40,32"></path><path d="M 133.8 112.5 L 133.8 152.9 L 161.7 152.9" id="49,31"></path><path d="M 161.7 175.6 L 161.7 168.0 L 189.6 168.0" id="39,30"></path><path d="M 161.7 175.6 L 161.7 183.1 L 189.6 183.1" id="39,29"></path><path d="M 161.7 228.5 L 161.7 198.3 L 189.6 198.3" id="38,28"></path><path d="M 161.7 228.5 L 161.7 213.4 L 189.6 213.4" id="38,27"></path><path d="M 161.7 228.5 L 161.7 228.5 L 189.6 228.5" id="38,26"></path><path d="M 161.7 228.5 L 161.7 243.6 L 189.6 243.6" id="38,25"></path><path d="M 161.7 228.5 L 161.7 258.8 L 189.6 258.8" id="38,24"></path><path d="M 133.8 226.0 L 133.8 273.9 L 161.7 273.9" id="48,23"></path><path d="M 133.8 296.6 L 133.8 289.0 L 161.7 289.0" id="47,22"></path><path d="M 133.8 296.6 L 133.8 304.2 L 161.7 304.2" id="47,21"></path><path d="M 133.8 326.9 L 133.8 319.3 L 161.7 319.3" id="46,20"></path><path d="M 133.8 326.9 L 133.8 334.4 L 161.7 334.4" id="46,19"></path><path d="M 105.8 334.4 L 105.8 349.6 L 133.8 349.6" id="51,18"></path><path d="M 105.8 334.4 L 105.8 364.7 L 133.8 364.7" id="51,17"></path><path d="M 133.8 395.0 L 133.8 379.8 L 161.7 379.8" id="45,16"></path><path d="M 133.8 395.0 L 133.8 395.0 L 161.7 395.0" id="45,15"></path><path d="M 133.8 395.0 L 133.8 410.1 L 161.7 410.1" id="45,14"></path><path d="M 133.8 447.9 L 133.8 425.2 L 161.7 425.2" id="44,13"></path><path d="M 133.8 447.9 L 133.8 440.4 L 161.7 440.4" id="44,12"></path><path d="M 133.8 447.9 L 133.8 455.5 L 161.7 455.5" id="44,11"></path><path d="M 133.8 447.9 L 133.8 470.6 L 161.7 470.6" id="44,10"></path><path d="M 133.8 508.4 L 133.8 485.7 L 161.7 485.7" id="43,9"></path><path d="M 133.8 508.4 L 133.8 500.9 L 161.7 500.9" id="43,8"></path><path d="M 133.8 508.4 L 133.8 516.0 L 161.7 516.0" id="43,7"></path><path d="M 133.8 508.4 L 133.8 531.1 L 161.7 531.1" id="43,6"></path><path d="M 133.8 576.5 L 133.8 546.3 L 161.7 546.3" id="42,5"></path><path d="M 133.8 576.5 L 133.8 561.4 L 161.7 561.4" id="42,4"></path><path d="M 133.8 576.5 L 133.8 576.5 L 161.7 576.5" id="42,3"></path><path d="M 133.8 576.5 L 133.8 591.7 L 161.7 591.7" id="42,2"></path><path d="M 133.8 576.5 L 133.8 606.8 L 161.7 606.8" id="42,1"></path><path d="M 105.8 510.0 L 105.8 621.9 L 133.8 621.9" id="50,0"></path></g><g class="toytree-Nodes" style="fill:rgb(40%,76.1%,64.7%);fill-opacity:1.0;stroke:rgb(0%,0%,0%);stroke-opacity:1.0;stroke-width:1"><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-0" transform="translate(133.773,621.921)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-1" transform="translate(161.698,606.790)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-2" transform="translate(161.698,591.659)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-3" transform="translate(161.698,576.529)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-4" transform="translate(161.698,561.398)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-5" transform="translate(161.698,546.267)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-6" transform="translate(161.698,531.136)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-7" transform="translate(161.698,516.005)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-8" transform="translate(161.698,500.874)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-9" transform="translate(161.698,485.743)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-10" transform="translate(161.698,470.612)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-11" transform="translate(161.698,455.482)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-12" transform="translate(161.698,440.351)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-13" transform="translate(161.698,425.220)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-14" transform="translate(161.698,410.089)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-15" transform="translate(161.698,394.958)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-16" transform="translate(161.698,379.827)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-17" transform="translate(133.773,364.696)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-18" transform="translate(133.773,349.565)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-19" transform="translate(161.698,334.435)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-20" transform="translate(161.698,319.304)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-21" transform="translate(161.698,304.173)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-22" transform="translate(161.698,289.042)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-23" transform="translate(161.698,273.911)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-24" transform="translate(189.622,258.780)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-25" transform="translate(189.622,243.649)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-26" transform="translate(189.622,228.518)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-27" transform="translate(189.622,213.388)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-28" transform="translate(189.622,198.257)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-29" transform="translate(189.622,183.126)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-30" transform="translate(189.622,167.995)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-31" transform="translate(161.698,152.864)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-32" transform="translate(189.622,137.733)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-33" transform="translate(189.622,122.602)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-34" transform="translate(189.622,107.471)"><circle r="0.0"></circle></g><g fill="rgb(100%,0%,0%)" fill-opacity="1.0" id="node-35" transform="translate(189.622,92.341)"><circle r="4.118055555555555"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-36" transform="translate(189.622,77.210)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-37" transform="translate(189.622,62.079)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-38" transform="translate(161.698,228.518)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-39" transform="translate(161.698,175.560)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-40" transform="translate(161.698,115.037)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-41" transform="translate(161.698,69.644)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-42" transform="translate(133.773,576.529)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-43" transform="translate(133.773,508.440)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-44" transform="translate(133.773,447.916)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-45" transform="translate(133.773,394.958)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-46" transform="translate(133.773,326.869)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-47" transform="translate(133.773,296.607)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-48" transform="translate(133.773,225.997)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-49" transform="translate(133.773,112.515)"><circle r="0.0"></circle></g><g fill="rgb(100%,0%,0%)" fill-opacity="1.0" id="node-50" transform="translate(105.849,509.953)"><circle r="12.5"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-51" transform="translate(105.849,334.435)"><circle r="0.0"></circle></g><g fill="rgb(100%,0%,0%)" fill-opacity="1.0" id="node-52" transform="translate(105.849,169.256)"><circle r="4.951388888888889"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-53" transform="translate(77.924,337.881)"><circle r="0.0"></circle></g><g fill="rgb(0%,100%,100%)" fill-opacity="1.0" id="node-54" transform="translate(50.000,337.881)"><circle r="0.0"></circle></g></g><g class="toytree-TipLabels" style="fill:rgb(14.9%,14.9%,14.9%);fill-opacity:1.0;font-family:helvetica;font-size:11px;font-weight:normal;stroke:none;white-space:pre"><g transform="translate(133.77,621.92)rotate(0)"><text x="15.00" y="2.81" style="">34</text></g><g transform="translate(161.70,606.79)rotate(0)"><text x="15.00" y="2.81" style="">36</text></g><g transform="translate(161.70,591.66)rotate(0)"><text x="15.00" y="2.81" style="">14</text></g><g transform="translate(161.70,576.53)rotate(0)"><text x="15.00" y="2.81" style="">17</text></g><g transform="translate(161.70,561.40)rotate(0)"><text x="15.00" y="2.81" style="">11</text></g><g transform="translate(161.70,546.27)rotate(0)"><text x="15.00" y="2.81" style="">22</text></g><g transform="translate(161.70,531.14)rotate(0)"><text x="15.00" y="2.81" style="">31</text></g><g transform="translate(161.70,516.01)rotate(0)"><text x="15.00" y="2.81" style="">21</text></g><g transform="translate(161.70,500.87)rotate(0)"><text x="15.00" y="2.81" style="">5</text></g><g transform="translate(161.70,485.74)rotate(0)"><text x="15.00" y="2.81" style="">25</text></g><g transform="translate(161.70,470.61)rotate(0)"><text x="15.00" y="2.81" style="">4</text></g><g transform="translate(161.70,455.48)rotate(0)"><text x="15.00" y="2.81" style="">19</text></g><g transform="translate(161.70,440.35)rotate(0)"><text x="15.00" y="2.81" style="">1</text></g><g transform="translate(161.70,425.22)rotate(0)"><text x="15.00" y="2.81" style="">3</text></g><g transform="translate(161.70,410.09)rotate(0)"><text x="15.00" y="2.81" style="">9</text></g><g transform="translate(161.70,394.96)rotate(0)"><text x="15.00" y="2.81" style="">10</text></g><g transform="translate(161.70,379.83)rotate(0)"><text x="15.00" y="2.81" style="">6</text></g><g transform="translate(133.77,364.70)rotate(0)"><text x="15.00" y="2.81" style="">24</text></g><g transform="translate(133.77,349.57)rotate(0)"><text x="15.00" y="2.81" style="">13</text></g><g transform="translate(161.70,334.43)rotate(0)"><text x="15.00" y="2.81" style="">30</text></g><g transform="translate(161.70,319.30)rotate(0)"><text x="15.00" y="2.81" style="">28</text></g><g transform="translate(161.70,304.17)rotate(0)"><text x="15.00" y="2.81" style="">18</text></g><g transform="translate(161.70,289.04)rotate(0)"><text x="15.00" y="2.81" style="">23</text></g><g transform="translate(161.70,273.91)rotate(0)"><text x="15.00" y="2.81" style="">15</text></g><g transform="translate(189.62,258.78)rotate(0)"><text x="15.00" y="2.81" style="">37</text></g><g transform="translate(189.62,243.65)rotate(0)"><text x="15.00" y="2.81" style="">16</text></g><g transform="translate(189.62,228.52)rotate(0)"><text x="15.00" y="2.81" style="">12</text></g><g transform="translate(189.62,213.39)rotate(0)"><text x="15.00" y="2.81" style="">8</text></g><g transform="translate(189.62,198.26)rotate(0)"><text x="15.00" y="2.81" style="">7</text></g><g transform="translate(189.62,183.13)rotate(0)"><text x="15.00" y="2.81" style="">20</text></g><g transform="translate(189.62,167.99)rotate(0)"><text x="15.00" y="2.81" style="">26</text></g><g transform="translate(161.70,152.86)rotate(0)"><text x="15.00" y="2.81" style="">33</text></g><g transform="translate(189.62,137.73)rotate(0)"><text x="15.00" y="2.81" style="">32</text></g><g transform="translate(189.62,122.60)rotate(0)"><text x="15.00" y="2.81" style="">35</text></g><g transform="translate(189.62,107.47)rotate(0)"><text x="15.00" y="2.81" style="">2</text></g><g transform="translate(189.62,92.34)rotate(0)"><text x="15.00" y="2.81" style="">0</text></g><g transform="translate(189.62,77.21)rotate(0)"><text x="15.00" y="2.81" style="">29</text></g><g transform="translate(189.62,62.08)rotate(0)"><text x="15.00" y="2.81" style="">27</text></g></g></g></g></g></svg><div class="toyplot-behavior"><script>(function()
{
var modules={};
})();</script></div></div></div></div>
</div>
<p>Alternatively, effects on internal nodes can also be translated through the tree onto the cell type level, allowing for a calculation of log-fold changes like in scCODA.
To visualize the log-fold changes of the cell types, we do the same plots as for scCODA, inspired by “High-resolution single-cell atlas reveals diversity and plasticity of tissue-resident neutrophils in non-small cell lung cancer”<span id="id17">[<a class="reference internal" href="#id318" title="Stefan Salcher, Gregor Sturm, Lena Horvath, Gerold Untergasser, Georgios Fotakis, Elisa Panizzolo, Agnieszka Martowicz, Georg Pall, Gabriele Gamerith, Martina Sykora, Florian Augustin, Katja Schmitz, Francesca Finotello, Dietmar Rieder, Sieghart Sopper, Dominik Wolf, Andreas Pircher, and Zlatko Trajanoski. High-resolution single-cell atlas reveals diversity and plasticity of tissue-resident neutrophils in non-small cell lung cancer. bioRxiv, 2022. URL: https://www.biorxiv.org/content/early/2022/05/10/2022.05.09.491204, arXiv:https://www.biorxiv.org/content/early/2022/05/10/2022.05.09.491204.full.pdf, doi:10.1101/2022.05.09.491204.">Salcher <em>et al.</em>, 2022</a>]</span>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">salmonella_effect</span> <span class="o">=</span> <span class="n">tasccoda_res</span><span class="o">.</span><span class="n">effect_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s2">&quot;condition[T.Salmonella]&quot;</span><span class="p">]</span>
<span class="n">hpoly_effect_3</span> <span class="o">=</span> <span class="n">tasccoda_res</span><span class="o">.</span><span class="n">effect_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s2">&quot;condition[T.Hpoly.Day3]&quot;</span><span class="p">]</span>
<span class="n">hpoly_effect_10</span> <span class="o">=</span> <span class="n">tasccoda_res</span><span class="o">.</span><span class="n">effect_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s2">&quot;condition[T.Hpoly.Day10]&quot;</span><span class="p">]</span>

<span class="n">credible_effects_salmonella</span> <span class="o">=</span> <span class="n">salmonella_effect</span><span class="p">[</span><span class="s2">&quot;Effect&quot;</span><span class="p">]</span> <span class="o">!=</span> <span class="mi">0</span>
<span class="n">credible_effects_hpoly_day3</span> <span class="o">=</span> <span class="n">hpoly_effect_3</span><span class="p">[</span><span class="s2">&quot;Effect&quot;</span><span class="p">]</span> <span class="o">!=</span> <span class="mi">0</span>
<span class="n">credible_effects_hpoly_day10</span> <span class="o">=</span> <span class="n">hpoly_effect_10</span><span class="p">[</span><span class="s2">&quot;Effect&quot;</span><span class="p">]</span> <span class="o">!=</span> <span class="mi">0</span>
<span class="p">(</span>
    <span class="n">alt</span><span class="o">.</span><span class="n">Chart</span><span class="p">(</span>
        <span class="n">salmonella_effect</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">credible_effects_salmonella</span><span class="p">]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(),</span>
        <span class="n">title</span><span class="o">=</span><span class="s2">&quot;Salmonella&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="o">.</span><span class="n">mark_bar</span><span class="p">()</span>
    <span class="o">.</span><span class="n">encode</span><span class="p">(</span>
        <span class="n">x</span><span class="o">=</span><span class="n">alt</span><span class="o">.</span><span class="n">X</span><span class="p">(</span><span class="s2">&quot;Cell Type&quot;</span><span class="p">,</span> <span class="n">sort</span><span class="o">=</span><span class="s2">&quot;y&quot;</span><span class="p">),</span>
        <span class="n">y</span><span class="o">=</span><span class="s2">&quot;log2-fold change&quot;</span><span class="p">,</span>
        <span class="n">color</span><span class="o">=</span><span class="n">alt</span><span class="o">.</span><span class="n">Color</span><span class="p">(</span><span class="s2">&quot;Cell Type&quot;</span><span class="p">),</span>
    <span class="p">)</span>
    <span class="o">|</span> <span class="n">alt</span><span class="o">.</span><span class="n">Chart</span><span class="p">(</span>
        <span class="n">hpoly_effect_3</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">credible_effects_hpoly_day3</span><span class="p">]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(),</span>
        <span class="n">title</span><span class="o">=</span><span class="s2">&quot;Heligmosomoides polygyrus day 3&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="o">.</span><span class="n">mark_bar</span><span class="p">()</span>
    <span class="o">.</span><span class="n">encode</span><span class="p">(</span>
        <span class="n">x</span><span class="o">=</span><span class="n">alt</span><span class="o">.</span><span class="n">X</span><span class="p">(</span><span class="s2">&quot;Cell Type&quot;</span><span class="p">,</span> <span class="n">sort</span><span class="o">=</span><span class="s2">&quot;y&quot;</span><span class="p">),</span>
        <span class="n">y</span><span class="o">=</span><span class="s2">&quot;log2-fold change&quot;</span><span class="p">,</span>
        <span class="n">color</span><span class="o">=</span><span class="n">alt</span><span class="o">.</span><span class="n">Color</span><span class="p">(</span><span class="s2">&quot;Cell Type&quot;</span><span class="p">),</span>
    <span class="p">)</span>
    <span class="o">|</span> <span class="n">alt</span><span class="o">.</span><span class="n">Chart</span><span class="p">(</span>
        <span class="n">hpoly_effect_10</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">credible_effects_hpoly_day10</span><span class="p">]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(),</span>
        <span class="n">title</span><span class="o">=</span><span class="s2">&quot;Heligmosomoides polygyrus day 10&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="o">.</span><span class="n">mark_bar</span><span class="p">()</span>
    <span class="o">.</span><span class="n">encode</span><span class="p">(</span>
        <span class="n">x</span><span class="o">=</span><span class="n">alt</span><span class="o">.</span><span class="n">X</span><span class="p">(</span><span class="s2">&quot;Cell Type&quot;</span><span class="p">,</span> <span class="n">sort</span><span class="o">=</span><span class="s2">&quot;y&quot;</span><span class="p">),</span>
        <span class="n">y</span><span class="o">=</span><span class="s2">&quot;log2-fold change&quot;</span><span class="p">,</span>
        <span class="n">color</span><span class="o">=</span><span class="n">alt</span><span class="o">.</span><span class="n">Color</span><span class="p">(</span><span class="s2">&quot;Cell Type&quot;</span><span class="p">),</span>
    <span class="p">)</span>
<span class="p">)</span><span class="o">.</span><span class="n">resolve_scale</span><span class="p">(</span><span class="n">y</span><span class="o">=</span><span class="s2">&quot;shared&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;shared&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
<div id="altair-viz-d7a834d6380a430699a35f3ecec5f5c5"></div>
<script type="text/javascript">
  var VEGA_DEBUG = (typeof VEGA_DEBUG == "undefined") ? {} : VEGA_DEBUG;
  (function(spec, embedOpt){
    let outputDiv = document.currentScript.previousElementSibling;
    if (outputDiv.id !== "altair-viz-d7a834d6380a430699a35f3ecec5f5c5") {
      outputDiv = document.getElementById("altair-viz-d7a834d6380a430699a35f3ecec5f5c5");
    }
    const paths = {
      "vega": "https://cdn.jsdelivr.net/npm//vega@5?noext",
      "vega-lib": "https://cdn.jsdelivr.net/npm//vega-lib?noext",
      "vega-lite": "https://cdn.jsdelivr.net/npm//vega-lite@4.17.0?noext",
      "vega-embed": "https://cdn.jsdelivr.net/npm//vega-embed@6?noext",
    };

    function maybeLoadScript(lib, version) {
      var key = `${lib.replace("-", "")}_version`;
      return (VEGA_DEBUG[key] == version) ?
        Promise.resolve(paths[lib]) :
        new Promise(function(resolve, reject) {
          var s = document.createElement('script');
          document.getElementsByTagName("head")[0].appendChild(s);
          s.async = true;
          s.onload = () => {
            VEGA_DEBUG[key] = version;
            return resolve(paths[lib]);
          };
          s.onerror = () => reject(`Error loading script: ${paths[lib]}`);
          s.src = paths[lib];
        });
    }

    function showError(err) {
      outputDiv.innerHTML = `<div class="error" style="color:red;">${err}</div>`;
      throw err;
    }

    function displayChart(vegaEmbed) {
      vegaEmbed(outputDiv, spec, embedOpt)
        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));
    }

    if(typeof define === "function" && define.amd) {
      requirejs.config({paths});
      require(["vega-embed"], displayChart, err => showError(`Error loading script: ${err.message}`));
    } else {
      maybeLoadScript("vega", "5")
        .then(() => maybeLoadScript("vega-lite", "4.17.0"))
        .then(() => maybeLoadScript("vega-embed", "6"))
        .catch(showError)
        .then(() => displayChart(vegaEmbed));
    }
  })({"config": {"view": {"continuousWidth": 400, "continuousHeight": 300}}, "hconcat": [{"data": {"name": "data-03cebe56cdf0e4611e5d78e513039df3"}, "mark": "bar", "encoding": {"color": {"field": "Cell Type", "type": "nominal"}, "x": {"field": "Cell Type", "sort": "y", "type": "nominal"}, "y": {"field": "log2-fold change", "type": "quantitative"}}, "title": "Salmonella"}, {"data": {"name": "data-d751713988987e9331980363e24189ce"}, "mark": "bar", "encoding": {"color": {"field": "Cell Type", "type": "nominal"}, "x": {"field": "Cell Type", "sort": "y", "type": "nominal"}, "y": {"field": "log2-fold change", "type": "quantitative"}}, "title": "Heligmosomoides polygyrus day 3"}, {"data": {"name": "data-fae1dc1aa23d27fdd458c1e563e8db5a"}, "mark": "bar", "encoding": {"color": {"field": "Cell Type", "type": "nominal"}, "x": {"field": "Cell Type", "sort": "y", "type": "nominal"}, "y": {"field": "log2-fold change", "type": "quantitative"}}, "title": "Heligmosomoides polygyrus day 10"}], "resolve": {"scale": {"color": "shared", "y": "shared"}}, "$schema": "https://vega.github.io/schema/vega-lite/v4.17.0.json", "datasets": {"data-03cebe56cdf0e4611e5d78e513039df3": [{"Cell Type": "37", "Effect": 0.719, "Median": 1.03, "HDI 3%": 0.134, "HDI 97%": 1.759, "SD": 0.421, "Expected Sample": 38.46481675008553, "log2-fold change": 0.8193670133528106}, {"Cell Type": "16", "Effect": 0.719, "Median": 0.968, "HDI 3%": 0.171, "HDI 97%": 1.636, "SD": 0.381, "Expected Sample": 76.99526857059027, "log2-fold change": 0.8193670133528109}, {"Cell Type": "12", "Effect": 0.888, "Median": 1.305, "HDI 3%": 0.288, "HDI 97%": 2.261, "SD": 0.513, "Expected Sample": 83.99388283767398, "log2-fold change": 1.0631824752630457}, {"Cell Type": "8", "Effect": 0.719, "Median": 0.921, "HDI 3%": 0.112, "HDI 97%": 1.587, "SD": 0.386, "Expected Sample": 15.576197890608823, "log2-fold change": 0.8193670133528109}, {"Cell Type": "7", "Effect": 0.719, "Median": 0.871, "HDI 3%": 0.042, "HDI 97%": 1.485, "SD": 0.383, "Expected Sample": 42.595302312088485, "log2-fold change": 0.8193670133528109}], "data-d751713988987e9331980363e24189ce": [], "data-fae1dc1aa23d27fdd458c1e563e8db5a": [{"Cell Type": "36", "Effect": -1.44, "Median": -1.44, "HDI 3%": -2.108, "HDI 97%": -0.764, "SD": 0.585, "Expected Sample": 3.4903511856491107, "log2-fold change": -1.0514356845253645}, {"Cell Type": "14", "Effect": -1.44, "Median": -1.471, "HDI 3%": -2.036, "HDI 97%": -0.929, "SD": 0.293, "Expected Sample": 3.6074543500691676, "log2-fold change": -1.051435684525365}, {"Cell Type": "17", "Effect": -1.44, "Median": -1.543, "HDI 3%": -2.185, "HDI 97%": -0.972, "SD": 0.322, "Expected Sample": 35.62338689461014, "log2-fold change": -1.0514356845253645}, {"Cell Type": "11", "Effect": -1.44, "Median": -1.517, "HDI 3%": -2.115, "HDI 97%": -0.974, "SD": 0.306, "Expected Sample": 22.488461544133138, "log2-fold change": -1.0514356845253647}, {"Cell Type": "22", "Effect": -1.44, "Median": -1.489, "HDI 3%": -2.05, "HDI 97%": -0.94, "SD": 0.296, "Expected Sample": 11.473067789514575, "log2-fold change": -1.051435684525365}, {"Cell Type": "31", "Effect": -1.44, "Median": -1.424, "HDI 3%": -1.987, "HDI 97%": -0.825, "SD": 0.342, "Expected Sample": 3.1835672901513647, "log2-fold change": -1.051435684525365}, {"Cell Type": "21", "Effect": -1.44, "Median": -1.453, "HDI 3%": -1.988, "HDI 97%": -0.896, "SD": 0.288, "Expected Sample": 8.457066443630879, "log2-fold change": -1.0514356845253643}, {"Cell Type": "5", "Effect": -1.44, "Median": -1.473, "HDI 3%": -2.035, "HDI 97%": -0.941, "SD": 0.29, "Expected Sample": 11.404435485586061, "log2-fold change": -1.051435684525365}, {"Cell Type": "25", "Effect": -1.44, "Median": -1.504, "HDI 3%": -2.083, "HDI 97%": -0.934, "SD": 0.307, "Expected Sample": 29.341484192563275, "log2-fold change": -1.0514356845253647}, {"Cell Type": "4", "Effect": -1.44, "Median": -1.442, "HDI 3%": -1.987, "HDI 97%": -0.913, "SD": 0.286, "Expected Sample": 14.139925014985804, "log2-fold change": -1.051435684525365}, {"Cell Type": "19", "Effect": -1.44, "Median": -1.418, "HDI 3%": -1.979, "HDI 97%": -0.862, "SD": 0.303, "Expected Sample": 9.079347759803387, "log2-fold change": -1.0514356845253647}, {"Cell Type": "1", "Effect": -1.44, "Median": -1.473, "HDI 3%": -2.048, "HDI 97%": -0.937, "SD": 0.298, "Expected Sample": 25.203991060962778, "log2-fold change": -1.0514356845253645}, {"Cell Type": "3", "Effect": -1.44, "Median": -1.44, "HDI 3%": -1.969, "HDI 97%": -0.888, "SD": 0.289, "Expected Sample": 16.102947945147665, "log2-fold change": -1.0514356845253645}, {"Cell Type": "9", "Effect": -1.44, "Median": -1.394, "HDI 3%": -2.046, "HDI 97%": -0.766, "SD": 0.413, "Expected Sample": 6.626006980724197, "log2-fold change": -1.0514356845253645}, {"Cell Type": "10", "Effect": -1.44, "Median": -1.443, "HDI 3%": -1.959, "HDI 97%": -0.89, "SD": 0.287, "Expected Sample": 20.6146576135338, "log2-fold change": -1.0514356845253647}, {"Cell Type": "6", "Effect": -1.44, "Median": -1.461, "HDI 3%": -2.002, "HDI 97%": -0.907, "SD": 0.294, "Expected Sample": 31.249468167960977, "log2-fold change": -1.0514356845253645}, {"Cell Type": "34", "Effect": -1.44, "Median": -1.4, "HDI 3%": -1.975, "HDI 97%": -0.814, "SD": 0.368, "Expected Sample": 3.1899407961131283, "log2-fold change": -1.0514356845253645}, {"Cell Type": "37", "Effect": -0.353, "Median": -0.623, "HDI 3%": -1.17, "HDI 97%": -0.04, "SD": 0.305, "Expected Sample": 31.18703168891626, "log2-fold change": 0.5167738249209385}, {"Cell Type": "16", "Effect": -0.353, "Median": -0.628, "HDI 3%": -1.171, "HDI 97%": -0.048, "SD": 0.301, "Expected Sample": 62.42727988044494, "log2-fold change": 0.5167738249209383}, {"Cell Type": "12", "Effect": -0.353, "Median": -0.626, "HDI 3%": -1.18, "HDI 97%": -0.079, "SD": 0.3, "Expected Sample": 57.51250241154267, "log2-fold change": 0.5167738249209385}, {"Cell Type": "8", "Effect": -0.353, "Median": -0.6, "HDI 3%": -1.131, "HDI 97%": -0.005, "SD": 0.305, "Expected Sample": 12.629083361125536, "log2-fold change": 0.5167738249209389}, {"Cell Type": "7", "Effect": -0.353, "Median": -0.603, "HDI 3%": -1.125, "HDI 97%": -0.005, "SD": 0.301, "Expected Sample": 34.53600342456115, "log2-fold change": 0.5167738249209383}, {"Cell Type": "20", "Effect": -0.353, "Median": -0.541, "HDI 3%": -1.11, "HDI 97%": 0.055, "SD": 0.322, "Expected Sample": 16.84413217728428, "log2-fold change": 0.5167738249209385}, {"Cell Type": "26", "Effect": -0.353, "Median": -0.506, "HDI 3%": -1.084, "HDI 97%": 0.077, "SD": 0.32, "Expected Sample": 24.581702304430486, "log2-fold change": 0.5167738249209383}, {"Cell Type": "15", "Effect": -0.353, "Median": -0.509, "HDI 3%": -1.049, "HDI 97%": 0.044, "SD": 0.308, "Expected Sample": 42.7770762476857, "log2-fold change": 0.5167738249209385}, {"Cell Type": "32", "Effect": -0.353, "Median": -0.368, "HDI 3%": -1.07, "HDI 97%": 0.287, "SD": 0.384, "Expected Sample": 10.329484280194572, "log2-fold change": 0.5167738249209385}, {"Cell Type": "35", "Effect": -0.353, "Median": -0.39, "HDI 3%": -1.08, "HDI 97%": 0.176, "SD": 0.348, "Expected Sample": 15.44063016273388, "log2-fold change": 0.5167738249209385}, {"Cell Type": "2", "Effect": -0.353, "Median": -0.418, "HDI 3%": -1.05, "HDI 97%": 0.149, "SD": 0.34, "Expected Sample": 10.676044047879687, "log2-fold change": 0.5167738249209383}, {"Cell Type": "0", "Effect": -0.586, "Median": -0.836, "HDI 3%": -2.053, "HDI 97%": 0.017, "SD": 0.591, "Expected Sample": 48.52127479903322, "log2-fold change": 0.1806258803938101}, {"Cell Type": "29", "Effect": -0.353, "Median": -0.403, "HDI 3%": -1.04, "HDI 97%": 0.121, "SD": 0.333, "Expected Sample": 24.07086907789912, "log2-fold change": 0.5167738249209385}, {"Cell Type": "27", "Effect": -0.353, "Median": -0.378, "HDI 3%": -1.007, "HDI 97%": 0.152, "SD": 0.33, "Expected Sample": 32.58989781626423, "log2-fold change": 0.5167738249209385}, {"Cell Type": "33", "Effect": -0.353, "Median": -0.351, "HDI 3%": -1.006, "HDI 97%": 0.173, "SD": 0.334, "Expected Sample": 8.864037471818532, "log2-fold change": 0.516773824920938}]}}, {"mode": "vega-lite"});
</script></div></div>
</div>
<p>Another insightful representation can be gained by plotting the effect sizes for each condition on the UMAP embedding, and comparing it to the cell type assignments:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;salmonella_effect_tasccoda&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span>
    <span class="n">tasccoda_res</span><span class="o">.</span><span class="n">effect_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[(</span><span class="s2">&quot;condition[T.Salmonella]&quot;</span><span class="p">,</span> <span class="n">c</span><span class="p">),</span> <span class="s2">&quot;Effect&quot;</span><span class="p">]</span>
    <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;nsbm_level_1&quot;</span><span class="p">]</span>
<span class="p">]</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;H.poly_3_effect_tasccoda&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span>
    <span class="n">tasccoda_res</span><span class="o">.</span><span class="n">effect_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[(</span><span class="s2">&quot;condition[T.Hpoly.Day3]&quot;</span><span class="p">,</span> <span class="n">c</span><span class="p">),</span> <span class="s2">&quot;Effect&quot;</span><span class="p">]</span>
    <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;nsbm_level_1&quot;</span><span class="p">]</span>
<span class="p">]</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;H.poly_10_effect_tasccoda&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span>
    <span class="n">tasccoda_res</span><span class="o">.</span><span class="n">effect_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[(</span><span class="s2">&quot;condition[T.Hpoly.Day10]&quot;</span><span class="p">,</span> <span class="n">c</span><span class="p">),</span> <span class="s2">&quot;Effect&quot;</span><span class="p">]</span>
    <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;nsbm_level_1&quot;</span><span class="p">]</span>
<span class="p">]</span>

<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span>
    <span class="n">color</span><span class="o">=</span><span class="p">[</span>
        <span class="s2">&quot;salmonella_effect_tasccoda&quot;</span><span class="p">,</span>
        <span class="s2">&quot;H.poly_3_effect_tasccoda&quot;</span><span class="p">,</span>
        <span class="s2">&quot;H.poly_10_effect_tasccoda&quot;</span><span class="p">,</span>
    <span class="p">],</span>
    <span class="n">ncols</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span>
    <span class="n">wspace</span><span class="o">=</span><span class="mf">0.25</span><span class="p">,</span>
    <span class="n">vcenter</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
    <span class="n">vmax</span><span class="o">=</span><span class="mf">1.5</span><span class="p">,</span>
    <span class="n">vmin</span><span class="o">=-</span><span class="mf">1.5</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;cell_label&quot;</span><span class="p">,</span> <span class="s2">&quot;nsbm_level_1&quot;</span><span class="p">],</span> <span class="n">ncols</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">wspace</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/compositional_71_0.png" src="../_images/compositional_71_0.png" />
<img alt="../_images/compositional_71_1.png" src="../_images/compositional_71_1.png" />
</div>
</div>
<p>The results are very similar to scCODA’s findings:</p>
<ul class="simple">
<li><p>For the Salmonella infection, we get an aggregated increase in clusters that approximately represent Enterocytes in the cell type clustering. This increase is even stronger for cluster 12, as indicated by the additional positive effect on the leaf level</p></li>
<li><p>For the Heligmosomoides polygyrus infection, we get no credible changes after 3 days. After 10 days, we pick up decreases in clusters that contain Stem- and transit-amplifying cells, as well as a less strong decrease of Enterocytes and Enterocyte progenitors, which was also picked up by scCODA.</p></li>
</ul>
</section>
<section id="without-labeled-clusters">
<h2><span class="section-number">18.6. </span>Without labeled clusters<a class="headerlink" href="#without-labeled-clusters" title="Permalink to this headline">#</a></h2>
<p>It is not always possible or practical to use precisely labeled clusters such as cell-type definitions, especially when we are interested in studying transitional states between cell type clusters, such as during developmental processes, or when we expect only a subpopulation of a cell type to be affected by the condition of interest. In such cases, determining compositional changes based on known annotations may not be appropriate.</p>
<p>A set of methods exist to detect compositional changes occuring in subpopulations of cells smaller than the cell type clusters, usually defined starting from a k-nearest neighbor (KNN) graph computed from similarities in the same low dimensional space used for clustering.</p>
<ul class="simple">
<li><p>DA-seq computes, for each cell, a score based on the relative prevalence of cells from both biological states in the cell’s neighborhood, using a range of k values<span id="id18">[<a class="reference internal" href="#id319" title="Jun Zhao, Ariel Jaffe, Henry Li, Ofir Lindenbaum, Esen Sefik, Ruaidhrí Jackson, Xiuyuan Cheng, Richard A. Flavell, and Yuval Kluger. Detection of differentially abundant cell subpopulations in scrna-seq data. Proceedings of the National Academy of Sciences, 118(22):e2100293118, 2021. URL: https://www.pnas.org/doi/abs/10.1073/pnas.2100293118, arXiv:https://www.pnas.org/doi/pdf/10.1073/pnas.2100293118, doi:10.1073/pnas.2100293118.">Zhao <em>et al.</em>, 2021</a>]</span>. The scores are used as input for a logistic classifier to predict the biological condition of each cell.</p></li>
<li><p>Milo assigns cells to partially overlapping neighborhoods on the KNN graph, then differential abundance (DA) testing is performed modelling cell counts witgh a generalized linear model (GLM) <span id="id19">[<a class="reference internal" href="#id320" title="Emma Dann, Neil C. Henderson, Sarah A. Teichmann, Michael D. Morgan, and John C. Marioni. Differential abundance testing on single-cell data using k-nearest neighbor graphs. Nature Biotechnology, 40(2):245-253, Feb 2022. URL: https://doi.org/10.1038/s41587-021-01033-z, doi:10.1038/s41587-021-01033-z.">Dann <em>et al.</em>, 2022</a>]</span>.</p></li>
<li><p>MELD calculates a relative likelihood estimate of observing each cell in every condition using graph-based density estimate<span id="id20">[<a class="reference internal" href="#id321" title="Daniel B. Burkhardt, Jay S. Stanley, Alexander Tong, Ana Luisa Perdigoto, Scott A. Gigante, Kevan C. Herold, Guy Wolf, Antonio J. Giraldez, David van Dijk, and Smita Krishnaswamy. Quantifying the effect of experimental perturbations at single-cell resolution. Nature Biotechnology, 39(5):619-629, May 2021. URL: https://doi.org/10.1038/s41587-020-00803-5, doi:10.1038/s41587-020-00803-5.">Burkhardt <em>et al.</em>, 2021</a>]</span>.</p></li>
</ul>
<p>These methods have unique strenghts and weaknesses. Because it relies on logistic classification, DA-seq is designed for pairwise comparisons between two biological conditions, but can’t be applied to test for differences associated with a continuous covariate (such as age or timepoints). DA-seq and Milo use the variance in the abundance statistic between replicate samples of the same condition to estimate the significance of the differential abundance, while MELD doesn’t use this information. While considering consistency across replicates reduces the number of false positives driven by one or a few samples, all KNN-based methods are sensitive to a loss of information if the conditions of interest and confounders, defined by technical or experimental sources of variation, are strongly correlated. The impact of confounders can be mitigated using batch integration methods before KNN graph construction and/or incorporating the confounding covariates in the model for DA testing, as we discuss further in the example below. Another limitation of KNN-based methods to bare in mind is that cells in a neighborhood may not necessarily represent a specific, unique biological subpopulation, because a cellular state may span over multiple neighborhoods. Reducing k for the KNN graph or constructing a graph on cells from a particular lineage of interest can help mitigate this issue and ensure the predicted effects are robust to the choice of parameters and to the data subset used<span id="id21">[<a class="reference internal" href="#id320" title="Emma Dann, Neil C. Henderson, Sarah A. Teichmann, Michael D. Morgan, and John C. Marioni. Differential abundance testing on single-cell data using k-nearest neighbor graphs. Nature Biotechnology, 40(2):245-253, Feb 2022. URL: https://doi.org/10.1038/s41587-021-01033-z, doi:10.1038/s41587-021-01033-z.">Dann <em>et al.</em>, 2022</a>]</span>.</p>
<p>Generally, if large differences are apparent in large clusters by visualization or the imbalances between cell types are of interest, direct analysis with cell-type aware methods, such as scCODA, might be more suitable. KNN-based methods are more powerful when we are interested in differences in cell abundances that might appear in transitional states between cell types or in a specific subset of cells of a given cell type.</p>
<p>We will now apply Milo to the Haber dataset to try to find over- or underrepresented neighborhoods of cells upon infection.</p>
<p>Milo is available as <a class="reference external" href="https://github.com/MarioniLab/miloR">miloR</a> for R users and <a class="reference external" href="https://github.com/emdann/milopy">milopy</a> for Python users in the scverse ecosystem. In the following demonstration, we will use milopy which is easiest to use with our AnnData object due to its scverse compatibility. Be aware that milopy in its current state also requires a working <a class="reference external" href="https://bioconductor.org/packages/release/bioc/html/edgeR.html">edgeR</a> installation.</p>
<p>First, we import <code class="docutils literal notranslate"><span class="pre">milopy</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">milopy</span>
<span class="kn">import</span> <span class="nn">milopy.core</span> <span class="k">as</span> <span class="nn">milo</span>
</pre></div>
</div>
</div>
</div>
<p>To perform DA analysis with Milo, we need to construct a KNN graph that is representative of the biological similarities between cells, as we do when performing clustering or UMAP visualization of a single-cell dataset. This means (A) building a common low-dimensional space for all samples and (B) minimizing cell-cell similarities driven by technical factors (i.e. batch effects).</p>
<p>We first use the standard scanpy workflow for dimensionality reduction to qualitatively assess whether we see a batch effect in this dataset.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&quot;haber_count.h5ad&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># use logcounts to calculate PCA and neighbors</span>
<span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;counts&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;logcounts&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;counts&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">adata</span><span class="o">.</span><span class="n">X</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;logcounts&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">highly_variable_genes</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span> <span class="n">n_top_genes</span><span class="o">=</span><span class="mi">3000</span><span class="p">,</span> <span class="n">subset</span><span class="o">=</span><span class="kc">False</span>
<span class="p">)</span>  <span class="c1"># 3k genes as used by authors for clustering</span>

<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">n_pcs</span><span class="o">=</span><span class="mi">30</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;condition&quot;</span><span class="p">,</span> <span class="s2">&quot;batch&quot;</span><span class="p">,</span> <span class="s2">&quot;cell_label&quot;</span><span class="p">],</span> <span class="n">ncols</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">wspace</span><span class="o">=</span><span class="mf">0.25</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/compositional_80_0.png" src="../_images/compositional_80_0.png" />
</div>
</div>
<p>While cell type clusters are broadly captured, we can see residual separation between batches, also for replicates of the same treatment. If we define neighbourhoods on this KNN graph we might have a large fraction of neighbourhoods containing cells from just one or a few batches. This could introduce false negatives, if the variance in number of cells between replicates is too low (e.g. 0 cells for all replicates) or too high (e.g. all zero cells except for one replicate with a large number of cells), but also false positives, especially when, like in this case, the number of replicates per condition is low.</p>
<p>To minimize these errors, we apply the scVI method to learn a batch-corrected latent space, as introduced in the <span class="xref myst">integration chapter</span>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">scvi</span>

<span class="n">adata_scvi</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[:,</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;highly_variable&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">scvi</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">SCVI</span><span class="o">.</span><span class="n">setup_anndata</span><span class="p">(</span><span class="n">adata_scvi</span><span class="p">,</span> <span class="n">layer</span><span class="o">=</span><span class="s2">&quot;counts&quot;</span><span class="p">,</span> <span class="n">batch_key</span><span class="o">=</span><span class="s2">&quot;batch&quot;</span><span class="p">)</span>
<span class="n">model_scvi</span> <span class="o">=</span> <span class="n">scvi</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">SCVI</span><span class="p">(</span><span class="n">adata_scvi</span><span class="p">)</span>
<span class="n">max_epochs_scvi</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">min</span><span class="p">([</span><span class="nb">round</span><span class="p">((</span><span class="mi">20000</span> <span class="o">/</span> <span class="n">adata</span><span class="o">.</span><span class="n">n_obs</span><span class="p">)</span> <span class="o">*</span> <span class="mi">400</span><span class="p">),</span> <span class="mi">400</span><span class="p">])</span>
<span class="n">model_scvi</span><span class="o">.</span><span class="n">train</span><span class="p">(</span><span class="n">max_epochs</span><span class="o">=</span><span class="n">max_epochs_scvi</span><span class="p">)</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;X_scVI&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">model_scvi</span><span class="o">.</span><span class="n">get_latent_representation</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>INFO:pytorch_lightning.utilities.seed:Global seed set to 0
INFO:pytorch_lightning.utilities.rank_zero:GPU available: False, used: False
INFO:pytorch_lightning.utilities.rank_zero:TPU available: False, using: 0 TPU cores
INFO:pytorch_lightning.utilities.rank_zero:IPU available: False, using: 0 IPUs
INFO:pytorch_lightning.utilities.rank_zero:HPU available: False, using: 0 HPUs
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 400/400: 100%|██████████| 400/400 [39:30&lt;00:00,  5.93s/it, loss=538, v_num=1]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">use_rep</span><span class="o">=</span><span class="s2">&quot;X_scVI&quot;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;condition&quot;</span><span class="p">,</span> <span class="s2">&quot;batch&quot;</span><span class="p">,</span> <span class="s2">&quot;cell_label&quot;</span><span class="p">],</span> <span class="n">ncols</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">wspace</span><span class="o">=</span><span class="mf">0.25</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/compositional_84_0.png" src="../_images/compositional_84_0.png" />
</div>
</div>
<p>Here we can see much better mixing between batches and cell labels form much more uniform clusters.</p>
<div class="admonition-will-batch-correction-remove-biological-differences-between-conditions admonition">
<p class="admonition-title">Will batch correction remove biological differences between conditions?</p>
<p>This really boils down to good experimental design. In an ideal set-up replicates from the same condition will be processed in different batches. This allows to estimate technical differences more accurately and possibly also incorporate the batch as a confounder in the linear model for differential abundance analysis (see below), to further minimize false positives. When, like in this example, batches are confounded with the biological condition of interest, we have to recognize that while we might be minimizing false positives, the rate of false negatives might also increase. The analyst can decide which type of error is more detrimental depending on the dataset and purpose of the differential abundance analysis.</p>
</div>
<section id="define-neighbourhoods">
<h3><span class="section-number">18.6.1. </span>Define neighbourhoods<a class="headerlink" href="#define-neighbourhoods" title="Permalink to this headline">#</a></h3>
<p>Milo is a KNN-based model, where cell abundance is quantified on neighbourhoods of cells. In Milo, a neighbourhood is defined as the group of cells connected by an edge to the same cell (<em>index cell</em>) in an undirected KNN graph. While we could in principle have one neighbourhood for each cell in the graph, this would be inefficient and significantly increase the multiple testing burden. Therefore Milo samples a refined set of cells as index cells for neighbourhoods, starting from a random sample of a fraction of cells. The initial proportion can be specified using the <code class="docutils literal notranslate"><span class="pre">prop</span></code> argument in the <code class="docutils literal notranslate"><span class="pre">milo.make_nhoods</span></code> function. As by default, we recommend using <code class="docutils literal notranslate"><span class="pre">prop=0.1</span></code> (10% of cells) and to reduce to 5% or 2% to increase scalability on large datasets (&gt; 100k cells).</p>
<p>If no <code class="docutils literal notranslate"><span class="pre">neighbors_key</span></code> parameter is specified, Milo uses the neighbours from <code class="docutils literal notranslate"><span class="pre">.obsp</span></code>. Therefore, ensure that <code class="docutils literal notranslate"><span class="pre">sc.pp.neighbors</span></code> was run on the correct representation, i.e. an integrated latent space if batch correction was required.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">milo</span><span class="o">.</span><span class="n">make_nhoods</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">prop</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Now the binary assignment of cells to neighbourhood is stored in <code class="docutils literal notranslate"><span class="pre">adata.obsm['nhoods']</span></code>. Here we can see that, as expected, the number of neighbourhoods should be less or equal to the number of cells in the graph times the <code class="docutils literal notranslate"><span class="pre">prop</span></code> parameter. In this case, less or equal than 984 neighbourhoods.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;nhoods&quot;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&lt;9842x851 sparse matrix of type &#39;&lt;class &#39;numpy.float32&#39;&gt;&#39;
	with 22705 stored elements in Compressed Sparse Row format&gt;
</pre></div>
</div>
</div>
</div>
<p>At this point we need to check the median number of cells in each neighbourhood, to make sure the neighbourhoods contain enough cells to detect differences between samples</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">nhood_size</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;nhoods&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">nhood_size</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;# cells in neighbourhood&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;# neighbouthoods&quot;</span><span class="p">);</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/compositional_92_0.png" src="../_images/compositional_92_0.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">np</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">nhood_size</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>26.0
</pre></div>
</div>
</div>
</div>
<p>We expect the minimum number of cells to be equal to the k parameter used during graph construction (k=10 in this case). To increase the statistical power for DA testing, we need a sufficient number of cells from all samples in the majority of the neighbourhoods. We can use the following rule of thumb: to have a median of 3 cells from each sample in a neighbourhood, the number of cells in a neighbourhood should be at least 3 times the number of samples. In this case, we have data from 10 samples. If we want to have on average 3 cells from each sample in a neighbourhood, the minimum number of cells should be 30.</p>
<p>Based on the plot above, we have a large number of neighbourhoods with less than 30 cells, which could lead to an underpowered test. To solve this, we just need to recompute the KNN graph using <code class="docutils literal notranslate"><span class="pre">n_neighbors=30</span></code>. To distinguish this KNN graph used for neighbourhood-level DA analysis from the graph used for UMAP building, we will store this as a distinct graph in <code class="docutils literal notranslate"><span class="pre">adata.obsp</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">use_rep</span><span class="o">=</span><span class="s2">&quot;X_scVI&quot;</span><span class="p">,</span> <span class="n">key_added</span><span class="o">=</span><span class="s2">&quot;milo&quot;</span><span class="p">)</span>
<span class="n">milo</span><span class="o">.</span><span class="n">make_nhoods</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">neighbors_key</span><span class="o">=</span><span class="s2">&quot;milo&quot;</span><span class="p">,</span> <span class="n">prop</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Let’s check that the distribution of neighbourhood sizes has shifted.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">nhood_size</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;nhoods&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">nhood_size</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;# cells in neighbourhood&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;# neighbouthoods&quot;</span><span class="p">);</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/compositional_97_0.png" src="../_images/compositional_97_0.png" />
</div>
</div>
</section>
<section id="count-cells-in-neighbourhoods">
<h3><span class="section-number">18.6.2. </span>Count cells in neighbourhoods<a class="headerlink" href="#count-cells-in-neighbourhoods" title="Permalink to this headline">#</a></h3>
<p>In the next step, Milo counts cells belonging to each of the samples (here identified by the <code class="docutils literal notranslate"><span class="pre">batch</span></code> column in <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code>).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">milo</span><span class="o">.</span><span class="n">count_nhoods</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">sample_col</span><span class="o">=</span><span class="s2">&quot;batch&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>This stores a neighbourhood-level AnnData object, where <code class="docutils literal notranslate"><span class="pre">nhood_adata.X</span></code> stores the number of cells from each sample in each neighbourhood</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s2">&quot;nhood_adata&quot;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 815 × 10
    obs: &#39;index_cell&#39;, &#39;kth_distance&#39;
    uns: &#39;sample_col&#39;
</pre></div>
</div>
</div>
</div>
<p>We can verify that the average number of cells per sample times the number of samples roughly corresponds to the number of cells in a neighbourhood.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mean_n_cells</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s2">&quot;nhood_adata&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">nhood_size</span><span class="p">,</span> <span class="n">mean_n_cells</span><span class="p">,</span> <span class="s2">&quot;.&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;# cells in nhood&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Mean # cells per sample in nhood&quot;</span><span class="p">);</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/compositional_104_0.png" src="../_images/compositional_104_0.png" />
</div>
</div>
<p>### Run differential abundance test on neighbourhoods</p>
<p>Milo uses edgeR’s QLF test to test if there are statistically significant differences between the number of cells from a condition of interest in each neighborhood.</p>
<p>Here we are interested in detecting in which neighbourhoods there is a significant increase or decrease of cells in response to infection. Since the <code class="docutils literal notranslate"><span class="pre">condition</span></code> covariate stores many different types of infection, we need to specify which conditions we need to contrast in our differential abundance test (following a convention used in R, by default the last level of the covariate against the rest will be used, in this case Salmonella vs rest). To specify the comparison, we use <a class="reference external" href="https://genomicsclass.github.io/book/pages/expressing_design_formula.html">the syntax used for GLMs in R</a>.</p>
<p>Let’s first test for differences associated with Salmonella infection.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">milo</span><span class="o">.</span><span class="n">DA_nhoods</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span> <span class="n">design</span><span class="o">=</span><span class="s2">&quot;~condition&quot;</span><span class="p">,</span> <span class="n">model_contrasts</span><span class="o">=</span><span class="s2">&quot;conditionSalmonella-conditionControl&quot;</span>
<span class="p">)</span>
<span class="n">milo_results_salmonella</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s2">&quot;nhood_adata&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">milo_results_salmonella</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>index_cell</th>
      <th>kth_distance</th>
      <th>logFC</th>
      <th>logCPM</th>
      <th>F</th>
      <th>PValue</th>
      <th>FDR</th>
      <th>SpatialFDR</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>B1_AAACGCACGAGGAC_Control_Stem</td>
      <td>1.352524</td>
      <td>-0.113817</td>
      <td>10.780782</td>
      <td>0.020652</td>
      <td>0.885754</td>
      <td>0.993142</td>
      <td>0.992048</td>
    </tr>
    <tr>
      <th>1</th>
      <td>B1_AAATACTGGCAAGG_Control_Stem</td>
      <td>1.228353</td>
      <td>-2.021994</td>
      <td>10.877369</td>
      <td>5.014459</td>
      <td>0.025312</td>
      <td>0.189257</td>
      <td>0.171766</td>
    </tr>
    <tr>
      <th>2</th>
      <td>B1_AACAATACTAGCCA_Control_TA</td>
      <td>1.248063</td>
      <td>-1.128037</td>
      <td>10.717958</td>
      <td>1.664835</td>
      <td>0.197190</td>
      <td>0.576656</td>
      <td>0.561321</td>
    </tr>
    <tr>
      <th>3</th>
      <td>B1_AAGCCTGACCCACT_Control_TA</td>
      <td>1.266929</td>
      <td>-0.088945</td>
      <td>10.929517</td>
      <td>0.012254</td>
      <td>0.911873</td>
      <td>0.993247</td>
      <td>0.992048</td>
    </tr>
    <tr>
      <th>4</th>
      <td>B1_AATGGCTGAGATGA_Control_TA.Early</td>
      <td>1.347708</td>
      <td>-1.162388</td>
      <td>10.822315</td>
      <td>1.674845</td>
      <td>0.195850</td>
      <td>0.576656</td>
      <td>0.561321</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>810</th>
      <td>B10_TTATGGCTTAACGC_Salmonella_TA.Early</td>
      <td>1.450556</td>
      <td>-0.223147</td>
      <td>10.540459</td>
      <td>0.074514</td>
      <td>0.784919</td>
      <td>0.964585</td>
      <td>0.962305</td>
    </tr>
    <tr>
      <th>811</th>
      <td>B10_TTCGTATGCATTGG_Salmonella_Stem</td>
      <td>1.218076</td>
      <td>0.511247</td>
      <td>10.953254</td>
      <td>0.410773</td>
      <td>0.521695</td>
      <td>0.864189</td>
      <td>0.855597</td>
    </tr>
    <tr>
      <th>812</th>
      <td>B10_TTCTTACTGAGGAC_Salmonella_Enterocyte</td>
      <td>0.993957</td>
      <td>1.669047</td>
      <td>11.062170</td>
      <td>5.164554</td>
      <td>0.023221</td>
      <td>0.183743</td>
      <td>0.165820</td>
    </tr>
    <tr>
      <th>813</th>
      <td>B10_TTTAGCTGGGTCAT_Salmonella_Enterocyte</td>
      <td>1.185833</td>
      <td>2.774240</td>
      <td>10.765315</td>
      <td>13.499471</td>
      <td>0.000249</td>
      <td>0.008106</td>
      <td>0.006883</td>
    </tr>
    <tr>
      <th>814</th>
      <td>B10_TTTCAGTGCGACAT_Salmonella_Stem</td>
      <td>1.188963</td>
      <td>-0.505190</td>
      <td>10.602136</td>
      <td>0.301956</td>
      <td>0.582757</td>
      <td>0.886808</td>
      <td>0.878134</td>
    </tr>
  </tbody>
</table>
<p>815 rows × 8 columns</p>
</div></div></div>
</div>
<p>For each neighbourhood, we calculate a set of statistics. The most important ones to understand are:</p>
<ul class="simple">
<li><p><strong>log-Fold Change (logFC):</strong> this represents the effect size of the difference in cell abundance and corresponds to the coefficient associated with the condition of interest in the GLM. If logFC &gt; 0 the neighbourhood is enriched in cells from the condition of interest, if logFC &lt; 0 the neighbourhood is depleted in cells from the condition of interest.</p></li>
<li><p><strong>Uncorrected p-value (PValue):</strong> this is the p-value for the QLF test before multiple testing correction</p></li>
<li><p><strong>SpatialFDR:</strong> this is the p-value adjusted for multiple testing to limit the false discovery rate. This is calculated adapting the weighted Benjamini-Hochberg (BH) correction introduced by Lun et al <span id="id22">[<a class="reference internal" href="#id328" title="Aaron T. L. Lun, Arianne C. Richard, and John C. Marioni. Testing for differential abundance in mass cytometry data. Nature Methods, 14(7):707–709, July 2017. doi:10.1038/nmeth.4295.">Lun <em>et al.</em>, 2017</a>]</span>, which accounts for the fact that because neighbourhoods are partially overlapping (i.e. one cell can belong to multiple neighbourhoods) the DA tests on different neighbourhoods are not completely independent. In practice, the BH correction is weighted by the reciprocal of the distance to the k-th nearest neighbor to the index cell (stored in <code class="docutils literal notranslate"><span class="pre">kth_distance</span></code>), which is used as a proxy for the amount of overlap with other neighbourhoods. You might notice that the SpatialFDR values are always lower or equal to the FDR values, calculated with a conventional BH correction.</p></li>
</ul>
<p>Before any exploration and interpretation of the results, we can visualize these statistics with a set of diagnostics plots to sanity check our statistical test:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">plot_milo_diagnostics</span><span class="p">(</span><span class="n">adata</span><span class="p">):</span>
    <span class="n">nhood_adata</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s2">&quot;nhood_adata&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">alpha</span> <span class="o">=</span> <span class="mf">0.1</span>  <span class="c1">## significance threshold</span>

    <span class="k">with</span> <span class="n">matplotlib</span><span class="o">.</span><span class="n">rc_context</span><span class="p">({</span><span class="s2">&quot;figure.figsize&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">12</span><span class="p">,</span> <span class="mi">12</span><span class="p">]}):</span>

        <span class="c1">## Check P-value histogram</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">nhood_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;PValue&quot;</span><span class="p">],</span> <span class="n">bins</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Uncorrected P-value&quot;</span><span class="p">)</span>

        <span class="c1">## Visualize extent of multiple-testing correction</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">nhood_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;PValue&quot;</span><span class="p">],</span> <span class="n">nhood_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;SpatialFDR&quot;</span><span class="p">],</span> <span class="n">s</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Uncorrected P-value&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;SpatialFDR&quot;</span><span class="p">)</span>

        <span class="c1">## Visualize volcano plot</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span>
            <span class="n">nhood_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;logFC&quot;</span><span class="p">],</span> <span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">nhood_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;SpatialFDR&quot;</span><span class="p">]),</span> <span class="n">s</span><span class="o">=</span><span class="mi">3</span>
        <span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">axhline</span><span class="p">(</span>
            <span class="n">y</span><span class="o">=-</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">alpha</span><span class="p">),</span>
            <span class="n">color</span><span class="o">=</span><span class="s2">&quot;red&quot;</span><span class="p">,</span>
            <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
            <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">int</span><span class="p">(</span><span class="n">alpha</span><span class="o">*</span><span class="mi">100</span><span class="p">)</span><span class="si">}</span><span class="s2"> % SpatialFDR&quot;</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;log-Fold Change&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;- log10(SpatialFDR)&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>

        <span class="c1">## Visualize MA plot</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">nhood_adata</span><span class="o">.</span><span class="n">obs</span>
        <span class="n">emp_null</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;SpatialFDR&quot;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">alpha</span><span class="p">][</span><span class="s2">&quot;logFC&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
        <span class="n">df</span><span class="p">[</span><span class="s2">&quot;Sig&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;SpatialFDR&quot;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="n">alpha</span>

        <span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">)</span>
        <span class="n">sns</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">df</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="s2">&quot;logCPM&quot;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s2">&quot;logFC&quot;</span><span class="p">,</span> <span class="n">hue</span><span class="o">=</span><span class="s2">&quot;Sig&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">axhline</span><span class="p">(</span><span class="n">y</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;grey&quot;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">axhline</span><span class="p">(</span><span class="n">y</span><span class="o">=</span><span class="n">emp_null</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;purple&quot;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;&lt; </span><span class="si">{</span><span class="nb">int</span><span class="p">(</span><span class="n">alpha</span><span class="o">*</span><span class="mi">100</span><span class="p">)</span><span class="si">}</span><span class="s2"> % SpatialFDR&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Mean log-counts&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;log-Fold Change&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>


<span class="n">plot_milo_diagnostics</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/compositional_110_0.png" src="../_images/compositional_110_0.png" />
</div>
</div>
<ol class="simple">
<li><p>The <strong>P-value histogram</strong> shows the distribution of P-values before multiple testing correction. By definition, we expect the p-values under the null hypothesis (&gt; significance level) to be uniformly distributed, while the peak of p-values close to zero represents the significant results. This gives you an idea of how conservative your test is, and it might help to spot early some pathological cases. For example, if the distribution of P-values looks bimodal, with a second peak close to 1, this might indicate that you have a large number of neighbourhoods with no variance between replicates of one condition (e.g. all replicates from one condition have 0 cells) which might indicate a residual batch effect or that you need to increase the size of neighbourhoods; if the p-value histogram is left-skewed this might indicate a <a class="reference external" href="https://github.com/MarioniLab/miloR/issues/220#issuecomment-1140812805">confounding covariate</a> has not been accounted for in the model. For other pathological cases and possible interpretations see <a class="reference external" href="http://varianceexplained.org/statistics/interpreting-pvalue-histogram/">this blogpost</a>.</p></li>
<li><p>For each neighbourhood we plot the uncorrected P-Value VS the p-value controlling for the Spatial FDR. Here we expect the adjusted p-values to be larger (so points above the diagonal). If the FDR correction is especially severe (i.e. many values close to 1) this might indicate a pathological case. You might be testing on too many neighbourhoods (you can reduce <code class="docutils literal notranslate"><span class="pre">prop</span></code> in <code class="docutils literal notranslate"><span class="pre">milo.make_nhoods</span></code>) or there might be too much overlap between neighbourhoods (you might need to decrease <em>k</em> when constructing the KNN graph)</p></li>
<li><p>The <strong>volcano plot</strong> gives us an idea of how many neighbourhoods show significant DA after multiple testing correction ( - log(SpatialFDR) &gt; 1) and shows how many neighbourhoods are enriched or depleted of cells from the condition of interest.</p></li>
<li><p>The <strong>MA plot</strong> shows the dependency between average number of cells per sample and the log-Fold Change of the test. In a balanced scenario, we expect points to be concentrated around logFC = 0, otherwise the shift might indicate a strong imbalance in average number of cells between samples from different conditions. For more tips on how to interpret the MA plot see <a class="reference external" href="https://github.com/MarioniLab/miloR/issues/208">https://github.com/MarioniLab/miloR/issues/208</a>.</p></li>
</ol>
<p>After sanity check, we can visualize the DA results for each neighbourhood by the position of the index cell on the UMAP embedding, to qualitatively assess which cell types may be most affected by the infection.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">milopy</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">build_nhood_graph</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="k">with</span> <span class="n">matplotlib</span><span class="o">.</span><span class="n">rc_context</span><span class="p">({</span><span class="s2">&quot;figure.figsize&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">10</span><span class="p">,</span> <span class="mi">10</span><span class="p">]}):</span>
    <span class="n">milopy</span><span class="o">.</span><span class="n">plot</span><span class="o">.</span><span class="n">plot_nhood_graph</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">min_size</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">plot_edges</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;cell_label&quot;</span><span class="p">,</span> <span class="n">legend_loc</span><span class="o">=</span><span class="s2">&quot;on data&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/compositional_113_0.png" src="../_images/compositional_113_0.png" />
<img alt="../_images/compositional_113_1.png" src="../_images/compositional_113_1.png" />
</div>
</div>
<p>This shows a set of neighbourhoods enriched upon Salmonella infection corresponding to mature enterocytes, and a depletion in a subset of stem cell neighbourhoods. For interpretation of results, it’s often useful to annotate neighbourhoods by the cell type cluster that they overlap with.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">milopy</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">annotate_nhoods</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">anno_col</span><span class="o">=</span><span class="s2">&quot;cell_label&quot;</span><span class="p">)</span>
<span class="c1"># Define as mixed if fraction of cells in nhood with same label is lower than 0.75</span>
<span class="n">nhood_adata</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s2">&quot;nhood_adata&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">nhood_adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
    <span class="n">nhood_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;nhood_annotation_frac&quot;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mf">0.75</span><span class="p">,</span> <span class="s2">&quot;nhood_annotation&quot;</span>
<span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;Mixed&quot;</span>
<span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s2">&quot;nhood_adata&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">nhood_adata</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">milopy</span><span class="o">.</span><span class="n">plot</span><span class="o">.</span><span class="n">plot_DA_beeswarm</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/compositional_116_0.png" src="../_images/compositional_116_0.png" />
</div>
</div>
<div class="admonition-what-about-the-compositional-effect admonition">
<p class="admonition-title">What about the compositional effect?</p>
<p>Comparing the Milo results to the scCODA results, here we identify a strong enrichment upon <em>Salmonella</em> infection in the Enterocytes, but also a depletion in a subset of Stem cells, similarly to what the original authors reported<span id="id23">[<a class="reference internal" href="#id313" title="Adam L. Haber, Moshe Biton, Noga Rogel, Rebecca H. Herbst, Karthik Shekhar, Christopher Smillie, Grace Burgin, Toni M. Delorey, Michael R. Howitt, Yarden Katz, Itay Tirosh, Semir Beyaz, Danielle Dionne, Mei Zhang, Raktima Raychowdhury, Wendy S. Garrett, Orit Rozenblatt-Rosen, Hai Ning Shi, Omer Yilmaz, Ramnik J. Xavier, and Aviv Regev. A single-cell survey of the small intestinal epithelium. Nature, 551(7680):333-339, Nov 2017. URL: https://doi.org/10.1038/nature24489, doi:10.1038/nature24489.">Haber <em>et al.</em>, 2017</a>]</span>. Although we don’t have a ground truth to verify whether the decrease in abundance of stem cells is real, it’s important to note that the GLM in Milo doesn’t explicitly model the compositional nature of cell abundances in neighborhoods, so in theory the results could be affected by compositional biases. In practice, performing a test on a large number of neighbourhoods alleviates this issue, since the effect in the opposite direction is distributed across thousands of neighborhoods rather than tens of cell types. Additionally, the test used in Milo uses the trimmed mean of M values normalization method (TMM normalization <span id="id24">[<a class="reference internal" href="#id329" title="Mark D. Robinson and Alicia Oshlack. A scaling normalization method for differential expression analysis of RNA-seq data. Genome Biology, 11(3):R25, March 2010. doi:10.1186/gb-2010-11-3-r25.">Robinson and Oshlack, 2010</a>]</span>) to estimate normalization factors robust to compositional differences across samples. In this particular example, residual compositional effects might be explained by (A) the relatively small number of neighborhoods (&lt; 1000), (B) the very large effect size in Enterocyte neighbourhoods or (C) the very small number of replicates per condition.</p>
</div>
<p>Of note, the GLM framework used by Milo allows to test for cell enrichment/depletion also for continuous covariates. We demonstrate this by testing for differential abundance along the <em>Heligmosomoides polygyrus</em> infection time course.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## Turn into continuous variable</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;Hpoly_timecourse&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;condition&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">reorder_categories</span><span class="p">(</span>
    <span class="p">[</span><span class="s2">&quot;Salmonella&quot;</span><span class="p">,</span> <span class="s2">&quot;Control&quot;</span><span class="p">,</span> <span class="s2">&quot;Hpoly.Day3&quot;</span><span class="p">,</span> <span class="s2">&quot;Hpoly.Day10&quot;</span><span class="p">]</span>
<span class="p">)</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;Hpoly_timecourse&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;Hpoly_timecourse&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">codes</span>

<span class="c1">## Here we exclude salmonella samples</span>
<span class="n">test_samples</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">batch</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">condition</span> <span class="o">!=</span> <span class="s2">&quot;Salmonella&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s2">&quot;str&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
<span class="p">)</span>
<span class="n">milo</span><span class="o">.</span><span class="n">DA_nhoods</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">design</span><span class="o">=</span><span class="s2">&quot;~ Hpoly_timecourse&quot;</span><span class="p">,</span> <span class="n">subset_samples</span><span class="o">=</span><span class="n">test_samples</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">plot_milo_diagnostics</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/compositional_120_0.png" src="../_images/compositional_120_0.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="n">matplotlib</span><span class="o">.</span><span class="n">rc_context</span><span class="p">({</span><span class="s2">&quot;figure.figsize&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">10</span><span class="p">,</span> <span class="mi">10</span><span class="p">]}):</span>
    <span class="n">milopy</span><span class="o">.</span><span class="n">plot</span><span class="o">.</span><span class="n">plot_nhood_graph</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">min_size</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">plot_edges</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/compositional_121_0.png" src="../_images/compositional_121_0.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">milopy</span><span class="o">.</span><span class="n">plot</span><span class="o">.</span><span class="n">plot_DA_beeswarm</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/compositional_122_0.png" src="../_images/compositional_122_0.png" />
</div>
</div>
<p>We can verify that the test captures a linear increase in cell numbers across the time course by plotting the number of cells per sample by condition in neighborhoods where significant enrichment or depletion was detected.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">plot_nhood_counts_by_cond</span><span class="p">(</span><span class="n">nhood_adata</span><span class="p">,</span> <span class="n">n_ixs</span><span class="p">):</span>
    <span class="n">pl_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">nhood_adata</span><span class="p">[</span><span class="n">n_ixs</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">A</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">nhood_adata</span><span class="o">.</span><span class="n">var_names</span><span class="p">)</span><span class="o">.</span><span class="n">melt</span><span class="p">(</span>
        <span class="n">var_name</span><span class="o">=</span><span class="s2">&quot;batch&quot;</span><span class="p">,</span> <span class="n">value_name</span><span class="o">=</span><span class="s2">&quot;n_cells&quot;</span>
    <span class="p">)</span>
    <span class="n">pl_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">pl_df</span><span class="p">,</span> <span class="n">nhood_adata</span><span class="o">.</span><span class="n">var</span><span class="p">)</span>
    <span class="n">pl_df</span> <span class="o">=</span> <span class="n">pl_df</span><span class="p">[</span><span class="n">pl_df</span><span class="p">[</span><span class="s2">&quot;Hpoly_timecourse&quot;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">]</span>
    <span class="n">pl_df</span><span class="p">[</span><span class="s2">&quot;log_n_cells&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">pl_df</span><span class="p">[</span><span class="s2">&quot;n_cells&quot;</span><span class="p">])</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">boxplot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">pl_df</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="s2">&quot;Hpoly_timecourse&quot;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s2">&quot;n_cells&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;lightblue&quot;</span><span class="p">)</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">stripplot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">pl_df</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="s2">&quot;Hpoly_timecourse&quot;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s2">&quot;n_cells&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;black&quot;</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;H.Poly time course&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;# cells&quot;</span><span class="p">)</span>


<span class="n">nhood_adata</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s2">&quot;nhood_adata&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

<span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="mi">121</span><span class="p">)</span>
<span class="n">entero_ixs</span> <span class="o">=</span> <span class="n">nhood_adata</span><span class="o">.</span><span class="n">obs_names</span><span class="p">[</span>
    <span class="p">(</span><span class="n">nhood_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;SpatialFDR&quot;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mf">0.1</span><span class="p">)</span>
    <span class="o">&amp;</span> <span class="p">(</span><span class="n">nhood_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;logFC&quot;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">)</span>
    <span class="o">&amp;</span> <span class="p">(</span><span class="n">nhood_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;nhood_annotation&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;Enterocyte&quot;</span><span class="p">)</span>
<span class="p">]</span>
<span class="n">plot_nhood_counts_by_cond</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s2">&quot;nhood_adata&quot;</span><span class="p">][:,</span> <span class="n">test_samples</span><span class="p">],</span> <span class="n">entero_ixs</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Enterocytes&quot;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="mi">122</span><span class="p">)</span>
<span class="n">tuft_ixs</span> <span class="o">=</span> <span class="n">nhood_adata</span><span class="o">.</span><span class="n">obs_names</span><span class="p">[</span>
    <span class="p">(</span><span class="n">nhood_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;SpatialFDR&quot;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mf">0.1</span><span class="p">)</span>
    <span class="o">&amp;</span> <span class="p">(</span><span class="n">nhood_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;logFC&quot;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span>
    <span class="o">&amp;</span> <span class="p">(</span><span class="n">nhood_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;nhood_annotation&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;Tuft&quot;</span><span class="p">)</span>
<span class="p">]</span>
<span class="n">plot_nhood_counts_by_cond</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s2">&quot;nhood_adata&quot;</span><span class="p">][:,</span> <span class="n">test_samples</span><span class="p">],</span> <span class="n">tuft_ixs</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Tuft cells&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/compositional_124_0.png" src="../_images/compositional_124_0.png" />
</div>
</div>
<p>Interestingly the DA test on the neighbourhoods detects an enrichment upon infection in Tuft cells and in a subset of goblet cells. We can characterize the difference between cell type subpopulations enriched upon infection by examining the mean gene expression profiles of cells in neighbourhoods. For example, if we take the neighbourhoods of Goblet cells, we can see that neighbourhoods enriched upon infection display a higher expression of Retnlb, which is a gene implicated in anti-parassitic immunity <span id="id25">[<a class="reference internal" href="#id313" title="Adam L. Haber, Moshe Biton, Noga Rogel, Rebecca H. Herbst, Karthik Shekhar, Christopher Smillie, Grace Burgin, Toni M. Delorey, Michael R. Howitt, Yarden Katz, Itay Tirosh, Semir Beyaz, Danielle Dionne, Mei Zhang, Raktima Raychowdhury, Wendy S. Garrett, Orit Rozenblatt-Rosen, Hai Ning Shi, Omer Yilmaz, Ramnik J. Xavier, and Aviv Regev. A single-cell survey of the small intestinal epithelium. Nature, 551(7680):333-339, Nov 2017. URL: https://doi.org/10.1038/nature24489, doi:10.1038/nature24489.">Haber <em>et al.</em>, 2017</a>]</span>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## Compute average Retnlb expression per neighbourhood</span>
<span class="c1"># (you can add mean expression for all genes using milopy.utils.add_nhood_expression)</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;Retnlb_expression&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">adata</span><span class="p">[:,</span> <span class="s2">&quot;Retnlb&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;logcounts&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span><span class="o">.</span><span class="n">ravel</span><span class="p">()</span>
<span class="p">)</span>
<span class="n">milopy</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">annotate_nhoods_continuous</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="s2">&quot;Retnlb_expression&quot;</span><span class="p">)</span>

<span class="c1">## Subset to Goblet cell neighbourhoods</span>
<span class="n">nhood_df</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s2">&quot;nhood_adata&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">nhood_df</span> <span class="o">=</span> <span class="n">nhood_df</span><span class="p">[</span><span class="n">nhood_df</span><span class="p">[</span><span class="s2">&quot;nhood_annotation&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;Goblet&quot;</span><span class="p">]</span>

<span class="n">sns</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">nhood_df</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="s2">&quot;logFC&quot;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s2">&quot;nhood_Retnlb_expression&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&lt;AxesSubplot:xlabel=&#39;logFC&#39;, ylabel=&#39;nhood_Retnlb_expression&#39;&gt;
</pre></div>
</div>
<img alt="../_images/compositional_126_1.png" src="../_images/compositional_126_1.png" />
</div>
</div>
<blockquote>
<div><p><strong>Accounting for confounding covariates:</strong> several confounding factors might affect cell abundances and proportions other than our condition of interest. For example, different set of samples might have been processed or sequenced in the same batch, or a set of samples could contain cells FAC-sorted using different markers to enrich a subset of populations of interest. As long as these factors are not completely correlated with the condition of interest, we can include these covariates in the model used for differential abundance testing, to estimate differential abundance associated with the condition of interest, while minimizing differences explained by the confounding factors. In Milo, we can express this type of test design using the syntax <code class="docutils literal notranslate"><span class="pre">~</span> <span class="pre">confounder</span> <span class="pre">+</span> <span class="pre">condition</span></code>.</p>
</div></blockquote>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## Make dummy confounder for the sake of this example</span>
<span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">42</span><span class="p">)</span>
<span class="n">nhood_adata</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s2">&quot;nhood_adata&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">conf_dict</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span>
    <span class="nb">zip</span><span class="p">(</span>
        <span class="n">nhood_adata</span><span class="o">.</span><span class="n">var_names</span><span class="p">,</span>
        <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">([</span><span class="s2">&quot;group1&quot;</span><span class="p">,</span> <span class="s2">&quot;group2&quot;</span><span class="p">],</span> <span class="n">nhood_adata</span><span class="o">.</span><span class="n">n_vars</span><span class="p">),</span>
    <span class="p">)</span>
<span class="p">)</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;dummy_confounder&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">conf_dict</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;batch&quot;</span><span class="p">]]</span>

<span class="n">milo</span><span class="o">.</span><span class="n">DA_nhoods</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">design</span><span class="o">=</span><span class="s2">&quot;~ dummy_confounder+condition&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s2">&quot;nhood_adata&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>index_cell</th>
      <th>kth_distance</th>
      <th>SpatialFDR</th>
      <th>Nhood_size</th>
      <th>nhood_annotation</th>
      <th>nhood_annotation_frac</th>
      <th>nhood_Retnlb_expression</th>
      <th>logFC</th>
      <th>logCPM</th>
      <th>F</th>
      <th>PValue</th>
      <th>FDR</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>B1_AAACGCACGAGGAC_Control_Stem</td>
      <td>1.352524</td>
      <td>0.954306</td>
      <td>60.0</td>
      <td>Stem</td>
      <td>0.933333</td>
      <td>0.011552</td>
      <td>0.255160</td>
      <td>10.780495</td>
      <td>0.088015</td>
      <td>0.766731</td>
      <td>0.957615</td>
    </tr>
    <tr>
      <th>1</th>
      <td>B1_AAATACTGGCAAGG_Control_Stem</td>
      <td>1.228353</td>
      <td>0.381188</td>
      <td>64.0</td>
      <td>Stem</td>
      <td>0.953125</td>
      <td>0.010830</td>
      <td>-1.767319</td>
      <td>10.877193</td>
      <td>3.287794</td>
      <td>0.069870</td>
      <td>0.412640</td>
    </tr>
    <tr>
      <th>2</th>
      <td>B1_AACAATACTAGCCA_Control_TA</td>
      <td>1.248063</td>
      <td>0.784777</td>
      <td>58.0</td>
      <td>Mixed</td>
      <td>0.637931</td>
      <td>0.011951</td>
      <td>-0.902394</td>
      <td>10.717705</td>
      <td>0.891019</td>
      <td>0.345257</td>
      <td>0.803956</td>
    </tr>
    <tr>
      <th>3</th>
      <td>B1_AAGCCTGACCCACT_Control_TA</td>
      <td>1.266929</td>
      <td>0.982088</td>
      <td>68.0</td>
      <td>Mixed</td>
      <td>0.676471</td>
      <td>0.010193</td>
      <td>0.079992</td>
      <td>10.929335</td>
      <td>0.008033</td>
      <td>0.928589</td>
      <td>0.981963</td>
    </tr>
    <tr>
      <th>4</th>
      <td>B1_AATGGCTGAGATGA_Control_TA.Early</td>
      <td>1.347708</td>
      <td>0.796504</td>
      <td>63.0</td>
      <td>Mixed</td>
      <td>0.507937</td>
      <td>0.000000</td>
      <td>-0.903908</td>
      <td>10.822015</td>
      <td>0.868622</td>
      <td>0.351392</td>
      <td>0.814884</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>810</th>
      <td>B10_TTATGGCTTAACGC_Salmonella_TA.Early</td>
      <td>1.450556</td>
      <td>0.982999</td>
      <td>47.0</td>
      <td>TA.Early</td>
      <td>0.829787</td>
      <td>0.000000</td>
      <td>0.070200</td>
      <td>10.540225</td>
      <td>0.006067</td>
      <td>0.937917</td>
      <td>0.982523</td>
    </tr>
    <tr>
      <th>811</th>
      <td>B10_TTCGTATGCATTGG_Salmonella_Stem</td>
      <td>1.218076</td>
      <td>0.831410</td>
      <td>71.0</td>
      <td>Stem</td>
      <td>0.873239</td>
      <td>0.029288</td>
      <td>0.723550</td>
      <td>10.952914</td>
      <td>0.668495</td>
      <td>0.413625</td>
      <td>0.848050</td>
    </tr>
    <tr>
      <th>812</th>
      <td>B10_TTCTTACTGAGGAC_Salmonella_Enterocyte</td>
      <td>0.993957</td>
      <td>0.317324</td>
      <td>67.0</td>
      <td>Enterocyte</td>
      <td>0.970149</td>
      <td>0.000000</td>
      <td>1.512371</td>
      <td>11.062816</td>
      <td>3.928519</td>
      <td>0.047541</td>
      <td>0.348469</td>
    </tr>
    <tr>
      <th>813</th>
      <td>B10_TTTAGCTGGGTCAT_Salmonella_Enterocyte</td>
      <td>1.185833</td>
      <td>0.020185</td>
      <td>53.0</td>
      <td>Enterocyte</td>
      <td>0.962264</td>
      <td>0.000000</td>
      <td>2.775218</td>
      <td>10.765690</td>
      <td>12.254156</td>
      <td>0.000469</td>
      <td>0.024479</td>
    </tr>
    <tr>
      <th>814</th>
      <td>B10_TTTCAGTGCGACAT_Salmonella_Stem</td>
      <td>1.188963</td>
      <td>0.895816</td>
      <td>50.0</td>
      <td>Stem</td>
      <td>0.900000</td>
      <td>0.013863</td>
      <td>-0.521905</td>
      <td>10.602009</td>
      <td>0.280400</td>
      <td>0.596468</td>
      <td>0.904999</td>
    </tr>
  </tbody>
</table>
<p>815 rows × 12 columns</p>
</div></div></div>
</div>
</section>
</section>
<section id="key-takeaways">
<h2><span class="section-number">18.7. </span>Key Takeaways<a class="headerlink" href="#key-takeaways" title="Permalink to this headline">#</a></h2>
<ol class="simple">
<li><p>If the primary interest lies in compositional changes among known cell-types or states, use scCODA or tascCODA to statistically evaluate changes in abundance.</p></li>
<li><p>KNN based methods like DA-Seq or MILO should be used if the data does not cluster distinctly, such as during developmental processes, if we are interested in differences in cell abundances that might appear in transitional states between cell types or in a specific subset of cells of a given cell type.</p></li>
</ol>
</section>
<section id="quiz">
<h2><span class="section-number">18.8. </span>Quiz<a class="headerlink" href="#quiz" title="Permalink to this headline">#</a></h2>
<ol class="simple">
<li><p>It is tricky to deduce compositional changes visually. Why?</p></li>
<li><p>Why is it necessary to interpret cell type abundances as proportions instead of absolute counts? What are the dangers of not doing so?</p></li>
<li><p>In which cases should tools that use cluster information, such as cell types be used, and in which cases tools that do not use cluster information?</p></li>
</ol>
</section>
<section id="references">
<h2><span class="section-number">18.9. </span>References<a class="headerlink" href="#references" title="Permalink to this headline">#</a></h2>
<div class="docutils container" id="id26">
<dl class="citation">
<dt class="label" id="id314"><span class="brackets"><a class="fn-backref" href="#id2">compAit82</a></span></dt>
<dd><p>J. Aitchison. The statistical analysis of compositional data. <em>Journal of the Royal Statistical Society: Series B (Methodological)</em>, 44(2):139–160, 1982. URL: <a class="reference external" href="https://rss.onlinelibrary.wiley.com/doi/abs/10.1111/j.2517-6161.1982.tb01195.x">https://rss.onlinelibrary.wiley.com/doi/abs/10.1111/j.2517-6161.1982.tb01195.x</a>, <a class="reference external" href="https://arxiv.org/abs/https://rss.onlinelibrary.wiley.com/doi/pdf/10.1111/j.2517-6161.1982.tb01195.x">arXiv:https://rss.onlinelibrary.wiley.com/doi/pdf/10.1111/j.2517-6161.1982.tb01195.x</a>, <a class="reference external" href="https://doi.org/https://doi.org/10.1111/j.2517-6161.1982.tb01195.x">doi:https://doi.org/10.1111/j.2517-6161.1982.tb01195.x</a>.</p>
</dd>
<dt class="label" id="id323"><span class="brackets"><a class="fn-backref" href="#id4">compBAH19</a></span></dt>
<dd><p>Barak Brill, Amnon Amir, and Ruth Heller. Testing for differential abundance in compositional counts data, with application to microbiome studies. <em>ArXiv</em>, April 2019. URL: <a class="reference external" href="http://arxiv.org/abs/1904.08937">http://arxiv.org/abs/1904.08937</a>, <a class="reference external" href="https://arxiv.org/abs/1904.08937">arXiv:1904.08937</a>.</p>
</dd>
<dt class="label" id="id321"><span class="brackets"><a class="fn-backref" href="#id20">compBST+21</a></span></dt>
<dd><p>Daniel B. Burkhardt, Jay S. Stanley, Alexander Tong, Ana Luisa Perdigoto, Scott A. Gigante, Kevan C. Herold, Guy Wolf, Antonio J. Giraldez, David van Dijk, and Smita Krishnaswamy. Quantifying the effect of experimental perturbations at single-cell resolution. <em>Nature Biotechnology</em>, 39(5):619–629, May 2021. URL: <a class="reference external" href="https://doi.org/10.1038/s41587-020-00803-5">https://doi.org/10.1038/s41587-020-00803-5</a>, <a class="reference external" href="https://doi.org/10.1038/s41587-020-00803-5">doi:10.1038/s41587-020-00803-5</a>.</p>
</dd>
<dt class="label" id="id316"><span class="brackets">compButtnerOMuller+21</span><span class="fn-backref">(<a href="#id8">1</a>,<a href="#id9">2</a>,<a href="#id10">3</a>)</span></dt>
<dd><p>M. Büttner, J. Ostner, C. L. Müller, F. J. Theis, and B. Schubert. Sccoda is a bayesian model for compositional single-cell data analysis. <em>Nature Communications</em>, 12(1):6876, Nov 2021. URL: <a class="reference external" href="https://doi.org/10.1038/s41467-021-27150-6">https://doi.org/10.1038/s41467-021-27150-6</a>, <a class="reference external" href="https://doi.org/10.1038/s41467-021-27150-6">doi:10.1038/s41467-021-27150-6</a>.</p>
</dd>
<dt class="label" id="id315"><span class="brackets"><a class="fn-backref" href="#id3">compCLO+19</a></span></dt>
<dd><p>Yue Cao, Yingxin Lin, John T. Ormerod, Pengyi Yang, Jean Y.H. Yang, and Kitty K. Lo. Scdc: single cell differential composition analysis. <em>BMC Bioinformatics</em>, 20(19):721, Dec 2019. URL: <a class="reference external" href="https://doi.org/10.1186/s12859-019-3211-9">https://doi.org/10.1186/s12859-019-3211-9</a>, <a class="reference external" href="https://doi.org/10.1186/s12859-019-3211-9">doi:10.1186/s12859-019-3211-9</a>.</p>
</dd>
<dt class="label" id="id320"><span class="brackets">compDHT+22</span><span class="fn-backref">(<a href="#id19">1</a>,<a href="#id21">2</a>)</span></dt>
<dd><p>Emma Dann, Neil C. Henderson, Sarah A. Teichmann, Michael D. Morgan, and John C. Marioni. Differential abundance testing on single-cell data using k-nearest neighbor graphs. <em>Nature Biotechnology</em>, 40(2):245–253, Feb 2022. URL: <a class="reference external" href="https://doi.org/10.1038/s41587-021-01033-z">https://doi.org/10.1038/s41587-021-01033-z</a>, <a class="reference external" href="https://doi.org/10.1038/s41587-021-01033-z">doi:10.1038/s41587-021-01033-z</a>.</p>
</dd>
<dt class="label" id="id327"><span class="brackets"><a class="fn-backref" href="#id7">compFRM+14</a></span></dt>
<dd><p>Andrew D Fernandes, Jennifer Ns Reid, Jean M Macklaim, Thomas A McMurrough, David R Edgell, and Gregory B Gloor. Unifying the analysis of high-throughput sequencing datasets: characterizing RNA-seq, 16S rRNA gene sequencing and selective growth experiments by compositional data analysis. <em>Microbiome</em>, 2:15, May 2014. URL: <a class="reference external" href="http://dx.doi.org/10.1186/2049-2618-2-15">http://dx.doi.org/10.1186/2049-2618-2-15</a>, <a class="reference external" href="https://doi.org/10.1186/2049-2618-2-15">doi:10.1186/2049-2618-2-15</a>.</p>
</dd>
<dt class="label" id="id325"><span class="brackets"><a class="fn-backref" href="#id5">compGMPGE17</a></span></dt>
<dd><p>Gregory B Gloor, Jean M Macklaim, Vera Pawlowsky-Glahn, and Juan J Egozcue. Microbiome datasets are compositional: and this is not optional. <em>Front. Microbiol.</em>, 8:2224, November 2017. URL: <a class="reference external" href="http://dx.doi.org/10.3389/fmicb.2017.02224">http://dx.doi.org/10.3389/fmicb.2017.02224</a>, <a class="reference external" href="https://doi.org/10.3389/fmicb.2017.02224">doi:10.3389/fmicb.2017.02224</a>.</p>
</dd>
<dt class="label" id="id313"><span class="brackets">compHBR+17</span><span class="fn-backref">(<a href="#id1">1</a>,<a href="#id12">2</a>,<a href="#id13">3</a>,<a href="#id14">4</a>,<a href="#id23">5</a>,<a href="#id25">6</a>)</span></dt>
<dd><p>Adam L. Haber, Moshe Biton, Noga Rogel, Rebecca H. Herbst, Karthik Shekhar, Christopher Smillie, Grace Burgin, Toni M. Delorey, Michael R. Howitt, Yarden Katz, Itay Tirosh, Semir Beyaz, Danielle Dionne, Mei Zhang, Raktima Raychowdhury, Wendy S. Garrett, Orit Rozenblatt-Rosen, Hai Ning Shi, Omer Yilmaz, Ramnik J. Xavier, and Aviv Regev. A single-cell survey of the small intestinal epithelium. <em>Nature</em>, 551(7680):333–339, Nov 2017. URL: <a class="reference external" href="https://doi.org/10.1038/nature24489">https://doi.org/10.1038/nature24489</a>, <a class="reference external" href="https://doi.org/10.1038/nature24489">doi:10.1038/nature24489</a>.</p>
</dd>
<dt class="label" id="id326"><span class="brackets"><a class="fn-backref" href="#id6">compLP20</a></span></dt>
<dd><p>Huang Lin and Shyamal Das Peddada. Analysis of compositions of microbiomes with bias correction. <em>Nat. Commun.</em>, 11(1):3514, July 2020. URL: <a class="reference external" href="http://dx.doi.org/10.1038/s41467-020-17041-7">http://dx.doi.org/10.1038/s41467-020-17041-7</a>, <a class="reference external" href="https://doi.org/10.1038/s41467-020-17041-7">doi:10.1038/s41467-020-17041-7</a>.</p>
</dd>
<dt class="label" id="id328"><span class="brackets"><a class="fn-backref" href="#id22">compLRM17</a></span></dt>
<dd><p>Aaron T. L. Lun, Arianne C. Richard, and John C. Marioni. Testing for differential abundance in mass cytometry data. <em>Nature Methods</em>, 14(7):707–709, July 2017. <a class="reference external" href="https://doi.org/10.1038/nmeth.4295">doi:10.1038/nmeth.4295</a>.</p>
</dd>
<dt class="label" id="id322"><span class="brackets"><a class="fn-backref" href="#id16">compMGC21</a></span></dt>
<dd><p>Leonardo Morelli, Valentina Giansanti, and Davide Cittaro. Nested stochastic block models applied to the analysis of single cell data. <em>BMC Bioinformatics</em>, 22(1):576, November 2021. URL: <a class="reference external" href="http://dx.doi.org/10.1186/s12859-021-04489-7">http://dx.doi.org/10.1186/s12859-021-04489-7</a>, <a class="reference external" href="https://doi.org/10.1186/s12859-021-04489-7">doi:10.1186/s12859-021-04489-7</a>.</p>
</dd>
<dt class="label" id="id317"><span class="brackets"><a class="fn-backref" href="#id15">compOCM21</a></span></dt>
<dd><p>Johannes Ostner, Salomé Carcy, and Christian L. Müller. Tasccoda: bayesian tree-aggregated analysis of compositional amplicon and single-cell data. <em>Frontiers in Genetics</em>, 2021. URL: <a class="reference external" href="https://www.frontiersin.org/article/10.3389/fgene.2021.766405">https://www.frontiersin.org/article/10.3389/fgene.2021.766405</a>, <a class="reference external" href="https://doi.org/10.3389/fgene.2021.766405">doi:10.3389/fgene.2021.766405</a>.</p>
</dd>
<dt class="label" id="id329"><span class="brackets"><a class="fn-backref" href="#id24">compRO10</a></span></dt>
<dd><p>Mark D. Robinson and Alicia Oshlack. A scaling normalization method for differential expression analysis of RNA-seq data. <em>Genome Biology</em>, 11(3):R25, March 2010. <a class="reference external" href="https://doi.org/10.1186/gb-2010-11-3-r25">doi:10.1186/gb-2010-11-3-r25</a>.</p>
</dd>
<dt class="label" id="id318"><span class="brackets">compSSH+22</span><span class="fn-backref">(<a href="#id11">1</a>,<a href="#id17">2</a>)</span></dt>
<dd><p>Stefan Salcher, Gregor Sturm, Lena Horvath, Gerold Untergasser, Georgios Fotakis, Elisa Panizzolo, Agnieszka Martowicz, Georg Pall, Gabriele Gamerith, Martina Sykora, Florian Augustin, Katja Schmitz, Francesca Finotello, Dietmar Rieder, Sieghart Sopper, Dominik Wolf, Andreas Pircher, and Zlatko Trajanoski. High-resolution single-cell atlas reveals diversity and plasticity of tissue-resident neutrophils in non-small cell lung cancer. <em>bioRxiv</em>, 2022. URL: <a class="reference external" href="https://www.biorxiv.org/content/early/2022/05/10/2022.05.09.491204">https://www.biorxiv.org/content/early/2022/05/10/2022.05.09.491204</a>, <a class="reference external" href="https://arxiv.org/abs/https://www.biorxiv.org/content/early/2022/05/10/2022.05.09.491204.full.pdf">arXiv:https://www.biorxiv.org/content/early/2022/05/10/2022.05.09.491204.full.pdf</a>, <a class="reference external" href="https://doi.org/10.1101/2022.05.09.491204">doi:10.1101/2022.05.09.491204</a>.</p>
</dd>
<dt class="label" id="id319"><span class="brackets"><a class="fn-backref" href="#id18">compZJL+21</a></span></dt>
<dd><p>Jun Zhao, Ariel Jaffe, Henry Li, Ofir Lindenbaum, Esen Sefik, Ruaidhrí Jackson, Xiuyuan Cheng, Richard A. Flavell, and Yuval Kluger. Detection of differentially abundant cell subpopulations in scrna-seq data. <em>Proceedings of the National Academy of Sciences</em>, 118(22):e2100293118, 2021. URL: <a class="reference external" href="https://www.pnas.org/doi/abs/10.1073/pnas.2100293118">https://www.pnas.org/doi/abs/10.1073/pnas.2100293118</a>, <a class="reference external" href="https://arxiv.org/abs/https://www.pnas.org/doi/pdf/10.1073/pnas.2100293118">arXiv:https://www.pnas.org/doi/pdf/10.1073/pnas.2100293118</a>, <a class="reference external" href="https://doi.org/10.1073/pnas.2100293118">doi:10.1073/pnas.2100293118</a>.</p>
</dd>
</dl>
</div>
</section>
<section id="contributors">
<h2><span class="section-number">18.10. </span>Contributors<a class="headerlink" href="#contributors" title="Permalink to this headline">#</a></h2>
<p>We gratefully acknowledge the contributions of:</p>
<section id="authors">
<h3><span class="section-number">18.10.1. </span>Authors<a class="headerlink" href="#authors" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p>Johannes Ostner</p></li>
<li><p>Emma Dann</p></li>
<li><p>Lukas Heumos</p></li>
<li><p>Anastasia Litinetskaya</p></li>
</ul>
</section>
<section id="reviewers">
<h3><span class="section-number">18.10.2. </span>Reviewers<a class="headerlink" href="#reviewers" title="Permalink to this headline">#</a></h3>
</section>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./conditions"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
    <a class='left-prev' id="prev-link" href="differential_gene_expression.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title"><span class="section-number">17. </span>Differential gene expression analysis</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="gsea_pathway.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title"><span class="section-number">19. </span>Gene set enrichment and pathway analysis</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By Lukas Heumos, Anna Schaar<br/>
  
      &copy; Copyright 2022.<br/>
    <div class="extra_footer">
      <div>
  Brought to you by
  <a href="https://www.github.com/theislab">Theislab</a>, with many thanks to the single-cell community as a whole!
</div>

    </div>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>