
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>18. Gene set enrichment and pathway analysis &#8212; Multimodal single-cell analysis</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="../_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" href="../_static/styles/sphinx-book-theme.css?digest=5115cc725059bd94278eecd172e13a965bf8f5a9" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/book.css" />
    <link rel="stylesheet" type="text/css" href="../_static/design-style.b7bb847fb20b106c3d81b95245e65545.min.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?digest=9c920249402e914e316237a7dbc6769907cce411"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js"></script>
    <script async="async" kind="hypothesis" src="https://hypothes.is/embed.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../_static/book.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="19. Perturbation modeling" href="perturbation_modeling.html" />
    <link rel="prev" title="17. Compositional analysis" href="compositional.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="en">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../_static/logo.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">Multimodal single-cell analysis</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Introduction
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../introduction/prior_art.html">
   1. Prior art
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../introduction/scrna_seq.html">
   2. Single-cell RNA sequencing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../introduction/raw_data_processing.html">
   3. Raw data processing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../introduction/analysis_tools.html">
   4. Analysis frameworks and tools
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../introduction/data_infrastructure.html">
   5. Data infrastructure
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../introduction/interoperability.html">
   6. Interoperability
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Preprocessing and visualization
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../preprocessing_visualization/quality_control.html">
   7. Quality Control
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../preprocessing_visualization/normalization.html">
   8. Normalization
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../preprocessing_visualization/feature_selection.html">
   9. Feature selection
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../preprocessing_visualization/dimensionality_reduction.html">
   10. Dimensionality Reduction
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Identifying cellular structure
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../cellular_structure/clustering.html">
   11. Clustering
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cellular_structure/integration.html">
   12. Data integration
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Inferring trajectories
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../trajectories/pseudotemporal.html">
   13. Pseudotemporal ordering
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../trajectories/rna_velocity.html">
   14. RNA velocity
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../trajectories/lineage_tracing.html">
   15. Lineage tracing
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Dealing with conditions
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="differential_gene_expression.html">
   16. Differential gene expression analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="compositional.html">
   17. Compositional analysis
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   18. Gene set enrichment and pathway analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="perturbation_modeling.html">
   19. Perturbation modeling
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Modeling mechanisms
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../mechanisms/grns.html">
   20. Gene regulatory networks
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../mechanisms/cell_cell_communication.html">
   21. Cell-cell communication
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Deconvolution
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../deconvolution/bulk_deconvolution.html">
   22. Bulk deconvolution
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Surface protein
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/quality_control.html">
   23. Quality control
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/normalization.html">
   24. Normalization
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/doublet_detection.html">
   25. Doublet detection
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/dimensionality_reduction.html">
   26. Dimensionality Reduction
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/batch_correction.html">
   27. Batch correction
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/annotation.html">
   28. Annotation
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  AIR repertoire
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../air_repertoire/ir_profiling.html">
   29. Immune Receptor Profiling
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../air_repertoire/clonotype.html">
   30. Clonotype analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../air_repertoire/specificity.html">
   34. Specificity analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../air_repertoire/multimodal_integration.html">
   35. Integrating AIR and transcriptomics
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multimodal integration
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../multimodal_integration/paired_integration.html">
   36. Paired integration
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../multimodal_integration/advanced_integration.html">
   37. Advanced integration
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Reproducibility
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../reproducibility/introduction.html">
   38. Reproducibility
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Outlook
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../outlook.html">
   39. Outlook
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Acknowledgements
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../acknowledgements.html">
   40. Acknowledgements
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Glossary
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../glossary.html">
   41. Glossary
  </a>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
            <div class="navbar_extra_footer">
            Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
            </div>
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right">
<div class="menu-dropdown menu-dropdown-launch-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Launch interactive content">
      <i class="fas fa-rocket"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://mybinder.org/v2/gh/theislab/single-cell-best-practices/master?urlpath=tree/conditions/gsea_pathway.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Launch on Binder"
>
  

<span class="headerbtn__icon-container">
  
    <img src="../_static/images/logo_binder.svg">
  </span>
<span class="headerbtn__text-container">Binder</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

<button onclick="toggleFullScreen()"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="bottom"
title="Fullscreen mode"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>

<div class="menu-dropdown menu-dropdown-repository-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Source repositories">
      <i class="fab fa-github"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://github.com/theislab/single-cell-best-practices"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Source repository"
>
  

<span class="headerbtn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="headerbtn__text-container">repository</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/theislab/single-cell-best-practices/issues/new?title=Issue%20on%20page%20%2Fconditions/gsea_pathway.html&body=Your%20issue%20content%20here."
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Open an issue"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="headerbtn__text-container">open issue</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/theislab/single-cell-best-practices/edit/master/conditions/gsea_pathway.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Edit this page"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="headerbtn__text-container">suggest edit</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

<div class="menu-dropdown menu-dropdown-download-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Download this page">
      <i class="fas fa-download"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="../_sources/conditions/gsea_pathway.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Download source file"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="headerbtn__text-container">.ipynb</span>
</a>

      </li>
      
      <li>
        
<button onclick="printPdf(this)"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="left"
title="Print to PDF"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="headerbtn__text-container">.pdf</span>
</button>

      </li>
      
    </ul>
  </div>
</div>
<label for="__page-toc"
  class="headerbtn headerbtn-page-toc"
  
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-list"></i>
  </span>

</label>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
    <div class="tocsection onthispage pt-5 pb-3">
        <i class="fas fa-list"></i> Contents
    </div>
    <nav id="bd-toc-nav" aria-label="Page">
        <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#motivation">
   18.1. Motivation
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#pathway-and-gene-set-collections">
   18.2. Pathway and gene set collections
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#null-hypotheses-in-gene-set-enrichment-analysis">
   18.3. Null hypotheses in gene set enrichment analysis
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#gene-set-tests-and-pathway-analysis">
   18.4. Gene set tests and pathway analysis
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#gene-set-test-vs-pathway-activity-inference">
     18.4.1. Gene set test vs. pathway activity inference
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#technical-considerations">
   18.5. Technical considerations
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#filtering-out-the-gene-sets-with-low-number-of-genes">
     18.5.1. Filtering out the gene sets with low number of genes
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#data-normalization">
     18.5.2. Data normalization
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#case-study-pathway-enrichment-analysis-and-activity-level-scoring-in-human-pbmc-single-cells">
   18.6. Case study: Pathway enrichment analysis and activity level scoring in human PBMC single cells
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#prepare-and-explore-the-data">
     18.6.1. Prepare and explore the data
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#cluster-level-gene-set-enrichment-analysis-with-fgsea">
     18.6.2. Cluster-level gene set enrichment analysis with fgsea
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#cell-level-pathway-activity-scoring-using-aucell">
     18.6.3. Cell-level pathway activity scoring using AUCell
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#gene-set-enrichment-for-complex-experimental-designs-using-limma-fry-and-pseudo-bulks">
     18.6.4. Gene set enrichment for complex experimental designs using limma-fry and pseudo-bulks
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#create-pseudo-bulk-samples-and-explore-the-data">
       18.6.4.1. Create pseudo-bulk samples and explore the data
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#fry-test-for-stimulated-vs-control">
       18.6.4.2. fry test for Stimulated vs Control
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#fry-test-for-the-comparison-between-two-stimulated-cell-types">
       18.6.4.3. fry test for the comparison between two stimulated cell types
      </a>
      <ul class="nav section-nav flex-column">
       <li class="toc-h5 nav-item toc-entry">
        <a class="reference internal nav-link" href="#on-the-effect-of-filtering-low-expression-genes">
         18.6.4.3.1. On the effect of  filtering low-expression genes
        </a>
       </li>
       <li class="toc-h5 nav-item toc-entry">
        <a class="reference internal nav-link" href="#a-note-on-the-redundancy-between-gene-sets-and-the-performance-of-preranked-and-fry-gene-set-tests">
         18.6.4.3.2. A note on the redundancy between gene sets and the performance of preranked and fry gene set tests
        </a>
       </li>
      </ul>
     </li>
    </ul>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#key-takeaways">
   18.7. Key Takeaways
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#quiz">
   18.8. Quiz
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#session-info">
   18.9. Session info
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#references">
   18.10. References
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#contributors">
   18.11. Contributors
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#authors">
     18.11.1. Authors
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#reviewers">
     18.11.2. Reviewers
    </a>
   </li>
  </ul>
 </li>
</ul>

    </nav>
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Gene set enrichment and pathway analysis</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#motivation">
   18.1. Motivation
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#pathway-and-gene-set-collections">
   18.2. Pathway and gene set collections
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#null-hypotheses-in-gene-set-enrichment-analysis">
   18.3. Null hypotheses in gene set enrichment analysis
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#gene-set-tests-and-pathway-analysis">
   18.4. Gene set tests and pathway analysis
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#gene-set-test-vs-pathway-activity-inference">
     18.4.1. Gene set test vs. pathway activity inference
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#technical-considerations">
   18.5. Technical considerations
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#filtering-out-the-gene-sets-with-low-number-of-genes">
     18.5.1. Filtering out the gene sets with low number of genes
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#data-normalization">
     18.5.2. Data normalization
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#case-study-pathway-enrichment-analysis-and-activity-level-scoring-in-human-pbmc-single-cells">
   18.6. Case study: Pathway enrichment analysis and activity level scoring in human PBMC single cells
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#prepare-and-explore-the-data">
     18.6.1. Prepare and explore the data
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#cluster-level-gene-set-enrichment-analysis-with-fgsea">
     18.6.2. Cluster-level gene set enrichment analysis with fgsea
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#cell-level-pathway-activity-scoring-using-aucell">
     18.6.3. Cell-level pathway activity scoring using AUCell
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#gene-set-enrichment-for-complex-experimental-designs-using-limma-fry-and-pseudo-bulks">
     18.6.4. Gene set enrichment for complex experimental designs using limma-fry and pseudo-bulks
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#create-pseudo-bulk-samples-and-explore-the-data">
       18.6.4.1. Create pseudo-bulk samples and explore the data
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#fry-test-for-stimulated-vs-control">
       18.6.4.2. fry test for Stimulated vs Control
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#fry-test-for-the-comparison-between-two-stimulated-cell-types">
       18.6.4.3. fry test for the comparison between two stimulated cell types
      </a>
      <ul class="nav section-nav flex-column">
       <li class="toc-h5 nav-item toc-entry">
        <a class="reference internal nav-link" href="#on-the-effect-of-filtering-low-expression-genes">
         18.6.4.3.1. On the effect of  filtering low-expression genes
        </a>
       </li>
       <li class="toc-h5 nav-item toc-entry">
        <a class="reference internal nav-link" href="#a-note-on-the-redundancy-between-gene-sets-and-the-performance-of-preranked-and-fry-gene-set-tests">
         18.6.4.3.2. A note on the redundancy between gene sets and the performance of preranked and fry gene set tests
        </a>
       </li>
      </ul>
     </li>
    </ul>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#key-takeaways">
   18.7. Key Takeaways
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#quiz">
   18.8. Quiz
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#session-info">
   18.9. Session info
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#references">
   18.10. References
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#contributors">
   18.11. Contributors
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#authors">
     18.11.1. Authors
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#reviewers">
     18.11.2. Reviewers
    </a>
   </li>
  </ul>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                
  <section class="tex2jax_ignore mathjax_ignore" id="gene-set-enrichment-and-pathway-analysis">
<h1><span class="section-number">18. </span>Gene set enrichment and pathway analysis<a class="headerlink" href="#gene-set-enrichment-and-pathway-analysis" title="Permalink to this headline">#</a></h1>
<section id="motivation">
<h2><span class="section-number">18.1. </span>Motivation<a class="headerlink" href="#motivation" title="Permalink to this headline">#</a></h2>
<p>Single-cell RNA-seq provides unprecedented insights into variations in cell types between conditions, tissue types, species and individuals. Differential gene expression analysis of the single-cell data is almost always followed by <em>gene set enrichment analysis</em>, where the aim is to identify gene programs, such as biological processes, gene ontologies or regulatory pathways that are over-represented in an experimental condition compared to control or other conditions, on the basis of differentially expressed (DE) genes.</p>
<p>To determine the pathways enriched in a cell type-specific manner between two conditions, first a relevant collection of gene set signatures is selected, where each gene set defines a biological processes (e.g. epithelial to mesechymal transition, metabolism etc) or pathway (e.g. MAPK signalling). For each gene set in the collection, DE genes present in the gene set are used to obtain a test statistic that is then used to assess the enrichment of the gene set. Depending on the type of the enrichment test chosen, gene expression measurements may or may not be used for the computation of test statistic.</p>
<p>In this chapter, we first provide an overview of different types of gene set enrichment tests, introduce some commonly used gene signature collections and discuss best practices for pathway enrichment and functional enrichment analysis in general.  We conclude the chapter by demonstrating three analytical approaches for gene set enrichment analysis. Note that we use the terms pathway analysis, pathway enrichment analysis, gene set enrichment analysis and functional analysis interchangeably in this chapter.</p>
</section>
<section id="pathway-and-gene-set-collections">
<h2><span class="section-number">18.2. </span>Pathway and gene set collections<a class="headerlink" href="#pathway-and-gene-set-collections" title="Permalink to this headline">#</a></h2>
<p>Gene sets are a curated list of gene names (or gene ids) that are known to be involved in a biological process through previous studies and/or experiments. The Molecular Signatures Database (MSigDB) <span id="id1">[<a class="reference internal" href="#id324" title="Arthur Liberzon, Aravind Subramanian, Reid Pinchback, Helga Thorvaldsdóttir, Pablo Tamayo, and Jill P Mesirov. Molecular signatures database (msigdb) 3.0. Bioinformatics, 27(12):1739–1740, 2011.">Liberzon <em>et al.</em>, 2011</a>, <a class="reference internal" href="#id323" title="Aravind Subramanian, Pablo Tamayo, Vamsi K Mootha, Sayan Mukherjee, Benjamin L Ebert, Michael A Gillette, Amanda Paulovich, Scott L Pomeroy, Todd R Golub, Eric S Lander, and others. Gene set enrichment analysis: a knowledge-based approach for interpreting genome-wide expression profiles. Proceedings of the National Academy of Sciences, 102(43):15545–15550, 2005.">Subramanian <em>et al.</em>, 2005</a>]</span> is the most comprehensive database consisting of 9 collections of gene sets. Some commonly used collections are C5, which is the gene ontology (GO) collection, C2 collection of curated gene signatures from published studies that are typically context (e.g. tissue, condition) specific, but also include KEGG and REACTOME gene signatures. For cancer studies, the Hallmark collection is commonly used, and for immunologic studies the C7 collection is a common choice. Note that these signatures are mainly derived from Bulk-seq measurements and measure continuous phenotypes. Recently and with the wide-spread availability of scRNA-seq datasets, databases have evolved that provide curated marker lists derived from published single cell studies that define cell types in various tissues and species. These include CellMarker <span id="id2">[<a class="reference internal" href="#id325" title="Xinxin Zhang, Yujia Lan, Jinyuan Xu, Fei Quan, Erjie Zhao, Chunyu Deng, Tao Luo, Liwen Xu, Gaoming Liao, Min Yan, and others. Cellmarker: a manually curated resource of cell markers in human and mouse. Nucleic acids research, 47(D1):D721–D728, 2019.">Zhang <em>et al.</em>, 2019</a>]</span> and PanglaoDB <span id="id3">[<a class="reference internal" href="#id326" title="Oscar Franzén, Li-Ming Gan, and Johan LM Björkegren. Panglaodb: a web server for exploration of mouse and human single-cell rna sequencing data. Database, 2019.">Franzén <em>et al.</em>, 2019</a>]</span>. Curated marker lists are not limited to those made available in databases, and can be curated by oneself.</p>
</section>
<section id="null-hypotheses-in-gene-set-enrichment-analysis">
<h2><span class="section-number">18.3. </span>Null hypotheses in gene set enrichment analysis<a class="headerlink" href="#null-hypotheses-in-gene-set-enrichment-analysis" title="Permalink to this headline">#</a></h2>
<p>Gene set tests can be <em>competitive</em> or <em>self-contained</em> as defined by Goeman and Buhlmann (2007) <span id="id4">[<a class="reference internal" href="#id317" title="Jelle J Goeman and Peter Bühlmann. Analyzing gene expression data in terms of gene sets: methodological issues. Bioinformatics, 23(8):980–987, 2007.">Goeman and Bühlmann, 2007</a>]</span>. Competitive gene set testing tests whether the genes in the set are highly ranked in terms of differential expression relative to the genes not in the set. The sampling unit here is genes, so the test can be done with a single sample (i.e. single-sample GSEA). The test requires genes that are not in the set (i.e background genes). In self-contained gene set testing, the sampling unit is the subject, so multiple samples per group are required, but it is not required to have genes that are not present in the set. A self-contained gene set test tests whether genes in the test set are differentially expressed without regard to any other gene measured in the dataset. These distinctions between the two null hypotheses make differences to the interpretation of gene set enrichment results. Note that in biological data there exist inter-gene correlations, that is the expression of genes in the same pathways are correlated. There are only a few tests that accomodate inter-gene correlations. We will discuss these methods later. Detailed explanations on various gene set tests can be found in <a class="reference external" href="https://bioconductor.org/packages/release/bioc/manuals/limma/man/limma.pdf"><em>limma</em> user manual</a>.</p>
</section>
<section id="gene-set-tests-and-pathway-analysis">
<h2><span class="section-number">18.4. </span>Gene set tests and pathway analysis<a class="headerlink" href="#gene-set-tests-and-pathway-analysis" title="Permalink to this headline">#</a></h2>
<p>In scRNA-seq data analysis, gene set enrichment is generally carried out on clusters of cells or cell types, one-at-a-time. Genes differentially expressed in a cluster or cell type are used to identify over-represented gene sets from the selected collection, using simple hypergeomtric tests or Fisher’s exact test (as in <em>Enrichr</em> {cite} <code class="docutils literal notranslate"><span class="pre">chen2013enrichr</span></code>), for example. Such tests do not require the actual gene expression measurements and read counts to compute an enrichment statistics, as they rely on testing how significant it is that an <span class="math notranslate nohighlight">\(X\)</span> number of genes in a gene set are differentially expressed in the experiment compared to the number of non-DE genes in the set.</p>
<p><em>fgsea</em> <span id="id5">[<a class="reference internal" href="#id313" title="Gennady Korotkevich, Vladimir Sukhov, Nikolay Budin, Boris Shpak, Maxim N Artyomov, and Alexey Sergushichev. Fast gene set enrichment analysis. BioRxiv, pages 060012, 2021.">Korotkevich <em>et al.</em>, 2021</a>]</span> is a more common tool for gene set enrichment test. <em>fgsae</em> is a computationally faster implementation of the well established <em>Gene Set Enrichment Analysis (GSEA)</em> algorithm <span id="id6">[<a class="reference internal" href="#id323" title="Aravind Subramanian, Pablo Tamayo, Vamsi K Mootha, Sayan Mukherjee, Benjamin L Ebert, Michael A Gillette, Amanda Paulovich, Scott L Pomeroy, Todd R Golub, Eric S Lander, and others. Gene set enrichment analysis: a knowledge-based approach for interpreting genome-wide expression profiles. Proceedings of the National Academy of Sciences, 102(43):15545–15550, 2005.">Subramanian <em>et al.</em>, 2005</a>]</span>, which computes enrichment statistics on the basis of some preranked gene-level test statistics. <em>fgsea</em> computes an enrichment score using some signed statistics of the genes in the gene set, such as  the t-statistics, log fold-changes (logFC) or p-values from the differential expression test. An empricial (estimated from the data) null distribution is computed for the enrichment score using some random gene sets of the same size, and a p-value is computed to determine the significance of the enrichment score. The p-values are then adjusted for multiple hypothesis testing. GSVA <span id="id7">[<a class="reference internal" href="#id332" title="Sonja Hänzelmann, Robert Castelo, and Justin Guinney. Gsva: gene set variation analysis for microarray and rna-seq data. BMC bioinformatics, 14(1):1–15, 2013.">Hänzelmann <em>et al.</em>, 2013</a>]</span> is another example of preranked gene set enrichment approaches. We should note that the pre-ranked gene set tests are not specific to single cell datasets and apply to Bulk-seq assays as well.</p>
<p>An alternative approach to test for gene set enrichment in a group of cells, that is clusters or cells of identical types, is to create pseudo-bulk samples from single cells and use gene set enrichment methods developed for Bulk RNA-seq. Several self-contained and competitive gene set enrichment tests, namely <em>fry</em> and <em>camera</em> are implemented in <em>limma</em> {cite} <code class="docutils literal notranslate"><span class="pre">ritchie2015limma</span></code>, which are compatible with the differential gene expression analysis framework through linear models and Empirical Bayes moderation of test statistics <span id="id8">[<a class="reference internal" href="#id334" title="Gordon K Smyth. Limma: linear models for microarray data. In Bioinformatics and computational biology solutions using R and Bioconductor, pages 397–420. Springer, 2005.">Smyth, 2005</a>]</span>. Linear models can accomodate complex experimental designs (e.g. subjects, perturbations, batches, nested contrasts, interactions etc) through the design matrix. In addition, the <code class="docutils literal notranslate"><span class="pre">camera</span></code> and <code class="docutils literal notranslate"><span class="pre">roast</span></code> gene set tests implemented in limma account for inter-gene correlations. Gene set tests in <em>limma</em> can also be applied to (properly transformed and normalised) single cell measurements without pseudo-bulk generation. However, there are currently no benchmarks that had assessed the accuracy of gene set test results when these methods are applied directly to single cells.</p>
<table class="colwidths-auto table">
<thead>
<tr class="row-odd"><th class="text-align:left head"><p>Test</p></th>
<th class="text-align:center head"><p>Bulk or SC</p></th>
<th class="text-align:center head"><p>Type of Null Hypothesis</p></th>
<th class="text-align:left head"><p>Input</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td class="text-align:left"><p>Hypergeometric</p></td>
<td class="text-align:center"><p>both</p></td>
<td class="text-align:center"><p>competitive</p></td>
<td class="text-align:left"><p>gene counts</p></td>
</tr>
<tr class="row-odd"><td class="text-align:left"><p>Fisher’s Exact</p></td>
<td class="text-align:center"><p>both</p></td>
<td class="text-align:center"><p>competitive</p></td>
<td class="text-align:left"><p>gene counts</p></td>
</tr>
<tr class="row-even"><td class="text-align:left"><p>GSEA<span class="math notranslate nohighlight">\(^*\)</span></p></td>
<td class="text-align:center"><p>bulk</p></td>
<td class="text-align:center"><p>competitive</p></td>
<td class="text-align:left"><p>gene ranks</p></td>
</tr>
<tr class="row-odd"><td class="text-align:left"><p>GSVA<span class="math notranslate nohighlight">\(^*\)</span></p></td>
<td class="text-align:center"><p>bulk</p></td>
<td class="text-align:center"><p>competitive</p></td>
<td class="text-align:left"><p>gene ranks</p></td>
</tr>
<tr class="row-even"><td class="text-align:left"><p>fgsea</p></td>
<td class="text-align:center"><p>both</p></td>
<td class="text-align:center"><p>competitive</p></td>
<td class="text-align:left"><p>gene ranks</p></td>
</tr>
<tr class="row-odd"><td class="text-align:left"><p>fry<span class="math notranslate nohighlight">\(^*\)</span></p></td>
<td class="text-align:center"><p>bulk</p></td>
<td class="text-align:center"><p>self-contained</p></td>
<td class="text-align:left"><p>expression matrix</p></td>
</tr>
<tr class="row-even"><td class="text-align:left"><p>camera<span class="math notranslate nohighlight">\(^*\)</span></p></td>
<td class="text-align:center"><p>bulk</p></td>
<td class="text-align:center"><p>competitive</p></td>
<td class="text-align:left"><p>expression matrix</p></td>
</tr>
<tr class="row-odd"><td class="text-align:left"><p>roast<span class="math notranslate nohighlight">\(^*\)</span></p></td>
<td class="text-align:center"><p>bulk</p></td>
<td class="text-align:center"><p>self-contained</p></td>
<td class="text-align:left"><p>expression matrix</p></td>
</tr>
</tbody>
</table>
<p>Table: Gene set tests, type of the applicable assays and Null Hypothesis they test</p>
<p><span class="math notranslate nohighlight">\(^*\)</span> These tests are practically applicable to single cell datasets, although their application to single cell may not be a common practice.</p>
<section id="gene-set-test-vs-pathway-activity-inference">
<h3><span class="section-number">18.4.1. </span>Gene set test vs. pathway activity inference<a class="headerlink" href="#gene-set-test-vs-pathway-activity-inference" title="Permalink to this headline">#</a></h3>
<p>Gene set tests test whether a pathway is enriched, in other words over-represented, in one condition compared to others, say, in healthy donors compared to severe COVID-19 patients in the monocyte population. An alternative approach is to simply score the activity of a pathway or gene signature, in absolute sense, in individual cells, rather than testing for a differential activity between conditions. Some of the widely used tools for inference of gene set activity in general (including pathway activity) in individual cells include <em>VISION</em> <span id="id9">[<a class="reference internal" href="#id312" title="David DeTomaso, Matthew G Jones, Meena Subramaniam, Tal Ashuach, Chun J Ye, and Nir Yosef. Functional interpretation of single cell similarity maps. Nature communications, 10(1):1–11, 2019.">DeTomaso <em>et al.</em>, 2019</a>]</span>, <em>AUCell</em> <span id="id10">[<a class="reference internal" href="#id318" title="Sara Aibar, Carmen Bravo González-Blas, Thomas Moerman, Vân Anh Huynh-Thu, Hana Imrichova, Gert Hulselmans, Florian Rambow, Jean-Christophe Marine, Pierre Geurts, Jan Aerts, and others. Scenic: single-cell regulatory network inference and clustering. Nature methods, 14(11):1083–1086, 2017.">Aibar <em>et al.</em>, 2017</a>]</span>, pathway overdispersion analysis using <em>Pagoda2</em> <span id="id11">[<a class="reference internal" href="#id314" title="Jean Fan, Neeraj Salathia, Rui Liu, Gwendolyn E Kaeser, Yun C Yung, Joseph L Herman, Fiona Kaper, Jian-Bing Fan, Kun Zhang, Jerold Chun, and others. Characterizing transcriptional heterogeneity through pathway and gene set overdispersion analysis. Nature methods, 13(3):241–244, 2016.">Fan <em>et al.</em>, 2016</a>, <a class="reference internal" href="#id329" title="Blue B Lake, Song Chen, Brandon C Sos, Jean Fan, Gwendolyn E Kaeser, Yun C Yung, Thu E Duong, Derek Gao, Jerold Chun, Peter V Kharchenko, and others. Integrative single-cell analysis of transcriptional and epigenetic states in the human adult brain. Nature biotechnology, 36(1):70–80, 2018.">Lake <em>et al.</em>, 2018</a>]</span> and simple combined z-score <span id="id12">[<a class="reference internal" href="#id328" title="Eunjung Lee, Han-Yu Chuang, Jong-Won Kim, Trey Ideker, and Doheon Lee. Inferring pathway activity toward precise disease classification. PLoS computational biology, 4(11):e1000217, 2008.">Lee <em>et al.</em>, 2008</a>]</span>.</p>
<p><em>DoRothEA</em> <span id="id13">[<a class="reference internal" href="#id321" title="Luz Garcia-Alonso, Christian H Holland, Mahmoud M Ibrahim, Denes Turei, and Julio Saez-Rodriguez. Benchmark and integration of resources for the estimation of human transcription factor activities. Genome research, 29(8):1363–1375, 2019.">Garcia-Alonso <em>et al.</em>, 2019</a>]</span> and <em>PROGENy</em> <span id="id14">[<a class="reference internal" href="#id320" title="Michael Schubert, Bertram Klinger, Martina Klünemann, Anja Sieber, Florian Uhlitz, Sascha Sauer, Mathew J Garnett, Nils Blüthgen, and Julio Saez-Rodriguez. Perturbation-response genes reveal signaling footprints in cancer gene expression. Nature communications, 9(1):1–11, 2018.">Schubert <em>et al.</em>, 2018</a>]</span> are among functional analysis tools developed to infer transcription factor (TF) - target activities originally in Bulk RNA data. Holland et al. <span id="id15">[<a class="reference internal" href="#id327" title="Christian H Holland, Jovan Tanevski, Javier Perales-Patón, Jan Gleixner, Manu P Kumar, Elisabetta Mereu, Brian A Joughin, Oliver Stegle, Douglas A Lauffenburger, Holger Heyn, and others. Robustness and applicability of transcription factor and pathway analysis tools on single-cell rna-seq data. Genome biology, 21(1):1–19, 2020.">Holland <em>et al.</em>, 2020</a>]</span> found that Bulk RNA-seq methods <em>DoRothEA</em> and <em>PROGENy</em> have optimal performance in simulated scRNA-seq data, and even partially outperform tools specifically designed for scRNA-seq analysis despite the drop-out events and low library sizes in single cell data. Holland et al. also concluded that pathway and TF activity inference is more sensitive to the choice of gene sets rather than the statistical methods. This observation though can be specific to functional enrichment analyses and be explained by the fact that TF-target relations are context-specific; that is TF-target associations in one cell type may actually differ from another cell type or tissue.</p>
<p>In contrast to Holand et al., Zhang et al. <span id="id16">[<a class="reference internal" href="#id316" title="Yaru Zhang, Yunlong Ma, Yukuan Huang, Yan Zhang, Qi Jiang, Meng Zhou, and Jianzhong Su. Benchmarking algorithms for pathway activity transformation of single-cell rna-seq data. Computational and structural biotechnology journal, 18:2953–2961, 2020.">Zhang <em>et al.</em>, 2020</a>]</span> found that single-cell-based tools, specifically Pagoda2, outperform bulk-base methods from three different aspects of accuracy, stability and scalability. It should be noted that pathway and gene set activity inference tools inherently do not account for batch effects or biological variations other than the biological variation of interest. Therefore, it is up to the data analyst to ensure that the differential gene expression analysis step has worked properly.</p>
<p>Furthermore, while the tools mentioned here score every gene set in individual cells, they are not able to select for the most biologically relevant gene sets among all scored gene sets. scDECAF (<a class="reference external" href="https://github.com/DavisLaboratory/scDECAF">https://github.com/DavisLaboratory/scDECAF</a>) is a gene set activity inference tool that allows data-driven selection of the most informative gene sets, thereby aids in dissecting meaningful cellular heterogeneity.</p>
</section>
</section>
<section id="technical-considerations">
<h2><span class="section-number">18.5. </span>Technical considerations<a class="headerlink" href="#technical-considerations" title="Permalink to this headline">#</a></h2>
<section id="filtering-out-the-gene-sets-with-low-number-of-genes">
<h3><span class="section-number">18.5.1. </span>Filtering out the gene sets with low number of genes<a class="headerlink" href="#filtering-out-the-gene-sets-with-low-number-of-genes" title="Permalink to this headline">#</a></h3>
<p>A common practice is to exclude any gene sets with a few genes overlapping the data or Highly Variable Genes (HVG) in the pre-processing step. Zhang et al. <span id="id17">[<a class="reference internal" href="#id316" title="Yaru Zhang, Yunlong Ma, Yukuan Huang, Yan Zhang, Qi Jiang, Meng Zhou, and Jianzhong Su. Benchmarking algorithms for pathway activity transformation of single-cell rna-seq data. Computational and structural biotechnology journal, 18:2953–2961, 2020.">Zhang <em>et al.</em>, 2020</a>]</span> found that the performance of both single-cell-based and bulk-based methods drops as gene coverage, that is the number of genes in pathways/gene sets, decreases. Holland et al <span id="id18">[<a class="reference internal" href="#id327" title="Christian H Holland, Jovan Tanevski, Javier Perales-Patón, Jan Gleixner, Manu P Kumar, Elisabetta Mereu, Brian A Joughin, Oliver Stegle, Douglas A Lauffenburger, Holger Heyn, and others. Robustness and applicability of transcription factor and pathway analysis tools on single-cell rna-seq data. Genome biology, 21(1):1–19, 2020.">Holland <em>et al.</em>, 2020</a>]</span> also found that gene sets of smaller size adversely impacts the performance of Bulk-seq <em>DoRothEA</em> and <em>PROGENy</em> on single cell data. These report collectively support that filtering gene sets with low gene counts, say less than 10 or 15 genes in the set, is beneficial in pathway analysis. Damian &amp; Gorfine (2004) <span id="id19">[<a class="reference internal" href="#id335" title="Doris Damian and Malka Gorfine. Statistical concerns about the gsea procedure. Nature genetics, 36(7):663–663, 2004.">Damian and Gorfine, 2004</a>]</span> attributed this to the fact gene variances in gene sets with a smaller number of genes are more likely to be large, whereas gene variances in larger gene sets tend to be smaller. This impacts the accuracy of the test statistics computed to test for enrichment. Zhang et al. additionally found that pathway analysis was susceptible to normalization procedures applied to gene expression measurements.</p>
</section>
<section id="data-normalization">
<h3><span class="section-number">18.5.2. </span>Data normalization<a class="headerlink" href="#data-normalization" title="Permalink to this headline">#</a></h3>
<p>Read counts in single cell experiments are typically normalised early on in the pre-processing pipeline to ensure that measuerments are comparable across cells of various library sizes. Zhang et al. <span id="id20">[<a class="reference internal" href="#id316" title="Yaru Zhang, Yunlong Ma, Yukuan Huang, Yan Zhang, Qi Jiang, Meng Zhou, and Jianzhong Su. Benchmarking algorithms for pathway activity transformation of single-cell rna-seq data. Computational and structural biotechnology journal, 18:2953–2961, 2020.">Zhang <em>et al.</em>, 2020</a>]</span> found that normalisation by <em>SCTransform</em> <span id="id21">[<a class="reference internal" href="#id330" title="Christoph Hafemeister and Rahul Satija. Normalization and variance stabilization of single-cell rna-seq data using regularized negative binomial regression. Genome biology, 20(1):1–15, 2019.">Hafemeister and Satija, 2019</a>]</span> and <em>scran</em> <span id="id22">[<a class="reference internal" href="#id331" title="Aaron TL Lun, Davis J McCarthy, and John C Marioni. A step-by-step workflow for low-level analysis of single-cell rna-seq data with bioconductor. F1000Research, 2016.">Lun <em>et al.</em>, 2016</a>]</span> generally improves the performance of both single-cell- and bulk-based pathway scoring tools. They found that the performance of <em>AUCell</em> (a rank-base method) and <em>z-score</em> (transformation to zero mean, unit standard deviation) is particularly affected by normalization with distinct methods.</p>
</section>
</section>
<section id="case-study-pathway-enrichment-analysis-and-activity-level-scoring-in-human-pbmc-single-cells">
<h2><span class="section-number">18.6. </span>Case study: Pathway enrichment analysis and activity level scoring in human PBMC single cells<a class="headerlink" href="#case-study-pathway-enrichment-analysis-and-activity-level-scoring-in-human-pbmc-single-cells" title="Permalink to this headline">#</a></h2>
<section id="prepare-and-explore-the-data">
<h3><span class="section-number">18.6.1. </span>Prepare and explore the data<a class="headerlink" href="#prepare-and-explore-the-data" title="Permalink to this headline">#</a></h3>
<p>We first download the 25K PBMC data and follow the standard <code class="docutils literal notranslate"><span class="pre">scanpy</span></code> workflow for normalisation of read counts and subsetting on the highly variable genes. The dataset contains untreated and IFN-<span class="math notranslate nohighlight">\(\beta\)</span> stimulated human PBMC cells <span id="id23">[<a class="reference internal" href="#id333" title="Hyun Min Kang, Meena Subramaniam, Sasha Targ, Michelle Nguyen, Lenka Maliskova, Elizabeth McCarthy, Eunice Wan, Simon Wong, Lauren Byrnes, Cristina M Lanata, and others. Multiplexed droplet single-cell rna-sequencing using natural genetic variation. Nature biotechnology, 36(1):89–94, 2018.">Kang <em>et al.</em>, 2018</a>]</span>. We explore patterns of variation in the data with UMAP representation of 4000 highly variable genes.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">anndata</span> <span class="k">as</span> <span class="nn">ad</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">anndata</span> <span class="k">as</span> <span class="nn">ad</span>
<span class="kn">import</span> <span class="nn">gdown</span>
<span class="kn">import</span> <span class="nn">anndata2ri</span>
<span class="kn">import</span> <span class="nn">rpy2</span>
<span class="kn">from</span> <span class="nn">rpy2.robjects</span> <span class="kn">import</span> <span class="n">r</span>
<span class="kn">import</span> <span class="nn">random</span>
<span class="kn">import</span> <span class="nn">session_info</span>


<span class="n">anndata2ri</span><span class="o">.</span><span class="n">activate</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>During startup - Warning message:
Setting LC_CTYPE failed, using &quot;C&quot; 
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span> <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%</span><span class="k">load_ext</span> rpy2.ipython
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
suppressPackageStartupMessages({
    library(SingleCellExperiment)
    library(fgsea)
    library(AUCell)
    library(ggplot2)
})
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span>wget -O <span class="s1">&#39;kang_counts_25k.h5ad&#39;</span> https://figshare.com/ndownloader/files/34464122
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>--2022-06-10 13:53:33--  https://figshare.com/ndownloader/files/34464122
Resolving figshare.com (figshare.com)... 34.251.76.130, 108.128.50.10, 2a05:d018:1f4:d000:4660:613f:c42c:9aed, ...
Connecting to figshare.com (figshare.com)|34.251.76.130|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://s3-eu-west-1.amazonaws.com/pfigshare-u-files/34464122/kang.h5ad?X-Amz-Algorithm=AWS4-HMAC-SHA256&amp;X-Amz-Credential=AKIAIYCQYOYV5JSSROOA/20220610/eu-west-1/s3/aws4_request&amp;X-Amz-Date=20220610T115333Z&amp;X-Amz-Expires=10&amp;X-Amz-SignedHeaders=host&amp;X-Amz-Signature=16dd4ce0a3085abdf129da1824f99da45cf39488f037a24a074698ae8756cbff [following]
--2022-06-10 13:53:33--  https://s3-eu-west-1.amazonaws.com/pfigshare-u-files/34464122/kang.h5ad?X-Amz-Algorithm=AWS4-HMAC-SHA256&amp;X-Amz-Credential=AKIAIYCQYOYV5JSSROOA/20220610/eu-west-1/s3/aws4_request&amp;X-Amz-Date=20220610T115333Z&amp;X-Amz-Expires=10&amp;X-Amz-SignedHeaders=host&amp;X-Amz-Signature=16dd4ce0a3085abdf129da1824f99da45cf39488f037a24a074698ae8756cbff
Resolving s3-eu-west-1.amazonaws.com (s3-eu-west-1.amazonaws.com)... 52.218.101.123
Connecting to s3-eu-west-1.amazonaws.com (s3-eu-west-1.amazonaws.com)|52.218.101.123|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 38356412 (37M) [application/octet-stream]
Saving to: &#39;kang_counts_25k.h5ad&#39;

100%[======================================&gt;] 38,356,412  45.2MB/s   in 0.8s   

2022-06-10 13:53:34 (45.2 MB/s) - &#39;kang_counts_25k.h5ad&#39; saved [38356412/38356412]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_25k</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&quot;kang_counts_25k.h5ad&quot;</span><span class="p">)</span>
<span class="n">adata_25k</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 24673 × 15706
    obs: &#39;nCount_RNA&#39;, &#39;nFeature_RNA&#39;, &#39;tsne1&#39;, &#39;tsne2&#39;, &#39;label&#39;, &#39;cluster&#39;, &#39;cell_type&#39;, &#39;replicate&#39;, &#39;nCount_SCT&#39;, &#39;nFeature_SCT&#39;, &#39;integrated_snn_res.0.4&#39;, &#39;seurat_clusters&#39;
    var: &#39;name&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_25k</span><span class="o">.</span><span class="n">X</span> <span class="o">=</span> <span class="n">adata_25k</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">A</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span> <span class="o">=</span> <span class="n">adata_25k</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;counts&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_total</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">target_sum</span><span class="o">=</span><span class="mf">1e4</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Let’s keep a copy of the full dataset for the subsequent analyses before we subset on highly variable genes:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">highly_variable_genes</span><span class="p">(</span>
    <span class="n">adata_</span><span class="p">,</span> <span class="n">n_top_genes</span><span class="o">=</span><span class="mi">4000</span><span class="p">,</span> <span class="n">flavor</span><span class="o">=</span><span class="s2">&quot;seurat_v3&quot;</span><span class="p">,</span> <span class="n">subset</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">layer</span><span class="o">=</span><span class="s2">&quot;counts&quot;</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/icb/soroor.hediyeh-zadeh/miniconda3/envs/pathway/lib/python3.8/site-packages/scanpy/preprocessing/_highly_variable_genes.py:62: UserWarning: `flavor=&#39;seurat_v3&#39;` expects raw count data, but non-integers were found.
  warnings.warn(
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 24673 × 4000
    obs: &#39;nCount_RNA&#39;, &#39;nFeature_RNA&#39;, &#39;tsne1&#39;, &#39;tsne2&#39;, &#39;label&#39;, &#39;cluster&#39;, &#39;cell_type&#39;, &#39;replicate&#39;, &#39;nCount_SCT&#39;, &#39;nFeature_SCT&#39;, &#39;integrated_snn_res.0.4&#39;, &#39;seurat_clusters&#39;
    var: &#39;name&#39;, &#39;highly_variable&#39;, &#39;highly_variable_rank&#39;, &#39;means&#39;, &#39;variances&#39;, &#39;variances_norm&#39;
    uns: &#39;log1p&#39;, &#39;hvg&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
    layers: &#39;counts&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 24673 × 15706
    obs: &#39;nCount_RNA&#39;, &#39;nFeature_RNA&#39;, &#39;tsne1&#39;, &#39;tsne2&#39;, &#39;label&#39;, &#39;cluster&#39;, &#39;cell_type&#39;, &#39;replicate&#39;, &#39;nCount_SCT&#39;, &#39;nFeature_SCT&#39;, &#39;integrated_snn_res.0.4&#39;, &#39;seurat_clusters&#39;
    var: &#39;name&#39;
    uns: &#39;log1p&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
    layers: &#39;counts&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">umap</span>

<span class="n">X</span> <span class="o">=</span> <span class="n">adata_</span><span class="o">.</span><span class="n">X</span>
<span class="n">trans</span> <span class="o">=</span> <span class="n">umap</span><span class="o">.</span><span class="n">UMAP</span><span class="p">(</span>
    <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
    <span class="n">random_state</span><span class="o">=</span><span class="mi">42</span><span class="p">,</span>
    <span class="n">min_dist</span><span class="o">=</span><span class="mf">0.4</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">adata_</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;X_umap&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">trans</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">X</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/icb/soroor.hediyeh-zadeh/miniconda3/envs/pathway/lib/python3.8/site-packages/tqdm/auto.py:22: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span>
    <span class="n">adata_</span><span class="p">,</span>
    <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;label&quot;</span><span class="p">,</span> <span class="s2">&quot;cell_type&quot;</span><span class="p">],</span>
    <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">ncols</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/gsea_pathway_35_0.png" src="../_images/gsea_pathway_35_0.png" />
</div>
</div>
<p>Note that here we manually imported the <code class="docutils literal notranslate"><span class="pre">UMAP</span></code> package instead of calling <code class="docutils literal notranslate"><span class="pre">sc.tl.umap</span></code>, since the latter computes UMAP representation on the Principal Components space of the data subsetted on HVGs, whereas the earlier applies the UMAP dimension reduction directly on the space of HVGs.</p>
<p>We generally recommend determining the differentially expressed genes as outlined in the <code class="docutils literal notranslate"><span class="pre">Differential</span> <span class="pre">gene</span> <span class="pre">expression</span></code> chapter. For simplicity, here we run a t-test using <code class="docutils literal notranslate"><span class="pre">rank_genes_groups</span></code> in <code class="docutils literal notranslate"><span class="pre">scanpy</span></code> to rank genes according to their test statistics for differential expression:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;group&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">label</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s2">&quot;string&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">cell_type</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;group&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata_</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">label</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s2">&quot;string&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="n">adata_</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">cell_type</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># find DE genes by t-test</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">rank_genes_groups</span><span class="p">(</span><span class="n">adata_</span><span class="p">,</span> <span class="s2">&quot;group&quot;</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s2">&quot;t-test&quot;</span><span class="p">,</span> <span class="n">key_added</span><span class="o">=</span><span class="s2">&quot;t-test&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Let’s extract the ranks for genes differentially expressed in response to IFN stimulation in the CD16 Monocyte (FCGR3A+ Monocytes) cluster. We use these ranks and the gene sets from REACTOME to find gene sets enriched in this cell population compared to all other populations using <code class="docutils literal notranslate"><span class="pre">fgsea</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">celltype_condition</span> <span class="o">=</span> <span class="s2">&quot;stim_FCGR3A+ Monocytes&quot;</span>  <span class="c1"># &#39;stimulated_B&#39;,  &#39;stimulated_CD8 T&#39;, &#39;stimulated_CD14 Mono&#39;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># extract ranked gene names and export to R</span>
<span class="n">ranked_genes_</span> <span class="o">=</span> <span class="n">adata_</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s2">&quot;t-test&quot;</span><span class="p">][</span><span class="s2">&quot;names&quot;</span><span class="p">][</span><span class="n">celltype_condition</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span> -i ranked_genes_
ranked_genes = ranked_genes_
head(ranked_genes)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[1] &quot;IFITM3&quot;  &quot;ISG15&quot;   &quot;TYROBP&quot;  &quot;TNFSF10&quot; &quot;S100A11&quot; &quot;MS4A7&quot;  
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># compute ranks and export to r</span>
<span class="n">stats_</span> <span class="o">=</span> <span class="n">adata_</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s2">&quot;t-test&quot;</span><span class="p">][</span><span class="s2">&quot;scores&quot;</span><span class="p">][</span><span class="n">celltype_condition</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span> -i stats_
stats = stats_
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
names(stats) = ranked_genes
</pre></div>
</div>
</div>
</div>
</section>
<section id="cluster-level-gene-set-enrichment-analysis-with-fgsea">
<h3><span class="section-number">18.6.2. </span>Cluster-level gene set enrichment analysis with fgsea<a class="headerlink" href="#cluster-level-gene-set-enrichment-analysis-with-fgsea" title="Permalink to this headline">#</a></h3>
<p>Download and read the <code class="docutils literal notranslate"><span class="pre">gmt</span></code> file for the REACTOME pathways annotated in the C2 collection of MSigDB.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span>wget -O <span class="s1">&#39;c2.cp.reactome.v7.5.1.symbols.gmt&#39;</span> https://figshare.com/ndownloader/files/35233771
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>--2022-06-10 13:56:13--  https://figshare.com/ndownloader/files/35233771
Resolving figshare.com (figshare.com)... 108.128.50.10, 34.251.76.130, 2a05:d018:1f4:d003:1aba:1dd8:169d:eaeb, ...
Connecting to figshare.com (figshare.com)|108.128.50.10|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://s3-eu-west-1.amazonaws.com/pfigshare-u-files/35233771/c2.cp.reactome.v7.5.1.symbols.gmt?X-Amz-Algorithm=AWS4-HMAC-SHA256&amp;X-Amz-Credential=AKIAIYCQYOYV5JSSROOA/20220610/eu-west-1/s3/aws4_request&amp;X-Amz-Date=20220610T115614Z&amp;X-Amz-Expires=10&amp;X-Amz-SignedHeaders=host&amp;X-Amz-Signature=276f271ee0c89ece05a1d3e499a724f340a32162f3abfb4965988f4ddffa041f [following]
--2022-06-10 13:56:14--  https://s3-eu-west-1.amazonaws.com/pfigshare-u-files/35233771/c2.cp.reactome.v7.5.1.symbols.gmt?X-Amz-Algorithm=AWS4-HMAC-SHA256&amp;X-Amz-Credential=AKIAIYCQYOYV5JSSROOA/20220610/eu-west-1/s3/aws4_request&amp;X-Amz-Date=20220610T115614Z&amp;X-Amz-Expires=10&amp;X-Amz-SignedHeaders=host&amp;X-Amz-Signature=276f271ee0c89ece05a1d3e499a724f340a32162f3abfb4965988f4ddffa041f
Resolving s3-eu-west-1.amazonaws.com (s3-eu-west-1.amazonaws.com)... 52.218.90.59
Connecting to s3-eu-west-1.amazonaws.com (s3-eu-west-1.amazonaws.com)|52.218.90.59|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 770046 (752K) [application/octet-stream]
Saving to: &#39;c2.cp.reactome.v7.5.1.symbols.gmt&#39;

100%[======================================&gt;] 770,046     2.64MB/s   in 0.3s   

2022-06-10 13:56:14 (2.64 MB/s) - &#39;c2.cp.reactome.v7.5.1.symbols.gmt&#39; saved [770046/770046]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
pathways &lt;- gmtPathways(&#39;c2.cp.reactome.v7.5.1.symbols.gmt&#39;)
</pre></div>
</div>
</div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">fgsea</span></code> uses the t-statistics from the t-test to rank the genes for the CD16 Monocyte phenotype upon IFN stimulation and computes p-values for each of the pathways annotated in the <code class="docutils literal notranslate"><span class="pre">gmt</span></code> file. <code class="docutils literal notranslate"><span class="pre">fgsea</span></code> also adjusts the p-values for Multiple hypotheis testing. Note that by default, <code class="docutils literal notranslate"><span class="pre">fgsea</span></code> keeps gene sets with a minimum size of 15 genes, and a maximum of 500 genes. We make a bar plot of top 20 pathways significantly enriched in stimulated CD16 Monocytes compared to all other cell types.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
fgseaRes &lt;- fgsea(pathways = pathways, 
                  stats    = stats,
                  eps      = 0.0,
                  minSize  = 15,
                  maxSize  = 500)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
ggdat &lt;- fgseaRes[order(fgseaRes$padj, decreasing=FALSE), ][1:20,]
ggplot(ggdat, aes(x=reorder(pathway, -log10(padj)), y = -log10(padj))) +
    geom_bar(stat=&quot;identity&quot;, fill = &quot;#51A0D5&quot;) + coord_flip() +
    xlab(&#39;Pathway&#39;) + theme_minimal()
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Fontconfig warning: ignoring UTF-8: not a valid region tag
</pre></div>
</div>
<img alt="../_images/gsea_pathway_54_1.png" src="../_images/gsea_pathway_54_1.png" />
</div>
</div>
<p>In the plot above, pathway names are given in the y-axis. The x-axis describes the <span class="math notranslate nohighlight">\(-\log_{10}\)</span>adjusted p-values. Therefore, the longer the height of the bar, the more significant the pathway is. Pathways are ordered by significance. The majority of interferon-related pathways are indeed ranked among the top 20 most significantly enriched pathways. Some IFN-related pathways include, REACTOME_INTERFERON_SIGNALING (ranked 1), REACTOME_INTERFERON_ALPHA_BETA_SIGNALING (ranked 2nd), REACTOME_INTERFERON_GAMMA_SIGNALING (ranked 15th). Overall, <code class="docutils literal notranslate"><span class="pre">fgsea</span></code> did a decent job in identifying the pathways known to be associated with interferon signalling, given that we know a priori that IFN-related pathways should be the top-ranked terms.</p>
<p>Let’s look at the raw output of <code class="docutils literal notranslate"><span class="pre">fgsea</span></code>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
head(ggdat, n=20)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>                                                                         pathway
 1:                                                REACTOME_INTERFERON_SIGNALING
 2:                                     REACTOME_INTERFERON_ALPHA_BETA_SIGNALING
 3:                                            REACTOME_NEUTROPHIL_DEGRANULATION
 4:                                         REACTOME_SELENOAMINO_ACID_METABOLISM
 5:                                 REACTOME_CYTOKINE_SIGNALING_IN_IMMUNE_SYSTEM
 6:         REACTOME_SRP_DEPENDENT_COTRANSLATIONAL_PROTEIN_TARGETING_TO_MEMBRANE
 7:                                   REACTOME_EUKARYOTIC_TRANSLATION_INITIATION
 8:                                   REACTOME_EUKARYOTIC_TRANSLATION_ELONGATION
 9:                                                REACTOME_INNATE_IMMUNE_SYSTEM
10:                                                     REACTOME_RRNA_PROCESSING
11:                                                         REACTOME_TRANSLATION
12:                   REACTOME_RESPONSE_OF_EIF2AK4_GCN2_TO_AMINO_ACID_DEFICIENCY
13:                                         REACTOME_NONSENSE_MEDIATED_DECAY_NMD
14:                                     REACTOME_CELLULAR_RESPONSE_TO_STARVATION
15:                                          REACTOME_INTERFERON_GAMMA_SIGNALING
16:                         REACTOME_ANTIVIRAL_MECHANISM_BY_IFN_STIMULATED_GENES
17: REACTOME_REGULATION_OF_MRNA_STABILITY_BY_PROTEINS_THAT_BIND_AU_RICH_ELEMENTS
18:           REACTOME_THE_ROLE_OF_GTSE1_IN_G2_M_PROGRESSION_AFTER_G2_CHECKPOINT
19:                                   REACTOME_MHC_CLASS_II_ANTIGEN_PRESENTATION
20:                                                   REACTOME_HEDGEHOG_ON_STATE
            pval         padj   log2err         ES       NES size
 1: 1.021127e-16 2.215845e-14 1.0574636  0.9014998  2.604920   52
 2: 2.714509e-14 2.945243e-12 0.9759947  0.9550021  2.275798   24
 3: 2.027122e-11 1.210167e-09 0.8634154  0.7687554  2.444866   85
 4: 2.230723e-11 1.210167e-09 0.8634154 -0.8686020 -2.157166   26
 5: 6.079513e-11 2.638508e-09 0.8390889  0.6583628  2.319097  165
 6: 8.025571e-11 2.902581e-09 0.8390889 -0.8793432 -2.131245   23
 7: 1.089888e-10 3.378654e-09 0.8390889 -0.8572025 -2.173018   28
 8: 2.341228e-10 5.644960e-09 0.8140358 -0.8522373 -2.116525   26
 9: 2.275888e-10 5.644960e-09 0.8266573  0.6355756  2.289050  178
10: 1.448735e-09 3.143756e-08 0.7881868 -0.8390513 -2.127004   28
11: 7.548158e-09 1.489046e-07 0.7477397 -0.7494213 -2.055180   43
12: 1.273755e-08 2.303373e-07 0.7477397 -0.8336699 -2.057261   25
13: 1.563379e-08 2.609640e-07 0.7337620 -0.8288527 -2.079210   27
14: 5.842353e-08 9.055647e-07 0.7195128 -0.7749049 -2.012610   33
15: 1.091666e-06 1.579276e-05 0.6435518  0.8531550  2.102721   26
16: 6.003687e-05 8.142501e-04 0.5573322  0.8654791  1.908613   17
17: 9.885811e-05 1.261895e-03 0.5384341  0.8802864  1.889005   15
18: 1.712869e-04 2.064959e-03 0.5188481  0.8723041  1.871876   15
19: 2.013967e-04 2.300162e-03 0.5188481  0.7971775  1.974569   27
20: 1.152294e-03 1.250239e-02 0.4550599  0.8382120  1.798717   15
                                      leadingEdge
 1:       IFITM3,ISG15,IFIT1,ISG20,IFIT3,IFI6,...
 2:       IFITM3,ISG15,IFIT1,ISG20,IFIT3,IFI6,...
 3:      TYROBP,S100A11,FCER1G,FTH1,CST3,PSAP,...
 4:          RPL5,RPL3,RPLP0,RPSA,RPL10A,RPS5,...
 5:      IFITM3,ISG15,IFIT1,ISG20,SOD2,CXCL10,...
 6:          RPL5,RPL3,RPLP0,RPSA,RPL10A,RPS5,...
 7:          RPL5,RPL3,RPLP0,RPSA,RPL10A,RPS5,...
 8:          RPL5,RPL3,RPLP0,RPSA,RPL10A,RPS5,...
 9:   ISG15,TYROBP,S100A11,FCGR3A,FCER1G,FTH1,...
10:          RPL5,RPL3,RPLP0,RPSA,RPL10A,RPS5,...
11:          RPL5,RPL3,RPLP0,RPSA,RPL10A,RPS5,...
12:          RPL5,RPL3,RPLP0,RPSA,RPL10A,RPS5,...
13:          RPL5,RPL3,RPLP0,RPSA,RPL10A,RPS5,...
14:          RPL5,RPL3,RPLP0,RPSA,RPL10A,RPS5,...
15:         GBP1,MT2A,IRF7,OAS1,OASL,HLA-DPA1,...
16:          ISG15,IFIT1,OAS1,OASL,MX1,UBE2L6,...
17:      PSMB9,PSMA4,PSME2,HSPA1A,HSPB1,HSPA8,...
18: PSMB9,PSMA4,PSME2,HSP90AA1,PSMB8,HSP90AB1,...
19:  CTSS,HLA-DPA1,CD74,HLA-DRA,CTSC,HLA-DRB1,...
20:           PSMB9,PSMA4,PSME2,PSMB8,UBC,UBB,...
</pre></div>
</div>
</div>
</div>
<p>In above, <code class="docutils literal notranslate"><span class="pre">pval</span></code> is the p-value for the enrichment test, <code class="docutils literal notranslate"><span class="pre">padj</span></code> is the adjusted p-value, <code class="docutils literal notranslate"><span class="pre">ES</span></code> and <code class="docutils literal notranslate"><span class="pre">NES</span></code> are enrichment scores and normalized enrichment scores respectively, and <code class="docutils literal notranslate"><span class="pre">size</span></code> indicates the number of genes in the pathway measured in the dataset. The leading genes that contribute to the enrichment score are shown in the last column. Note that enrichment scores are signed. Therefore, a negative scores suggests the pathway is down-regulated and a positive score is indicative of up-regulation of genes in the pathway or gene set.</p>
<p>We apply some minor pre-processing here, including putting the count data back in the <code class="docutils literal notranslate"><span class="pre">.X</span></code> slot of the anndata object to proceed to the next part - that is, scoring the pathway activity levels per cell, for which we use another Bioconductor package, <code class="docutils literal notranslate"><span class="pre">AUCell</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">del</span> <span class="n">adata_</span><span class="o">.</span><span class="n">uns</span>
<span class="k">del</span> <span class="n">adata_</span><span class="o">.</span><span class="n">obs</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># put counts back</span>
<span class="n">adata_</span><span class="o">.</span><span class="n">X</span> <span class="o">=</span> <span class="n">adata_</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;counts&quot;</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="cell-level-pathway-activity-scoring-using-aucell">
<h3><span class="section-number">18.6.3. </span>Cell-level pathway activity scoring using AUCell<a class="headerlink" href="#cell-level-pathway-activity-scoring-using-aucell" title="Permalink to this headline">#</a></h3>
<p>Unlike the previous approach where we assessed gene set <em>enrichment</em> per <em>cluster</em> (or rather cell type), one can <em>score</em> the activity level of pathways and gene sets in each individual cell, that is based on absolute gene expression in the cell, regardless of expression of genes in the other cells. This we can achieve by activity scoring tools such as <code class="docutils literal notranslate"><span class="pre">AUCell</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span> -i adata_
adata_
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/icb/soroor.hediyeh-zadeh/miniconda3/envs/pathway/lib/python3.8/site-packages/anndata2ri/r2py.py:102: FutureWarning: X.dtype being converted to np.float32 from float64. In the next version of anndata (0.9) conversion will not be automatic. Pass dtype explicitly to avoid this warning. Pass `AnnData(X, dtype=X.dtype, ...)` to get the future behavour.
  return AnnData(exprs, obs, var, uns, obsm or None, layers=layers)
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>class: SingleCellExperiment 
dim: 4000 24673 
metadata(0):
assays(2): X counts
rownames(4000): AL627309.1 HES4 ... C21orf58 S100B
rowData names(6): name highly_variable ... variances variances_norm
colnames(24673): AAACATACATTTCC-1 AAACATACCAGAAA-1 ... TTTGCATGGGACGA-2
  TTTGCATGTCTTAC-2
colData names(0):
reducedDimNames(2): PCA UMAP
mainExpName: NULL
altExpNames(0):
</pre></div>
</div>
</div>
</div>
<p>The AUC scores are calculated in two steps: 1. Building the rankings <code class="docutils literal notranslate"><span class="pre">AUCell_buildRankings</span></code> and 2. Calculate enrichment <code class="docutils literal notranslate"><span class="pre">AUCell_calcAUC</span></code>. This is equivalent to using <code class="docutils literal notranslate"><span class="pre">AUCell_run</span></code> in Bioconductor 3.15 and R 4.2.0.
Since we are using R 4.1.2 and an older Biocondutor version, we demonstrate the old two-step <code class="docutils literal notranslate"><span class="pre">AUCell</span></code> workflow. If you have the latest version of R and Biocondutor, the one-step call to  <code class="docutils literal notranslate"><span class="pre">AUCell_run</span></code> will do the same job as the one presented here:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
# Bioconductor 3.15 and R 4.2.0
# adata_aucell &lt;- AUCell_run(adata_ , pathways)

cells_rankings &lt;- AUCell_buildRankings(adata_, plotStats=FALSE)
cells_AUC &lt;- AUCell_calcAUC(pathways, cells_rankings)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>R[write to console]: Quantiles for the number of genes detected by cell: 
(Non-detected genes are shuffled at the end of the ranking. Keep it in mind when choosing the threshold for calculating the AUC).
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span> min   1%   5%  10%  50% 100% 
   5   77   94  104  137  480 
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>R[write to console]: Warning in .AUCell_calcAUC(geneSets = geneSets, rankings = rankings, nCores = nCores,  :
R[write to console]: 
 
R[write to console]:  Using only the first 200 genes (aucMaxRank) to calculate the AUC.

R[write to console]: Warning in .AUCell_calcAUC(geneSets = geneSets, rankings = rankings, nCores = nCores,  :
R[write to console]: 
 
R[write to console]:  The following gene sets will be excluded from the analysis(less than 20% of their genes are available):
REACTOME_INTERLEUKIN_6_SIGNALING, REACTOME_APOPTOSIS, REACTOME_HEMOSTASIS, REACTOME_INTRINSIC_PATHWAY_FOR_APOPTOSIS, REACTOME_MAPK3_ERK1_ACTIVATION, REACTOME_RECOGNITION_OF_DNA_DAMAGE_BY_PCNA_CONTAINING_REPLICATION_COMPLEX, REACTOME_RECOGNITION_AND_ASSOCIATION_OF_DNA_GLYCOSYLASE_WITH_SITE_CONTAINING_AN_AFFECTED_PURINE, REACTOME_ACTIVATION_OF_BAD_AND_TRANSLOCATION_TO_MITOCHONDRIA, REACTOME_RELEASE_OF_APOPTOTIC_FACTORS_FROM_THE_MITOCHONDRIA, REACTOME_FORMATION_OF_APOPTOSOME, REACTOME_ACTIVATION_OF_CASPASES_THROUGH_APOPTOSOME_MEDIATED_CLEAVAGE, REACTOME_CYTOCHROME_C_MEDIATED_APOPTOTIC_RESPONSE, REACTOME_APOPTOTIC_CLEAVAGE_OF_CELLULAR_PROTEINS, REACTOME_SMAC_XIAP_REGULATED_APOPTOTIC_RESPONSE, REACTOME_APOPTOTIC_FACTOR_MEDIATED_RESPONSE, REACTOME_OPIOID_SIGNALLING, REACTOME_PKA_MEDIATED_PHOSPHORYLATION_OF_CREB, REACTOME_CA_DEPENDENT_EVENTS, REACTOME_G_PROTEIN_MEDIATED_EVENTS, REACTOME_ELE [... truncated]

R[write to console]: Genes in the gene sets NOT available in the dataset: 
	REACTOME_TRANSLESION_SYNTHESIS_BY_Y_FAMILY_DNA_POLYMERASES_BYPASSES_LESIONS_ON_DNA_TEMPLATE: 	30 (77% of 39)
	REACTOME_TRANSLESION_SYNTHESIS_BY_POLH: 	14 (74% of 19)
	REACTOME_DISPLACEMENT_OF_DNA_GLYCOSYLASE_BY_APEX1: 	7 (78% of 9)
	REACTOME_POLB_DEPENDENT_LONG_PATCH_BASE_EXCISION_REPAIR: 	4 (50% of 8)
	REACTOME_RESOLUTION_OF_AP_SITES_VIA_THE_MULTIPLE_NUCLEOTIDE_PATCH_REPLACEMENT_PATHWAY: 	17 (68% of 25)
	REACTOME_ACTIVATION_OF_NOXA_AND_TRANSLOCATION_TO_MITOCHONDRIA: 	2 (40% of 5)
	REACTOME_BH3_ONLY_PROTEINS_ASSOCIATE_WITH_AND_INACTIVATE_ANTI_APOPTOTIC_BCL_2_MEMBERS: 	7 (78% of 9)
	REACTOME_CAMK_IV_MEDIATED_PHOSPHORYLATION_OF_CREB: 	7 (70% of 10)
	REACTOME_PRESYNAPTIC_DEPOLARIZATION_AND_CALCIUM_CHANNEL_OPENING: 	8 (73% of 11)
	REACTOME_NEUROTRANSMITTER_CLEARANCE: 	7 (64% of 11)
	REACTOME_INHIBITION_OF_REPLICATION_INITIATION_OF_DAMAGED_DNA_BY_RB1_E2F1: 	9 (69% of 13)
	REACTOME_E2F_ENABLED_INHIBITION_OF_PRE_REPLICATION_COMPLEX_FORMATION: 	7 (78% of 9)
	REACTOME_E2F_MEDIATED_REGULATION_OF_DNA_REPLICATION: 	16 (73% of 22)
	REACTOME_GPVI_MEDIATED_ACTIVATION_CASCADE: 	25 (71% of 35)
	REACTOME_ANTIVIRAL_MECHANISM_BY_IFN_STIMULATED_GENES: 	65 (79% of 82)
	REACTOME_ROS_AND_RNS_PRODUCTION_IN_PHAGOCYTES: 	27 (75% of 36)
	REACTOME_SIGNALING_BY_ERBB2: 	35 (70% of 50)
	REACTOME_SIGNALING_BY_ERBB2_IN_CANCER: 	18 (69% of 26)
	REACTOME_CONSTITUTIVE_SIGNALING_BY_LIGAND_RESPONSIVE_EGFR_CANCER_VARIANTS: 	13 (68% of 19)
	REACTOME_SIGNALING_BY_ERBB4: 	46 (79% of 58)
	REACTOME_CROSS_PRESENTATION_OF_PARTICULATE_EXOGENOUS_ANTIGENS_PHAGOSOMES: 	6 (75% of 8)
	REACTOME_ERYTHROCYTES_TAKE_UP_CARBON_DIOXIDE_AND_RELEASE_OXYGEN: 	9 (69% of 13)
	REACTOME_ERYTHROCYTES_TAKE_UP_OXYGEN_AND_RELEASE_CARBON_DIOXIDE: 	5 (56% of 9)
	REACTOME_SHC1_EVENTS_IN_ERBB2_SIGNALING: 	16 (73% of 22)
	REACTOME_PI3K_EVENTS_IN_ERBB4_SIGNALING: 	7 (70% of 10)
	REACTOME_NUCLEAR_SIGNALING_BY_ERBB4: 	25 (78% of 32)
	REACTOME_DOWNREGULATION_OF_ERBB4_SIGNALING: 	7 (78% of 9)
	REACTOME_CYTOKINE_SIGNALING_IN_IMMUNE_SYSTEM: 	554 (77% of 719)
	REACTOME_CA2_ACTIVATED_K_CHANNELS: 	7 (78% of 9)
	REACTOME_ACROSOME_REACTION_AND_SPERM_OOCYTE_MEMBRANE_BINDING: 	4 (67% of 6)
	REACTOME_GRB7_EVENTS_IN_ERBB2_SIGNALING: 	1 (20% of 5)
	REACTOME_DOWNREGULATION_OF_ERBB2_ERBB3_SIGNALING: 	8 (62% of 13)
	REACTOME_TRANSCRIPTION_OF_E2F_TARGETS_UNDER_NEGATIVE_CONTROL_BY_DREAM_COMPLEX: 	11 (58% of 19)
	REACTOME_TRANSCRIPTION_OF_E2F_TARGETS_UNDER_NEGATIVE_CONTROL_BY_P107_RBL1_AND_P130_RBL2_IN_COMPLEX_WITH_HDAC1: 	8 (50% of 16)
	REACTOME_ABC_TRANSPORTERS_IN_LIPID_HOMEOSTASIS: 	12 (67% of 18)
	REACTOME_ACTIVATION_OF_PUMA_AND_TRANSLOCATION_TO_MITOCHONDRIA: 	6 (67% of 9)
	REACTOME_APOPTOSIS_INDUCED_DNA_FRAGMENTATION: 	10 (77% of 13)
	REACTOME_INTRINSIC_PATHWAY_OF_FIBRIN_CLOT_FORMATION: 	18 (78% of 23)
	REACTOME_COMMON_PATHWAY_OF_FIBRIN_CLOT_FORMATION: 	17 (77% of 22)
	REACTOME_FORMATION_OF_FIBRIN_CLOT_CLOTTING_CASCADE: 	31 (79% of 39)
	REACTOME_SIGNALING_BY_SCF_KIT: 	33 (77% of 43)
	REACTOME_REGULATION_OF_KIT_SIGNALING: 	11 (69% of 16)
	REACTOME_COLLAGEN_DEGRADATION: 	48 (75% of 64)
	REACTOME_TETRAHYDROBIOPTERIN_BH4_SYNTHESIS_RECYCLING_SALVAGE_AND_REGULATION: 	6 (60% of 10)
	REACTOME_COLLAGEN_FORMATION: 	68 (76% of 90)
	REACTOME_REVERSIBLE_HYDRATION_OF_CARBON_DIOXIDE: 	9 (75% of 12)
	REACTOME_ACYL_CHAIN_REMODELING_OF_CL: 	4 (67% of 6)
	REACTOME_DAG_AND_IP3_SIGNALING: 	32 (78% of 41)
	REACTOME_G0_AND_EARLY_G1: 	17 (63% of 27)
	REACTOME_CYTOSOLIC_SULFONATION_OF_SMALL_MOLECULES: 	19 (79% of 24)
	REACTOME_POLO_LIKE_KINASE_MEDIATED_EVENTS: 	12 (75% of 16)
	REACTOME_EUKARYOTIC_TRANSLATION_ELONGATION: 	68 (72% of 94)
	REACTOME_ACTIVATION_OF_MATRIX_METALLOPROTEINASES: 	25 (76% of 33)
	REACTOME_DEGRADATION_OF_CYSTEINE_AND_HOMOCYSTEINE: 	11 (79% of 14)
	REACTOME_EARLY_PHASE_OF_HIV_LIFE_CYCLE: 	11 (79% of 14)
	REACTOME_SYNTHESIS_OF_GLYCOSYLPHOSPHATIDYLINOSITOL_GPI: 	14 (78% of 18)
	REACTOME_ATTACHMENT_OF_GPI_ANCHOR_TO_UPAR: 	4 (57% of 7)
	REACTOME_GLUCAGON_SIGNALING_IN_METABOLIC_REGULATION: 	26 (79% of 33)
	REACTOME_CHREBP_ACTIVATES_METABOLIC_GENE_EXPRESSION: 	6 (75% of 8)
	REACTOME_SIGNALING_BY_EGFR_IN_CANCER: 	18 (72% of 25)
	REACTOME_NEF_AND_SIGNAL_TRANSDUCTION: 	4 (50% of 8)
	REACTOME_THE_ROLE_OF_NEF_IN_HIV_1_REPLICATION_AND_DISEASE_PATHOGENESIS: 	22 (79% of 28)
	REACTOME_COLLAGEN_BIOSYNTHESIS_AND_MODIFYING_ENZYMES: 	53 (79% of 67)
	REACTOME_REGULATION_OF_CHOLESTEROL_BIOSYNTHESIS_BY_SREBP_SREBF: 	43 (78% of 55)
	REACTOME_TRAFFICKING_AND_PROCESSING_OF_ENDOSOMAL_TLR: 	8 (62% of 13)
	REACTOME_INFLUENZA_INFECTION: 	125 (80% of 157)
	REACTOME_NEUROTOXICITY_OF_CLOSTRIDIUM_TOXINS: 	7 (70% of 10)
	REACTOME_TICAM1_RIP1_MEDIATED_IKK_COMPLEX_RECRUITMENT: 	15 (79% of 19)
	REACTOME_SIGNALING_BY_TGF_BETA_RECEPTOR_COMPLEX: 	59 (79% of 75)
	REACTOME_POLYMERASE_SWITCHING_ON_THE_C_STRAND_OF_THE_TELOMERE: 	19 (73% of 26)
	REACTOME_PROCESSIVE_SYNTHESIS_ON_THE_C_STRAND_OF_THE_TELOMERE: 	15 (79% of 19)
	REACTOME_TELOMERE_C_STRAND_LAGGING_STRAND_SYNTHESIS: 	24 (71% of 34)
	REACTOME_TELOMERE_C_STRAND_SYNTHESIS_INITIATION: 	10 (77% of 13)
	REACTOME_ASSEMBLY_OF_THE_HIV_VIRION: 	12 (75% of 16)
	REACTOME_ACTIVATION_OF_ATR_IN_RESPONSE_TO_REPLICATION_STRESS: 	27 (73% of 37)
	REACTOME_UNWINDING_OF_DNA: 	7 (58% of 12)
	REACTOME_SRP_DEPENDENT_COTRANSLATIONAL_PROTEIN_TARGETING_TO_MEMBRANE: 	90 (80% of 113)
	REACTOME_APOBEC3G_MEDIATED_RESISTANCE_TO_HIV_1_INFECTION: 	3 (60% of 5)
	REACTOME_EXTENSION_OF_TELOMERES: 	40 (78% of 51)
	REACTOME_SIGNALING_BY_PDGF: 	45 (78% of 58)
	REACTOME_CELLULAR_HEXOSE_TRANSPORT: 	16 (76% of 21)
	REACTOME_FGFR2_LIGAND_BINDING_AND_ACTIVATION: 	15 (75% of 20)
	REACTOME_FGFR2B_LIGAND_BINDING_AND_ACTIVATION: 	7 (70% of 10)
	REACTOME_GAP_JUNCTION_DEGRADATION: 	9 (75% of 12)
	REACTOME_PRE_NOTCH_PROCESSING_IN_GOLGI: 	14 (78% of 18)
	REACTOME_CHOLESTEROL_BIOSYNTHESIS: 	18 (72% of 25)
	REACTOME_P75NTR_SIGNALS_VIA_NF_KB: 	12 (75% of 16)
	REACTOME_NRAGE_SIGNALS_DEATH_THROUGH_JNK: 	45 (76% of 59)
	REACTOME_P75_NTR_RECEPTOR_MEDIATED_SIGNALLING: 	75 (77% of 97)
	REACTOME_SIGNALING_BY_VEGF: 	82 (77% of 106)
	REACTOME_GRB2_EVENTS_IN_ERBB2_SIGNALING: 	12 (75% of 16)
	REACTOME_PI3K_EVENTS_IN_ERBB2_SIGNALING: 	10 (62% of 16)
	REACTOME_VITAMIN_C_ASCORBATE_METABOLISM: 	5 (62% of 8)
	REACTOME_IMMUNOREGULATORY_INTERACTIONS_BETWEEN_A_LYMPHOID_AND_A_NON_LYMPHOID_CELL: 	149 (78% of 191)
	REACTOME_NEGATIVE_REGULATION_OF_THE_PI3K_AKT_NETWORK: 	88 (78% of 113)
	REACTOME_G_PROTEIN_ACTIVATION: 	17 (71% of 24)
	REACTOME_METABOLISM_OF_NITRIC_OXIDE_NOS3_ACTIVATION_AND_REGULATION: 	11 (73% of 15)
	REACTOME_ASSEMBLY_OF_COLLAGEN_FIBRILS_AND_OTHER_MULTIMERIC_STRUCTURES: 	43 (70% of 61)
	REACTOME_TCR_SIGNALING: 	89 (72% of 124)
	REACTOME_GENERATION_OF_SECOND_MESSENGER_MOLECULES: 	20 (51% of 39)
	REACTOME_CALCINEURIN_ACTIVATES_NFAT: 	6 (67% of 9)
	REACTOME_ENOS_ACTIVATION: 	7 (64% of 11)
	REACTOME_CELL_DEATH_SIGNALLING_VIA_NRAGE_NRIF_AND_NADE: 	58 (76% of 76)
	REACTOME_P75NTR_RECRUITS_SIGNALLING_COMPLEXES: 	10 (77% of 13)
	REACTOME_NF_KB_IS_ACTIVATED_AND_SIGNALS_SURVIVAL: 	10 (77% of 13)
	REACTOME_PECAM1_INTERACTIONS: 	7 (58% of 12)
	REACTOME_TIE2_SIGNALING: 	14 (78% of 18)
	REACTOME_EGFR_INTERACTS_WITH_PHOSPHOLIPASE_C_GAMMA: 	6 (67% of 9)
	REACTOME_MHC_CLASS_II_ANTIGEN_PRESENTATION: 	99 (79% of 126)
	REACTOME_SYNTHESIS_OF_EPOXY_EET_AND_DIHYDROXYEICOSATRIENOIC_ACIDS_DHET: 	6 (75% of 8)
	REACTOME_SYNTHESIS_OF_5_EICOSATETRAENOIC_ACIDS: 	6 (67% of 9)
	REACTOME_SYNTHESIS_OF_LIPOXINS_LX: 	3 (50% of 6)
	REACTOME_SYNTHESIS_OF_12_EICOSATETRAENOIC_ACID_DERIVATIVES: 	4 (57% of 7)
	REACTOME_SYNTHESIS_OF_15_EICOSATETRAENOIC_ACID_DERIVATIVES: 	4 (67% of 6)
	REACTOME_HYALURONAN_METABOLISM: 	13 (76% of 17)
	REACTOME_HYALURONAN_BIOSYNTHESIS_AND_EXPORT: 	2 (40% of 5)
	REACTOME_INTEGRIN_CELL_SURFACE_INTERACTIONS: 	65 (76% of 85)
	REACTOME_ABACAVIR_TRANSMEMBRANE_TRANSPORT: 	3 (60% of 5)
	REACTOME_ABACAVIR_TRANSPORT_AND_METABOLISM: 	7 (70% of 10)
	REACTOME_SYNTHESIS_OF_PROSTAGLANDINS_PG_AND_THROMBOXANES_TX: 	11 (73% of 15)
	REACTOME_DAP12_INTERACTIONS: 	33 (75% of 44)
	REACTOME_TGF_BETA_RECEPTOR_SIGNALING_IN_EMT_EPITHELIAL_TO_MESENCHYMAL_TRANSITION: 	11 (69% of 16)
	REACTOME_TRANSCRIPTIONAL_ACTIVITY_OF_SMAD2_SMAD3_SMAD4_HETEROTRIMER: 	36 (78% of 46)
	REACTOME_SMAD2_SMAD3_SMAD4_HETEROTRIMER_REGULATES_TRANSCRIPTION: 	23 (72% of 32)
	REACTOME_NOTCH2_INTRACELLULAR_DOMAIN_REGULATES_TRANSCRIPTION: 	9 (75% of 12)
	REACTOME_PI3K_AKT_SIGNALING_IN_CANCER: 	83 (79% of 105)
	REACTOME_CONSTITUTIVE_SIGNALING_BY_ABERRANT_PI3K_IN_CANCER: 	57 (73% of 78)
	REACTOME_CROSSLINKING_OF_COLLAGEN_FIBRILS: 	13 (72% of 18)
	REACTOME_SELENOAMINO_ACID_METABOLISM: 	92 (78% of 118)
	REACTOME_DAP12_SIGNALING: 	18 (62% of 29)
	REACTOME_ACTIVATION_OF_GENE_EXPRESSION_BY_SREBF_SREBP: 	32 (76% of 42)
	REACTOME_THE_CANONICAL_RETINOID_CYCLE_IN_RODS_TWILIGHT_VISION: 	17 (74% of 23)
	REACTOME_CONDENSATION_OF_PROMETAPHASE_CHROMOSOMES: 	8 (73% of 11)
	REACTOME_THE_PHOTOTRANSDUCTION_CASCADE: 	25 (74% of 34)
	REACTOME_EPH_EPHRIN_SIGNALING: 	70 (76% of 92)
	REACTOME_POU5F1_OCT4_SOX2_NANOG_REPRESS_GENES_RELATED_TO_DIFFERENTIATION: 	6 (60% of 10)
	REACTOME_SYNDECAN_INTERACTIONS: 	21 (78% of 27)
	REACTOME_NON_INTEGRIN_MEMBRANE_ECM_INTERACTIONS: 	47 (80% of 59)
	REACTOME_ECM_PROTEOGLYCANS: 	60 (79% of 76)
	REACTOME_SCAVENGING_BY_CLASS_B_RECEPTORS: 	4 (67% of 6)
	REACTOME_SCAVENGING_BY_CLASS_A_RECEPTORS: 	15 (79% of 19)
	REACTOME_SCAVENGING_BY_CLASS_F_RECEPTORS: 	4 (67% of 6)
	REACTOME_REGULATION_OF_INNATE_IMMUNE_RESPONSES_TO_CYTOSOLIC_DNA: 	11 (73% of 15)
	REACTOME_GLYCOGEN_STORAGE_DISEASES: 	12 (75% of 16)
	REACTOME_GLYCOGEN_SYNTHESIS: 	11 (69% of 16)
	REACTOME_REGULATION_OF_HSF1_MEDIATED_HEAT_SHOCK_RESPONSE: 	65 (79% of 82)
	REACTOME_HSP90_CHAPERONE_CYCLE_FOR_STEROID_HORMONE_RECEPTORS_SHR_IN_THE_PRESENCE_OF_LIGAND: 	42 (74% of 57)
	REACTOME_HSF1_ACTIVATION: 	20 (65% of 31)
	REACTOME_CELLULAR_RESPONSE_TO_HEAT_STRESS: 	80 (79% of 101)
	REACTOME_ATTENUATION_PHASE: 	16 (57% of 28)
	REACTOME_HSF1_DEPENDENT_TRANSACTIVATION: 	25 (66% of 38)
	REACTOME_DEFECTIVE_ST3GAL3_CAUSES_MCT12_AND_EIEE15: 	6 (75% of 8)
	REACTOME_CLASS_A_1_RHODOPSIN_LIKE_RECEPTORS: 	264 (80% of 331)
	REACTOME_NETRIN_1_SIGNALING: 	34 (68% of 50)
	REACTOME_SEMAPHORIN_INTERACTIONS: 	50 (78% of 64)
	REACTOME_NCAM_SIGNALING_FOR_NEURITE_OUT_GROWTH: 	48 (76% of 63)
	REACTOME_PEPTIDE_LIGAND_BINDING_RECEPTORS: 	146 (74% of 198)
	REACTOME_SIGNALING_BY_ROBO_RECEPTORS: 	168 (77% of 218)
	REACTOME_DEACTIVATION_OF_THE_BETA_CATENIN_TRANSACTIVATING_COMPLEX: 	31 (74% of 42)
	REACTOME_MYOCLONIC_EPILEPSY_OF_LAFORA: 	6 (67% of 9)
	REACTOME_CHEMOKINE_RECEPTORS_BIND_CHEMOKINES: 	27 (47% of 58)
	REACTOME_REGULATION_OF_INSULIN_LIKE_GROWTH_FACTOR_IGF_TRANSPORT_AND_UPTAKE_BY_INSULIN_LIKE_GROWTH_FACTOR_BINDING_PROTEINS_IGFBPS: 	98 (79% of 124)
	REACTOME_GLUCAGON_LIKE_PEPTIDE_1_GLP1_REGULATES_INSULIN_SECRETION: 	32 (76% of 42)
	REACTOME_ABC_FAMILY_PROTEINS_MEDIATED_TRANSPORT: 	80 (78% of 103)
	REACTOME_BETA_CATENIN_INDEPENDENT_WNT_SIGNALING: 	113 (77% of 146)
	REACTOME_VASOPRESSIN_LIKE_RECEPTORS: 	3 (50% of 6)
	REACTOME_COSTIMULATION_BY_THE_CD28_FAMILY: 	47 (64% of 74)
	REACTOME_CD28_CO_STIMULATION: 	23 (70% of 33)
	REACTOME_CD28_DEPENDENT_PI3K_AKT_SIGNALING: 	15 (68% of 22)
	REACTOME_CD28_DEPENDENT_VAV1_PATHWAY: 	8 (67% of 12)
	REACTOME_CTLA4_INHIBITORY_SIGNALING: 	15 (71% of 21)
	REACTOME_ALPHA_OXIDATION_OF_PHYTANATE: 	4 (67% of 6)
	REACTOME_BETA_OXIDATION_OF_PRISTANOYL_COA: 	7 (78% of 9)
	REACTOME_PD_1_SIGNALING: 	15 (54% of 28)
	REACTOME_PEROXISOMAL_LIPID_METABOLISM: 	21 (72% of 29)
	REACTOME_EICOSANOID_LIGAND_BINDING_RECEPTORS: 	10 (67% of 15)
	REACTOME_LEUKOTRIENE_RECEPTORS: 	3 (60% of 5)
	REACTOME_PROSTANOID_LIGAND_RECEPTORS: 	6 (67% of 9)
	REACTOME_ADP_SIGNALLING_THROUGH_P2Y_PURINOCEPTOR_12: 	16 (73% of 22)
	REACTOME_G_BETA_GAMMA_SIGNALLING_THROUGH_PI3KGAMMA: 	18 (72% of 25)
	REACTOME_SIGNAL_AMPLIFICATION: 	26 (79% of 33)
	REACTOME_PROSTACYCLIN_SIGNALLING_THROUGH_PROSTACYCLIN_RECEPTOR: 	13 (68% of 19)
	REACTOME_EPHB_MEDIATED_FORWARD_SIGNALING: 	28 (67% of 42)
	REACTOME_EPHA_MEDIATED_GROWTH_CONE_COLLAPSE: 	21 (72% of 29)
	REACTOME_EPHRIN_SIGNALING: 	14 (74% of 19)
	REACTOME_G_PROTEIN_BETA_GAMMA_SIGNALLING: 	25 (78% of 32)
	REACTOME_SEMA3A_PAK_DEPENDENT_AXON_REPULSION: 	11 (69% of 16)
	REACTOME_ADRENALINE_NORADRENALINE_INHIBITS_INSULIN_SECRETION: 	20 (71% of 28)
	REACTOME_SEMA4D_IN_SEMAPHORIN_SIGNALING: 	19 (79% of 24)
	REACTOME_CA2_PATHWAY: 	43 (69% of 62)
	REACTOME_G_ALPHA_12_13_SIGNALLING_EVENTS: 	60 (75% of 80)
	REACTOME_PLATELET_HOMEOSTASIS: 	68 (79% of 86)
	REACTOME_REDUCTION_OF_CYTOSOLIC_CA_LEVELS: 	8 (67% of 12)
	REACTOME_PLATELET_CALCIUM_HOMEOSTASIS: 	22 (79% of 28)
	REACTOME_ADP_SIGNALLING_THROUGH_P2Y_PURINOCEPTOR_1: 	19 (76% of 25)
	REACTOME_G_ALPHA_I_SIGNALLING_EVENTS: 	248 (79% of 314)
	REACTOME_G_ALPHA_Z_SIGNALLING_EVENTS: 	37 (77% of 48)
	REACTOME_DCC_MEDIATED_ATTRACTIVE_SIGNALING: 	8 (57% of 14)
	REACTOME_NCAM1_INTERACTIONS: 	32 (76% of 42)
	REACTOME_LYSOSPHINGOLIPID_AND_LPA_RECEPTORS: 	10 (71% of 14)
	REACTOME_TIGHT_JUNCTION_INTERACTIONS: 	23 (77% of 30)
	REACTOME_GLUCAGON_TYPE_LIGAND_RECEPTORS: 	25 (76% of 33)
	REACTOME_SYNTHESIS_SECRETION_AND_DEACYLATION_OF_GHRELIN: 	15 (79% of 19)
	REACTOME_REGULATION_OF_INSULIN_SECRETION: 	62 (79% of 78)
	REACTOME_BICARBONATE_TRANSPORTERS: 	7 (70% of 10)
	REACTOME_CATION_COUPLED_CHLORIDE_COTRANSPORTERS: 	5 (71% of 7)
	REACTOME_REGULATION_OF_COMMISSURAL_AXON_PATHFINDING_BY_SLIT_AND_ROBO: 	7 (70% of 10)
	REACTOME_INACTIVATION_OF_CDC42_AND_RAC1: 	5 (62% of 8)
	REACTOME_THROMBOXANE_SIGNALLING_THROUGH_TP_RECEPTOR: 	17 (71% of 24)
	REACTOME_GP1B_IX_V_ACTIVATION_SIGNALLING: 	8 (67% of 12)
	REACTOME_VASOPRESSIN_REGULATES_RENAL_WATER_HOMEOSTASIS_VIA_AQUAPORINS: 	33 (77% of 43)
	REACTOME_PASSIVE_TRANSPORT_BY_AQUAPORINS: 	10 (77% of 13)
	REACTOME_PROTON_COUPLED_MONOCARBOXYLATE_TRANSPORT: 	4 (67% of 6)
	REACTOME_UNBLOCKING_OF_NMDA_RECEPTORS_GLUTAMATE_BINDING_AND_ACTIVATION: 	14 (67% of 21)
	REACTOME_BINDING_OF_TCF_LEF_CTNNB1_TO_TARGET_GENE_PROMOTERS: 	4 (50% of 8)
	REACTOME_CREB1_PHOSPHORYLATION_THROUGH_THE_ACTIVATION_OF_ADENYLATE_CYCLASE: 	8 (67% of 12)
	REACTOME_CREB1_PHOSPHORYLATION_THROUGH_THE_ACTIVATION_OF_CAMKII_CAMKK_CAMKIV_CASCASDE: 	5 (62% of 8)
	REACTOME_CREB1_PHOSPHORYLATION_THROUGH_NMDA_RECEPTOR_MEDIATED_ACTIVATION_OF_RAS_SIGNALING: 	21 (75% of 28)
	REACTOME_RAS_ACTIVATION_UPON_CA2_INFLUX_THROUGH_NMDA_RECEPTOR: 	13 (65% of 20)
	REACTOME_RELAXIN_RECEPTORS: 	6 (75% of 8)
	REACTOME_INTERACTION_BETWEEN_L1_AND_ANKYRINS: 	24 (77% of 31)
	REACTOME_SMOOTH_MUSCLE_CONTRACTION: 	30 (79% of 38)
	REACTOME_AQUAPORIN_MEDIATED_TRANSPORT: 	40 (77% of 52)
	REACTOME_TYPE_I_HEMIDESMOSOME_ASSEMBLY: 	8 (73% of 11)
	REACTOME_NRCAM_INTERACTIONS: 	5 (71% of 7)
	REACTOME_CHL1_INTERACTIONS: 	5 (56% of 9)
	REACTOME_INTERLEUKIN_12_FAMILY_SIGNALING: 	41 (72% of 57)
	REACTOME_SIGNALING_BY_INTERLEUKINS: 	364 (79% of 463)
	REACTOME_OTHER_INTERLEUKIN_SIGNALING: 	19 (79% of 24)
	REACTOME_AUF1_HNRNP_D0_BINDS_AND_DESTABILIZES_MRNA: 	42 (76% of 55)
	REACTOME_IONOTROPIC_ACTIVITY_OF_KAINATE_RECEPTORS: 	6 (60% of 10)
	REACTOME_ACTIVATION_OF_KAINATE_RECEPTORS_UPON_GLUTAMATE_BINDING: 	20 (67% of 30)
	REACTOME_INTERLEUKIN_2_FAMILY_SIGNALING: 	32 (73% of 44)
	REACTOME_MITOTIC_G1_PHASE_AND_G1_S_TRANSITION: 	119 (80% of 149)
	REACTOME_THROMBIN_SIGNALLING_THROUGH_PROTEINASE_ACTIVATED_RECEPTORS_PARS: 	25 (78% of 32)
	REACTOME_REPRESSION_OF_WNT_TARGET_GENES: 	8 (57% of 14)
	REACTOME_INTERCONVERSION_OF_NUCLEOTIDE_DI_AND_TRIPHOSPHATES: 	21 (72% of 29)
	REACTOME_PRESYNAPTIC_FUNCTION_OF_KAINATE_RECEPTORS: 	15 (71% of 21)
	REACTOME_INTERLEUKIN_3_INTERLEUKIN_5_AND_GM_CSF_SIGNALING: 	35 (73% of 48)
	REACTOME_RIPK1_MEDIATED_REGULATED_NECROSIS: 	22 (76% of 29)
	REACTOME_VEGFR2_MEDIATED_VASCULAR_PERMEABILITY: 	20 (74% of 27)
	REACTOME_VEGFR2_MEDIATED_CELL_PROLIFERATION: 	14 (74% of 19)
	REACTOME_DISEASES_OF_IMMUNE_SYSTEM: 	24 (77% of 31)
	REACTOME_UPTAKE_AND_FUNCTION_OF_DIPHTHERIA_TOXIN: 	2 (33% of 6)
	REACTOME_UPTAKE_AND_ACTIONS_OF_BACTERIAL_TOXINS: 	21 (72% of 29)
	REACTOME_SYNTHESIS_OF_DIPHTHAMIDE_EEF2: 	5 (62% of 8)
	REACTOME_RA_BIOSYNTHESIS_PATHWAY: 	17 (77% of 22)
	REACTOME_ION_HOMEOSTASIS: 	42 (78% of 54)
	REACTOME_PIWI_INTERACTING_RNA_PIRNA_BIOGENESIS: 	22 (76% of 29)
	REACTOME_IRAK4_DEFICIENCY_TLR2_4: 	13 (72% of 18)
	REACTOME_CLEC7A_DECTIN_1_INDUCES_NFAT_ACTIVATION: 	7 (64% of 11)
	REACTOME_ABC_TRANSPORTER_DISORDERS: 	62 (79% of 78)
	REACTOME_BBSOME_MEDIATED_CARGO_TARGETING_TO_CILIUM: 	17 (74% of 23)
	REACTOME_INTRAFLAGELLAR_TRANSPORT: 	39 (72% of 54)
	REACTOME_RHO_GTPASES_ACTIVATE_IQGAPS: 	24 (75% of 32)
	REACTOME_RHO_GTPASES_ACTIVATE_PAKS: 	15 (71% of 21)
	REACTOME_ACTIVATION_OF_SMO: 	13 (72% of 18)
	REACTOME_CONSTITUTIVE_SIGNALING_BY_EGFRVIII: 	11 (73% of 15)
	REACTOME_PCNA_DEPENDENT_LONG_PATCH_BASE_EXCISION_REPAIR: 	15 (71% of 21)
	REACTOME_FRUCTOSE_METABOLISM: 	5 (71% of 7)
	REACTOME_PHOSPHOLIPASE_C_MEDIATED_CASCADE_FGFR2: 	14 (78% of 18)
	REACTOME_TRANSLESION_SYNTHESIS_BY_POLK: 	13 (76% of 17)
	REACTOME_TERMINATION_OF_TRANSLESION_DNA_SYNTHESIS: 	23 (72% of 32)
	REACTOME_RESPONSE_TO_METAL_IONS: 	10 (71% of 14)
	REACTOME_METALLOTHIONEINS_BIND_METALS: 	7 (64% of 11)
	REACTOME_DISEASES_OF_CARBOHYDRATE_METABOLISM: 	27 (79% of 34)
	REACTOME_RHO_GTPASES_ACTIVATE_WASPS_AND_WAVES: 	25 (69% of 36)
	REACTOME_RHO_GTPASES_ACTIVATE_FORMINS: 	108 (77% of 140)
	REACTOME_TNFR2_NON_CANONICAL_NF_KB_PATHWAY: 	75 (74% of 101)
	REACTOME_RHO_GTPASES_ACTIVATE_NADPH_OXIDASES: 	18 (75% of 24)
	REACTOME_TNFS_BIND_THEIR_PHYSIOLOGICAL_RECEPTORS: 	16 (55% of 29)
	REACTOME_NEGATIVE_REGULATION_OF_MAPK_PATHWAY: 	33 (77% of 43)
	REACTOME_TNF_RECEPTOR_SUPERFAMILY_TNFSF_MEMBERS_MEDIATING_NON_CANONICAL_NF_KB_PATHWAY: 	11 (65% of 17)
	REACTOME_MAPK_FAMILY_SIGNALING_CASCADES: 	258 (79% of 327)
	REACTOME_HDR_THROUGH_MMEJ_ALT_NHEJ: 	7 (70% of 10)
	REACTOME_HDR_THROUGH_HOMOLOGOUS_RECOMBINATION_HRR: 	52 (78% of 67)
	REACTOME_RESOLUTION_OF_D_LOOP_STRUCTURES_THROUGH_SYNTHESIS_DEPENDENT_STRAND_ANNEALING_SDSA: 	20 (77% of 26)
	REACTOME_HOMOLOGOUS_DNA_PAIRING_AND_STRAND_EXCHANGE: 	32 (76% of 42)
	REACTOME_DNA_DAMAGE_RECOGNITION_IN_GG_NER: 	29 (76% of 38)
	REACTOME_GAP_FILLING_DNA_REPAIR_SYNTHESIS_AND_LIGATION_IN_GG_NER: 	19 (76% of 25)
	REACTOME_DUAL_INCISION_IN_GG_NER: 	32 (78% of 41)
	REACTOME_FANCONI_ANEMIA_PATHWAY: 	30 (77% of 39)
	REACTOME_INTERLEUKIN_6_FAMILY_SIGNALING: 	18 (75% of 24)
	REACTOME_INTERLEUKIN_10_SIGNALING: 	30 (65% of 46)
	REACTOME_ERBB2_REGULATES_CELL_MOTILITY: 	10 (67% of 15)
	REACTOME_INTERLEUKIN_4_AND_INTERLEUKIN_13_SIGNALING: 	81 (73% of 111)
	REACTOME_IL_6_TYPE_CYTOKINE_RECEPTOR_LIGAND_INTERACTIONS: 	13 (76% of 17)
	REACTOME_TP53_REGULATES_TRANSCRIPTION_OF_CELL_CYCLE_GENES: 	37 (76% of 49)
	REACTOME_METAL_SEQUESTRATION_BY_ANTIMICROBIAL_PROTEINS: 	3 (50% of 6)
	REACTOME_TP53_REGULATES_TRANSCRIPTION_OF_GENES_INVOLVED_IN_G2_CELL_CYCLE_ARREST: 	14 (78% of 18)
	REACTOME_TP53_REGULATES_TRANSCRIPTION_OF_ADDITIONAL_CELL_CYCLE_GENES_WHOSE_EXACT_ROLE_IN_THE_P53_PATHWAY_REMAIN_UNCERTAIN: 	16 (76% of 21)
	REACTOME_TP53_REGULATES_TRANSCRIPTION_OF_GENES_INVOLVED_IN_G1_CELL_CYCLE_ARREST: 	11 (79% of 14)
	REACTOME_REGULATION_OF_TP53_ACTIVITY_THROUGH_METHYLATION: 	14 (74% of 19)
	REACTOME_GLYCEROPHOSPHOLIPID_CATABOLISM: 	5 (71% of 7)
	REACTOME_CDC6_ASSOCIATION_WITH_THE_ORC_ORIGIN_COMPLEX: 	8 (73% of 11)
	REACTOME_DNA_REPLICATION_INITIATION: 	5 (62% of 8)
	REACTOME_ACTIVATION_OF_THE_PRE_REPLICATIVE_COMPLEX: 	24 (73% of 33)
	REACTOME_POLYMERASE_SWITCHING: 	9 (64% of 14)
	REACTOME_PROCESSIVE_SYNTHESIS_ON_THE_LAGGING_STRAND: 	10 (67% of 15)
	REACTOME_LAGGING_STRAND_SYNTHESIS: 	13 (65% of 20)
	REACTOME_DNA_STRAND_ELONGATION: 	20 (62% of 32)
	REACTOME_G1_S_SPECIFIC_TRANSCRIPTION: 	18 (62% of 29)
	REACTOME_CYCLIN_D_ASSOCIATED_EVENTS_IN_G1: 	36 (77% of 47)
	REACTOME_SYNTHESIS_OF_DNA: 	93 (78% of 120)
	REACTOME_S_PHASE: 	126 (78% of 162)
	REACTOME_MITOTIC_SPINDLE_CHECKPOINT: 	88 (79% of 111)
	REACTOME_GLYCOGEN_BREAKDOWN_GLYCOGENOLYSIS: 	10 (67% of 15)
	REACTOME_GALACTOSE_CATABOLISM: 	3 (60% of 5)
	REACTOME_BRANCHED_CHAIN_AMINO_ACID_CATABOLISM: 	16 (76% of 21)
	REACTOME_HISTIDINE_CATABOLISM: 	5 (62% of 8)
	REACTOME_LYSINE_CATABOLISM: 	9 (75% of 12)
	REACTOME_TRYPTOPHAN_CATABOLISM: 	10 (71% of 14)
	REACTOME_MRNA_EDITING_C_TO_U_CONVERSION: 	4 (50% of 8)
	REACTOME_EUKARYOTIC_TRANSLATION_INITIATION: 	92 (77% of 120)
	REACTOME_ACTIVATION_OF_THE_MRNA_UPON_BINDING_OF_THE_CAP_BINDING_COMPLEX_AND_EIFS_AND_SUBSEQUENT_BINDING_TO_43S: 	46 (77% of 60)
	REACTOME_RNA_POLYMERASE_II_TRANSCRIPTION: 	1095 (80% of 1375)
	REACTOME_RESOLUTION_OF_ABASIC_SITES_AP_SITES: 	28 (74% of 38)
	REACTOME_MRNA_EDITING: 	5 (50% of 10)
	REACTOME_DISSOLUTION_OF_FIBRIN_CLOT: 	8 (62% of 13)
	REACTOME_PLATELET_ADHESION_TO_EXPOSED_COLLAGEN: 	8 (53% of 15)
	REACTOME_PLATELET_ACTIVATION_SIGNALING_AND_AGGREGATION: 	202 (77% of 261)
	REACTOME_RESPONSE_TO_ELEVATED_PLATELET_CYTOSOLIC_CA2: 	100 (76% of 132)
	REACTOME_PLATELET_AGGREGATION_PLUG_FORMATION: 	31 (79% of 39)
	REACTOME_SYNTHESIS_OF_KETONE_BODIES: 	6 (75% of 8)
	REACTOME_INSULIN_RECEPTOR_RECYCLING: 	20 (77% of 26)
	REACTOME_TRANSPORT_OF_FATTY_ACIDS: 	5 (62% of 8)
	REACTOME_TRANSPORT_OF_NUCLEOSIDES_AND_FREE_PURINE_AND_PYRIMIDINE_BASES_ACROSS_THE_PLASMA_MEMBRANE: 	9 (75% of 12)
	REACTOME_INTERFERON_GAMMA_SIGNALING: 	67 (72% of 93)
	REACTOME_REGULATION_OF_IFNG_SIGNALING: 	11 (79% of 14)
	REACTOME_ADVANCED_GLYCOSYLATION_ENDPRODUCT_RECEPTOR_SIGNALING: 	10 (77% of 13)
	REACTOME_ERBB2_ACTIVATES_PTK6_SIGNALING: 	8 (62% of 13)
	REACTOME_SIGNALING_BY_PTK6: 	41 (76% of 54)
	REACTOME_PTK6_REGULATES_RTKS_AND_THEIR_EFFECTORS_AKT1_AND_DOK1: 	7 (78% of 9)
	REACTOME_PTK6_EXPRESSION: 	3 (60% of 5)
	REACTOME_SYNAPTIC_ADHESION_LIKE_MOLECULES: 	16 (76% of 21)
	REACTOME_BUTYROPHILIN_BTN_FAMILY_INTERACTIONS: 	9 (75% of 12)
	REACTOME_DOWNREGULATION_OF_ERBB2_SIGNALING: 	20 (69% of 29)
	REACTOME_REELIN_SIGNALLING_PATHWAY: 	3 (60% of 5)
	REACTOME_TFAP2A_ACTS_AS_A_TRANSCRIPTIONAL_REPRESSOR_DURING_RETINOIC_ACID_INDUCED_CELL_DIFFERENTIATION: 	3 (60% of 5)
	REACTOME_CREB3_FACTORS_ACTIVATE_GENES: 	7 (78% of 9)
	REACTOME_INLA_MEDIATED_ENTRY_OF_LISTERIA_MONOCYTOGENES_INTO_HOST_CELLS: 	7 (78% of 9)
	REACTOME_PROTEIN_METHYLATION: 	12 (71% of 17)
	REACTOME_RUNX1_AND_FOXP3_CONTROL_THE_DEVELOPMENT_OF_REGULATORY_T_LYMPHOCYTES_TREGS: 	3 (30% of 10)
	REACTOME_RUNX1_REGULATES_TRANSCRIPTION_OF_GENES_INVOLVED_IN_DIFFERENTIATION_OF_KERATINOCYTES: 	6 (75% of 8)
	REACTOME_RUNX1_REGULATES_TRANSCRIPTION_OF_GENES_INVOLVED_IN_DIFFERENTIATION_OF_MYELOID_CELLS: 	5 (71% of 7)
	REACTOME_RUNX2_REGULATES_OSTEOBLAST_DIFFERENTIATION: 	17 (71% of 24)
	REACTOME_RUNX2_REGULATES_BONE_DEVELOPMENT: 	23 (74% of 31)
	REACTOME_COLLAGEN_CHAIN_TRIMERIZATION: 	33 (75% of 44)
	REACTOME_REGULATION_OF_PTEN_LOCALIZATION: 	7 (78% of 9)
	REACTOME_GENE_AND_PROTEIN_EXPRESSION_BY_JAK_STAT_SIGNALING_AFTER_INTERLEUKIN_12_STIMULATION: 	24 (63% of 38)
	REACTOME_RUNX3_REGULATES_WNT_SIGNALING: 	5 (62% of 8)
	REACTOME_TRANSCRIPTIONAL_REGULATION_BY_E2F6: 	27 (79% of 34)
	REACTOME_HDL_ASSEMBLY: 	6 (75% of 8)
	REACTOME_CHYLOMICRON_CLEARANCE: 	3 (60% of 5)
	REACTOME_VLDL_CLEARANCE: 	4 (67% of 6)
	REACTOME_LIPID_PARTICLE_ORGANIZATION: 	2 (33% of 6)
	REACTOME_G_BETA_GAMMA_SIGNALLING_THROUGH_CDC42: 	14 (70% of 20)
	REACTOME_METABOLISM_OF_COFACTORS: 	14 (74% of 19)
	REACTOME_GLYCOGEN_METABOLISM: 	18 (67% of 27)
	REACTOME_OAS_ANTIVIRAL_RESPONSE: 	7 (78% of 9)
	REACTOME_INTERLEUKIN_9_SIGNALING: 	7 (78% of 9)
	REACTOME_LOSS_OF_FUNCTION_OF_MECP2_IN_RETT_SYNDROME: 	10 (77% of 13)
	REACTOME_SIGNALING_BY_TGFB_FAMILY_MEMBERS: 	82 (79% of 104)
	REACTOME_EXTRA_NUCLEAR_ESTROGEN_SIGNALING: 	58 (77% of 75)
	REACTOME_REGULATION_OF_EXPRESSION_OF_SLITS_AND_ROBOS: 	134 (78% of 172)
	REACTOME_ROBO_RECEPTORS_BIND_AKAP5: 	5 (56% of 9)
	REACTOME_INTERLEUKIN_18_SIGNALING: 	6 (75% of 8)
	REACTOME_RHOBTB2_GTPASE_CYCLE: 	18 (78% of 23)
	REACTOME_RHOV_GTPASE_CYCLE: 	29 (78% of 37)
	REACTOME_TICAM1_DEPENDENT_ACTIVATION_OF_IRF3_IRF7: 	7 (58% of 12)
	REACTOME_TICAM1_TRAF6_DEPENDENT_INDUCTION_OF_TAK1_COMPLEX: 	7 (64% of 11)
	REACTOME_INTERLEUKIN_36_PATHWAY: 	5 (71% of 7)
	REACTOME_BIOSYNTHESIS_OF_SPECIALIZED_PRORESOLVING_MEDIATORS_SPMS: 	14 (74% of 19)
	REACTOME_BIOSYNTHESIS_OF_EPA_DERIVED_SPMS: 	4 (67% of 6)
	REACTOME_INTERLEUKIN_2_SIGNALING: 	7 (58% of 12)
	REACTOME_INTERLEUKIN_12_SIGNALING: 	32 (68% of 47)
	REACTOME_INTERLEUKIN_23_SIGNALING: 	7 (78% of 9)
	REACTOME_INTERLEUKIN_21_SIGNALING: 	7 (70% of 10)
	REACTOME_ERYTHROPOIETIN_ACTIVATES_PHOSPHOINOSITIDE_3_KINASE_PI3K: 	8 (67% of 12)
	REACTOME_ERYTHROPOIETIN_ACTIVATES_PHOSPHOLIPASE_C_GAMMA_PLCG: 	5 (71% of 7)
	REACTOME_ACTIVATED_NTRK2_SIGNALS_THROUGH_FYN: 	5 (71% of 7)
	REACTOME_NTRK2_ACTIVATES_RAC1: 	3 (60% of 5)
	REACTOME_ACTIVATED_NTRK2_SIGNALS_THROUGH_CDK5: 	4 (67% of 6)
	REACTOME_BLOOD_GROUP_SYSTEMS_BIOSYNTHESIS: 	16 (76% of 21)
	REACTOME_RHOF_GTPASE_CYCLE: 	33 (79% of 42)
	REACTOME_INTERFERON_ALPHA_BETA_SIGNALING: 	47 (66% of 71)
	REACTOME_INTERLEUKIN_RECEPTOR_SHC_SIGNALING: 	19 (70% of 27)
	REACTOME_REGULATION_OF_SIGNALING_BY_CBL: 	15 (68% of 22)
	REACTOME_INTERFERON_SIGNALING: 	151 (74% of 203)
	REACTOME_IRON_UPTAKE_AND_TRANSPORT: 	46 (79% of 58)
	REACTOME_NONSENSE_MEDIATED_DECAY_NMD: 	89 (77% of 116)
	REACTOME_NEGATIVE_REGULATORS_OF_DDX58_IFIH1_SIGNALING: 	27 (77% of 35)
	REACTOME_ION_TRANSPORT_BY_P_TYPE_ATPASES: 	40 (73% of 55)
	REACTOME_ACTIVATION_OF_IRF3_IRF7_MEDIATED_BY_TBK1_IKK_EPSILON: 	12 (71% of 17)
	REACTOME_IRAK1_RECRUITS_IKK_COMPLEX: 	11 (79% of 14)
	REACTOME_IRAK2_MEDIATED_ACTIVATION_OF_TAK1_COMPLEX: 	7 (70% of 10)
	REACTOME_TRAF6_MEDIATED_INDUCTION_OF_TAK1_COMPLEX_WITHIN_TLR4_COMPLEX: 	12 (75% of 16)
	REACTOME_MOLYBDENUM_COFACTOR_BIOSYNTHESIS: 	3 (50% of 6)
	REACTOME_DISEASES_OF_BASE_EXCISION_REPAIR: 	3 (60% of 5)
	REACTOME_FLT3_SIGNALING: 	30 (79% of 38)
	REACTOME_ASSEMBLY_AND_CELL_SURFACE_PRESENTATION_OF_NMDA_RECEPTORS: 	35 (80% of 44)
	REACTOME_CHAPERONE_MEDIATED_AUTOPHAGY: 	14 (64% of 22)
	REACTOME_NEGATIVE_REGULATION_OF_NMDA_RECEPTOR_MEDIATED_NEURONAL_TRANSMISSION: 	13 (62% of 21)
	REACTOME_FOXO_MEDIATED_TRANSCRIPTION_OF_CELL_CYCLE_GENES: 	13 (76% of 17)
	REACTOME_ACTIVATION_OF_RAC1_DOWNSTREAM_OF_NMDARS: 	5 (71% of 7)
	REACTOME_LONG_TERM_POTENTIATION: 	15 (65% of 23)
	REACTOME_RESPONSE_OF_EIF2AK4_GCN2_TO_AMINO_ACID_DEFICIENCY: 	77 (75% of 102)
	REACTOME_PREVENTION_OF_PHAGOSOMAL_LYSOSOMAL_FUSION: 	7 (78% of 9)
	REACTOME_MODULATION_BY_MTB_OF_HOST_IMMUNE_SYSTEM: 	5 (71% of 7)
	REACTOME_ALPHA_PROTEIN_KINASE_1_SIGNALING_PATHWAY: 	7 (64% of 11)
	REACTOME_RAS_PROCESSING: 	19 (79% of 24)
	REACTOME_ABERRANT_REGULATION_OF_MITOTIC_G1_S_TRANSITION_IN_CANCER_DUE_TO_RB1_DEFECTS: 	13 (76% of 17)
	REACTOME_CD163_MEDIATING_AN_ANTI_INFLAMMATORY_RESPONSE: 	7 (78% of 9)
	REACTOME_KILLING_MECHANISMS: 	9 (75% of 12)
	REACTOME_PEXOPHAGY: 	8 (73% of 11)
	REACTOME_SIGNALING_BY_ERBB2_ECD_MUTANTS: 	11 (69% of 16)
	REACTOME_DEFECTIVE_FACTOR_IX_CAUSES_HEMOPHILIA_B: 	7 (78% of 9)
	REACTOME_DEFECTIVE_F9_ACTIVATION: 	4 (67% of 6)
	REACTOME_NERVOUS_SYSTEM_DEVELOPMENT: 	454 (78% of 580)
	REACTOME_DISEASES_OF_DNA_REPAIR: 	26 (76% of 34)
	REACTOME_FLT3_SIGNALING_IN_DISEASE: 	22 (79% of 28)
	REACTOME_MATURATION_OF_PROTEIN_3A: 	7 (78% of 9)
	REACTOME_ALK_MUTANTS_BIND_TKIS: 	9 (75% of 12)
	REACTOME_SIGNALING_BY_FLT3_FUSION_PROTEINS: 	15 (79% of 19)
	REACTOME_NEGATIVE_REGULATION_OF_FLT3: 	11 (73% of 15)
	REACTOME_FLT3_SIGNALING_THROUGH_SRC_FAMILY_KINASES: 	4 (67% of 6)
	REACTOME_FLT3_SIGNALING_BY_CBL_MUTANTS: 	5 (71% of 7)
	REACTOME_CELLULAR_RESPONSE_TO_STARVATION: 	124 (79% of 157)
	REACTOME_NUCLEAR_EVENTS_STIMULATED_BY_ALK_SIGNALING_IN_CANCER: 	15 (79% of 19)
	REACTOME_TRAF6_MEDIATED_IRF7_ACTIVATION_IN_TLR7_8_OR_9_SIGNALING: 	9 (69% of 13)
	REACTOME_SERINE_BIOSYNTHESIS: 	7 (78% of 9)
	REACTOME_KINESINS: 	47 (77% of 61)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
dim(cells_AUC)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[1]   432 24673
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>

aucell_scores &lt;- data.frame(cells_AUC@assays@data$AUC)
rownames(aucell_scores) &lt;- cells_AUC@NAMES
colnames(aucell_scores) &lt;- colnames(adata_)
</pre></div>
</div>
</div>
</div>
<p>Export the results back to python</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span> -o aucell_scores  
aucell_scores = aucell_scores
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span> -o adata_ 
adata_ = adata_
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">aucell_scores</span> <span class="o">=</span> <span class="n">aucell_scores</span><span class="o">.</span><span class="n">T</span>
</pre></div>
</div>
</div>
</div>
<p>We now add the scores for the interferon-related REACTOME pathways to the anndata object and annotate the activity level of these pathways in each of the cells on the UMAP:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ifn_pathways</span> <span class="o">=</span> <span class="p">[</span>
    <span class="s2">&quot;REACTOME_INTERFERON_SIGNALING&quot;</span><span class="p">,</span>
    <span class="s2">&quot;REACTOME_INTERFERON_ALPHA_BETA_SIGNALING&quot;</span><span class="p">,</span>
    <span class="s2">&quot;REACTOME_INTERFERON_GAMMA_SIGNALING&quot;</span><span class="p">,</span>
<span class="p">]</span>

<span class="n">adata_</span><span class="o">.</span><span class="n">obs</span> <span class="o">=</span> <span class="n">aucell_scores</span><span class="p">[</span><span class="n">ifn_pathways</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;condition&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">label</span>
<span class="n">adata_</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">cell_type</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/tmp/ipykernel_88192/3278589065.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  adata_.obs[&#39;condition&#39;] = adata.obs.label
/tmp/ipykernel_88192/3278589065.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  adata_.obs[&#39;cell_type&#39;] = adata.obs.cell_type
</pre></div>
</div>
</div>
</div>
<p>Plot the scores on the umap</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span>
    <span class="n">adata_</span><span class="p">,</span>
    <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;condition&quot;</span><span class="p">,</span> <span class="s2">&quot;cell_type&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">ifn_pathways</span><span class="p">,</span>
    <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">ncols</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
    <span class="n">wspace</span><span class="o">=</span><span class="mf">0.3</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/gsea_pathway_77_0.png" src="../_images/gsea_pathway_77_0.png" />
</div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">AUCell</span></code> scores the pathways well-known to be implicated in interferon signalling high in IFN-stimulated cells, while cells in the control condition generally have low scores for these pathways, demonstrating that gene set scoring with <code class="docutils literal notranslate"><span class="pre">AUCell</span></code> has been successful. Also note that the scores are generally larger for terms that are ranked higher in the gene set enrichment test results by <code class="docutils literal notranslate"><span class="pre">fgsea</span></code>. The concordance between pathway activity scores by <code class="docutils literal notranslate"><span class="pre">AUCell</span></code> and gene set enrichment test by <code class="docutils literal notranslate"><span class="pre">fgsea</span></code> is promising, given that we know a priori that IFN-related pathways should be the top-ranked terms. In addition, the effect of IFN stimulation is very large in this dataset and this contributes to the performance of the methods here.</p>
</section>
<section id="gene-set-enrichment-for-complex-experimental-designs-using-limma-fry-and-pseudo-bulks">
<h3><span class="section-number">18.6.4. </span>Gene set enrichment for complex experimental designs using limma-fry and pseudo-bulks<a class="headerlink" href="#gene-set-enrichment-for-complex-experimental-designs-using-limma-fry-and-pseudo-bulks" title="Permalink to this headline">#</a></h3>
<p>In cluster-level t-test approach, differentially expressed genes are found by comparing a cluster to all other clusters, which in this case includes both control and stimulated cells. Linear models allow us to compare cells in the stimulated condition only to those in the control group, resulting in more accurate identification of genes responding to the stimulus. Indeed, linear models can accommodate for complex experimental designs, for example, identification of gene sets enriched in <code class="docutils literal notranslate"><span class="pre">Cell</span> <span class="pre">type</span> <span class="pre">A</span> <span class="pre">in</span> <span class="pre">treatment</span> <span class="pre">1</span></code> compared to <code class="docutils literal notranslate"><span class="pre">Cell</span> <span class="pre">type</span> <span class="pre">A</span> <span class="pre">in</span> <span class="pre">treatment</span> <span class="pre">2</span></code>; that is, across perturbation and across cell types effects, while adjusting for batch effects, between-individual variations, gender and strain differences in mouse models etc.</p>
<p>In the next section, we demonstrate a limma-fry workflow that generalize to realistic data analysis routines, say, for single-cell case control studies. We first create pseudo-bulk replicates per cell type and condition (3 replicates per condition - cell type combination). We then find gene sets enriched in stimulated compared to control cells in a cell type. We also assess gene set enrichment between two stimulated cell type populations to find differences in signalling pathways.</p>
<section id="create-pseudo-bulk-samples-and-explore-the-data">
<h4><span class="section-number">18.6.4.1. </span>Create pseudo-bulk samples and explore the data<a class="headerlink" href="#create-pseudo-bulk-samples-and-explore-the-data" title="Permalink to this headline">#</a></h4>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># adapted from DE chapter</span>


<span class="k">def</span> <span class="nf">aggregate_and_filter</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span>
    <span class="n">cell_identity</span><span class="p">,</span>
    <span class="n">sample_key</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">condition_key</span><span class="o">=</span><span class="s2">&quot;condition&quot;</span><span class="p">,</span>
    <span class="n">cell_identity_key</span><span class="o">=</span><span class="s2">&quot;cell_identity&quot;</span><span class="p">,</span>
    <span class="n">obs_to_keep</span><span class="o">=</span><span class="p">[],</span>  <span class="c1"># which additional metadata to keep, e.g. gender, age, etc.</span>
    <span class="n">replicates_per_sample</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span>
    <span class="n">NUM_OF_CELL_PER_SAMPLE</span><span class="o">=</span><span class="mi">75</span><span class="p">,</span>
<span class="p">):</span>

    <span class="c1"># subset adata to the given cell identity</span>
    <span class="n">adata_cell_pop</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">cell_identity_key</span><span class="p">]</span> <span class="o">==</span> <span class="n">cell_identity</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="c1"># re-filter for this cell population</span>
    <span class="c1">#     sc.pp.filter_genes(adata_cell_pop, min_cells=3) # leave this out for now</span>

    <span class="k">if</span> <span class="n">sample_key</span><span class="p">:</span>

        <span class="c1"># if cells are from different samples and/or patients, make replicates per sample - condition - cell type</span>
        <span class="c1"># no use of condition_key. may need updates</span>

        <span class="c1"># check which donors to keep according to the number of cells specified with NUM_OF_CELL_PER_DONOR</span>
        <span class="n">size_by_donor</span> <span class="o">=</span> <span class="n">adata_cell_pop</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="n">sample_key</span><span class="p">])</span><span class="o">.</span><span class="n">size</span><span class="p">()</span>
        <span class="n">donors_to_drop</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">donor</span>
            <span class="k">for</span> <span class="n">donor</span> <span class="ow">in</span> <span class="n">size_by_donor</span><span class="o">.</span><span class="n">index</span>
            <span class="k">if</span> <span class="n">size_by_donor</span><span class="p">[</span><span class="n">donor</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">NUM_OF_CELL_PER_SAMPLE</span>
        <span class="p">]</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="o">*</span><span class="n">adata_cell_pop</span><span class="o">.</span><span class="n">var_names</span><span class="p">,</span> <span class="o">*</span><span class="n">obs_to_keep</span><span class="p">])</span>

        <span class="n">adata_cell_pop</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">sample_key</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata_cell_pop</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">sample_key</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span>
            <span class="s2">&quot;category&quot;</span>
        <span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">donor</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span>
            <span class="n">donors</span> <span class="o">:=</span> <span class="n">adata_cell_pop</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">sample_key</span><span class="p">]</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">categories</span>
        <span class="p">):</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Processing donor </span><span class="si">{</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="si">}</span><span class="s2"> out of </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">donors</span><span class="p">)</span><span class="si">}</span><span class="s2">...&quot;</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\r</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">donor</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">donors_to_drop</span><span class="p">:</span>
                <span class="n">adata_donor</span> <span class="o">=</span> <span class="n">adata_cell_pop</span><span class="p">[</span><span class="n">adata_cell_pop</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">sample_key</span><span class="p">]</span> <span class="o">==</span> <span class="n">donor</span><span class="p">]</span>
                <span class="c1"># create replicates for each donor</span>
                <span class="n">indices</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">adata_donor</span><span class="o">.</span><span class="n">obs_names</span><span class="p">)</span>
                <span class="n">random</span><span class="o">.</span><span class="n">shuffle</span><span class="p">(</span><span class="n">indices</span><span class="p">)</span>
                <span class="n">indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array_split</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">indices</span><span class="p">),</span> <span class="n">replicates_per_patient</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">rep_idx</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">indices</span><span class="p">):</span>
                    <span class="n">adata_replicate</span> <span class="o">=</span> <span class="n">adata_donor</span><span class="p">[</span><span class="n">rep_idx</span><span class="p">]</span>
                    <span class="c1"># specify how to aggregate: sum gene expression for each gene for each donor and also keep the condition information</span>
                    <span class="n">agg_dict</span> <span class="o">=</span> <span class="p">{</span><span class="n">gene</span><span class="p">:</span> <span class="s2">&quot;sum&quot;</span> <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">adata_replicate</span><span class="o">.</span><span class="n">var_names</span><span class="p">}</span>
                    <span class="k">for</span> <span class="n">obs</span> <span class="ow">in</span> <span class="n">obs_to_keep</span><span class="p">:</span>
                        <span class="n">agg_dict</span><span class="p">[</span><span class="n">obs</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;first&quot;</span>
                    <span class="c1"># create a df with all genes, donor and condition info</span>
                    <span class="n">df_donor</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">adata_replicate</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">A</span><span class="p">)</span>
                    <span class="n">df_donor</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">adata_replicate</span><span class="o">.</span><span class="n">obs_names</span>
                    <span class="n">df_donor</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="n">adata_replicate</span><span class="o">.</span><span class="n">var_names</span>
                    <span class="n">df_donor</span> <span class="o">=</span> <span class="n">df_donor</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">adata_replicate</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">obs_to_keep</span><span class="p">])</span>
                    <span class="c1"># aggregate</span>
                    <span class="n">df_donor</span> <span class="o">=</span> <span class="n">df_donor</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">sample_key</span><span class="p">)</span><span class="o">.</span><span class="n">agg</span><span class="p">(</span><span class="n">agg_dict</span><span class="p">)</span>
                    <span class="n">df_donor</span><span class="p">[</span><span class="n">sample_key</span><span class="p">]</span> <span class="o">=</span> <span class="n">donor</span>
                    <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;donor_</span><span class="si">{</span><span class="n">donor</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">cell_identity</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df_donor</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">donor</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>

        <span class="c1"># if no samples and/or patients, make replicates per condition-cell type</span>

        <span class="n">size_by_donor</span> <span class="o">=</span> <span class="n">adata_cell_pop</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="n">condition_key</span><span class="p">])</span><span class="o">.</span><span class="n">size</span><span class="p">()</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="o">*</span><span class="n">adata_cell_pop</span><span class="o">.</span><span class="n">var_names</span><span class="p">,</span> <span class="o">*</span><span class="n">obs_to_keep</span><span class="p">])</span>

        <span class="n">adata_cell_pop</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">condition_key</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata_cell_pop</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">condition_key</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span>
            <span class="s2">&quot;category&quot;</span>
        <span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">donor</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span>
            <span class="n">donors</span> <span class="o">:=</span> <span class="n">adata_cell_pop</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">condition_key</span><span class="p">]</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">categories</span>
        <span class="p">):</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Processing condition </span><span class="si">{</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="si">}</span><span class="s2"> out of </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">donors</span><span class="p">)</span><span class="si">}</span><span class="s2">...&quot;</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\r</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">adata_donor</span> <span class="o">=</span> <span class="n">adata_cell_pop</span><span class="p">[</span><span class="n">adata_cell_pop</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">condition_key</span><span class="p">]</span> <span class="o">==</span> <span class="n">donor</span><span class="p">]</span>
            <span class="c1"># create replicates for each donor</span>
            <span class="n">indices</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">adata_donor</span><span class="o">.</span><span class="n">obs_names</span><span class="p">)</span>
            <span class="n">random</span><span class="o">.</span><span class="n">shuffle</span><span class="p">(</span><span class="n">indices</span><span class="p">)</span>
            <span class="n">indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array_split</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">indices</span><span class="p">),</span> <span class="n">replicates_per_sample</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">rep_idx</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">indices</span><span class="p">):</span>
                <span class="n">adata_replicate</span> <span class="o">=</span> <span class="n">adata_donor</span><span class="p">[</span><span class="n">rep_idx</span><span class="p">]</span>
                <span class="k">if</span> <span class="p">(</span>
                    <span class="n">adata_replicate</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">NUM_OF_CELL_PER_SAMPLE</span>
                <span class="p">):</span>  <span class="c1"># this enures mean 75 cells per replicate</span>
                    <span class="c1"># specify how to aggregate: sum gene expression for each gene for each donor and also keep the condition information</span>
                    <span class="n">agg_dict</span> <span class="o">=</span> <span class="p">{</span><span class="n">gene</span><span class="p">:</span> <span class="s2">&quot;sum&quot;</span> <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">adata_replicate</span><span class="o">.</span><span class="n">var_names</span><span class="p">}</span>
                    <span class="k">for</span> <span class="n">obs</span> <span class="ow">in</span> <span class="n">obs_to_keep</span><span class="p">:</span>
                        <span class="n">agg_dict</span><span class="p">[</span><span class="n">obs</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;first&quot;</span>
                    <span class="c1"># create a df with all genes, donor and condition info</span>
                    <span class="n">df_donor</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">adata_replicate</span><span class="o">.</span><span class="n">X</span><span class="p">)</span>
                    <span class="n">df_donor</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">adata_replicate</span><span class="o">.</span><span class="n">obs_names</span>
                    <span class="n">df_donor</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="n">adata_replicate</span><span class="o">.</span><span class="n">var_names</span>
                    <span class="n">df_donor</span> <span class="o">=</span> <span class="n">df_donor</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">adata_replicate</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">obs_to_keep</span><span class="p">])</span>
                    <span class="c1"># aggregate</span>
                    <span class="n">df_donor</span> <span class="o">=</span> <span class="n">df_donor</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">condition_key</span><span class="p">)</span><span class="o">.</span><span class="n">agg</span><span class="p">(</span><span class="n">agg_dict</span><span class="p">)</span>
                    <span class="n">df_donor</span><span class="p">[</span><span class="n">condition_key</span><span class="p">]</span> <span class="o">=</span> <span class="n">donor</span>
                    <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cell_identity</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">donor</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df_donor</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">donor</span><span class="p">]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">pass</span>

    <span class="c1"># create AnnData object from the df</span>
    <span class="n">adata_cell_pop</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span>
        <span class="n">df</span><span class="p">[</span><span class="n">adata_cell_pop</span><span class="o">.</span><span class="n">var_names</span><span class="p">],</span> <span class="n">obs</span><span class="o">=</span><span class="n">df</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="n">adata_cell_pop</span><span class="o">.</span><span class="n">var_names</span><span class="p">)</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">adata_cell_pop</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">obs_to_keep</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">,</span> <span class="s2">&quot;label&quot;</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 24673 × 15706
    obs: &#39;nCount_RNA&#39;, &#39;nFeature_RNA&#39;, &#39;tsne1&#39;, &#39;tsne2&#39;, &#39;label&#39;, &#39;cluster&#39;, &#39;cell_type&#39;, &#39;replicate&#39;, &#39;nCount_SCT&#39;, &#39;nFeature_SCT&#39;, &#39;integrated_snn_res.0.4&#39;, &#39;seurat_clusters&#39;, &#39;group&#39;
    var: &#39;name&#39;
    uns: &#39;log1p&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
    layers: &#39;counts&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ads</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">celltype</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">cell_type</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">categories</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Processing cell type </span><span class="si">{</span><span class="n">celltype</span><span class="si">}</span><span class="s2">...&quot;</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\r</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">ads</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
        <span class="n">aggregate_and_filter</span><span class="p">(</span>
            <span class="n">adata</span><span class="p">,</span>
            <span class="n">celltype</span><span class="p">,</span>
            <span class="n">condition_key</span><span class="o">=</span><span class="s2">&quot;label&quot;</span><span class="p">,</span>
            <span class="n">cell_identity_key</span><span class="o">=</span><span class="s2">&quot;cell_type&quot;</span><span class="p">,</span>
            <span class="n">obs_to_keep</span><span class="o">=</span><span class="n">obs_to_keep</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="p">)</span>
<span class="n">pb_data</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">ads</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Processing condition 2 out of 2....
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/tmp/ipykernel_88192/491905042.py:93: FutureWarning: X.dtype being converted to np.float32 from float64. In the next version of anndata (0.9) conversion will not be automatic. Pass dtype explicitly to avoid this warning. Pass `AnnData(X, dtype=X.dtype, ...)` to get the future behavour.
  adata_cell_pop = sc.AnnData(df[adata_cell_pop.var_names], obs=df.drop(columns=adata_cell_pop.var_names))
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Processing condition 2 out of 2...s......
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/tmp/ipykernel_88192/491905042.py:93: FutureWarning: X.dtype being converted to np.float32 from object. In the next version of anndata (0.9) conversion will not be automatic. Pass dtype explicitly to avoid this warning. Pass `AnnData(X, dtype=X.dtype, ...)` to get the future behavour.
  adata_cell_pop = sc.AnnData(df[adata_cell_pop.var_names], obs=df.drop(columns=adata_cell_pop.var_names))
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pb_data</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 42 × 15706
    obs: &#39;cell_type&#39;, &#39;label&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Does PC1 captures a meaningful biological or technical fact?</span>
<span class="n">pb_data</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;lib_size&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pb_data</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">pb_data</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">pb_data</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">pb_data</span><span class="o">.</span><span class="n">obs</span><span class="p">,</span> <span class="n">ncols</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">250</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/gsea_pathway_89_0.png" src="../_images/gsea_pathway_89_0.png" />
</div>
</div>
<p>The CD14 Monocyte and CD4 T cells form the two largest clusters in the data as evident in the UMAP representations. So, we observe that PC1 captures cluster size instead of biological variability (i.e. variation due to cell type, or stimulus effect) - that is the proportion and composition of cell types bias the observed variability in pseudo-bulk data, since the library size (sum of the reads per sample) is higher in more abundant cell types.</p>
<p>We adjust for differences in library sizes between pseudo-bulks by applying <span class="math notranslate nohighlight">\(\log_2\)</span> Counts-Per-Million (CPM) transformation, a common normalisation technique in bulk RNA-seq.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># we will need to work with log-CPMs for fry</span>
<span class="n">y</span> <span class="o">=</span> <span class="n">pb_data</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">T</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span> -i y
logCPM &lt;- edgeR::cpm(y, log=TRUE, prior.count = 2)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span> -o logCPM
logCPM = data.frame(logCPM)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pb_data</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s2">&quot;logCPM_allgenes&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">logCPM</span><span class="o">.</span><span class="n">T</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pb_data</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;logCPM_pca&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">logCPM</span><span class="o">.</span><span class="n">T</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">(),</span> <span class="n">return_info</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">embedding</span><span class="p">(</span><span class="n">pb_data</span><span class="p">,</span> <span class="s2">&quot;logCPM_pca&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">pb_data</span><span class="o">.</span><span class="n">obs</span><span class="p">,</span> <span class="n">ncols</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">250</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/gsea_pathway_96_0.png" src="../_images/gsea_pathway_96_0.png" />
</div>
</div>
<p>After adjusting for differences in library sizes, PC1 now captures difference between lymphoid (T, NK, B) and myeloid (Mono, DC) populations. This looks good! However, the variation due to administration of stimulus (i.e. difference between control and stimulated pseduo-replicates) is still not captured by any of the first two PCs. Ideally, the variation of interest has to be detectable in top few PCs of the pseudo-bulk data. In this case, since we are indeed interested in stimulation effect per cell type, we proceed to gene set testing. We re-iterate that the purpose of plotting PCs is to explore various axes of variability in the data and to spot unwanted variabilities that can substantial influence the test results. Users may proceed with the rest of the analyses should they be satisfied with the the variations in their data.</p>
<p>We first set up the design and contrast matrices. Let’s remind ourselves that a design matrix is a mathematical representation of group membership (i.e. the group or condition to which a sample belongs), and contrast matrices are mathematical representations of comparisons of interest for the differential test.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">groups</span> <span class="o">=</span> <span class="n">pb_data</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">label</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s2">&quot;string&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="n">pb_data</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">cell_type</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span> -i groups
group &lt;-  as.factor(gsub(&quot; |\\+&quot;,&quot;_&quot;, groups))
design &lt;- model.matrix(~ 0 + group)
head(design)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>     [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10] [,11] [,12] [,13] [,14]
[1,]    0    0    1    0    0    0    0    0    0     0     0     0     0     0
[2,]    0    0    1    0    0    0    0    0    0     0     0     0     0     0
[3,]    0    0    1    0    0    0    0    0    0     0     0     0     0     0
[4,]    0    0    0    0    0    0    0    0    0     1     0     0     0     0
[5,]    0    0    0    0    0    0    0    0    0     1     0     0     0     0
[6,]    0    0    0    0    0    0    0    0    0     1     0     0     0     0
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
colnames(design)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span> [1] &quot;groupctrl_B_cells&quot;           &quot;groupctrl_CD14__Monocytes&quot;  
 [3] &quot;groupctrl_CD4_T_cells&quot;       &quot;groupctrl_CD8_T_cells&quot;      
 [5] &quot;groupctrl_Dendritic_cells&quot;   &quot;groupctrl_FCGR3A__Monocytes&quot;
 [7] &quot;groupctrl_NK_cells&quot;          &quot;groupstim_B_cells&quot;          
 [9] &quot;groupstim_CD14__Monocytes&quot;   &quot;groupstim_CD4_T_cells&quot;      
[11] &quot;groupstim_CD8_T_cells&quot;       &quot;groupstim_Dendritic_cells&quot;  
[13] &quot;groupstim_FCGR3A__Monocytes&quot; &quot;groupstim_NK_cells&quot;         
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span> 
kang_pbmc_con &lt;- limma::makeContrasts(
    
    # the effect if stimulus in CD16 Monocyte cells
    groupstim_FCGR3A__Monocytes - groupctrl_FCGR3A__Monocytes,
    
    # the effect of stimulus in CD16 Monocytes compared to CD8 T Cells
    (groupstim_FCGR3A__Monocytes - groupctrl_FCGR3A__Monocytes) - (groupstim_CD8_T_cells - groupctrl_CD8_T_cells), 
    levels = design
)
</pre></div>
</div>
</div>
</div>
<p>Index the genes annotated in each pathway in our data as follows:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">var_names</span> <span class="o">=</span> <span class="n">pb_data</span><span class="o">.</span><span class="n">var_names</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span> -i var_names
rownames(logCPM) &lt;- var_names
idx &lt;- limma::ids2indices(pathways, rownames(logCPM))
</pre></div>
</div>
</div>
</div>
<p>As done in the <code class="docutils literal notranslate"><span class="pre">fgsea</span></code> method, let’s remove gene sets with less than 15 genes</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
keep_gs &lt;- lapply(idx, FUN=function(x) length(x) &gt;= 15)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
idx &lt;- idx[unlist(keep_gs)]
</pre></div>
</div>
</div>
</div>
<p>Now that we have set up the design and contrast matrices, and have indexed the genes in each pathway in our data, we can call <code class="docutils literal notranslate"><span class="pre">fry()</span></code> to test for enriched pathways in each of the contrasts we set above:</p>
</section>
<section id="fry-test-for-stimulated-vs-control">
<h4><span class="section-number">18.6.4.2. </span>fry test for Stimulated vs Control<a class="headerlink" href="#fry-test-for-stimulated-vs-control" title="Permalink to this headline">#</a></h4>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
res &lt;- limma::fry(logCPM, index = idx, design = design, contrast = kang_pbmc_con[,1])
</pre></div>
</div>
</div>
</div>
<p>Let’s print top few lines of fry’s output</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
NUM_CHARS = 80
# trim long pathway names
rownames(res) &lt;- sapply(rownames(res), FUN=function(x) ifelse(nchar(x) &gt; NUM_CHARS, substr(x,1, NUM_CHARS), x))
head(res)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>                                                                                 NGenes
REACTOME_EUKARYOTIC_TRANSLATION_INITIATION                                          116
REACTOME_ACTIVATION_OF_THE_MRNA_UPON_BINDING_OF_THE_CAP_BINDING_COMPLEX_AND_EIFS     58
REACTOME_SELENOAMINO_ACID_METABOLISM                                                102
REACTOME_METABOLISM_OF_RNA                                                          632
REACTOME_RESPIRATORY_ELECTRON_TRANSPORT_ATP_SYNTHESIS_BY_CHEMIOSMOTIC_COUPLING_A     93
REACTOME_RESPIRATORY_ELECTRON_TRANSPORT                                              90
                                                                                 Direction
REACTOME_EUKARYOTIC_TRANSLATION_INITIATION                                            Down
REACTOME_ACTIVATION_OF_THE_MRNA_UPON_BINDING_OF_THE_CAP_BINDING_COMPLEX_AND_EIFS      Down
REACTOME_SELENOAMINO_ACID_METABOLISM                                                  Down
REACTOME_METABOLISM_OF_RNA                                                            Down
REACTOME_RESPIRATORY_ELECTRON_TRANSPORT_ATP_SYNTHESIS_BY_CHEMIOSMOTIC_COUPLING_A      Down
REACTOME_RESPIRATORY_ELECTRON_TRANSPORT                                               Down
                                                                                       PValue
REACTOME_EUKARYOTIC_TRANSLATION_INITIATION                                       1.984338e-23
REACTOME_ACTIVATION_OF_THE_MRNA_UPON_BINDING_OF_THE_CAP_BINDING_COMPLEX_AND_EIFS 5.765406e-22
REACTOME_SELENOAMINO_ACID_METABOLISM                                             6.850612e-22
REACTOME_METABOLISM_OF_RNA                                                       1.216535e-21
REACTOME_RESPIRATORY_ELECTRON_TRANSPORT_ATP_SYNTHESIS_BY_CHEMIOSMOTIC_COUPLING_A 3.453102e-21
REACTOME_RESPIRATORY_ELECTRON_TRANSPORT                                          3.458450e-21
                                                                                          FDR
REACTOME_EUKARYOTIC_TRANSLATION_INITIATION                                       1.764076e-20
REACTOME_ACTIVATION_OF_THE_MRNA_UPON_BINDING_OF_THE_CAP_BINDING_COMPLEX_AND_EIFS 2.030065e-19
REACTOME_SELENOAMINO_ACID_METABOLISM                                             2.030065e-19
REACTOME_METABOLISM_OF_RNA                                                       2.703750e-19
REACTOME_RESPIRATORY_ELECTRON_TRANSPORT_ATP_SYNTHESIS_BY_CHEMIOSMOTIC_COUPLING_A 3.807220e-19
REACTOME_RESPIRATORY_ELECTRON_TRANSPORT                                          3.807220e-19
                                                                                 PValue.Mixed
REACTOME_EUKARYOTIC_TRANSLATION_INITIATION                                       1.519405e-43
REACTOME_ACTIVATION_OF_THE_MRNA_UPON_BINDING_OF_THE_CAP_BINDING_COMPLEX_AND_EIFS 9.580683e-38
REACTOME_SELENOAMINO_ACID_METABOLISM                                             1.571595e-45
REACTOME_METABOLISM_OF_RNA                                                       7.665100e-80
REACTOME_RESPIRATORY_ELECTRON_TRANSPORT_ATP_SYNTHESIS_BY_CHEMIOSMOTIC_COUPLING_A 5.037806e-55
REACTOME_RESPIRATORY_ELECTRON_TRANSPORT                                          1.407025e-54
                                                                                    FDR.Mixed
REACTOME_EUKARYOTIC_TRANSLATION_INITIATION                                       5.651678e-43
REACTOME_ACTIVATION_OF_THE_MRNA_UPON_BINDING_OF_THE_CAP_BINDING_COMPLEX_AND_EIFS 2.820274e-37
REACTOME_SELENOAMINO_ACID_METABOLISM                                             6.528728e-45
REACTOME_METABOLISM_OF_RNA                                                       2.620875e-78
REACTOME_RESPIRATORY_ELECTRON_TRANSPORT_ATP_SYNTHESIS_BY_CHEMIOSMOTIC_COUPLING_A 3.670991e-54
REACTOME_RESPIRATORY_ELECTRON_TRANSPORT                                          9.927343e-54
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
NUM_CHARS &lt;- 50
ggdat &lt;- res[1:20,]
ggdat$pathway &lt;- sapply(rownames(ggdat), FUN=function(x) ifelse(nchar(x) &gt; NUM_CHARS, substr(x,1, NUM_CHARS), x))
ggplot(ggdat, aes(x=reorder(pathway, -log10(FDR)), y = -log10(FDR))) +
    geom_bar(stat=&quot;identity&quot;, fill = &quot;#9370DB&quot;) + coord_flip() +
    xlab(&#39;Pathway&#39;) + theme_minimal()
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/gsea_pathway_113_0.png" src="../_images/gsea_pathway_113_0.png" />
</div>
</div>
</section>
<section id="fry-test-for-the-comparison-between-two-stimulated-cell-types">
<h4><span class="section-number">18.6.4.3. </span>fry test for the comparison between two stimulated cell types<a class="headerlink" href="#fry-test-for-the-comparison-between-two-stimulated-cell-types" title="Permalink to this headline">#</a></h4>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
res &lt;- limma::fry(logCPM, index = idx, design = design, contrast = kang_pbmc_con[,2])
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
ggdat &lt;- res[1:20,]
ggdat$pathway &lt;- sapply(rownames(ggdat), FUN=function(x) ifelse(nchar(x) &gt; NUM_CHARS, substr(x,1, NUM_CHARS), x))
ggplot(ggdat, aes(x=reorder(pathway, -log10(FDR)), y = -log10(FDR))) +
    geom_bar(stat=&quot;identity&quot;, fill = &quot;#8B008B&quot;) + coord_flip() +
    xlab(&#39;Pathway&#39;) + theme_minimal()
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/gsea_pathway_116_0.png" src="../_images/gsea_pathway_116_0.png" />
</div>
</div>
<p>As demonstrated above, limma-fry can accomodate gene set enrichment tests for datasets and research problems with complex experimental designs. Both <code class="docutils literal notranslate"><span class="pre">fgsea</span></code> and <code class="docutils literal notranslate"><span class="pre">fry</span></code> provide insights into the direction of enrichment (ES field in <code class="docutils literal notranslate"><span class="pre">fgsea</span></code> and Direction field in <code class="docutils literal notranslate"><span class="pre">fry</span></code>). They both can be applied to clusters of cells or pseudo-bulk samples. However,  Unlike <code class="docutils literal notranslate"><span class="pre">fgsea</span></code>, more flexible tests can be carried out with <code class="docutils literal notranslate"><span class="pre">fry</span></code>. In addition, <code class="docutils literal notranslate"><span class="pre">fry</span></code> can reveal if genes in a pathway are changing between the experimental conditions but in consistent or inconsistent directions. Pathways in which the genes change in consistent direction are identified with <code class="docutils literal notranslate"><span class="pre">FDR</span></code> &lt; 0.05. Pathways in which the genes are DE between the conditions but they change in different, inconsistent directions can be identified where <code class="docutils literal notranslate"><span class="pre">FDR</span></code> &gt; 0.05, but <code class="docutils literal notranslate"><span class="pre">FDR.Mixed</span></code> &lt; 0.05 (assuming 0.05 is the desired significance level). <code class="docutils literal notranslate"><span class="pre">fry</span></code> is bidirectional, applicable to arbitrary designs and works well with small number of samples (although this may not be an issue in single cell). Therefore, the results by <code class="docutils literal notranslate"><span class="pre">fry</span></code> might be of more interest biologically.</p>
<section id="on-the-effect-of-filtering-low-expression-genes">
<h5><span class="section-number">18.6.4.3.1. </span>On the effect of  filtering low-expression genes<a class="headerlink" href="#on-the-effect-of-filtering-low-expression-genes" title="Permalink to this headline">#</a></h5>
<p>As mentioned before, Ideally, the variation of interest has to be detectable in top few PCs of the pseudo-bulk data.
Let’s remove genes with low expression in the data, apply <span class="math notranslate nohighlight">\(\log_2\)</span>CPM transformation and repeat the PCA plots:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span> -i y
keep &lt;- edgeR::filterByExpr(y) # in real analysis, supply the desig matrix to the function to retain as more genes as possible
y &lt;- y[keep,]
logCPM &lt;- edgeR::cpm(y, log=TRUE, prior.count = 2)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>R[write to console]: No group or design set. Assuming all samples belong to one group.
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span> -o logCPM
logCPM = data.frame(logCPM)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pb_data</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s2">&quot;logCPM_FLE&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">logCPM</span><span class="o">.</span><span class="n">T</span>  <span class="c1"># FLE for filter low exprs</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pb_data</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;logCPM_FLE_pca&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">logCPM</span><span class="o">.</span><span class="n">T</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">(),</span> <span class="n">return_info</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">embedding</span><span class="p">(</span><span class="n">pb_data</span><span class="p">,</span> <span class="s2">&quot;logCPM_FLE_pca&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">pb_data</span><span class="o">.</span><span class="n">obs</span><span class="p">,</span> <span class="n">ncols</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">250</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/gsea_pathway_123_0.png" src="../_images/gsea_pathway_123_0.png" />
</div>
</div>
<p>Here, “logCPM_FLE” denotes filtering for low expressed genes followed by <span class="math notranslate nohighlight">\(\log_2\)</span>CPM transformation. We can now clearly observe that PC1 captures cell type effect and PC2 captures the treatment effect, when low-expressed genes are removed and differences between library sizes are adjusted by <span class="math notranslate nohighlight">\(\log_2\)</span>CPM transformation.</p>
<p>Since in this case study we are indeed interested in stimulation effect per cell type, and this variation is better preseved before gene filtering, we presented the enrichment test results on unfiltered data.
In practice, filtering low abundance genes and computation of normalisation factors by <code class="docutils literal notranslate"><span class="pre">edgeR::calcNormFactors</span></code> are standard part of bulk RNA-seq analysis workflow. Should we have been interested in global effects of IFN stimulation, we should have used the filtered data. Additionally, one can note that <code class="docutils literal notranslate"><span class="pre">design</span> <span class="pre">&lt;-</span> <span class="pre">model.matrix(~</span> <span class="pre">0</span> <span class="pre">+</span> <span class="pre">lineage</span> <span class="pre">+</span> <span class="pre">group)</span></code> would take into account differences (that is baseline expression differences) between myeloid and lymphoid lineages, improving the separation of pseudo-bulk samples by IFN stimulation, possibly along PC1.  In this case study, we were interested in cell type-specific effects, hence we stayed with a model of data whereby the variability along PC1 is by cell type. The choice of design matrix has to be carefully considered to align with the biological question of interest.</p>
</section>
<section id="a-note-on-the-redundancy-between-gene-sets-and-the-performance-of-preranked-and-fry-gene-set-tests">
<h5><span class="section-number">18.6.4.3.2. </span>A note on the redundancy between gene sets and the performance of preranked and fry gene set tests<a class="headerlink" href="#a-note-on-the-redundancy-between-gene-sets-and-the-performance-of-preranked-and-fry-gene-set-tests" title="Permalink to this headline">#</a></h5>
<p>Generally, there can be a large overlap between closely related gene sets. This overlap impacts the rank of the gene sets in the enrichment results and can compromise the final interpretation. For example, the cells in Kang et al. are treated with IFN-<span class="math notranslate nohighlight">\(\beta\)</span>. Therefore, one would expect to see the term REACTOME_INTERFERON_ALPHA_BETA_SIGNALING as the top ranked term. While this term is indeed the top rank term in the output of <code class="docutils literal notranslate"><span class="pre">fry</span></code>, in the output of <code class="docutils literal notranslate"><span class="pre">fgsea</span></code>  REACTOME_INTERFERON_SIGNALING is the top rank term. This term has a larger number of genes (52) compared to REACTOME_INTERFERON_ALPHA_BETA_SIGNALING (24 genes), and most of those genes are shared between the two terms. This illustrates another difference between preranked gene set tests such as <code class="docutils literal notranslate"><span class="pre">fgsea</span></code> and <code class="docutils literal notranslate"><span class="pre">fry</span></code>, in preventing the larger gene sets from dominating the enrichment results. The better performance of <code class="docutils literal notranslate"><span class="pre">fry</span></code> is due to more accurate estimation of gene expression variances, hence more sensitive DE gene results.</p>
</section>
</section>
</section>
</section>
<section id="key-takeaways">
<h2><span class="section-number">18.7. </span>Key Takeaways<a class="headerlink" href="#key-takeaways" title="Permalink to this headline">#</a></h2>
<ul class="simple">
<li><p>Normalise your data using standard scRNA-seq normalisation methods namely SCTransform and scran, and filter gene sets with low gene coverage in your data prior to pathway analysis.</p></li>
<li><p>Be aware of different types of gene set testing tests (i.e. competitive vs self-contained) and use one that suits your application.</p></li>
<li><p>Be aware of differences between gene set enrichment and gene set activity inference. <em>fgsea</em> is the widely used gene set test in single-cell studies; Pagoda 2 is found to outperform other pathway activity scoring tools. If your datasets has complex experimental design, consider pseudo-bulk analysis with gene set tests implemented in <em>limma</em>, as they are compatible with the linear model framework can additionally account for inter-gene correlations.</p></li>
</ul>
</section>
<section id="quiz">
<h2><span class="section-number">18.8. </span>Quiz<a class="headerlink" href="#quiz" title="Permalink to this headline">#</a></h2>
<ol class="simple">
<li><p>What is the difference between gene set enrichment tests and activity scoring?</p></li>
<li><p>Describe examples of settings where gene set tests should be used? Can you outline examples of settings where pathway activity scoring methods are applicable?</p></li>
<li><p>What are the two types of Null Hypotheses in gene set enrichment tests. Explain the difference between the two types.</p></li>
<li><p>What is the most important preprocessing step in pathway analysis? What are the consequences if it is not conducted properly?</p></li>
<li><p>Name one gene set testing and one gene set activity scoring algorithm and explain it briefly.</p></li>
</ol>
</section>
<section id="session-info">
<h2><span class="section-number">18.9. </span>Session info<a class="headerlink" href="#session-info" title="Permalink to this headline">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
sessionInfo()
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>R version 4.1.2 (2021-11-01)
Platform: x86_64-conda-linux-gnu (64-bit)
Running under: CentOS Linux 7 (Core)

Matrix products: default
BLAS/LAPACK: /mnt/home/icb/soroor.hediyeh-zadeh/miniconda3/envs/pathway/lib/libopenblasp-r0.3.18.so

locale:
 [1] LC_CTYPE=C                 LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats4    tools     stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] ggplot2_3.3.6               AUCell_1.16.0              
 [3] fgsea_1.20.0                SingleCellExperiment_1.16.0
 [5] SummarizedExperiment_1.24.0 Biobase_2.54.0             
 [7] GenomicRanges_1.46.1        GenomeInfoDb_1.30.0        
 [9] IRanges_2.28.0              S4Vectors_0.32.3           
[11] BiocGenerics_0.40.0         MatrixGenerics_1.6.0       
[13] matrixStats_0.62.0         

loaded via a namespace (and not attached):
 [1] locfit_1.5-9.5         Rcpp_1.0.8.3           lattice_0.20-45       
 [4] png_0.1-7              Biostrings_2.62.0      digest_0.6.29         
 [7] utf8_1.2.2             mime_0.12              R6_2.5.1              
[10] RSQLite_2.2.8          httr_1.4.3             pillar_1.7.0          
[13] zlibbioc_1.40.0        rlang_1.0.2            data.table_1.14.2     
[16] annotate_1.72.0        blob_1.2.3             R.oo_1.24.0           
[19] R.utils_2.11.0         Matrix_1.4-1           labeling_0.4.2        
[22] statmod_1.4.36         BiocParallel_1.28.3    RCurl_1.98-1.6        
[25] bit_4.0.4              munsell_0.5.0          shiny_1.7.1           
[28] DelayedArray_0.20.0    compiler_4.1.2         httpuv_1.6.5          
[31] pkgconfig_2.0.3        htmltools_0.5.2        KEGGREST_1.34.0       
[34] tibble_3.1.7           gridExtra_2.3          GenomeInfoDbData_1.2.7
[37] edgeR_3.36.0           XML_3.99-0.9           fansi_1.0.3           
[40] withr_2.5.0            crayon_1.5.1           later_1.2.0           
[43] R.methodsS3_1.8.1      bitops_1.0-7           grid_4.1.2            
[46] xtable_1.8-4           GSEABase_1.56.0        gtable_0.3.0          
[49] lifecycle_1.0.1        DBI_1.1.2              magrittr_2.0.3        
[52] scales_1.2.0           graph_1.72.0           cli_3.3.0             
[55] cachem_1.0.6           farver_2.1.0           XVector_0.34.0        
[58] promises_1.2.0.1       limma_3.50.1           ellipsis_0.3.2        
[61] vctrs_0.4.1            fastmatch_1.1-3        bit64_4.0.5           
[64] glue_1.6.2             parallel_4.1.2         fastmap_1.1.0         
[67] AnnotationDbi_1.56.1   colorspace_2.0-3       memoise_2.0.1         
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">session_info</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><details>
<summary>Click to view session information</summary>
<pre>
-----
anndata             0.8.0
anndata2ri          0.0.0
gdown               4.4.0
numpy               1.21.6
pandas              1.2.5
rpy2                3.5.1
scanpy              1.9.1
session_info        1.0.0
umap                0.5.2
-----
</pre>
<details>
<summary>Click to view modules imported as dependencies</summary>
<pre>
PIL                         9.0.1
asttokens                   NA
backcall                    0.2.0
beta_ufunc                  NA
binom_ufunc                 NA
brotli                      NA
bs4                         4.11.1
certifi                     2021.10.08
cffi                        1.15.0
charset_normalizer          2.0.4
colorama                    0.4.4
cycler                      0.10.0
cython_runtime              NA
dateutil                    2.8.2
debugpy                     1.5.1
decorator                   5.1.1
dunamai                     1.12.0
entrypoints                 0.4
executing                   0.8.3
filelock                    3.6.0
get_version                 3.5.4
h5py                        3.6.0
hypergeom_ufunc             NA
idna                        3.3
importlib_metadata          NA
ipykernel                   6.9.1
jedi                        0.18.1
jinja2                      3.0.3
joblib                      1.1.0
kiwisolver                  1.4.2
llvmlite                    0.38.0
markupsafe                  2.1.1
matplotlib                  3.5.1
matplotlib_inline           NA
mpl_toolkits                NA
natsort                     7.1.1
nbinom_ufunc                NA
numba                       0.55.1
packaging                   21.3
parso                       0.8.3
pexpect                     4.8.0
pickleshare                 0.7.5
pkg_resources               NA
prompt_toolkit              3.0.20
ptyprocess                  0.7.0
pure_eval                   0.2.2
pycparser                   2.21
pydev_ipython               NA
pydevconsole                NA
pydevd                      2.6.0
pydevd_concurrency_analyser NA
pydevd_file_utils           NA
pydevd_plugins              NA
pydevd_tracing              NA
pygments                    2.11.2
pynndescent                 0.5.4
pyparsing                   3.0.4
pytz                        2022.1
requests                    2.27.1
scipy                       1.8.0
setuptools                  61.2.0
six                         1.16.0
sklearn                     1.0.2
skmisc                      0.1.4
socks                       1.7.1
soupsieve                   2.3.1
sphinxcontrib               NA
stack_data                  0.2.0
statsmodels                 0.13.2
threadpoolctl               2.2.0
tornado                     6.1
tqdm                        4.64.0
traitlets                   5.1.1
typing_extensions           NA
tzlocal                     NA
urllib3                     1.26.9
wcwidth                     0.2.5
zipp                        NA
zmq                         22.3.0
</pre>
</details> <!-- seems like this ends pre, so might as well be explicit -->
<pre>
-----
IPython             8.2.0
jupyter_client      7.2.2
jupyter_core        4.9.2
-----
Python 3.8.12 | packaged by conda-forge | (default, Oct 12 2021, 21:59:51) [GCC 9.4.0]
Linux-3.10.0-1160.25.1.el7.x86_64-x86_64-with-glibc2.10
-----
Session information updated at 2022-06-10 15:00
</pre>
</details></div></div>
</div>
</section>
<section id="references">
<h2><span class="section-number">18.10. </span>References<a class="headerlink" href="#references" title="Permalink to this headline">#</a></h2>
<div class="docutils container" id="id24">
<dl class="citation">
<dt class="label" id="id318"><span class="brackets"><a class="fn-backref" href="#id10">gspaAGonzalezBM+17</a></span></dt>
<dd><p>Sara Aibar, Carmen Bravo González-Blas, Thomas Moerman, Vân Anh Huynh-Thu, Hana Imrichova, Gert Hulselmans, Florian Rambow, Jean-Christophe Marine, Pierre Geurts, Jan Aerts, and others. Scenic: single-cell regulatory network inference and clustering. <em>Nature methods</em>, 14(11):1083–1086, 2017.</p>
</dd>
<dt class="label" id="id335"><span class="brackets"><a class="fn-backref" href="#id19">gspaDG04</a></span></dt>
<dd><p>Doris Damian and Malka Gorfine. Statistical concerns about the gsea procedure. <em>Nature genetics</em>, 36(7):663–663, 2004.</p>
</dd>
<dt class="label" id="id312"><span class="brackets"><a class="fn-backref" href="#id9">gspaDJS+19</a></span></dt>
<dd><p>David DeTomaso, Matthew G Jones, Meena Subramaniam, Tal Ashuach, Chun J Ye, and Nir Yosef. Functional interpretation of single cell similarity maps. <em>Nature communications</em>, 10(1):1–11, 2019.</p>
</dd>
<dt class="label" id="id314"><span class="brackets"><a class="fn-backref" href="#id11">gspaFSL+16</a></span></dt>
<dd><p>Jean Fan, Neeraj Salathia, Rui Liu, Gwendolyn E Kaeser, Yun C Yung, Joseph L Herman, Fiona Kaper, Jian-Bing Fan, Kun Zhang, Jerold Chun, and others. Characterizing transcriptional heterogeneity through pathway and gene set overdispersion analysis. <em>Nature methods</em>, 13(3):241–244, 2016.</p>
</dd>
<dt class="label" id="id326"><span class="brackets"><a class="fn-backref" href="#id3">gspaFranzenGBjorkegren19</a></span></dt>
<dd><p>Oscar Franzén, Li-Ming Gan, and Johan LM Björkegren. Panglaodb: a web server for exploration of mouse and human single-cell rna sequencing data. <em>Database</em>, 2019.</p>
</dd>
<dt class="label" id="id321"><span class="brackets"><a class="fn-backref" href="#id13">gspaGAHI+19</a></span></dt>
<dd><p>Luz Garcia-Alonso, Christian H Holland, Mahmoud M Ibrahim, Denes Turei, and Julio Saez-Rodriguez. Benchmark and integration of resources for the estimation of human transcription factor activities. <em>Genome research</em>, 29(8):1363–1375, 2019.</p>
</dd>
<dt class="label" id="id317"><span class="brackets"><a class="fn-backref" href="#id4">gspaGBuhlmann07</a></span></dt>
<dd><p>Jelle J Goeman and Peter Bühlmann. Analyzing gene expression data in terms of gene sets: methodological issues. <em>Bioinformatics</em>, 23(8):980–987, 2007.</p>
</dd>
<dt class="label" id="id330"><span class="brackets"><a class="fn-backref" href="#id21">gspaHS19</a></span></dt>
<dd><p>Christoph Hafemeister and Rahul Satija. Normalization and variance stabilization of single-cell rna-seq data using regularized negative binomial regression. <em>Genome biology</em>, 20(1):1–15, 2019.</p>
</dd>
<dt class="label" id="id327"><span class="brackets">gspaHTPPaton+20</span><span class="fn-backref">(<a href="#id15">1</a>,<a href="#id18">2</a>)</span></dt>
<dd><p>Christian H Holland, Jovan Tanevski, Javier Perales-Patón, Jan Gleixner, Manu P Kumar, Elisabetta Mereu, Brian A Joughin, Oliver Stegle, Douglas A Lauffenburger, Holger Heyn, and others. Robustness and applicability of transcription factor and pathway analysis tools on single-cell rna-seq data. <em>Genome biology</em>, 21(1):1–19, 2020.</p>
</dd>
<dt class="label" id="id332"><span class="brackets"><a class="fn-backref" href="#id7">gspaHanzelmannCG13</a></span></dt>
<dd><p>Sonja Hänzelmann, Robert Castelo, and Justin Guinney. Gsva: gene set variation analysis for microarray and rna-seq data. <em>BMC bioinformatics</em>, 14(1):1–15, 2013.</p>
</dd>
<dt class="label" id="id333"><span class="brackets"><a class="fn-backref" href="#id23">gspaKST+18</a></span></dt>
<dd><p>Hyun Min Kang, Meena Subramaniam, Sasha Targ, Michelle Nguyen, Lenka Maliskova, Elizabeth McCarthy, Eunice Wan, Simon Wong, Lauren Byrnes, Cristina M Lanata, and others. Multiplexed droplet single-cell rna-sequencing using natural genetic variation. <em>Nature biotechnology</em>, 36(1):89–94, 2018.</p>
</dd>
<dt class="label" id="id313"><span class="brackets"><a class="fn-backref" href="#id5">gspaKSB+21</a></span></dt>
<dd><p>Gennady Korotkevich, Vladimir Sukhov, Nikolay Budin, Boris Shpak, Maxim N Artyomov, and Alexey Sergushichev. Fast gene set enrichment analysis. <em>BioRxiv</em>, pages 060012, 2021.</p>
</dd>
<dt class="label" id="id329"><span class="brackets"><a class="fn-backref" href="#id11">gspaLCS+18</a></span></dt>
<dd><p>Blue B Lake, Song Chen, Brandon C Sos, Jean Fan, Gwendolyn E Kaeser, Yun C Yung, Thu E Duong, Derek Gao, Jerold Chun, Peter V Kharchenko, and others. Integrative single-cell analysis of transcriptional and epigenetic states in the human adult brain. <em>Nature biotechnology</em>, 36(1):70–80, 2018.</p>
</dd>
<dt class="label" id="id328"><span class="brackets"><a class="fn-backref" href="#id12">gspaLCK+08</a></span></dt>
<dd><p>Eunjung Lee, Han-Yu Chuang, Jong-Won Kim, Trey Ideker, and Doheon Lee. Inferring pathway activity toward precise disease classification. <em>PLoS computational biology</em>, 4(11):e1000217, 2008.</p>
</dd>
<dt class="label" id="id324"><span class="brackets"><a class="fn-backref" href="#id1">gspaLSP+11</a></span></dt>
<dd><p>Arthur Liberzon, Aravind Subramanian, Reid Pinchback, Helga Thorvaldsdóttir, Pablo Tamayo, and Jill P Mesirov. Molecular signatures database (msigdb) 3.0. <em>Bioinformatics</em>, 27(12):1739–1740, 2011.</p>
</dd>
<dt class="label" id="id331"><span class="brackets"><a class="fn-backref" href="#id22">gspaLMM16</a></span></dt>
<dd><p>Aaron TL Lun, Davis J McCarthy, and John C Marioni. A step-by-step workflow for low-level analysis of single-cell rna-seq data with bioconductor. <em>F1000Research</em>, 2016.</p>
</dd>
<dt class="label" id="id320"><span class="brackets"><a class="fn-backref" href="#id14">gspaSKKlunemann+18</a></span></dt>
<dd><p>Michael Schubert, Bertram Klinger, Martina Klünemann, Anja Sieber, Florian Uhlitz, Sascha Sauer, Mathew J Garnett, Nils Blüthgen, and Julio Saez-Rodriguez. Perturbation-response genes reveal signaling footprints in cancer gene expression. <em>Nature communications</em>, 9(1):1–11, 2018.</p>
</dd>
<dt class="label" id="id334"><span class="brackets"><a class="fn-backref" href="#id8">gspaSmy05</a></span></dt>
<dd><p>Gordon K Smyth. Limma: linear models for microarray data. In <em>Bioinformatics and computational biology solutions using R and Bioconductor</em>, pages 397–420. Springer, 2005.</p>
</dd>
<dt class="label" id="id323"><span class="brackets">gspaSTM+05</span><span class="fn-backref">(<a href="#id1">1</a>,<a href="#id6">2</a>)</span></dt>
<dd><p>Aravind Subramanian, Pablo Tamayo, Vamsi K Mootha, Sayan Mukherjee, Benjamin L Ebert, Michael A Gillette, Amanda Paulovich, Scott L Pomeroy, Todd R Golub, Eric S Lander, and others. Gene set enrichment analysis: a knowledge-based approach for interpreting genome-wide expression profiles. <em>Proceedings of the National Academy of Sciences</em>, 102(43):15545–15550, 2005.</p>
</dd>
<dt class="label" id="id325"><span class="brackets"><a class="fn-backref" href="#id2">gspaZLX+19</a></span></dt>
<dd><p>Xinxin Zhang, Yujia Lan, Jinyuan Xu, Fei Quan, Erjie Zhao, Chunyu Deng, Tao Luo, Liwen Xu, Gaoming Liao, Min Yan, and others. Cellmarker: a manually curated resource of cell markers in human and mouse. <em>Nucleic acids research</em>, 47(D1):D721–D728, 2019.</p>
</dd>
<dt class="label" id="id316"><span class="brackets">gspaZMH+20</span><span class="fn-backref">(<a href="#id16">1</a>,<a href="#id17">2</a>,<a href="#id20">3</a>)</span></dt>
<dd><p>Yaru Zhang, Yunlong Ma, Yukuan Huang, Yan Zhang, Qi Jiang, Meng Zhou, and Jianzhong Su. Benchmarking algorithms for pathway activity transformation of single-cell rna-seq data. <em>Computational and structural biotechnology journal</em>, 18:2953–2961, 2020.</p>
</dd>
</dl>
</div>
</section>
<section id="contributors">
<h2><span class="section-number">18.11. </span>Contributors<a class="headerlink" href="#contributors" title="Permalink to this headline">#</a></h2>
<p>We gratefully acknowledge the contributions of:</p>
<section id="authors">
<h3><span class="section-number">18.11.1. </span>Authors<a class="headerlink" href="#authors" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p>Lukas Heumos</p></li>
<li><p>Anastasia Litinetskaya</p></li>
<li><p>Soroor Hediyeh-Zadeh</p></li>
</ul>
</section>
<section id="reviewers">
<h3><span class="section-number">18.11.2. </span>Reviewers<a class="headerlink" href="#reviewers" title="Permalink to this headline">#</a></h3>
</section>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./conditions"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
    <a class='left-prev' id="prev-link" href="compositional.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title"><span class="section-number">17. </span>Compositional analysis</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="perturbation_modeling.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title"><span class="section-number">19. </span>Perturbation modeling</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By Lukas Heumos, Anna Schaar<br/>
  
      &copy; Copyright 2022.<br/>
    <div class="extra_footer">
      <div>
  Brought to you by
  <a href="https://www.github.com/theislab">Theislab</a>, with many thanks to the single-cell community as a whole!
</div>

    </div>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>