
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>12. Data integration &#8212; Cross-modal single-cell analysis</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="../_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" href="../_static/styles/sphinx-book-theme.css?digest=5115cc725059bd94278eecd172e13a965bf8f5a9" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/book.css" />
    <link rel="stylesheet" type="text/css" href="../_static/design-style.b7bb847fb20b106c3d81b95245e65545.min.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?digest=9c920249402e914e316237a7dbc6769907cce411"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js"></script>
    <script async="async" kind="hypothesis" src="https://hypothes.is/embed.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../_static/book.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="13. Lineage tracing" href="../trajectories/lineage_tracing.html" />
    <link rel="prev" title="11. Clustering" href="clustering.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="en">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../_static/logo.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">Cross-modal single-cell analysis</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Introduction
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../introduction/prior_art.html">
   1. Prior art
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../introduction/experimental_data_collection.html">
   2. Introduction to single-cell RNA sequencing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../introduction/raw_data_processing.html">
   3. Raw data processing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../introduction/analysis_tools.html">
   4. Computational single-cell analysis frameworks and tools
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../introduction/data_infrastructure.html">
   5. Data infrastructure
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../introduction/interoperability.html">
   6. Interoperability
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Preprocessing and visualization
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../preprocessing_visualization/quality_control.html">
   7. Quality Control
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../preprocessing_visualization/normalization.html">
   8. Normalization
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../preprocessing_visualization/feature_selection.html">
   9. Feature selection
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../preprocessing_visualization/dimensionality_reduction.html">
   10. Dimensionality Reduction
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Identifying cellular structure
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="clustering.html">
   11. Clustering
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   12. Data integration
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Inferring trajectories
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../trajectories/lineage_tracing.html">
   13. Lineage tracing
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Dealing with conditions
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../conditions/differential_gene_expression.html">
   14. Differential gene expression analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../conditions/gsea_pathway.html">
   15. Gene set enrichment and pathway analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../conditions/perturbation_modeling.html">
   16. Perturbation modeling
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Deconvolution
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../deconvolution/motivation.html">
   17. Motivation
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Reproducibility
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../reproducibility/introduction.html">
   18. Reproducibility
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Outlook
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../outlook.html">
   19. Outlook
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Acknowledgements
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../acknowledgements.html">
   20. Acknowledgements
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Glossary
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../glossary.html">
   21. Glossary
  </a>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
            <div class="navbar_extra_footer">
            Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
            </div>
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right">
<div class="menu-dropdown menu-dropdown-launch-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Launch interactive content">
      <i class="fas fa-rocket"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://mybinder.org/v2/gh/theislab/extended-single-cell-tutorial/master?urlpath=tree/cellular_structure/integration.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Launch on Binder"
>
  

<span class="headerbtn__icon-container">
  
    <img src="../_static/images/logo_binder.svg">
  </span>
<span class="headerbtn__text-container">Binder</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

<button onclick="toggleFullScreen()"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="bottom"
title="Fullscreen mode"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>

<div class="menu-dropdown menu-dropdown-repository-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Source repositories">
      <i class="fab fa-github"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://github.com/theislab/extended-single-cell-tutorial"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Source repository"
>
  

<span class="headerbtn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="headerbtn__text-container">repository</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/theislab/extended-single-cell-tutorial/issues/new?title=Issue%20on%20page%20%2Fcellular_structure/integration.html&body=Your%20issue%20content%20here."
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Open an issue"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="headerbtn__text-container">open issue</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/theislab/extended-single-cell-tutorial/edit/master/cellular_structure/integration.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Edit this page"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="headerbtn__text-container">suggest edit</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

<div class="menu-dropdown menu-dropdown-download-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Download this page">
      <i class="fas fa-download"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="../_sources/cellular_structure/integration.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Download source file"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="headerbtn__text-container">.ipynb</span>
</a>

      </li>
      
      <li>
        
<button onclick="printPdf(this)"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="left"
title="Print to PDF"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="headerbtn__text-container">.pdf</span>
</button>

      </li>
      
    </ul>
  </div>
</div>
<label for="__page-toc"
  class="headerbtn headerbtn-page-toc"
  
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-list"></i>
  </span>

</label>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
    <div class="tocsection onthispage pt-5 pb-3">
        <i class="fas fa-list"></i> Contents
    </div>
    <nav id="bd-toc-nav" aria-label="Page">
        <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#dataset">
   12.1. Dataset
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#unintegrated">
   12.2. Unintegrated
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#batch-aware-feature-selection">
   12.3. Batch-aware feature selection
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#variational-autoencoder-based">
   12.4. Variational Autoencoder based
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#data-preparation">
     12.4.1. Data preparation
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#building-the-model">
     12.4.2. Building the model
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#training-the-model">
     12.4.3. Training the model
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#extracting-the-embedding">
     12.4.4. Extracting the embedding
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#calculate-corrected-umap">
     12.4.5. Calculate corrected UMAP
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#with-cell-labels">
   12.5. With cell labels
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#graph-based">
   12.6. Graph based
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#mutual-nearest-neighbors-based">
   12.7. Mutual nearest neighbors based
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#benchmarking-your-own-integration">
   12.8. Benchmarking your own integration
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#key-takeaways">
   12.9. Key Takeaways
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#quiz">
   12.10. Quiz
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#session-information">
   12.11. Session information
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#references">
   12.12. References
  </a>
 </li>
</ul>

    </nav>
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Data integration</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#dataset">
   12.1. Dataset
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#unintegrated">
   12.2. Unintegrated
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#batch-aware-feature-selection">
   12.3. Batch-aware feature selection
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#variational-autoencoder-based">
   12.4. Variational Autoencoder based
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#data-preparation">
     12.4.1. Data preparation
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#building-the-model">
     12.4.2. Building the model
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#training-the-model">
     12.4.3. Training the model
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#extracting-the-embedding">
     12.4.4. Extracting the embedding
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#calculate-corrected-umap">
     12.4.5. Calculate corrected UMAP
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#with-cell-labels">
   12.5. With cell labels
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#graph-based">
   12.6. Graph based
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#mutual-nearest-neighbors-based">
   12.7. Mutual nearest neighbors based
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#benchmarking-your-own-integration">
   12.8. Benchmarking your own integration
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#key-takeaways">
   12.9. Key Takeaways
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#quiz">
   12.10. Quiz
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#session-information">
   12.11. Session information
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#references">
   12.12. References
  </a>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                
  <section class="tex2jax_ignore mathjax_ignore" id="data-integration">
<h1><span class="section-number">12. </span>Data integration<a class="headerlink" href="#data-integration" title="Permalink to this headline">#</a></h1>
<p>A central challenge in most scRNA-seq data analyses is presented by batch effects. Batch effects are transcriptomic manifestations of handling cells in distinct groups or “batches”. The origins of these effects are diverse and difficult to pin down. Some batch effect sources might be technical such as differences in sample handling, experimental protocols, or sequencing depths, but also biological effects such as donor variation, tissues, or sampling location are often interpreted as a batch effect. Removing these effects is crucial to enable joint analysis that can focus on finding common structure in the data across batches. Often it is only after removing these effects that one can find rare cellular identities of which there are few per batch, and whose identification was previously obscured by differences between batches.</p>
<p>The removal of batch effects in data has previously been divided into two subtasks: batch correction and data integration <span id="id1">[<a class="reference internal" href="../introduction/prior_art.html#id34" title="Malte D Luecken and Fabian J Theis. Current best practices in single-cell rna-seq analysis: a tutorial. Molecular Systems Biology, 15(6):e8746, 2019. URL: https://www.embopress.org/doi/abs/10.15252/msb.20188746, arXiv:https://www.embopress.org/doi/pdf/10.15252/msb.20188746, doi:https://doi.org/10.15252/msb.20188746.">Luecken and Theis, 2019</a>]</span>. These subtasks differ in the complexity of the batch effect that must be removed. Batch correction methods deal with batch effects between samples in the same experiment where cell identity compositions are consistent, and the effect is quasi-linear. In contrast, data integration methods must deal with complex, often nested, batch effects between datasets often generated with different protocols where cell identities may not be shared across batches.</p>
<p>Approaches to remove batch effects in scRNA-seq data often involve three steps: dimensionality reduction, modeling and removing the batch effect, and projection back into a high-dimensional space.</p>
<p>Types of data integration methods</p>
<p>Benchmarks</p>
<p>scIB</p>
<p>Ocular atlas</p>
<p>Hemberg</p>
<p>Tran</p>
<p>Others</p>
<p>Different output types</p>
<p>Features</p>
<p>Embedding</p>
<p>Graph</p>
<p>General pros/cons</p>
<p>With existing labels</p>
<section id="dataset">
<h2><span class="section-number">12.1. </span>Dataset<a class="headerlink" href="#dataset" title="Permalink to this headline">#</a></h2>
<p>The dataset we will use to demonstrate data integration contains several samples of bone marrow mononuclear cells. These samples were originally created for the Open Problems in Single-Cell Analysis <a class="reference external" href="https://openproblems.bio/neurips_2021/">NeurIPS Competition 2021</a>. The 10x Multiome protocol was used which measures both RNA expression (scRNA-seq) and chromatin accessibility (scATAC-seq) in the same cells. The version we use here has been pre-processed as shown in previous chapters.</p>
<p><strong>!!! TODO !!!</strong> - Add instructions for getting the dataset and check the code for loading it is correct</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Python packages</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">scvi</span>
<span class="kn">import</span> <span class="nn">bbknn</span>
<span class="kn">import</span> <span class="nn">scib</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">matplotlib</span>

<span class="c1"># R interface</span>
<span class="kn">from</span> <span class="nn">rpy2.robjects</span> <span class="kn">import</span> <span class="n">pandas2ri</span>
<span class="kn">from</span> <span class="nn">rpy2.robjects</span> <span class="kn">import</span> <span class="n">r</span>
<span class="kn">import</span> <span class="nn">rpy2.rinterface_lib.callbacks</span>
<span class="kn">import</span> <span class="nn">anndata2ri</span>

<span class="n">pandas2ri</span><span class="o">.</span><span class="n">activate</span><span class="p">()</span>
<span class="n">anndata2ri</span><span class="o">.</span><span class="n">activate</span><span class="p">()</span>

<span class="o">%</span><span class="k">load_ext</span> rpy2.ipython
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
# R packages
library(Seurat)
</pre></div>
</div>
</div>
</div>
<p>Let’s read in the dataset using <strong>scanpy</strong> to get an <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> object.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s2">&quot;../../datasets/openproblems_bmmc_multiome_genes_filtered.h5ad&quot;</span><span class="p">)</span>
<span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;logcounts&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span>
<span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 69249 × 129921
    obs: &#39;GEX_pct_counts_mt&#39;, &#39;GEX_n_counts&#39;, &#39;GEX_n_genes&#39;, &#39;GEX_size_factors&#39;, &#39;GEX_phase&#39;, &#39;ATAC_nCount_peaks&#39;, &#39;ATAC_atac_fragments&#39;, &#39;ATAC_reads_in_peaks_frac&#39;, &#39;ATAC_blacklist_fraction&#39;, &#39;ATAC_nucleosome_signal&#39;, &#39;cell_type&#39;, &#39;batch&#39;, &#39;ATAC_pseudotime_order&#39;, &#39;GEX_pseudotime_order&#39;, &#39;Samplename&#39;, &#39;Site&#39;, &#39;DonorNumber&#39;, &#39;Modality&#39;, &#39;VendorLot&#39;, &#39;DonorID&#39;, &#39;DonorAge&#39;, &#39;DonorBMI&#39;, &#39;DonorBloodType&#39;, &#39;DonorRace&#39;, &#39;Ethnicity&#39;, &#39;DonorGender&#39;, &#39;QCMeds&#39;, &#39;DonorSmoker&#39;
    var: &#39;feature_types&#39;, &#39;gene_id&#39;
    uns: &#39;ATAC_gene_activity_var_names&#39;, &#39;dataset_id&#39;, &#39;genome&#39;, &#39;organism&#39;
    obsm: &#39;ATAC_gene_activity&#39;, &#39;ATAC_lsi_full&#39;, &#39;ATAC_lsi_red&#39;, &#39;ATAC_umap&#39;, &#39;GEX_X_pca&#39;, &#39;GEX_X_umap&#39;
    layers: &#39;counts&#39;, &#39;logcounts&#39;
</pre></div>
</div>
</div>
</div>
<p>The full dataset contains 69249 cells and measurements for 129921 features. There are two versions of the expression matrix, <code class="docutils literal notranslate"><span class="pre">counts</span></code> which contains the raw count values and <code class="docutils literal notranslate"><span class="pre">logcounts</span></code> which contains normalised log-counts (these values are also stored in <code class="docutils literal notranslate"><span class="pre">adata.X</span></code>).</p>
<p>The <code class="docutils literal notranslate"><span class="pre">obs</span></code> slot contains several variables, some of which were calculated during pre-processing (for quality control) and others that contain metadata about the samples. The ones we are interesed in here are:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">cell_type</span></code> - The annotated label for each cell</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">batch</span></code> - The sequencing batch for each cell</p></li>
</ul>
<p>For a real analysis it would be important to consider more variables but to keep it simple here we will only look at these.</p>
<p>We define variables to hold these names so that it is clear how we are using them in the code. This also helps with reproducibility because if we decided to change one of them for whatever reason we can be sure it has changed in the whole notebook.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">label_key</span> <span class="o">=</span> <span class="s2">&quot;cell_type&quot;</span>
<span class="n">batch_key</span> <span class="o">=</span> <span class="s2">&quot;batch&quot;</span>
</pre></div>
</div>
</div>
</div>
<p>Let’s have a look at the different batches and how many cells we have for each.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">batch_key</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>s4d8     9876
s4d1     8023
s3d10    6781
s1d2     6740
s1d1     6224
s2d4     6111
s2d5     4895
s3d3     4325
s4d9     4325
s1d3     4279
s2d1     4220
s3d7     1771
s3d6     1679
Name: batch, dtype: int64
</pre></div>
</div>
</div>
</div>
<p>There are 13 different different batches in the dataset. During this experiment multiple samples were taken from a set of donors and sequenced at different facilities so the names here are a combination of the sample number and the donor. For simplicity and to reduce computational time we will select four samples to use.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">keep_batches</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;s4d8&quot;</span><span class="p">,</span> <span class="s2">&quot;s1d1&quot;</span><span class="p">,</span> <span class="s2">&quot;s2d1&quot;</span><span class="p">,</span> <span class="s2">&quot;s3d7&quot;</span><span class="p">]</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">batch_key</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">keep_batches</span><span class="p">)]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 22091 × 129921
    obs: &#39;GEX_pct_counts_mt&#39;, &#39;GEX_n_counts&#39;, &#39;GEX_n_genes&#39;, &#39;GEX_size_factors&#39;, &#39;GEX_phase&#39;, &#39;ATAC_nCount_peaks&#39;, &#39;ATAC_atac_fragments&#39;, &#39;ATAC_reads_in_peaks_frac&#39;, &#39;ATAC_blacklist_fraction&#39;, &#39;ATAC_nucleosome_signal&#39;, &#39;cell_type&#39;, &#39;batch&#39;, &#39;ATAC_pseudotime_order&#39;, &#39;GEX_pseudotime_order&#39;, &#39;Samplename&#39;, &#39;Site&#39;, &#39;DonorNumber&#39;, &#39;Modality&#39;, &#39;VendorLot&#39;, &#39;DonorID&#39;, &#39;DonorAge&#39;, &#39;DonorBMI&#39;, &#39;DonorBloodType&#39;, &#39;DonorRace&#39;, &#39;Ethnicity&#39;, &#39;DonorGender&#39;, &#39;QCMeds&#39;, &#39;DonorSmoker&#39;
    var: &#39;feature_types&#39;, &#39;gene_id&#39;
    uns: &#39;ATAC_gene_activity_var_names&#39;, &#39;dataset_id&#39;, &#39;genome&#39;, &#39;organism&#39;
    obsm: &#39;ATAC_gene_activity&#39;, &#39;ATAC_lsi_full&#39;, &#39;ATAC_lsi_red&#39;, &#39;ATAC_umap&#39;, &#39;GEX_X_pca&#39;, &#39;GEX_X_umap&#39;
    layers: &#39;counts&#39;, &#39;logcounts&#39;
</pre></div>
</div>
</div>
</div>
<p>After subsetting to select these batches we are left with 22091 cells.</p>
<p>We have two annotations for the features store in <code class="docutils literal notranslate"><span class="pre">var</span></code>:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">feature_types</span></code> - The type of each feature</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">gene_id</span></code> - The gene associated with each feature</p></li>
</ul>
<p>Let’s have a look at the feature types.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;feature_types&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>ATAC    116490
GEX      13431
Name: feature_types, dtype: int64
</pre></div>
</div>
</div>
</div>
<p>We can see that there are over 100,000 ATAC features but only around 13,000 gene expression (“GEX”) features. Integration of multiple modalities is a complex problem that will be described in other chapters so for now we will subset to only the gene expression. We also perform a simple filtering to make sure we have no features will zero counts (this is necessary because by selecting a subset of samples we may have removed all the cells which expressed a particular feature).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[:,</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;feature_types&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;GEX&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_cells</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 22091 × 13431
    obs: &#39;GEX_pct_counts_mt&#39;, &#39;GEX_n_counts&#39;, &#39;GEX_n_genes&#39;, &#39;GEX_size_factors&#39;, &#39;GEX_phase&#39;, &#39;ATAC_nCount_peaks&#39;, &#39;ATAC_atac_fragments&#39;, &#39;ATAC_reads_in_peaks_frac&#39;, &#39;ATAC_blacklist_fraction&#39;, &#39;ATAC_nucleosome_signal&#39;, &#39;cell_type&#39;, &#39;batch&#39;, &#39;ATAC_pseudotime_order&#39;, &#39;GEX_pseudotime_order&#39;, &#39;Samplename&#39;, &#39;Site&#39;, &#39;DonorNumber&#39;, &#39;Modality&#39;, &#39;VendorLot&#39;, &#39;DonorID&#39;, &#39;DonorAge&#39;, &#39;DonorBMI&#39;, &#39;DonorBloodType&#39;, &#39;DonorRace&#39;, &#39;Ethnicity&#39;, &#39;DonorGender&#39;, &#39;QCMeds&#39;, &#39;DonorSmoker&#39;
    var: &#39;feature_types&#39;, &#39;gene_id&#39;, &#39;n_cells&#39;
    uns: &#39;ATAC_gene_activity_var_names&#39;, &#39;dataset_id&#39;, &#39;genome&#39;, &#39;organism&#39;
    obsm: &#39;ATAC_gene_activity&#39;, &#39;ATAC_lsi_full&#39;, &#39;ATAC_lsi_red&#39;, &#39;ATAC_umap&#39;, &#39;GEX_X_pca&#39;, &#39;GEX_X_umap&#39;
    layers: &#39;counts&#39;, &#39;logcounts&#39;
</pre></div>
</div>
</div>
</div>
<p>Because of the subsetting we also need to renormalise the data.</p>
<p><strong>!!! TODO !!!</strong> - Check if there is a better way to handle this. Found it was necessary to stop HVGs failing but better if we can reuse the existing values.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">X</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;counts&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_total</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;logcounts&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<p>We will use this dataset to demonstrate integration.</p>
<p>Most integration methods require a single object containing all the samples and a batch variable (like we have here). If instead you have separate objects for each of your samples you can join them using the <strong>anndata</strong> <code class="docutils literal notranslate"><span class="pre">concat()</span></code> function. See the the <a class="reference external" href="https://anndata.readthedocs.io/en/stable/concatenation.html">concatenation tutorial</a> for more details. Similar functionality exists in other ecosystems.</p>
<blockquote>
<div><p><strong>Integrating UMI and full-length data</strong></p>
<p>Integrating samples from UMI and full-length protocols can present additional challenges. This is because full-length protocols are affected by gene-length bias (longer genes will be more highly expressed) while UMI data is not <span id="id2">[<a class="reference internal" href="#id160" title="Belinda Phipson, Luke Zappia, and Alicia Oshlack. Gene length and detection bias in single cell RNA sequencing protocols. F1000Res., 6:595, April 2017.">Phipson <em>et al.</em>, 2017</a>]</span>.
Because of this it is generally recommended to transform counts for full-length samples into a unit which corrects for gene-length (such as transcripts per million (TPM)) before attempting integration.</p>
<p>This case applies to the samples we use here.
The celseq protocol produces full-length data but the values we use here are inferred UMI counts calculated by the authors of the original study.
This conversion process produces floating point numbers rather than integers which results in warnings in some of the following sections.</p>
</div></blockquote>
</section>
<section id="unintegrated">
<h2><span class="section-number">12.2. </span>Unintegrated<a class="headerlink" href="#unintegrated" title="Permalink to this headline">#</a></h2>
<p>It is always recommended to look at the raw data before performing any integration. This can give some indication of how big any batch effects are and what might be causing them. For some experiments it might even suggest that integration is not required if samples already overlap. This is not uncommon for mouse or cell line studies from a single lab for example where most of the variables which contribute to batch effects can be controlled.</p>
<p>We will perform highly variable gene (HVG) selection, PCA and UMAP dimensionality reduction as we have seen in previous chapters.</p>
<p><strong>!!! TODO !!!</strong> - Check this is done the same way as we decided was best practice in earlier chapters</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">highly_variable_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 22091 × 13431
    obs: &#39;GEX_pct_counts_mt&#39;, &#39;GEX_n_counts&#39;, &#39;GEX_n_genes&#39;, &#39;GEX_size_factors&#39;, &#39;GEX_phase&#39;, &#39;ATAC_nCount_peaks&#39;, &#39;ATAC_atac_fragments&#39;, &#39;ATAC_reads_in_peaks_frac&#39;, &#39;ATAC_blacklist_fraction&#39;, &#39;ATAC_nucleosome_signal&#39;, &#39;cell_type&#39;, &#39;batch&#39;, &#39;ATAC_pseudotime_order&#39;, &#39;GEX_pseudotime_order&#39;, &#39;Samplename&#39;, &#39;Site&#39;, &#39;DonorNumber&#39;, &#39;Modality&#39;, &#39;VendorLot&#39;, &#39;DonorID&#39;, &#39;DonorAge&#39;, &#39;DonorBMI&#39;, &#39;DonorBloodType&#39;, &#39;DonorRace&#39;, &#39;Ethnicity&#39;, &#39;DonorGender&#39;, &#39;QCMeds&#39;, &#39;DonorSmoker&#39;
    var: &#39;feature_types&#39;, &#39;gene_id&#39;, &#39;n_cells&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;
    uns: &#39;ATAC_gene_activity_var_names&#39;, &#39;dataset_id&#39;, &#39;genome&#39;, &#39;organism&#39;, &#39;log1p&#39;, &#39;hvg&#39;, &#39;pca&#39;, &#39;neighbors&#39;, &#39;umap&#39;
    obsm: &#39;ATAC_gene_activity&#39;, &#39;ATAC_lsi_full&#39;, &#39;ATAC_lsi_red&#39;, &#39;ATAC_umap&#39;, &#39;GEX_X_pca&#39;, &#39;GEX_X_umap&#39;, &#39;X_pca&#39;, &#39;X_umap&#39;
    varm: &#39;PCs&#39;
    layers: &#39;counts&#39;, &#39;logcounts&#39;
    obsp: &#39;distances&#39;, &#39;connectivities&#39;
</pre></div>
</div>
</div>
</div>
<p>This adds several new items to our AnnData object. The <code class="docutils literal notranslate"><span class="pre">var</span></code> slot now includes means, dispersions and the selected variable genes. In the <code class="docutils literal notranslate"><span class="pre">obsp</span></code> slot we have distances and connectivities for our KNN graph and in <code class="docutils literal notranslate"><span class="pre">obsm</span></code> are the PCA and UMAP embeddings.</p>
<p>Let’s plot the UMAP, colouring the points by cell identity and batch labels. If the dataset had not already been labelled (which is often the case) we would only be able to consider the batch labels. This is done by</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="n">batch_key</span> <span class="o">+</span> <span class="s2">&quot;_colors&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span>
    <span class="s2">&quot;#1b9e77&quot;</span><span class="p">,</span>
    <span class="s2">&quot;#d95f02&quot;</span><span class="p">,</span>
    <span class="s2">&quot;#7570b3&quot;</span><span class="p">,</span>
    <span class="s2">&quot;#e7298a&quot;</span><span class="p">,</span>
<span class="p">]</span>  <span class="c1"># Set custom colours for batches</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="n">label_key</span><span class="p">,</span> <span class="n">batch_key</span><span class="p">],</span> <span class="n">wspace</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/integration_27_0.png" src="../_images/integration_27_0.png" />
</div>
</div>
<p>Often when looking at these plots you will see clear separation between batches. In this case what we see is more subtle and while cells from the same label are generally near each other there is a shift between batchs. If we were to perform a clustering analysis using this raw data we would probably end up with some clusters containing a single batch which would be difficult to interpret at the annotation stage. We are also likely to overlook rare cell types which are not common enough in any single sample to produce their own cluster.</p>
<p>Now that we have confirmed there are batch effects to correct we can move on to the different integration methods. If the batched perfectly overlaid each other then there would be no need to perform integration.</p>
</section>
<section id="batch-aware-feature-selection">
<h2><span class="section-number">12.3. </span>Batch-aware feature selection<a class="headerlink" href="#batch-aware-feature-selection" title="Permalink to this headline">#</a></h2>
<p>As shown in previous chapters we often select a subset of genes to use for our analysis in order to reduce noise and processing time. We do the same thing when we have multiple samples however, it is important that gene selection is performed in a batch-aware way. This is because genes that are variable across the whole dataset could be capturing batch effects rather than the biological signals we are interested in.</p>
<p>We can perform batch aware highly variable gene selection by setting the <code class="docutils literal notranslate"><span class="pre">batch_key</span></code> argument. <strong>scanpy</strong> will then calculate HVGs for each batch separately and combine the results by selecting those genes that are highly variable in the most batches.</p>
<p><em><strong>!!! TODO !!!</strong></em> - Check which HVG method to use here</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">highly_variable_genes</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span> <span class="n">n_top_genes</span><span class="o">=</span><span class="mi">2000</span><span class="p">,</span> <span class="n">flavor</span><span class="o">=</span><span class="s2">&quot;cell_ranger&quot;</span><span class="p">,</span> <span class="n">batch_key</span><span class="o">=</span><span class="n">batch_key</span>
<span class="p">)</span>
<span class="n">adata</span>
<span class="n">adata</span><span class="o">.</span><span class="n">var</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/icb/luke.zappia/miniconda3/envs/bp-integration/lib/python3.8/site-packages/pandas/core/indexing.py:1637: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_block(indexer, value, name)
</pre></div>
</div>
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>feature_types</th>
      <th>gene_id</th>
      <th>n_cells</th>
      <th>highly_variable</th>
      <th>means</th>
      <th>dispersions</th>
      <th>dispersions_norm</th>
      <th>highly_variable_nbatches</th>
      <th>highly_variable_intersection</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>AL627309.5</th>
      <td>GEX</td>
      <td>ENSG00000241860</td>
      <td>188</td>
      <td>False</td>
      <td>0.004929</td>
      <td>0.673549</td>
      <td>0.007133</td>
      <td>1</td>
      <td>False</td>
    </tr>
    <tr>
      <th>LINC01409</th>
      <td>GEX</td>
      <td>ENSG00000237491</td>
      <td>764</td>
      <td>False</td>
      <td>0.018855</td>
      <td>0.646017</td>
      <td>-0.733194</td>
      <td>0</td>
      <td>False</td>
    </tr>
    <tr>
      <th>LINC01128</th>
      <td>GEX</td>
      <td>ENSG00000228794</td>
      <td>881</td>
      <td>False</td>
      <td>0.022610</td>
      <td>0.685406</td>
      <td>0.199297</td>
      <td>1</td>
      <td>False</td>
    </tr>
    <tr>
      <th>NOC2L</th>
      <td>GEX</td>
      <td>ENSG00000188976</td>
      <td>977</td>
      <td>False</td>
      <td>0.024214</td>
      <td>0.674026</td>
      <td>-0.391298</td>
      <td>0</td>
      <td>False</td>
    </tr>
    <tr>
      <th>KLHL17</th>
      <td>GEX</td>
      <td>ENSG00000187961</td>
      <td>145</td>
      <td>False</td>
      <td>0.003606</td>
      <td>0.653598</td>
      <td>-0.247074</td>
      <td>0</td>
      <td>False</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>MT-ND5</th>
      <td>GEX</td>
      <td>ENSG00000198786</td>
      <td>10920</td>
      <td>False</td>
      <td>0.369542</td>
      <td>0.583063</td>
      <td>-1.156446</td>
      <td>0</td>
      <td>False</td>
    </tr>
    <tr>
      <th>MT-ND6</th>
      <td>GEX</td>
      <td>ENSG00000198695</td>
      <td>3819</td>
      <td>False</td>
      <td>0.090493</td>
      <td>0.640347</td>
      <td>-0.343394</td>
      <td>0</td>
      <td>False</td>
    </tr>
    <tr>
      <th>MT-CYB</th>
      <td>GEX</td>
      <td>ENSG00000198727</td>
      <td>13859</td>
      <td>False</td>
      <td>0.549443</td>
      <td>0.558554</td>
      <td>-1.268154</td>
      <td>0</td>
      <td>False</td>
    </tr>
    <tr>
      <th>AL592183.1</th>
      <td>GEX</td>
      <td>ENSG00000273748</td>
      <td>1730</td>
      <td>False</td>
      <td>0.046654</td>
      <td>0.691756</td>
      <td>0.096679</td>
      <td>0</td>
      <td>False</td>
    </tr>
    <tr>
      <th>AC240274.1</th>
      <td>GEX</td>
      <td>ENSG00000271254</td>
      <td>159</td>
      <td>False</td>
      <td>0.003217</td>
      <td>0.499162</td>
      <td>-0.219673</td>
      <td>0</td>
      <td>False</td>
    </tr>
  </tbody>
</table>
<p>13431 rows × 9 columns</p>
</div></div></div>
</div>
<p>We can see there are now some addtional columns in <code class="docutils literal notranslate"><span class="pre">var</span></code>:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">highly_variable_nbatches</span></code> - The number of batches where each gene was found to be highly variable</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">highly_variable_intersection</span></code> - Whether each genes was highly variable in every batch</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">highly_variable</span></code> - Whether each genes was selected as highly variable after combining the results from each batch</p></li>
</ul>
<p>Let’s check how many batches each gene was variable in:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">n_batches</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;highly_variable_nbatches&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">n_batches</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">kind</span><span class="o">=</span><span class="s2">&quot;bar&quot;</span><span class="p">)</span>
<span class="n">n_batches</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>0    9404
1    1967
2     832
4     685
3     543
Name: highly_variable_nbatches, dtype: int64
</pre></div>
</div>
<img alt="../_images/integration_33_1.png" src="../_images/integration_33_1.png" />
</div>
</div>
<p>The first thing we notice is that most genes are not highly variable. This is typically the case but it can depend on how different the samples we are trying to integrate are. The overlap then decreases as we add more samples with relatively few genes being highly variable in all four batches. By selecting the top 2000 genes we have selected all HVGs that are present in three or four batches and most of those that are present in two batches.</p>
<blockquote>
<div><p><strong>How many genes to use?</strong></p>
<p>This is a question which doesn’t have a clear answer. The authors of the <strong>scvi-tools</strong> package which we use below recommend between 1000 and 10000 genes but how many depends on the context including the complexity of the dataset and the number of batches.
In general it is better to select slightly too many genes than to select too few and risk removing genes which are important for a rare cell type.
However, more genes will also increase the time required to run the integration methods.</p>
</div></blockquote>
<p>We will create an object with just the selected genes to use for integration.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_hvg</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[:,</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;highly_variable&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">adata_hvg</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 22091 × 2000
    obs: &#39;GEX_pct_counts_mt&#39;, &#39;GEX_n_counts&#39;, &#39;GEX_n_genes&#39;, &#39;GEX_size_factors&#39;, &#39;GEX_phase&#39;, &#39;ATAC_nCount_peaks&#39;, &#39;ATAC_atac_fragments&#39;, &#39;ATAC_reads_in_peaks_frac&#39;, &#39;ATAC_blacklist_fraction&#39;, &#39;ATAC_nucleosome_signal&#39;, &#39;cell_type&#39;, &#39;batch&#39;, &#39;ATAC_pseudotime_order&#39;, &#39;GEX_pseudotime_order&#39;, &#39;Samplename&#39;, &#39;Site&#39;, &#39;DonorNumber&#39;, &#39;Modality&#39;, &#39;VendorLot&#39;, &#39;DonorID&#39;, &#39;DonorAge&#39;, &#39;DonorBMI&#39;, &#39;DonorBloodType&#39;, &#39;DonorRace&#39;, &#39;Ethnicity&#39;, &#39;DonorGender&#39;, &#39;QCMeds&#39;, &#39;DonorSmoker&#39;
    var: &#39;feature_types&#39;, &#39;gene_id&#39;, &#39;n_cells&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable_nbatches&#39;, &#39;highly_variable_intersection&#39;
    uns: &#39;ATAC_gene_activity_var_names&#39;, &#39;dataset_id&#39;, &#39;genome&#39;, &#39;organism&#39;, &#39;log1p&#39;, &#39;hvg&#39;, &#39;pca&#39;, &#39;neighbors&#39;, &#39;umap&#39;, &#39;batch_colors&#39;, &#39;cell_type_colors&#39;
    obsm: &#39;ATAC_gene_activity&#39;, &#39;ATAC_lsi_full&#39;, &#39;ATAC_lsi_red&#39;, &#39;ATAC_umap&#39;, &#39;GEX_X_pca&#39;, &#39;GEX_X_umap&#39;, &#39;X_pca&#39;, &#39;X_umap&#39;
    varm: &#39;PCs&#39;
    layers: &#39;counts&#39;, &#39;logcounts&#39;
    obsp: &#39;distances&#39;, &#39;connectivities&#39;
</pre></div>
</div>
</div>
</div>
</section>
<section id="variational-autoencoder-based">
<h2><span class="section-number">12.4. </span>Variational Autoencoder based<a class="headerlink" href="#variational-autoencoder-based" title="Permalink to this headline">#</a></h2>
<p>The first integration method we will use is <strong>scVI</strong>, a method based on a conditional variational autoencoder. A <a class="reference external" href="https://towardsdatascience.com/understanding-variational-autoencoders-vaes-f70510919f73">variational autoencoder</a> is a type of artificial neural network which attempts to reduce the dimensionality of a dataset and reconstruct its distribution. The conditional part refers to explicitly considering differences between conditions (in this case batches) during this process in order to remove their effects. In benchmarking studies <strong>scVI</strong> has been shown to perform well across a range of datasets with a good balance of batch correction while conserving biological variability <span id="id3">[<a class="reference internal" href="#id158" title="Malte D. Luecken, M. Büttner, K. Chaichoompu, A. Danese, M. Interlandi, M. F. Mueller, D. C. Strobl, L. Zappia, M. Dugas, M. Colomé-Tatché, and Fabian J. Theis. Benchmarking atlas-level data integration in single-cell genomics. Nature Methods, 19(1):41-50, Jan 2022. URL: https://doi.org/10.1038/s41592-021-01336-8, doi:10.1038/s41592-021-01336-8.">Luecken <em>et al.</em>, 2022</a>]</span>. <strong>scVI</strong> models raw counts directly, so it is important that we provide it with a count matrix rather than a normalized expression matrix.</p>
<p>First let’s make a copy of our dataset to use for this integration. Normally it is not necessary to do this but as we will demonstrate multiple integration methods making a copy makes it easier to show what has been added by each method.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_scvi</span> <span class="o">=</span> <span class="n">adata_hvg</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<section id="data-preparation">
<h3><span class="section-number">12.4.1. </span>Data preparation<a class="headerlink" href="#data-preparation" title="Permalink to this headline">#</a></h3>
<p>The first step in using <strong>scVI</strong> is prepare our AnnData object. This step stores some information required by <strong>scVI</strong> such as which expression matrix to use and what the batch key is.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">scvi</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">SCVI</span><span class="o">.</span><span class="n">setup_anndata</span><span class="p">(</span><span class="n">adata_scvi</span><span class="p">,</span> <span class="n">layer</span><span class="o">=</span><span class="s2">&quot;counts&quot;</span><span class="p">,</span> <span class="n">batch_key</span><span class="o">=</span><span class="n">batch_key</span><span class="p">)</span>
<span class="n">adata_scvi</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 22091 × 2000
    obs: &#39;GEX_pct_counts_mt&#39;, &#39;GEX_n_counts&#39;, &#39;GEX_n_genes&#39;, &#39;GEX_size_factors&#39;, &#39;GEX_phase&#39;, &#39;ATAC_nCount_peaks&#39;, &#39;ATAC_atac_fragments&#39;, &#39;ATAC_reads_in_peaks_frac&#39;, &#39;ATAC_blacklist_fraction&#39;, &#39;ATAC_nucleosome_signal&#39;, &#39;cell_type&#39;, &#39;batch&#39;, &#39;ATAC_pseudotime_order&#39;, &#39;GEX_pseudotime_order&#39;, &#39;Samplename&#39;, &#39;Site&#39;, &#39;DonorNumber&#39;, &#39;Modality&#39;, &#39;VendorLot&#39;, &#39;DonorID&#39;, &#39;DonorAge&#39;, &#39;DonorBMI&#39;, &#39;DonorBloodType&#39;, &#39;DonorRace&#39;, &#39;Ethnicity&#39;, &#39;DonorGender&#39;, &#39;QCMeds&#39;, &#39;DonorSmoker&#39;, &#39;_scvi_batch&#39;, &#39;_scvi_labels&#39;
    var: &#39;feature_types&#39;, &#39;gene_id&#39;, &#39;n_cells&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable_nbatches&#39;, &#39;highly_variable_intersection&#39;
    uns: &#39;ATAC_gene_activity_var_names&#39;, &#39;dataset_id&#39;, &#39;genome&#39;, &#39;organism&#39;, &#39;log1p&#39;, &#39;hvg&#39;, &#39;pca&#39;, &#39;neighbors&#39;, &#39;umap&#39;, &#39;batch_colors&#39;, &#39;cell_type_colors&#39;, &#39;_scvi_uuid&#39;, &#39;_scvi_manager_uuid&#39;
    obsm: &#39;ATAC_gene_activity&#39;, &#39;ATAC_lsi_full&#39;, &#39;ATAC_lsi_red&#39;, &#39;ATAC_umap&#39;, &#39;GEX_X_pca&#39;, &#39;GEX_X_umap&#39;, &#39;X_pca&#39;, &#39;X_umap&#39;
    varm: &#39;PCs&#39;
    layers: &#39;counts&#39;, &#39;logcounts&#39;
    obsp: &#39;distances&#39;, &#39;connectivities&#39;
</pre></div>
</div>
</div>
</div>
<p>The fields created by <strong>scVI</strong> are prefixed with <code class="docutils literal notranslate"><span class="pre">_scvi</span></code>. These are designed for internal use and should not be manually modified. The general advice from <strong>scVI</strong> is that we should not to make any changes to our object until after the model is trained. On other datasets you may see a warning about <code class="docutils literal notranslate"><span class="pre">adata.layers[&quot;counts&quot;]</span></code> containing unnormalised count data. This usually means you should check that the layer provided to the setup function does actually contain count values but it can also happen if you have values from performing gene length correction on data from a full-lenght protocol or from another quantification method that does not produce integer counts.</p>
</section>
<section id="building-the-model">
<h3><span class="section-number">12.4.2. </span>Building the model<a class="headerlink" href="#building-the-model" title="Permalink to this headline">#</a></h3>
<p>We can now construct an <strong>scVI</strong> model object. As well as the <strong>scVI</strong> model we use here the <strong>scvi-tools</strong> package contains various other models (we will use the <strong>scANVI</strong> model below).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">model_scvi</span> <span class="o">=</span> <span class="n">scvi</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">SCVI</span><span class="p">(</span><span class="n">adata_scvi</span><span class="p">)</span>
<span class="n">model_scvi</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">SCVI Model with the following params: 
n_hidden: <span style="color: #008080; text-decoration-color: #008080; font-weight: bold">128</span>, n_latent: <span style="color: #008080; text-decoration-color: #008080; font-weight: bold">10</span>, n_layers: <span style="color: #008080; text-decoration-color: #008080; font-weight: bold">1</span>, dropout_rate: <span style="color: #008080; text-decoration-color: #008080; font-weight: bold">0.1</span>, dispersion: gene, 
gene_likelihood: zinb, latent_distribution: normal
Training status: Not Trained
</pre>
</div><div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
</div>
</div>
<p>The model object contains the provided AnnData object as well as the neural network for the model itself. You can see that currently the model is not trained. If we wanted to modify the structure of the network we could provide additional arguments to the model construction function but here we just use the defaults.</p>
<p>We can also print a more detailed description of the model that show us where things are stored in the associated AnnData object.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">model_scvi</span><span class="o">.</span><span class="n">view_anndata_setup</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">Anndata setup with scvi-tools version <span style="color: #008080; text-decoration-color: #008080; font-weight: bold">0.16</span>.<span style="color: #008080; text-decoration-color: #008080; font-weight: bold">4</span>.
</pre>
</div><div class="output text_html"><pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">
</pre>
</div><div class="output text_html"><pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">Setup via `SCVI.setup_anndata` with arguments:
</pre>
</div><div class="output text_html"><pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"><span style="font-weight: bold">{</span>
<span style="color: #7fbf7f; text-decoration-color: #7fbf7f">│   </span><span style="color: #008000; text-decoration-color: #008000">'layer'</span>: <span style="color: #008000; text-decoration-color: #008000">'counts'</span>,
<span style="color: #7fbf7f; text-decoration-color: #7fbf7f">│   </span><span style="color: #008000; text-decoration-color: #008000">'batch_key'</span>: <span style="color: #008000; text-decoration-color: #008000">'batch'</span>,
<span style="color: #7fbf7f; text-decoration-color: #7fbf7f">│   </span><span style="color: #008000; text-decoration-color: #008000">'labels_key'</span>: <span style="color: #800080; text-decoration-color: #800080; font-style: italic">None</span>,
<span style="color: #7fbf7f; text-decoration-color: #7fbf7f">│   </span><span style="color: #008000; text-decoration-color: #008000">'size_factor_key'</span>: <span style="color: #800080; text-decoration-color: #800080; font-style: italic">None</span>,
<span style="color: #7fbf7f; text-decoration-color: #7fbf7f">│   </span><span style="color: #008000; text-decoration-color: #008000">'categorical_covariate_keys'</span>: <span style="color: #800080; text-decoration-color: #800080; font-style: italic">None</span>,
<span style="color: #7fbf7f; text-decoration-color: #7fbf7f">│   </span><span style="color: #008000; text-decoration-color: #008000">'continuous_covariate_keys'</span>: <span style="color: #800080; text-decoration-color: #800080; font-style: italic">None</span>
<span style="font-weight: bold">}</span>
</pre>
</div><div class="output text_html"><pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">
</pre>
</div><div class="output text_html"><pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"><span style="font-style: italic">         Summary Statistics         </span>
┏━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━┓
┃<span style="font-weight: bold">     Summary Stat Key     </span>┃<span style="font-weight: bold"> Value </span>┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━┩
│<span style="color: #0087ff; text-decoration-color: #0087ff">         n_cells          </span>│<span style="color: #af00d7; text-decoration-color: #af00d7"> 22091 </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">          n_vars          </span>│<span style="color: #af00d7; text-decoration-color: #af00d7"> 2000  </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">         n_batch          </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">   4   </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">         n_labels         </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">   1   </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff"> n_extra_categorical_covs </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">   0   </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff"> n_extra_continuous_covs  </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">   0   </span>│
└──────────────────────────┴───────┘
</pre>
</div><div class="output text_html"><pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"><span style="font-style: italic">               Data Registry                </span>
┏━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃<span style="font-weight: bold"> Registry Key </span>┃<span style="font-weight: bold">    scvi-tools Location    </span>┃
┡━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│<span style="color: #0087ff; text-decoration-color: #0087ff">      X       </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">  adata.layers['counts']   </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">    batch     </span>│<span style="color: #af00d7; text-decoration-color: #af00d7"> adata.obs['_scvi_batch']  </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">    labels    </span>│<span style="color: #af00d7; text-decoration-color: #af00d7"> adata.obs['_scvi_labels'] </span>│
└──────────────┴───────────────────────────┘
</pre>
</div><div class="output text_html"><pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"><span style="font-style: italic">                  batch State Registry                   </span>
┏━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━┓
┃<span style="font-weight: bold">  Source Location   </span>┃<span style="font-weight: bold"> Categories </span>┃<span style="font-weight: bold"> scvi-tools Encoding </span>┃
┡━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━┩
│<span style="color: #0087ff; text-decoration-color: #0087ff"> adata.obs['batch'] </span>│<span style="color: #008000; text-decoration-color: #008000">    s1d1    </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">          0          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                    </span>│<span style="color: #008000; text-decoration-color: #008000">    s2d1    </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">          1          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                    </span>│<span style="color: #008000; text-decoration-color: #008000">    s3d7    </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">          2          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                    </span>│<span style="color: #008000; text-decoration-color: #008000">    s4d8    </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">          3          </span>│
└────────────────────┴────────────┴─────────────────────┘
</pre>
</div><div class="output text_html"><pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"><span style="font-style: italic">                     labels State Registry                      </span>
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━┓
┃<span style="font-weight: bold">      Source Location      </span>┃<span style="font-weight: bold"> Categories </span>┃<span style="font-weight: bold"> scvi-tools Encoding </span>┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━┩
│<span style="color: #0087ff; text-decoration-color: #0087ff"> adata.obs['_scvi_labels'] </span>│<span style="color: #008000; text-decoration-color: #008000">     0      </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">          0          </span>│
└───────────────────────────┴────────────┴─────────────────────┘
</pre>
</div></div>
</div>
<p>Here we can see exactly what information has been assigned by <strong>scVI</strong> and even details like how each different batch is encoded in the model.</p>
</section>
<section id="training-the-model">
<h3><span class="section-number">12.4.3. </span>Training the model<a class="headerlink" href="#training-the-model" title="Permalink to this headline">#</a></h3>
<p>The model will be trained for a given number of <em>epochs</em> (a training iteration where every cell is passed through the network). The following code is a suggested heuristic for reducing the number of epochs for larger datasets. Since this dataset has slightly more than 20000 cells we get a value slightly below the default value of 400.</p>
<p><strong>!!! TODO !!!</strong> - Check this is still the recommended thing to do with new scvi-tools versions</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">max_epochs_scvi</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">min</span><span class="p">([</span><span class="nb">round</span><span class="p">((</span><span class="mi">20000</span> <span class="o">/</span> <span class="n">adata</span><span class="o">.</span><span class="n">n_obs</span><span class="p">)</span> <span class="o">*</span> <span class="mi">400</span><span class="p">),</span> <span class="mi">400</span><span class="p">])</span>
<span class="n">max_epochs_scvi</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>362
</pre></div>
</div>
</div>
</div>
<p>We now train the model for the selected number of epochs (this will take ~30-60 minutes depending on the computer you are using).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">model_scvi</span><span class="o">.</span><span class="n">train</span><span class="p">(</span><span class="n">max_epochs</span><span class="o">=</span><span class="n">max_epochs_scvi</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/icb/luke.zappia/miniconda3/envs/bp-integration/lib/python3.8/site-packages/torchmetrics/utilities/prints.py:36: UserWarning: Torchmetrics v0.9 introduced a new argument class property called `full_state_update` that has
                not been set for this class (ElboMetric). The property determines if `update` by
                default needs access to the full metric state. If this is not the case, significant speedups can be
                achieved and we recommend setting this to `False`.
                We provide an checking function
                `from torchmetrics.utilities import check_forward_no_full_state`
                that can be used to check if the `full_state_update=True` (old and potential slower behaviour,
                default for now) or if `full_state_update=False` can be used safely.
                
  warnings.warn(*args, **kwargs)
GPU available: False, used: False
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
Set SLURM handle signals.
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 362/362: 100%|██████████| 362/362 [44:26&lt;00:00,  7.37s/it, loss=535, v_num=1]
</pre></div>
</div>
</div>
</div>
<blockquote>
<div><p><strong>Early stopping</strong></p>
<p>An alternative to manually setting the number of epochs is to set <code class="docutils literal notranslate"><span class="pre">early_stopping=True</span></code> in the training function. This will let <strong>scVI</strong> decide to stop training early depending on the convergence of the model.
The exact conditions for stopping can be controlled by other parameters.</p>
</div></blockquote>
</section>
<section id="extracting-the-embedding">
<h3><span class="section-number">12.4.4. </span>Extracting the embedding<a class="headerlink" href="#extracting-the-embedding" title="Permalink to this headline">#</a></h3>
<p>The main result we want to extract from the trained model is the latent representation for each cell. This is an embedding where the batch effects have been removed. We store this in <code class="docutils literal notranslate"><span class="pre">obsm</span></code> with the key <code class="docutils literal notranslate"><span class="pre">X_scvi</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_scvi</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;X_scVI&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">model_scvi</span><span class="o">.</span><span class="n">get_latent_representation</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="calculate-corrected-umap">
<h3><span class="section-number">12.4.5. </span>Calculate corrected UMAP<a class="headerlink" href="#calculate-corrected-umap" title="Permalink to this headline">#</a></h3>
<p>We will now visualise the data as we did before integrated. We calculate a new UMAP embedding but instead of finding nearest neighbours in PCA space we start with the corrected representation from <strong>scVI</strong>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata_scvi</span><span class="p">,</span> <span class="n">use_rep</span><span class="o">=</span><span class="s2">&quot;X_scVI&quot;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_scvi</span><span class="p">)</span>
<span class="n">adata_scvi</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 22091 × 2000
    obs: &#39;GEX_pct_counts_mt&#39;, &#39;GEX_n_counts&#39;, &#39;GEX_n_genes&#39;, &#39;GEX_size_factors&#39;, &#39;GEX_phase&#39;, &#39;ATAC_nCount_peaks&#39;, &#39;ATAC_atac_fragments&#39;, &#39;ATAC_reads_in_peaks_frac&#39;, &#39;ATAC_blacklist_fraction&#39;, &#39;ATAC_nucleosome_signal&#39;, &#39;cell_type&#39;, &#39;batch&#39;, &#39;ATAC_pseudotime_order&#39;, &#39;GEX_pseudotime_order&#39;, &#39;Samplename&#39;, &#39;Site&#39;, &#39;DonorNumber&#39;, &#39;Modality&#39;, &#39;VendorLot&#39;, &#39;DonorID&#39;, &#39;DonorAge&#39;, &#39;DonorBMI&#39;, &#39;DonorBloodType&#39;, &#39;DonorRace&#39;, &#39;Ethnicity&#39;, &#39;DonorGender&#39;, &#39;QCMeds&#39;, &#39;DonorSmoker&#39;, &#39;_scvi_batch&#39;, &#39;_scvi_labels&#39;
    var: &#39;feature_types&#39;, &#39;gene_id&#39;, &#39;n_cells&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable_nbatches&#39;, &#39;highly_variable_intersection&#39;
    uns: &#39;ATAC_gene_activity_var_names&#39;, &#39;dataset_id&#39;, &#39;genome&#39;, &#39;organism&#39;, &#39;log1p&#39;, &#39;hvg&#39;, &#39;pca&#39;, &#39;neighbors&#39;, &#39;umap&#39;, &#39;batch_colors&#39;, &#39;cell_type_colors&#39;, &#39;_scvi_uuid&#39;, &#39;_scvi_manager_uuid&#39;
    obsm: &#39;ATAC_gene_activity&#39;, &#39;ATAC_lsi_full&#39;, &#39;ATAC_lsi_red&#39;, &#39;ATAC_umap&#39;, &#39;GEX_X_pca&#39;, &#39;GEX_X_umap&#39;, &#39;X_pca&#39;, &#39;X_umap&#39;, &#39;X_scVI&#39;
    varm: &#39;PCs&#39;
    layers: &#39;counts&#39;, &#39;logcounts&#39;
    obsp: &#39;distances&#39;, &#39;connectivities&#39;
</pre></div>
</div>
</div>
</div>
<p>Once we have the new UMAP representation we can plot it coloured by batch and identity labels as before.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_scvi</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="n">label_key</span><span class="p">,</span> <span class="n">batch_key</span><span class="p">],</span> <span class="n">wspace</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/integration_64_0.png" src="../_images/integration_64_0.png" />
</div>
</div>
<p>This looks better! Before, the various batches were shifted apart from each other. Now, the batches are more overlapping and we have a single blob for each cell identity label.</p>
<p>In many cases we would not already have identity labels so from this stage we would continue with clustering, annotation and further analysis as described in other sections.</p>
</section>
</section>
<section id="with-cell-labels">
<h2><span class="section-number">12.5. </span>With cell labels<a class="headerlink" href="#with-cell-labels" title="Permalink to this headline">#</a></h2>
<p>When performing integration with <strong>scVI</strong> we pretended that we didn’t already have any cell labels (although we showed them in plots). While this scenario is common there are some cases where we do know something about cell identity in advance. Most often this is when we want to combine one or more publicly available datasets with data from a new study. When we have labels for at least some of the cells we can use <strong>scANVI</strong>. This is an extension of the <strong>scVI</strong> model that knows about cell labels as well as batches. Because it has this extra information it can try to keep the differences between cell labels while removing batch effects. Benchmarking suggests that <strong>scANVI</strong> tends to better preserve biological signals compared to <strong>scVI</strong> but sometimes it is not as effective at removing batch effects. While we have labels for all cells here it is also possible to use <strong>scANVI</strong> in a semi-supervised manner where labels are only provided for some cells.</p>
<blockquote>
<div><p><strong>Label harmonization</strong></p>
<p>If you are using <strong>scANVI</strong> to integrate multiple datasets for which you already have labels it is important to first perform <em>label harmonization</em>.
This refers to a process of checking that labels are consistent across the datasets that are being integrated.
How best to do this is an open question but often requires input from subject-matter experts.</p>
</div></blockquote>
<p>We start by creating a <strong>scANVI</strong> model object. Note that because <strong>scANVI</strong> refines an <strong>scVI</strong> model we provide that rather than an AnnData object. If we had not already trained an <strong>scVI</strong> model we would need to do that first. We also provide a key for the column of <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code> which contains our cell labels as well as the label which corresponds to unlabelled cells. In this case all of our cells are labelled so we just provide a dummy value but in most cases it is important to check that this is set correctly so that <strong>scANVI</strong> knows which label to ignore.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Normally we would need to run scVI first but we have already done that here</span>
<span class="c1"># model_scvi = scvi.model.SCVI(adata_scvi) etc.</span>
<span class="n">model_scanvi</span> <span class="o">=</span> <span class="n">scvi</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">SCANVI</span><span class="o">.</span><span class="n">from_scvi_model</span><span class="p">(</span>
    <span class="n">model_scvi</span><span class="p">,</span> <span class="n">labels_key</span><span class="o">=</span><span class="n">label_key</span><span class="p">,</span> <span class="n">unlabeled_category</span><span class="o">=</span><span class="s2">&quot;unlabelled&quot;</span>
<span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">model_scanvi</span><span class="p">)</span>
<span class="n">model_scanvi</span><span class="o">.</span><span class="n">view_anndata_setup</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">ScanVI Model with the following params: 
unlabeled_category: unlabelled, n_hidden: <span style="color: #008080; text-decoration-color: #008080; font-weight: bold">128</span>, n_latent: <span style="color: #008080; text-decoration-color: #008080; font-weight: bold">10</span>, n_layers: <span style="color: #008080; text-decoration-color: #008080; font-weight: bold">1</span>, dropout_rate: <span style="color: #008080; text-decoration-color: #008080; font-weight: bold">0.1</span>, 
dispersion: gene, gene_likelihood: zinb
Training status: Not Trained
</pre>
</div><div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
<div class="output text_html"><pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">Anndata setup with scvi-tools version <span style="color: #008080; text-decoration-color: #008080; font-weight: bold">0.16</span>.<span style="color: #008080; text-decoration-color: #008080; font-weight: bold">4</span>.
</pre>
</div><div class="output text_html"><pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">
</pre>
</div><div class="output text_html"><pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">Setup via `SCANVI.setup_anndata` with arguments:
</pre>
</div><div class="output text_html"><pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"><span style="font-weight: bold">{</span>
<span style="color: #7fbf7f; text-decoration-color: #7fbf7f">│   </span><span style="color: #008000; text-decoration-color: #008000">'labels_key'</span>: <span style="color: #008000; text-decoration-color: #008000">'cell_type'</span>,
<span style="color: #7fbf7f; text-decoration-color: #7fbf7f">│   </span><span style="color: #008000; text-decoration-color: #008000">'unlabeled_category'</span>: <span style="color: #008000; text-decoration-color: #008000">'unlabelled'</span>,
<span style="color: #7fbf7f; text-decoration-color: #7fbf7f">│   </span><span style="color: #008000; text-decoration-color: #008000">'layer'</span>: <span style="color: #008000; text-decoration-color: #008000">'counts'</span>,
<span style="color: #7fbf7f; text-decoration-color: #7fbf7f">│   </span><span style="color: #008000; text-decoration-color: #008000">'batch_key'</span>: <span style="color: #008000; text-decoration-color: #008000">'batch'</span>,
<span style="color: #7fbf7f; text-decoration-color: #7fbf7f">│   </span><span style="color: #008000; text-decoration-color: #008000">'size_factor_key'</span>: <span style="color: #800080; text-decoration-color: #800080; font-style: italic">None</span>,
<span style="color: #7fbf7f; text-decoration-color: #7fbf7f">│   </span><span style="color: #008000; text-decoration-color: #008000">'categorical_covariate_keys'</span>: <span style="color: #800080; text-decoration-color: #800080; font-style: italic">None</span>,
<span style="color: #7fbf7f; text-decoration-color: #7fbf7f">│   </span><span style="color: #008000; text-decoration-color: #008000">'continuous_covariate_keys'</span>: <span style="color: #800080; text-decoration-color: #800080; font-style: italic">None</span>
<span style="font-weight: bold">}</span>
</pre>
</div><div class="output text_html"><pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">
</pre>
</div><div class="output text_html"><pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"><span style="font-style: italic">         Summary Statistics         </span>
┏━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━┓
┃<span style="font-weight: bold">     Summary Stat Key     </span>┃<span style="font-weight: bold"> Value </span>┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━┩
│<span style="color: #0087ff; text-decoration-color: #0087ff">         n_cells          </span>│<span style="color: #af00d7; text-decoration-color: #af00d7"> 22091 </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">          n_vars          </span>│<span style="color: #af00d7; text-decoration-color: #af00d7"> 2000  </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">         n_batch          </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">   4   </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">         n_labels         </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">  22   </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff"> n_extra_categorical_covs </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">   0   </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff"> n_extra_continuous_covs  </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">   0   </span>│
└──────────────────────────┴───────┘
</pre>
</div><div class="output text_html"><pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"><span style="font-style: italic">               Data Registry                </span>
┏━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃<span style="font-weight: bold"> Registry Key </span>┃<span style="font-weight: bold">    scvi-tools Location    </span>┃
┡━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│<span style="color: #0087ff; text-decoration-color: #0087ff">      X       </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">  adata.layers['counts']   </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">    batch     </span>│<span style="color: #af00d7; text-decoration-color: #af00d7"> adata.obs['_scvi_batch']  </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">    labels    </span>│<span style="color: #af00d7; text-decoration-color: #af00d7"> adata.obs['_scvi_labels'] </span>│
└──────────────┴───────────────────────────┘
</pre>
</div><div class="output text_html"><pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"><span style="font-style: italic">                  batch State Registry                   </span>
┏━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━┓
┃<span style="font-weight: bold">  Source Location   </span>┃<span style="font-weight: bold"> Categories </span>┃<span style="font-weight: bold"> scvi-tools Encoding </span>┃
┡━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━┩
│<span style="color: #0087ff; text-decoration-color: #0087ff"> adata.obs['batch'] </span>│<span style="color: #008000; text-decoration-color: #008000">    s1d1    </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">          0          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                    </span>│<span style="color: #008000; text-decoration-color: #008000">    s2d1    </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">          1          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                    </span>│<span style="color: #008000; text-decoration-color: #008000">    s3d7    </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">          2          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                    </span>│<span style="color: #008000; text-decoration-color: #008000">    s4d8    </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">          3          </span>│
└────────────────────┴────────────┴─────────────────────┘
</pre>
</div><div class="output text_html"><pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"><span style="font-style: italic">                        labels State Registry                         </span>
┏━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━┓
┃<span style="font-weight: bold">    Source Location     </span>┃<span style="font-weight: bold">     Categories      </span>┃<span style="font-weight: bold"> scvi-tools Encoding </span>┃
┡━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━┩
│<span style="color: #0087ff; text-decoration-color: #0087ff"> adata.obs['cell_type'] </span>│<span style="color: #008000; text-decoration-color: #008000">        B1 B         </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">          0          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                        </span>│<span style="color: #008000; text-decoration-color: #008000">  CD4+ T activated   </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">          1          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                        </span>│<span style="color: #008000; text-decoration-color: #008000">    CD4+ T naive     </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">          2          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                        </span>│<span style="color: #008000; text-decoration-color: #008000">       CD8+ T        </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">          3          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                        </span>│<span style="color: #008000; text-decoration-color: #008000">     CD14+ Mono      </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">          4          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                        </span>│<span style="color: #008000; text-decoration-color: #008000">     CD16+ Mono      </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">          5          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                        </span>│<span style="color: #008000; text-decoration-color: #008000">    Erythroblast     </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">          6          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                        </span>│<span style="color: #008000; text-decoration-color: #008000">      G/M prog       </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">          7          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                        </span>│<span style="color: #008000; text-decoration-color: #008000">         HSC         </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">          8          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                        </span>│<span style="color: #008000; text-decoration-color: #008000"> ID2-hi myeloid prog </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">          9          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                        </span>│<span style="color: #008000; text-decoration-color: #008000">         ILC         </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">         10          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                        </span>│<span style="color: #008000; text-decoration-color: #008000">     Lymph prog      </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">         11          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                        </span>│<span style="color: #008000; text-decoration-color: #008000">      MK/E prog      </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">         12          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                        </span>│<span style="color: #008000; text-decoration-color: #008000">         NK          </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">         13          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                        </span>│<span style="color: #008000; text-decoration-color: #008000">    Naive CD20+ B    </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">         14          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                        </span>│<span style="color: #008000; text-decoration-color: #008000">     Normoblast      </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">         15          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                        </span>│<span style="color: #008000; text-decoration-color: #008000">     Plasma cell     </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">         16          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                        </span>│<span style="color: #008000; text-decoration-color: #008000">   Proerythroblast   </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">         17          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                        </span>│<span style="color: #008000; text-decoration-color: #008000">   Transitional B    </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">         18          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                        </span>│<span style="color: #008000; text-decoration-color: #008000">        cDC2         </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">         19          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                        </span>│<span style="color: #008000; text-decoration-color: #008000">         pDC         </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">         20          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">                        </span>│<span style="color: #008000; text-decoration-color: #008000">     unlabelled      </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">         21          </span>│
└────────────────────────┴─────────────────────┴─────────────────────┘
</pre>
</div></div>
</div>
<p>This model object is very similar to what we saw before for <strong>scVI</strong>. As mentioned previously we could modify the structure of the model network but here we just use the default.</p>
<p>Again we have a heuristic for selecting the number of training epochs. Note that this is much fewer than before as we are just refining the <strong>scVI</strong> model, rather than training a whole network from scratch.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">max_epochs_scanvi</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">min</span><span class="p">([</span><span class="mi">10</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">([</span><span class="mi">2</span><span class="p">,</span> <span class="nb">round</span><span class="p">(</span><span class="n">max_epochs_scvi</span> <span class="o">/</span> <span class="mf">3.0</span><span class="p">)])]))</span>
<span class="n">model_scanvi</span><span class="o">.</span><span class="n">train</span><span class="p">(</span><span class="n">max_epochs</span><span class="o">=</span><span class="n">max_epochs_scanvi</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span><span class=" -Color -Color-Blue">INFO    </span> Training for <span class=" -Color -Color-Bold -Color-Bold-Cyan">10</span> epochs.                                                             
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/icb/luke.zappia/miniconda3/envs/bp-integration/lib/python3.8/site-packages/torchmetrics/utilities/prints.py:36: UserWarning: Torchmetrics v0.9 introduced a new argument class property called `full_state_update` that has
                not been set for this class (ElboMetric). The property determines if `update` by
                default needs access to the full metric state. If this is not the case, significant speedups can be
                achieved and we recommend setting this to `False`.
                We provide an checking function
                `from torchmetrics.utilities import check_forward_no_full_state`
                that can be used to check if the `full_state_update=True` (old and potential slower behaviour,
                default for now) or if `full_state_update=False` can be used safely.
                
  warnings.warn(*args, **kwargs)
GPU available: False, used: False
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
Set SLURM handle signals.
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 10/10: 100%|██████████| 10/10 [01:50&lt;00:00, 11.01s/it, loss=626, v_num=1]
</pre></div>
</div>
</div>
</div>
<p>We can extract the new latent representation from the model and create a new UMAP embedding as we did for <strong>scVI</strong>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_scanvi</span> <span class="o">=</span> <span class="n">adata_scvi</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">adata_scanvi</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;X_scANVI&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">model_scanvi</span><span class="o">.</span><span class="n">get_latent_representation</span><span class="p">()</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata_scanvi</span><span class="p">,</span> <span class="n">use_rep</span><span class="o">=</span><span class="s2">&quot;X_scANVI&quot;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_scanvi</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_scanvi</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="n">label_key</span><span class="p">,</span> <span class="n">batch_key</span><span class="p">],</span> <span class="n">wspace</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/integration_72_0.png" src="../_images/integration_72_0.png" />
</div>
</div>
<p>By looking at the UMAP representation it is difficult to tell the difference between <strong>scANVI</strong> and <strong>scVI</strong> but when the effect of the refinement has been quantified it has been shown to often better preserve biological variation.</p>
</section>
<section id="graph-based">
<h2><span class="section-number">12.6. </span>Graph based<a class="headerlink" href="#graph-based" title="Permalink to this headline">#</a></h2>
<p>The next method we will look at is <strong>BBKNN</strong> or “Batch Balanced KNN”. This is a very different approach to <strong>scVI</strong>, rather than using a neural network to embed cells in a batch-corrected space it instead modifies how the KNN graph used for clustering and embedding is constructed. As we have seen in previous chapters the normal KNN procedure connects cells to the most similar cells across the whole dataset. The change that <strong>BBKNN</strong> makes is to enforce that cells are connected to cells from other batches. While this is a simple modification it can be quite effective, particularly when there are very strong batch effects. However, as the result is an integrated graph it can have limited downstream uses as few packages will accept this as an input.</p>
<p>An important parameter for <strong>BBKNN</strong> is the number of neighbours per batch. A suggested heuristic for this is to use 25 if there are less than 10000 cells or the default of 3 if there are more than 10000.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">neighbors_within_batch</span> <span class="o">=</span> <span class="mi">25</span> <span class="k">if</span> <span class="n">adata_hvg</span><span class="o">.</span><span class="n">n_obs</span> <span class="o">&gt;</span> <span class="mi">10000</span> <span class="k">else</span> <span class="mi">3</span>
</pre></div>
</div>
</div>
</div>
<p>Before using <strong>BBKNN</strong> we first perform a PCA as we would before building a normal KNN graph. Unlike <strong>scVI</strong> which models raw counts here we start with the log-normalised expression matrix.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_bbknn</span> <span class="o">=</span> <span class="n">adata_hvg</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">adata_bbknn</span><span class="o">.</span><span class="n">X</span> <span class="o">=</span> <span class="n">adata_bbknn</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;logcounts&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">adata_bbknn</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>We can now run <strong>BBKNN</strong>, replacing the call to the <strong>scanpy</strong> <code class="docutils literal notranslate"><span class="pre">neighbors()</span></code> function in a standard workflow. An important difference is to make sure the <code class="docutils literal notranslate"><span class="pre">batch_key</span></code> argument is set which specifies a column in <code class="docutils literal notranslate"><span class="pre">adata_hvg.obs</span></code> which contains batch labels.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">bbknn</span><span class="o">.</span><span class="n">bbknn</span><span class="p">(</span>
    <span class="n">adata_bbknn</span><span class="p">,</span> <span class="n">batch_key</span><span class="o">=</span><span class="n">batch_key</span><span class="p">,</span> <span class="n">neighbors_within_batch</span><span class="o">=</span><span class="n">neighbors_within_batch</span>
<span class="p">)</span>
<span class="n">adata_bbknn</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 22091 × 2000
    obs: &#39;GEX_pct_counts_mt&#39;, &#39;GEX_n_counts&#39;, &#39;GEX_n_genes&#39;, &#39;GEX_size_factors&#39;, &#39;GEX_phase&#39;, &#39;ATAC_nCount_peaks&#39;, &#39;ATAC_atac_fragments&#39;, &#39;ATAC_reads_in_peaks_frac&#39;, &#39;ATAC_blacklist_fraction&#39;, &#39;ATAC_nucleosome_signal&#39;, &#39;cell_type&#39;, &#39;batch&#39;, &#39;ATAC_pseudotime_order&#39;, &#39;GEX_pseudotime_order&#39;, &#39;Samplename&#39;, &#39;Site&#39;, &#39;DonorNumber&#39;, &#39;Modality&#39;, &#39;VendorLot&#39;, &#39;DonorID&#39;, &#39;DonorAge&#39;, &#39;DonorBMI&#39;, &#39;DonorBloodType&#39;, &#39;DonorRace&#39;, &#39;Ethnicity&#39;, &#39;DonorGender&#39;, &#39;QCMeds&#39;, &#39;DonorSmoker&#39;
    var: &#39;feature_types&#39;, &#39;gene_id&#39;, &#39;n_cells&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable_nbatches&#39;, &#39;highly_variable_intersection&#39;
    uns: &#39;ATAC_gene_activity_var_names&#39;, &#39;dataset_id&#39;, &#39;genome&#39;, &#39;organism&#39;, &#39;log1p&#39;, &#39;hvg&#39;, &#39;pca&#39;, &#39;neighbors&#39;, &#39;umap&#39;, &#39;batch_colors&#39;, &#39;cell_type_colors&#39;
    obsm: &#39;ATAC_gene_activity&#39;, &#39;ATAC_lsi_full&#39;, &#39;ATAC_lsi_red&#39;, &#39;ATAC_umap&#39;, &#39;GEX_X_pca&#39;, &#39;GEX_X_umap&#39;, &#39;X_pca&#39;, &#39;X_umap&#39;
    varm: &#39;PCs&#39;
    layers: &#39;counts&#39;, &#39;logcounts&#39;
    obsp: &#39;distances&#39;, &#39;connectivities&#39;
</pre></div>
</div>
</div>
</div>
<p>Unlike the default <strong>scanpy</strong> function <strong>BBKNN</strong> does not allow specifying a key for storing results so they are always stored under the default “neighbors” key.</p>
<p>We can use this new integrated graph just like we would a normal KNN graph to construct a UMAP embedding.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_bbknn</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_bbknn</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="n">label_key</span><span class="p">,</span> <span class="n">batch_key</span><span class="p">],</span> <span class="n">wspace</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/integration_82_0.png" src="../_images/integration_82_0.png" />
</div>
</div>
<p>This integration is also improved compared to the unintegrated data with cell identities grouped together but we sill see some shifts between batches.</p>
</section>
<section id="mutual-nearest-neighbors-based">
<h2><span class="section-number">12.7. </span>Mutual nearest neighbors based<a class="headerlink" href="#mutual-nearest-neighbors-based" title="Permalink to this headline">#</a></h2>
<p>Some downstream applications cannot accept an integrated embedding or neighborhood graph and require a corrected expression matrix. One method that can produce this output is <strong>Seurat</strong>. The <strong>Seurat</strong> integration method belongs to a class of methods which is built on the idea of <em>mutual nearest neighbors</em> (which <strong>Seurat</strong> calls <em>anchors</em>). These are pairs of cells from two different datasets which are in the neighborhood of each other when the datasets are placed in the same space. After finding these cells they can be used to align the two datasets and correct the differences between them. Seurat has also been found to one of the top mixing methods <span id="id4">[<a class="reference internal" href="#id159" title="Hoa Thi Nhu Tran, Kok Siong Ang, Marion Chevrier, Xiaomeng Zhang, Nicole Yee Shin Lee, Michelle Goh, and Jinmiao Chen. A benchmark of batch-effect correction methods for single-cell rna sequencing data. Genome Biology, 21(1):12, Jan 2020. URL: https://doi.org/10.1186/s13059-019-1850-9, doi:10.1186/s13059-019-1850-9.">Tran <em>et al.</em>, 2020</a>]</span>.</p>
<p>As <strong>Seurat</strong> is an R package we must transfer our data from Python to R. Here we prepare the object to convert so that it can be handled by <strong>rpy2</strong> and <strong>anndata2ri</strong>.</p>
<p><strong>!!! TODO !!!</strong> - See if can avoid densifying</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_seurat</span> <span class="o">=</span> <span class="n">adata_hvg</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="c1"># Convert categorical columns to strings</span>
<span class="n">adata_seurat</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">batch_key</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata_seurat</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">batch_key</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
<span class="n">adata_seurat</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">label_key</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata_seurat</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">label_key</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
<span class="c1"># Densify expression matrices</span>
<span class="n">adata_seurat</span><span class="o">.</span><span class="n">X</span> <span class="o">=</span> <span class="n">adata_seurat</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">todense</span><span class="p">()</span>
<span class="n">adata_seurat</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;counts&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata_seurat</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;counts&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">todense</span><span class="p">()</span>
<span class="n">adata_seurat</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;logcounts&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata_seurat</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;logcounts&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">todense</span><span class="p">()</span>
<span class="k">del</span> <span class="n">adata_seurat</span><span class="o">.</span><span class="n">obsm</span>  <span class="c1"># Also contains sparse matrices we don&#39;t need</span>
<span class="c1"># Delete uns as this can contain arbitrary objects which are difficult to convert</span>
<span class="k">del</span> <span class="n">adata_seurat</span><span class="o">.</span><span class="n">uns</span>
<span class="n">adata_seurat</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 22091 × 2000
    obs: &#39;GEX_pct_counts_mt&#39;, &#39;GEX_n_counts&#39;, &#39;GEX_n_genes&#39;, &#39;GEX_size_factors&#39;, &#39;GEX_phase&#39;, &#39;ATAC_nCount_peaks&#39;, &#39;ATAC_atac_fragments&#39;, &#39;ATAC_reads_in_peaks_frac&#39;, &#39;ATAC_blacklist_fraction&#39;, &#39;ATAC_nucleosome_signal&#39;, &#39;cell_type&#39;, &#39;batch&#39;, &#39;ATAC_pseudotime_order&#39;, &#39;GEX_pseudotime_order&#39;, &#39;Samplename&#39;, &#39;Site&#39;, &#39;DonorNumber&#39;, &#39;Modality&#39;, &#39;VendorLot&#39;, &#39;DonorID&#39;, &#39;DonorAge&#39;, &#39;DonorBMI&#39;, &#39;DonorBloodType&#39;, &#39;DonorRace&#39;, &#39;Ethnicity&#39;, &#39;DonorGender&#39;, &#39;QCMeds&#39;, &#39;DonorSmoker&#39;
    var: &#39;feature_types&#39;, &#39;gene_id&#39;, &#39;n_cells&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable_nbatches&#39;, &#39;highly_variable_intersection&#39;
    varm: &#39;PCs&#39;
    layers: &#39;counts&#39;, &#39;logcounts&#39;
    obsp: &#39;distances&#39;, &#39;connectivities&#39;
</pre></div>
</div>
</div>
</div>
<p>The prepared object is now available in R as a SingleCellExperiment object thanks to <strong>anndata2ri</strong>. Note that this is transposed compared to an AnnData object so our observations (cells) are now the columns and our variables (genes) are now the rows).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span> -i adata_seurat
adata_seurat
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/icb/luke.zappia/miniconda3/envs/bp-integration/lib/python3.8/site-packages/rpy2/robjects/pandas2ri.py:261: DeprecationWarning: The global conversion available with activate() is deprecated and will be removed in the next major release. Use a local converter.
  warnings.warn(&#39;The global conversion available with activate() &#39;
/home/icb/luke.zappia/miniconda3/envs/bp-integration/lib/python3.8/site-packages/rpy2/robjects/numpy2ri.py:195: DeprecationWarning: The global conversion available with activate() is deprecated and will be removed in the next major release. Use a local converter.
  warnings.warn(&#39;The global conversion available with activate() &#39;
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>class: SingleCellExperiment 
dim: 2000 22091 
metadata(0):
assays(3): X counts logcounts
rownames(2000): GPR153 ESPN ... DNASE1L1 CLIC2
rowData names(9): feature_types gene_id ... highly_variable_nbatches
  highly_variable_intersection
colnames(22091): TAGTTGTCACCCTCAC-1-s1d1 CTATGGCCATAACGGG-1-s1d1 ...
  AAGTCTATCTTAGTCT-14-s4d8 GAAGCCTGTTTGGGTA-14-s4d8
colData names(28): GEX_pct_counts_mt GEX_n_counts ... QCMeds
  DonorSmoker
reducedDimNames(0):
spikeNames(0):
altExpNames(0):
</pre></div>
</div>
</div>
</div>
<p><strong>Seurat</strong> uses it’s own object to store data. Helpfully the authors provide a function to convert from SingleCellExperiment. We just provide the object and tell <strong>Seurat</strong> which assays (layers in our AnnData object) contain raw counts and normalised expression (which <strong>Seurat</strong> stores in a slot called “data”).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span> -i adata_seurat
seurat &lt;- as.Seurat(adata_seurat, counts = &quot;counts&quot;, data = &quot;logcounts&quot;)
seurat
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>An object of class Seurat 
2000 features across 22091 samples within 1 assay 
Active assay: RNA (2000 features, 0 variable features)
</pre></div>
</div>
</div>
</div>
<p>Unlike some of the other methods we have seen which take a single object and a batch key, the <strong>Seurat</strong> integration functions require a list of objects. We create this using the <code class="docutils literal notranslate"><span class="pre">SplitObject()</span></code> function.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span> -i batch_key
batch_list &lt;- SplitObject(seurat, split.by = batch_key)
batch_list
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>$s1d1
An object of class Seurat 
2000 features across 6224 samples within 1 assay 
Active assay: RNA (2000 features, 0 variable features)

$s2d1
An object of class Seurat 
2000 features across 4220 samples within 1 assay 
Active assay: RNA (2000 features, 0 variable features)

$s3d7
An object of class Seurat 
2000 features across 1771 samples within 1 assay 
Active assay: RNA (2000 features, 0 variable features)

$s4d8
An object of class Seurat 
2000 features across 9876 samples within 1 assay 
Active assay: RNA (2000 features, 0 variable features)
</pre></div>
</div>
</div>
</div>
<p>We can now use this list to find anchors for each pairs of datasets. Usually you would identify batch-aware highly variable genes first (using the <code class="docutils literal notranslate"><span class="pre">FindVariableFeatures()</span></code> and <code class="docutils literal notranslate"><span class="pre">SelectIntegrationFeatures()</span></code> functions) but as we have done that already we tell <strong>Seurat</strong> to use all the features in the object.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
anchors &lt;- FindIntegrationAnchors(batch_list, anchor.features = rownames(seurat))
anchors
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>  |                                                  | 0 % ~calculating   |+++++++++++++                                     | 25% ~03s           |+++++++++++++++++++++++++                         | 50% ~01s           |++++++++++++++++++++++++++++++++++++++            | 75% ~01s           |++++++++++++++++++++++++++++++++++++++++++++++++++| 100% elapsed=03s  
  |                                                  | 0 % ~calculating   |+++++++++                                         | 17% ~02m 10s       |+++++++++++++++++                                 | 33% ~01m 22s       |+++++++++++++++++++++++++                         | 50% ~52s           |++++++++++++++++++++++++++++++++++                | 67% ~55s           |++++++++++++++++++++++++++++++++++++++++++        | 83% ~31s           |++++++++++++++++++++++++++++++++++++++++++++++++++| 100% elapsed=03m 04s
An AnchorSet object containing 57268 anchors between 4 Seurat objects 
 This can be used as input to IntegrateData or TransferData.
</pre></div>
</div>
</div>
</div>
<p><strong>Seurat</strong> can then use the anchors to compute a transformation which maps one dataset onto another. This is done in a pairwise way until all the datasets are merged. By default <strong>Seurat</strong> will determine a merge order so that more similar datasets are merged together first but it is also possible to define this order.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
integrated &lt;- IntegrateData(anchors)
integrated
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>An object of class Seurat 
4000 features across 22091 samples within 2 assays 
Active assay: integrated (2000 features, 2000 variable features)
 1 other assay present: RNA
</pre></div>
</div>
</div>
</div>
<p>The result is another Seurat object but notice now that the active assay is called “integrated”. This contains the corrected expression matrix which is the final output of the integration.</p>
<p>Here we extract that matrix and prepare it for transfer back to Python.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span> -o integrated_expr
# Extract the integrated expression matrix
integrated_expr &lt;- as.matrix(GetAssayData(integrated))
# Make sure the rows and columns are in the same order as the original object
integrated_expr &lt;- integrated_expr[rownames(seurat), colnames(seurat)]
# Transpose the matrix to AnnData format
integrated_expr &lt;- t(integrated_expr)
print(integrated_expr[1:10, 1:10])
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>                        GPR153 ESPN TNFRSF25 TNFRSF9 SPSB1 KAZN     PADI2 PADI4
TAGTTGTCACCCTCAC-1-s1d1      0    0        0       0     0    0 0.0000000     0
CTATGGCCATAACGGG-1-s1d1      0    0        0       0     0    0 0.0000000     0
CCGCACACAGGTTAAA-1-s1d1      0    0        0       0     0    0 0.1853299     0
TCATTTGGTAATGGAA-1-s1d1      0    0        0       0     0    0 0.0000000     0
ACCACATAGGTGTCCA-1-s1d1      0    0        0       0     0    0 0.0000000     0
TGGATTGGTTTGCGAA-1-s1d1      0    0        0       0     0    0 0.0000000     0
GTGAGCGAGTAAAGGT-1-s1d1      0    0        0       0     0    0 0.0000000     0
GACTTAGGTTGCGCGA-1-s1d1      0    0        0       0     0    0 0.0000000     0
GCCTTACTCGTTACAA-1-s1d1      0    0        0       0     0    0 0.0000000     0
GTAAGGTCAATAACCT-1-s1d1      0    0        0       0     0    0 0.0000000     0
                        CDA  ELOA-AS1
TAGTTGTCACCCTCAC-1-s1d1   0 0.0000000
CTATGGCCATAACGGG-1-s1d1   0 0.0000000
CCGCACACAGGTTAAA-1-s1d1   0 0.0000000
TCATTTGGTAATGGAA-1-s1d1   0 0.0000000
ACCACATAGGTGTCCA-1-s1d1   0 0.0000000
TGGATTGGTTTGCGAA-1-s1d1   0 0.0000000
GTGAGCGAGTAAAGGT-1-s1d1   0 0.3227825
GACTTAGGTTGCGCGA-1-s1d1   0 0.0000000
GCCTTACTCGTTACAA-1-s1d1   0 0.0000000
GTAAGGTCAATAACCT-1-s1d1   0 0.0000000
</pre></div>
</div>
</div>
</div>
<p>We will now store the corrected expression matrix as a layer in our AnnData object. We also set <code class="docutils literal notranslate"><span class="pre">adata.X</span></code> to use this matrix.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_seurat</span><span class="o">.</span><span class="n">X</span> <span class="o">=</span> <span class="n">integrated_expr</span>
<span class="n">adata_seurat</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;seurat&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">integrated_expr</span>
<span class="nb">print</span><span class="p">(</span><span class="n">adata_seurat</span><span class="p">)</span>
<span class="n">adata</span><span class="o">.</span><span class="n">X</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 22091 × 2000
    obs: &#39;GEX_pct_counts_mt&#39;, &#39;GEX_n_counts&#39;, &#39;GEX_n_genes&#39;, &#39;GEX_size_factors&#39;, &#39;GEX_phase&#39;, &#39;ATAC_nCount_peaks&#39;, &#39;ATAC_atac_fragments&#39;, &#39;ATAC_reads_in_peaks_frac&#39;, &#39;ATAC_blacklist_fraction&#39;, &#39;ATAC_nucleosome_signal&#39;, &#39;cell_type&#39;, &#39;batch&#39;, &#39;ATAC_pseudotime_order&#39;, &#39;GEX_pseudotime_order&#39;, &#39;Samplename&#39;, &#39;Site&#39;, &#39;DonorNumber&#39;, &#39;Modality&#39;, &#39;VendorLot&#39;, &#39;DonorID&#39;, &#39;DonorAge&#39;, &#39;DonorBMI&#39;, &#39;DonorBloodType&#39;, &#39;DonorRace&#39;, &#39;Ethnicity&#39;, &#39;DonorGender&#39;, &#39;QCMeds&#39;, &#39;DonorSmoker&#39;
    var: &#39;feature_types&#39;, &#39;gene_id&#39;, &#39;n_cells&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;highly_variable_nbatches&#39;, &#39;highly_variable_intersection&#39;
    varm: &#39;PCs&#39;
    layers: &#39;counts&#39;, &#39;logcounts&#39;, &#39;seurat&#39;
    obsp: &#39;distances&#39;, &#39;connectivities&#39;
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&lt;22091x13431 sparse matrix of type &#39;&lt;class &#39;numpy.float32&#39;&gt;&#39;
	with 23917995 stored elements in Compressed Sparse Row format&gt;
</pre></div>
</div>
</div>
</div>
<p>Now that we have the results of our integration we can calculate a UMAP and plot it as we have for the other methods.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">adata_seurat</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata_seurat</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_seurat</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_seurat</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="n">label_key</span><span class="p">,</span> <span class="n">batch_key</span><span class="p">],</span> <span class="n">wspace</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>... storing &#39;cell_type&#39; as categorical
... storing &#39;batch&#39; as categorical
</pre></div>
</div>
<img alt="../_images/integration_102_1.png" src="../_images/integration_102_1.png" />
</div>
</div>
<p>As we have previously seen, the batches are mixed while the labels are separated. It is tempting to select an integration based on the UMAPs but this does not fully represent the quality of an integration. In the next section we present some approaches to more rigorously evaluate integration methods.</p>
</section>
<section id="benchmarking-your-own-integration">
<h2><span class="section-number">12.8. </span>Benchmarking your own integration<a class="headerlink" href="#benchmarking-your-own-integration" title="Permalink to this headline">#</a></h2>
<p>The methods demonstrated here are selected based on results from benchmarking experiments including the <a class="reference external" href="https://theislab.github.io/scib-reproducibility/">single-cell integration benchmarking project</a>. This project also produced a software package called <strong>scib</strong> that can be used to run a range of integration methods as well as the metrics that were used for evaluation. In this section we show how to use this package to evaluate the quality of an integration.</p>
<blockquote>
<div><p><strong>What is the ground truth?</strong></p>
<p>Some of these metrics, particularly those that evaluate conservation of biological variation, require a known ground truth to compare to. Usually this is a cell identity label but can sometimes be other information such as known trajectories. Because of this requirement it is difficult to evaluate integration for a completely new dataset where it is unclear what biological signal should be preserved.</p>
</div></blockquote>
<p>The <strong>scib</strong> metrics can be run individually but there are also wrappers for running multiple metrics at once. Here we run a subset of the metrics which are quick to compute using the <code class="docutils literal notranslate"><span class="pre">metrics_fast()</span></code> function. This function takes a few arguments: the original unintegrated dataset, the integrated dataset, a batch key and a label key. Depending on the output of the integration method we might also need to supply additional arguments, for example here we specify the embedding to use for <strong>scVI</strong> and <strong>scANVI</strong> with the <code class="docutils literal notranslate"><span class="pre">embed</span></code> argument. You can also control how some metrics are run with additional arguments.</p>
<p>Let’s run the metrics for each of the integrations we have performed above, as well as the unintegrated data (after highly variable gene selection).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">metrics_scvi</span> <span class="o">=</span> <span class="n">scib</span><span class="o">.</span><span class="n">metrics</span><span class="o">.</span><span class="n">metrics_fast</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span> <span class="n">adata_scvi</span><span class="p">,</span> <span class="n">batch_key</span><span class="p">,</span> <span class="n">label_key</span><span class="p">,</span> <span class="n">embed</span><span class="o">=</span><span class="s2">&quot;X_scVI&quot;</span>
<span class="p">)</span>
<span class="n">metrics_scanvi</span> <span class="o">=</span> <span class="n">scib</span><span class="o">.</span><span class="n">metrics</span><span class="o">.</span><span class="n">metrics_fast</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span> <span class="n">adata_scanvi</span><span class="p">,</span> <span class="n">batch_key</span><span class="p">,</span> <span class="n">label_key</span><span class="p">,</span> <span class="n">embed</span><span class="o">=</span><span class="s2">&quot;X_scANVI&quot;</span>
<span class="p">)</span>
<span class="n">metrics_bbknn</span> <span class="o">=</span> <span class="n">scib</span><span class="o">.</span><span class="n">metrics</span><span class="o">.</span><span class="n">metrics_fast</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">adata_bbknn</span><span class="p">,</span> <span class="n">batch_key</span><span class="p">,</span> <span class="n">label_key</span><span class="p">)</span>
<span class="n">metrics_seurat</span> <span class="o">=</span> <span class="n">scib</span><span class="o">.</span><span class="n">metrics</span><span class="o">.</span><span class="n">metrics_fast</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">adata_seurat</span><span class="p">,</span> <span class="n">batch_key</span><span class="p">,</span> <span class="n">label_key</span><span class="p">)</span>
<span class="n">metrics_hvg</span> <span class="o">=</span> <span class="n">scib</span><span class="o">.</span><span class="n">metrics</span><span class="o">.</span><span class="n">metrics_fast</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">adata_hvg</span><span class="p">,</span> <span class="n">batch_key</span><span class="p">,</span> <span class="n">label_key</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Silhouette score...
PC regression...
Isolated labels ASW...
Graph connectivity...
Silhouette score...
PC regression...
Isolated labels ASW...
Graph connectivity...
Silhouette score...
PC regression...
Isolated labels ASW...
Graph connectivity...
Silhouette score...
PC regression...
Isolated labels ASW...
Graph connectivity...
Silhouette score...
PC regression...
Isolated labels ASW...
Graph connectivity...
</pre></div>
</div>
</div>
</div>
<p>Here is an example of what one of the metrics results looks like for a single integration:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">metrics_hvg</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>0</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>NMI_cluster/label</th>
      <td>NaN</td>
    </tr>
    <tr>
      <th>ARI_cluster/label</th>
      <td>NaN</td>
    </tr>
    <tr>
      <th>ASW_label</th>
      <td>0.550467</td>
    </tr>
    <tr>
      <th>ASW_label/batch</th>
      <td>0.909000</td>
    </tr>
    <tr>
      <th>PCR_batch</th>
      <td>0.507870</td>
    </tr>
    <tr>
      <th>cell_cycle_conservation</th>
      <td>NaN</td>
    </tr>
    <tr>
      <th>isolated_label_F1</th>
      <td>NaN</td>
    </tr>
    <tr>
      <th>isolated_label_silhouette</th>
      <td>0.475478</td>
    </tr>
    <tr>
      <th>graph_conn</th>
      <td>0.987786</td>
    </tr>
    <tr>
      <th>kBET</th>
      <td>NaN</td>
    </tr>
    <tr>
      <th>iLISI</th>
      <td>NaN</td>
    </tr>
    <tr>
      <th>cLISI</th>
      <td>NaN</td>
    </tr>
    <tr>
      <th>hvg_overlap</th>
      <td>1.000000</td>
    </tr>
    <tr>
      <th>trajectory</th>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>Each row is a different metric and the values show the score for each metrics. Scores are between 0 and 1, where 1 is a good performance and 0 is a poor performance. Because we have only run the fast metrics here, some of the metrics have <code class="docutils literal notranslate"><span class="pre">NaN</span></code> scores. Also note that some metrics cannot be used with some output formats which can also be a reason for <code class="docutils literal notranslate"><span class="pre">NaN</span></code> values being returned.</p>
<p>To compare the methods it is useful to have all the metrics results in one table. This code combines them and tidies them into a more convenient format.</p>
<p><strong>!!! TODO !!!</strong> - Check if HVG overlap should be excluded (maybe should only be calculated on expression output)</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Concatenate metrics results</span>
<span class="n">metrics</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span>
    <span class="p">[</span><span class="n">metrics_scvi</span><span class="p">,</span> <span class="n">metrics_scanvi</span><span class="p">,</span> <span class="n">metrics_bbknn</span><span class="p">,</span> <span class="n">metrics_seurat</span><span class="p">,</span> <span class="n">metrics_hvg</span><span class="p">],</span>
    <span class="n">axis</span><span class="o">=</span><span class="s2">&quot;columns&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="c1"># Set methods as column names</span>
<span class="n">metrics</span> <span class="o">=</span> <span class="n">metrics</span><span class="o">.</span><span class="n">set_axis</span><span class="p">(</span>
    <span class="p">[</span><span class="s2">&quot;scVI&quot;</span><span class="p">,</span> <span class="s2">&quot;scANVI&quot;</span><span class="p">,</span> <span class="s2">&quot;BBKNN&quot;</span><span class="p">,</span> <span class="s2">&quot;Seurat&quot;</span><span class="p">,</span> <span class="s2">&quot;Unintegrated&quot;</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="s2">&quot;columns&quot;</span>
<span class="p">)</span>
<span class="c1"># Select only the fast metrics</span>
<span class="n">metrics</span> <span class="o">=</span> <span class="n">metrics</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
    <span class="p">[</span>
        <span class="s2">&quot;ASW_label&quot;</span><span class="p">,</span>
        <span class="s2">&quot;ASW_label/batch&quot;</span><span class="p">,</span>
        <span class="s2">&quot;PCR_batch&quot;</span><span class="p">,</span>
        <span class="s2">&quot;isolated_label_silhouette&quot;</span><span class="p">,</span>
        <span class="s2">&quot;graph_conn&quot;</span><span class="p">,</span>
        <span class="s2">&quot;hvg_overlap&quot;</span><span class="p">,</span>
    <span class="p">],</span>
    <span class="p">:,</span>
<span class="p">]</span>
<span class="c1"># Transpose so that metrics are columns and methods are rows</span>
<span class="n">metrics</span> <span class="o">=</span> <span class="n">metrics</span><span class="o">.</span><span class="n">T</span>
<span class="n">metrics</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>ASW_label</th>
      <th>ASW_label/batch</th>
      <th>PCR_batch</th>
      <th>isolated_label_silhouette</th>
      <th>graph_conn</th>
      <th>hvg_overlap</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>scVI</th>
      <td>0.574195</td>
      <td>0.899081</td>
      <td>0.846422</td>
      <td>0.513492</td>
      <td>0.973778</td>
      <td>1.0000</td>
    </tr>
    <tr>
      <th>scANVI</th>
      <td>0.594502</td>
      <td>0.900252</td>
      <td>0.813801</td>
      <td>0.491256</td>
      <td>0.976798</td>
      <td>1.0000</td>
    </tr>
    <tr>
      <th>BBKNN</th>
      <td>0.555377</td>
      <td>0.912848</td>
      <td>0.507860</td>
      <td>0.471376</td>
      <td>0.994674</td>
      <td>1.0000</td>
    </tr>
    <tr>
      <th>Seurat</th>
      <td>0.554370</td>
      <td>0.922662</td>
      <td>0.826818</td>
      <td>0.456714</td>
      <td>0.972622</td>
      <td>0.5945</td>
    </tr>
    <tr>
      <th>Unintegrated</th>
      <td>0.550467</td>
      <td>0.909000</td>
      <td>0.507870</td>
      <td>0.475478</td>
      <td>0.987786</td>
      <td>1.0000</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>We now have all the scores in one table with metrics as columns and methods as rows. Styling the table with a gradient can make it easier to see the differences between scores.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">metrics</span><span class="o">.</span><span class="n">style</span><span class="o">.</span><span class="n">background_gradient</span><span class="p">(</span><span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;Blues&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style  type="text/css" >
#T_ccb02_row0_col0{
            background-color:  #5fa6d1;
            color:  #000000;
        }#T_ccb02_row0_col1,#T_ccb02_row2_col2,#T_ccb02_row3_col3,#T_ccb02_row3_col4,#T_ccb02_row3_col5,#T_ccb02_row4_col0,#T_ccb02_row4_col2{
            background-color:  #f7fbff;
            color:  #000000;
        }#T_ccb02_row0_col2,#T_ccb02_row0_col3,#T_ccb02_row0_col5,#T_ccb02_row1_col0,#T_ccb02_row1_col5,#T_ccb02_row2_col4,#T_ccb02_row2_col5,#T_ccb02_row3_col1,#T_ccb02_row4_col5{
            background-color:  #08306b;
            color:  #f1f1f1;
        }#T_ccb02_row0_col4{
            background-color:  #edf4fc;
            color:  #000000;
        }#T_ccb02_row1_col1{
            background-color:  #eef5fc;
            color:  #000000;
        }#T_ccb02_row1_col2{
            background-color:  #084990;
            color:  #f1f1f1;
        }#T_ccb02_row1_col3{
            background-color:  #4896c8;
            color:  #000000;
        }#T_ccb02_row1_col4{
            background-color:  #d2e3f3;
            color:  #000000;
        }#T_ccb02_row2_col0{
            background-color:  #e1edf8;
            color:  #000000;
        }#T_ccb02_row2_col1{
            background-color:  #4f9bcb;
            color:  #000000;
        }#T_ccb02_row2_col3{
            background-color:  #c3daee;
            color:  #000000;
        }#T_ccb02_row3_col0{
            background-color:  #e6f0f9;
            color:  #000000;
        }#T_ccb02_row3_col2{
            background-color:  #083e81;
            color:  #f1f1f1;
        }#T_ccb02_row4_col1{
            background-color:  #8cc0dd;
            color:  #000000;
        }#T_ccb02_row4_col3{
            background-color:  #add0e6;
            color:  #000000;
        }#T_ccb02_row4_col4{
            background-color:  #3181bd;
            color:  #000000;
        }</style><table id="T_ccb02_" ><thead>    <tr>        <th class="blank level0" ></th>        <th class="col_heading level0 col0" >ASW_label</th>        <th class="col_heading level0 col1" >ASW_label/batch</th>        <th class="col_heading level0 col2" >PCR_batch</th>        <th class="col_heading level0 col3" >isolated_label_silhouette</th>        <th class="col_heading level0 col4" >graph_conn</th>        <th class="col_heading level0 col5" >hvg_overlap</th>    </tr></thead><tbody>
                <tr>
                        <th id="T_ccb02_level0_row0" class="row_heading level0 row0" >scVI</th>
                        <td id="T_ccb02_row0_col0" class="data row0 col0" >0.574195</td>
                        <td id="T_ccb02_row0_col1" class="data row0 col1" >0.899081</td>
                        <td id="T_ccb02_row0_col2" class="data row0 col2" >0.846422</td>
                        <td id="T_ccb02_row0_col3" class="data row0 col3" >0.513492</td>
                        <td id="T_ccb02_row0_col4" class="data row0 col4" >0.973778</td>
                        <td id="T_ccb02_row0_col5" class="data row0 col5" >1.000000</td>
            </tr>
            <tr>
                        <th id="T_ccb02_level0_row1" class="row_heading level0 row1" >scANVI</th>
                        <td id="T_ccb02_row1_col0" class="data row1 col0" >0.594502</td>
                        <td id="T_ccb02_row1_col1" class="data row1 col1" >0.900252</td>
                        <td id="T_ccb02_row1_col2" class="data row1 col2" >0.813801</td>
                        <td id="T_ccb02_row1_col3" class="data row1 col3" >0.491256</td>
                        <td id="T_ccb02_row1_col4" class="data row1 col4" >0.976798</td>
                        <td id="T_ccb02_row1_col5" class="data row1 col5" >1.000000</td>
            </tr>
            <tr>
                        <th id="T_ccb02_level0_row2" class="row_heading level0 row2" >BBKNN</th>
                        <td id="T_ccb02_row2_col0" class="data row2 col0" >0.555377</td>
                        <td id="T_ccb02_row2_col1" class="data row2 col1" >0.912848</td>
                        <td id="T_ccb02_row2_col2" class="data row2 col2" >0.507860</td>
                        <td id="T_ccb02_row2_col3" class="data row2 col3" >0.471376</td>
                        <td id="T_ccb02_row2_col4" class="data row2 col4" >0.994674</td>
                        <td id="T_ccb02_row2_col5" class="data row2 col5" >1.000000</td>
            </tr>
            <tr>
                        <th id="T_ccb02_level0_row3" class="row_heading level0 row3" >Seurat</th>
                        <td id="T_ccb02_row3_col0" class="data row3 col0" >0.554370</td>
                        <td id="T_ccb02_row3_col1" class="data row3 col1" >0.922662</td>
                        <td id="T_ccb02_row3_col2" class="data row3 col2" >0.826818</td>
                        <td id="T_ccb02_row3_col3" class="data row3 col3" >0.456714</td>
                        <td id="T_ccb02_row3_col4" class="data row3 col4" >0.972622</td>
                        <td id="T_ccb02_row3_col5" class="data row3 col5" >0.594500</td>
            </tr>
            <tr>
                        <th id="T_ccb02_level0_row4" class="row_heading level0 row4" >Unintegrated</th>
                        <td id="T_ccb02_row4_col0" class="data row4 col0" >0.550467</td>
                        <td id="T_ccb02_row4_col1" class="data row4 col1" >0.909000</td>
                        <td id="T_ccb02_row4_col2" class="data row4 col2" >0.507870</td>
                        <td id="T_ccb02_row4_col3" class="data row4 col3" >0.475478</td>
                        <td id="T_ccb02_row4_col4" class="data row4 col4" >0.987786</td>
                        <td id="T_ccb02_row4_col5" class="data row4 col5" >1.000000</td>
            </tr>
    </tbody></table></div></div>
</div>
<p>For some metrics the scores tend to be in a relatively small range. To emphasise the differences between methods and place each metric on the same scale, we scale them so that the worst performer gets a score of 0, the best performer gets a score of 1 and the others are somewhere in between.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">metrics_scaled</span> <span class="o">=</span> <span class="p">(</span><span class="n">metrics</span> <span class="o">-</span> <span class="n">metrics</span><span class="o">.</span><span class="n">min</span><span class="p">())</span> <span class="o">/</span> <span class="p">(</span><span class="n">metrics</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">-</span> <span class="n">metrics</span><span class="o">.</span><span class="n">min</span><span class="p">())</span>
<span class="n">metrics_scaled</span><span class="o">.</span><span class="n">style</span><span class="o">.</span><span class="n">background_gradient</span><span class="p">(</span><span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;Blues&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style  type="text/css" >
#T_2b413_row0_col0{
            background-color:  #5fa6d1;
            color:  #000000;
        }#T_2b413_row0_col1,#T_2b413_row2_col2,#T_2b413_row3_col3,#T_2b413_row3_col4,#T_2b413_row3_col5,#T_2b413_row4_col0,#T_2b413_row4_col2{
            background-color:  #f7fbff;
            color:  #000000;
        }#T_2b413_row0_col2,#T_2b413_row0_col3,#T_2b413_row0_col5,#T_2b413_row1_col0,#T_2b413_row1_col5,#T_2b413_row2_col4,#T_2b413_row2_col5,#T_2b413_row3_col1,#T_2b413_row4_col5{
            background-color:  #08306b;
            color:  #f1f1f1;
        }#T_2b413_row0_col4{
            background-color:  #edf4fc;
            color:  #000000;
        }#T_2b413_row1_col1{
            background-color:  #eef5fc;
            color:  #000000;
        }#T_2b413_row1_col2{
            background-color:  #084990;
            color:  #f1f1f1;
        }#T_2b413_row1_col3{
            background-color:  #4896c8;
            color:  #000000;
        }#T_2b413_row1_col4{
            background-color:  #d2e3f3;
            color:  #000000;
        }#T_2b413_row2_col0{
            background-color:  #e1edf8;
            color:  #000000;
        }#T_2b413_row2_col1{
            background-color:  #4f9bcb;
            color:  #000000;
        }#T_2b413_row2_col3{
            background-color:  #c3daee;
            color:  #000000;
        }#T_2b413_row3_col0{
            background-color:  #e6f0f9;
            color:  #000000;
        }#T_2b413_row3_col2{
            background-color:  #083e81;
            color:  #f1f1f1;
        }#T_2b413_row4_col1{
            background-color:  #8cc0dd;
            color:  #000000;
        }#T_2b413_row4_col3{
            background-color:  #add0e6;
            color:  #000000;
        }#T_2b413_row4_col4{
            background-color:  #3181bd;
            color:  #000000;
        }</style><table id="T_2b413_" ><thead>    <tr>        <th class="blank level0" ></th>        <th class="col_heading level0 col0" >ASW_label</th>        <th class="col_heading level0 col1" >ASW_label/batch</th>        <th class="col_heading level0 col2" >PCR_batch</th>        <th class="col_heading level0 col3" >isolated_label_silhouette</th>        <th class="col_heading level0 col4" >graph_conn</th>        <th class="col_heading level0 col5" >hvg_overlap</th>    </tr></thead><tbody>
                <tr>
                        <th id="T_2b413_level0_row0" class="row_heading level0 row0" >scVI</th>
                        <td id="T_2b413_row0_col0" class="data row0 col0" >0.538832</td>
                        <td id="T_2b413_row0_col1" class="data row0 col1" >0.000000</td>
                        <td id="T_2b413_row0_col2" class="data row0 col2" >1.000000</td>
                        <td id="T_2b413_row0_col3" class="data row0 col3" >1.000000</td>
                        <td id="T_2b413_row0_col4" class="data row0 col4" >0.052415</td>
                        <td id="T_2b413_row0_col5" class="data row0 col5" >1.000000</td>
            </tr>
            <tr>
                        <th id="T_2b413_level0_row1" class="row_heading level0 row1" >scANVI</th>
                        <td id="T_2b413_row1_col0" class="data row1 col0" >1.000000</td>
                        <td id="T_2b413_row1_col1" class="data row1 col1" >0.049643</td>
                        <td id="T_2b413_row1_col2" class="data row1 col2" >0.903648</td>
                        <td id="T_2b413_row1_col3" class="data row1 col3" >0.608373</td>
                        <td id="T_2b413_row1_col4" class="data row1 col4" >0.189377</td>
                        <td id="T_2b413_row1_col5" class="data row1 col5" >1.000000</td>
            </tr>
            <tr>
                        <th id="T_2b413_level0_row2" class="row_heading level0 row2" >BBKNN</th>
                        <td id="T_2b413_row2_col0" class="data row2 col0" >0.111505</td>
                        <td id="T_2b413_row2_col1" class="data row2 col1" >0.583824</td>
                        <td id="T_2b413_row2_col2" class="data row2 col2" >0.000000</td>
                        <td id="T_2b413_row2_col3" class="data row2 col3" >0.258229</td>
                        <td id="T_2b413_row2_col4" class="data row2 col4" >1.000000</td>
                        <td id="T_2b413_row2_col5" class="data row2 col5" >1.000000</td>
            </tr>
            <tr>
                        <th id="T_2b413_level0_row3" class="row_heading level0 row3" >Seurat</th>
                        <td id="T_2b413_row3_col0" class="data row3 col0" >0.088620</td>
                        <td id="T_2b413_row3_col1" class="data row3 col1" >1.000000</td>
                        <td id="T_2b413_row3_col2" class="data row3 col2" >0.942096</td>
                        <td id="T_2b413_row3_col3" class="data row3 col3" >0.000000</td>
                        <td id="T_2b413_row3_col4" class="data row3 col4" >0.000000</td>
                        <td id="T_2b413_row3_col5" class="data row3 col5" >0.000000</td>
            </tr>
            <tr>
                        <th id="T_2b413_level0_row4" class="row_heading level0 row4" >Unintegrated</th>
                        <td id="T_2b413_row4_col0" class="data row4 col0" >0.000000</td>
                        <td id="T_2b413_row4_col1" class="data row4 col1" >0.420647</td>
                        <td id="T_2b413_row4_col2" class="data row4 col2" >0.000030</td>
                        <td id="T_2b413_row4_col3" class="data row4 col3" >0.330475</td>
                        <td id="T_2b413_row4_col4" class="data row4 col4" >0.687636</td>
                        <td id="T_2b413_row4_col5" class="data row4 col5" >1.000000</td>
            </tr>
    </tbody></table></div></div>
</div>
<p>The values now better represent the differences between methods (and better match the colour scale). However, it is important to note that the scaled scores can only be used to compare the relative performance of this specific set of integrations. If we wanted to add another method we would need to perform the scaling again. We also can’t say that an integration is definitively “good”, only that it is better than the other methods we have tried. This scaling emphasises differences between methods. For example, if we had metric scores of 0.92, 0.94 and 0.96 these would be scaled to 0, 0.5 and 1.0. This makes the first method appear to score much worse, even though it is only slightly lower than the other two and still got a very high score. This effect is bigger when you are comparing few methods and when they get similar raw scores. Whether you look at raw or scaled scores depends on whether you want to focus on absolute performance or difference in performance between methods.</p>
<p>The evaluation metrics can be grouped into two categories, those that measure the removal of batch effects and those that measure the conservation of biological variation. We can calculate summary scores for each of these categories by taking the mean of the scaled values for each group. This kind of summary score wouldn’t make sense with raw values as some metrics consistently produce higher scores than others.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">metrics_scaled</span><span class="p">[</span><span class="s2">&quot;Batch&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">metrics_scaled</span><span class="p">[</span>
    <span class="p">[</span><span class="s2">&quot;ASW_label/batch&quot;</span><span class="p">,</span> <span class="s2">&quot;PCR_batch&quot;</span><span class="p">,</span> <span class="s2">&quot;graph_conn&quot;</span><span class="p">]</span>
<span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">metrics_scaled</span><span class="p">[</span><span class="s2">&quot;Bio&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">metrics_scaled</span><span class="p">[</span>
    <span class="p">[</span><span class="s2">&quot;ASW_label&quot;</span><span class="p">,</span> <span class="s2">&quot;isolated_label_silhouette&quot;</span><span class="p">,</span> <span class="s2">&quot;hvg_overlap&quot;</span><span class="p">]</span>
<span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">metrics_scaled</span><span class="o">.</span><span class="n">style</span><span class="o">.</span><span class="n">background_gradient</span><span class="p">(</span><span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;Blues&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style  type="text/css" >
#T_4dd57_row0_col0{
            background-color:  #5fa6d1;
            color:  #000000;
        }#T_4dd57_row0_col1,#T_4dd57_row0_col6,#T_4dd57_row2_col2,#T_4dd57_row3_col3,#T_4dd57_row3_col4,#T_4dd57_row3_col5,#T_4dd57_row3_col7,#T_4dd57_row4_col0,#T_4dd57_row4_col2{
            background-color:  #f7fbff;
            color:  #000000;
        }#T_4dd57_row0_col2,#T_4dd57_row0_col3,#T_4dd57_row0_col5,#T_4dd57_row1_col0,#T_4dd57_row1_col5,#T_4dd57_row1_col7,#T_4dd57_row2_col4,#T_4dd57_row2_col5,#T_4dd57_row3_col1,#T_4dd57_row3_col6,#T_4dd57_row4_col5{
            background-color:  #08306b;
            color:  #f1f1f1;
        }#T_4dd57_row0_col4{
            background-color:  #edf4fc;
            color:  #000000;
        }#T_4dd57_row0_col7{
            background-color:  #083776;
            color:  #f1f1f1;
        }#T_4dd57_row1_col1{
            background-color:  #eef5fc;
            color:  #000000;
        }#T_4dd57_row1_col2{
            background-color:  #084990;
            color:  #f1f1f1;
        }#T_4dd57_row1_col3{
            background-color:  #4896c8;
            color:  #000000;
        }#T_4dd57_row1_col4{
            background-color:  #d2e3f3;
            color:  #000000;
        }#T_4dd57_row1_col6{
            background-color:  #e3eef9;
            color:  #000000;
        }#T_4dd57_row2_col0{
            background-color:  #e1edf8;
            color:  #000000;
        }#T_4dd57_row2_col1{
            background-color:  #4f9bcb;
            color:  #000000;
        }#T_4dd57_row2_col3{
            background-color:  #c3daee;
            color:  #000000;
        }#T_4dd57_row2_col6{
            background-color:  #4b98ca;
            color:  #000000;
        }#T_4dd57_row2_col7{
            background-color:  #68acd5;
            color:  #000000;
        }#T_4dd57_row3_col0{
            background-color:  #e6f0f9;
            color:  #000000;
        }#T_4dd57_row3_col2{
            background-color:  #083e81;
            color:  #f1f1f1;
        }#T_4dd57_row4_col1{
            background-color:  #8cc0dd;
            color:  #000000;
        }#T_4dd57_row4_col3{
            background-color:  #add0e6;
            color:  #000000;
        }#T_4dd57_row4_col4{
            background-color:  #3181bd;
            color:  #000000;
        }#T_4dd57_row4_col6{
            background-color:  #eaf3fb;
            color:  #000000;
        }#T_4dd57_row4_col7{
            background-color:  #6dafd7;
            color:  #000000;
        }</style><table id="T_4dd57_" ><thead>    <tr>        <th class="blank level0" ></th>        <th class="col_heading level0 col0" >ASW_label</th>        <th class="col_heading level0 col1" >ASW_label/batch</th>        <th class="col_heading level0 col2" >PCR_batch</th>        <th class="col_heading level0 col3" >isolated_label_silhouette</th>        <th class="col_heading level0 col4" >graph_conn</th>        <th class="col_heading level0 col5" >hvg_overlap</th>        <th class="col_heading level0 col6" >Batch</th>        <th class="col_heading level0 col7" >Bio</th>    </tr></thead><tbody>
                <tr>
                        <th id="T_4dd57_level0_row0" class="row_heading level0 row0" >scVI</th>
                        <td id="T_4dd57_row0_col0" class="data row0 col0" >0.538832</td>
                        <td id="T_4dd57_row0_col1" class="data row0 col1" >0.000000</td>
                        <td id="T_4dd57_row0_col2" class="data row0 col2" >1.000000</td>
                        <td id="T_4dd57_row0_col3" class="data row0 col3" >1.000000</td>
                        <td id="T_4dd57_row0_col4" class="data row0 col4" >0.052415</td>
                        <td id="T_4dd57_row0_col5" class="data row0 col5" >1.000000</td>
                        <td id="T_4dd57_row0_col6" class="data row0 col6" >0.350805</td>
                        <td id="T_4dd57_row0_col7" class="data row0 col7" >0.846277</td>
            </tr>
            <tr>
                        <th id="T_4dd57_level0_row1" class="row_heading level0 row1" >scANVI</th>
                        <td id="T_4dd57_row1_col0" class="data row1 col0" >1.000000</td>
                        <td id="T_4dd57_row1_col1" class="data row1 col1" >0.049643</td>
                        <td id="T_4dd57_row1_col2" class="data row1 col2" >0.903648</td>
                        <td id="T_4dd57_row1_col3" class="data row1 col3" >0.608373</td>
                        <td id="T_4dd57_row1_col4" class="data row1 col4" >0.189377</td>
                        <td id="T_4dd57_row1_col5" class="data row1 col5" >1.000000</td>
                        <td id="T_4dd57_row1_col6" class="data row1 col6" >0.380890</td>
                        <td id="T_4dd57_row1_col7" class="data row1 col7" >0.869458</td>
            </tr>
            <tr>
                        <th id="T_4dd57_level0_row2" class="row_heading level0 row2" >BBKNN</th>
                        <td id="T_4dd57_row2_col0" class="data row2 col0" >0.111505</td>
                        <td id="T_4dd57_row2_col1" class="data row2 col1" >0.583824</td>
                        <td id="T_4dd57_row2_col2" class="data row2 col2" >0.000000</td>
                        <td id="T_4dd57_row2_col3" class="data row2 col3" >0.258229</td>
                        <td id="T_4dd57_row2_col4" class="data row2 col4" >1.000000</td>
                        <td id="T_4dd57_row2_col5" class="data row2 col5" >1.000000</td>
                        <td id="T_4dd57_row2_col6" class="data row2 col6" >0.527941</td>
                        <td id="T_4dd57_row2_col7" class="data row2 col7" >0.456578</td>
            </tr>
            <tr>
                        <th id="T_4dd57_level0_row3" class="row_heading level0 row3" >Seurat</th>
                        <td id="T_4dd57_row3_col0" class="data row3 col0" >0.088620</td>
                        <td id="T_4dd57_row3_col1" class="data row3 col1" >1.000000</td>
                        <td id="T_4dd57_row3_col2" class="data row3 col2" >0.942096</td>
                        <td id="T_4dd57_row3_col3" class="data row3 col3" >0.000000</td>
                        <td id="T_4dd57_row3_col4" class="data row3 col4" >0.000000</td>
                        <td id="T_4dd57_row3_col5" class="data row3 col5" >0.000000</td>
                        <td id="T_4dd57_row3_col6" class="data row3 col6" >0.647365</td>
                        <td id="T_4dd57_row3_col7" class="data row3 col7" >0.029540</td>
            </tr>
            <tr>
                        <th id="T_4dd57_level0_row4" class="row_heading level0 row4" >Unintegrated</th>
                        <td id="T_4dd57_row4_col0" class="data row4 col0" >0.000000</td>
                        <td id="T_4dd57_row4_col1" class="data row4 col1" >0.420647</td>
                        <td id="T_4dd57_row4_col2" class="data row4 col2" >0.000030</td>
                        <td id="T_4dd57_row4_col3" class="data row4 col3" >0.330475</td>
                        <td id="T_4dd57_row4_col4" class="data row4 col4" >0.687636</td>
                        <td id="T_4dd57_row4_col5" class="data row4 col5" >1.000000</td>
                        <td id="T_4dd57_row4_col6" class="data row4 col6" >0.369438</td>
                        <td id="T_4dd57_row4_col7" class="data row4 col7" >0.443492</td>
            </tr>
    </tbody></table></div></div>
</div>
<p>Plotting the two summary scores against each other gives an indication of the priorities of each method. Some will be biased towards batch correction while others will favour retaining biological variation.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">()</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">metrics_scaled</span><span class="o">.</span><span class="n">plot</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span>
    <span class="n">x</span><span class="o">=</span><span class="s2">&quot;Batch&quot;</span><span class="p">,</span>
    <span class="n">y</span><span class="o">=</span><span class="s2">&quot;Bio&quot;</span><span class="p">,</span>
    <span class="n">c</span><span class="o">=</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">metrics_scaled</span><span class="p">)),</span>
    <span class="n">colormap</span><span class="o">=</span><span class="n">matplotlib</span><span class="o">.</span><span class="n">cm</span><span class="o">.</span><span class="n">get_cmap</span><span class="p">(</span><span class="s2">&quot;Set1&quot;</span><span class="p">),</span>
    <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span>
<span class="p">)</span>

<span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">metrics_scaled</span><span class="p">[[</span><span class="s2">&quot;Batch&quot;</span><span class="p">,</span> <span class="s2">&quot;Bio&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
        <span class="n">k</span><span class="p">,</span>
        <span class="n">v</span><span class="p">,</span>
        <span class="n">xytext</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="o">-</span><span class="mi">3</span><span class="p">),</span>
        <span class="n">textcoords</span><span class="o">=</span><span class="s2">&quot;offset points&quot;</span><span class="p">,</span>
        <span class="n">family</span><span class="o">=</span><span class="s2">&quot;sans-serif&quot;</span><span class="p">,</span>
        <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span>
    <span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/integration_118_0.png" src="../_images/integration_118_0.png" />
</div>
</div>
<p>In our small example scenario it looks as though both <strong>BBKNN</strong> and <strong>Seurat</strong> have removed slightly more of the batch effects but also removed more more biological variation. Both <strong>scVI</strong> and <strong>scANVI</strong> score more highly on retaining biological variation, with <strong>scANVI</strong> being slightly better.</p>
<p>To get an overall score for each method we can combine the two summary scores. The <strong>scIB</strong> paper suggests a weighting of 40% batch correction and 60% biological conservation but you may prefer to weight things differently depending on the priorities for your dataset.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">metrics_scaled</span><span class="p">[</span><span class="s2">&quot;Overall&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mf">0.4</span> <span class="o">*</span> <span class="n">metrics_scaled</span><span class="p">[</span><span class="s2">&quot;Batch&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="mf">0.6</span> <span class="o">*</span> <span class="n">metrics_scaled</span><span class="p">[</span><span class="s2">&quot;Bio&quot;</span><span class="p">]</span>
<span class="n">metrics_scaled</span><span class="o">.</span><span class="n">style</span><span class="o">.</span><span class="n">background_gradient</span><span class="p">(</span><span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;Blues&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style  type="text/css" >
#T_8dd3f_row0_col0{
            background-color:  #5fa6d1;
            color:  #000000;
        }#T_8dd3f_row0_col1,#T_8dd3f_row0_col6,#T_8dd3f_row2_col2,#T_8dd3f_row3_col3,#T_8dd3f_row3_col4,#T_8dd3f_row3_col5,#T_8dd3f_row3_col7,#T_8dd3f_row3_col8,#T_8dd3f_row4_col0,#T_8dd3f_row4_col2{
            background-color:  #f7fbff;
            color:  #000000;
        }#T_8dd3f_row0_col2,#T_8dd3f_row0_col3,#T_8dd3f_row0_col5,#T_8dd3f_row1_col0,#T_8dd3f_row1_col5,#T_8dd3f_row1_col7,#T_8dd3f_row1_col8,#T_8dd3f_row2_col4,#T_8dd3f_row2_col5,#T_8dd3f_row3_col1,#T_8dd3f_row3_col6,#T_8dd3f_row4_col5{
            background-color:  #08306b;
            color:  #f1f1f1;
        }#T_8dd3f_row0_col4{
            background-color:  #edf4fc;
            color:  #000000;
        }#T_8dd3f_row0_col7{
            background-color:  #083776;
            color:  #f1f1f1;
        }#T_8dd3f_row0_col8{
            background-color:  #084184;
            color:  #f1f1f1;
        }#T_8dd3f_row1_col1{
            background-color:  #eef5fc;
            color:  #000000;
        }#T_8dd3f_row1_col2{
            background-color:  #084990;
            color:  #f1f1f1;
        }#T_8dd3f_row1_col3{
            background-color:  #4896c8;
            color:  #000000;
        }#T_8dd3f_row1_col4{
            background-color:  #d2e3f3;
            color:  #000000;
        }#T_8dd3f_row1_col6{
            background-color:  #e3eef9;
            color:  #000000;
        }#T_8dd3f_row2_col0{
            background-color:  #e1edf8;
            color:  #000000;
        }#T_8dd3f_row2_col1{
            background-color:  #4f9bcb;
            color:  #000000;
        }#T_8dd3f_row2_col3{
            background-color:  #c3daee;
            color:  #000000;
        }#T_8dd3f_row2_col6{
            background-color:  #4b98ca;
            color:  #000000;
        }#T_8dd3f_row2_col7{
            background-color:  #68acd5;
            color:  #000000;
        }#T_8dd3f_row2_col8{
            background-color:  #63a8d3;
            color:  #000000;
        }#T_8dd3f_row3_col0{
            background-color:  #e6f0f9;
            color:  #000000;
        }#T_8dd3f_row3_col2{
            background-color:  #083e81;
            color:  #f1f1f1;
        }#T_8dd3f_row4_col1{
            background-color:  #8cc0dd;
            color:  #000000;
        }#T_8dd3f_row4_col3{
            background-color:  #add0e6;
            color:  #000000;
        }#T_8dd3f_row4_col4{
            background-color:  #3181bd;
            color:  #000000;
        }#T_8dd3f_row4_col6{
            background-color:  #eaf3fb;
            color:  #000000;
        }#T_8dd3f_row4_col7{
            background-color:  #6dafd7;
            color:  #000000;
        }#T_8dd3f_row4_col8{
            background-color:  #a8cee4;
            color:  #000000;
        }</style><table id="T_8dd3f_" ><thead>    <tr>        <th class="blank level0" ></th>        <th class="col_heading level0 col0" >ASW_label</th>        <th class="col_heading level0 col1" >ASW_label/batch</th>        <th class="col_heading level0 col2" >PCR_batch</th>        <th class="col_heading level0 col3" >isolated_label_silhouette</th>        <th class="col_heading level0 col4" >graph_conn</th>        <th class="col_heading level0 col5" >hvg_overlap</th>        <th class="col_heading level0 col6" >Batch</th>        <th class="col_heading level0 col7" >Bio</th>        <th class="col_heading level0 col8" >Overall</th>    </tr></thead><tbody>
                <tr>
                        <th id="T_8dd3f_level0_row0" class="row_heading level0 row0" >scVI</th>
                        <td id="T_8dd3f_row0_col0" class="data row0 col0" >0.538832</td>
                        <td id="T_8dd3f_row0_col1" class="data row0 col1" >0.000000</td>
                        <td id="T_8dd3f_row0_col2" class="data row0 col2" >1.000000</td>
                        <td id="T_8dd3f_row0_col3" class="data row0 col3" >1.000000</td>
                        <td id="T_8dd3f_row0_col4" class="data row0 col4" >0.052415</td>
                        <td id="T_8dd3f_row0_col5" class="data row0 col5" >1.000000</td>
                        <td id="T_8dd3f_row0_col6" class="data row0 col6" >0.350805</td>
                        <td id="T_8dd3f_row0_col7" class="data row0 col7" >0.846277</td>
                        <td id="T_8dd3f_row0_col8" class="data row0 col8" >0.648088</td>
            </tr>
            <tr>
                        <th id="T_8dd3f_level0_row1" class="row_heading level0 row1" >scANVI</th>
                        <td id="T_8dd3f_row1_col0" class="data row1 col0" >1.000000</td>
                        <td id="T_8dd3f_row1_col1" class="data row1 col1" >0.049643</td>
                        <td id="T_8dd3f_row1_col2" class="data row1 col2" >0.903648</td>
                        <td id="T_8dd3f_row1_col3" class="data row1 col3" >0.608373</td>
                        <td id="T_8dd3f_row1_col4" class="data row1 col4" >0.189377</td>
                        <td id="T_8dd3f_row1_col5" class="data row1 col5" >1.000000</td>
                        <td id="T_8dd3f_row1_col6" class="data row1 col6" >0.380890</td>
                        <td id="T_8dd3f_row1_col7" class="data row1 col7" >0.869458</td>
                        <td id="T_8dd3f_row1_col8" class="data row1 col8" >0.674031</td>
            </tr>
            <tr>
                        <th id="T_8dd3f_level0_row2" class="row_heading level0 row2" >BBKNN</th>
                        <td id="T_8dd3f_row2_col0" class="data row2 col0" >0.111505</td>
                        <td id="T_8dd3f_row2_col1" class="data row2 col1" >0.583824</td>
                        <td id="T_8dd3f_row2_col2" class="data row2 col2" >0.000000</td>
                        <td id="T_8dd3f_row2_col3" class="data row2 col3" >0.258229</td>
                        <td id="T_8dd3f_row2_col4" class="data row2 col4" >1.000000</td>
                        <td id="T_8dd3f_row2_col5" class="data row2 col5" >1.000000</td>
                        <td id="T_8dd3f_row2_col6" class="data row2 col6" >0.527941</td>
                        <td id="T_8dd3f_row2_col7" class="data row2 col7" >0.456578</td>
                        <td id="T_8dd3f_row2_col8" class="data row2 col8" >0.485123</td>
            </tr>
            <tr>
                        <th id="T_8dd3f_level0_row3" class="row_heading level0 row3" >Seurat</th>
                        <td id="T_8dd3f_row3_col0" class="data row3 col0" >0.088620</td>
                        <td id="T_8dd3f_row3_col1" class="data row3 col1" >1.000000</td>
                        <td id="T_8dd3f_row3_col2" class="data row3 col2" >0.942096</td>
                        <td id="T_8dd3f_row3_col3" class="data row3 col3" >0.000000</td>
                        <td id="T_8dd3f_row3_col4" class="data row3 col4" >0.000000</td>
                        <td id="T_8dd3f_row3_col5" class="data row3 col5" >0.000000</td>
                        <td id="T_8dd3f_row3_col6" class="data row3 col6" >0.647365</td>
                        <td id="T_8dd3f_row3_col7" class="data row3 col7" >0.029540</td>
                        <td id="T_8dd3f_row3_col8" class="data row3 col8" >0.276670</td>
            </tr>
            <tr>
                        <th id="T_8dd3f_level0_row4" class="row_heading level0 row4" >Unintegrated</th>
                        <td id="T_8dd3f_row4_col0" class="data row4 col0" >0.000000</td>
                        <td id="T_8dd3f_row4_col1" class="data row4 col1" >0.420647</td>
                        <td id="T_8dd3f_row4_col2" class="data row4 col2" >0.000030</td>
                        <td id="T_8dd3f_row4_col3" class="data row4 col3" >0.330475</td>
                        <td id="T_8dd3f_row4_col4" class="data row4 col4" >0.687636</td>
                        <td id="T_8dd3f_row4_col5" class="data row4 col5" >1.000000</td>
                        <td id="T_8dd3f_row4_col6" class="data row4 col6" >0.369438</td>
                        <td id="T_8dd3f_row4_col7" class="data row4 col7" >0.443492</td>
                        <td id="T_8dd3f_row4_col8" class="data row4 col8" >0.413870</td>
            </tr>
    </tbody></table></div></div>
</div>
<p>Let’s make a quick bar chart to visualise the overall performance.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">metrics_scaled</span><span class="o">.</span><span class="n">plot</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="n">y</span><span class="o">=</span><span class="s2">&quot;Overall&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&lt;AxesSubplot:&gt;
</pre></div>
</div>
<img alt="../_images/integration_122_1.png" src="../_images/integration_122_1.png" />
</div>
</div>
<p>As we have already seen <strong>scVI</strong> and <strong>scANVI</strong> are the best performers with <strong>scANVI</strong> scoring slightly higher. It is important to note that this is just an example of how to run these metrics for this specific dataset, not a proper evaluation of these methods. For that you should refer to existing benchmarking publications.</p>
<p>Existing benchmarks have suggested methods which generally perform well, but performance can also be quite variable across scenarios. For some analyses it may be worthwhile performing your own evaluation of integration. The <strong>scib</strong> package makes this process easier, but it can still be a significant undertaking, relying on a good knowledge of the ground truth and interpretation of the metrics.</p>
</section>
<section id="key-takeaways">
<h2><span class="section-number">12.9. </span>Key Takeaways<a class="headerlink" href="#key-takeaways" title="Permalink to this headline">#</a></h2>
<ol class="simple">
<li><p>Visualize your data before attempting to correct for batch effects to assess the extent of the issue. Batch effect correction is not always required and it might mask the biological variation of interest.</p></li>
<li><p>If cell labels are available and biological variation is the most important, the usage of scANVI is advised.</p></li>
<li><p>Consider running several integration methods on your dataset and evaluate them with the scIB metrics to use the generated embedding that is most robust for your use-case.</p></li>
</ol>
</section>
<section id="quiz">
<h2><span class="section-number">12.10. </span>Quiz<a class="headerlink" href="#quiz" title="Permalink to this headline">#</a></h2>
<ol class="simple">
<li><p>Explain the difference between batch effects and data integration challenges.</p></li>
<li><p>What are sources of batch effects?</p></li>
<li><p>What is the difference between technical and biological variation?</p></li>
<li><p>How does one evaluate whether the integration worked well or not? What are useful metrics for this purpose?</p></li>
</ol>
</section>
<section id="session-information">
<h2><span class="section-number">12.11. </span>Session information<a class="headerlink" href="#session-information" title="Permalink to this headline">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">session_info</span>

<span class="n">session_info</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/icb/luke.zappia/miniconda3/envs/bp-integration/lib/python3.8/site-packages/jupyterlab_server/app.py:31: DeprecationWarning: metadata {&#39;default&#39;: &#39;2.8.2&#39;} was set from the constructor. With traitlets 4.1, metadata should be set using the .tag() method, e.g., Int().tag(key1=&#39;value1&#39;, key2=&#39;value2&#39;)
  app_version = Unicode(&#39;&#39;, help=&#39;The version of the application.&#39;,
/home/icb/luke.zappia/miniconda3/envs/bp-integration/lib/python3.8/site-packages/session_info/main.py:213: DeprecationWarning: Accessing jsonschema.__version__ is deprecated and will be removed in a future release. Use importlib.metadata directly to query for jsonschema&#39;s version.
  mod_version = _find_version(mod.__version__)
</pre></div>
</div>
<div class="output text_html"><details>
<summary>Click to view session information</summary>
<pre>
-----
anndata             0.7.6
anndata2ri          1.0.6
bbknn               NA
matplotlib          3.4.3
numpy               1.18.1
pandas              1.2.5
rpy2                3.4.5
scanpy              1.8.1
scib                1.0.0
scvi                0.16.4
session_info        1.0.0
-----
</pre>
<details>
<summary>Click to view modules imported as dependencies</summary>
<pre>
PIL                         9.1.1
absl                        NA
annoy                       NA
anyio                       NA
attr                        21.4.0
babel                       2.10.3
backcall                    0.2.0
backports                   NA
beta_ufunc                  NA
binom_ufunc                 NA
brotli                      NA
certifi                     2022.06.15
cffi                        1.15.0
charset_normalizer          2.0.12
chex                        0.1.3
colorama                    0.4.5
cycler                      0.10.0
cython_runtime              NA
dateutil                    2.8.2
debugpy                     1.6.0
decorator                   5.1.1
defusedxml                  0.7.1
deprecate                   0.3.2
docrep                      0.3.2
dunamai                     1.12.0
entrypoints                 0.4
flatbuffers                 2.0
flax                        0.5.0
fsspec                      2022.5.0
get_version                 3.5.4
google                      NA
h5py                        2.10.0
idna                        3.3
igraph                      0.9.8
importlib_resources         NA
ipykernel                   6.4.1
ipython_genutils            0.2.0
ipywidgets                  7.6.5
jax                         0.3.5
jaxlib                      0.3.2
jedi                        0.18.1
jinja2                      3.1.2
joblib                      1.1.0
json5                       NA
jsonschema                  4.6.0
jupyter_server              1.11.0
jupyterlab_server           2.8.2
kiwisolver                  1.4.3
llvmlite                    0.38.1
louvain                     0.7.0
markupsafe                  2.1.1
matplotlib_inline           NA
mpl_toolkits                NA
msgpack                     1.0.4
multipledispatch            0.6.0
natsort                     8.1.0
nbclassic                   NA
nbformat                    5.1.3
nbinom_ufunc                NA
numba                       0.55.1
numexpr                     2.7.3
numpyro                     0.9.1
opt_einsum                  v3.3.0
optax                       0.1.2
packaging                   21.3
parso                       0.8.3
pexpect                     4.8.0
pickleshare                 0.7.5
pkg_resources               NA
prometheus_client           NA
prompt_toolkit              3.0.29
ptyprocess                  0.7.0
pvectorc                    NA
pycparser                   2.21
pydev_ipython               NA
pydevconsole                NA
pydevd                      2.8.0
pydevd_file_utils           NA
pydevd_plugins              NA
pydevd_tracing              NA
pygments                    2.12.0
pynndescent                 0.5.4
pyparsing                   3.0.9
pyro                        1.7.0
pyrsistent                  NA
pytorch_lightning           1.5.10
pytz                        2022.1
pytz_deprecation_shim       NA
requests                    2.28.0
rich                        NA
scipy                       1.7.1
seaborn                     0.11.2
send2trash                  NA
setuptools                  59.5.0
sinfo                       0.3.1
six                         1.16.0
sklearn                     1.0.2
sniffio                     1.2.0
socks                       1.7.1
sphinxcontrib               NA
statsmodels                 0.12.2
storemagic                  NA
tables                      3.6.1
tensorboard                 2.9.1
terminado                   0.15.0
texttable                   1.6.4
threadpoolctl               3.1.0
toolz                       0.11.2
torch                       1.10.2
torchmetrics                0.9.1
tornado                     6.1
tqdm                        4.64.0
traitlets                   5.3.0
tree                        0.1.7
typing_extensions           NA
tzlocal                     NA
umap                        0.5.1
urllib3                     1.26.9
wcwidth                     0.2.5
websocket                   1.3.3
yaml                        6.0
zipp                        NA
zmq                         23.1.0
</pre>
</details> <!-- seems like this ends pre, so might as well be explicit -->
<pre>
-----
IPython             7.28.0
jupyter_client      7.0.4
jupyter_core        4.8.1
jupyterlab          3.1.14
notebook            6.4.4
-----
Python 3.8.12 | packaged by conda-forge | (default, Jan 30 2022, 23:42:07) [GCC 9.4.0]
Linux-3.10.0-1160.25.1.el7.x86_64-x86_64-with-glibc2.10
-----
Session information updated at 2022-06-23 19:39
</pre>
</details></div></div>
</div>
</section>
<section id="references">
<h2><span class="section-number">12.12. </span>References<a class="headerlink" href="#references" title="Permalink to this headline">#</a></h2>
<div class="docutils container" id="id5">
<dl class="citation">
<dt class="label" id="id33"><span class="brackets"><a class="fn-backref" href="#id1">LT19</a></span></dt>
<dd><p>Malte D Luecken and Fabian J Theis. Current best practices in single-cell rna-seq analysis: a tutorial. <em>Molecular Systems Biology</em>, 15(6):e8746, 2019. URL: <a class="reference external" href="https://www.embopress.org/doi/abs/10.15252/msb.20188746">https://www.embopress.org/doi/abs/10.15252/msb.20188746</a>, <a class="reference external" href="https://arxiv.org/abs/https://www.embopress.org/doi/pdf/10.15252/msb.20188746">arXiv:https://www.embopress.org/doi/pdf/10.15252/msb.20188746</a>, <a class="reference external" href="https://doi.org/https://doi.org/10.15252/msb.20188746">doi:https://doi.org/10.15252/msb.20188746</a>.</p>
</dd>
<dt class="label" id="id158"><span class="brackets"><a class="fn-backref" href="#id3">LButtnerC+22</a></span></dt>
<dd><p>Malte D. Luecken, M. Büttner, K. Chaichoompu, A. Danese, M. Interlandi, M. F. Mueller, D. C. Strobl, L. Zappia, M. Dugas, M. Colomé-Tatché, and Fabian J. Theis. Benchmarking atlas-level data integration in single-cell genomics. <em>Nature Methods</em>, 19(1):41–50, Jan 2022. URL: <a class="reference external" href="https://doi.org/10.1038/s41592-021-01336-8">https://doi.org/10.1038/s41592-021-01336-8</a>, <a class="reference external" href="https://doi.org/10.1038/s41592-021-01336-8">doi:10.1038/s41592-021-01336-8</a>.</p>
</dd>
<dt class="label" id="id160"><span class="brackets"><a class="fn-backref" href="#id2">PZO17</a></span></dt>
<dd><p>Belinda Phipson, Luke Zappia, and Alicia Oshlack. Gene length and detection bias in single cell RNA sequencing protocols. <em>F1000Res.</em>, 6:595, April 2017.</p>
</dd>
<dt class="label" id="id159"><span class="brackets"><a class="fn-backref" href="#id4">TAC+20</a></span></dt>
<dd><p>Hoa Thi Nhu Tran, Kok Siong Ang, Marion Chevrier, Xiaomeng Zhang, Nicole Yee Shin Lee, Michelle Goh, and Jinmiao Chen. A benchmark of batch-effect correction methods for single-cell rna sequencing data. <em>Genome Biology</em>, 21(1):12, Jan 2020. URL: <a class="reference external" href="https://doi.org/10.1186/s13059-019-1850-9">https://doi.org/10.1186/s13059-019-1850-9</a>, <a class="reference external" href="https://doi.org/10.1186/s13059-019-1850-9">doi:10.1186/s13059-019-1850-9</a>.</p>
</dd>
</dl>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./cellular_structure"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
    <a class='left-prev' id="prev-link" href="clustering.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title"><span class="section-number">11. </span>Clustering</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="../trajectories/lineage_tracing.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title"><span class="section-number">13. </span>Lineage tracing</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By Theislab<br/>
  
      &copy; Copyright 2022.<br/>
    <div class="extra_footer">
      <div>
  Brought to you by
  <a href="https://www.github.com/theislab">Theislab</a>, with many thanks to the single-cell community as a whole!
</div>

    </div>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>