
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>36. Integrating AIR and transcriptomics &#8212; Single-cell best practices</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="../_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" href="../_static/styles/sphinx-book-theme.css?digest=5115cc725059bd94278eecd172e13a965bf8f5a9" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/book.css" />
    <link rel="stylesheet" type="text/css" href="../_static/design-style.b7bb847fb20b106c3d81b95245e65545.min.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?digest=9c920249402e914e316237a7dbc6769907cce411"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js"></script>
    <script async="async" kind="hypothesis" src="https://hypothes.is/embed.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../_static/book.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="37. Paired integration" href="../multimodal_integration/paired_integration.html" />
    <link rel="prev" title="35. Specificity analysis" href="specificity.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="en">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../_static/logo.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">Single-cell best practices</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Introduction
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../introduction/prior_art.html">
   1. Prior art
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../introduction/scrna_seq.html">
   2. Single-cell RNA sequencing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../introduction/raw_data_processing.html">
   3. Raw data processing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../introduction/analysis_tools.html">
   4. Analysis frameworks and tools
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../introduction/data_infrastructure.html">
   5. Data infrastructure
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../introduction/interoperability.html">
   6. Interoperability
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Preprocessing and visualization
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../preprocessing_visualization/quality_control.html">
   7. Quality Control
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../preprocessing_visualization/normalization.html">
   8. Normalization
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../preprocessing_visualization/feature_selection.html">
   9. Feature selection
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../preprocessing_visualization/dimensionality_reduction.html">
   10. Dimensionality Reduction
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Identifying cellular structure
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../cellular_structure/clustering.html">
   11. Clustering
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cellular_structure/annotation.html">
   12. Annotation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cellular_structure/integration.html">
   13. Data integration
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Inferring trajectories
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../trajectories/pseudotemporal.html">
   14. Pseudotemporal ordering
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../trajectories/rna_velocity.html">
   15. RNA velocity
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../trajectories/lineage_tracing.html">
   16. Lineage tracing
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Dealing with conditions
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../conditions/differential_gene_expression.html">
   17. Differential gene expression analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../conditions/compositional.html">
   18. Compositional analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../conditions/gsea_pathway.html">
   19. Gene set enrichment and pathway analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../conditions/perturbation_modeling.html">
   20. Perturbation modeling
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Modeling mechanisms
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../mechanisms/grns.html">
   21. Gene regulatory networks
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../mechanisms/cell_cell_communication.html">
   22. Cell-cell communication
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Deconvolution
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../deconvolution/bulk_deconvolution.html">
   23. Bulk deconvolution
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Surface protein
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/quality_control.html">
   24. Quality control
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/normalization.html">
   25. Normalization
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/doublet_detection.html">
   26. Doublet detection
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/dimensionality_reduction.html">
   27. Dimensionality Reduction
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/batch_correction.html">
   28. Batch correction
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/annotation.html">
   29. Annotation
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  AIR repertoire
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="ir_profiling.html">
   30. Immune Receptor Profiling
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="clonotype.html">
   31. Clonotype analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="specificity.html">
   35. Specificity analysis
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   36. Integrating AIR and transcriptomics
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multimodal integration
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../multimodal_integration/paired_integration.html">
   37. Paired integration
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../multimodal_integration/advanced_integration.html">
   38. Advanced integration
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Reproducibility
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../reproducibility/introduction.html">
   39. Reproducibility
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Outlook
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../outlook.html">
   40. Outlook
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Acknowledgements
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../acknowledgements.html">
   41. Acknowledgements
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Glossary
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../glossary.html">
   42. Glossary
  </a>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
            <div class="navbar_extra_footer">
            Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
            </div>
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right">
<div class="menu-dropdown menu-dropdown-launch-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Launch interactive content">
      <i class="fas fa-rocket"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://mybinder.org/v2/gh/theislab/single-cell-best-practices/master?urlpath=tree/air_repertoire/multimodal_integration.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Launch on Binder"
>
  

<span class="headerbtn__icon-container">
  
    <img src="../_static/images/logo_binder.svg">
  </span>
<span class="headerbtn__text-container">Binder</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

<button onclick="toggleFullScreen()"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="bottom"
title="Fullscreen mode"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>

<div class="menu-dropdown menu-dropdown-repository-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Source repositories">
      <i class="fab fa-github"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://github.com/theislab/single-cell-best-practices"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Source repository"
>
  

<span class="headerbtn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="headerbtn__text-container">repository</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/theislab/single-cell-best-practices/issues/new?title=Issue%20on%20page%20%2Fair_repertoire/multimodal_integration.html&body=Your%20issue%20content%20here."
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Open an issue"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="headerbtn__text-container">open issue</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/theislab/single-cell-best-practices/edit/master/air_repertoire/multimodal_integration.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Edit this page"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="headerbtn__text-container">suggest edit</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

<div class="menu-dropdown menu-dropdown-download-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Download this page">
      <i class="fas fa-download"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="../_sources/air_repertoire/multimodal_integration.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Download source file"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="headerbtn__text-container">.ipynb</span>
</a>

      </li>
      
      <li>
        
<button onclick="printPdf(this)"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="left"
title="Print to PDF"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="headerbtn__text-container">.pdf</span>
</button>

      </li>
      
    </ul>
  </div>
</div>
<label for="__page-toc"
  class="headerbtn headerbtn-page-toc"
  
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-list"></i>
  </span>

</label>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
    <div class="tocsection onthispage pt-5 pb-3">
        <i class="fas fa-list"></i> Contents
    </div>
    <nav id="bd-toc-nav" aria-label="Page">
        <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#motivation">
   36.1. Motivation
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#data-preparation">
   36.2. Data Preparation
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#uni-modal-analysis-with-multimodal-conditions">
   36.3. Uni-modal Analysis with multimodal conditions
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#air-analysis-on-rna-clusters">
     36.3.1. AIR analysis on RNA clusters
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#gex-analysis-on-air-clusters">
     36.3.2. GEX analysis on AIR clusters
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#multimodal-integration">
   36.4. Multimodal Integration
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#integrating-tcr-and-gex">
     36.4.1. Integrating TCR and GEX
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#tessa">
       36.4.1.1. TESSA
      </a>
      <ul class="nav section-nav flex-column">
       <li class="toc-h5 nav-item toc-entry">
        <a class="reference internal nav-link" href="#data-preprocessing">
         36.4.1.1.1. Data Preprocessing
        </a>
       </li>
       <li class="toc-h5 nav-item toc-entry">
        <a class="reference internal nav-link" href="#running-the-model">
         36.4.1.1.2. Running the model
        </a>
       </li>
       <li class="toc-h5 nav-item toc-entry">
        <a class="reference internal nav-link" href="#output">
         36.4.1.1.3. Output
        </a>
       </li>
      </ul>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#conga">
       36.4.1.2. CoNGA
      </a>
      <ul class="nav section-nav flex-column">
       <li class="toc-h5 nav-item toc-entry">
        <a class="reference internal nav-link" href="#data-preperation">
         36.4.1.2.1. Data Preperation
        </a>
       </li>
       <li class="toc-h5 nav-item toc-entry">
        <a class="reference internal nav-link" href="#executing-conga">
         36.4.1.2.2. Executing CoNGA
        </a>
       </li>
       <li class="toc-h5 nav-item toc-entry">
        <a class="reference internal nav-link" href="#id5">
         36.4.1.2.3. Output
        </a>
       </li>
      </ul>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#mvtcr">
       36.4.1.3. mvTCR
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#integrating-bcr-and-gex">
     36.4.2. Integrating BCR and GEX
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#benisse">
       36.4.2.1. Benisse
      </a>
      <ul class="nav section-nav flex-column">
       <li class="toc-h5 nav-item toc-entry">
        <a class="reference internal nav-link" href="#id10">
         36.4.2.1.1. Running the model
        </a>
       </li>
       <li class="toc-h5 nav-item toc-entry">
        <a class="reference internal nav-link" href="#id11">
         36.4.2.1.2. Output
        </a>
       </li>
      </ul>
     </li>
    </ul>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#key-takeaways">
   36.5. Key Takeaways
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#questions">
   36.6. Questions
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#references">
   36.7. References
  </a>
 </li>
</ul>

    </nav>
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Integrating AIR and transcriptomics</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#motivation">
   36.1. Motivation
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#data-preparation">
   36.2. Data Preparation
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#uni-modal-analysis-with-multimodal-conditions">
   36.3. Uni-modal Analysis with multimodal conditions
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#air-analysis-on-rna-clusters">
     36.3.1. AIR analysis on RNA clusters
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#gex-analysis-on-air-clusters">
     36.3.2. GEX analysis on AIR clusters
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#multimodal-integration">
   36.4. Multimodal Integration
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#integrating-tcr-and-gex">
     36.4.1. Integrating TCR and GEX
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#tessa">
       36.4.1.1. TESSA
      </a>
      <ul class="nav section-nav flex-column">
       <li class="toc-h5 nav-item toc-entry">
        <a class="reference internal nav-link" href="#data-preprocessing">
         36.4.1.1.1. Data Preprocessing
        </a>
       </li>
       <li class="toc-h5 nav-item toc-entry">
        <a class="reference internal nav-link" href="#running-the-model">
         36.4.1.1.2. Running the model
        </a>
       </li>
       <li class="toc-h5 nav-item toc-entry">
        <a class="reference internal nav-link" href="#output">
         36.4.1.1.3. Output
        </a>
       </li>
      </ul>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#conga">
       36.4.1.2. CoNGA
      </a>
      <ul class="nav section-nav flex-column">
       <li class="toc-h5 nav-item toc-entry">
        <a class="reference internal nav-link" href="#data-preperation">
         36.4.1.2.1. Data Preperation
        </a>
       </li>
       <li class="toc-h5 nav-item toc-entry">
        <a class="reference internal nav-link" href="#executing-conga">
         36.4.1.2.2. Executing CoNGA
        </a>
       </li>
       <li class="toc-h5 nav-item toc-entry">
        <a class="reference internal nav-link" href="#id5">
         36.4.1.2.3. Output
        </a>
       </li>
      </ul>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#mvtcr">
       36.4.1.3. mvTCR
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#integrating-bcr-and-gex">
     36.4.2. Integrating BCR and GEX
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#benisse">
       36.4.2.1. Benisse
      </a>
      <ul class="nav section-nav flex-column">
       <li class="toc-h5 nav-item toc-entry">
        <a class="reference internal nav-link" href="#id10">
         36.4.2.1.1. Running the model
        </a>
       </li>
       <li class="toc-h5 nav-item toc-entry">
        <a class="reference internal nav-link" href="#id11">
         36.4.2.1.2. Output
        </a>
       </li>
      </ul>
     </li>
    </ul>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#key-takeaways">
   36.5. Key Takeaways
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#questions">
   36.6. Questions
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#references">
   36.7. References
  </a>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                
  <div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># 1. Introduction                 =&gt; Done
# 2. Groups                       =&gt; Done
# 2.1 Apply TCR on RNA clusters   =&gt; Done
# 2.2 Apply RNA on TCR clusters   =&gt; TODO wait for conditions of previous chapters
# 3. Multimodal integration       =&gt; Done
# 3.1. TCR                        =&gt; Done
# 3.1.1 TESSA                     =&gt; Done
# 3.1.2 CONGA                     =&gt; Needs refactoring
# 3.1.3 mvTCR                     =&gt; Done
# 3.2. BCR                        =&gt; Done
# 3.2.1. Benisse                  =&gt; Done
# 4. Take Home messages           =&gt; Done
# 5. Quiz                         =&gt; Done
</pre></div>
</div>
</div>
</div>
<section class="tex2jax_ignore mathjax_ignore" id="integrating-air-and-transcriptomics">
<h1><span class="section-number">36. </span>Integrating AIR and transcriptomics<a class="headerlink" href="#integrating-air-and-transcriptomics" title="Permalink to this headline">#</a></h1>
<section id="motivation">
<h2><span class="section-number">36.1. </span>Motivation<a class="headerlink" href="#motivation" title="Permalink to this headline">#</a></h2>
<p>With the development of novel single cell technology, IR sequencing is combined often with other omics layers such as with transcriptomics. For B- and T-cells this enables the analysis of various characteristics on a single cell level: while the transcriptome provides insights into the current state of the cell, the IR is indicative for the cell’s specificity and thereby explains the cell’s fate upon infection or vaccination.</p>
</section>
<section id="data-preparation">
<h2><span class="section-number">36.2. </span>Data Preparation<a class="headerlink" href="#data-preparation" title="Permalink to this headline">#</a></h2>
<p>First, we will combine the GEX with AIR data. Note, that we can do this also at earlier stages of single cell analysis: e.g. we can fuse both modalities already before preprocessing to filter GEX doublets. When fusing GEX and AIR data, a left join is often performed. I.e., only cells with GEX are kept for analysis. Therefore, the order of fusing and filtering is not of great importance. However, it can be convenient to visualize the AIR information in an GEX UMAP in early stages. E.g. missing AIR information in clusters can help during cell type assignment.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>import warnings

warnings.filterwarnings(
    &quot;ignore&quot;,
    &quot;.*IProgress not found*&quot;,
)
warnings.simplefilter(action=&quot;ignore&quot;, category=FutureWarning)

import scanpy as sc
import scirpy as ir
import pandas as pd
import numpy as np
import seaborn as sb
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>path_data = &quot;./data&quot;
path_gex_tcr = f&quot;{path_data}/TCR_00_GEX.h5ad&quot;
path_gex_bcr = f&quot;{path_data}/BCR_00_GEX.h5ad&quot;
path_tcr = f&quot;{path_data}/TCR_01_preprocessed.h5ad&quot;
path_bcr = f&quot;{path_data}/BCR_01_preprocessed.h5ad&quot;

path_tmp = f&quot;{path_data}/tmp&quot;
path_res = f&quot;{path_data}/res&quot;
</pre></div>
</div>
</div>
</div>
<p>First we will load the Gene expression data processed by the authors. You can download the full dataset from <a class="reference external" href="https://www.ebi.ac.uk/arrayexpress/files/E-MTAB-10026/E-MTAB-10026.processed.4.zip">https://www.ebi.ac.uk/arrayexpress/files/E-MTAB-10026/E-MTAB-10026.processed.4.zip</a> . Since we are working only with the B- and T-cells of selected donors, we provide a downsampled dataset for you:</p>
<p>! wget -O $path_bcr_input -nc # todo file link <a class="reference external" href="https://figshare.com/ndownloader/files/35574338">https://figshare.com/ndownloader/files/35574338</a></p>
<p>! wget -O $path_bcr_input -nc # todo file link <a class="reference external" href="https://figshare.com/ndownloader/files/35574338">https://figshare.com/ndownloader/files/35574338</a></p>
<p>Let’s load the GEX and the previously annotated TCR data. Both AnnData objects can then easily be merged via scirpy into a shared AnnData object, which will store IR information in adata.obs and GEX in <code class="docutils literal notranslate"><span class="pre">adata.X</span></code>. Here a left join is conducted: only AIRs from cells, where there is scRNA data, are added to the AnnData object. While the dataset can include cells other than B- and T-cells and, therefore, AIRs without scRNA-seq indicate missing values. Since a majority of downstream analysis is conducted on transcriptomics level, these cells are excluded.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata_tc = sc.read(path_gex_tcr)
adata_tcr = sc.read(path_tcr)

ir.pp.merge_with_ir(adata_tc, adata_tcr)
</pre></div>
</div>
</div>
</div>
<p>Let’s calculate Leiden clustering on the T cell transcriptome.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pp.neighbors(adata_tc)
sc.tl.leiden(adata_tc)
</pre></div>
</div>
</div>
</div>
<p>To get an overview, we will plot the data with cluster assigment as a UMAP visualisation.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.tl.umap(adata_tc)
sc.pl.umap(adata_tc, color=[&quot;full_clustering&quot;, &quot;has_ir&quot;, &quot;leiden&quot;], ncols=1)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/multimodal_integration_14_0.png" src="../_images/multimodal_integration_14_0.png" />
</div>
</div>
<p>Here, we can see, that the majority of cells belong to the CD8+ effector T cells. Also, most of the cells express an TCR. The presence of AIRs can guide cell type annotation to separate clusters of other immune cells from T- and B-cells. Additional, missing AIRs can also be used for quality control: e.g. if the data is sorted for T cells, but large clusters miss TCRs, it should be investigated on a transcriptomic level, whether these cells were wrongly sorted.</p>
<p>Similar, we merge with the BCR and GEX data for B cells.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata_bc = sc.read(path_gex_bcr)
adata_bcr = sc.read(path_bcr)

ir.pp.merge_with_ir(adata_bc, adata_bcr)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pp.neighbors(adata_bc)
sc.tl.leiden(adata_bc)
sc.tl.umap(adata_bc)
sc.pl.umap(adata_bc, color=[&quot;full_clustering&quot;, &quot;has_ir&quot;, &quot;leiden&quot;], ncols=1)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/multimodal_integration_18_0.png" src="../_images/multimodal_integration_18_0.png" />
</div>
</div>
<p>In this visualization, we see a clear separation between Plasmablasts and B cells.</p>
</section>
<section id="uni-modal-analysis-with-multimodal-conditions">
<h2><span class="section-number">36.3. </span>Uni-modal Analysis with multimodal conditions<a class="headerlink" href="#uni-modal-analysis-with-multimodal-conditions" title="Permalink to this headline">#</a></h2>
<p>While studies often provide paired measurements for both modalities, these are often analyzed individually utilizing only limited shared information. Often one modality (VDJ or transcriptome) is used to provide conditions, on which the other modality is then analyzed. E.g. we can observe how cells from the same clonal lineage adapt to perturbation by DEG analysis of these cells in different points in time.</p>
<p>Previously, we covered these unimodal analysis in other sections of this book. Here, we will focus, on how we can apply these techniques across VDJ and other modalities, rather than explaining the underlying approaches. To find a more detailed explanation please refer to the corresponding chapters.</p>
<section id="air-analysis-on-rna-clusters">
<h3><span class="section-number">36.3.1. </span>AIR analysis on RNA clusters<a class="headerlink" href="#air-analysis-on-rna-clusters" title="Permalink to this headline">#</a></h3>
<p>We have shown sequence and diversity analysis of AIRs in chapter (LINK UNIMODAL SEQUENCE ANALYSIS). When paired VDJ- and transcriptome data is available, we can perform similar analysis on cell clusters defined in the RNA-space. How these RNA-clusters are defined, is depending on the study design. Selected ways to define RNA clusters for AIR analysis:</p>
<ul class="simple">
<li><p><strong>Leiden Clusters</strong>: Arguably the easiest way to define subsets of cells within count data, is via clustering algorithms. Usually, Leiden clustering is performed and highly expressed genes of the clusters are used to determine the cells’ state. We can now conduct the TCR analysis on clusters of interest (e.g. an activated cluster that is likely disease-specific).</p></li>
<li><p><strong>Marker Genes</strong>: We can select AIR cells based on specific marker genes, that are indicative for specific cell states. Depending on your research questions, you should select your marker genes accordingly (e.g. markers for proliferation, activation, suppression). Naturally, this also extends to scores defined on collections of marker genes.</p></li>
<li><p><strong>Cell Types</strong>: Several sub cell types exist for B- and T-Cells for various degrees of hierarchy. After these cell types are annotated via transcriptome or other modalities, we can investigate how the repertoire composition differs among this label. This analysis can either be performed within sample (how does the repertoire differ between cell types) or within a cell type across samples (e.g. how does the repertoire of CD8 effector T-cells change between disease and healthy patients).</p></li>
</ul>
<p><strong>Leiden Clusters</strong>: Above we clustered the gene expression data via the Leiden algorithm. We can now use the resulting groups to performance any kind of sequence analysis, e.g. spectratyping. For that, we simply replace the entry which defines the group parameter (here: color) with the column name of the Leiden clustering (“leiden”).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>ir.pl.spectratype(
    adata_bc,
    color=&quot;leiden&quot;,
    viztype=&quot;curve&quot;,
    curve_layout=&quot;shifted&quot;,
    fig_kws={&quot;figsize&quot;: [8, 4]},
    kde_kws={&quot;kde_norm&quot;: False},
)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/icb/felix.drost/miniconda3/envs/best_practice/lib/python3.8/site-packages/scirpy/pl/base.py:262: UserWarning: FixedFormatter should only be used together with FixedLocator
  ax.set_yticklabels(order)
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&lt;AxesSubplot:title={&#39;center&#39;:&#39;Spectratype of IR_VJ_1_junction_aa by leiden&#39;}, xlabel=&#39;IR_VJ_1_junction_aa length&#39;&gt;
</pre></div>
</div>
<img alt="../_images/multimodal_integration_23_2.png" src="../_images/multimodal_integration_23_2.png" />
</div>
</div>
<p>The BCR length of most Leiden clusters follows a similar distribution. However, it is clear to see, that cluster 5 and 9 mainly contain BCRs with longer sequence length. These cells stem from a subcluster of Plasmablasts and Plasma cells. A diverging length profile can be caused by an immune reaction. It might therefore be useful to further investigate these clusters. #update interpretation after subsample decision.</p>
<p><strong>Marker Genes</strong>: Next, we can define our groups based on individual genes. As an example, we here pick Interferon Gamma as a marker for pro-inflammatory immune response. This will provide us with a selection of cells which are currently initiating an immune response.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pl.umap(adata_tc, color=&quot;IFNG&quot;)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/multimodal_integration_26_0.png" src="../_images/multimodal_integration_26_0.png" />
</div>
</div>
<p>When visualizing it, we observe, that it is present in a subset of CD8 effector T cells. Next, we will select cells with high IFNg score (arbitrary threshold: 2.5) and annotate this selection in the AnnData object. To use it as category, we will store it as string. We can visualize the selected cells in the UMAP.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata_tc.obs[&quot;elevated_IFNG&quot;] = adata_tc[:, &quot;IFNG&quot;].X.todense() &gt; 3
adata_tc.obs[&quot;elevated_IFNG&quot;] = adata_tc.obs[&quot;elevated_IFNG&quot;].astype(str)
sc.pl.umap(adata_tc, color=&quot;elevated_IFNG&quot;, groups=&quot;True&quot;)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>... storing &#39;elevated_IFNG&#39; as categorical
</pre></div>
</div>
<img alt="../_images/multimodal_integration_28_1.png" src="../_images/multimodal_integration_28_1.png" />
</div>
</div>
<p>Next, we will generate frequency plots to see, whether the TCRs from cells with elevated IFNg expression, follow specific sequence patterns compared to the remaining cells. First, we select all cells with annotated CDR3β, followed by separation into background cells and cells with elevated IFNg. Finally, we construct the motifs and plot them.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from palmotif import compute_pal_motif, mpl_logo, svg_logo
from IPython.display import SVG

df_sequences = adata_tc[~adata_tc.obs[&quot;IR_VDJ_1_junction_aa&quot;].isna()].obs[
    [&quot;IR_VDJ_1_junction_aa&quot;, &quot;elevated_IFNG&quot;]
]

seqs_elevated = df_sequences[df_sequences[&quot;elevated_IFNG&quot;] == &quot;True&quot;][
    &quot;IR_VDJ_1_junction_aa&quot;
]
seqs_background = df_sequences[df_sequences[&quot;elevated_IFNG&quot;] == &quot;False&quot;][
    &quot;IR_VDJ_1_junction_aa&quot;
]
centroid = df_sequences[&quot;IR_VDJ_1_junction_aa&quot;].value_counts().index[0]

motif_ifng, _ = compute_pal_motif(
    seqs=seqs_elevated,
    centroid=centroid,
)
motif_back, _ = compute_pal_motif(
    seqs=seqs_background,
    centroid=centroid,
)

svg_logo(motif_ifng, &quot;tmp/motif_ifng.svg&quot;)
svg_logo(motif_back, &quot;tmp/motif_back.svg&quot;)
display(SVG(&quot;tmp/motif_ifng.svg&quot;))
display(SVG(&quot;tmp/motif_back.svg&quot;))
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/multimodal_integration_30_0.svg" src="../_images/multimodal_integration_30_0.svg" /><img alt="../_images/multimodal_integration_30_1.svg" src="../_images/multimodal_integration_30_1.svg" /></div>
</div>
<p>While the frequency of amino acids at specific positions changes slightly, no large clear differing motifs can be seen between the two cell sets. Therefore, there doesn’t exist any evidence, that the selected subset of T cells share specificity to the same target epitope.</p>
<p><strong>Cell Type</strong>: Here, we apply the diversity between different cell types annotated on the transcriptomic level (stored in “full_clustering”). We see, that proliferating and effector CD8+ T cells have the lowest diversity. Since these cell types contain larger clones, this was to be expected. Such an analysis could be used as a sanity check for the cell type assignment or to compare the diversity of cell types over multiple studies.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># TODO: remove clonotype definition once this is handled by clonotype chapter
# ir.pl.alpha_diversity(adata_tc, groupby=&quot;full_clustering&quot;, target_col=&quot;clone_id&quot;, figsize=[8, 4])
</pre></div>
</div>
</div>
</div>
</section>
<section id="gex-analysis-on-air-clusters">
<h3><span class="section-number">36.3.2. </span>GEX analysis on AIR clusters<a class="headerlink" href="#gex-analysis-on-air-clusters" title="Permalink to this headline">#</a></h3>
<p>Similarly, we can perform the standard analysis of transcriptome (LINK UNIMODAL GEX ANALYSIS), but on groups derived in the AIR space. E.g.:</p>
<ul class="simple">
<li><p><strong>Clonotype</strong>: AIR clonotypes can be used as clusters, if sufficient clonal expansion is present. This is especially intriguing, since the clonotype serves as a proxy for the cells specificity. Cells recognizing the same antigen can therefore be tracked over conditions such as time, treatment, or stimulation, to observe their change in</p></li>
<li><p><strong>Clonotype Network</strong>: To increase the amount of cells to track, we can use networks of clonotypes with similar AIR sequence instead of clonotypes. While these cells are not ancestrally related, they are likely to share the same specificity.</p></li>
<li><p><strong>Disease Specificity</strong>: We previously queried databases, to detect epitope specific AIRs. We can now use this annotation, to investigate, how B- or T-cells reactive to selected diseases differ from other cells. Here, we must be cautious, since the dataset query provides us only with positive annotation: by same or similar AIR, we identify cells specific to a certain epitope. However, a missing pair of AIR and epitope in a database is not indicative of this binding being impossible.</p></li>
</ul>
<p>As for the previous subsection, the selected condition and analysis are highly depended on your research question. In the previous notebooks, we annotated these conditions in <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code>. Since they can, hence, be applied for DEG analysis in the same fashion, we will only showcase DEG by &lt;what_ever_works_best&gt;.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.tl.rank_genes_groups(adata_tc, groupby=&quot;elevated_IFNG&quot;)
sc.pl.rank_genes_groups(adata_tc, groupby=&quot;elevated_IFNG&quot;)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>WARNING: Default of the method has been changed to &#39;t-test&#39; from &#39;t-test_overestim_var&#39;
</pre></div>
</div>
<img alt="../_images/multimodal_integration_35_1.png" src="../_images/multimodal_integration_35_1.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># TODO adapt to disease specific cells =&gt; Covid B cells against rest
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="multimodal-integration">
<h2><span class="section-number">36.4. </span>Multimodal Integration<a class="headerlink" href="#multimodal-integration" title="Permalink to this headline">#</a></h2>
<p>In the previous section we have shown, how modality specific analysis is performed on clusters derived from other modalities. However, this underutilizes the additional information provided by paired data, since AIRs and gene expression are interlinked: adaptive immune cells recognizing the same epitopes, will undergo a similar development upon activation. As for other Omics combinations, there has been an interest lately in integrating AIR and GEX for deriving clusters or a shared representation.</p>
<p>These models are still novel and are not part of a standard “best-practice” analysis pipelines. We assume that these or similar approaches will be increasingly used in the future and therefore want to showcase these integration models as an outlook. However, you should be careful when applying these models, since they are still experimental and not independently evaluated for their strengths and weaknesses.</p>
<section id="integrating-tcr-and-gex">
<h3><span class="section-number">36.4.1. </span>Integrating TCR and GEX<a class="headerlink" href="#integrating-tcr-and-gex" title="Permalink to this headline">#</a></h3>
<p>Only recently, three methods were developed to jointly utilize TCR sequences and transcriptome each aiming at different aspects of analysis:</p>
<ul class="simple">
<li><p><strong>TESSA</strong>: applies a Bayesian model on a TCR and GEX embedding for clustering T cells of similar transcriptome and sequences.</p></li>
<li><p><strong>Conga</strong>: clusters T cell clones through a graph-theoretic approach by distances of GEX and TCR</p></li>
<li><p><strong>mvTCR</strong>: uses Deep Variational Autoencoders to derive a shared embedding of GEX and TCR.</p></li>
</ul>
<p>All multimodal approaches, require paired GEX and AIR data. We will therefore start by filtering all cells without CDR3β.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata_tc = adata_tc[~adata_tc.obs[&quot;IR_VDJ_1_junction_aa&quot;].isna()].copy()
</pre></div>
</div>
</div>
</div>
<section id="tessa">
<h4><span class="section-number">36.4.1.1. </span>TESSA<a class="headerlink" href="#tessa" title="Permalink to this headline">#</a></h4>
<p>Zhang et al. developed TCR functional landscape estimation supervised with scRNA-seq analysis (TESSA) <span id="id1">[<a class="reference internal" href="#id424" title="Ze Zhang, Danyi Xiong, Xinlei Wang, Hongyu Liu, and Tao Wang. Mapping the functional landscape of t cell receptor repertoires by single-t cell transcriptomics. Nature methods, 18(1):92–99, 2021.">Zhang <em>et al.</em>, 2021</a>]</span>, which aims at embedding and clustering T cell clones based on their TCR sequence and transcriptome via Bayesian modelling. The CDR3β sequence is first compressed to a 30-dimensional numeric representation using a pretrained autoencoder. Following, the dimensions are upweighted to correlate the TCR representation with the gene expression of similar TCR-groups, thereby assigning importance of TCR position for explaining the cells’ gene expression. In an interactive process, weights and groups are updated until convergence to reach a maximal alignment between both modalities.</p>
<p>TESSA produced clusters of high purity when embedding T-cells with known epitope specificity from {cite}’10x2019new’, surpasing the uni-modal model GLIPH <span id="id2">[<a class="reference internal" href="specificity.html#id415" title="Jacob Glanville, Huang Huang, Allison Nau, Olivia Hatton, Lisa E Wagar, Florian Rubelt, Xuhuai Ji, Arnold Han, Sheri M Krams, Christina Pettus, and others. Identifying specificity groups in the t cell receptor repertoire. Nature, 547(7661):94–98, 2017.">Glanville <em>et al.</em>, 2017</a>]</span>, which is commonly used for clustering TCR sequences. Further, cluster centrally was indicative for higher avidity clones shown by clonal expansion and high ADT counts. Using TESSA on data from <span id="id3">[<a class="reference internal" href="#id425" title="Kathryn E Yost, Ansuman T Satpathy, Daniel K Wells, Yanyan Qi, Chunlin Wang, Robin Kageyama, Katherine L McNamara, Jeffrey M Granja, Kavita Y Sarin, Ryanne A Brown, and others. Clonal replacement of tumor-specific t cells following pd-1 blockade. Nature medicine, 25(8):1251–1259, 2019.">Yost <em>et al.</em>, 2019</a>]</span>, the author detected novel clusters of responder T cell in patients undergoing PD-1 blockade.</p>
<p>The code and instructions for installation can be found <a class="reference external" href="https://github.com/jcao89757/TESSA">here</a>.</p>
<section id="data-preprocessing">
<h5><span class="section-number">36.4.1.1.1. </span>Data Preprocessing<a class="headerlink" href="#data-preprocessing" title="Permalink to this headline">#</a></h5>
<p>TESSA requires several TCR and GEX files in specific format.</p>
<p>First, we save a “.csv”-File with the following columns:</p>
<ul class="simple">
<li><p><strong>contig_id</strong>: cell barcodes as index column</p></li>
<li><p><strong>cdr3</strong>: CDR3β amino acid sequence trimmed of starting C and ending F</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>df_tcr = adata_tc.obs
df_tcr.index.name = &quot;contig_id&quot;

# trimm cdr3 sequence
df_tcr[&quot;cdr3&quot;] = [seq[1:-1] for seq in df_tcr[&quot;IR_VDJ_1_junction_aa&quot;]]

# select only columns needed
df_tcr = df_tcr[[&quot;cdr3&quot;]]

df_tcr.to_csv(f&quot;{path_tmp}/TESSA_tcrs.csv&quot;)
df_tcr.head(5)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>cdr3</th>
    </tr>
    <tr>
      <th>contig_id</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>S11_AAACCTGAGATTACCC-1</th>
      <td>ASSLDARDRGRVTEAF</td>
    </tr>
    <tr>
      <th>S11_AAACCTGGTAATTGGA-1</th>
      <td>ASSPGTGTYGYT</td>
    </tr>
    <tr>
      <th>S11_AAACCTGGTAGCACGA-1</th>
      <td>ASSIPGAVHEQY</td>
    </tr>
    <tr>
      <th>S11_AAACCTGGTCTCAACA-1</th>
      <td>ASSLDARDRGRVTEAF</td>
    </tr>
    <tr>
      <th>S11_AAACCTGTCGCCGTGA-1</th>
      <td>ASSPQTGVARYGYT</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>We will filter the 10% most variable genes, as done in the TESSA publication.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>n_genes = adata_tc.shape[1] // 10
sc.pp.highly_variable_genes(adata_tc, n_top_genes=n_genes)
adata_tessa = adata_tc[:, adata_tc.var[&quot;highly_variable&quot;]].copy()
</pre></div>
</div>
</div>
</div>
<p>We now store the GEX matrix to a “.csv”-File. TESSA requires the transposed matrix of the adata object. I.e. the rows represent the different genes and the columns indicate a specific cell. For this, we will store the count data in a DataFrame with the right index and column names. Following, we transpose the DataFrame.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>count_mat = adata_tessa.X.A
df_counts = pd.DataFrame(count_mat)

df_counts.index = adata_tessa.obs.index
df_counts.index.name = &quot;&quot;
df_counts.columns = adata_tessa.var.index

df_counts = df_counts.transpose()

df_counts.to_csv(f&quot;{path_tmp}/TESSA_gex.csv&quot;)
df_counts.head()
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>S11_AAACCTGAGATTACCC-1</th>
      <th>S11_AAACCTGGTAATTGGA-1</th>
      <th>S11_AAACCTGGTAGCACGA-1</th>
      <th>S11_AAACCTGGTCTCAACA-1</th>
      <th>S11_AAACCTGTCGCCGTGA-1</th>
      <th>S11_AAACCTGTCTTATCTG-1</th>
      <th>S11_AAACGGGCAAGAAAGG-1</th>
      <th>S11_AAACGGGGTAGCGATG-1</th>
      <th>S11_AAACGGGGTCGGCACT-1</th>
      <th>S11_AAACGGGGTTACGGAG-1</th>
      <th>...</th>
      <th>S12_TTTGGTTGTCAGAGGT-1</th>
      <th>S12_TTTGGTTTCAAGAAGT-1</th>
      <th>S12_TTTGGTTTCTACTATC-1</th>
      <th>S12_TTTGGTTTCTCGCTTG-1</th>
      <th>S12_TTTGTCACACGGTAGA-1</th>
      <th>S12_TTTGTCAGTAGAAAGG-1</th>
      <th>S12_TTTGTCAGTCCAACTA-1</th>
      <th>S12_TTTGTCAGTGAGTGAC-1</th>
      <th>S12_TTTGTCATCCACTCCA-1</th>
      <th>S12_TTTGTCATCGCATGAT-1</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>KLHL17</th>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.00000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0000</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.00000</td>
      <td>0.000000</td>
      <td>...</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>1.662003</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>HES4</th>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.00000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0000</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.00000</td>
      <td>0.000000</td>
      <td>...</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>ISG15</th>
      <td>1.845827</td>
      <td>0.0</td>
      <td>1.54457</td>
      <td>2.441434</td>
      <td>2.430966</td>
      <td>0.9284</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1.64661</td>
      <td>1.565189</td>
      <td>...</td>
      <td>1.210719</td>
      <td>0.0</td>
      <td>1.581859</td>
      <td>1.511785</td>
      <td>0.000000</td>
      <td>2.015501</td>
      <td>0.0</td>
      <td>2.088418</td>
      <td>1.611682</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>AGRN</th>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.00000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0000</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.00000</td>
      <td>0.000000</td>
      <td>...</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>TTLL10</th>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.00000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0000</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.00000</td>
      <td>0.000000</td>
      <td>...</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 5281 columns</p>
</div></div></div>
</div>
</section>
<section id="running-the-model">
<h5><span class="section-number">36.4.1.1.2. </span>Running the model<a class="headerlink" href="#running-the-model" title="Permalink to this headline">#</a></h5>
<p>We need to provide different setting options, most of which specify input or output directories. We will summarize them in a dictionary first.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>settings_full = {
    # Input files
    &quot;tcr&quot;: f&quot;{path_tmp}/TESSA_tcrs.csv&quot;,
    &quot;exp&quot;: f&quot;{path_tmp}/TESSA_gex.csv&quot;,
    # TESSA models
    &quot;model&quot;: f&quot;TESSA/BriseisEncoder/TrainedEncoder.h5&quot;,
    &quot;embeding_vectors&quot;: f&quot;TESSA/BriseisEncoder/Atchley_factors.csv&quot;,
    # Output files
    &quot;output_TCR&quot;: f&quot;{path_res}/TESSA_tcr_embedding.csv&quot;,
    &quot;output_log&quot;: f&quot;{path_res}/TESSA_log.log&quot;,
    &quot;output_tessa&quot;: f&quot;{path_res}&quot;,
    &quot;within_sample_networks&quot;: &quot;FALSE&quot;,
}
</pre></div>
</div>
</div>
</div>
<p>In the next step, we will create the command for running TESSA by adding setting key and setting option.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cmd_tessa = &quot;source ~/.bashrc &amp;&amp;&quot;
cmd_tessa += &quot;conda activate TESSA &amp;&amp; &quot;

cmd_tessa += &quot;python TESSA/Tessa_main.py&quot;
for key, value in settings_full.items():
    cmd_tessa += f&quot; -{key} {value}&quot;
cmd_tessa
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&#39;source ~/.bashrc &amp;&amp;conda activate TESSA &amp;&amp; python TESSA/Tessa_main.py -tcr ./data/tmp/TESSA_tcrs.csv -exp ./data/tmp/TESSA_gex.csv -model TESSA/BriseisEncoder/TrainedEncoder.h5 -embeding_vectors TESSA/BriseisEncoder/Atchley_factors.csv -output_TCR ./data/res/TESSA_tcr_embedding.csv -output_log ./data/res/TESSA_log.log -output_tessa ./data/res -within_sample_networks FALSE&#39;
</pre></div>
</div>
</div>
</div>
<p>Finally, we run the model on the specified settings by calling the command. Depending on your compute ressources this might take several minutes.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>!$cmd_tessa
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Using TensorFlow backend.
/home/icb/felix.drost/miniconda3/envs/TESSA/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:523: FutureWarning: Passing (type, 1) or &#39;1type&#39; as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / &#39;(1,)type&#39;.
  _np_qint8 = np.dtype([(&quot;qint8&quot;, np.int8, 1)])
/home/icb/felix.drost/miniconda3/envs/TESSA/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:524: FutureWarning: Passing (type, 1) or &#39;1type&#39; as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / &#39;(1,)type&#39;.
  _np_quint8 = np.dtype([(&quot;quint8&quot;, np.uint8, 1)])
/home/icb/felix.drost/miniconda3/envs/TESSA/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:525: FutureWarning: Passing (type, 1) or &#39;1type&#39; as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / &#39;(1,)type&#39;.
  _np_qint16 = np.dtype([(&quot;qint16&quot;, np.int16, 1)])
/home/icb/felix.drost/miniconda3/envs/TESSA/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:526: FutureWarning: Passing (type, 1) or &#39;1type&#39; as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / &#39;(1,)type&#39;.
  _np_quint16 = np.dtype([(&quot;quint16&quot;, np.uint16, 1)])
/home/icb/felix.drost/miniconda3/envs/TESSA/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:527: FutureWarning: Passing (type, 1) or &#39;1type&#39; as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / &#39;(1,)type&#39;.
  _np_qint32 = np.dtype([(&quot;qint32&quot;, np.int32, 1)])
/home/icb/felix.drost/miniconda3/envs/TESSA/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:532: FutureWarning: Passing (type, 1) or &#39;1type&#39; as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / &#39;(1,)type&#39;.
  np_resource = np.dtype([(&quot;resource&quot;, np.ubyte, 1)])
sys:1: FutureWarning: arrays to stack must be passed as a &quot;sequence&quot; type such as list or tuple. Support for non-sequence iterables such as generators is deprecated as of NumPy 1.16 and will raise an error in the future.
2022-08-08 09:43:34.047654: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: SSE4.1 SSE4.2 AVX AVX2 AVX512F FMA
2022-08-08 09:43:34.050857: I tensorflow/core/common_runtime/process_util.cc:69] Creating new thread pool with default inter op setting: 2. Tune using inter_op_parallelism_threads for best performance.
Warning message:
package ‘LaplacesDemon’ was built under R version 3.5.1 
Warning message:
package ‘Rtsne’ was built under R version 3.5.1 

Initialization
[1] 123

Iteration round: 1 
  # clusters: 751 
  Clustering rate: 0 
  Recent b acceptance rate: 0 

Iteration round: 2 
  # clusters: 171 
  Clustering rate: 0.916 
  Recent b acceptance rate: 0 

Iteration round: 3 
  # clusters: 127 
  Clustering rate: 0.997 
  Recent b acceptance rate: 0 

Iteration round: 4 
  # clusters: 126 
  Clustering rate: 0.997 
  Recent b acceptance rate: 0 

Iteration round: 5 
  # clusters: 125 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 6 
  # clusters: 125 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 7 
  # clusters: 124 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 8 
  # clusters: 124 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 9 
  # clusters: 125 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 10 
  # clusters: 124 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 11 
  # clusters: 124 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 12 
  # clusters: 124 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 13 
  # clusters: 124 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 14 
  # clusters: 124 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 15 
  # clusters: 124 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 16 
  # clusters: 124 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 17 
  # clusters: 125 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 18 
  # clusters: 125 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 19 
  # clusters: 124 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 20 
  # clusters: 124 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 21 
  # clusters: 124 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 22 
  # clusters: 124 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 23 
  # clusters: 124 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 24 
  # clusters: 125 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 25 
  # clusters: 124 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 26 
  # clusters: 124 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 27 
  # clusters: 124 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 28 
  # clusters: 125 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 29 
  # clusters: 124 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 30 
  # clusters: 124 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 31 
  # clusters: 124 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 32 
  # clusters: 124 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 33 
  # clusters: 124 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 34 
  # clusters: 124 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 35 
  # clusters: 124 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 36 
  # clusters: 125 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 37 
  # clusters: 124 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 38 
  # clusters: 125 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 39 
  # clusters: 124 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 40 
  # clusters: 124 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 41 
  # clusters: 125 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 42 
  # clusters: 124 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 43 
  # clusters: 126 
  Clustering rate: 0.997 
  Recent b acceptance rate: 0 

Iteration round: 44 
  # clusters: 125 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 45 
  # clusters: 125 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 46 
  # clusters: 125 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 47 
  # clusters: 124 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 48 
  # clusters: 125 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 49 
  # clusters: 124 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 50 
  # clusters: 124 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 51 
  # clusters: 124 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 52 
  # clusters: 124 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 53 
  # clusters: 124 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 54 
  # clusters: 124 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 55 
  # clusters: 125 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 56 
  # clusters: 124 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 57 
  # clusters: 125 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 58 
  # clusters: 125 
  Clustering rate: 0.999 
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>  Recent b acceptance rate: 0 

Iteration round: 59 
  # clusters: 125 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 60 
  # clusters: 124 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 61 
  # clusters: 126 
  Clustering rate: 0.997 
  Recent b acceptance rate: 0 

Iteration round: 62 
  # clusters: 126 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 63 
  # clusters: 126 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 64 
  # clusters: 126 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 65 
  # clusters: 125 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 66 
  # clusters: 125 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 67 
  # clusters: 125 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 68 
  # clusters: 126 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 69 
  # clusters: 125 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 70 
  # clusters: 126 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 71 
  # clusters: 126 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 72 
  # clusters: 126 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 73 
  # clusters: 126 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 74 
  # clusters: 126 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 75 
  # clusters: 127 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 76 
  # clusters: 126 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 77 
  # clusters: 126 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 78 
  # clusters: 126 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 79 
  # clusters: 126 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 80 
  # clusters: 126 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 81 
  # clusters: 126 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 82 
  # clusters: 127 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 83 
  # clusters: 127 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 84 
  # clusters: 126 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 85 
  # clusters: 126 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 86 
  # clusters: 126 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 87 
  # clusters: 126 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 88 
  # clusters: 126 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 89 
  # clusters: 127 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 90 
  # clusters: 126 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 91 
  # clusters: 126 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 92 
  # clusters: 128 
  Clustering rate: 0.997 
  Recent b acceptance rate: 0 

Iteration round: 93 
  # clusters: 126 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 94 
  # clusters: 127 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 95 
  # clusters: 127 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 96 
  # clusters: 126 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 97 
  # clusters: 126 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 98 
  # clusters: 126 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 99 
  # clusters: 126 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 100 
  # clusters: 127 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 101 
  # clusters: 126 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 102 
  # clusters: 127 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 103 
  # clusters: 126 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 104 
  # clusters: 127 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 105 
  # clusters: 126 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 106 
  # clusters: 126 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 107 
  # clusters: 126 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 108 
  # clusters: 126 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 109 
  # clusters: 127 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 110 
  # clusters: 127 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 111 
  # clusters: 126 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 112 
  # clusters: 126 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 113 
  # clusters: 126 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 114 
  # clusters: 126 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 115 
  # clusters: 127 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 116 
  # clusters: 126 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 117 
  # clusters: 127 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 118 
  # clusters: 126 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 119 
  # clusters: 128 
  Clustering rate: 0.997 
  Recent b acceptance rate: 0 

Iteration round: 120 
  # clusters: 126 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 121 
  # clusters: 126 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 122 
  # clusters: 126 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 123 
  # clusters: 127 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 124 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 125 
  # clusters: 128 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 126 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 127 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 128 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 129 
  # clusters: 128 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 130 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 131 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 132 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 133 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 134 
  # clusters: 129 
  Clustering rate: 0.997 
  Recent b acceptance rate: 0 

Iteration round: 135 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 136 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 137 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 138 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 139 
  # clusters: 128 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 140 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 141 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 142 
  # clusters: 128 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 143 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 144 
  # clusters: 127 
  Clustering rate: 1 
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>  Recent b acceptance rate: 0 

Iteration round: 145 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 146 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 147 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 148 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 149 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 150 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 151 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 152 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 153 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 154 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 155 
  # clusters: 128 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 156 
  # clusters: 128 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 157 
  # clusters: 128 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 158 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 159 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 160 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 161 
  # clusters: 128 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 162 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 163 
  # clusters: 129 
  Clustering rate: 0.997 
  Recent b acceptance rate: 0 

Iteration round: 164 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 165 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 166 
  # clusters: 128 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 167 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 168 
  # clusters: 129 
  Clustering rate: 0.997 
  Recent b acceptance rate: 0 

Iteration round: 169 
  # clusters: 128 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 170 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 171 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 172 
  # clusters: 128 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 173 
  # clusters: 128 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 174 
  # clusters: 128 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 175 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 176 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 177 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 178 
  # clusters: 128 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 179 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 180 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 181 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 182 
  # clusters: 128 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 183 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 184 
  # clusters: 128 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 185 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 186 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 187 
  # clusters: 128 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 188 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 189 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 190 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 191 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 192 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 193 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 194 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 195 
  # clusters: 128 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 196 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 197 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 198 
  # clusters: 128 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 199 
  # clusters: 128 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 200 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 201 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 202 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 203 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 204 
  # clusters: 128 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 205 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 206 
  # clusters: 128 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 207 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 208 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 209 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 210 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 211 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 212 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 213 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 214 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 215 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 216 
  # clusters: 128 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 217 
  # clusters: 128 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 218 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 219 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 220 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 221 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 222 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 223 
  # clusters: 128 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 224 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 225 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 226 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 227 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 228 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 229 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 230 
  # clusters: 127 
  Clustering rate: 1 
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>  Recent b acceptance rate: 0 

Iteration round: 231 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 232 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 233 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 234 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 235 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 236 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 237 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 238 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 239 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 240 
  # clusters: 129 
  Clustering rate: 0.997 
  Recent b acceptance rate: 0 

Iteration round: 241 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 242 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 243 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 244 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 245 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 246 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 247 
  # clusters: 128 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 248 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 249 
  # clusters: 128 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 250 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 251 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 252 
  # clusters: 128 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 253 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 254 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 255 
  # clusters: 127 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 256 
  # clusters: 129 
  Clustering rate: 0.997 
  Recent b acceptance rate: 0 

Iteration round: 257 
  # clusters: 129 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 258 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 259 
  # clusters: 129 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 260 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 261 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 262 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 263 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 264 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 265 
  # clusters: 129 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 266 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 267 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 268 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 269 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 270 
  # clusters: 129 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 271 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 272 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 273 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 274 
  # clusters: 129 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 275 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 276 
  # clusters: 129 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 277 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 278 
  # clusters: 131 
  Clustering rate: 0.996 
  Recent b acceptance rate: 0 

Iteration round: 279 
  # clusters: 129 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 280 
  # clusters: 130 
  Clustering rate: 0.997 
  Recent b acceptance rate: 0 

Iteration round: 281 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 282 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 283 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 284 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 285 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 286 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 287 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 288 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 289 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 290 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 291 
  # clusters: 129 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 292 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 293 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 294 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 295 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 296 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 297 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 298 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 299 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 300 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 301 
  # clusters: 130 
  Clustering rate: 0.997 
  Recent b acceptance rate: 0 

Iteration round: 302 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 303 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 304 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 305 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 306 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 307 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 308 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 309 
  # clusters: 130 
  Clustering rate: 0.997 
  Recent b acceptance rate: 0 

Iteration round: 310 
  # clusters: 129 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 311 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 312 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 313 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 314 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 315 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 316 
  # clusters: 128 
  Clustering rate: 1 
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>  Recent b acceptance rate: 0 

Iteration round: 317 
  # clusters: 129 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 318 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 319 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 320 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 321 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 322 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 323 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 324 
  # clusters: 129 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 325 
  # clusters: 129 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 326 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 327 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 328 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 329 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 330 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 331 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 332 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 333 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 334 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 335 
  # clusters: 129 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 336 
  # clusters: 129 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 337 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 338 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 339 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 340 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 341 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 342 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 343 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 344 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 345 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 346 
  # clusters: 129 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 347 
  # clusters: 129 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 348 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 349 
  # clusters: 129 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 350 
  # clusters: 129 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 351 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 352 
  # clusters: 129 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 353 
  # clusters: 129 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 354 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 355 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 356 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 357 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 358 
  # clusters: 129 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 359 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 360 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 361 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 362 
  # clusters: 129 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 363 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 364 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 365 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 366 
  # clusters: 129 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 367 
  # clusters: 129 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 368 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 369 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 370 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 371 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 372 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 373 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 374 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 375 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 376 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 377 
  # clusters: 129 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 378 
  # clusters: 129 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 379 
  # clusters: 129 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 380 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 381 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 382 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 383 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 384 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 385 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 386 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 387 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 388 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 389 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 390 
  # clusters: 129 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 391 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 392 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 393 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 394 
  # clusters: 129 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 395 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 396 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 397 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 398 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 399 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 400 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 401 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 402 
  # clusters: 128 
  Clustering rate: 1 
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>  Recent b acceptance rate: 0 

Iteration round: 403 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 404 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 405 
  # clusters: 129 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 406 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 407 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 408 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 409 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 410 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 411 
  # clusters: 129 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 412 
  # clusters: 130 
  Clustering rate: 0.997 
  Recent b acceptance rate: 0 

Iteration round: 413 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 414 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 415 
  # clusters: 129 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 416 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 417 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 418 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 419 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 420 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 421 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 422 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 423 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 424 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 425 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 426 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 427 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 428 
  # clusters: 129 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 429 
  # clusters: 129 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 430 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 431 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 432 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 433 
  # clusters: 129 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 434 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 435 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 436 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 437 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 438 
  # clusters: 129 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 439 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 440 
  # clusters: 130 
  Clustering rate: 0.997 
  Recent b acceptance rate: 0 

Iteration round: 441 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 442 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 443 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 444 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 445 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 446 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 447 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 448 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 449 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 450 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 451 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 452 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 453 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 454 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 455 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 456 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 457 
  # clusters: 130 
  Clustering rate: 0.997 
  Recent b acceptance rate: 0 

Iteration round: 458 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 459 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 460 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 461 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 462 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 463 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 464 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 465 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 466 
  # clusters: 131 
  Clustering rate: 0.996 
  Recent b acceptance rate: 0 

Iteration round: 467 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 468 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 469 
  # clusters: 128 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 470 
  # clusters: 130 
  Clustering rate: 0.997 
  Recent b acceptance rate: 0 

Iteration round: 471 
  # clusters: 129 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 472 
  # clusters: 130 
  Clustering rate: 0.997 
  Recent b acceptance rate: 0 

Iteration round: 473 
  # clusters: 130 
  Clustering rate: 0.997 
  Recent b acceptance rate: 0 

Iteration round: 474 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 475 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 476 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 477 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 478 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 479 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 480 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 481 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 482 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 483 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 484 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 485 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 486 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 487 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 488 
  # clusters: 129 
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 489 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 490 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 491 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 492 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 493 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 494 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 495 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 496 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 497 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 498 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 499 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 500 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 501 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 502 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 503 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 504 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 505 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 506 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 507 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 508 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 509 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 510 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 511 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 512 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 513 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 514 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 515 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 516 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 517 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 518 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 519 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 520 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 521 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 522 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 523 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 524 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 525 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 526 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 527 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 528 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 529 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 530 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 531 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 532 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 533 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 534 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 535 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 536 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 537 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 538 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 539 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 540 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 541 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 542 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 543 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 544 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 545 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 546 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 547 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 548 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 549 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 550 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 551 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 552 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 553 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 554 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 555 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 556 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 557 
  # clusters: 131 
  Clustering rate: 0.997 
  Recent b acceptance rate: 0 

Iteration round: 558 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 559 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 560 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 561 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 562 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 563 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 564 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 565 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 566 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 567 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 568 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 569 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 570 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 571 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 572 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 573 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 574 
  # clusters: 129 
  Clustering rate: 1 
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>  Recent b acceptance rate: 0 

Iteration round: 575 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 576 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 577 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 578 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 579 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 580 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 581 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 582 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 583 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 584 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 585 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 586 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 587 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 588 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 589 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 590 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 591 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 592 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 593 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 594 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 595 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 596 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 597 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 598 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 599 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 600 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 601 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 602 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 603 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 604 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 605 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 606 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 607 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 608 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 609 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 610 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 611 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 612 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 613 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 614 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 615 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 616 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 617 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 618 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 619 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 620 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 621 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 622 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 623 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 624 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 625 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 626 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 627 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 628 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 629 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 630 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 631 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 632 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 633 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 634 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 635 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 636 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 637 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 638 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 639 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 640 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 641 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 642 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 643 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 644 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 645 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 646 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 647 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 648 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 649 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 650 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 651 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 652 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 653 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 654 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 655 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 656 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 657 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 658 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 659 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 660 
  # clusters: 129 
  Clustering rate: 1 
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>  Recent b acceptance rate: 0 

Iteration round: 661 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 662 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 663 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 664 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 665 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 666 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 667 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 668 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 669 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 670 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 671 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 672 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 673 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 674 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 675 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 676 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 677 
  # clusters: 131 
  Clustering rate: 0.997 
  Recent b acceptance rate: 0 

Iteration round: 678 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 679 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 680 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 681 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 682 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 683 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 684 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 685 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 686 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 687 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 688 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 689 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 690 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 691 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 692 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 693 
  # clusters: 131 
  Clustering rate: 0.997 
  Recent b acceptance rate: 0 

Iteration round: 694 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 695 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 696 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 697 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 698 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 699 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 700 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 701 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 702 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 703 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 704 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 705 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 706 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 707 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 708 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 709 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 710 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 711 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 712 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 713 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 714 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 715 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 716 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 717 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 718 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 719 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 720 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 721 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 722 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 723 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 724 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 725 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 726 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 727 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 728 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 729 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 730 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 731 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 732 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 733 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 734 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 735 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 736 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 737 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 738 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 739 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 740 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 741 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 742 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 743 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 744 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 745 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 746 
  # clusters: 131 
  Clustering rate: 0.997 
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>  Recent b acceptance rate: 0 

Iteration round: 747 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 748 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 749 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 750 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 751 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 752 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 753 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 754 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 755 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 756 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 757 
  # clusters: 131 
  Clustering rate: 0.997 
  Recent b acceptance rate: 0 

Iteration round: 758 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 759 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 760 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 761 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 762 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 763 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 764 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 765 
  # clusters: 131 
  Clustering rate: 0.997 
  Recent b acceptance rate: 0 

Iteration round: 766 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 767 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 768 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 769 
  # clusters: 131 
  Clustering rate: 0.997 
  Recent b acceptance rate: 0 

Iteration round: 770 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 771 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 772 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 773 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 774 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 775 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 776 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 777 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 778 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 779 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 780 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 781 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 782 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 783 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 784 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 785 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 786 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 787 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 788 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 789 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 790 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 791 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 792 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 793 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 794 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 795 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 796 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 797 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 798 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 799 
  # clusters: 131 
  Clustering rate: 0.997 
  Recent b acceptance rate: 0 

Iteration round: 800 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 801 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 802 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 803 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 804 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 805 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 806 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 807 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 808 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 809 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 810 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 811 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 812 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 813 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 814 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 815 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 816 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 817 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 818 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 819 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 820 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 821 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 822 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 823 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 824 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 825 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 826 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 827 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 828 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 829 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 830 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 831 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 832 
  # clusters: 129 
  Clustering rate: 1 
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>  Recent b acceptance rate: 0 

Iteration round: 833 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 834 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 835 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 836 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 837 
  # clusters: 129 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 838 
  # clusters: 130 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 839 
  # clusters: 131 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 840 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 841 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 842 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 843 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 844 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 845 
  # clusters: 131 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 846 
  # clusters: 132 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 847 
  # clusters: 131 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 848 
  # clusters: 132 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 849 
  # clusters: 131 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 850 
  # clusters: 132 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 851 
  # clusters: 132 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 852 
  # clusters: 131 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 853 
  # clusters: 132 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 854 
  # clusters: 131 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 855 
  # clusters: 131 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 856 
  # clusters: 131 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 857 
  # clusters: 131 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 858 
  # clusters: 131 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 859 
  # clusters: 132 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 860 
  # clusters: 131 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 861 
  # clusters: 131 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 862 
  # clusters: 131 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 863 
  # clusters: 131 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 864 
  # clusters: 131 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 865 
  # clusters: 131 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 866 
  # clusters: 131 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 867 
  # clusters: 131 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 868 
  # clusters: 131 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 869 
  # clusters: 131 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 870 
  # clusters: 131 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 871 
  # clusters: 131 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 872 
  # clusters: 132 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 873 
  # clusters: 131 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 874 
  # clusters: 131 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 875 
  # clusters: 131 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 876 
  # clusters: 132 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 877 
  # clusters: 131 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 878 
  # clusters: 131 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 879 
  # clusters: 131 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 880 
  # clusters: 131 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 881 
  # clusters: 131 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 882 
  # clusters: 131 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 883 
  # clusters: 131 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 884 
  # clusters: 132 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 885 
  # clusters: 131 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 886 
  # clusters: 131 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 887 
  # clusters: 131 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 888 
  # clusters: 131 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 889 
  # clusters: 131 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 890 
  # clusters: 131 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 891 
  # clusters: 131 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 892 
  # clusters: 131 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 893 
  # clusters: 131 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 894 
  # clusters: 131 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 895 
  # clusters: 131 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 896 
  # clusters: 132 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 897 
  # clusters: 131 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 898 
  # clusters: 131 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 899 
  # clusters: 131 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 900 
  # clusters: 131 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 901 
  # clusters: 131 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 902 
  # clusters: 131 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 903 
  # clusters: 132 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 904 
  # clusters: 132 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 905 
  # clusters: 131 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 906 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 907 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 908 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 909 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 910 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 911 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 912 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 913 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 914 
  # clusters: 131 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 915 
  # clusters: 131 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 916 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 917 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 918 
  # clusters: 131 
  Clustering rate: 0.999 
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>  Recent b acceptance rate: 0 

Iteration round: 919 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 920 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 921 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 922 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 923 
  # clusters: 131 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 924 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 925 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 926 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 927 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 928 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 929 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 930 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 931 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 932 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 933 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 934 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 935 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 936 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 937 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 938 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 939 
  # clusters: 131 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 940 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 941 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 942 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 943 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 944 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 945 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 946 
  # clusters: 131 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 947 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 948 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 949 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 950 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 951 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 952 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 953 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 954 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 955 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 956 
  # clusters: 131 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 957 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 958 
  # clusters: 131 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 959 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 960 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 961 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 962 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 963 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 964 
  # clusters: 131 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 965 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 966 
  # clusters: 132 
  Clustering rate: 0.997 
  Recent b acceptance rate: 0 

Iteration round: 967 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 968 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 969 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 970 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 971 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 972 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 973 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 974 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 975 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 976 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 977 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 978 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 979 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 980 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 981 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 982 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 983 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 984 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 985 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 986 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 987 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 988 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 989 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 990 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 991 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 992 
  # clusters: 131 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 993 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 994 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 995 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 996 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 997 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 998 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 

Iteration round: 999 
  # clusters: 131 
  Clustering rate: 0.999 
  Recent b acceptance rate: 0 

Iteration round: 1000 
  # clusters: 130 
  Clustering rate: 1 
  Recent b acceptance rate: 0 
</pre></div>
</div>
</div>
</div>
</section>
<section id="output">
<h5><span class="section-number">36.4.1.1.3. </span>Output<a class="headerlink" href="#output" title="Permalink to this headline">#</a></h5>
<p>We now recieve three output files: hat we can further use for downstream tasks:</p>
<ul class="simple">
<li><p><strong>TESSA_tcr_embedding.csv</strong>: Embedding of TCRs (without GEX information)</p></li>
<li><p><strong>result_meta.csv</strong>: cluster assignment by cell</p></li>
<li><p><strong>tessa_final.RData</strong>: Containing inferred model parameters (e.g. weighting vector b)</p></li>
</ul>
<p>Here, we will use the first two files as embedding and clustering. We will store them in AnnData format for easier handling. To avoid the clustering to be dominated by highly expanded clones, we reduce the cell-level data to a clone-level by removing duplicate TCRs.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>tessa_embedding = pd.read_csv(f&quot;{path_res}/TESSA_tcr_embedding.csv&quot;, index_col=0)
tessa_embedding = tessa_embedding.drop_duplicates()
tessa_obs = adata_tc[adata_tc.obs.index.isin(tessa_embedding.index)].obs
tessa_embedding = sc.AnnData(X=tessa_embedding.values, obs=tessa_obs)
</pre></div>
</div>
</div>
</div>
<p>Next, we will add the TESSA cluster assignment to the AnnData object. Since TESSA internally changes the barcode slightly, we will need to adjust for that. Additionally, we drop the duplicate TCRs within a cluster.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>clustering = pd.read_csv(f&quot;{path_res}/result_meta.csv&quot;, index_col=0)
clustering.index = clustering.index.str.replace(&quot;.&quot;, &quot;-&quot;)
clustering = clustering[clustering.index.isin(tessa_embedding.obs.index)]
tessa_embedding.obs[&quot;TESSA_cluster&quot;] = clustering[&quot;cluster_number&quot;].values

print(f&quot;Unique clones: {tessa_embedding.obs[&#39;cdr3&#39;].nunique()}&quot;)
print(f&quot;Unique clusters: {tessa_embedding.obs[&#39;TESSA_cluster&#39;].nunique()}&quot;)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Unique clones: 751
Unique clusters: 132
</pre></div>
</div>
</div>
</div>
<p>You can see, that 751 clones were grouped to 144 clusters, which have similar TCR sequences and are likely to be functionally related. For TSNE visualisation {cite}’’, we will highlight the ten largest clusters in the TCR embedding space.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pp.neighbors(tessa_embedding)
sc.tl.tsne(tessa_embedding)

top_10_clusters = tessa_embedding.obs[&quot;TESSA_cluster&quot;].value_counts().head(10).index
tessa_embedding.obs[&quot;TESSA_cluster_top10&quot;] = tessa_embedding.obs[&quot;TESSA_cluster&quot;].apply(
    lambda x: x if x in top_10_clusters else np.nan
)

sc.pl.tsne(tessa_embedding, color=&quot;TESSA_cluster_top10&quot;)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>WARNING: Consider installing the package MulticoreTSNE (https://github.com/DmitryUlyanov/Multicore-TSNE). Even for n_jobs=1 this speeds up the computation considerably and might yield better converged results.
... storing &#39;TESSA_cluster&#39; as categorical
... storing &#39;TESSA_cluster_top10&#39; as categorical
</pre></div>
</div>
<img alt="../_images/multimodal_integration_59_1.png" src="../_images/multimodal_integration_59_1.png" />
</div>
</div>
<p>For plotting the cluster assignemnt in the GEX space, we add the assignment to the original adata object. When plotting, we can see that TESSA clusters also share similar phenotypes at GEX level.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>mapping_dict = dict(
    tessa_embedding.obs[[&quot;IR_VDJ_1_junction_aa&quot;, &quot;TESSA_cluster_top10&quot;]].values
)
adata_tc.obs[&quot;TESSA_cluster_top10&quot;] = adata_tc.obs[&quot;IR_VDJ_1_junction_aa&quot;].map(
    mapping_dict
)
sc.pl.umap(adata_tc, color=&quot;TESSA_cluster_top10&quot;)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>... storing &#39;cdr3&#39; as categorical
... storing &#39;TESSA_cluster_top10&#39; as categorical
</pre></div>
</div>
<img alt="../_images/multimodal_integration_61_1.png" src="../_images/multimodal_integration_61_1.png" />
</div>
</div>
<p>We can see in both visualisations, that the clusters are related on a TCR and GEX level. However, the full clustering can not directly observed in a single modality. Due to similar TCR and GEX profile, these cells might be specific to the same epitopes. Examplatory, we could use this annotation for DEG analysis between cell networks as described above.</p>
</section>
</section>
<section id="conga">
<h4><span class="section-number">36.4.1.2. </span>CoNGA<a class="headerlink" href="#conga" title="Permalink to this headline">#</a></h4>
<p>clonotype neighbor graph analysis (CoNGA) <span id="id4">[<a class="reference internal" href="#id426" title="Stefan A Schattgen, Kate Guion, Jeremy Chase Crawford, Aisha Souquette, Alvaro Martinez Barrio, Michael JT Stubbington, Paul G Thomas, and Philip Bradley. Integrating t cell receptor sequences and transcriptional profiles by clonotype neighbor graph analysis (conga). Nature Biotechnology, 40(1):54–63, 2022.">Schattgen <em>et al.</em>, 2022</a>]</span> uses similarity graphs at a clonotype level on GEX and TCR, for which the well known distance metric TCRDist ‘cite’{dash2017quantifiable} is used. Based on the graph neighborhood, clusters are formed by shared GEX and TCR assignments. These so called “CoNGA clusters” thereby share similar receptor sequences as well as similar gene expression. CoNGA clusters contained clonotypes with highly alike TCRs based on e.g. length and physio-chemical properties and were shown to capture specificity. The following explanation runs the main CoNGA pipeline. Since the different aspects of the pipeline are explained here <a class="reference external" href="https://github.com/phbradley/conga/blob/master/colab_conga_pipeline.ipynb">https://github.com/phbradley/conga/blob/master/colab_conga_pipeline.ipynb</a>, the following description is mainly focused on how to transform the data and run the pipeline.</p>
<section id="data-preperation">
<h5><span class="section-number">36.4.1.2.1. </span>Data Preperation<a class="headerlink" href="#data-preperation" title="Permalink to this headline">#</a></h5>
<p>CoNGA conveniently supports several input types for the gene expression matrix including h5ad-files. For the clonotype information, we need to create a clone_file first by running a preprocessing script on the Cell Ranger output.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>import conga

df_tcr = pd.read_csv(
    f&quot;{path_data}/TCR_00_read_aligned.csv&quot;,
    # index_col=&#39;barcode&#39;,
)
df_tcr = df_tcr.drop(&quot;Unnamed: 0&quot;, axis=1)
df_tcr = df_tcr[df_tcr[&quot;barcode&quot;].isin(adata_tc.obs.index)]

path_reduced_tcr = f&quot;{path_tmp}/conga_tcrs.csv&quot;
df_tcr.to_csv(path_reduced_tcr)

path_conga_clones = f&quot;{path_tmp}/conga_clone_file.tsv&quot;
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/icb/felix.drost/miniconda3/envs/best_practice/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3135: DtypeWarning: Columns (38,39,40) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cmd_conga_pp = (
    &quot;python conga/scripts/setup_10x_for_conga.py &quot;
    f&quot;--filtered_contig_annotations_csvfile {path_reduced_tcr} &quot;
    f&quot;--output_clones_file {path_conga_clones} &quot;
    &quot;--organism human&quot;
)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>!$cmd_conga_pp
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>repeat?? 2 289 (&#39;TRBV18*01&#39;, &#39;TRBJ1-4*01&#39;, &#39;CASSIGGQIYEKLFF&#39;, &#39;tgtgccagtagtattggagggcagatatatgaaaaactgtttttt&#39;) (&#39;TRBV19*01&#39;, &#39;TRBJ1-4*01&#39;, &#39;CASSIGGQIYEKLFF&#39;, &#39;tgtgccagtagtattggagggcagatatatgaaaaactgtttttt&#39;)
repeat?? 2 88 (&#39;TRAV18*01&#39;, &#39;TRAJ40*01&#39;, &#39;CALSPQTPGTYKYIF&#39;, &#39;tgtgctctgagtccccaaaccccaggaacctacaaatacatcttt&#39;) (&#39;TRAV19*01&#39;, &#39;TRAJ40*01&#39;, &#39;CALSPQTPGTYKYIF&#39;, &#39;tgtgctctgagtccccaaaccccaggaacctacaaatacatcttt&#39;)
repeat?? 480 11 (&#39;TRBV11-2*01&#39;, &#39;TRBJ2-1*01&#39;, &#39;CASSLDLLLRYNEQFF&#39;, &#39;tgtgccagcagcttagatctactactaagatacaatgagcagttcttc&#39;) (&#39;TRBV7-2*01&#39;, &#39;TRBJ2-1*01&#39;, &#39;CASSLDLLLRYNEQFF&#39;, &#39;tgtgccagcagcttagatctactactaagatacaatgagcagttcttc&#39;)
ab_counts: [((1, 1), 2292), ((2, 2), 1986), ((1, 2), 155), ((2, 3), 121), ((2, 1), 119), ((3, 2), 106), ((3, 3), 65), ((4, 2), 14)]
old_unpaired_barcodes: 414 old_paired_barcodes: 4867 new_stringent_paired_barcodes: 2411
compute tcrdist distance matrix for 623 clonotypes
Using C++ TCRdist calculator
Read 623 paired tcrs from file ./data/tmp/conga_clone_file_AB.dist_50_kpcs_tcrs.tsv
making ./data/tmp/conga_clone_file_AB.dist_50_kpcs_tcrdists.txt
......
running KernelPCA with None kernel distance matrix shape= (623, 623) D.max()= 445.0 force_Dmax= None
writing TCRdist kernel PCs to outfile: ./data/tmp/conga_clone_file_AB.dist_50_kpcs
If this all worked you should be able to pass ./data/tmp/conga_clone_file.tsv as the --clones_file argument to run_conga.py
DONE
</pre></div>
</div>
</div>
</div>
<p>The CoNGA pipeline requires unlogged counts for the gene expression. Since this is not provided in our SARS-CoV-2 dataset, we will revert the log1p transformation by its inverse and save the results to a separate file. If you run the analysis on your own dataset, you can directly link to your raw count data.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>path_conga_gex = f&quot;{path_tmp}/conga_gex.h5ad&quot;

adata_conga = adata_tc.copy()
adata_conga.X = np.expm1(adata_tc.X)
sc.write(adata=adata_conga, filename=path_conga_gex)
</pre></div>
</div>
</div>
</div>
</section>
<section id="executing-conga">
<h5><span class="section-number">36.4.1.2.2. </span>Executing CoNGA<a class="headerlink" href="#executing-conga" title="Permalink to this headline">#</a></h5>
<p>We can execute the main CoNGA pipeline via the following command line call, which specifies the input and output directories.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cmd_conga = f&quot;cd {path_res} &amp;&amp; &quot;

cmd_conga += (
    &quot;python ../../conga/scripts/run_conga.py &quot;
    f&quot;--graph_vs_graph --graph_vs_graph_stats --graph_vs_features --tcr_clumping --find_hotspot_features &quot;
    f&quot;--gex_data ../../{path_conga_gex} --gex_data_type h5ad &quot;
    f&quot;--clones_file ../../{path_conga_clones} &quot;
    f&quot;--organism human --outfile_prefix conga&quot;
)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>!$cmd_conga
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>WARNING: If you miss a compact list, please try `print_header`!
The `sinfo` package has changed name and is now called `session_info` to become more discoverable and self-explanatory. The `sinfo` PyPI package will be kept around to avoid breaking old installs and you can downgrade to 0.3.2 if you want to use it without seeing this message. For the latest features and bug fixes, please install `session_info` instead. The usage and defaults also changed slightly, so please review the latest README at https://gitlab.com/joelostblom/session_info.
-----
anndata     0.7.6
scanpy      1.7.0
sinfo       0.3.4
-----
PIL                 9.2.0
beta_ufunc          NA
binom_ufunc         NA
cairocffi           1.3.0
cffi                1.15.0
conga               NA
cycler              0.10.0
cython_runtime      NA
dateutil            2.8.2
defusedxml          0.7.1
dunamai             1.12.0
get_version         3.5.4
h5py                3.7.0
hypergeom_ufunc     NA
igraph              0.9.11
importlib_metadata  NA
joblib              1.1.0
kiwisolver          1.4.4
legacy_api_wrap     1.2
leidenalg           0.8.4
llvmlite            0.35.0
matplotlib          3.5.2
mpl_toolkits        NA
natsort             8.1.0
nbinom_ufunc        NA
ncf_ufunc           NA
numba               0.52.0
numexpr             2.8.3
numpy               1.20.3
packaging           21.3
pandas              1.2.3
pkg_resources       NA
pyparsing           3.0.4
pytoml              NA
pytz                2022.1
scipy               1.9.0
setuptools          61.2.0
setuptools_scm      NA
six                 1.16.0
sklearn             0.24.1
statsmodels         0.13.2
tables              3.7.0
texttable           1.6.4
typing_extensions   NA
wcwidth             0.2.5
yaml                6.0
zipp                NA
-----
Python 3.8.13 (default, Mar 28 2022, 11:38:47) [GCC 7.5.0]
Linux-3.10.0-1160.11.1.el7.x86_64-x86_64-with-glibc2.17
80 logical CPU cores, x86_64
-----
Session information updated at 2022-08-08 09:50

reading: ../.././data/tmp/conga_gex.h5ad of type h5ad
/lustre/groups/imm01/workspace/felix.drost/BestPractice2/notebooks/multimodal_vdj/conga/conga/preprocess.py:233: DeprecationWarning: Use is_view instead of isview, isview will be removed in the future.
  if adata.isview: # ran into trouble with AnnData views vs copies
total barcodes: 5281 (5281, 24929)
reading: ../.././data/tmp/conga_clone_file.tsv
reading: ../.././data/tmp/conga_clone_file_AB.dist_50_kpcs
Reducing to the 2411 barcodes (out of 5281) with paired TCR sequence data
AnnData object with n_obs × n_vars = 2411 × 24929
    obs: &#39;sample_id&#39;, &#39;n_genes&#39;, &#39;n_genes_by_counts&#39;, &#39;total_counts&#39;, &#39;total_counts_mt&#39;, &#39;pct_counts_mt&#39;, &#39;full_clustering&#39;, &#39;initial_clustering_x&#39;, &#39;Resample&#39;, &#39;Collection_Day_x&#39;, &#39;Sex_x&#39;, &#39;Age_interval&#39;, &#39;Swab_result_x&#39;, &#39;Status_x&#39;, &#39;Smoker_x&#39;, &#39;Status_on_day_collection_x&#39;, &#39;Status_on_day_collection_summary_x&#39;, &#39;Days_from_onset_x&#39;, &#39;Site&#39;, &#39;time_after_LPS_x&#39;, &#39;Worst_Clinical_Status_x&#39;, &#39;Outcome_x&#39;, &#39;patient_id_x&#39;, &#39;is_cell&#39;, &#39;high_confidence&#39;, &#39;multi_chain&#39;, &#39;extra_chains&#39;, &#39;IR_VJ_1_c_call&#39;, &#39;IR_VJ_2_c_call&#39;, &#39;IR_VDJ_1_c_call&#39;, &#39;IR_VDJ_2_c_call&#39;, &#39;IR_VJ_1_consensus_count&#39;, &#39;IR_VJ_2_consensus_count&#39;, &#39;IR_VDJ_1_consensus_count&#39;, &#39;IR_VDJ_2_consensus_count&#39;, &#39;IR_VJ_1_d_call&#39;, &#39;IR_VJ_2_d_call&#39;, &#39;IR_VDJ_1_d_call&#39;, &#39;IR_VDJ_2_d_call&#39;, &#39;IR_VJ_1_duplicate_count&#39;, &#39;IR_VJ_2_duplicate_count&#39;, &#39;IR_VDJ_1_duplicate_count&#39;, &#39;IR_VDJ_2_duplicate_count&#39;, &#39;IR_VJ_1_j_call&#39;, &#39;IR_VJ_2_j_call&#39;, &#39;IR_VDJ_1_j_call&#39;, &#39;IR_VDJ_2_j_call&#39;, &#39;IR_VJ_1_junction&#39;, &#39;IR_VJ_2_junction&#39;, &#39;IR_VDJ_1_junction&#39;, &#39;IR_VDJ_2_junction&#39;, &#39;IR_VJ_1_junction_aa&#39;, &#39;IR_VJ_2_junction_aa&#39;, &#39;IR_VDJ_1_junction_aa&#39;, &#39;IR_VDJ_2_junction_aa&#39;, &#39;IR_VJ_1_locus&#39;, &#39;IR_VJ_2_locus&#39;, &#39;IR_VDJ_1_locus&#39;, &#39;IR_VDJ_2_locus&#39;, &#39;IR_VJ_1_productive&#39;, &#39;IR_VJ_2_productive&#39;, &#39;IR_VDJ_1_productive&#39;, &#39;IR_VDJ_2_productive&#39;, &#39;IR_VJ_1_v_call&#39;, &#39;IR_VJ_2_v_call&#39;, &#39;IR_VDJ_1_v_call&#39;, &#39;IR_VDJ_2_v_call&#39;, &#39;has_ir&#39;, &#39;Centre&#39;, &#39;Sample&#39;, &#39;patient_id_y&#39;, &#39;Collection_Day_y&#39;, &#39;Sex_y&#39;, &#39;Swab_result_y&#39;, &#39;Status_y&#39;, &#39;Smoker_y&#39;, &#39;Status_on_day_collection_y&#39;, &#39;Status_on_day_collection_summary_y&#39;, &#39;Days_from_onset_y&#39;, &#39;time_after_LPS_y&#39;, &#39;Worst_Clinical_Status_y&#39;, &#39;Outcome_y&#39;, &#39;initial_clustering_y&#39;, &#39;study_id&#39;, &#39;AgeRange&#39;, &#39;Age&#39;, &#39;receptor_type&#39;, &#39;receptor_subtype&#39;, &#39;chain_pairing&#39;, &#39;leiden&#39;, &#39;elevated_IFNG&#39;, &#39;cdr3&#39;, &#39;TESSA_cluster_top10&#39;, &#39;va&#39;, &#39;ja&#39;, &#39;cdr3a&#39;, &#39;cdr3a_nucseq&#39;, &#39;vb&#39;, &#39;jb&#39;, &#39;cdr3b&#39;, &#39;cdr3b_nucseq&#39;
    var: &#39;feature_types&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;
    uns: &#39;TESSA_cluster_top10_colors&#39;, &#39;elevated_IFNG_colors&#39;, &#39;full_clustering_colors&#39;, &#39;has_ir_colors&#39;, &#39;hvg&#39;, &#39;leiden&#39;, &#39;leiden_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;, &#39;umap&#39;, &#39;conga_results&#39;, &#39;conga_stats&#39;, &#39;organism&#39;
    obsm: &#39;X_pca&#39;, &#39;X_pca_harmony&#39;, &#39;X_umap&#39;, &#39;X_pca_tcr&#39;
    layers: &#39;raw&#39;
    obsp: &#39;connectivities&#39;, &#39;distances&#39;
min_genes_per_cell not set. Using default 200
max_genes_per_cell not set. Using default 2500
max_percent_mito not set. Using default 0.1
/home/icb/felix.drost/miniconda3/envs/best_practice/lib/python3.8/site-packages/seaborn/rcmod.py:82: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
  if LooseVersion(mpl.__version__) &gt;= &quot;3.0&quot;:
/home/icb/felix.drost/miniconda3/envs/best_practice/lib/python3.8/site-packages/setuptools/_distutils/version.py:351: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
  other = LooseVersion(other)
... storing &#39;va&#39; as categorical
... storing &#39;ja&#39; as categorical
... storing &#39;cdr3a&#39; as categorical
... storing &#39;cdr3a_nucseq&#39; as categorical
... storing &#39;vb&#39; as categorical
... storing &#39;jb&#39; as categorical
... storing &#39;cdr3b&#39; as categorical
... storing &#39;cdr3b_nucseq&#39; as categorical
making: conga_qc_violins.png
making: conga_counts_vs_genes.png
filtered out 29 cells with more than 2500 genes
filtered out 0 cells with more than 0.1 percent mito
get_feature_types_varname: feature_types feature_type_counts: [(&#39;Gene Expression&#39;, 12921), (&#39;Antibody Capture&#39;, 192)]
num antibody features: 192
Removed 192 antibody features from adata, using colname feature_types
excluding 100 TR genes (87 variable)
excluding 7 sexlinked genes
total of 1216 variable genes (2382, 12921)
Normalize and logging matrix...
get_feature_types_varname: feature_types feature_type_counts: [(&#39;Gene Expression&#39;, 1216)]
feature_types counter: [(&#39;Gene Expression&#39;, 12921), (&#39;Antibody Capture&#39;, 192)]
normalize_and_log_the_raw_matrix: adata.raw.X.max()=  1011.645
run reduce_to_single_cell_per_clone
compute pca to find rep cell for each clone (2382, 1216)
num_clones: 607
normalize_and_log_the_raw_matrix:: matrix is already logged
choose representative cell for clone: 0 607 (2382, 1216)
reduce from 2382 cells to 607 cells (one per clonotype)
run cluster_and_tsne_and_umap
computing X_pca_gex using sc.tl.pca
/home/icb/felix.drost/miniconda3/envs/best_practice/lib/python3.8/site-packages/numba/core/consts.py:114: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  return getattr(value, expr.attr)
/home/icb/felix.drost/miniconda3/envs/best_practice/lib/python3.8/site-packages/numba/core/ir_utils.py:2097: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  defn = getattr(defn, x, False)
/home/icb/felix.drost/miniconda3/envs/best_practice/lib/python3.8/site-packages/numba/core/typing/context.py:338: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  attrval = getattr(typ.pymod, attr)
/home/icb/felix.drost/miniconda3/envs/best_practice/lib/python3.8/site-packages/numba/core/typing/context.py:338: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  attrval = getattr(typ.pymod, attr)
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/icb/felix.drost/miniconda3/envs/best_practice/lib/python3.8/site-packages/numba/core/typing/context.py:338: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  attrval = getattr(typ.pymod, attr)
running 1D UMAP gex
ran leiden clustering: leiden_gex
running 1D UMAP tcr
ran leiden clustering: leiden_tcr
compute D gex 607
argpartitions: 0.01 607 gex
argpartitions: 0.1 607 gex
calculate nndists: gex 0.1
DONE calculating nndists: gex 0.1
compute D tcr 607
argpartitions: 0.01 607 tcr
argpartitions: 0.1 607 tcr
calculate nndists: tcr 0.1
DONE calculating nndists: tcr 0.1
setup_tcr_cluster_names: [&#39;0&#39;, &#39;1_AV8&#39;, &#39;2&#39;, &#39;3_AV12&#39;, &#39;4&#39;, &#39;5_av21&#39;, &#39;6_BV20&#39;, &#39;7_av38&#39;, &#39;8_AV19&#39;, &#39;9_bv6&#39;, &#39;10_av29&#39;, &#39;11_AV26&#39;, &#39;12_AV9&#39;]
parse_tcr_junctions: 0 607
success_rate: 85.43
success_rate: 94.11
util.run_command: cmd= /lustre/groups/imm01/workspace/felix.drost/BestPractice2/notebooks/multimodal_vdj/conga/tcrdist_cpp/bin/calc_distributions -f conga_0.43453522085181073_tcr_clumping_tcrs.tsv -m 96 -d /lustre/groups/imm01/workspace/felix.drost/BestPractice2/notebooks/multimodal_vdj/conga/tcrdist_cpp/db/tcrdist_info_human.txt -a conga_0.43453522085181073_tcr_clumping_bg_achains.tsv -b conga_0.43453522085181073_tcr_clumping_bg_bchains.tsv -o conga_0.43453522085181073_tcr_clumping_dists.txt
Read 607 paired tcrs from file conga_0.43453522085181073_tcr_clumping_tcrs.tsv
Read 50000 single chain tcrs from file conga_0.43453522085181073_tcr_clumping_bg_achains.tsv
Read 50000 single chain tcrs from file conga_0.43453522085181073_tcr_clumping_bg_bchains.tsv
making conga_0.43453522085181073_tcr_clumping_dists.txt
......
util.run_command: cmd= /lustre/groups/imm01/workspace/felix.drost/BestPractice2/notebooks/multimodal_vdj/conga/tcrdist_cpp/bin/find_neighbors -f conga_0.43453522085181073_tcr_clumping_tcrs.tsv -t 96 -d /lustre/groups/imm01/workspace/felix.drost/BestPractice2/notebooks/multimodal_vdj/conga/tcrdist_cpp/db/tcrdist_info_human.txt -o conga_0.43453522085181073_tcr_clumping -a conga_0.43453522085181073_tcr_clumping_agroups.txt -b conga_0.43453522085181073_tcr_clumping_bgroups.txt
Read 607 paired tcrs from file conga_0.43453522085181073_tcr_clumping_tcrs.tsv
making conga_0.43453522085181073_tcr_clumping_nbr96_indices.txt and conga_0.43453522085181073_tcr_clumping_nbr96_distances.txt
......
saved tcr_clumping results to tsvfile: conga_tcr_clumping.tsv
making: conga_tcr_clumping.png
saved graph_vs_graph_stats results to tsvfile: conga_graph_vs_graph_stats.tsv
find_neighbor_neighbor_interactions:
find_neighbor_cluster_interactions:
find_neighbor_cluster_interactions:
find_neighbor_neighbor_interactions:
find_neighbor_cluster_interactions:
find_neighbor_cluster_interactions:
saved graph_vs_graph results to tsvfile: conga_graph_vs_graph.tsv
/lustre/groups/imm01/workspace/felix.drost/BestPractice2/notebooks/multimodal_vdj/conga/conga/correlations.py:1174: RuntimeWarning: invalid value encountered in log2
  logfoldchanges = np.log2((np.expm1(mean_fg) + 1e-9) /
making tcr score table, #features= 168
making tcr score table, #features= 168
making tcr score table, #features= 168
saved tcr_graph_vs_gex_features results to tsvfile: conga_tcr_graph_vs_gex_features.tsv
saved tcr_genes_vs_gex_features results to tsvfile: conga_tcr_genes_vs_gex_features.tsv
saved gex_graph_vs_tcr_features results to tsvfile: conga_gex_graph_vs_tcr_features.tsv
fastcluster is not available. Consider installing for faster performance.
computing pairwise X_pca_gex distances, (607, 40)
computing linkage matrix from pairwise X_pca_gex distances
making clustermap; num_features= 8
making: conga_graph_vs_features_gex_clustermap.png
writing help message to file: conga_graph_vs_features_gex_clustermap.png_README.txt
fastcluster is not available. Consider installing for faster performance.
computing pairwise X_pca_tcr distances, (607, 50)
computing linkage matrix from pairwise X_pca_tcr distances
making clustermap; num_features= 8
making: conga_graph_vs_features_tcr_clustermap.png
writing help message to file: conga_graph_vs_features_tcr_clustermap.png_README.txt
making: conga_tcr_graph_vs_gex_features_plot.png
writing help message to file: conga_tcr_graph_vs_gex_features_plot.png_README.txt
making: conga_tcr_graph_vs_gex_features_panels.png
writing help message to file: conga_tcr_graph_vs_gex_features_panels.png_README.txt
making: conga_tcr_genes_vs_gex_features_panels.png
writing help message to file: conga_tcr_genes_vs_gex_features_panels.png_README.txt
making: conga_gex_graph_vs_tcr_features_plot.png
writing help message to file: conga_gex_graph_vs_tcr_features_plot.png_README.txt
making: conga_gex_graph_vs_tcr_features_panels.png
writing help message to file: conga_gex_graph_vs_tcr_features_panels.png_README.txt
making: conga_graph_vs_summary.png
writing help message to file: conga_graph_vs_summary.png_README.txt
conga.correlations.find_hotspots:: Using pval_threshold = 0.05
computing H matrix      0    607     0.000000
computing H matrix      0    607     0.000000
conga.correlations.find_hotspots:: Using pval_threshold = 0.05
computing H matrix      0    607     0.000000
computing H matrix      0    607     0.000000
saved hotspot_features results to tsvfile: conga_hotspot_features.tsv
making: conga_hotspot_gex_features_0.010_nbrs_tcr_plot_umap_nbr_avg.png
making: conga_hotspot_gex_features_0.010_nbrs_tcr_plot_umap_nbr_avg.png
fastcluster is not available. Consider installing for faster performance.
computing pairwise X_pca_tcr distances, (607, 50)
computing linkage matrix from pairwise X_pca_tcr distances
making clustermap; num_features= 14
making: conga_hotspot_gex_features_0.010_nbrs_tcr_plot_clustermap_nbr_avg.png
making: conga_hotspot_tcr_features_0.010_nbrs_gex_plot_umap_nbr_avg.png
making: conga_hotspot_tcr_features_0.010_nbrs_gex_plot_umap_nbr_avg.png
fastcluster is not available. Consider installing for faster performance.
computing pairwise X_pca_gex distances, (607, 40)
computing linkage matrix from pairwise X_pca_gex distances
making clustermap; num_features= 2
making: conga_hotspot_tcr_features_0.010_nbrs_gex_plot_clustermap_nbr_avg.png
making: conga_hotspot_combo_features_0.010_nbrs_gex_plot_umap_nbr_avg.png
making: conga_hotspot_combo_features_0.010_nbrs_gex_plot_umap_nbr_avg.png
fastcluster is not available. Consider installing for faster performance.
computing pairwise X_pca_gex distances, (607, 40)
computing linkage matrix from pairwise X_pca_gex distances
making clustermap; num_features= 16
making: conga_hotspot_combo_features_0.010_nbrs_gex_plot_clustermap_nbr_avg.png
making: conga_hotspot_combo_features_0.010_nbrs_tcr_plot_umap_nbr_avg.png
making: conga_hotspot_combo_features_0.010_nbrs_tcr_plot_umap_nbr_avg.png
fastcluster is not available. Consider installing for faster performance.
computing pairwise X_pca_tcr distances, (607, 50)
computing linkage matrix from pairwise X_pca_tcr distances
making clustermap; num_features= 16
making: conga_hotspot_combo_features_0.010_nbrs_tcr_plot_clustermap_nbr_avg.png
making: conga_hotspot_gex_features_0.100_nbrs_tcr_plot_umap_nbr_avg.png
making: conga_hotspot_gex_features_0.100_nbrs_tcr_plot_umap_nbr_avg.png
fastcluster is not available. Consider installing for faster performance.
computing pairwise X_pca_tcr distances, (607, 50)
computing linkage matrix from pairwise X_pca_tcr distances
making clustermap; num_features= 4
making: conga_hotspot_gex_features_0.100_nbrs_tcr_plot_clustermap_nbr_avg.png
making: conga_hotspot_tcr_features_0.100_nbrs_gex_plot_umap_nbr_avg.png
making: conga_hotspot_tcr_features_0.100_nbrs_gex_plot_umap_nbr_avg.png
fastcluster is not available. Consider installing for faster performance.
computing pairwise X_pca_gex distances, (607, 40)
computing linkage matrix from pairwise X_pca_gex distances
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>making clustermap; num_features= 2
/home/icb/felix.drost/miniconda3/envs/best_practice/lib/python3.8/site-packages/seaborn/matrix.py:805: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).
  self._figure = plt.figure(figsize=figsize)
making: conga_hotspot_tcr_features_0.100_nbrs_gex_plot_clustermap_nbr_avg.png
making: conga_hotspot_combo_features_0.100_nbrs_gex_plot_umap_nbr_avg.png
making: conga_hotspot_combo_features_0.100_nbrs_gex_plot_umap_nbr_avg.png
writing help message to file: conga_hotspot_combo_features_0.100_nbrs_gex_plot_umap_nbr_avg.png_README.txt
fastcluster is not available. Consider installing for faster performance.
computing pairwise X_pca_gex distances, (607, 40)
computing linkage matrix from pairwise X_pca_gex distances
making clustermap; num_features= 6
making: conga_hotspot_combo_features_0.100_nbrs_gex_plot_clustermap_nbr_avg.png
writing help message to file: conga_hotspot_combo_features_0.100_nbrs_gex_plot_clustermap_nbr_avg.png_README.txt
making: conga_hotspot_combo_features_0.100_nbrs_tcr_plot_umap_nbr_avg.png
making: conga_hotspot_combo_features_0.100_nbrs_tcr_plot_umap_nbr_avg.png
writing help message to file: conga_hotspot_combo_features_0.100_nbrs_tcr_plot_umap_nbr_avg.png_README.txt
fastcluster is not available. Consider installing for faster performance.
computing pairwise X_pca_tcr distances, (607, 50)
computing linkage matrix from pairwise X_pca_tcr distances
making clustermap; num_features= 6
making: conga_hotspot_combo_features_0.100_nbrs_tcr_plot_clustermap_nbr_avg.png
writing help message to file: conga_hotspot_combo_features_0.100_nbrs_tcr_plot_clustermap_nbr_avg.png_README.txt
gex_indegree_bias: nbr_frac= 0.0100 DescribeResult(nobs=607, minmax=(0.0, 10.166666666666666), mean=1.0, variance=2.018243491015769, skewness=3.039506092173335, kurtosis=11.769470027172124)
tcr_indegree_bias: nbr_frac= 0.0100 DescribeResult(nobs=607, minmax=(0.0, 4.833333333333333), mean=1.0, variance=0.39998166483315, skewness=1.5095755547526837, kurtosis=4.479541829909785)
indegree_bias_correlation: nbr_frac= 0.0100 LinregressResult(slope=-0.021985010220304323, intercept=1.0219850102203043, rvalue=-0.04938481261795842, pvalue=0.22438694011492175, stderr=0.018076957449519378, intercept_stderr=0.03138796265299426)
gex_indegree_bias: nbr_frac= 0.1000 DescribeResult(nobs=607, minmax=(0.0, 4.75), mean=1.0, variance=0.981480564723139, skewness=1.101749242325695, kurtosis=0.4703199539035019)
tcr_indegree_bias: nbr_frac= 0.1000 DescribeResult(nobs=607, minmax=(0.13333333333333333, 5.266666666666667), mean=1.0, variance=0.29885038503850386, skewness=1.8851191979890307, kurtosis=7.218176649626399)
indegree_bias_correlation: nbr_frac= 0.1000 LinregressResult(slope=-0.019649747477813798, intercept=1.019649747477814, rvalue=-0.03560991325639319, pvalue=0.38113660770777213, stderr=0.022419857025912687, intercept_stderr=0.03154644878557476)
making: conga_results_summary.html
make_html_summary:: no results for content_tag graph_vs_graph_logos
make_html_summary:: no results for content_tag tcr_clumping_logos
make_html_summary:: no results for content_tag tcr_db_match
make_html_summary:: no results for content_tag tcr_db_match_plot
make_html_summary:: no results for content_tag gex_clusters_tcrdist_trees
make_html_summary:: no results for content_tag conga_threshold_tcrdist_tree
make_html_summary:: no results for content_tag batch_umaps
DONE
</pre></div>
</div>
</div>
</div>
</section>
<section id="id5">
<h5><span class="section-number">36.4.1.2.3. </span>Output<a class="headerlink" href="#id5" title="Permalink to this headline">#</a></h5>
<p>The results of the pipeline are stored in  <code class="docutils literal notranslate"><span class="pre">./data/res/conga_results_summary.html</span></code> with an explanation for each result. Additional annotation and plots are available under <code class="docutils literal notranslate"><span class="pre">./data/res_conga/</span></code>. Explaining all results, is out of scope of this tutorial. We, therefore, refer to the very detailed explanation in the web summary and the colab notebook.</p>
</section>
</section>
<section id="mvtcr">
<h4><span class="section-number">36.4.1.3. </span>mvTCR<a class="headerlink" href="#mvtcr" title="Permalink to this headline">#</a></h4>
<p>mvTCR by An et al. is a multiview Variational autoencoder that compresses TCR sequence and gene expression into a lower-dimensional representation <span id="id6">[<a class="reference internal" href="#id428" title="Yang An, Felix Drost, Fabian Theis, Benjamin Schubert, and Mohammad Lotfollahi. Jointly learning t-cell receptor and transcriptomic information to decipher the immune response. bioRxiv, 2021.">An <em>et al.</em>, 2021</a>]</span>. Two deep learning architectures - Transformer and Multi-layer perceptron - extract information from both TCR and GEX respectively, before they are fused to derive the joint space. Following, the trained models can be used to embed similar data.</p>
<p>The authors showed, that multi-modal models can better capture antigen specificity, than uni-modal embeddings on the data from <span id="id7">[<a class="reference internal" href="#id427" title="10x Genomics. A new way of exploring immunity–linking highly multiplexed antigen recognition to immune repertoire and phenotype. Tech. rep, 2019.">10x Genomics, 2019</a>]</span> for prediction and clustering. Additionally, they showed that cell type and cell functionality are preserved in the embedding space on a SARS-CoV-2 dataset from <span id="id8">[<a class="reference internal" href="#id429" title="David S Fischer, Meshal Ansari, Karolin I Wagner, Sebastian Jarosch, Yiqi Huang, Christoph H Mayr, Maximilian Strunz, Niklas J Lang, Elvira D’Ippolito, Monika Hammel, and others. Single-cell rna sequencing reveals ex vivo signatures of sars-cov-2-reactive t cells through ‘reverse phenotyping’. Nature communications, 12(1):1–14, 2021.">Fischer <em>et al.</em>, 2021</a>]</span>. Code and further explanations are available <a class="reference external" href="https://github.com/SchubertLab/mvTCR">here</a>.</p>
<p>mvTCR relies on the AnnData format and we therefore need to prepare our AnnData object for training. For that, the several functions are provided.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>import sys

sys.path.append(&quot;mvTCR&quot;)
sys.path.append(&quot;.&quot;)

import tcr_embedding.utils_training as utils
from tcr_embedding.models.model_selection import run_model_selection
from tcr_embedding.utils_preprocessing import group_shuffle_split, encode_tcr
</pre></div>
</div>
</div>
</div>
<p>The AnnData object requires a clonotype assignement (see notebook X). Since mvTCR uses both chains of the primary receptor, we will use this information for defining the clonotype as well. # todo =&gt; use the one provided</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata_mvtcr = adata_tc[~adata_tc.obs[&quot;IR_VJ_1_junction_aa&quot;].isna()].copy()
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>ir.tl.define_clonotypes(
    adata_mvtcr, key_added=&quot;clonotype&quot;, receptor_arms=&quot;all&quot;, dual_ir=&quot;primary_only&quot;
)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 728/728 [00:00&lt;00:00, 797.36it/s]
</pre></div>
</div>
</div>
</div>
<p>The model requires a numeric encoding of the CDR3α and CDR3β chain. For that, we need to provide the columns containing the amino acid sequences. Additional, the pad attribute indicates the maximal length of the sequence. If we do not want to embedd aditional data afterwards, we can set it to the maximal sequence length.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>pad = max(
    adata_mvtcr.obs[&quot;IR_VJ_1_junction_aa&quot;].str.len().max(),
    adata_mvtcr.obs[&quot;IR_VDJ_1_junction_aa&quot;].str.len().max(),
)
encode_tcr(
    adata_mvtcr,
    column_cdr3a=&quot;IR_VJ_1_junction_aa&quot;,
    column_cdr3b=&quot;IR_VDJ_1_junction_aa&quot;,
    pad=pad,
)
</pre></div>
</div>
</div>
</div>
<p>The data needs to be divided into a training and validation data. This is done by setting the values <code class="docutils literal notranslate"><span class="pre">train</span></code> or <code class="docutils literal notranslate"><span class="pre">val</span></code> for each cell in <code class="docutils literal notranslate"><span class="pre">adata.obs['set']</span></code>. By choosing the <code class="docutils literal notranslate"><span class="pre">clonotype</span></code> and 0.2, the dataset used to evaluate the model’s performance will contain approximately 20% of unique clonotypes to the training set. This setting can generally be used for data analysis.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>train, val = group_shuffle_split(adata_mvtcr, group_col=&quot;clonotype&quot;, val_split=0.2)
adata_mvtcr.obs[&quot;set&quot;] = &quot;train&quot;
adata_mvtcr.obs.loc[val.obs.index, &quot;set&quot;] = &quot;val&quot;
</pre></div>
</div>
</div>
</div>
<p>We need to provide various information, that are used during training:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>params_experiment = {
    &quot;study_name&quot;: &quot;test_haniffa&quot;,  # Name that identifies the study
    &quot;model_name&quot;: &quot;moe&quot;,  # Type of mixture model used during training, authors suggest using Mixture of Export (moe)
    &quot;balanced_sampling&quot;: &quot;clonotype&quot;,  # Column containing the id for clonotypes
    &quot;comet_workspace&quot;: None,  # Can be used to log experiments via comet-ml
    &quot;save_path&quot;: path_res,  # Output path, were the selected models are stored
}
</pre></div>
</div>
</div>
</div>
<p>mvTCR runs a hyperparameter optimization to choose the best model architecture. For general analysis, the optimization mode <code class="docutils literal notranslate"><span class="pre">pseudo_metric</span></code> is most suited. Here, the best model is determined by the capability of preserving multiple characteristics of the dataset. In the following, we choose identical TCR sequence (<code class="docutils literal notranslate"><span class="pre">clonotype</span></code>) and cell type (<code class="docutils literal notranslate"><span class="pre">functional.cluster</span></code>) both equally weighted (1 and 1). If after training the model, the latent space is dominated by one modality, we can rerun the model with different weightings.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>params_optimization = {
    &quot;name&quot;: &quot;pseudo_metric&quot;,
    &quot;prediction_labels&quot;: {&quot;clonotype&quot;: 1, &quot;full_clustering&quot;: 1},
}
</pre></div>
</div>
</div>
</div>
<p>Finally, we can run the model selection. Especially for larger datasets an extensive model search was used by the authors. As a showcase, we here only train one model. We recommend using a computer with GPU, even then the model will train several minutes.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>n_runs = 1
run_model_selection(adata_mvtcr, params_experiment, params_optimization, n_runs)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span><span class=" -Color -Color-Green">[I 2022-08-08 09:52:40,771]</span> A new study created in RDB with name: test_haniffa
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 200/200 [09:20&lt;00:00,  2.80s/it]
<span class=" -Color -Color-Green">[I 2022-08-08 10:02:06,885]</span> Trial 0 finished with value: 1.6926555212450274 and parameters: {&#39;dropout&#39;: 0.1, &#39;activation&#39;: &#39;linear&#39;, &#39;rna_hidden&#39;: 1500, &#39;hdim&#39;: 200, &#39;shared_hidden&#39;: 100, &#39;rna_num_layers&#39;: 1, &#39;tfmr_encoding_layers&#39;: 4, &#39;loss_weights_kl&#39;: 4.0428727350273357e-07, &#39;loss_weights_tcr&#39;: 0.034702669886504146, &#39;lr&#39;: 1.0994335574766187e-05, &#39;zdim&#39;: 50, &#39;tfmr_embedding_size&#39;: 16, &#39;tfmr_num_heads&#39;: 8, &#39;tfmr_dropout&#39;: 0.15000000000000002}. Best is trial 0 with value: 1.6926555212450274.
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Study statistics:
  Number of finished trials: 1
  Number of pruned trials: 0
  Number of complete trials: 1
Best trial: 
  trial_0
  Value: 1.6926555212450274
</pre></div>
</div>
</div>
</div>
<p>We now have a trained model, that we can use to embed data to a representation capturing TCR and GEX, simultaneously. For this, we will load the best model from our studies folder.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>best_trial = 0
path_model = f&quot;../data/res/trial_{best_trial}/best_model_by_metric.pt&quot;
model = utils.load_model(adata_mvtcr, path_model)
</pre></div>
</div>
</div>
</div>
<p>In a next step, we can now embed this data and assign the annotation to our new embedded AnnData object.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>mvtcr_embedding = model.get_latent(adata_mvtcr, metadata=[])
mvtcr_embedding.obs = adata_mvtcr.obs.copy()
</pre></div>
</div>
</div>
</div>
<p>This embedding can now be used for data analysis such as visualization. For this we will select the top 10 most abundant clonotypes within the sample.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>top_10_clones = mvtcr_embedding.obs[&quot;clonotype&quot;].value_counts().head(10).index
mvtcr_embedding.obs[&quot;clonotypes_top10&quot;] = mvtcr_embedding.obs[&quot;clonotype&quot;].apply(
    lambda x: x if x in top_10_clones else np.nan
)
</pre></div>
</div>
</div>
</div>
<p>Now, we will be using UMAPs to show cell type assignment and the top 10 clonotypes.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pp.neighbors(mvtcr_embedding)
sc.tl.umap(mvtcr_embedding)
sc.pl.umap(mvtcr_embedding, color=[&quot;full_clustering&quot;, &quot;clonotypes_top10&quot;], ncols=1)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>... storing &#39;clonotype&#39; as categorical
... storing &#39;set&#39; as categorical
... storing &#39;clonotypes_top10&#39; as categorical
</pre></div>
</div>
<img alt="../_images/multimodal_integration_97_1.png" src="../_images/multimodal_integration_97_1.png" />
</div>
</div>
<p>Here, we visualize the TCR-GEX embedding. It is clear to see, that the data separates by clonotype. However, the data forms additional subclusters and structures within the clones, which is influenced by the GEX. Depending on your analysis, you might need to adjust the weighting between GEX and TCR and run a longer parameter search. Otherwise, different downstream task such as clustering can now be performed on this embedding using Scanpy functions.</p>
</section>
</section>
<section id="integrating-bcr-and-gex">
<h3><span class="section-number">36.4.2. </span>Integrating BCR and GEX<a class="headerlink" href="#integrating-bcr-and-gex" title="Permalink to this headline">#</a></h3>
<p>The three models above were developed TCRs. However, Conga and mvTCR might be applicable for B cell due to similar structures of TCR and BCR. However, both models have not been evaluated for B cells in their publication. TESSA relies on a pre-trained TCR autoencoder and should therefore not be used for BCR-GEX integration. The authors recently published <a class="reference external" href="https://github.com/wooyongc/Benisse">Benisse</a>, which is similar to TESSA but applied on B cells <span id="id9">[<a class="reference internal" href="#id430" title="Ze Zhang, Woo Yong Chang, Kaiwen Wang, Yuqiu Yang, Xinlei Wang, Chen Yao, Tuoqi Wu, Li Wang, and Tao Wang. Interpreting the b-cell receptor repertoire with single-cell gene expression using benisse. Nature Machine Intelligence, pages 1–9, 2022.">Zhang <em>et al.</em>, 2022</a>]</span>.</p>
<p>Similar as for TCR, multimodal BCR integration tools are still novel and have not been used in standard analysis pipelines. Therefore, we recommend caution, when applying these models.</p>
<section id="benisse">
<h4><span class="section-number">36.4.2.1. </span>Benisse<a class="headerlink" href="#benisse" title="Permalink to this headline">#</a></h4>
<p>Benisse embeds the heavy-chain of the BCR into a numeric embedding using a pre-trained model. The model works on a clonotype level: i.e., GEX of cells with the same BCR are averaged to form representative per clone. Following, Benisse detects clones that are similar on a BCR- and GEX-level by constructing a sparse graph structure.</p>
<p>Since the model has a similar setup as TESSA, we will follow the same processing steps and only comment, when details differ from above.</p>
<p>Due to long runtime, we first subsample the B-cell dataset to 1000 cells of one donor.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata_benisse = adata_bc[~adata_bc.obs[&quot;IR_VDJ_1_junction_aa&quot;].isna()]
adata_benisse = adata_benisse[adata_benisse.obs[&quot;sample_id&quot;] == &quot;MH9143275&quot;]
sc.pp.subsample(adata_benisse, n_obs=1000, random_state=0)
print(f&quot;Amount of cells: {len(adata_benisse)}&quot;)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Amount of cells: 1000
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>df_bcr = adata_benisse.obs
df_bcr.index.name = &quot;contigs&quot;

df_bcr = df_bcr.rename(columns={&quot;IR_VDJ_1_junction_aa&quot;: &quot;cdr3&quot;})

df_bcr = df_bcr[[&quot;cdr3&quot;]]

df_bcr.to_csv(f&quot;{path_tmp}/BENISSE_bcrs.csv&quot;)
df_bcr.head(5)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>cdr3</th>
    </tr>
    <tr>
      <th>contigs</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>TTCTCAAAGAATTCCC-MH9143275</th>
      <td>CAGGSQWEVKLDYW</td>
    </tr>
    <tr>
      <th>TATCAGGAGTGAACAT-MH9143275</th>
      <td>CARPRSLIAAAGAFDIW</td>
    </tr>
    <tr>
      <th>CATCCACCAACGCACC-MH9143275</th>
      <td>CASEEVTMVRGVMFPYGMDVW</td>
    </tr>
    <tr>
      <th>GTCGTAATCACCGTAA-MH9143275</th>
      <td>CAREGMVYVDYW</td>
    </tr>
    <tr>
      <th>AACTCTTAGCACCGCT-MH9143275</th>
      <td>CATQNPGYSSSWDDRGAFDIW</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>We will filter the 1,000 most variable genes, as provided in the sample data file in the GitHub repository.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pp.highly_variable_genes(adata_benisse, n_top_genes=1000)
adata_benisse = adata_benisse[:, adata_benisse.var[&quot;highly_variable&quot;]].copy()

count_mat = adata_benisse.X.A
df_counts = pd.DataFrame(count_mat)

df_counts.index = adata_benisse.obs.index
df_counts.index.name = &quot;&quot;
df_counts.columns = adata_benisse.var.index

df_counts = df_counts.transpose()

df_counts.to_csv(f&quot;{path_tmp}/BENISSE_gex.csv&quot;)
df_counts.head()
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>TTCTCAAAGAATTCCC-MH9143275</th>
      <th>TATCAGGAGTGAACAT-MH9143275</th>
      <th>CATCCACCAACGCACC-MH9143275</th>
      <th>GTCGTAATCACCGTAA-MH9143275</th>
      <th>AACTCTTAGCACCGCT-MH9143275</th>
      <th>ACTATCTCACAGGAGT-MH9143275</th>
      <th>CAGTAACCAGGAATCG-MH9143275</th>
      <th>ACTGTCCTCTCGGACG-MH9143275</th>
      <th>TCGGGACCAGCATGAG-MH9143275</th>
      <th>ATTGGACAGCCATCGC-MH9143275</th>
      <th>...</th>
      <th>CACAGTAAGTGTCCAT-MH9143275</th>
      <th>GTCAAGTCAAGGACTG-MH9143275</th>
      <th>TAGTGGTAGCGATTCT-MH9143275</th>
      <th>GTGCAGCGTCTCCCTA-MH9143275</th>
      <th>CTGGTCTCATCCTTGC-MH9143275</th>
      <th>CTTTGCGAGCTGCCCA-MH9143275</th>
      <th>TACACGATCGGGAGTA-MH9143275</th>
      <th>GGGCACTAGGATGTAT-MH9143275</th>
      <th>TCACAAGGTGGTGTAG-MH9143275</th>
      <th>TGCGCAGTCTCAAGTG-MH9143275</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>HES4</th>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.00000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.00000</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>CPTP</th>
      <td>0.334078</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.00000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.44491</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.565442</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>THAP3</th>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.225355</td>
      <td>0.0</td>
      <td>0.45654</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.00000</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>VAMP3</th>
      <td>0.334078</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.00000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.00000</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.565442</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>CASP9</th>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.225355</td>
      <td>0.0</td>
      <td>0.00000</td>
      <td>0.355897</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.00000</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.291166</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 1000 columns</p>
</div></div></div>
</div>
<p>Unlike TESSA, BENISSE additionally requires an adapted contigs file as specified from the 10x cellranger output. Here, we will filter the original contigs file to contain only the barcodes of the selected donors.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>path_bcr_contigs = f&quot;{path_data}/BCR_00_read_aligned.csv&quot;
df_bcr_contigs = pd.read_csv(path_bcr_contigs, index_col=0, low_memory=False)
df_bcr_contigs = df_bcr_contigs[df_bcr_contigs[&quot;barcode&quot;].isin(adata_benisse.obs.index)]
df_bcr_contigs.to_csv(f&quot;{path_tmp}/BENISSE_bcr_contigs.csv&quot;)
</pre></div>
</div>
</div>
</div>
<section id="id10">
<h5><span class="section-number">36.4.2.1.1. </span>Running the model<a class="headerlink" href="#id10" title="Permalink to this headline">#</a></h5>
<p>First, we need to embed the BCR sequences via the pre-trained encoder:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>settings_embedding = {
    &quot;input_data&quot;: f&quot;{path_tmp}/BENISSE_bcrs.csv&quot;,
    &quot;output_data&quot;: f&quot;{path_res}/BENISSE_encoded_bcrs.csv&quot;,
    &quot;cuda&quot;: &quot;True&quot;,
}
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cmd_bcr_embedding = &quot;source ~/.bashrc &amp;&amp; &quot;
cmd_bcr_embedding += &quot;conda activate BENISSE &amp;&amp; &quot;

cmd_bcr_embedding += &quot;python Benisse/AchillesEncoder.py&quot;
for key, value in settings_embedding.items():
    cmd_bcr_embedding += f&quot; --{key} {value}&quot;
cmd_bcr_embedding
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&#39;source ~/.bashrc &amp;&amp; conda activate BENISSE &amp;&amp; python Benisse/AchillesEncoder.py --input_data ./data/tmp/BENISSE_bcrs.csv --output_data ./data/res/BENISSE_encoded_bcrs.csv --cuda True&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>!$cmd_bcr_embedding
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Reading file:
./data/tmp/BENISSE_bcrs.csv
Loading trained model...
Embedding BCR sequences...
BCR successfully encoded. The output file is generated here:  ./data/res/BENISSE_encoded_bcrs.csv
</pre></div>
</div>
</div>
</div>
<p>Following, we need to create the graph representation. We will use the default values provided by the authors here.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>max_iter = 100
cmd_benisse = [
    &quot;source ~/.bashrc &amp;&amp;&quot;,
    &quot;conda activate BENISSE &amp;&amp;&quot;,
    &quot;export R_LIBS=$CONDA_PREFIX/lib/R/library &amp;&amp;&quot;,
    &quot;Rscript Benisse/Benisse.R&quot;,
    f&quot;{path_tmp}/BENISSE_gex.csv&quot;,
    f&quot;{path_tmp}/BENISSE_bcr_contigs.csv&quot;,
    f&quot;{path_res}/BENISSE_encoded_bcrs.csv&quot;,
    f&quot;{path_res}&quot;,
    f&quot;1610 1 {max_iter} 1 1 10 1e-10&quot;,
]
cmd_benisse = &quot; &quot;.join(cmd_benisse)
cmd_benisse
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&#39;source ~/.bashrc &amp;&amp; conda activate BENISSE &amp;&amp; export R_LIBS=$CONDA_PREFIX/lib/R/library &amp;&amp; Rscript Benisse/Benisse.R ./data/tmp/BENISSE_gex.csv ./data/tmp/BENISSE_bcr_contigs.csv ./data/res/BENISSE_encoded_bcrs.csv ./data/res 1610 1 100 1 1 10 1e-10&#39;
</pre></div>
</div>
</div>
</div>
<p>Even with reduced number of iteration, running the code might take several minutes.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>!$cmd_benisse
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Warning message:
package ‘data.table’ was built under R version 4.0.5 
Loading required package: igraph

Attaching package: ‘igraph’

The following objects are masked from ‘package:stats’:

    decompose, spectrum

The following object is masked from ‘package:base’:

    union

Warning messages:
1: package ‘ggplot2’ was built under R version 4.0.5 
2: package ‘igraph’ was built under R version 4.0.3 
?25h?25h?25h?25h?25h?25h?25h?25h?25h?25h?25h?25h?25h?25h?25h?25h?25h?25h?25h?25h?25h?25h?25h?25h?25h?25h?25h?25h?25h?25h?25h[1] &quot;Begin initiation.&quot;
[1] &quot;Initiation success.&quot;
[1] &quot;Iteration: 1&quot;
[1] &quot;Iteration: 2&quot;
[1] &quot;Iteration: 3&quot;
[1] &quot;Iteration: 4&quot;
[1] &quot;Iteration: 5&quot;
[1] &quot;Iteration: 6&quot;
[1] &quot;Iteration: 7&quot;
[1] &quot;Iteration: 8&quot;
[1] &quot;Iteration: 9&quot;
[1] &quot;Iteration: 10&quot;
[1] &quot;Iteration: 11&quot;
[1] &quot;Iteration: 12&quot;
[1] &quot;Iteration: 13&quot;
[1] &quot;Iteration: 14&quot;
[1] &quot;Iteration: 15&quot;
[1] &quot;Iteration: 16&quot;
[1] &quot;Iteration: 17&quot;
[1] &quot;Iteration: 18&quot;
[1] &quot;Iteration: 19&quot;
[1] &quot;Iteration: 20&quot;
[1] &quot;Iteration: 21&quot;
[1] &quot;Iteration: 22&quot;
[1] &quot;Iteration: 23&quot;
[1] &quot;Iteration: 24&quot;
[1] &quot;Iteration: 25&quot;
[1] &quot;Iteration: 26&quot;
[1] &quot;Iteration: 27&quot;
[1] &quot;Iteration: 28&quot;
[1] &quot;Iteration: 29&quot;
[1] &quot;Iteration: 30&quot;
[1] &quot;Iteration: 31&quot;
[1] &quot;Iteration: 32&quot;
[1] 0
[1] &quot;Convergence: iteration 32&quot;
?25h?25h?25h?25h?25h?25h?25h?25h?25h?25h?25h?25h?25h
</pre></div>
</div>
</div>
</div>
</section>
<section id="id11">
<h5><span class="section-number">36.4.2.1.2. </span>Output<a class="headerlink" href="#id11" title="Permalink to this headline">#</a></h5>
<p>We now recieve several output files, which are described in more detail in the authors GitHub repository. In the following, we will examine:</p>
<ul class="simple">
<li><p><strong>connectionplot.pdf</strong>: Visualisation of the derived clustering</p></li>
<li><p><strong>clone_annotation.csv</strong>: cluster assignment by cell</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from IPython.display import IFrame

IFrame(f&quot;{path_res}/connectionplot.pdf&quot;, width=600, height=600)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
<iframe
    width="600"
    height="600"
    src="./data/res/connectionplot.pdf"
    frameborder="0"
    allowfullscreen

></iframe>
</div></div>
</div>
<p>The connection plot prowides us with a visualisation of the graph build by Benisse. Each clonotype is represented by a node in the learned embedding space. These clustering is informed by BCR as well as gene expression similarity.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>benisse_clusters = pd.read_csv(f&quot;{path_res}/clone_annotation.csv&quot;, index_col=0)
benisse_clusters.index = [el.replace(&quot;.&quot;, &quot;-&quot;) for el in benisse_clusters.index]
adata_benisse_out = adata_benisse[
    adata_benisse.obs.index.isin(benisse_clusters.index)
].copy()
adata_benisse_out.obs[&quot;benisse_clusters&quot;] = benisse_clusters.loc[
    adata_benisse_out.obs.index
][&quot;graph_label&quot;]
sc.pp.neighbors(adata_benisse_out)
sc.tl.umap(adata_benisse_out)

top_10_clusters = (
    adata_benisse_out.obs[&quot;benisse_clusters&quot;].value_counts().head(10).index
)
adata_benisse_out.obs[&quot;clusters_top10&quot;] = adata_benisse_out.obs[
    &quot;benisse_clusters&quot;
].apply(lambda x: x if x in top_10_clusters else np.nan)
sc.pl.umap(adata_benisse_out, color=&quot;clusters_top10&quot;)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>... storing &#39;benisse_clusters&#39; as categorical
... storing &#39;clusters_top10&#39; as categorical
</pre></div>
</div>
<img alt="../_images/multimodal_integration_118_1.png" src="../_images/multimodal_integration_118_1.png" />
</div>
</div>
<p>Here, we projected the ten largest clusters upon the transcriptome space. While some clusters follow the same phenotype, other clusters separate in the RNA space. The cluster annotation could be used for downstream analysis (see above) on BCR or transcriptomic site or to infer ancestral relationships between cells (see paper).</p>
</section>
</section>
</section>
</section>
<section id="key-takeaways">
<h2><span class="section-number">36.5. </span>Key Takeaways<a class="headerlink" href="#key-takeaways" title="Permalink to this headline">#</a></h2>
<ul class="simple">
<li><p>For multimodal datasets of AIR and GEX, typically, one modality is used for grouping the cells to perform standard uni-modal analysis on the other modality (e.g. Sequence analysis on leiden clusters).</p></li>
<li><p>Cell functionality (determined by IR) and cell state (observed via GEX) are interlinked. It has been shown, that cells with alike AIR sequences can share similar phenotypes.</p></li>
<li><p>Due to the inherent structural difference between count matrices (GEX) and amino acid sequences (IR), it is difficult to directly fuse both modalities. Therefore, several methods were developed, recently, to utilize paired GEX-AIR data for to derive clusters, or an embedding. However, these approaches are still novel and not part of a standard analysis pipeline.</p></li>
</ul>
</section>
<section id="questions">
<h2><span class="section-number">36.6. </span>Questions<a class="headerlink" href="#questions" title="Permalink to this headline">#</a></h2>
<p>Why could it be useful to integrate AIR sequence information with gene expression?</p>
<ul class="simple">
<li><p>GEX can be used to improve AIR sequence reads.</p></li>
<li><p>+ Both modalities provide different insights into the cell, while still being interdependent.</p></li>
<li><p>Since both modalities capture the same information, integrating them provides an additional quality check.</p></li>
<li><p>For most cells either GEX or AIR is measured. Integrating thereby allows analysis of all cells.</p></li>
</ul>
<p>What information provides us the AIR sequence, that is not directly captured in GEX?</p>
<ul class="simple">
<li><p>A count matrix between cell and antibody-tagged epitope bindings.</p></li>
<li><p>The IR sequence can be used for demultiplexing between different donors.</p></li>
<li><p>+ The cell’s clonotype and thereby cell ancestory is defined by the AIR sequence.</p></li>
<li><p>+ The AIR sequence determines specificity and is therefor a barcode for recognizing the same epitope.</p></li>
</ul>
<p>On what premise rely multi-modal integration approaches?</p>
<ul class="simple">
<li><p>+ Cells of same or alike AIRs often have a similar phenotype.</p></li>
<li><p>Information of AIR and GEX provide orthogonal information to each other, since they are independent.</p></li>
<li><p>Knowledge is transfered between large gene expression datasets into which AIR data can be mapped.</p></li>
<li><p>Each cell occurs clonally expanded and thereby provides multiple gene profiles.</p></li>
</ul>
</section>
<section id="references">
<h2><span class="section-number">36.7. </span>References<a class="headerlink" href="#references" title="Permalink to this headline">#</a></h2>
<div class="docutils container" id="id12">
<dl class="citation">
<dt class="label" id="id427"><span class="brackets"><a class="fn-backref" href="#id7">10xG19</a></span></dt>
<dd><p>10x Genomics. A new way of exploring immunity–linking highly multiplexed antigen recognition to immune repertoire and phenotype. <em>Tech. rep</em>, 2019.</p>
</dd>
<dt class="label" id="id428"><span class="brackets"><a class="fn-backref" href="#id6">ADT+21</a></span></dt>
<dd><p>Yang An, Felix Drost, Fabian Theis, Benjamin Schubert, and Mohammad Lotfollahi. Jointly learning t-cell receptor and transcriptomic information to decipher the immune response. <em>bioRxiv</em>, 2021.</p>
</dd>
<dt class="label" id="id429"><span class="brackets"><a class="fn-backref" href="#id8">FAW+21</a></span></dt>
<dd><p>David S Fischer, Meshal Ansari, Karolin I Wagner, Sebastian Jarosch, Yiqi Huang, Christoph H Mayr, Maximilian Strunz, Niklas J Lang, Elvira D’Ippolito, Monika Hammel, and others. Single-cell rna sequencing reveals ex vivo signatures of sars-cov-2-reactive t cells through ‘reverse phenotyping’. <em>Nature communications</em>, 12(1):1–14, 2021.</p>
</dd>
<dt class="label" id="id411"><span class="brackets"><a class="fn-backref" href="#id2">GHN+17</a></span></dt>
<dd><p>Jacob Glanville, Huang Huang, Allison Nau, Olivia Hatton, Lisa E Wagar, Florian Rubelt, Xuhuai Ji, Arnold Han, Sheri M Krams, Christina Pettus, and others. Identifying specificity groups in the t cell receptor repertoire. <em>Nature</em>, 547(7661):94–98, 2017.</p>
</dd>
<dt class="label" id="id426"><span class="brackets"><a class="fn-backref" href="#id4">SGC+22</a></span></dt>
<dd><p>Stefan A Schattgen, Kate Guion, Jeremy Chase Crawford, Aisha Souquette, Alvaro Martinez Barrio, Michael JT Stubbington, Paul G Thomas, and Philip Bradley. Integrating t cell receptor sequences and transcriptional profiles by clonotype neighbor graph analysis (conga). <em>Nature Biotechnology</em>, 40(1):54–63, 2022.</p>
</dd>
<dt class="label" id="id425"><span class="brackets"><a class="fn-backref" href="#id3">YSW+19</a></span></dt>
<dd><p>Kathryn E Yost, Ansuman T Satpathy, Daniel K Wells, Yanyan Qi, Chunlin Wang, Robin Kageyama, Katherine L McNamara, Jeffrey M Granja, Kavita Y Sarin, Ryanne A Brown, and others. Clonal replacement of tumor-specific t cells following pd-1 blockade. <em>Nature medicine</em>, 25(8):1251–1259, 2019.</p>
</dd>
<dt class="label" id="id430"><span class="brackets"><a class="fn-backref" href="#id9">ZCW+22</a></span></dt>
<dd><p>Ze Zhang, Woo Yong Chang, Kaiwen Wang, Yuqiu Yang, Xinlei Wang, Chen Yao, Tuoqi Wu, Li Wang, and Tao Wang. Interpreting the b-cell receptor repertoire with single-cell gene expression using benisse. <em>Nature Machine Intelligence</em>, pages 1–9, 2022.</p>
</dd>
<dt class="label" id="id424"><span class="brackets"><a class="fn-backref" href="#id1">ZXW+21</a></span></dt>
<dd><p>Ze Zhang, Danyi Xiong, Xinlei Wang, Hongyu Liu, and Tao Wang. Mapping the functional landscape of t cell receptor repertoires by single-t cell transcriptomics. <em>Nature methods</em>, 18(1):92–99, 2021.</p>
</dd>
</dl>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./air_repertoire"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
    <a class='left-prev' id="prev-link" href="specificity.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title"><span class="section-number">35. </span>Specificity analysis</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="../multimodal_integration/paired_integration.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title"><span class="section-number">37. </span>Paired integration</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By Lukas Heumos, Anna Schaar<br/>
  
      &copy; Copyright 2022.<br/>
    <div class="extra_footer">
      <div>
  Brought to you by
  <a href="https://www.github.com/theislab">Theislab</a>, with many thanks to the single-cell community as a whole!
</div>

    </div>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>