
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>38. Advanced integration &#8212; Single-cell best practices</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="../_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" href="../_static/styles/sphinx-book-theme.css?digest=5115cc725059bd94278eecd172e13a965bf8f5a9" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/book.css" />
    <link rel="stylesheet" type="text/css" href="../_static/design-style.b7bb847fb20b106c3d81b95245e65545.min.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?digest=9c920249402e914e316237a7dbc6769907cce411"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js"></script>
    <script async="async" kind="hypothesis" src="https://hypothes.is/embed.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../_static/book.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="39. Reproducibility" href="../reproducibility/introduction.html" />
    <link rel="prev" title="37. Paired integration" href="paired_integration.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="en">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../_static/logo.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">Single-cell best practices</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Introduction
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../introduction/prior_art.html">
   1. Prior art
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../introduction/scrna_seq.html">
   2. Single-cell RNA sequencing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../introduction/raw_data_processing.html">
   3. Raw data processing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../introduction/analysis_tools.html">
   4. Analysis frameworks and tools
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../introduction/data_infrastructure.html">
   5. Data infrastructure
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../introduction/interoperability.html">
   6. Interoperability
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Preprocessing and visualization
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../preprocessing_visualization/quality_control.html">
   7. Quality Control
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../preprocessing_visualization/normalization.html">
   8. Normalization
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../preprocessing_visualization/feature_selection.html">
   9. Feature selection
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../preprocessing_visualization/dimensionality_reduction.html">
   10. Dimensionality Reduction
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Identifying cellular structure
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../cellular_structure/clustering.html">
   11. Clustering
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cellular_structure/annotation.html">
   12. Annotation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cellular_structure/integration.html">
   13. Data integration
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Inferring trajectories
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../trajectories/pseudotemporal.html">
   14. Pseudotemporal ordering
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../trajectories/rna_velocity.html">
   15. RNA velocity
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../trajectories/lineage_tracing.html">
   16. Lineage tracing
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Dealing with conditions
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../conditions/differential_gene_expression.html">
   17. Differential gene expression analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../conditions/compositional.html">
   18. Compositional analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../conditions/gsea_pathway.html">
   19. Gene set enrichment and pathway analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../conditions/perturbation_modeling.html">
   20. Perturbation modeling
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Modeling mechanisms
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../mechanisms/grns.html">
   21. Gene regulatory networks
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../mechanisms/cell_cell_communication.html">
   22. Cell-cell communication
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Deconvolution
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../deconvolution/bulk_deconvolution.html">
   23. Bulk deconvolution
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Surface protein
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/quality_control.html">
   24. Quality control
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/normalization.html">
   25. Normalization
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/doublet_detection.html">
   26. Doublet detection
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/dimensionality_reduction.html">
   27. Dimensionality Reduction
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/batch_correction.html">
   28. Batch correction
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/annotation.html">
   29. Annotation
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  AIR repertoire
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../air_repertoire/ir_profiling.html">
   30. Immune Receptor Profiling
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../air_repertoire/clonotype.html">
   31. Clonotype analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../air_repertoire/specificity.html">
   35. Specificity analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../air_repertoire/multimodal_integration.html">
   36. Integrating AIR and transcriptomics
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multimodal integration
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="paired_integration.html">
   37. Paired integration
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   38. Advanced integration
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Reproducibility
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../reproducibility/introduction.html">
   39. Reproducibility
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Outlook
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../outlook.html">
   40. Outlook
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Acknowledgements
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../acknowledgements.html">
   41. Acknowledgements
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Glossary
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../glossary.html">
   42. Glossary
  </a>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
            <div class="navbar_extra_footer">
            Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
            </div>
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right">
<div class="menu-dropdown menu-dropdown-launch-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Launch interactive content">
      <i class="fas fa-rocket"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://mybinder.org/v2/gh/theislab/single-cell-best-practices/master?urlpath=tree/multimodal_integration/advanced_integration.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Launch on Binder"
>
  

<span class="headerbtn__icon-container">
  
    <img src="../_static/images/logo_binder.svg">
  </span>
<span class="headerbtn__text-container">Binder</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

<button onclick="toggleFullScreen()"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="bottom"
title="Fullscreen mode"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>

<div class="menu-dropdown menu-dropdown-repository-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Source repositories">
      <i class="fab fa-github"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://github.com/theislab/single-cell-best-practices"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Source repository"
>
  

<span class="headerbtn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="headerbtn__text-container">repository</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/theislab/single-cell-best-practices/issues/new?title=Issue%20on%20page%20%2Fmultimodal_integration/advanced_integration.html&body=Your%20issue%20content%20here."
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Open an issue"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="headerbtn__text-container">open issue</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/theislab/single-cell-best-practices/edit/master/multimodal_integration/advanced_integration.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Edit this page"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="headerbtn__text-container">suggest edit</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

<div class="menu-dropdown menu-dropdown-download-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Download this page">
      <i class="fas fa-download"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="../_sources/multimodal_integration/advanced_integration.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Download source file"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="headerbtn__text-container">.ipynb</span>
</a>

      </li>
      
      <li>
        
<button onclick="printPdf(this)"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="left"
title="Print to PDF"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="headerbtn__text-container">.pdf</span>
</button>

      </li>
      
    </ul>
  </div>
</div>
<label for="__page-toc"
  class="headerbtn headerbtn-page-toc"
  
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-list"></i>
  </span>

</label>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
    <div class="tocsection onthispage pt-5 pb-3">
        <i class="fas fa-list"></i> Contents
    </div>
    <nav id="bd-toc-nav" aria-label="Page">
        <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#motivation">
   38.1. Motivation
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#environment-setup">
   38.2. Environment setup
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#prepare-data">
   38.3. Prepare data
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#unpaired-integration-with-graph-linked-unified-embedding-glue">
   38.4. Unpaired integration with Graph Linked Unified Embedding (GLUE)
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#graph-construction">
     38.4.1. Graph construction
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#configure-data">
     38.4.2. Configure data
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#integration-with-partially-overlapping-data-and-query-to-reference-mapping-with-totalvi">
   38.5. Integration with partially overlapping data and query-to-reference mapping with totalVI
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#query-to-reference-mapping">
     38.5.1. Query-to-reference mapping
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#query-mapping-and-imputation-with-seurat-s-wnn">
   38.6. Query mapping and imputation with Seurat’s WNN
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#mapping-adt-onto-an-rna-atlas-with-bridge-integration">
   38.7. Mapping ADT onto an RNA atlas with bridge integration
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#trimodal-integration-and-query-to-reference-mapping-with-multigrate">
   38.8. Trimodal integration and query-to-reference mapping with multigrate
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#building-the-reference">
     38.8.1. Building the reference
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#querying-the-trimodal-reference">
     38.8.2. Querying the trimodal reference
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#session-info">
   38.9. Session info
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#references">
   38.10. References
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#contributors">
   38.11. Contributors
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#authors">
     38.11.1. Authors
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#reviewers">
     38.11.2. Reviewers
    </a>
   </li>
  </ul>
 </li>
</ul>

    </nav>
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Advanced integration</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#motivation">
   38.1. Motivation
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#environment-setup">
   38.2. Environment setup
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#prepare-data">
   38.3. Prepare data
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#unpaired-integration-with-graph-linked-unified-embedding-glue">
   38.4. Unpaired integration with Graph Linked Unified Embedding (GLUE)
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#graph-construction">
     38.4.1. Graph construction
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#configure-data">
     38.4.2. Configure data
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#integration-with-partially-overlapping-data-and-query-to-reference-mapping-with-totalvi">
   38.5. Integration with partially overlapping data and query-to-reference mapping with totalVI
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#query-to-reference-mapping">
     38.5.1. Query-to-reference mapping
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#query-mapping-and-imputation-with-seurat-s-wnn">
   38.6. Query mapping and imputation with Seurat’s WNN
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#mapping-adt-onto-an-rna-atlas-with-bridge-integration">
   38.7. Mapping ADT onto an RNA atlas with bridge integration
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#trimodal-integration-and-query-to-reference-mapping-with-multigrate">
   38.8. Trimodal integration and query-to-reference mapping with multigrate
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#building-the-reference">
     38.8.1. Building the reference
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#querying-the-trimodal-reference">
     38.8.2. Querying the trimodal reference
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#session-info">
   38.9. Session info
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#references">
   38.10. References
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#contributors">
   38.11. Contributors
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#authors">
     38.11.1. Authors
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#reviewers">
     38.11.2. Reviewers
    </a>
   </li>
  </ul>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                
  <section class="tex2jax_ignore mathjax_ignore" id="advanced-integration">
<span id="multimodal-integration-advanced-integration"></span><h1><span class="section-number">38. </span>Advanced integration<a class="headerlink" href="#advanced-integration" title="Permalink to this headline">#</a></h1>
<section id="motivation">
<h2><span class="section-number">38.1. </span>Motivation<a class="headerlink" href="#motivation" title="Permalink to this headline">#</a></h2>
<p>In this notebook, we showcase more advanced methods and techniques for multimodal integration. Examples of such advanced techniques are unpaired integration, integration of partially overlapping data or multimodal query-to-reference mapping. These are especially useful when the measurements of various modalities were not done jointly per cell, but originate from different experiments.</p>
<p>We discuss each of these cases in more detail in the corresponding sections below.</p>
</section>
<section id="environment-setup">
<h2><span class="section-number">38.2. </span>Environment setup<a class="headerlink" href="#environment-setup" title="Permalink to this headline">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>import warnings

warnings.simplefilter(action=&quot;ignore&quot;)

import scglue
import networkx as nx
import pandas as pd
import numpy as np
import scanpy as sc
import anndata
import multigrate as mtg
import scvi
from scvi.model import TOTALVI

import rpy2.rinterface_lib.callbacks
import anndata2ri
import logging

import seaborn as sns
import matplotlib.pyplot as plt

from rpy2.robjects import r
from rpy2.robjects import pandas2ri

rpy2.rinterface_lib.callbacks.logger.setLevel(logging.ERROR)

pandas2ri.activate()
anndata2ri.activate()

%load_ext rpy2.ipython
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Global seed set to 0
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>%%R
suppressPackageStartupMessages({
    library(SingleCellExperiment)
    library(Seurat)
})
set.seed(123)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>    WARNING: The R package &quot;reticulate&quot; does not
    consider that it could be called from a Python process. This
    results in a quasi-obligatory segfault when rpy2 is evaluating
    R code using it. On the hand, rpy2 is accounting for the
    fact that it might already be running embedded in a Python
    process. This is why:
    - Python -&gt; rpy2 -&gt; R -&gt; reticulate: crashes
    - R -&gt; reticulate -&gt; Python -&gt; rpy2: works

    The issue with reticulate is tracked here:
    https://github.com/rstudio/reticulate/issues/208
    
</pre></div>
</div>
</div>
</div>
</section>
<section id="prepare-data">
<h2><span class="section-number">38.3. </span>Prepare data<a class="headerlink" href="#prepare-data" title="Permalink to this headline">#</a></h2>
<p>We continue using Multiome and CITE-seq data from the NeurIPS 2021 single cell competition <span id="id1">[<a class="reference internal" href="#id445" title="Malte D Luecken, Daniel Bernard Burkhardt, Robrecht Cannoodt, Christopher Lance, Aditi Agrawal, Hananeh Aliee, Ann T Chen, Louise Deconinck, Angela M Detweiler, Alejandro A Granados, Shelly Huynh, Laura Isacco, Yang Joon Kim, Dominik Klein, BONY DE KUMAR, Sunil Kuppasani, Heiko Lickert, Aaron McGeever, Honey Mekonen, Joaquin Caceres Melgarejo, Maurizio Morri, Michaela Müller, Norma Neff, Sheryl Paul, Bastian Rieck, Kaylie Schneider, Scott Steelman, Michael Sterr, Daniel J. Treacy, Alexander Tong, Alexandra-Chloe Villani, Guilin Wang, Jia Yan, Ce Zhang, Angela Oliveira Pisco, Smita Krishnaswamy, Fabian J Theis, and Jonathan M. Bloom. A sandbox for prediction and integration of dna, rna, and proteins in single cells. In Thirty-fifth Conference on Neural Information Processing Systems Datasets and Benchmarks Track (Round 2). 2021. URL: https://openreview.net/forum?id=gN35BGa1Rt.">Luecken <em>et al.</em>, 2021</a>]</span>.</p>
<p>We will need:</p>
<ul class="simple">
<li><p>RNA-seq part of the multiome and ADT from the CITE-seq data for unpaired integration with <a class="reference external" href="#glue">GLUE</a></p></li>
<li><p>paired gene expression and protein from the CITE-seq data for query-to-reference mapping with <a class="reference external" href="#totalvi">totalVI</a></p></li>
<li><p>RNA-seq query from the CITE-seq data for query mapping with <a class="reference external" href="#seurat">Seurat v4</a></p></li>
<li><p>CITE-seq data for <a class="reference external" href="#bridge">bridge</a> mapping</p></li>
<li><p>NeurIPS multiome and NeurIPS CITE-seq for trimodal integration and query-to-reference mapping with <a class="reference external" href="#multigrate">multigrate</a></p></li>
</ul>
<p>Since we showcase query-to-reference mapping functionality of some methods, we split the data into a reference and a query. For the methods that we only use for integration, we also use the reference batches.</p>
<p>We note that the batch names for Multiome and CITE-seq data are the same, but they are actually not the same cells.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cite_reference_batches = [
    &quot;s1d1&quot;,
    &quot;s1d2&quot;,
    &quot;s1d3&quot;,
]  # need for totalVI, multigrate and bridge (RNA part)
multiome_reference_batches = [&quot;s1d1&quot;, &quot;s1d2&quot;, &quot;s1d3&quot;]  # need for GLUE and multigrate
cite_query_batches = [&quot;s2d1&quot;, &quot;s2d4&quot;]  # need for totalVI, multigrate and bridge
multiome_query_batches = [&quot;s2d1&quot;, &quot;s2d4&quot;]  # need for multigrate
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>rna_multiome = sc.read(
    &quot;/lustre/groups/ml01/workspace/anastasia.litinetskaya/data/neurips-multiome/rna_hvg.h5ad&quot;
)
rna_multiome
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 69249 × 4000
    obs: &#39;GEX_pct_counts_mt&#39;, &#39;GEX_n_counts&#39;, &#39;GEX_n_genes&#39;, &#39;GEX_size_factors&#39;, &#39;GEX_phase&#39;, &#39;ATAC_nCount_peaks&#39;, &#39;ATAC_atac_fragments&#39;, &#39;ATAC_reads_in_peaks_frac&#39;, &#39;ATAC_blacklist_fraction&#39;, &#39;ATAC_nucleosome_signal&#39;, &#39;cell_type&#39;, &#39;batch&#39;, &#39;ATAC_pseudotime_order&#39;, &#39;GEX_pseudotime_order&#39;, &#39;Samplename&#39;, &#39;Site&#39;, &#39;DonorNumber&#39;, &#39;Modality&#39;, &#39;VendorLot&#39;, &#39;DonorID&#39;, &#39;DonorAge&#39;, &#39;DonorBMI&#39;, &#39;DonorBloodType&#39;, &#39;DonorRace&#39;, &#39;Ethnicity&#39;, &#39;DonorGender&#39;, &#39;QCMeds&#39;, &#39;DonorSmoker&#39;
    var: &#39;feature_types&#39;, &#39;gene_id&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;
    uns: &#39;ATAC_gene_activity_var_names&#39;, &#39;Site_colors&#39;, &#39;batch_colors&#39;, &#39;cell_type_colors&#39;, &#39;dataset_id&#39;, &#39;genome&#39;, &#39;hvg&#39;, &#39;organism&#39;
    obsm: &#39;ATAC_gene_activity&#39;, &#39;ATAC_lsi_full&#39;, &#39;ATAC_lsi_red&#39;, &#39;ATAC_umap&#39;, &#39;GEX_X_pca&#39;, &#39;GEX_X_umap&#39;, &#39;X_umap&#39;
    layers: &#39;counts&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>atac_multiome = sc.read(
    &quot;/lustre/groups/ml01/workspace/anastasia.litinetskaya/data/trimodal_neurips/atac_hvf_muon.h5ad&quot;
)
atac_multiome
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 69249 × 20000
    obs: &#39;GEX_pct_counts_mt&#39;, &#39;GEX_n_counts&#39;, &#39;GEX_n_genes&#39;, &#39;GEX_size_factors&#39;, &#39;GEX_phase&#39;, &#39;ATAC_nCount_peaks&#39;, &#39;ATAC_atac_fragments&#39;, &#39;ATAC_reads_in_peaks_frac&#39;, &#39;ATAC_blacklist_fraction&#39;, &#39;ATAC_nucleosome_signal&#39;, &#39;cell_type&#39;, &#39;batch&#39;, &#39;ATAC_pseudotime_order&#39;, &#39;GEX_pseudotime_order&#39;, &#39;Samplename&#39;, &#39;Site&#39;, &#39;DonorNumber&#39;, &#39;Modality&#39;, &#39;VendorLot&#39;, &#39;DonorID&#39;, &#39;DonorAge&#39;, &#39;DonorBMI&#39;, &#39;DonorBloodType&#39;, &#39;DonorRace&#39;, &#39;Ethnicity&#39;, &#39;DonorGender&#39;, &#39;QCMeds&#39;, &#39;DonorSmoker&#39;, &#39;technology&#39;, &#39;cell_type_l2&#39;, &#39;cell_type_l1&#39;, &#39;cell_type_l3&#39;, &#39;assay&#39;, &#39;split&#39;
    var: &#39;feature_types&#39;, &#39;gene_id&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;
    uns: &#39;ATAC_gene_activity_var_names&#39;, &#39;dataset_id&#39;, &#39;genome&#39;, &#39;hvg&#39;, &#39;log1p&#39;, &#39;organism&#39;
    obsm: &#39;ATAC_gene_activity&#39;, &#39;ATAC_lsi_full&#39;, &#39;ATAC_lsi_red&#39;, &#39;ATAC_umap&#39;, &#39;GEX_X_pca&#39;, &#39;GEX_X_umap&#39;
    layers: &#39;binary&#39;, &#39;counts&#39;, &#39;cpm&#39;, &#39;tf-idf-binary&#39;, &#39;tf-idf-counts&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>rna_cite = sc.read(
    &quot;/lustre/groups/ml01/workspace/anastasia.litinetskaya/data/neurips-cite/rna_hvg.h5ad&quot;
)
rna_cite
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 90261 × 4000
    obs: &#39;GEX_n_genes_by_counts&#39;, &#39;GEX_pct_counts_mt&#39;, &#39;GEX_size_factors&#39;, &#39;GEX_phase&#39;, &#39;ADT_n_antibodies_by_counts&#39;, &#39;ADT_total_counts&#39;, &#39;ADT_iso_count&#39;, &#39;cell_type&#39;, &#39;batch&#39;, &#39;ADT_pseudotime_order&#39;, &#39;GEX_pseudotime_order&#39;, &#39;Samplename&#39;, &#39;Site&#39;, &#39;DonorNumber&#39;, &#39;Modality&#39;, &#39;VendorLot&#39;, &#39;DonorID&#39;, &#39;DonorAge&#39;, &#39;DonorBMI&#39;, &#39;DonorBloodType&#39;, &#39;DonorRace&#39;, &#39;Ethnicity&#39;, &#39;DonorGender&#39;, &#39;QCMeds&#39;, &#39;DonorSmoker&#39;, &#39;is_train&#39;
    var: &#39;feature_types&#39;, &#39;gene_id&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;
    uns: &#39;dataset_id&#39;, &#39;genome&#39;, &#39;hvg&#39;, &#39;organism&#39;
    obsm: &#39;ADT_X_pca&#39;, &#39;ADT_X_umap&#39;, &#39;ADT_isotype_controls&#39;, &#39;GEX_X_pca&#39;, &#39;GEX_X_umap&#39;
    layers: &#39;counts&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adt_cite = sc.read(&quot;/lustre/groups/ml01/workspace/daniel.strobl/neurips_cite_pp.h5ad&quot;)
adt_cite
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 120502 × 136
    obs: &#39;donor&#39;, &#39;batch&#39;, &#39;n_genes_by_counts&#39;, &#39;log1p_n_genes_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;n_counts&#39;, &#39;outliers&#39;
    var: &#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;n_cells_by_counts&#39;, &#39;mean_counts&#39;, &#39;log1p_mean_counts&#39;, &#39;pct_dropout_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;
    uns: &#39;batch_colors&#39;, &#39;donor_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;umap&#39;
    obsm: &#39;X_isotypes&#39;, &#39;X_pca&#39;, &#39;X_pcahm&#39;, &#39;X_umap&#39;
    varm: &#39;PCs&#39;
    layers: &#39;counts&#39;
    obsp: &#39;connectivities&#39;, &#39;distances&#39;
</pre></div>
</div>
</div>
</div>
<p>There are some differences in <code class="docutils literal notranslate"><span class="pre">.obs_names</span></code> of RNA and ADT of CITE-seq data, so we update them to make sure they allign between modalities.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>rna_cite.obs_names
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Index([&#39;GCATTAGCATAAGCGG-1-s1d1&#39;, &#39;TACAGGTGTTAGAGTA-1-s1d1&#39;,
       &#39;AGGATCTAGGTCTACT-1-s1d1&#39;, &#39;GTAGAAAGTGACACAG-1-s1d1&#39;,
       &#39;TCCGAAAAGGATCATA-1-s1d1&#39;, &#39;CTCCCAATCCATTGGA-1-s1d1&#39;,
       &#39;GACCAATCAATTTCGG-1-s1d1&#39;, &#39;TTCCGGTAGTTGTAAG-1-s1d1&#39;,
       &#39;ACCTGTCAGGACTGGT-1-s1d1&#39;, &#39;TTCGATTTCAGGACAG-1-s1d1&#39;,
       ...
       &#39;TCTTCCTAGCCAACCC-1-s4d9&#39;, &#39;TTCCACGGTTGAGGAC-1-s4d9&#39;,
       &#39;ATTCCTAGTCCAAGAG-1-s4d9&#39;, &#39;GCGGAAAGTACGCGTC-1-s4d9&#39;,
       &#39;TAACTTCAGATACAGT-1-s4d9&#39;, &#39;GAATCACCACGGAAGT-1-s4d9&#39;,
       &#39;GCTGGGTGTACGGATG-1-s4d9&#39;, &#39;TCGAAGTGTGACAGGT-1-s4d9&#39;,
       &#39;GCAGGCTGTTGCATAC-1-s4d9&#39;, &#39;ACGTAACAGGTCTACT-1-s4d9&#39;],
      dtype=&#39;object&#39;, length=90261)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adt_cite.obs_names
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Index([&#39;AAACCCAAGGATGGCT-1-0-0-0-0-0-0-0-0-0-0-0&#39;,
       &#39;AAACCCAAGGCCTAGA-1-0-0-0-0-0-0-0-0-0-0-0&#39;,
       &#39;AAACCCAAGTGAGTGC-1-0-0-0-0-0-0-0-0-0-0-0&#39;,
       &#39;AAACCCACAAGAGGCT-1-0-0-0-0-0-0-0-0-0-0-0&#39;,
       &#39;AAACCCACATCGTGGC-1-0-0-0-0-0-0-0-0-0-0-0&#39;,
       &#39;AAACCCACATTCTCTA-1-0-0-0-0-0-0-0-0-0-0-0&#39;,
       &#39;AAACCCAGTCCGCAGT-1-0-0-0-0-0-0-0-0-0-0-0&#39;,
       &#39;AAACCCAGTGCATACT-1-0-0-0-0-0-0-0-0-0-0-0&#39;,
       &#39;AAACCCAGTTGACGGA-1-0-0-0-0-0-0-0-0-0-0-0&#39;,
       &#39;AAACCCATCGATACTG-1-0-0-0-0-0-0-0-0-0-0-0&#39;,
       ...
       &#39;TTTGGTTGTCGTACTA-1-1&#39;, &#39;TTTGGTTTCACCTCGT-1-1&#39;, &#39;TTTGGTTTCGAGAACG-1-1&#39;,
       &#39;TTTGGTTTCGATACGT-1-1&#39;, &#39;TTTGTTGGTAGCTTTG-1-1&#39;, &#39;TTTGTTGGTCCAATCA-1-1&#39;,
       &#39;TTTGTTGGTGACTAAA-1-1&#39;, &#39;TTTGTTGTCCCTCTCC-1-1&#39;, &#39;TTTGTTGTCTAGAGCT-1-1&#39;,
       &#39;TTTGTTGTCTCTGCCA-1-1&#39;],
      dtype=&#39;object&#39;, length=120502)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adt_cite.obs_names = [
    name.split(&quot;-&quot;)[0] + &quot;-&quot; + name.split(&quot;-&quot;)[1] + &quot;-&quot; + batch
    for batch, name in zip(adt_cite.obs[&quot;donor&quot;], adt_cite.obs_names)
]
</pre></div>
</div>
</div>
</div>
<p>We subset to the cells that are present in both modalities.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>common_idx = list(set(rna_cite.obs_names).intersection(set(adt_cite.obs_names)))
rna_cite = rna_cite[common_idx].copy()
adt_cite = adt_cite[common_idx].copy()
</pre></div>
</div>
</div>
</div>
<p>We also copy some of the metadata from RNA to ADT.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adt_cite.obs = adt_cite.obs.join(rna_cite.obs[[&quot;Samplename&quot;, &quot;cell_type&quot;]])
</pre></div>
</div>
</div>
</div>
<p>Also we make sure that cells are the same in multiome data.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>assert np.sum(rna_multiome.obs_names != atac_multiome.obs_names) == 0
</pre></div>
</div>
</div>
</div>
<p>Because batch names are the same for different modalities and datasets, we will need to update a bunch of batch columns in <code class="docutils literal notranslate"><span class="pre">.obs</span></code> so we create a helper function for this.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>def update_obs_column(
    adata, obs_column_name, suffix, how=&quot;right&quot;, new_obs_column_name=None
):
    if new_obs_column_name is None:
        new_obs_column_name = f&quot;new_{obs_column_name}&quot;
    # otherwise can&#39;t use + operator with categorical columns
    adata.obs[obs_column_name] = adata.obs[obs_column_name].astype(&quot;str&quot;).copy()
    # create new column in .obs
    if how == &quot;right&quot;:
        adata.obs[new_obs_column_name] = f&quot;_{suffix}&quot;
        adata.obs[new_obs_column_name] = (
            adata.obs[obs_column_name] + adata.obs[new_obs_column_name]
        )
    else:
        adata.obs[new_obs_column_name] = f&quot;{suffix}_&quot;
        adata.obs[new_obs_column_name] = (
            adata.obs[new_obs_column_name] + adata.obs[obs_column_name]
        )
    return adata
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>rna_multiome = update_obs_column(rna_multiome, &quot;batch&quot;, &quot;_rna_multiome&quot;)
atac_multiome = update_obs_column(atac_multiome, &quot;batch&quot;, &quot;_atac_multiome&quot;)
rna_cite = update_obs_column(rna_cite, &quot;batch&quot;, &quot;_rna_cite&quot;)
adt_cite = update_obs_column(adt_cite, &quot;donor&quot;, &quot;_adt_cite&quot;, &quot;batch&quot;)
</pre></div>
</div>
</div>
</div>
<p>Finally we subset to the correct batches the reference datasets and the query.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># query
rna_multiome_query = rna_multiome[
    rna_multiome.obs[&quot;batch&quot;].isin(multiome_query_batches)
].copy()
atac_multiome_query = atac_multiome[
    atac_multiome.obs[&quot;batch&quot;].isin(multiome_query_batches)
].copy()
rna_cite_query = rna_cite[rna_cite.obs[&quot;batch&quot;].isin(cite_query_batches)].copy()
adt_cite_query = adt_cite[adt_cite.obs[&quot;donor&quot;].isin(cite_query_batches)].copy()
# reference
rna_multiome = rna_multiome[
    rna_multiome.obs[&quot;batch&quot;].isin(multiome_reference_batches)
].copy()
atac_multiome = atac_multiome[
    atac_multiome.obs[&quot;batch&quot;].isin(multiome_reference_batches)
].copy()
rna_cite = rna_cite[rna_cite.obs[&quot;batch&quot;].isin(cite_reference_batches)].copy()
adt_cite = adt_cite[adt_cite.obs[&quot;donor&quot;].isin(cite_reference_batches)].copy()
</pre></div>
</div>
</div>
</div>
<p><a id='glue'></a></p>
</section>
<section id="unpaired-integration-with-graph-linked-unified-embedding-glue">
<h2><span class="section-number">38.4. </span>Unpaired integration with Graph Linked Unified Embedding (GLUE)<a class="headerlink" href="#unpaired-integration-with-graph-linked-unified-embedding-glue" title="Permalink to this headline">#</a></h2>
<p>Opposed to paired integration that we demonstrated before, it is also possible to perform completely unpaired integration. In this case, there is no intersection of cell barcodes or features. Hence, we need some prior knowledge to connect different modalities.</p>
<p>GLUE <span id="id2">[<a class="reference internal" href="#id446" title="Zhi-Jie Cao and Ge Gao. Multi-omics single-cell data integration and regulatory inference with graph-linked embedding. Nature Biotechnology, May 2022. URL: https://doi.org/10.1038/s41587-022-01284-4, doi:10.1038/s41587-022-01284-4.">Cao and Gao, 2022</a>]</span> is a deep learning model for unpaired integration which makes use of a regulatory graph helping connect features from different modalities. The model is based on conditional variational autoencoders, where the model learns to reconstruct while simultaneously allowing for batch correction. To guide the integration, GLUE learns an embedding for each modality for each feature by utilizing a prior knowledge graph. We demonstrate how to use GLUE to integrate unpaired RNA and ADT data using RNA part of Multiome data and ADT part of CITE-seq data from the NeurIPS competiiton (<a class="reference external" href="https://openproblems.bio/neurips_2021/">https://openproblems.bio/neurips_2021/</a>). To construct the graph, we connect nodes from RNA modality to nodes from ADT modality if and only if the RNA node is a protein encoding gene of a given protein from ADT modality. The output of the GLUE model is a representation of each cell in a shared latent space.</p>
<p>We refer the reader to GLUE tutorial to see how one can integrate unpaired RNA and ATAC with GLUE and for more details about the model <a class="reference external" href="https://scglue.readthedocs.io/en/latest/tutorials.html">https://scglue.readthedocs.io/en/latest/tutorials.html</a>.</p>
<p>We follow the GLUE tutorial on how to preprocess RNA-seq. We log-normalize the raw counts, scale and calculate PCA.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>rna_multiome.X = rna_multiome.layers[&quot;counts&quot;].copy()
sc.pp.normalize_total(rna_multiome)
sc.pp.log1p(rna_multiome)
sc.pp.scale(rna_multiome)
sc.tl.pca(rna_multiome, n_comps=100, svd_solver=&quot;auto&quot;)
</pre></div>
</div>
</div>
</div>
<p>For ADT counts, we use CLR-normalized counts from <code class="docutils literal notranslate"><span class="pre">.X</span></code> and PCA calculated on the normalized values.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>np.max(adt_cite.X)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>69.27085
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.tl.pca(adt_cite, n_comps=100, svd_solver=&quot;auto&quot;)
</pre></div>
</div>
</div>
</div>
<p>We need to make sure that protein encoding gene names and the corresponding protein names align in both data objects. We will need this later for the graph construction.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>{
    &quot;CD103&quot;: &quot;ITGAE&quot;,
    &quot;CD11b&quot;: &quot;ITGAM&quot;,
    &quot;CD11c&quot;: &quot;ITGAX&quot;,
    &quot;CD122&quot;: &quot;IL2RB&quot;,
    &quot;CD124&quot;: &quot;IL4R&quot;,
    &quot;CD127&quot;: &quot;IL7R&quot;,
    &quot;CD13&quot;: &quot;ANPEP&quot;,
    &quot;CD137&quot;: &quot;TNFRSF9&quot;,
    &quot;CD152&quot;: &quot;CTLA4&quot;,
    &quot;CD154&quot;: &quot;CD40LG&quot;,
    &quot;CD16&quot;: &quot;FCGR3A&quot;,
    &quot;CD161&quot;: &quot;KLRB1&quot;,
    &quot;CD185&quot;: &quot;CXCR5&quot;,
    &quot;CD194&quot;: &quot;CCR4&quot;,
    &quot;CD196&quot;: &quot;CCR6&quot;,
    &quot;CD20&quot;: &quot;MS4A1&quot;,
    &quot;CD21&quot;: &quot;CR2&quot;,
    &quot;CD23&quot;: &quot;FCER2&quot;,
    &quot;CD25&quot;: &quot;IL2RA&quot;,
    &quot;CD26&quot;: &quot;DPP4&quot;,
    &quot;CD268&quot;: &quot;TNFRSF13C&quot;,
    &quot;CD272&quot;: &quot;BTLA&quot;,
    &quot;CD278&quot;: &quot;ICOS&quot;,
    &quot;CD29&quot;: &quot;ITGB1&quot;,
    &quot;CD3&quot;: &quot;CD3G&quot;,
    &quot;CD303&quot;: &quot;CLEC4C&quot;,
    &quot;CD304&quot;: &quot;NRP1&quot;,
    &quot;CD314&quot;: &quot;KLRK1&quot;,
    &quot;CD319&quot;: &quot;SLAMF7&quot;,
    &quot;CD335&quot;: &quot;NCR1&quot;,
    &quot;CD35&quot;: &quot;CR1&quot;,
    &quot;CD352&quot;: &quot;SLAMF6&quot;,
    &quot;CD39&quot;: &quot;ENTPD1&quot;,
    &quot;CD49a&quot;: &quot;ITGA1&quot;,
    &quot;CD49f&quot;: &quot;ITGA6&quot;,
    &quot;CD54&quot;: &quot;ICAM1&quot;,
    &quot;CD56&quot;: &quot;NCAM1&quot;,
    &quot;CD62L&quot;: &quot;SELL&quot;,
    &quot;CD71&quot;: &quot;TFRC&quot;,
    &quot;CD73&quot;: &quot;NT5E&quot;,
    &quot;CD79b&quot;: &quot;CD79B&quot;,
    &quot;CD8&quot;: &quot;CD8A&quot;,
    &quot;CD85j&quot;: &quot;LILRB1&quot;,
    &quot;CD88&quot;: &quot;C5AR1&quot;,
    &quot;CD94&quot;: &quot;KLRD1&quot;,
    &quot;CD95&quot;: &quot;FAS&quot;,
    &quot;HLA-DR&quot;: &quot;HLA-DRA&quot;,
    &quot;IgD&quot;: &quot;IGHD&quot;,
    &quot;IgM&quot;: &quot;IGHM&quot;,
    &quot;TCR&quot;: &quot;TRAC&quot;,
}
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adt_cite.var_names = adt_cite.var_names.str.strip(&quot;-1&quot;)
adt_cite.var_names = [
    rename_proteins[name] if name in rename_proteins else name
    for name in adt_cite.var_names
]
</pre></div>
</div>
</div>
</div>
<section id="graph-construction">
<h3><span class="section-number">38.4.1. </span>Graph construction<a class="headerlink" href="#graph-construction" title="Permalink to this headline">#</a></h3>
<p>Before we can run GLUE integration, we need to construct a graph with prior knowledge on how feature from the two modalities are connected. The node set has to be the union of the features from all the modalities, and the edge weights have to between 0 and 1. To integrate ADT with RNA, we construct the graph the following way: we set the edge weight to 1 if the protein name is the same as the gene name and 0 for all the other edges.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>p = np.array(adt_cite.var_names)
r = np.array(rna_multiome.var_names)
# mask entries are set to 1 where protein name is the same as gene name
mask = np.repeat(p.reshape(-1, 1), r.shape[0], axis=1) == r
mask = np.array(mask)
</pre></div>
</div>
</div>
</div>
<p>We rename the features so they differ for ADT and RNA modalities.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>rna_vars = [v + &quot;_rna&quot; for v in rna_multiome.var_names]
prot_vars = [v + &quot;_prot&quot; for v in adt_cite.var_names]
rna_multiome.var_names = rna_vars
adt_cite.var_names = prot_vars
</pre></div>
</div>
</div>
</div>
<p>GLUE also requires each node to have an self-loop with weight 1, so we add these here too.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adj = pd.DataFrame(mask, index=prot_vars, columns=rna_vars)
diag_edges = adj[adj &gt; 0].stack().index.tolist()
diag_edges = [(n1, n2, {&quot;weight&quot;: 1.0, &quot;sign&quot;: 1}) for n1, n2 in diag_edges]
self_loop_rna = [(g, g, {&quot;weight&quot;: 1.0, &quot;sign&quot;: 1}) for g in rna_vars]
self_loop_prot = [(g, g, {&quot;weight&quot;: 1.0, &quot;sign&quot;: 1}) for g in prot_vars]
</pre></div>
</div>
</div>
</div>
<p>Next we construct the actual graph object.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>graph = nx.Graph()
graph.add_nodes_from(rna_vars)
graph.add_nodes_from(prot_vars)
graph.add_edges_from(diag_edges)
graph.add_edges_from(self_loop_prot)
graph.add_edges_from(self_loop_rna)
</pre></div>
</div>
</div>
</div>
<p>We should have 4136 nodes which corresponds to 4000 genes and 136 proteins. We also have 4186 edges with non-zero weights: 4136 self-loops and 50 connections between genes and proteins with the same name.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>graph.number_of_nodes(), graph.number_of_edges()
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(4136, 4186)
</pre></div>
</div>
</div>
</div>
</section>
<section id="configure-data">
<h3><span class="section-number">38.4.2. </span>Configure data<a class="headerlink" href="#configure-data" title="Permalink to this headline">#</a></h3>
<p>In this section we again follow the GLUE tutorial to configure the model. First, we set up the RNA encoder-decoder pair to learn to reconstruct raw counts that we assume follow NB distribution. We specify PCA embedding to be used with in the encoder.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>scglue.models.configure_dataset(
    rna_multiome,
    &quot;NB&quot;,
    use_highly_variable=False,
    use_batch=&quot;Samplename&quot;,
    use_layer=&quot;counts&quot;,
    use_rep=&quot;X_pca&quot;,
)
</pre></div>
</div>
</div>
</div>
<p>Next, we set up the ADT encoder and decoder to reconstruct normalized counts following the normal distribution and  to use PCA embeddings.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>scglue.models.configure_dataset(
    adt_cite,
    &quot;Normal&quot;,
    use_highly_variable=False,
    use_batch=&quot;Samplename&quot;,
    use_rep=&quot;X_pca&quot;,
)
</pre></div>
</div>
</div>
</div>
<p>We initialize and train the final model.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>glue = scglue.models.fit_SCGLUE(
    {&quot;rna&quot;: rna_multiome, &quot;adt&quot;: adt_cite},
    graph,
)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[INFO] fit_SCGLUE: Pretraining SCGLUE model...
[INFO] autodevice: Using GPU 0 as computation device.
[INFO] check_graph: Checking variable coverage...
[INFO] check_graph: Checking edge attributes...
[INFO] check_graph: Checking self-loops...
[INFO] check_graph: Checking graph symmetry...
[INFO] check_graph: All checks passed!
[INFO] SCGLUEModel: Setting `graph_batch_size` = 1457
[INFO] SCGLUEModel: Setting `max_epochs` = 102
[INFO] SCGLUEModel: Setting `patience` = 9
[INFO] SCGLUEModel: Setting `reduce_lr_patience` = 5
[INFO] SCGLUETrainer: Using training directory: &quot;/tmp/GLUETMPzimh2zsp&quot;
[INFO] SCGLUETrainer: [Epoch 10] train={&#39;g_nll&#39;: 0.192, &#39;g_kl&#39;: 0.04, &#39;g_elbo&#39;: 0.233, &#39;x_rna_nll&#39;: 0.241, &#39;x_rna_kl&#39;: 0.006, &#39;x_rna_elbo&#39;: 0.248, &#39;x_adt_nll&#39;: 1.928, &#39;x_adt_kl&#39;: 0.208, &#39;x_adt_elbo&#39;: 2.137, &#39;dsc_loss&#39;: 0.666, &#39;vae_loss&#39;: 2.394, &#39;gen_loss&#39;: 2.36}, val={&#39;g_nll&#39;: 0.188, &#39;g_kl&#39;: 0.04, &#39;g_elbo&#39;: 0.228, &#39;x_rna_nll&#39;: 0.243, &#39;x_rna_kl&#39;: 0.006, &#39;x_rna_elbo&#39;: 0.249, &#39;x_adt_nll&#39;: 1.874, &#39;x_adt_kl&#39;: 0.199, &#39;x_adt_elbo&#39;: 2.072, &#39;dsc_loss&#39;: 0.664, &#39;vae_loss&#39;: 2.331, &#39;gen_loss&#39;: 2.298}, 4.3s elapsed
[INFO] SCGLUETrainer: [Epoch 20] train={&#39;g_nll&#39;: 0.196, &#39;g_kl&#39;: 0.041, &#39;g_elbo&#39;: 0.238, &#39;x_rna_nll&#39;: 0.237, &#39;x_rna_kl&#39;: 0.006, &#39;x_rna_elbo&#39;: 0.243, &#39;x_adt_nll&#39;: 1.845, &#39;x_adt_kl&#39;: 0.174, &#39;x_adt_elbo&#39;: 2.019, &#39;dsc_loss&#39;: 0.678, &#39;vae_loss&#39;: 2.271, &#39;gen_loss&#39;: 2.237}, val={&#39;g_nll&#39;: 0.196, &#39;g_kl&#39;: 0.041, &#39;g_elbo&#39;: 0.238, &#39;x_rna_nll&#39;: 0.237, &#39;x_rna_kl&#39;: 0.006, &#39;x_rna_elbo&#39;: 0.242, &#39;x_adt_nll&#39;: 1.808, &#39;x_adt_kl&#39;: 0.172, &#39;x_adt_elbo&#39;: 1.98, &#39;dsc_loss&#39;: 0.681, &#39;vae_loss&#39;: 2.232, &#39;gen_loss&#39;: 2.198}, 4.5s elapsed
[INFO] SCGLUETrainer: [Epoch 30] train={&#39;g_nll&#39;: 0.185, &#39;g_kl&#39;: 0.042, &#39;g_elbo&#39;: 0.227, &#39;x_rna_nll&#39;: 0.236, &#39;x_rna_kl&#39;: 0.006, &#39;x_rna_elbo&#39;: 0.242, &#39;x_adt_nll&#39;: 1.834, &#39;x_adt_kl&#39;: 0.17, &#39;x_adt_elbo&#39;: 2.005, &#39;dsc_loss&#39;: 0.68, &#39;vae_loss&#39;: 2.255, &#39;gen_loss&#39;: 2.221}, val={&#39;g_nll&#39;: 0.157, &#39;g_kl&#39;: 0.042, &#39;g_elbo&#39;: 0.199, &#39;x_rna_nll&#39;: 0.237, &#39;x_rna_kl&#39;: 0.005, &#39;x_rna_elbo&#39;: 0.242, &#39;x_adt_nll&#39;: 1.799, &#39;x_adt_kl&#39;: 0.167, &#39;x_adt_elbo&#39;: 1.965, &#39;dsc_loss&#39;: 0.674, &#39;vae_loss&#39;: 2.216, &#39;gen_loss&#39;: 2.182}, 4.4s elapsed
Epoch 00031: reducing learning rate of group 0 to 2.0000e-04.
Epoch 00031: reducing learning rate of group 0 to 2.0000e-04.
[INFO] LRScheduler: Learning rate reduction: step 1
[INFO] SCGLUETrainer: [Epoch 40] train={&#39;g_nll&#39;: 0.159, &#39;g_kl&#39;: 0.042, &#39;g_elbo&#39;: 0.201, &#39;x_rna_nll&#39;: 0.236, &#39;x_rna_kl&#39;: 0.006, &#39;x_rna_elbo&#39;: 0.241, &#39;x_adt_nll&#39;: 1.82, &#39;x_adt_kl&#39;: 0.171, &#39;x_adt_elbo&#39;: 1.99, &#39;dsc_loss&#39;: 0.678, &#39;vae_loss&#39;: 2.239, &#39;gen_loss&#39;: 2.205}, val={&#39;g_nll&#39;: 0.17, &#39;g_kl&#39;: 0.042, &#39;g_elbo&#39;: 0.212, &#39;x_rna_nll&#39;: 0.238, &#39;x_rna_kl&#39;: 0.005, &#39;x_rna_elbo&#39;: 0.244, &#39;x_adt_nll&#39;: 1.788, &#39;x_adt_kl&#39;: 0.171, &#39;x_adt_elbo&#39;: 1.959, &#39;dsc_loss&#39;: 0.67, &#39;vae_loss&#39;: 2.211, &#39;gen_loss&#39;: 2.177}, 4.3s elapsed
Epoch 00044: reducing learning rate of group 0 to 2.0000e-05.
Epoch 00044: reducing learning rate of group 0 to 2.0000e-05.
[INFO] LRScheduler: Learning rate reduction: step 2
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2022-10-14 13:52:55,001 ignite.handlers.early_stopping.EarlyStopping INFO: EarlyStopping: Stop training
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[INFO] EarlyStopping: Restoring checkpoint &quot;41&quot;...
[INFO] EarlyStopping: Restoring checkpoint &quot;41&quot;...
[INFO] fit_SCGLUE: Estimating balancing weight...
[INFO] estimate_balancing_weight: Clustering cells...
[INFO] estimate_balancing_weight: Matching clusters...
[INFO] estimate_balancing_weight: Matching array shape = (22, 22)...
[INFO] estimate_balancing_weight: Estimating balancing weight...
[INFO] fit_SCGLUE: Fine-tuning SCGLUE model...
[INFO] check_graph: Checking variable coverage...
[INFO] check_graph: Checking edge attributes...
[INFO] check_graph: Checking self-loops...
[INFO] check_graph: Checking graph symmetry...
[INFO] check_graph: All checks passed!
[INFO] SCGLUEModel: Setting `graph_batch_size` = 1457
[INFO] SCGLUEModel: Setting `align_burnin` = 17
[INFO] SCGLUEModel: Setting `max_epochs` = 102
[INFO] SCGLUEModel: Setting `patience` = 9
[INFO] SCGLUEModel: Setting `reduce_lr_patience` = 5
[INFO] SCGLUETrainer: Using training directory: &quot;/tmp/GLUETMPqkn8cw51&quot;
[INFO] SCGLUETrainer: [Epoch 10] train={&#39;g_nll&#39;: 0.153, &#39;g_kl&#39;: 0.042, &#39;g_elbo&#39;: 0.195, &#39;x_rna_nll&#39;: 0.237, &#39;x_rna_kl&#39;: 0.006, &#39;x_rna_elbo&#39;: 0.243, &#39;x_adt_nll&#39;: 1.834, &#39;x_adt_kl&#39;: 0.169, &#39;x_adt_elbo&#39;: 2.003, &#39;dsc_loss&#39;: 0.663, &#39;vae_loss&#39;: 2.254, &#39;gen_loss&#39;: 2.22}, val={&#39;g_nll&#39;: 0.145, &#39;g_kl&#39;: 0.042, &#39;g_elbo&#39;: 0.187, &#39;x_rna_nll&#39;: 0.238, &#39;x_rna_kl&#39;: 0.006, &#39;x_rna_elbo&#39;: 0.244, &#39;x_adt_nll&#39;: 1.783, &#39;x_adt_kl&#39;: 0.169, &#39;x_adt_elbo&#39;: 1.952, &#39;dsc_loss&#39;: 0.688, &#39;vae_loss&#39;: 2.203, &#39;gen_loss&#39;: 2.169}, 4.5s elapsed
[INFO] SCGLUETrainer: [Epoch 20] train={&#39;g_nll&#39;: 0.136, &#39;g_kl&#39;: 0.042, &#39;g_elbo&#39;: 0.178, &#39;x_rna_nll&#39;: 0.239, &#39;x_rna_kl&#39;: 0.006, &#39;x_rna_elbo&#39;: 0.244, &#39;x_adt_nll&#39;: 1.828, &#39;x_adt_kl&#39;: 0.168, &#39;x_adt_elbo&#39;: 1.996, &#39;dsc_loss&#39;: 0.634, &#39;vae_loss&#39;: 2.247, &#39;gen_loss&#39;: 2.216}, val={&#39;g_nll&#39;: 0.147, &#39;g_kl&#39;: 0.042, &#39;g_elbo&#39;: 0.189, &#39;x_rna_nll&#39;: 0.239, &#39;x_rna_kl&#39;: 0.005, &#39;x_rna_elbo&#39;: 0.244, &#39;x_adt_nll&#39;: 1.784, &#39;x_adt_kl&#39;: 0.17, &#39;x_adt_elbo&#39;: 1.954, &#39;dsc_loss&#39;: 0.661, &#39;vae_loss&#39;: 2.206, &#39;gen_loss&#39;: 2.173}, 4.4s elapsed
Epoch 00025: reducing learning rate of group 0 to 2.0000e-04.
Epoch 00025: reducing learning rate of group 0 to 2.0000e-04.
[INFO] LRScheduler: Learning rate reduction: step 1
[INFO] SCGLUETrainer: [Epoch 30] train={&#39;g_nll&#39;: 0.138, &#39;g_kl&#39;: 0.042, &#39;g_elbo&#39;: 0.18, &#39;x_rna_nll&#39;: 0.238, &#39;x_rna_kl&#39;: 0.005, &#39;x_rna_elbo&#39;: 0.243, &#39;x_adt_nll&#39;: 1.82, &#39;x_adt_kl&#39;: 0.168, &#39;x_adt_elbo&#39;: 1.988, &#39;dsc_loss&#39;: 0.66, &#39;vae_loss&#39;: 2.238, &#39;gen_loss&#39;: 2.205}, val={&#39;g_nll&#39;: 0.139, &#39;g_kl&#39;: 0.042, &#39;g_elbo&#39;: 0.181, &#39;x_rna_nll&#39;: 0.239, &#39;x_rna_kl&#39;: 0.005, &#39;x_rna_elbo&#39;: 0.245, &#39;x_adt_nll&#39;: 1.773, &#39;x_adt_kl&#39;: 0.168, &#39;x_adt_elbo&#39;: 1.941, &#39;dsc_loss&#39;: 0.662, &#39;vae_loss&#39;: 2.193, &#39;gen_loss&#39;: 2.16}, 4.4s elapsed
Epoch 00033: reducing learning rate of group 0 to 2.0000e-05.
Epoch 00033: reducing learning rate of group 0 to 2.0000e-05.
[INFO] LRScheduler: Learning rate reduction: step 2
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2022-10-14 13:56:17,639 ignite.handlers.early_stopping.EarlyStopping INFO: EarlyStopping: Stop training
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[INFO] EarlyStopping: Restoring checkpoint &quot;35&quot;...
[INFO] EarlyStopping: Restoring checkpoint &quot;35&quot;...
</pre></div>
</div>
</div>
</div>
<p>Now we can obtain the latent representation for both modalities and concatenate them into one AnnData object for later visualization.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>rna_multiome.obsm[&quot;X_glue&quot;] = glue.encode_data(&quot;rna&quot;, rna_multiome)
adt_cite.obsm[&quot;X_glue&quot;] = glue.encode_data(&quot;adt&quot;, adt_cite)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adt_cite.obs[&quot;modality&quot;] = &quot;ADT&quot;
rna_multiome.obs[&quot;modality&quot;] = &quot;RNA&quot;

combined = anndata.concat([rna_multiome, adt_cite])
</pre></div>
</div>
</div>
</div>
<p>Finally, we visualize the integrated latent space.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pp.neighbors(combined, use_rep=&quot;X_glue&quot;, metric=&quot;cosine&quot;)
sc.tl.umap(combined)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pl.umap(combined, color=[&quot;cell_type&quot;, &quot;modality&quot;, &quot;Samplename&quot;], ncols=1)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/advanced_integration_57_0.png" src="../_images/advanced_integration_57_0.png" />
</div>
</div>
<p>We see that the batches were well integrated as well as the two modalities. Yet again, scIB could be used to evaluate the integration.</p>
<p><a id='totalvi'></a></p>
</section>
</section>
<section id="integration-with-partially-overlapping-data-and-query-to-reference-mapping-with-totalvi">
<h2><span class="section-number">38.5. </span>Integration with partially overlapping data and query-to-reference mapping with totalVI<a class="headerlink" href="#integration-with-partially-overlapping-data-and-query-to-reference-mapping-with-totalvi" title="Permalink to this headline">#</a></h2>
<p>Another multimodal data integration scenario is when we have paired and unpaired data available at the same time, for instance, a CITE-seq and an RNA-seq datasets. This scenario is also known as mosaic integration. We might be interested in mapping the datasets into a shared latent space to make use of all the data available to us. Another use case would be to impute missing modalities (i.e. proteins abundance for RNA-seq dataset). We show how to perform these two tasks, partially overlapping integration and imputation, using totalVI <span id="id3">[<a class="reference internal" href="paired_integration.html#id439" title="Adam Gayoso, Zoë Steier, Romain Lopez, Jeffrey Regier, Kristopher L. Nazor, Aaron Streets, and Nir Yosef. Joint probabilistic modeling of single-cell multi-omic data with totalvi. Nature Methods, 18(3):272-282, Mar 2021. URL: https://doi.org/10.1038/s41592-020-01050-x, doi:10.1038/s41592-020-01050-x.">Gayoso <em>et al.</em>, 2021</a>]</span>.</p>
<p>We set the protein counts of one of the batches to zeros to model the missing proteins. This helps later with mapping of the new RNA-only queries.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata = rna_cite.copy()
adata.obsm[&quot;protein_counts&quot;] = adt_cite.layers[&quot;counts&quot;].A.copy()
adata.obsm[&quot;protein_counts&quot;][adata.obs[&quot;batch&quot;] == &quot;s1d3&quot;] = 0.0
adata
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 16294 × 4000
    obs: &#39;GEX_n_genes_by_counts&#39;, &#39;GEX_pct_counts_mt&#39;, &#39;GEX_size_factors&#39;, &#39;GEX_phase&#39;, &#39;ADT_n_antibodies_by_counts&#39;, &#39;ADT_total_counts&#39;, &#39;ADT_iso_count&#39;, &#39;cell_type&#39;, &#39;batch&#39;, &#39;ADT_pseudotime_order&#39;, &#39;GEX_pseudotime_order&#39;, &#39;Samplename&#39;, &#39;Site&#39;, &#39;DonorNumber&#39;, &#39;Modality&#39;, &#39;VendorLot&#39;, &#39;DonorID&#39;, &#39;DonorAge&#39;, &#39;DonorBMI&#39;, &#39;DonorBloodType&#39;, &#39;DonorRace&#39;, &#39;Ethnicity&#39;, &#39;DonorGender&#39;, &#39;QCMeds&#39;, &#39;DonorSmoker&#39;, &#39;is_train&#39;, &#39;new_batch&#39;
    var: &#39;feature_types&#39;, &#39;gene_id&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;
    uns: &#39;dataset_id&#39;, &#39;genome&#39;, &#39;hvg&#39;, &#39;organism&#39;
    obsm: &#39;ADT_X_pca&#39;, &#39;ADT_X_umap&#39;, &#39;ADT_isotype_controls&#39;, &#39;GEX_X_pca&#39;, &#39;GEX_X_umap&#39;, &#39;protein_counts&#39;
    layers: &#39;counts&#39;
</pre></div>
</div>
</div>
</div>
<p>Next, we setup the AnnData, specify the model parameters and train the model.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>TOTALVI.setup_anndata(
    adata,
    layer=&quot;counts&quot;,
    batch_key=&quot;batch&quot;,
    protein_expression_obsm_key=&quot;protein_counts&quot;,
)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span><span class=" -Color -Color-Blue">INFO    </span> Generating sequential column names                                                  
<span class=" -Color -Color-Blue">INFO    </span> Found batches with missing protein expression                                       
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>arches_params = dict(
    use_layer_norm=&quot;both&quot;,
    use_batch_norm=&quot;none&quot;,
    n_layers_decoder=2,
    n_layers_encoder=2,
)

vae = TOTALVI(adata, **arches_params)
vae.train()
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span><span class=" -Color -Color-Blue">INFO    </span> Computing empirical prior initialization for protein background.                    
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>GPU available: True, used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 214/400:  54%|█████▎    | 214/400 [06:33&lt;06:35,  2.12s/it, loss=714, v_num=1]Epoch 00214: reducing learning rate of group 0 to 2.4000e-03.
Epoch 336/400:  84%|████████▍ | 336/400 [10:28&lt;01:58,  1.86s/it, loss=707, v_num=1]Epoch 00336: reducing learning rate of group 0 to 1.4400e-03.
Epoch 372/400:  93%|█████████▎| 372/400 [11:41&lt;00:54,  1.93s/it, loss=693, v_num=1]Epoch 00372: reducing learning rate of group 0 to 8.6400e-04.
Epoch 386/400:  96%|█████████▋| 386/400 [12:10&lt;00:26,  1.89s/it, loss=697, v_num=1]
Monitored metric elbo_validation did not improve in the last 45 records. Best score: 1450.084. Signaling Trainer to stop.
</pre></div>
</div>
</div>
</div>
<p>Now we obtain the latent representation and visualize it.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata.obsm[&quot;X_totalvi&quot;] = vae.get_latent_representation()
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pp.neighbors(adata, use_rep=&quot;X_totalvi&quot;)
sc.tl.umap(adata)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pl.umap(adata, color=[&quot;cell_type&quot;, &quot;batch&quot;], ncols=1)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/advanced_integration_67_0.png" src="../_images/advanced_integration_67_0.png" />
</div>
</div>
<p>We see that batches as well as paired/RNA-only data are well integrated.</p>
<section id="query-to-reference-mapping">
<h3><span class="section-number">38.5.1. </span>Query-to-reference mapping<a class="headerlink" href="#query-to-reference-mapping" title="Permalink to this headline">#</a></h3>
<p>Now we demontrate how to map new unimodal (RNA-only) and multimodal query (CITE-seq) onto the above reference.</p>
<p>We mimic RNA-only query by setting the protein counts of one of the two batches to zero.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>query = rna_cite_query.copy()
query.obsm[&quot;protein_counts&quot;] = adt_cite_query.layers[&quot;counts&quot;].A.copy()
query.obsm[&quot;protein_counts&quot;][query.obs[&quot;batch&quot;] == &quot;s2d4&quot;] = 0.0
</pre></div>
</div>
</div>
</div>
<p>We create some additional columns in <code class="docutils literal notranslate"><span class="pre">.obs</span></code> to help with the visualization later.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata.obs[&quot;dataset_name&quot;] = &quot;Reference&quot;
query.obs[&quot;dataset_name&quot;] = &quot;Query&quot;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata.obs[&quot;dataset_name_fine&quot;] = &quot;CITE reference&quot;
adata.obs[&quot;dataset_name_fine&quot;][adata.obs[&quot;batch&quot;] == &quot;s1d3&quot;] = &quot;RNA reference&quot;
query.obs[&quot;dataset_name_fine&quot;] = &quot;CITE query&quot;
query.obs[&quot;dataset_name_fine&quot;][query.obs[&quot;batch&quot;] == &quot;s2d4&quot;] = &quot;RNA query&quot;
</pre></div>
</div>
</div>
</div>
<p>Next, we update and fine-tune the model.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>vae_q = TOTALVI.load_query_data(
    query,
    vae,
)
vae_q.train(
    plan_kwargs=dict(weight_decay=0.0, scale_adversarial_loss=0.0),
)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span><span class=" -Color -Color-Blue">INFO    </span> Found batches with missing protein expression                                       
<span class=" -Color -Color-Blue">INFO    </span> Computing empirical prior initialization for protein background.                    
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>GPU available: True, used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 168/400:  42%|████▏     | 168/400 [05:13&lt;06:54,  1.79s/it, loss=557, v_num=1]Epoch 00168: reducing learning rate of group 0 to 2.4000e-03.
Epoch 182/400:  46%|████▌     | 182/400 [05:38&lt;06:45,  1.86s/it, loss=557, v_num=1]
Monitored metric elbo_validation did not improve in the last 45 records. Best score: 1074.102. Signaling Trainer to stop.
</pre></div>
</div>
</div>
</div>
<p>Finally, we obtain the latent representation for the query and visualize it together with the reference. For this we concatenate reference and the query and recalculate the neighors and the UMAP.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>query.obsm[&quot;X_totalvi_scarches&quot;] = vae_q.get_latent_representation(query)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata.obsm[&quot;X_totalvi_scarches&quot;] = adata.obsm[&quot;X_totalvi&quot;]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>full_data = adata.concatenate(query, batch_key=&quot;concat_batch&quot;)
full_data
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 32340 × 4000
    obs: &#39;GEX_n_genes_by_counts&#39;, &#39;GEX_pct_counts_mt&#39;, &#39;GEX_size_factors&#39;, &#39;GEX_phase&#39;, &#39;ADT_n_antibodies_by_counts&#39;, &#39;ADT_total_counts&#39;, &#39;ADT_iso_count&#39;, &#39;cell_type&#39;, &#39;batch&#39;, &#39;ADT_pseudotime_order&#39;, &#39;GEX_pseudotime_order&#39;, &#39;Samplename&#39;, &#39;Site&#39;, &#39;DonorNumber&#39;, &#39;Modality&#39;, &#39;VendorLot&#39;, &#39;DonorID&#39;, &#39;DonorAge&#39;, &#39;DonorBMI&#39;, &#39;DonorBloodType&#39;, &#39;DonorRace&#39;, &#39;Ethnicity&#39;, &#39;DonorGender&#39;, &#39;QCMeds&#39;, &#39;DonorSmoker&#39;, &#39;is_train&#39;, &#39;new_batch&#39;, &#39;_scvi_labels&#39;, &#39;_scvi_batch&#39;, &#39;dataset_name&#39;, &#39;dataset_name_fine&#39;, &#39;concat_batch&#39;
    var: &#39;feature_types&#39;, &#39;gene_id&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;
    obsm: &#39;ADT_X_pca&#39;, &#39;ADT_X_umap&#39;, &#39;ADT_isotype_controls&#39;, &#39;GEX_X_pca&#39;, &#39;GEX_X_umap&#39;, &#39;protein_counts&#39;, &#39;X_totalvi_scarches&#39;
    layers: &#39;counts&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pp.neighbors(full_data, use_rep=&quot;X_totalvi_scarches&quot;)
sc.tl.umap(full_data)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pl.umap(
    full_data,
    color=[&quot;cell_type&quot;, &quot;batch&quot;, &quot;dataset_name&quot;, &quot;dataset_name_fine&quot;],
    ncols=1,
)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/advanced_integration_81_0.png" src="../_images/advanced_integration_81_0.png" />
</div>
</div>
<p><a id='seurat'></a></p>
</section>
</section>
<section id="query-mapping-and-imputation-with-seurat-s-wnn">
<h2><span class="section-number">38.6. </span>Query mapping and imputation with Seurat’s WNN<a class="headerlink" href="#query-mapping-and-imputation-with-seurat-s-wnn" title="Permalink to this headline">#</a></h2>
<p>Seurat v4 allows mapping of new RNA-seq queries onto a multimodal reference integrated with weighted-nearest-neighbor (WNN) <span id="id4">[<a class="reference internal" href="paired_integration.html#id437" title="Yuhan Hao, Stephanie Hao, Erica Andersen-Nissen, William M. Mauck, Shiwei Zheng, Andrew Butler, Maddie J. Lee, Aaron J. Wilk, Charlotte Darby, Michael Zager, Paul Hoffman, Marlon Stoeckius, Efthymia Papalexi, Eleni P. Mimitou, Jaison Jain, Avi Srivastava, Tim Stuart, Lamar M. Fleming, Bertrand Yeung, Angela J. Rogers, Juliana M. McElrath, Catherine A. Blish, Raphael Gottardo, Peter Smibert, and Rahul Satija. Integrated analysis of multimodal single-cell data. Cell, 184(13):3573-3587.e29, 2021. URL: https://www.sciencedirect.com/science/article/pii/S0092867421005833, doi:https://doi.org/10.1016/j.cell.2021.04.048.">Hao <em>et al.</em>, 2021</a>]</span>. We use the reference we built in the paired integration notebook.</p>
<p>The mapping is based on finding anchors between the query and the reference. This approach also allows for missing information transfer from the reference to the query. Here we show how to predict cell types and how to impute missing modalities (protein abundance in this case).</p>
<p>We note that since Seurat v4 in an R library we will need to work with <code class="docutils literal notranslate"><span class="pre">anndata2ri</span></code> package (<a class="reference external" href="https://github.com/theislab/anndata2ri">https://github.com/theislab/anndata2ri</a>) to move all of our data from Python to R.</p>
<p>First, let’s take a look at our reference and the query.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>%%R
ref &lt;- readRDS(cite, file = &quot;wnn_ref.rds&quot;)
ref
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>An object of class Seurat 
4136 features across 16294 samples within 2 assays 
Active assay: RNA (4000 features, 4000 variable features)
 1 other assay present: ADT
 4 dimensional reductions calculated: pca, harmony_pca, spca, wnn.umap
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>rna_cite_query
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>View of AnnData object with n_obs × n_vars = 16046 × 4000
    obs: &#39;GEX_n_genes_by_counts&#39;, &#39;GEX_pct_counts_mt&#39;, &#39;GEX_size_factors&#39;, &#39;GEX_phase&#39;, &#39;ADT_n_antibodies_by_counts&#39;, &#39;ADT_total_counts&#39;, &#39;ADT_iso_count&#39;, &#39;cell_type&#39;, &#39;batch&#39;, &#39;ADT_pseudotime_order&#39;, &#39;GEX_pseudotime_order&#39;, &#39;Samplename&#39;, &#39;Site&#39;, &#39;DonorNumber&#39;, &#39;Modality&#39;, &#39;VendorLot&#39;, &#39;DonorID&#39;, &#39;DonorAge&#39;, &#39;DonorBMI&#39;, &#39;DonorBloodType&#39;, &#39;DonorRace&#39;, &#39;Ethnicity&#39;, &#39;DonorGender&#39;, &#39;QCMeds&#39;, &#39;DonorSmoker&#39;, &#39;is_train&#39;, &#39;new_batch&#39;
    var: &#39;feature_types&#39;, &#39;gene_id&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;
    uns: &#39;dataset_id&#39;, &#39;genome&#39;, &#39;hvg&#39;, &#39;organism&#39;
    obsm: &#39;ADT_X_pca&#39;, &#39;ADT_X_umap&#39;, &#39;ADT_isotype_controls&#39;, &#39;GEX_X_pca&#39;, &#39;GEX_X_umap&#39;
    layers: &#39;counts&#39;
</pre></div>
</div>
</div>
</div>
<p>We need to preprocess the query first, so we use CLR-normalized counts to calculate PCA.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pp.pca(rna_cite_query)
rna_cite_query
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 16046 × 4000
    obs: &#39;GEX_n_genes_by_counts&#39;, &#39;GEX_pct_counts_mt&#39;, &#39;GEX_size_factors&#39;, &#39;GEX_phase&#39;, &#39;ADT_n_antibodies_by_counts&#39;, &#39;ADT_total_counts&#39;, &#39;ADT_iso_count&#39;, &#39;cell_type&#39;, &#39;batch&#39;, &#39;ADT_pseudotime_order&#39;, &#39;GEX_pseudotime_order&#39;, &#39;Samplename&#39;, &#39;Site&#39;, &#39;DonorNumber&#39;, &#39;Modality&#39;, &#39;VendorLot&#39;, &#39;DonorID&#39;, &#39;DonorAge&#39;, &#39;DonorBMI&#39;, &#39;DonorBloodType&#39;, &#39;DonorRace&#39;, &#39;Ethnicity&#39;, &#39;DonorGender&#39;, &#39;QCMeds&#39;, &#39;DonorSmoker&#39;, &#39;is_train&#39;, &#39;new_batch&#39;
    var: &#39;feature_types&#39;, &#39;gene_id&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;
    uns: &#39;dataset_id&#39;, &#39;genome&#39;, &#39;hvg&#39;, &#39;organism&#39;, &#39;pca&#39;
    obsm: &#39;ADT_X_pca&#39;, &#39;ADT_X_umap&#39;, &#39;ADT_isotype_controls&#39;, &#39;GEX_X_pca&#39;, &#39;GEX_X_umap&#39;, &#39;X_pca&#39;
    varm: &#39;PCs&#39;
    layers: &#39;counts&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>np.max(rna_cite_query.X)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>9.061238
</pre></div>
</div>
</div>
</div>
<p>Next, we move the query from Python to R.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata_ = sc.AnnData(rna_cite_query.X.copy())
adata_.obs_names = rna_cite_query.obs_names.copy()
adata_.var_names = rna_cite_query.var_names.copy()
adata_.obs[&quot;cell_type&quot;] = rna_cite_query.obs[&quot;cell_type&quot;].copy()
adata_.obs[&quot;batch&quot;] = rna_cite_query.obs[&quot;batch&quot;].copy()
adata_.obsm[&quot;X_pca&quot;] = rna_cite_query.obsm[&quot;X_pca&quot;].copy()
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>%%R -i adata_
query = as.Seurat(adata_, data=&#39;X&#39;, counts=NULL)
query
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>An object of class Seurat 
4000 features across 16046 samples within 1 assay 
Active assay: originalexp (4000 features, 0 variable features)
 1 dimensional reduction calculated: PCA
</pre></div>
</div>
</div>
</div>
<p>Now we need to find anchors between the reference and the query. We specify that we want to use SPCA dimentionality reduction from the reference.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>%%R
anchors &lt;- FindTransferAnchors(
  reference = ref,
  query = query,
  reference.reduction = &quot;spca&quot;,
  dims = 1:20
)
</pre></div>
</div>
</div>
</div>
<p>Next, we map query onto the reference using the saved UMAP model to make sure that the UMAP projections stay the same for the reference and the query is mapped onto it. We additionally specify what information we want to transfer from the reference to the query with <code class="docutils literal notranslate"><span class="pre">refdata</span></code> parameter: in our case we want to predict the cell type and the protein counts.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>%%R
query &lt;- MapQuery(
  anchorset = anchors,
  query = query,
  reference = ref,
  refdata = list(
    cell_type = &quot;cell_type&quot;,
    predicted_ADT = &quot;ADT&quot;
  ),
  reference.reduction = &quot;spca&quot;,
  reduction.model = &quot;wnn.umap&quot;,
  verbose=FALSE
)
</pre></div>
</div>
</div>
</div>
<p>Let’s visualize the reference and the query together and take a look at predicted vs true cell labels.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>%%R
ref$id &lt;- &#39;reference&#39;
query$id &lt;- &#39;query&#39;
refquery &lt;-  merge(ref, query)
refquery[[&quot;umap&quot;]] &lt;- merge(ref[[&quot;wnn.umap&quot;]], query[[&quot;ref.umap&quot;]])
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>%%R
p1 &lt;- DimPlot(refquery, reduction = &quot;umap&quot;, group.by = &quot;cell_type&quot;, label = TRUE, label.size = 3, repel = TRUE) + NoLegend()
p2 &lt;- DimPlot(refquery, reduction = &quot;umap&quot;, group.by = &quot;predicted.cell_type&quot;, label = TRUE, label.size = 3, repel = TRUE) + NoLegend()
p1 + p2
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Fontconfig warning: ignoring UTF-8: not a valid region tag
</pre></div>
</div>
<img alt="../_images/advanced_integration_97_1.png" src="../_images/advanced_integration_97_1.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>%%R
DimPlot(refquery, reduction = &quot;umap&quot;, group.by = &quot;id&quot;, label = TRUE, label.size = 3, repel = TRUE) + NoLegend()
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/advanced_integration_98_0.png" src="../_images/advanced_integration_98_0.png" />
</div>
</div>
<p><a id='bridge'></a></p>
</section>
<section id="mapping-adt-onto-an-rna-atlas-with-bridge-integration">
<h2><span class="section-number">38.7. </span>Mapping ADT onto an RNA atlas with bridge integration<a class="headerlink" href="#mapping-adt-onto-an-rna-atlas-with-bridge-integration" title="Permalink to this headline">#</a></h2>
<p>Bridge <span id="id5">[<a class="reference internal" href="#id447" title="Yuhan Hao, Tim Stuart, Madeline Kowalski, Saket Choudhary, Paul Hoffman, Austin Hartman, Avi Srivastava, Gesmira Molla, Shaista Madad, Carlos Fernandez-Granda, and Rahul Satija. Dictionary learning for integrative, multimodal, and scalable single-cell analysis. bioRxiv, 2022. URL: https://www.biorxiv.org/content/early/2022/02/26/2022.02.24.481684, arXiv:https://www.biorxiv.org/content/early/2022/02/26/2022.02.24.481684.full.pdf, doi:10.1101/2022.02.24.481684.">Hao <em>et al.</em>, 2022</a>]</span> is a method for mapping new modalities onto an RNA-seq reference. To do so we need a so-called “bridge” dataset that contains paired data (RNA-seq and the new modality we want to map).</p>
<p>Since we want to showcase the bridging functionality, we assume that our reference consist of only one batch. We refer the reader to the integration vignette if there is a need to integrate batches in the reference (<a class="reference external" href="https://satijalab.org/seurat/articles/integration_introduction.html">https://satijalab.org/seurat/articles/integration_introduction.html</a>). As the bridge dataset we use one batch of the CITE-seq query, and as the actual ADT query we use ADT modality of the second batch in the CITE-seq query.</p>
<p>First, we subset the datasets to reference, bridge and query as described above.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>rna_cite_ref = rna_cite[rna_cite.obs[&quot;batch&quot;] == &quot;s1d1&quot;].copy()
rna_cite_bridge = rna_cite_query[rna_cite_query.obs[&quot;batch&quot;] == &quot;s2d1&quot;].copy()
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adt_cite_bridge = adt_cite_query[adt_cite_query.obs[&quot;donor&quot;] == &quot;s2d1&quot;].copy()
adt_cite_query = adt_cite_query[adt_cite_query.obs[&quot;donor&quot;] == &quot;s2d4&quot;].copy()
</pre></div>
</div>
</div>
</div>
<p>Next, we prepare the reference. We move the object from Python to R, normalize the data with SCTransform and calculate PCA. We also calculate the UMAP (and keep the model) that we will later use to map query onto.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata_ = sc.AnnData(rna_cite_ref.layers[&quot;counts&quot;].A.copy())
adata_.obs_names = rna_cite_ref.obs_names.copy()
adata_.var_names = rna_cite_ref.var_names.copy()
adata_.obs[&quot;cell_type&quot;] = rna_cite_ref.obs[&quot;cell_type&quot;].copy()
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>%%R -i adata_
ref = as.Seurat(adata_, data=NULL, counts=&#39;X&#39;)
ref
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>An object of class Seurat 
4000 features across 5219 samples within 1 assay 
Active assay: originalexp (4000 features, 0 variable features)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>%%R
ref &lt;- RenameAssays(object = ref, originalexp = &quot;RNA&quot;, verbose=FALSE) 
ref &lt;- SCTransform(ref, verbose=FALSE)
ref &lt;- RunPCA(ref, verbose=FALSE)
ref &lt;- RunUMAP(ref, dims = 1:50, return.model=TRUE, verbose=FALSE)
</pre></div>
</div>
</div>
</div>
<p>Next, we prepare the bridge dataset. We move both RNA and ADT objects to R, follow the same normalization procedure as above for RNA, perform CLR-normalization and calculate PCA for ADT counts.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata_ = sc.AnnData(adt_cite_bridge.layers[&quot;counts&quot;].A.copy())
adata_.obs_names = adt_cite_bridge.obs_names.copy()
adata_.var_names = adt_cite_bridge.var_names.copy()
adata_.obs[&quot;cell_type&quot;] = adt_cite_bridge.obs[&quot;cell_type&quot;].copy()
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata_
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 10465 × 136
    obs: &#39;cell_type&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>%%R -i adata_
adt &lt;- as.Seurat(adata_, data=NULL, counts=&#39;X&#39;)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata_ = sc.AnnData(rna_cite_bridge.layers[&quot;counts&quot;].A.copy())
adata_.obs_names = rna_cite_bridge.obs_names.copy()
adata_.var_names = rna_cite_bridge.var_names.copy()
adata_.obs[&quot;cell_type&quot;] = rna_cite_bridge.obs[&quot;cell_type&quot;].copy()
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>%%R -i adata_
rna = as.Seurat(adata_, data=NULL, counts=&#39;X&#39;)
cite &lt;- rna
cite &lt;- RenameAssays(object = cite, originalexp = &quot;RNA&quot;) 
cite[[&quot;ADT&quot;]] &lt;- CreateAssayObject(counts = adt@assays$originalexp@data)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>%%R
DefaultAssay(cite) &lt;- &quot;RNA&quot;
VariableFeatures(cite) &lt;- rownames(cite)
cite &lt;- SCTransform(cite, verbose = FALSE) 
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>%%R
DefaultAssay(cite) &lt;- &quot;ADT&quot;
VariableFeatures(cite) &lt;- rownames(cite)
cite &lt;- NormalizeData(cite, normalization.method = &#39;CLR&#39;, margin = 2, verbose=FALSE)
cite &lt;- ScaleData(cite, verbose=FALSE)
</pre></div>
</div>
</div>
</div>
<p>Finally, we prepare the ADT query the same way as the bridge ADT above.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata_ = sc.AnnData(adt_cite_query.layers[&quot;counts&quot;].A.copy())
adata_.obs_names = adt_cite_query.obs_names.copy()
adata_.var_names = adt_cite_query.var_names.copy()
adata_.obs[&quot;cell_type&quot;] = adt_cite_query.obs[&quot;cell_type&quot;].copy()
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>%%R -i adata_
query &lt;- as.Seurat(adata_, data=NULL, counts=&#39;X&#39;)
query &lt;- RenameAssays(object = query, originalexp = &quot;ADT&quot;, verbose=FALSE) 
VariableFeatures(query) &lt;- rownames(query)
query &lt;- NormalizeData(query, normalization.method = &#39;CLR&#39;, margin = 2, verbose=FALSE)
query &lt;- ScaleData(query, verbose=FALSE)
query &lt;- RunPCA(query, verbose=FALSE, reduction.name = &#39;apca&#39;)
</pre></div>
</div>
</div>
</div>
<p>Now we can perform the bridge integration. The final step of query mapping is the same as with Seurat v4 query mapping so we can also predict cell types for the query data.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>%%R
dims.adt &lt;- 1:50
dims.rna &lt;- 1:50

DefaultAssay(cite) &lt;-  &quot;RNA&quot;
DefaultAssay(ref) &lt;- &quot;SCT&quot;
obj.rna.ext &lt;- PrepareBridgeReference(
    reference = ref,
    bridge = cite,
    bridge.query.assay = &quot;ADT&quot;,
    reference.reduction = &quot;pca&quot;,
    reference.dims = dims.rna,
    normalization.method = &quot;SCT&quot;,
    verbose = FALSE
)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>  |                                                  | 0 % ~calculating   |++++++++++++++++++++++++++++++++++++++++++++++++++| 100% elapsed=00s  
  |                                                  | 0 % ~calculating   |++++++++++++++++++++++++++++++++++++++++++++++++++| 100% elapsed=00s  
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>%%R
bridge.anchor &lt;- FindBridgeTransferAnchors(
    extended.reference = obj.rna.ext, 
    query = query,
    reduction = &quot;pcaproject&quot;,
    dims = dims.adt,
    verbose = FALSE
)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>  |                                                  | 0 % ~calculating   |++++++++++++++++++++++++++++++++++++++++++++++++++| 100% elapsed=00s  
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>%%R
query &lt;- MapQuery(
    anchorset = bridge.anchor, 
    reference = ref, 
    query = query, 
    refdata = list(
        cell_type = &quot;cell_type&quot;
    ),
    reduction.model = &quot;umap&quot;,
    verbose = FALSE
)
</pre></div>
</div>
</div>
</div>
<p>Let’s visualize the result.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>%%R
p1 &lt;- DimPlot(query, group.by = &quot;predicted.cell_type&quot;, reduction = &quot;ref.umap&quot;, label = TRUE) + NoLegend()
p2 &lt;- DimPlot(query, group.by = &quot;cell_type&quot;, reduction = &quot;ref.umap&quot;, label = TRUE) + NoLegend()
p1 + p2
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/advanced_integration_122_0.png" src="../_images/advanced_integration_122_0.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>%%R
ref$id &lt;- &#39;reference&#39;
query$id &lt;- &#39;query&#39;
refquery &lt;-  merge(ref, query)
refquery[[&quot;umap&quot;]] &lt;- merge(ref[[&quot;umap&quot;]], query[[&quot;ref.umap&quot;]])
DimPlot(refquery, group.by = &#39;id&#39;, shuffle = TRUE)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/advanced_integration_123_0.png" src="../_images/advanced_integration_123_0.png" />
</div>
</div>
<p><a id='multigrate'></a></p>
</section>
<section id="trimodal-integration-and-query-to-reference-mapping-with-multigrate">
<h2><span class="section-number">38.8. </span>Trimodal integration and query-to-reference mapping with multigrate<a class="headerlink" href="#trimodal-integration-and-query-to-reference-mapping-with-multigrate" title="Permalink to this headline">#</a></h2>
<p>In our final example, we will build a trimodal reference atlas using RNA, ATAC and ADT as three modalities. Multigrate <span id="id6">[]</span> is a deep-learning method based on conditional autoencoders. Data from each modality is fed into a separate encoder which outputs the parameters of the marginal distribution, then the parameters of the joint distribution are calculated with product of experts (PoE) <span id="id7">[]</span>. The conditional decoders then learn to reconstruct the original input data while simultaneously correction for batch effects. PoE allows the integration of data with missing measurements for some modalities by setting the corresponding marginals to a non-informative distribution. Multigrate also follows scArches <span id="id8">[]</span> framework to allow query-to-reference mapping of unimodal as well as multimodal queries.</p>
<section id="building-the-reference">
<h3><span class="section-number">38.8.1. </span>Building the reference<a class="headerlink" href="#building-the-reference" title="Permalink to this headline">#</a></h3>
<p>First, we load data that has 4000 highly variable genes that are common for the scRNA-seq and scRNA-seq from the CITE-seq and the multiome experiment respectively.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>rna1 = sc.read(
    &quot;/lustre/groups/ml01/workspace/anastasia.litinetskaya/data/trimodal_neurips/rna_hvg_cite.h5ad&quot;
)
rna1_ref = rna1[adt_cite.obs_names].copy()
rna1_query = rna1[adt_cite_query.obs_names].copy()
rna1_ref.shape, rna1_query.shape
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>((16294, 4000), (16046, 4000))
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>rna2 = sc.read(
    &quot;/lustre/groups/ml01/workspace/anastasia.litinetskaya/data/trimodal_neurips/rna_hvg_multiome.h5ad&quot;
)
rna2_ref = rna2[atac_multiome.obs_names].copy()
rna2_query = rna2[atac_multiome_query.obs_names].copy()
rna2_ref.shape, rna2_query.shape
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>((17243, 4000), (10331, 4000))
</pre></div>
</div>
</div>
</div>
<p>Next, we concatenate all the data into one AnnData object specifying how the data is paired and which layers to use.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata = mtg.data.organize_multiome_anndatas(
    adatas=[[rna1_ref, rna2_ref], [None, atac_multiome], [adt_cite, None]],
    layers=[[&quot;counts&quot;, &quot;counts&quot;], [None, &quot;cpm&quot;], [None, None]],
)
adata
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 33537 × 24136
    obs: &#39;GEX_n_genes_by_counts&#39;, &#39;GEX_pct_counts_mt&#39;, &#39;GEX_size_factors&#39;, &#39;GEX_phase&#39;, &#39;ADT_n_antibodies_by_counts&#39;, &#39;ADT_total_counts&#39;, &#39;ADT_iso_count&#39;, &#39;cell_type&#39;, &#39;batch&#39;, &#39;ADT_pseudotime_order&#39;, &#39;GEX_pseudotime_order&#39;, &#39;Samplename&#39;, &#39;Site&#39;, &#39;DonorNumber&#39;, &#39;Modality&#39;, &#39;VendorLot&#39;, &#39;DonorID&#39;, &#39;DonorAge&#39;, &#39;DonorBMI&#39;, &#39;DonorBloodType&#39;, &#39;DonorRace&#39;, &#39;Ethnicity&#39;, &#39;DonorGender&#39;, &#39;QCMeds&#39;, &#39;DonorSmoker&#39;, &#39;is_train&#39;, &#39;GEX_n_counts&#39;, &#39;GEX_n_genes&#39;, &#39;ATAC_nCount_peaks&#39;, &#39;ATAC_atac_fragments&#39;, &#39;ATAC_reads_in_peaks_frac&#39;, &#39;ATAC_blacklist_fraction&#39;, &#39;ATAC_nucleosome_signal&#39;, &#39;ATAC_pseudotime_order&#39;, &#39;technology&#39;, &#39;cell_type_l2&#39;, &#39;cell_type_l1&#39;, &#39;cell_type_l3&#39;, &#39;assay&#39;, &#39;split&#39;, &#39;group&#39;, &#39;donor&#39;, &#39;log1p_n_genes_by_counts&#39;, &#39;log1p_total_counts&#39;, &#39;n_counts&#39;, &#39;n_genes_by_counts&#39;, &#39;new_donor&#39;, &#39;outliers&#39;, &#39;total_counts&#39;, &#39;new_batch&#39;
    var: &#39;modality&#39;
    uns: &#39;modality_lengths&#39;
    layers: &#39;counts&#39;
</pre></div>
</div>
</div>
</div>
<p>Next, we register covariates that the model will correct for in the latent space.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>mtg.model.MultiVAE.setup_anndata(
    adata,
    categorical_covariate_keys=[&quot;Modality&quot;, &quot;Samplename&quot;],
    rna_indices_end=4000,
)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>WARNING:jax._src.lib.xla_bridge:No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
</pre></div>
</div>
</div>
</div>
<p>Next, we initialize the model and train it.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>model = mtg.model.MultiVAE(
    adata,
    losses=[&quot;nb&quot;, &quot;mse&quot;, &quot;mse&quot;],
    loss_coefs={
        &quot;kl&quot;: 1e-3,
        &quot;integ&quot;: 10000,
    },
    integrate_on=&quot;Modality&quot;,
    mmd=&quot;marginal&quot;,
)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>model.train()
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs
/lustre/groups/ml01/workspace/anastasia.litinetskaya/miniconda3/envs/multigrate/lib/python3.8/site-packages/pytorch_lightning/trainer/configuration_validator.py:267: LightningDeprecationWarning: The `Callback.on_epoch_end` hook was deprecated in v1.6 and will be removed in v1.8. Please use `Callback.on_&lt;train/validation/test&gt;_epoch_end` instead.
  rank_zero_deprecation(
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 200/200: 100%|██████████| 200/200 [32:36&lt;00:00,  8.96s/it, loss=2.37e+03, v_num=1]
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>`Trainer.fit` stopped: `max_epochs=200` reached.
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 200/200: 100%|██████████| 200/200 [32:36&lt;00:00,  9.78s/it, loss=2.37e+03, v_num=1]
</pre></div>
</div>
</div>
</div>
<p>Finally, we obtain the latent representation, save in explicitely in <code class="docutils literal notranslate"><span class="pre">.obsm['latent_ref']</span></code> as we will overwrite <code class="docutils literal notranslate"><span class="pre">.obsm['latent']</span></code> later when we work with fine-tuned model after query mapping and visualize the result.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>model.get_latent_representation()
adata.obsm[&quot;latent_ref&quot;] = adata.obsm[&quot;latent&quot;].copy()
adata
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 33537 × 24136
    obs: &#39;GEX_n_genes_by_counts&#39;, &#39;GEX_pct_counts_mt&#39;, &#39;GEX_size_factors&#39;, &#39;GEX_phase&#39;, &#39;ADT_n_antibodies_by_counts&#39;, &#39;ADT_total_counts&#39;, &#39;ADT_iso_count&#39;, &#39;cell_type&#39;, &#39;batch&#39;, &#39;ADT_pseudotime_order&#39;, &#39;GEX_pseudotime_order&#39;, &#39;Samplename&#39;, &#39;Site&#39;, &#39;DonorNumber&#39;, &#39;Modality&#39;, &#39;VendorLot&#39;, &#39;DonorID&#39;, &#39;DonorAge&#39;, &#39;DonorBMI&#39;, &#39;DonorBloodType&#39;, &#39;DonorRace&#39;, &#39;Ethnicity&#39;, &#39;DonorGender&#39;, &#39;QCMeds&#39;, &#39;DonorSmoker&#39;, &#39;is_train&#39;, &#39;GEX_n_counts&#39;, &#39;GEX_n_genes&#39;, &#39;ATAC_nCount_peaks&#39;, &#39;ATAC_atac_fragments&#39;, &#39;ATAC_reads_in_peaks_frac&#39;, &#39;ATAC_blacklist_fraction&#39;, &#39;ATAC_nucleosome_signal&#39;, &#39;ATAC_pseudotime_order&#39;, &#39;technology&#39;, &#39;cell_type_l2&#39;, &#39;cell_type_l1&#39;, &#39;cell_type_l3&#39;, &#39;assay&#39;, &#39;split&#39;, &#39;group&#39;, &#39;donor&#39;, &#39;log1p_n_genes_by_counts&#39;, &#39;log1p_total_counts&#39;, &#39;n_counts&#39;, &#39;n_genes_by_counts&#39;, &#39;new_donor&#39;, &#39;outliers&#39;, &#39;total_counts&#39;, &#39;new_batch&#39;, &#39;size_factors&#39;, &#39;_scvi_batch&#39;
    var: &#39;modality&#39;
    uns: &#39;modality_lengths&#39;, &#39;_scvi_uuid&#39;, &#39;_scvi_manager_uuid&#39;
    obsm: &#39;_scvi_extra_categorical_covs&#39;, &#39;latent&#39;, &#39;latent_ref&#39;
    layers: &#39;counts&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pp.neighbors(adata, use_rep=&quot;latent&quot;)
sc.tl.umap(adata)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pl.umap(adata, color=[&quot;cell_type&quot;, &quot;Modality&quot;, &quot;Samplename&quot;], ncols=1, frameon=False)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/advanced_integration_138_0.png" src="../_images/advanced_integration_138_0.png" />
</div>
</div>
</section>
<section id="querying-the-trimodal-reference">
<h3><span class="section-number">38.8.2. </span>Querying the trimodal reference<a class="headerlink" href="#querying-the-trimodal-reference" title="Permalink to this headline">#</a></h3>
<p>We repeat the same steps for the setting up the query as for the reference before.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>query = mtg.data.organize_multiome_anndatas(
    adatas=[
        [rna1_query, rna2_query],
        [None, atac_multiome_query],
        [adt_cite_query, None],
    ],
    layers=[[&quot;counts&quot;, &quot;counts&quot;], [None, &quot;cpm&quot;], [None, None]],
)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>mtg.model.MultiVAE.setup_anndata(
    query,
    categorical_covariate_keys=[&quot;Modality&quot;, &quot;Samplename&quot;],
    rna_indices_end=4000,
)
</pre></div>
</div>
</div>
</div>
<p>Let’s imitate unimodal queries by masking with zeros the snRNA part of one multiome batch and ADT part of one CITE-seq batch in the query. We get one ATAC-only query batch and one scRNA-only query batch respectively.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>idx_atac_query = query.obs[&quot;Samplename&quot;] == &quot;site2_donor4_multiome&quot;
idx_scrna_query = query.obs[&quot;Samplename&quot;] == &quot;site2_donor1_cite&quot;

idx_mutiome_query = query.obs[&quot;Samplename&quot;] == &quot;site2_donor1_multiome&quot;
idx_cite_query = query.obs[&quot;Samplename&quot;] == &quot;site2_donor4_cite&quot;

np.sum(idx_atac_query), np.sum(idx_scrna_query), np.sum(idx_mutiome_query), np.sum(
    idx_cite_query
),
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(6111, 10465, 4220, 5581)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>query[idx_atac_query, :4000].X = 0
query[idx_scrna_query, 4000:].X = 0
</pre></div>
</div>
</div>
</div>
<p>We update the model by adding new weights to the new batches in the query and fine-tune those weights.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>q_model = mtg.model.MultiVAE.load_query_data(query, model)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>q_model.train(weight_decay=0)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs
/lustre/groups/ml01/workspace/anastasia.litinetskaya/miniconda3/envs/multigrate/lib/python3.8/site-packages/pytorch_lightning/trainer/configuration_validator.py:267: LightningDeprecationWarning: The `Callback.on_epoch_end` hook was deprecated in v1.6 and will be removed in v1.8. Please use `Callback.on_&lt;train/validation/test&gt;_epoch_end` instead.
  rank_zero_deprecation(
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 185/200:  92%|█████████▎| 185/200 [18:05&lt;01:28,  5.87s/it, loss=2.13e+03, v_num=1]
Monitored metric reconstruction_loss_validation did not improve in the last 50 records. Best score: 1422.694. Signaling Trainer to stop.
</pre></div>
</div>
</div>
</div>
<p>We obtain the latent representation for query and the reference from the updated model. Note that the representation of the reference is the same as before up to some sampling noise.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>q_model.get_latent_representation(adata=query)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>q_model.get_latent_representation(adata=adata)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span><span class=" -Color -Color-Blue">INFO    </span> Input AnnData not setup with scvi-tools. attempting to transfer AnnData setup                             
</pre></div>
</div>
</div>
</div>
<p>Finally, we concatenate the reference and the query, and visualize both on a UMAP.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata.obs[&quot;reference&quot;] = &quot;reference&quot;
query.obs[&quot;reference&quot;] = &quot;query&quot;

adata.obs[&quot;type_of_query&quot;] = &quot;reference&quot;
query.obs.loc[idx_atac_query, &quot;type_of_query&quot;] = &quot;ATAC query&quot;
query.obs.loc[idx_scrna_query, &quot;type_of_query&quot;] = &quot;scRNA query&quot;
query.obs.loc[idx_mutiome_query, &quot;type_of_query&quot;] = &quot;multiome query&quot;
query.obs.loc[idx_cite_query, &quot;type_of_query&quot;] = &quot;CITE-seq query&quot;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata_both = anndata.concat([adata, query])
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pp.neighbors(adata_both, use_rep=&quot;latent&quot;)
sc.tl.umap(adata_both)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pl.umap(
    adata_both,
    color=[&quot;cell_type&quot;, &quot;Modality&quot;, &quot;Samplename&quot;, &quot;reference&quot;],
    ncols=1,
    frameon=False,
)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/advanced_integration_155_0.png" src="../_images/advanced_integration_155_0.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pl.umap(
    adata_both, color=&quot;type_of_query&quot;, ncols=1, frameon=False, groups=[&quot;ATAC query&quot;]
)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/advanced_integration_156_0.png" src="../_images/advanced_integration_156_0.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pl.umap(
    adata_both, color=&quot;type_of_query&quot;, ncols=1, frameon=False, groups=[&quot;CITE-seq query&quot;]
)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/advanced_integration_157_0.png" src="../_images/advanced_integration_157_0.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pl.umap(
    adata_both, color=&quot;type_of_query&quot;, ncols=1, frameon=False, groups=[&quot;multiome query&quot;]
)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/advanced_integration_158_0.png" src="../_images/advanced_integration_158_0.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pl.umap(
    adata_both, color=&quot;type_of_query&quot;, ncols=1, frameon=False, groups=[&quot;scRNA query&quot;]
)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/advanced_integration_159_0.png" src="../_images/advanced_integration_159_0.png" />
</div>
</div>
<p>We observe that the query cell types are mapped to the corresponding cell types in the reference. We also note that unimodal query mapping is possible with Multigrate even though there was no unimodal data in the reference.</p>
</section>
</section>
<section id="session-info">
<h2><span class="section-number">38.9. </span>Session info<a class="headerlink" href="#session-info" title="Permalink to this headline">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>%%R
sessionInfo()
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>R version 4.1.3 (2022-03-10)
Platform: x86_64-conda-linux-gnu (64-bit)
Running under: CentOS Linux 7 (Core)

Matrix products: default
BLAS/LAPACK: /lustre/groups/ml01/workspace/anastasia.litinetskaya/miniconda3/envs/bridge/lib/libopenblasp-r0.3.20.so

locale:
 [1] LC_CTYPE=C                 LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats4    tools     stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] sp_1.5-0                    SeuratObject_4.1.0         
 [3] Seurat_4.1.1.9001           SingleCellExperiment_1.16.0
 [5] SummarizedExperiment_1.24.0 Biobase_2.54.0             
 [7] GenomicRanges_1.46.1        GenomeInfoDb_1.30.0        
 [9] IRanges_2.28.0              S4Vectors_0.32.3           
[11] BiocGenerics_0.40.0         MatrixGenerics_1.6.0       
[13] matrixStats_0.62.0         

loaded via a namespace (and not attached):
  [1] Rtsne_0.16             colorspace_2.0-3       deldir_1.0-6          
  [4] ellipsis_0.3.2         ggridges_0.5.3         XVector_0.34.0        
  [7] RcppHNSW_0.4.1         spatstat.data_2.2-0    leiden_0.4.2          
 [10] listenv_0.8.0          ggrepel_0.9.1          RSpectra_0.16-1       
 [13] fansi_1.0.3            codetools_0.2-18       splines_4.1.3         
 [16] polyclip_1.10-0        jsonlite_1.8.0         ica_1.0-3             
 [19] cluster_2.1.3          png_0.1-7              rgeos_0.5-9           
 [22] uwot_0.1.11            spatstat.sparse_2.1-1  shiny_1.7.2           
 [25] sctransform_0.3.3      compiler_4.1.3         httr_1.4.3            
 [28] assertthat_0.2.1       Matrix_1.4-1           fastmap_1.1.0         
 [31] lazyeval_0.2.2         cli_3.3.0              later_1.2.0           
 [34] htmltools_0.5.3        igraph_1.3.1           gtable_0.3.0          
 [37] glue_1.6.2             GenomeInfoDbData_1.2.7 RANN_2.6.1            
 [40] reshape2_1.4.4         dplyr_1.0.9            Rcpp_1.0.9            
 [43] scattermore_0.8        vctrs_0.4.1            nlme_3.1-158          
 [46] progressr_0.10.1       lmtest_0.9-40          spatstat.random_2.2-0 
 [49] stringr_1.4.0          globals_0.15.1         mime_0.12             
 [52] miniUI_0.1.1.1         lifecycle_1.0.1        irlba_2.3.5           
 [55] goftest_1.2-3          future_1.27.0          zlibbioc_1.40.0       
 [58] MASS_7.3-58            zoo_1.8-10             scales_1.2.0          
 [61] spatstat.core_2.4-4    spatstat.utils_2.3-1   promises_1.2.0.1      
 [64] parallel_4.1.3         RColorBrewer_1.1-3     reticulate_1.25       
 [67] pbapply_1.5-0          gridExtra_2.3          ggplot2_3.3.6         
 [70] rpart_4.1.16           stringi_1.7.8          fastDummies_1.6.3     
 [73] rlang_1.0.4            pkgconfig_2.0.3        bitops_1.0-7          
 [76] lattice_0.20-45        tensor_1.5             ROCR_1.0-11           
 [79] purrr_0.3.4            patchwork_1.1.1        htmlwidgets_1.5.4     
 [82] cowplot_1.1.1          tidyselect_1.1.2       parallelly_1.32.1     
 [85] RcppAnnoy_0.0.19       plyr_1.8.7             magrittr_2.0.3        
 [88] R6_2.5.1               generics_0.1.3         DelayedArray_0.20.0   
 [91] DBI_1.1.3              mgcv_1.8-40            pillar_1.8.0          
 [94] fitdistrplus_1.1-8     abind_1.4-5            survival_3.3-1        
 [97] RCurl_1.98-1.7         tibble_3.1.8           future.apply_1.9.0    
[100] KernSmooth_2.23-20     utf8_1.2.2             spatstat.geom_2.4-0   
[103] plotly_4.10.0          grid_4.1.3             data.table_1.14.2     
[106] digest_0.6.29          xtable_1.8-4           tidyr_1.2.0           
[109] httpuv_1.6.5           munsell_0.5.0          viridisLite_0.4.0     
</pre></div>
</div>
</div>
</div>
</section>
<section id="references">
<h2><span class="section-number">38.10. </span>References<a class="headerlink" href="#references" title="Permalink to this headline">#</a></h2>
<div class="docutils container" id="id9">
<dl class="citation">
<dt class="label" id="id446"><span class="brackets"><a class="fn-backref" href="#id2">CG22</a></span></dt>
<dd><p>Zhi-Jie Cao and Ge Gao. Multi-omics single-cell data integration and regulatory inference with graph-linked embedding. <em>Nature Biotechnology</em>, May 2022. URL: <a class="reference external" href="https://doi.org/10.1038/s41587-022-01284-4">https://doi.org/10.1038/s41587-022-01284-4</a>, <a class="reference external" href="https://doi.org/10.1038/s41587-022-01284-4">doi:10.1038/s41587-022-01284-4</a>.</p>
</dd>
<dt class="label" id="id438"><span class="brackets"><a class="fn-backref" href="#id3">GSL+21</a></span></dt>
<dd><p>Adam Gayoso, Zoë Steier, Romain Lopez, Jeffrey Regier, Kristopher L. Nazor, Aaron Streets, and Nir Yosef. Joint probabilistic modeling of single-cell multi-omic data with totalvi. <em>Nature Methods</em>, 18(3):272–282, Mar 2021. URL: <a class="reference external" href="https://doi.org/10.1038/s41592-020-01050-x">https://doi.org/10.1038/s41592-020-01050-x</a>, <a class="reference external" href="https://doi.org/10.1038/s41592-020-01050-x">doi:10.1038/s41592-020-01050-x</a>.</p>
</dd>
<dt class="label" id="id436"><span class="brackets"><a class="fn-backref" href="#id4">HHAN+21</a></span></dt>
<dd><p>Yuhan Hao, Stephanie Hao, Erica Andersen-Nissen, William M. Mauck, Shiwei Zheng, Andrew Butler, Maddie J. Lee, Aaron J. Wilk, Charlotte Darby, Michael Zager, Paul Hoffman, Marlon Stoeckius, Efthymia Papalexi, Eleni P. Mimitou, Jaison Jain, Avi Srivastava, Tim Stuart, Lamar M. Fleming, Bertrand Yeung, Angela J. Rogers, Juliana M. McElrath, Catherine A. Blish, Raphael Gottardo, Peter Smibert, and Rahul Satija. Integrated analysis of multimodal single-cell data. <em>Cell</em>, 184(13):3573–3587.e29, 2021. URL: <a class="reference external" href="https://www.sciencedirect.com/science/article/pii/S0092867421005833">https://www.sciencedirect.com/science/article/pii/S0092867421005833</a>, <a class="reference external" href="https://doi.org/https://doi.org/10.1016/j.cell.2021.04.048">doi:https://doi.org/10.1016/j.cell.2021.04.048</a>.</p>
</dd>
<dt class="label" id="id447"><span class="brackets"><a class="fn-backref" href="#id5">HSK+22</a></span></dt>
<dd><p>Yuhan Hao, Tim Stuart, Madeline Kowalski, Saket Choudhary, Paul Hoffman, Austin Hartman, Avi Srivastava, Gesmira Molla, Shaista Madad, Carlos Fernandez-Granda, and Rahul Satija. Dictionary learning for integrative, multimodal, and scalable single-cell analysis. <em>bioRxiv</em>, 2022. URL: <a class="reference external" href="https://www.biorxiv.org/content/early/2022/02/26/2022.02.24.481684">https://www.biorxiv.org/content/early/2022/02/26/2022.02.24.481684</a>, <a class="reference external" href="https://arxiv.org/abs/https://www.biorxiv.org/content/early/2022/02/26/2022.02.24.481684.full.pdf">arXiv:https://www.biorxiv.org/content/early/2022/02/26/2022.02.24.481684.full.pdf</a>, <a class="reference external" href="https://doi.org/10.1101/2022.02.24.481684">doi:10.1101/2022.02.24.481684</a>.</p>
</dd>
<dt class="label" id="id445"><span class="brackets"><a class="fn-backref" href="#id1">LBC+21</a></span></dt>
<dd><p>Malte D Luecken, Daniel Bernard Burkhardt, Robrecht Cannoodt, Christopher Lance, Aditi Agrawal, Hananeh Aliee, Ann T Chen, Louise Deconinck, Angela M Detweiler, Alejandro A Granados, Shelly Huynh, Laura Isacco, Yang Joon Kim, Dominik Klein, BONY DE KUMAR, Sunil Kuppasani, Heiko Lickert, Aaron McGeever, Honey Mekonen, Joaquin Caceres Melgarejo, Maurizio Morri, Michaela Müller, Norma Neff, Sheryl Paul, Bastian Rieck, Kaylie Schneider, Scott Steelman, Michael Sterr, Daniel J. Treacy, Alexander Tong, Alexandra-Chloe Villani, Guilin Wang, Jia Yan, Ce Zhang, Angela Oliveira Pisco, Smita Krishnaswamy, Fabian J Theis, and Jonathan M. Bloom. A sandbox for prediction and integration of dna, rna, and proteins in single cells. In <em>Thirty-fifth Conference on Neural Information Processing Systems Datasets and Benchmarks Track (Round 2)</em>. 2021. URL: <a class="reference external" href="https://openreview.net/forum?id=gN35BGa1Rt">https://openreview.net/forum?id=gN35BGa1Rt</a>.</p>
</dd>
</dl>
</div>
</section>
<section id="contributors">
<h2><span class="section-number">38.11. </span>Contributors<a class="headerlink" href="#contributors" title="Permalink to this headline">#</a></h2>
<p>We gratefully acknowledge the contributions of:</p>
<section id="authors">
<h3><span class="section-number">38.11.1. </span>Authors<a class="headerlink" href="#authors" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p>Anastasia Litinetskaya</p></li>
</ul>
</section>
<section id="reviewers">
<h3><span class="section-number">38.11.2. </span>Reviewers<a class="headerlink" href="#reviewers" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p>Lukas Heumos</p></li>
</ul>
</section>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./multimodal_integration"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
    <a class='left-prev' id="prev-link" href="paired_integration.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title"><span class="section-number">37. </span>Paired integration</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="../reproducibility/introduction.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title"><span class="section-number">39. </span>Reproducibility</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By Lukas Heumos, Anna Schaar<br/>
  
      &copy; Copyright 2022.<br/>
    <div class="extra_footer">
      <div>
  Brought to you by
  <a href="https://www.github.com/theislab">Theislab</a>, with many thanks to the single-cell community as a whole!
</div>

    </div>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>